// ngx-tethys/esm2015/skeleton/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PortalModule } from '@angular/cdk/portal';
import { ThySkeletonComponent } from './skeleton.component';
import { ThySkeletonListComponent } from './stylized/list.component';
import { ThySkeletonBulletListComponent } from './stylized/bullet-list.component';
import { ThySkeletonTitleComponent } from './stylized/title.component';
import { ThySkeletonAvatarComponent } from './stylized/avatar.component';
import { ThySkeletonParagraphComponent } from './stylized/paragraph.component';
import * as ɵngcc0 from '@angular/core';
export class ThySkeletonModule {
}
ThySkeletonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySkeletonModule });
ThySkeletonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySkeletonModule_Factory(t) { return new (t || ThySkeletonModule)(); }, providers: [], imports: [[CommonModule, PortalModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySkeletonModule, { declarations: function () { return [ThySkeletonComponent,
        ThySkeletonListComponent,
        ThySkeletonBulletListComponent,
        ThySkeletonAvatarComponent,
        ThySkeletonTitleComponent,
        ThySkeletonParagraphComponent]; }, imports: function () { return [CommonModule, PortalModule]; }, exports: function () { return [ThySkeletonComponent,
        ThySkeletonListComponent,
        ThySkeletonBulletListComponent,
        ThySkeletonTitleComponent,
        ThySkeletonAvatarComponent,
        ThySkeletonParagraphComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThySkeletonComponent,
                    ThySkeletonListComponent,
                    ThySkeletonBulletListComponent,
                    ThySkeletonAvatarComponent,
                    ThySkeletonTitleComponent,
                    ThySkeletonParagraphComponent
                ],
                imports: [CommonModule, PortalModule],
                entryComponents: [ThySkeletonComponent],
                exports: [
                    ThySkeletonComponent,
                    ThySkeletonListComponent,
                    ThySkeletonBulletListComponent,
                    ThySkeletonTitleComponent,
                    ThySkeletonAvatarComponent,
                    ThySkeletonParagraphComponent
                ],
                providers: []
            }]
    }], null, null); })();
