// ngx-tethys/esm2015/skeleton/skeleton.component.js
import { Component, Input, Inject, PLATFORM_ID, HostBinding, TemplateRef, ContentChild } from '@angular/core';
import { generateRandomStr } from 'ngx-tethys/util';
import { isPlatformBrowser } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["content"];
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1__svg_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const content_r8 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", content_r8);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1_Template, 2, 1, "ng-container", 7);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.contentTemplates);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_5__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_5__svg_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r3.customTemplate);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r4.animationValues[0])("dur", ctx_r4.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r5.animationValues[1])("dur", ctx_r5.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r6.animationValues[2])("dur", ctx_r6.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 2);
    ɵngcc0.ɵɵelement(1, "rect", 3);
    ɵngcc0.ɵɵelementStart(2, "defs");
    ɵngcc0.ɵɵelementStart(3, "clipPath");
    ɵngcc0.ɵɵtemplate(4, ThySkeletonComponent__svg_svg_0__svg_ng_container_4_Template, 2, 1, "ng-container", 1);
    ɵngcc0.ɵɵtemplate(5, ThySkeletonComponent__svg_svg_0__svg_ng_container_5_Template, 2, 1, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "linearGradient");
    ɵngcc0.ɵɵelementStart(7, "stop", 4);
    ɵngcc0.ɵɵtemplate(8, ThySkeletonComponent__svg_svg_0__svg_ng_container_8_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "stop", 5);
    ɵngcc0.ɵɵtemplate(10, ThySkeletonComponent__svg_svg_0__svg_ng_container_10_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "stop", 6);
    ɵngcc0.ɵɵtemplate(12, ThySkeletonComponent__svg_svg_0__svg_ng_container_12_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattributeInterpolate2("viewBox", "0 0 ", ctx_r0.thyViewBoxWidth, " ", ctx_r0.thyViewBoxHeight, "");
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r0.thyStyle);
    ɵngcc0.ɵɵattribute("preserveAspectRatio", ctx_r0.thyPreserveAspectRatio)("height", ctx_r0.thyHeight)("width", ctx_r0.thyWidth);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r0.fillStyle);
    ɵngcc0.ɵɵattribute("clip-path", ctx_r0.clipPath)("width", ctx_r0.thyWidth)("height", ctx_r0.thyHeight);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("id", ctx_r0.idClip);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.contentTemplates);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("id", ctx_r0.idGradient);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thyPrimaryColor)("stop-opacity", ctx_r0.thyPrimaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thySecondaryColor)("stop-opacity", ctx_r0.thySecondaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thyPrimaryColor)("stop-opacity", ctx_r0.thyPrimaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
} }
function ThySkeletonComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementContainerEnd();
} }
const _c1 = ["*"];
export class ThySkeletonComponent {
    constructor(platformId) {
        this.platformId = platformId;
        this.addSkeletonClass = true;
        this.thyAnimate = true;
        this.thyBaseUrl = '';
        this.thyWidth = '100%';
        this.thyHeight = '100%';
        this.thyViewBoxWidth = 400;
        this.thyViewBoxHeight = 130;
        this.thySpeed = 2;
        this.thyPreserveAspectRatio = 'none'; // xMidYMid meet
        this.thyPrimaryColor = '#f0f0f0';
        this.thySecondaryColor = '#e0e0e0';
        this.thyPrimaryOpacity = 1;
        this.thySecondaryOpacity = 1;
        this.thyIgnoreBaseUrl = false;
        this.thyLoadingDone = false;
        this.contentTemplates = [];
        this.idClip = generateRandomStr();
        this.idGradient = generateRandomStr();
        this.defaultAnimation = ['-3; 1', '-2; 2', '-1; 3'];
        this.rtlAnimation = ['1; -3', '2; -2', '3; -1'];
    }
    ngOnInit() {
        this.animationValues = this.thyRtl ? this.rtlAnimation : this.defaultAnimation;
        if (this.thyBaseUrl === '' && !this.thyIgnoreBaseUrl && isPlatformBrowser(this.platformId)) {
            this.thyBaseUrl = window.location.pathname;
        }
        this.setFillStyle();
        this.setClipPath();
    }
    ngOnChanges(changes) {
        if (changes['thyBaseUrl']) {
            if (changes['thyBaseUrl'].previousValue !== changes['thyBaseUrl'].currentValue) {
                this.setFillStyle();
                this.setClipPath();
            }
        }
    }
    setFillStyle() {
        this.fillStyle = {
            fill: `url(${this.thyBaseUrl}#${this.idGradient})`
        };
    }
    setClipPath() {
        this.clipPath = `url(${this.thyBaseUrl}#${this.idClip})`;
    }
    assignInputProperties(input) {
        this.thyAnimate = input.thyAnimate;
        this.thyBaseUrl = input.thyBaseUrl;
        this.thyHeight = input.thyHeight;
        this.thyWidth = input.thyWidth;
        this.thyViewBoxHeight = input.thyViewBoxHeight;
        this.thyViewBoxWidth = input.thyViewBoxWidth;
        this.thyPreserveAspectRatio = input.thyPreserveAspectRatio;
        this.thyPrimaryColor = input.thyPrimaryColor;
        this.thyPrimaryOpacity = input.thyPrimaryOpacity;
        this.thySecondaryColor = input.thySecondaryColor;
        this.thySecondaryOpacity = input.thySecondaryOpacity;
        this.thyRtl = this.thyRtl;
        this.thySpeed = this.thySpeed;
        this.thyUniqueKey = this.thyUniqueKey;
    }
    addTemplate(template) {
        if (template) {
            this.contentTemplates.push(template);
        }
    }
}
ThySkeletonComponent.ɵfac = function ThySkeletonComponent_Factory(t) { return new (t || ThySkeletonComponent)(ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID)); };
ThySkeletonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonComponent, selectors: [["thy-skeleton"]], contentQueries: function ThySkeletonComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customTemplate = _t.first);
    } }, hostVars: 2, hostBindings: function ThySkeletonComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-skeleton", ctx.addSkeletonClass);
    } }, inputs: { thyAnimate: "thyAnimate", thyBaseUrl: "thyBaseUrl", thyWidth: "thyWidth", thyHeight: "thyHeight", thyViewBoxWidth: "thyViewBoxWidth", thyViewBoxHeight: "thyViewBoxHeight", thySpeed: "thySpeed", thyPreserveAspectRatio: "thyPreserveAspectRatio", thyPrimaryColor: "thyPrimaryColor", thySecondaryColor: "thySecondaryColor", thyPrimaryOpacity: "thyPrimaryOpacity", thySecondaryOpacity: "thySecondaryOpacity", thyIgnoreBaseUrl: "thyIgnoreBaseUrl", thyLoadingDone: "thyLoadingDone", thyRtl: "thyRtl", thyUniqueKey: "thyUniqueKey", thyStyle: "thyStyle" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c1, decls: 2, vars: 2, consts: [[3, "ngStyle", 4, "ngIf"], [4, "ngIf"], [3, "ngStyle"], ["x", "0", "y", "0", 3, "ngStyle"], ["offset", "0%"], ["offset", "50%"], ["offset", "100%"], [4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["attributeName", "offset", "repeatCount", "indefinite"]], template: function ThySkeletonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThySkeletonComponent__svg_svg_0_Template, 13, 23, "svg", 0);
        ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent_ng_container_1_Template, 2, 0, "ng-container", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.thyLoadingDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyLoadingDone);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
ThySkeletonComponent.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
ThySkeletonComponent.propDecorators = {
    addSkeletonClass: [{ type: HostBinding, args: ['class.thy-skeleton',] }],
    thyAnimate: [{ type: Input }],
    thyBaseUrl: [{ type: Input }],
    thyWidth: [{ type: Input }],
    thyHeight: [{ type: Input }],
    thyViewBoxWidth: [{ type: Input }],
    thyViewBoxHeight: [{ type: Input }],
    thySpeed: [{ type: Input }],
    thyPreserveAspectRatio: [{ type: Input }],
    thyPrimaryColor: [{ type: Input }],
    thySecondaryColor: [{ type: Input }],
    thyPrimaryOpacity: [{ type: Input }],
    thySecondaryOpacity: [{ type: Input }],
    thyUniqueKey: [{ type: Input }],
    thyRtl: [{ type: Input }],
    thyStyle: [{ type: Input }],
    thyIgnoreBaseUrl: [{ type: Input }],
    thyLoadingDone: [{ type: Input }],
    customTemplate: [{ type: ContentChild, args: ['content',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton',
                template: "<svg\n  *ngIf=\"!thyLoadingDone\"\n  attr.viewBox=\"0 0 {{ thyViewBoxWidth }} {{ thyViewBoxHeight }}\"\n  [attr.preserveAspectRatio]=\"thyPreserveAspectRatio\"\n  [ngStyle]=\"thyStyle\"\n  [attr.height]=\"thyHeight\"\n  [attr.width]=\"thyWidth\"\n>\n  <rect\n    [attr.clip-path]=\"clipPath\"\n    [ngStyle]=\"fillStyle\"\n    x=\"0\"\n    y=\"0\"\n    [attr.width]=\"thyWidth\"\n    [attr.height]=\"thyHeight\"\n  />\n\n  <defs>\n    <clipPath [attr.id]=\"idClip\">\n      <ng-container *ngIf=\"contentTemplates\">\n        <ng-container *ngFor=\"let content of contentTemplates\">\n          <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n        </ng-container>\n      </ng-container>\n\n      <ng-container *ngIf=\"customTemplate\">\n        <ng-container *ngTemplateOutlet=\"customTemplate\"></ng-container>\n      </ng-container>\n    </clipPath>\n\n    <linearGradient [attr.id]=\"idGradient\">\n      <stop offset=\"0%\" [attr.stop-color]=\"thyPrimaryColor\" [attr.stop-opacity]=\"thyPrimaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[0]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n\n      <stop offset=\"50%\" [attr.stop-color]=\"thySecondaryColor\" [attr.stop-opacity]=\"thySecondaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[1]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n\n      <stop offset=\"100%\" [attr.stop-color]=\"thyPrimaryColor\" [attr.stop-opacity]=\"thyPrimaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[2]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n    </linearGradient>\n  </defs>\n</svg>\n<ng-container *ngIf=\"thyLoadingDone\">\n  <ng-content></ng-content>\n</ng-container>\n"
            }]
    }], function () { return [{ type: String, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, { addSkeletonClass: [{
            type: HostBinding,
            args: ['class.thy-skeleton']
        }], thyAnimate: [{
            type: Input
        }], thyBaseUrl: [{
            type: Input
        }], thyWidth: [{
            type: Input
        }], thyHeight: [{
            type: Input
        }], thyViewBoxWidth: [{
            type: Input
        }], thyViewBoxHeight: [{
            type: Input
        }], thySpeed: [{
            type: Input
        }], thyPreserveAspectRatio: [{
            type: Input
        }], thyPrimaryColor: [{
            type: Input
        }], thySecondaryColor: [{
            type: Input
        }], thyPrimaryOpacity: [{
            type: Input
        }], thySecondaryOpacity: [{
            type: Input
        }], thyIgnoreBaseUrl: [{
            type: Input
        }], thyLoadingDone: [{
            type: Input
        }], thyRtl: [{
            type: Input
        }], thyUniqueKey: [{
            type: Input
        }], thyStyle: [{
            type: Input
        }], customTemplate: [{
            type: ContentChild,
            args: ['content']
        }] }); })();
