// ngx-tethys/esm2015/core/click-positioner.js
import { Injectable } from '@angular/core';
import { ThyClickDispatcher } from './event-dispatchers/index';
import * as i0 from "@angular/core";
import * as i1 from "./event-dispatchers/click-dispatcher";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './event-dispatchers/index';
export class ThyClickPositioner {
    constructor(clickDispatcher) {
        this.clickDispatcher = clickDispatcher;
        this.lastPosition = null;
        this.initialized = false;
    }
    get lastClickPosition() {
        return this.lastPosition;
    }
    runTaskUseLastPosition(task) {
        setTimeout(() => {
            task(this.lastClickPosition);
        });
    }
    initialize() {
        if (this.initialized) {
            return;
        }
        this.initialized = true;
        this.clickDispatcher.clicked(0).subscribe((event) => {
            this.lastPosition = { x: event.clientX, y: event.clientY };
        });
    }
}
ThyClickPositioner.ɵfac = function ThyClickPositioner_Factory(t) { return new (t || ThyClickPositioner)(ɵngcc0.ɵɵinject(ɵngcc1.ThyClickDispatcher)); };
ThyClickPositioner.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyClickPositioner_Factory() { return new ThyClickPositioner(i0.ɵɵinject(i1.ThyClickDispatcher)); }, token: ThyClickPositioner, providedIn: "root" });
ThyClickPositioner.ctorParameters = () => [
    { type: ThyClickDispatcher }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyClickPositioner, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.ThyClickDispatcher }]; }, null); })();
