// ngx-tethys/esm2015/core/scroll.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import { reqAnimFrame } from './request-animation';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as ɵngcc0 from '@angular/core';
function easeInOutCubic(t, b, c, d) {
    const cc = c - b;
    let tt = t / (d / 2);
    if (tt < 1) {
        return (cc / 2) * tt * tt * tt + b;
    }
    else {
        return (cc / 2) * ((tt -= 2) * tt * tt + 2) + b;
    }
}
export class ThyScrollService {
    constructor(document) {
        this.document = document;
    }
    /** Set the position of the scroll bar of `element`. */
    setScrollTop(element, topValue = 0) {
        if (element === window) {
            this.document.body.scrollTop = topValue;
            this.document.documentElement.scrollTop = topValue;
        }
        else {
            element.scrollTop = topValue;
        }
    }
    /** Get the position of the scoll bar of `element`. */
    getScroll(element, top = true) {
        const target = element ? element : window;
        const prop = top ? 'pageYOffset' : 'pageXOffset';
        const method = top ? 'scrollTop' : 'scrollLeft';
        const isWindow = target === window;
        let ret = isWindow ? target[prop] : target[method];
        if (isWindow && typeof ret !== 'number') {
            ret = this.document.documentElement[method];
        }
        return ret;
    }
    /**
     * Scroll `element` to some position with animation.
     *
     * @param container container, `window` by default
     * @param topValue Scroll to `top`, 0 by default
     * @param easing Transition curve, `easeInOutCubic` by default
     * @param callback callback invoked when transition is done
     */
    scrollTo(container, topValue = 0, easing, callback) {
        const target = container ? container : window;
        const scrollTop = this.getScroll(target);
        const startTime = Date.now();
        const frameFunc = () => {
            const timestamp = Date.now();
            const time = timestamp - startTime;
            this.setScrollTop(target, (easing || easeInOutCubic)(time, scrollTop, topValue, 450));
            if (time < 450) {
                reqAnimFrame(frameFunc);
            }
            else {
                if (callback) {
                    callback();
                }
            }
        };
        reqAnimFrame(frameFunc);
    }
}
ThyScrollService.ɵfac = function ThyScrollService_Factory(t) { return new (t || ThyScrollService)(ɵngcc0.ɵɵinject(DOCUMENT)); };
ThyScrollService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyScrollService_Factory() { return new ThyScrollService(i0.ɵɵinject(i1.DOCUMENT)); }, token: ThyScrollService, providedIn: "root" });
ThyScrollService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyScrollService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, null); })();
