// ngx-tethys/esm2015/core/update-host-class.service.js
import { Injectable, Renderer2, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class UpdateHostClassService {
    constructor(renderer) {
        this.renderer = renderer;
        this._classNames = [];
    }
    initializeElement(element) {
        if (element instanceof ElementRef) {
            this._hostElement = element.nativeElement;
        }
        else {
            this._hostElement = element;
        }
        return this;
    }
    updateClass(classNames) {
        if (this._classNames) {
            this._classNames.forEach(className => {
                if (classNames.indexOf(className) < 0) {
                    this.removeClass(className);
                }
            });
        }
        const newClasses = [];
        classNames.forEach(className => {
            if (className) {
                newClasses.push(className);
                if (this._classNames.indexOf(className) < 0) {
                    this.addClass(className);
                }
            }
        });
        this._classNames = newClasses;
        return this;
    }
    updateClassByMap(classMap) {
        const newClasses = [];
        for (const key in classMap) {
            if (classMap.hasOwnProperty(key) && classMap[key]) {
                newClasses.push(key);
            }
        }
        this.updateClass(newClasses);
    }
    addClass(className) {
        this.renderer.addClass(this._hostElement, className);
        return this;
    }
    removeClass(className) {
        this.renderer.removeClass(this._hostElement, className);
        return this;
    }
}
UpdateHostClassService.ɵfac = function UpdateHostClassService_Factory(t) { return new (t || UpdateHostClassService)(ɵngcc0.ɵɵinject(ɵngcc0.Renderer2)); };
UpdateHostClassService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: UpdateHostClassService, factory: UpdateHostClassService.ɵfac });
UpdateHostClassService.ctorParameters = () => [
    { type: Renderer2 }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(UpdateHostClassService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.Renderer2 }]; }, null); })();
