// ngx-tethys/esm2015/timeline/timeline-item.component.js
import { Component, Input, HostBinding, ContentChild, TemplateRef, ViewChild, ChangeDetectorRef, ChangeDetectionStrategy } from '@angular/core';
import { ThyTimelineService } from './timeline.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './timeline.service';
import * as ɵngcc2 from '@angular/common';

const _c0 = ["dot"];
const _c1 = ["description"];
const _c2 = ["timelineItem"];
function ThyTimelineItemComponent_ng_template_0_3_ng_template_0_Template(rf, ctx) { }
function ThyTimelineItemComponent_ng_template_0_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTimelineItemComponent_ng_template_0_3_ng_template_0_Template, 0, 0, "ng-template", 7);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r2.dot);
} }
function ThyTimelineItemComponent_ng_template_0_7_ng_template_0_Template(rf, ctx) { }
function ThyTimelineItemComponent_ng_template_0_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTimelineItemComponent_ng_template_0_7_ng_template_0_Template, 0, 0, "ng-template", 7);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r3.description);
} }
function ThyTimelineItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelement(1, "div", 2);
    ɵngcc0.ɵɵelementStart(2, "div", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTimelineItemComponent_ng_template_0_3_Template, 1, 1, undefined, 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 5);
    ɵngcc0.ɵɵprojection(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 6);
    ɵngcc0.ɵɵtemplate(7, ThyTimelineItemComponent_ng_template_0_7_Template, 1, 1, undefined, 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("thy-timeline-item-reverse-last", ctx_r1.isLast && ctx_r1.reverse)("thy-timeline-item-last", ctx_r1.isLast)("thy-timeline-item-reverse-first", ctx_r1.isFirst && ctx_r1.reverse)("thy-timeline-item-right", (ctx_r1.thyPosition || ctx_r1.position) === "right")("thy-timeline-item-left", (ctx_r1.thyPosition || ctx_r1.position) === "left");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵclassProp("thy-timeline-item-dot-primary", ctx_r1.color === "primary")("thy-timeline-item-dot-success", ctx_r1.color === "success")("thy-timeline-item-dot-warning", ctx_r1.color === "warning")("thy-timeline-item-dot-danger", ctx_r1.color === "danger")("thy-timeline-item-dot-info", ctx_r1.color === "info")("thy-timeline-item-dot-custom", !!ctx_r1.dot);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.dot);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.description);
} }
const _c3 = ["*"];
export class ThyTimelineItemComponent {
    constructor(cdr, timelineService) {
        this.cdr = cdr;
        this.timelineService = timelineService;
        this.color = 'primary';
        this.isLast = false;
        this.isFirst = false;
        this.reverse = false;
    }
    set thyColor(value) {
        if (value) {
            this.color = value;
        }
    }
    detectChanges() {
        this.cdr.detectChanges();
    }
    ngOnChanges(changes) {
        this.timelineService.markForCheck();
    }
    ngOnInit() { }
}
ThyTimelineItemComponent.ɵfac = function ThyTimelineItemComponent_Factory(t) { return new (t || ThyTimelineItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTimelineService)); };
ThyTimelineItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTimelineItemComponent, selectors: [["thy-timeline-item"]], contentQueries: function ThyTimelineItemComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dot = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.description = _t.first);
    } }, viewQuery: function ThyTimelineItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c2, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, hostVars: 2, hostBindings: function ThyTimelineItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.className);
    } }, inputs: { thyColor: "thyColor", thyPosition: "thyPosition" }, exportAs: ["ThyTimelineItem"], features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c3, decls: 2, vars: 0, consts: [["timelineItem", ""], [1, "thy-timeline-item"], [1, "thy-timeline-item-tail"], [1, "thy-timeline-item-dot"], [4, "ngIf"], [1, "thy-timeline-item-content"], [1, "thy-timeline-item-content-other-side"], [3, "ngTemplateOutlet"]], template: function ThyTimelineItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThyTimelineItemComponent_ng_template_0_Template, 8, 24, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgTemplateOutlet], encapsulation: 2, changeDetection: 0 });
ThyTimelineItemComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ThyTimelineService }
];
ThyTimelineItemComponent.propDecorators = {
    template: [{ type: ViewChild, args: ['timelineItem', { static: false },] }],
    className: [{ type: HostBinding, args: ['class',] }],
    thyColor: [{ type: Input }],
    thyPosition: [{ type: Input }],
    dot: [{ type: ContentChild, args: ['dot', { static: false },] }],
    description: [{ type: ContentChild, args: ['description', { static: false },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTimelineItemComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                selector: 'thy-timeline-item',
                template: "<ng-template #timelineItem>\n  <div\n    class=\"thy-timeline-item\"\n    [class.thy-timeline-item-reverse-last]=\"isLast && reverse\"\n    [class.thy-timeline-item-last]=\"isLast\"\n    [class.thy-timeline-item-reverse-first]=\"isFirst && reverse\"\n    [class.thy-timeline-item-right]=\"(thyPosition || position) === 'right'\"\n    [class.thy-timeline-item-left]=\"(thyPosition || position) === 'left'\"\n  >\n    <div class=\"thy-timeline-item-tail\"></div>\n    <div\n      class=\"thy-timeline-item-dot\"\n      [class.thy-timeline-item-dot-primary]=\"color === 'primary'\"\n      [class.thy-timeline-item-dot-success]=\"color === 'success'\"\n      [class.thy-timeline-item-dot-warning]=\"color === 'warning'\"\n      [class.thy-timeline-item-dot-danger]=\"color === 'danger'\"\n      [class.thy-timeline-item-dot-info]=\"color === 'info'\"\n      [class.thy-timeline-item-dot-custom]=\"!!dot\"\n    >\n      <ng-template *ngIf=\"dot\" [ngTemplateOutlet]=\"dot\"></ng-template>\n    </div>\n    <div class=\"thy-timeline-item-content\">\n      <ng-content></ng-content>\n    </div>\n    <div class=\"thy-timeline-item-content-other-side\">\n      <ng-template *ngIf=\"description\" [ngTemplateOutlet]=\"description\"></ng-template>\n    </div>\n  </div>\n</ng-template>\n",
                exportAs: 'ThyTimelineItem'
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.ThyTimelineService }]; }, { thyColor: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['timelineItem', { static: false }]
        }], className: [{
            type: HostBinding,
            args: ['class']
        }], thyPosition: [{
            type: Input
        }], dot: [{
            type: ContentChild,
            args: ['dot', { static: false }]
        }], description: [{
            type: ContentChild,
            args: ['description', { static: false }]
        }] }); })();
