// ngx-tethys/esm2015/timeline/timeline.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyTimelineComponent } from './timeline.component';
import { ThyTimelineItemComponent } from './timeline-item.component';
import * as ɵngcc0 from '@angular/core';
export class ThyTimelineModule {
}
ThyTimelineModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTimelineModule });
ThyTimelineModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTimelineModule_Factory(t) { return new (t || ThyTimelineModule)(); }, providers: [], imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTimelineModule, { declarations: function () { return [ThyTimelineComponent, ThyTimelineItemComponent]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyTimelineComponent, ThyTimelineItemComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTimelineModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyTimelineComponent, ThyTimelineItemComponent],
                entryComponents: [],
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyTimelineComponent, ThyTimelineItemComponent],
                providers: []
            }]
    }], null, null); })();
