// ngx-tethys/esm2015/skeleton/stylized/avatar.component.js
import { Component, TemplateRef, ViewChild } from '@angular/core';
import { ThySkeletonComponent } from '../skeleton.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../skeleton.component';

const _c0 = ["content"];
function ThySkeletonAvatarComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "circle", 1);
} }
export class ThySkeletonAvatarComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonAvatarComponent.ɵfac = function ThySkeletonAvatarComponent_Factory(t) { return new (t || ThySkeletonAvatarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThySkeletonComponent)); };
ThySkeletonAvatarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonAvatarComponent, selectors: [["thy-skeleton-avatar-template"], ["", "thySkeletonAvatarTemplate", ""]], viewQuery: function ThySkeletonAvatarComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, decls: 2, vars: 0, consts: [["content", ""], ["cx", "30", "cy", "30", "r", "30"]], template: function ThySkeletonAvatarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonAvatarComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonAvatarComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonAvatarComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonAvatarComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-avatar-template,[thySkeletonAvatarTemplate]',
                template: `
        <ng-template #content>
            <svg:circle cx="30" cy="30" r="30" />
        </ng-template>
    `
            }]
    }], function () { return [{ type: ɵngcc1.ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();
