// ngx-tethys/esm2015/skeleton/stylized/bullet-list.component.js
import { Component, Input, ViewChild, TemplateRef } from '@angular/core';
import { ThySkeletonComponent } from '../skeleton.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../skeleton.component';
import * as ɵngcc2 from '@angular/common';

const _c0 = ["content"];
function ThySkeletonBulletListComponent_ng_template_0_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(1, "circle", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("cx", item_r3.circle.cx)("cy", item_r3.circle.cy);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("x", item_r3.rect.x)("y", item_r3.rect.y);
} }
function ThySkeletonBulletListComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThySkeletonBulletListComponent_ng_template_0_ng_container_0_Template, 3, 4, "ng-container", 1);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.items);
} }
export class ThySkeletonBulletListComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
        this.thyCount = 5;
        this.items = [];
    }
    ngOnInit() {
        for (let i = 0; i <= this.thyCount; i++) {
            this.items.push({
                circle: {
                    cx: 10,
                    cy: i * 30 + 20,
                    r: 8
                },
                rect: {
                    x: 25,
                    y: i * 2 * 15 + 15,
                    rx: 5,
                    ry: 5,
                    width: '90%',
                    height: 10
                }
            });
        }
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonBulletListComponent.ɵfac = function ThySkeletonBulletListComponent_Factory(t) { return new (t || ThySkeletonBulletListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThySkeletonComponent)); };
ThySkeletonBulletListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonBulletListComponent, selectors: [["thy-skeleton-bullet-list-template"]], viewQuery: function ThySkeletonBulletListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, inputs: { thyCount: "thyCount" }, decls: 2, vars: 0, consts: [["content", ""], [4, "ngFor", "ngForOf"], ["r", "8"], ["rx", "5", "ry", "5", "width", "90%", "height", "10"]], template: function ThySkeletonBulletListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonBulletListComponent_ng_template_0_Template, 1, 1, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc2.NgForOf], encapsulation: 2 });
ThySkeletonBulletListComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonBulletListComponent.propDecorators = {
    thyCount: [{ type: Input }],
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonBulletListComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-bullet-list-template',
                template: `
        <ng-template #content>
            <ng-container *ngFor="let item of items">
                <svg:circle [attr.cx]="item.circle.cx" [attr.cy]="item.circle.cy" r="8" />
                <svg:rect [attr.x]="item.rect.x" [attr.y]="item.rect.y" rx="5" ry="5" width="90%" height="10" />
            </ng-container>
        </ng-template>
    `
            }]
    }], function () { return [{ type: ɵngcc1.ThySkeletonComponent }]; }, { thyCount: [{
            type: Input
        }], contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();
