// ngx-tethys/esm2015/skeleton/stylized/list.component.js
import { Component, ViewChild, TemplateRef } from '@angular/core';
import { ThySkeletonComponent } from '../skeleton.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../skeleton.component';

const _c0 = ["content"];
function ThySkeletonListComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
    ɵngcc0.ɵɵelement(1, "rect", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelement(3, "rect", 4);
    ɵngcc0.ɵɵelement(4, "rect", 5);
    ɵngcc0.ɵɵelement(5, "rect", 6);
} }
export class ThySkeletonListComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonListComponent.ɵfac = function ThySkeletonListComponent_Factory(t) { return new (t || ThySkeletonListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThySkeletonComponent)); };
ThySkeletonListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonListComponent, selectors: [["thy-skeleton-list-template"]], viewQuery: function ThySkeletonListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, decls: 2, vars: 0, consts: [["content", ""], ["x", "0", "y", "0", "rx", "3", "ry", "3", "width", "250", "height", "10"], ["x", "20", "y", "20", "rx", "3", "ry", "3", "width", "220", "height", "10"], ["x", "20", "y", "40", "rx", "3", "ry", "3", "width", "170", "height", "10"], ["x", "0", "y", "60", "rx", "3", "ry", "3", "width", "250", "height", "10"], ["x", "20", "y", "80", "rx", "3", "ry", "3", "width", "200", "height", "10"], ["x", "20", "y", "100", "rx", "3", "ry", "3", "width", "80", "height", "10"]], template: function ThySkeletonListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonListComponent_ng_template_0_Template, 6, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonListComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonListComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonListComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-list-template',
                template: `
        <ng-template #content>
            <svg:rect x="0" y="0" rx="3" ry="3" width="250" height="10" />
            <svg:rect x="20" y="20" rx="3" ry="3" width="220" height="10" />
            <svg:rect x="20" y="40" rx="3" ry="3" width="170" height="10" />
            <svg:rect x="0" y="60" rx="3" ry="3" width="250" height="10" />
            <svg:rect x="20" y="80" rx="3" ry="3" width="200" height="10" />
            <svg:rect x="20" y="100" rx="3" ry="3" width="80" height="10" />
        </ng-template>
    `
            }]
    }], function () { return [{ type: ɵngcc1.ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();
