// ngx-tethys/esm2015/skeleton/stylized/paragraph.component.js
import { Component, TemplateRef, ViewChild } from '@angular/core';
import { ThySkeletonComponent } from '../skeleton.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../skeleton.component';

const _c0 = ["content"];
function ThySkeletonParagraphComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
    ɵngcc0.ɵɵelement(1, "rect", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelement(3, "rect", 4);
} }
export class ThySkeletonParagraphComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonParagraphComponent.ɵfac = function ThySkeletonParagraphComponent_Factory(t) { return new (t || ThySkeletonParagraphComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThySkeletonComponent)); };
ThySkeletonParagraphComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonParagraphComponent, selectors: [["thy-skeleton-paragraph-template"], ["", "thySkeletonParagraphTemplate", ""]], viewQuery: function ThySkeletonParagraphComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, exportAs: ["thySkeletonParagraphTemplate"], decls: 2, vars: 0, consts: [["content", ""], ["x", "0", "y", "0", "rx", "2", "ry", "2", "width", "38%", "height", "10"], ["x", "20", "y", "20", "rx", "2", "ry", "2", "width", "92%", "height", "10"], ["x", "20", "y", "40", "rx", "2", "ry", "2", "width", "92%", "height", "10"], ["x", "0", "y", "60", "rx", "2", "ry", "2", "width", "61%", "height", "10"]], template: function ThySkeletonParagraphComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonParagraphComponent_ng_template_0_Template, 4, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonParagraphComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonParagraphComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonParagraphComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-paragraph-template,[thySkeletonParagraphTemplate]',
                template: `
        <ng-template #content>
            <svg:rect x="0" y="0" rx="2" ry="2" width="38%" height="10" />
            <svg:rect x="20" y="20" rx="2" ry="2" width="92%" height="10" />
            <svg:rect x="20" y="40" rx="2" ry="2" width="92%" height="10" />
            <svg:rect x="0" y="60" rx="2" ry="2" width="61%" height="10" />
        </ng-template>
    `,
                exportAs: 'thySkeletonParagraphTemplate'
            }]
    }], function () { return [{ type: ɵngcc1.ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();
