// ngx-tethys/esm2015/skeleton/stylized/title.component.js
import { Component, TemplateRef, ViewChild, Input } from '@angular/core';
import { ThySkeletonComponent } from '../skeleton.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../skeleton.component';

const _c0 = ["content"];
function ThySkeletonTitleComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("width", ctx_r1.width)("height", ctx_r1.height);
} }
export class ThySkeletonTitleComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
        this.width = 100;
        this.height = 10;
    }
    set thyWidth(value) {
        if (value) {
            this.width = value;
        }
    }
    set thyHeight(value) {
        if (value) {
            this.height = value;
        }
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonTitleComponent.ɵfac = function ThySkeletonTitleComponent_Factory(t) { return new (t || ThySkeletonTitleComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThySkeletonComponent)); };
ThySkeletonTitleComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonTitleComponent, selectors: [["thy-skeleton-title-template"], ["", "thySkeletonTitleTemplate", ""]], viewQuery: function ThySkeletonTitleComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, inputs: { thyWidth: "thyWidth", thyHeight: "thyHeight" }, exportAs: ["thySkeletonTitleTemplate"], decls: 2, vars: 0, consts: [["content", ""], ["x", "15", "y", "13", "rx", "4", "ry", "4"]], template: function ThySkeletonTitleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonTitleComponent_ng_template_0_Template, 1, 2, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonTitleComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonTitleComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }],
    thyWidth: [{ type: Input }],
    thyHeight: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonTitleComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-title-template,[thySkeletonTitleTemplate]',
                template: `
        <ng-template #content>
            <svg:rect x="15" y="13" rx="4" ry="4" [attr.width]="width" [attr.height]="height" />
        </ng-template>
    `,
                exportAs: 'thySkeletonTitleTemplate'
            }]
    }], function () { return [{ type: ɵngcc1.ThySkeletonComponent }]; }, { thyWidth: [{
            type: Input
        }], thyHeight: [{
            type: Input
        }], contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();
