// ngx-tethys/esm2015/core/event-dispatchers/click-dispatcher.js
import { Injectable, Inject, NgZone } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { ThyEventDispatcher } from './event-dispatcher';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as ɵngcc0 from '@angular/core';
const DEFAULT_CLICKED_TIME = 100;
export class ThyClickDispatcher extends ThyEventDispatcher {
    constructor(document, ngZone) {
        super(document, ngZone, 'click');
    }
    clicked(auditTimeInMs = DEFAULT_CLICKED_TIME) {
        return this.subscribe(auditTimeInMs);
    }
}
ThyClickDispatcher.ɵfac = function ThyClickDispatcher_Factory(t) { return new (t || ThyClickDispatcher)(ɵngcc0.ɵɵinject(DOCUMENT), ɵngcc0.ɵɵinject(ɵngcc0.NgZone)); };
ThyClickDispatcher.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyClickDispatcher_Factory() { return new ThyClickDispatcher(i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i0.NgZone)); }, token: ThyClickDispatcher, providedIn: "root" });
ThyClickDispatcher.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: NgZone }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyClickDispatcher, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc0.NgZone }]; }, null); })();
