// ngx-tethys/esm2015/core/event-dispatchers/event-dispatcher.js
import { Injectable, NgZone } from '@angular/core';
import { fromEvent, Subject, Observable } from 'rxjs';
import { auditTime } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
const DEFAULT_EVENT_TIME = 100;
export class ThyEventDispatcher {
    constructor(document, ngZone, eventName) {
        this.document = document;
        this.ngZone = ngZone;
        this.eventName = eventName;
        this._globalSubscription = null;
        this._event$ = new Subject();
        this._subscriptionCount = 0;
    }
    _addGlobalListener() {
        this._globalSubscription = this.ngZone.runOutsideAngular(() => {
            return fromEvent(this.document, this.eventName).subscribe((event) => {
                this._event$.next(event);
            });
        });
    }
    _removeGlobalListener() {
        if (this._globalSubscription) {
            this._globalSubscription.unsubscribe();
            this._globalSubscription = null;
        }
    }
    get globalSubscription() {
        return this._globalSubscription;
    }
    subscribe(auditTimeInMs = DEFAULT_EVENT_TIME) {
        return new Observable(observer => {
            if (!this._globalSubscription) {
                this._addGlobalListener();
            }
            // In the case of a 0ms delay, use an observable without auditTime
            // since it does add a perceptible delay in processing overhead.
            const subscription = auditTimeInMs > 0 ? this._event$.pipe(auditTime(auditTimeInMs)).subscribe(observer) : this._event$.subscribe(observer);
            this._subscriptionCount++;
            return () => {
                subscription.unsubscribe();
                this._subscriptionCount--;
                if (!this._subscriptionCount) {
                    this._removeGlobalListener();
                }
            };
        });
    }
    ngOnDestroy() {
        this._removeGlobalListener();
        this._event$.complete();
    }
}
ThyEventDispatcher.ɵfac = function ThyEventDispatcher_Factory(t) { return new (t || ThyEventDispatcher)(ɵngcc0.ɵɵinject(undefined), ɵngcc0.ɵɵinject(ɵngcc0.NgZone), ɵngcc0.ɵɵinject(String)); };
ThyEventDispatcher.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyEventDispatcher, factory: ThyEventDispatcher.ɵfac });
ThyEventDispatcher.ctorParameters = () => [
    { type: undefined },
    { type: NgZone },
    { type: String }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyEventDispatcher, [{
        type: Injectable
    }], function () { return [{ type: undefined }, { type: ɵngcc0.NgZone }, { type: String }]; }, null); })();
