// ngx-tethys/esm2015/core/event-dispatchers/keyboard-dispatcher.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable, NgZone } from '@angular/core';
import { ThyEventDispatcher } from './event-dispatcher';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as ɵngcc0 from '@angular/core';
const DEFAULT_KEYDOWN_TIME = 100;
export class ThyKeyboardDispatcher extends ThyEventDispatcher {
    constructor(document, ngZone) {
        super(document, ngZone, 'keydown');
    }
    keydown(auditTimeInMs = DEFAULT_KEYDOWN_TIME) {
        return this.subscribe(auditTimeInMs);
    }
}
ThyKeyboardDispatcher.ɵfac = function ThyKeyboardDispatcher_Factory(t) { return new (t || ThyKeyboardDispatcher)(ɵngcc0.ɵɵinject(DOCUMENT), ɵngcc0.ɵɵinject(ɵngcc0.NgZone)); };
ThyKeyboardDispatcher.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyKeyboardDispatcher_Factory() { return new ThyKeyboardDispatcher(i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i0.NgZone)); }, token: ThyKeyboardDispatcher, providedIn: "root" });
ThyKeyboardDispatcher.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: NgZone }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyKeyboardDispatcher, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc0.NgZone }]; }, null); })();
