// ngx-tethys/esm2015/raster/module.js
import { NgModule } from '@angular/core';
import { ThyRowDirective } from './thy-row.directive';
import { ThyColDirective } from './thy-col.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyRasterModule {
}
ThyRasterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyRasterModule });
ThyRasterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyRasterModule_Factory(t) { return new (t || ThyRasterModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyRasterModule, { declarations: function () { return [ThyRowDirective, ThyColDirective]; }, exports: function () { return [ThyRowDirective, ThyColDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRasterModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyRowDirective, ThyColDirective],
                exports: [ThyRowDirective, ThyColDirective],
                imports: []
            }]
    }], null, null); })();
