// ngx-tethys/esm2015/raster/thy-row.directive.js
import { Directive, Input, Renderer2, ElementRef } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { mixinUnsubscribe, MixinBase } from 'ngx-tethys/core';
import { isString } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
const _MixinBase = mixinUnsubscribe(MixinBase);
export class ThyRowDirective extends _MixinBase {
    constructor(renderer, elementRef) {
        super();
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.thyAlign = null;
        this.thyJustify = null;
        this.actualGutter$ = new ReplaySubject(1);
    }
    ngOnInit() {
        this._setGutterStyle();
    }
    ngOnChanges() {
        this._setGutterStyle();
    }
    ngAfterViewInit() { }
    _setGutterStyle() {
        const [horizontalGutter, verticalGutter] = this._getGutter();
        this.actualGutter$.next([horizontalGutter, verticalGutter]);
        const renderGutter = (name, gutter) => {
            const nativeElement = this.elementRef.nativeElement;
            this.renderer.setStyle(nativeElement, name, `-${gutter / 2}px`);
        };
        if (horizontalGutter > 0) {
            renderGutter('margin-left', horizontalGutter);
            renderGutter('margin-right', horizontalGutter);
        }
        if (verticalGutter > 0) {
            renderGutter('margin-top', verticalGutter);
            renderGutter('margin-bottom', verticalGutter);
        }
    }
    _getGutter() {
        if (isString(this.thyGutter)) {
            throw Error(`thyGutter value can not be string type`);
        }
        return [this.thyGutter, 0];
    }
    ngOnDestroy() {
        super.ngOnDestroy();
    }
}
ThyRowDirective.ɵfac = function ThyRowDirective_Factory(t) { return new (t || ThyRowDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyRowDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyRowDirective, selectors: [["", "thyRow", ""]], hostAttrs: [1, "thy-row"], inputs: { thyAlign: "thyAlign", thyJustify: "thyJustify", thyGutter: "thyGutter" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature] });
ThyRowDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
ThyRowDirective.propDecorators = {
    thyGutter: [{ type: Input }],
    thyAlign: [{ type: Input }],
    thyJustify: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRowDirective, [{
        type: Directive,
        args: [{
                selector: '[thyRow]',
                host: {
                    class: 'thy-row'
                }
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { thyAlign: [{
            type: Input
        }], thyJustify: [{
            type: Input
        }], thyGutter: [{
            type: Input
        }] }); })();
