// ngx-tethys/esm2015/slide/slide-footer/slide-footer.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThySlideFooterComponent {
    constructor() {
        this.slideLayoutFooter = true;
    }
    ngOnInit() { }
}
ThySlideFooterComponent.ɵfac = function ThySlideFooterComponent_Factory(t) { return new (t || ThySlideFooterComponent)(); };
ThySlideFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySlideFooterComponent, selectors: [["thy-slide-footer"]], hostVars: 2, hostBindings: function ThySlideFooterComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-slide-footer", ctx.slideLayoutFooter);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThySlideFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThySlideFooterComponent.propDecorators = {
    slideLayoutFooter: [{ type: HostBinding, args: ['class.thy-slide-footer',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySlideFooterComponent, [{
        type: Component,
        args: [{
                selector: 'thy-slide-footer',
                template: '<ng-content></ng-content>'
            }]
    }], function () { return []; }, { slideLayoutFooter: [{
            type: HostBinding,
            args: ['class.thy-slide-footer']
        }] }); })();
