// ngx-tethys/esm2015/testing/data-transfer.js
function removeFromArray(array, item) {
    const index = array.indexOf(item);
    if (index >= 0) {
        array.splice(index, 1);
    }
}
export class FakeDataTransferItem {
    constructor(file, kind) {
        this.file = file;
        this.kind = kind;
    }
    getAsFile() {
        return this.file;
    }
    getAsString(callback) {
        throw new Error('Method not implemented.');
    }
    webkitGetAsEntry() {
        return this.file;
    }
}
export class FakeDataTransfer {
    constructor(files, items) {
        this.dataByFormat = {};
        this.dropEffect = 'none';
        this.effectAllowed = 'all';
        this.types = [];
        this.getData = function (dataFormat) {
            return this.dataByFormat[dataFormat];
        };
        this.setDragImage = function () {
            // don't do anything (the stub just makes sure there is no error thrown if someone tries to call the method)
        };
        this.files = files;
        this.items = items;
    }
    clearData(dataFormat) {
        if (dataFormat) {
            delete this.dataByFormat[dataFormat];
            removeFromArray(this.types, dataFormat);
        }
        else {
            this.dataByFormat = {};
            this.types = [];
        }
    }
    setData(dataFormat, data) {
        this.dataByFormat[dataFormat] = data;
        if (this.types.indexOf(dataFormat) < 0) {
            this.types.push(dataFormat);
        }
        return true;
    }
}
