// ngx-tethys/esm2015/input/input.directive.js
import { Directive, ElementRef, HostBinding, Input, Renderer2 } from '@angular/core';
import { UpdateHostClassService } from 'ngx-tethys/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
const inputGroupSizeMap = {
    xs: ['form-control-xs'],
    sm: ['form-control-sm'],
    md: ['form-control-md'],
    lg: ['form-control-lg']
};
export class ThyInputDirective {
    constructor(updateHostClassService, elementRef, render) {
        this.updateHostClassService = updateHostClassService;
        this.elementRef = elementRef;
        this.render = render;
        this._isFormControl = true;
        this.initialized = false;
        this.updateHostClassService.initializeElement(elementRef.nativeElement);
    }
    set thySize(size) {
        if (size && inputGroupSizeMap[size]) {
            this.updateHostClassService.updateClass(inputGroupSizeMap[size]);
        }
        else {
            this.updateHostClassService.updateClass([]);
        }
    }
    set thyAutocomplete(value) {
        this.autocomplete = coerceBooleanProperty(value);
        if (this.initialized) {
            this.setAutocomplete();
        }
    }
    ngOnInit() {
        this.initialized = true;
        this.setAutocomplete();
    }
    setAutocomplete() {
        this.render.setAttribute(this.elementRef.nativeElement, 'autocomplete', this.autocomplete ? 'on' : 'off');
    }
}
ThyInputDirective.ɵfac = function ThyInputDirective_Factory(t) { return new (t || ThyInputDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyInputDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyInputDirective, selectors: [["", "thyInput", ""]], hostVars: 2, hostBindings: function ThyInputDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("form-control", ctx._isFormControl);
    } }, inputs: { thySize: "thySize", thyAutocomplete: "thyAutocomplete" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])] });
ThyInputDirective.ctorParameters = () => [
    { type: UpdateHostClassService },
    { type: ElementRef },
    { type: Renderer2 }
];
ThyInputDirective.propDecorators = {
    _isFormControl: [{ type: HostBinding, args: ['class.form-control',] }],
    thySize: [{ type: Input }],
    thyAutocomplete: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyInputDirective, [{
        type: Directive,
        args: [{
                selector: '[thyInput]',
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { _isFormControl: [{
            type: HostBinding,
            args: ['class.form-control']
        }], thySize: [{
            type: Input
        }], thyAutocomplete: [{
            type: Input
        }] }); })();
