// ngx-tethys/esm2015/input/input.component.js
import { Component, ContentChild, EventEmitter, forwardRef, HostBinding, Input, Output, TemplateRef, ViewChild, ViewEncapsulation, NgZone } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { UpdateHostClassService } from 'ngx-tethys/core';
import { take } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/shared';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from './input.directive';
import * as ɵngcc5 from 'ngx-tethys/icon';

const _c0 = ["append"];
const _c1 = ["prepend"];
const _c2 = ["eye"];
function ThyInputComponent_div_0_2_ng_template_0_Template(rf, ctx) { }
function ThyInputComponent_div_0_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyInputComponent_div_0_2_ng_template_0_Template, 0, 0, "ng-template");
} }
function ThyInputComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5, 6);
    ɵngcc0.ɵɵtemplate(2, ThyInputComponent_div_0_2_Template, 1, 0, undefined, 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.prependTemplate);
} }
function ThyInputComponent_div_5_2_ng_template_0_Template(rf, ctx) { }
function ThyInputComponent_div_5_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyInputComponent_div_5_2_ng_template_0_Template, 0, 0, "ng-template");
} }
function ThyInputComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8, 9);
    ɵngcc0.ɵɵtemplate(2, ThyInputComponent_div_5_2_Template, 1, 0, undefined, 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r2.appendTemplate);
} }
function ThyInputComponent_ng_template_6_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 13);
} }
function ThyInputComponent_ng_template_6_thy_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 14);
} }
function ThyInputComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵlistener("click", function ThyInputComponent_ng_template_6_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.togglePasswordType(); });
    ɵngcc0.ɵɵtemplate(1, ThyInputComponent_ng_template_6_thy_icon_1_Template, 1, 0, "thy-icon", 11);
    ɵngcc0.ɵɵtemplate(2, ThyInputComponent_ng_template_6_thy_icon_2_Template, 1, 0, "thy-icon", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.isPassword(ctx_r4.type));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r4.isPassword(ctx_r4.type));
} }
export const CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => ThyInputComponent),
    multi: true
};
const noop = () => { };
const ɵ0 = noop;
const password = 'password';
export class ThyInputComponent {
    constructor(ngZone) {
        this.ngZone = ngZone;
        this.placeholder = '';
        this.thyAutofocus = false;
        this.readonly = false;
        this.focus = new EventEmitter();
        this.blur = new EventEmitter();
        this._type = 'text';
        this.onTouchedCallback = noop;
        this.onChangeCallback = noop;
        this._isSearchContainer = true;
        this._isFormControl = true;
        this._isFocus = false;
        this.disabled = false;
    }
    set thyType(value) {
        this.type = value;
    }
    set thyAutocomplete(value) {
        this.autocomplete = value;
    }
    ngOnInit() {
        this.ngZone.onStable.pipe(take(1)).subscribe(() => {
            if (this.isPassword(this.type)) {
                this.appendTemplate = this.eyeTemplate;
            }
        });
    }
    writeValue(value) {
        this.value = value;
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    onModelChange() {
        this.onChangeCallback(this.value);
    }
    onInputFocus(event) {
        this._isFocus = true;
        this.showLabel = true;
        this.focus.emit(event);
    }
    onInputBlur(event) {
        this._isFocus = false;
        this.showLabel = false;
        this.blur.emit(event);
    }
    isPassword(value) {
        return value === password;
    }
    togglePasswordType() {
        this.type = this.isPassword(this.type) ? 'text' : 'password';
    }
}
ThyInputComponent.ɵfac = function ThyInputComponent_Factory(t) { return new (t || ThyInputComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyInputComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyInputComponent, selectors: [["thy-input"]], contentQueries: function ThyInputComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.appendTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.prependTemplate = _t.first);
    } }, viewQuery: function ThyInputComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c2, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.eyeTemplate = _t.first);
    } }, hostVars: 8, hostBindings: function ThyInputComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-input", ctx._isSearchContainer)("form-control", ctx._isFormControl)("form-control-active", ctx._isFocus)("disabled", ctx.disabled);
    } }, inputs: { placeholder: "placeholder", thyAutofocus: "thyAutofocus", readonly: "readonly", thyType: "thyType", type: "type", thyAutocomplete: "thyAutocomplete", thySize: "thySize", thyLabelText: "thyLabelText" }, outputs: { focus: "focus", blur: "blur" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService, CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR])], decls: 8, vars: 16, consts: [["class", "input-prepend", 4, "ngIf"], ["thyInput", "", 3, "thySize", "thyAutofocus", "type", "placeholder", "disabled", "thyAutocomplete", "ngModel", "readonly", "ngModelChange", "focus", "blur"], ["input", ""], ["class", "input-append", 4, "ngIf"], ["eye", ""], [1, "input-prepend"], ["prepend", ""], [4, "ngTemplateOutlet"], [1, "input-append"], ["append", ""], ["href", "javascript:;", 1, "link-secondary", "input-password-icon", 3, "click"], ["thyIconName", "eye", 4, "ngIf"], ["thyIconName", "eye-invisible", 4, "ngIf"], ["thyIconName", "eye"], ["thyIconName", "eye-invisible"]], template: function ThyInputComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyInputComponent_div_0_Template, 3, 1, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "span");
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "input", 1, 2);
        ɵngcc0.ɵɵlistener("ngModelChange", function ThyInputComponent_Template_input_ngModelChange_3_listener($event) { return ctx.value = $event; })("ngModelChange", function ThyInputComponent_Template_input_ngModelChange_3_listener() { return ctx.onModelChange(); })("focus", function ThyInputComponent_Template_input_focus_3_listener($event) { return ctx.onInputFocus($event); })("blur", function ThyInputComponent_Template_input_blur_3_listener($event) { return ctx.onInputBlur($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ThyInputComponent_div_5_Template, 3, 1, "div", 3);
        ɵngcc0.ɵɵtemplate(6, ThyInputComponent_ng_template_6_Template, 3, 2, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.prependTemplate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMapInterpolate1("input-label input-label-", ctx.thySize, "");
        ɵngcc0.ɵɵclassProp("active", ctx.showLabel);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyLabelText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thySize", ctx.thySize)("thyAutofocus", ctx.thyAutofocus)("type", ctx.type)("placeholder", ctx.placeholder)("disabled", ctx.disabled)("thyAutocomplete", ctx.autocomplete)("ngModel", ctx.value)("readonly", ctx.readonly);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.appendTemplate);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyAutofocusDirective, ɵngcc3.DefaultValueAccessor, ɵngcc4.ThyInputDirective, ɵngcc3.NgControlStatus, ɵngcc3.NgModel, ɵngcc1.NgTemplateOutlet, ɵngcc5.ThyIconComponent], encapsulation: 2 });
ThyInputComponent.ctorParameters = () => [
    { type: NgZone }
];
ThyInputComponent.propDecorators = {
    placeholder: [{ type: Input }],
    thySize: [{ type: Input }],
    thyAutofocus: [{ type: Input }],
    type: [{ type: Input }],
    thyType: [{ type: Input }],
    thyLabelText: [{ type: Input }],
    readonly: [{ type: Input }],
    thyAutocomplete: [{ type: Input }],
    focus: [{ type: Output }],
    blur: [{ type: Output }],
    appendTemplate: [{ type: ContentChild, args: ['append',] }],
    prependTemplate: [{ type: ContentChild, args: ['prepend',] }],
    eyeTemplate: [{ type: ViewChild, args: ['eye', { static: true },] }],
    _isSearchContainer: [{ type: HostBinding, args: ['class.thy-input',] }],
    _isFormControl: [{ type: HostBinding, args: ['class.form-control',] }],
    _isFocus: [{ type: HostBinding, args: ['class.form-control-active',] }],
    disabled: [{ type: HostBinding, args: ['class.disabled',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyInputComponent, [{
        type: Component,
        args: [{
                selector: 'thy-input',
                template: "<div #prepend class=\"input-prepend\" *ngIf=\"prependTemplate\">\n  <ng-template *ngTemplateOutlet=\"prependTemplate\"></ng-template>\n</div>\n<span class=\"input-label input-label-{{ thySize }}\" [class.active]=\"showLabel\">{{ thyLabelText }}</span>\n<input\n  #input\n  thyInput\n  [thySize]=\"thySize\"\n  [thyAutofocus]=\"thyAutofocus\"\n  [type]=\"type\"\n  [placeholder]=\"placeholder\"\n  [disabled]=\"disabled\"\n  [thyAutocomplete]=\"autocomplete\"\n  [(ngModel)]=\"value\"\n  (ngModelChange)=\"onModelChange()\"\n  (focus)=\"onInputFocus($event)\"\n  (blur)=\"onInputBlur($event)\"\n  [readonly]=\"readonly\"\n/>\n<div #append class=\"input-append\" *ngIf=\"appendTemplate\">\n  <ng-template *ngTemplateOutlet=\"appendTemplate\"></ng-template>\n</div>\n\n<ng-template #eye>\n  <a href=\"javascript:;\" class=\"link-secondary input-password-icon\" (click)=\"togglePasswordType()\">\n    <thy-icon *ngIf=\"isPassword(type)\" thyIconName=\"eye\"></thy-icon>\n    <thy-icon *ngIf=\"!isPassword(type)\" thyIconName=\"eye-invisible\"></thy-icon>\n  </a>\n</ng-template>\n",
                providers: [UpdateHostClassService, CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR],
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc0.NgZone }]; }, { placeholder: [{
            type: Input
        }], thyAutofocus: [{
            type: Input
        }], readonly: [{
            type: Input
        }], focus: [{
            type: Output
        }], blur: [{
            type: Output
        }], _isSearchContainer: [{
            type: HostBinding,
            args: ['class.thy-input']
        }], _isFormControl: [{
            type: HostBinding,
            args: ['class.form-control']
        }], _isFocus: [{
            type: HostBinding,
            args: ['class.form-control-active']
        }], disabled: [{
            type: HostBinding,
            args: ['class.disabled']
        }], thyType: [{
            type: Input
        }], type: [{
            type: Input
        }], thyAutocomplete: [{
            type: Input
        }], appendTemplate: [{
            type: ContentChild,
            args: ['append']
        }], thySize: [{
            type: Input
        }], thyLabelText: [{
            type: Input
        }], prependTemplate: [{
            type: ContentChild,
            args: ['prepend']
        }], eyeTemplate: [{
            type: ViewChild,
            args: ['eye', { static: true }]
        }] }); })();
export { ɵ0 };
