// ngx-tethys/esm2015/input/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyInputDirective } from './input.directive';
import { ThyInputComponent } from './input.component';
import { ThyInputGroupComponent } from './input-group.component';
import { ThyInputSearchComponent } from './input-search.component';
import { FormsModule } from '@angular/forms';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyInputModule {
}
ThyInputModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyInputModule });
ThyInputModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyInputModule_Factory(t) { return new (t || ThyInputModule)(); }, imports: [[CommonModule, FormsModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyInputModule, { declarations: function () { return [ThyInputDirective, ThyInputComponent, ThyInputGroupComponent, ThyInputSearchComponent]; }, imports: function () { return [CommonModule, FormsModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyInputDirective, ThyInputComponent, ThyInputGroupComponent, ThyInputSearchComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyInputModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ThySharedModule, ThyIconModule],
                declarations: [ThyInputDirective, ThyInputComponent, ThyInputGroupComponent, ThyInputSearchComponent],
                exports: [ThyInputDirective, ThyInputComponent, ThyInputGroupComponent, ThyInputSearchComponent]
            }]
    }], null, null); })();
