// ngx-tethys/esm2015/date-range/optional-dates/optional-dates.component.js
import { Component } from '@angular/core';
import { ThyPopover } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-tethys/action-menu';
import * as ɵngcc4 from 'ngx-tethys/date-picker';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from 'ngx-tethys/icon';

function OptionalDateRangesComponent_thy_action_menu_0_a_1_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 7);
    ɵngcc0.ɵɵelement(1, "thy-icon", 8);
    ɵngcc0.ɵɵelementEnd();
} }
function OptionalDateRangesComponent_thy_action_menu_0_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 6);
    ɵngcc0.ɵɵlistener("click", function OptionalDateRangesComponent_thy_action_menu_0_a_1_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const dateRange_r3 = ctx.$implicit; const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5._selectDateRange(dateRange_r3); });
    ɵngcc0.ɵɵelementStart(1, "span", 4);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, OptionalDateRangesComponent_thy_action_menu_0_a_1_span_3_Template, 2, 0, "span", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const dateRange_r3 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(dateRange_r3 == null ? null : dateRange_r3.text);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", dateRange_r3.key === (ctx_r1.selectedDate == null ? null : ctx_r1.selectedDate.key));
} }
function OptionalDateRangesComponent_thy_action_menu_0_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 7);
    ɵngcc0.ɵɵelement(1, "thy-icon", 8);
    ɵngcc0.ɵɵelementEnd();
} }
function OptionalDateRangesComponent_thy_action_menu_0_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-action-menu", 1);
    ɵngcc0.ɵɵtemplate(1, OptionalDateRangesComponent_thy_action_menu_0_a_1_Template, 4, 2, "a", 2);
    ɵngcc0.ɵɵelementStart(2, "a", 3);
    ɵngcc0.ɵɵlistener("ngModelChange", function OptionalDateRangesComponent_thy_action_menu_0_Template_a_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.selectedDate = $event; })("ngModelChange", function OptionalDateRangesComponent_thy_action_menu_0_Template_a_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9._selectedCustomDate($event); });
    ɵngcc0.ɵɵelementStart(3, "span", 4);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, OptionalDateRangesComponent_thy_action_menu_0_span_5_Template, 2, 0, "span", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.optionalDateRanges);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r0.selectedDate)("thyMinDate", ctx_r0.minDate)("thyMaxDate", ctx_r0.maxDate);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.customValue);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customKey === (ctx_r0.selectedDate == null ? null : ctx_r0.selectedDate.key));
} }
export class OptionalDateRangesComponent {
    constructor(thyPopover) {
        this.thyPopover = thyPopover;
        this.hiddenMenu = false;
        this.customValue = '自定义';
    }
    ngOnInit() { }
    _selectDateRange(dateRange) {
        this.selectedDate = dateRange;
        this.selectedDateRange(dateRange);
        this.thyPopover.close();
    }
    _selectedCustomDate(date) {
        this.selectedDate = {
            begin: date.begin,
            end: date.end,
            key: this.customKey,
            text: this.customValue
        };
        this.selectedDateRange(this.selectedDate);
    }
}
OptionalDateRangesComponent.ɵfac = function OptionalDateRangesComponent_Factory(t) { return new (t || OptionalDateRangesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover)); };
OptionalDateRangesComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: OptionalDateRangesComponent, selectors: [["date-range-optional"]], decls: 1, vars: 1, consts: [["class", "thy-date-range-action-menu-container", 4, "ngIf"], [1, "thy-date-range-action-menu-container"], ["thyActionMenuItem", "", "href", "javascript:;", 3, "click", 4, "ngFor", "ngForOf"], ["thyActionMenuItem", "", "href", "javascript:;", "thyRangePicker", "", 3, "ngModel", "thyMinDate", "thyMaxDate", "ngModelChange"], ["thyActionMenuItemName", ""], ["thyActionMenuItemExtendIcon", "", 4, "ngIf"], ["thyActionMenuItem", "", "href", "javascript:;", 3, "click"], ["thyActionMenuItemExtendIcon", ""], ["thyIconName", "check", 1, "text-primary"]], template: function OptionalDateRangesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, OptionalDateRangesComponent_thy_action_menu_0_Template, 6, 6, "thy-action-menu", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.hiddenMenu);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.ThyActionMenuComponent, ɵngcc2.NgForOf, ɵngcc3.ThyActionMenuItemDirective, ɵngcc4.ThyRangePickerDirective, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc3.ThyActionMenuItemNameDirective, ɵngcc3.ThyActionMenuItemExtendIconDirective, ɵngcc6.ThyIconComponent], encapsulation: 2 });
OptionalDateRangesComponent.ctorParameters = () => [
    { type: ThyPopover }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(OptionalDateRangesComponent, [{
        type: Component,
        args: [{
                selector: 'date-range-optional',
                template: "<thy-action-menu *ngIf=\"!hiddenMenu\" class=\"thy-date-range-action-menu-container\">\n  <a thyActionMenuItem href=\"javascript:;\" *ngFor=\"let dateRange of optionalDateRanges\" (click)=\"_selectDateRange(dateRange)\">\n    <span thyActionMenuItemName>{{ dateRange?.text }}</span>\n    <span thyActionMenuItemExtendIcon *ngIf=\"dateRange.key === selectedDate?.key\">\n      <thy-icon class=\"text-primary\" thyIconName=\"check\"></thy-icon>\n    </span>\n  </a>\n  <a\n    thyActionMenuItem\n    href=\"javascript:;\"\n    thyRangePicker\n    [(ngModel)]=\"selectedDate\"\n    [thyMinDate]=\"minDate\"\n    [thyMaxDate]=\"maxDate\"\n    (ngModelChange)=\"_selectedCustomDate($event)\"\n  >\n    <span thyActionMenuItemName>{{ customValue }}</span>\n    <span thyActionMenuItemExtendIcon *ngIf=\"customKey === selectedDate?.key\">\n      <thy-icon class=\"text-primary\" thyIconName=\"check\"></thy-icon>\n    </span>\n  </a>\n</thy-action-menu>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }]; }, null); })();
