// ngx-tethys/esm2015/date-picker/lib/lib-packer.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CalendarFooterComponent } from './calendar/calendar-footer.component';
import { CalendarHeaderComponent } from './calendar/calendar-header.component';
import { DecadePanelComponent } from './decade/decade-panel.component';
import { MonthPanelComponent } from './month/month-panel.component';
import { DatePopupComponent } from './popups/date-popup.component';
import { InnerPopupComponent } from './popups/inner-popup.component';
import { YearPanelComponent } from './year/year-panel.component';
import { MonthTableComponent } from './month/month-table.component';
import { DateTableComponent } from './date/date-table.component';
import { DateTableCellComponent } from './date/date-table-cell.component';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyTimePickerModule } from 'ngx-tethys/time-picker';
import * as ɵngcc0 from '@angular/core';
export class LibPackerModule {
}
LibPackerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LibPackerModule });
LibPackerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LibPackerModule_Factory(t) { return new (t || LibPackerModule)(); }, imports: [[CommonModule, FormsModule, ThyButtonModule, ThyIconModule, ThyTimePickerModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LibPackerModule, { declarations: function () { return [CalendarHeaderComponent,
        CalendarFooterComponent,
        DateTableComponent,
        YearPanelComponent,
        MonthTableComponent,
        MonthPanelComponent,
        DecadePanelComponent,
        InnerPopupComponent,
        DatePopupComponent,
        DateTableCellComponent]; }, imports: function () { return [CommonModule, FormsModule, ThyButtonModule, ThyIconModule, ThyTimePickerModule]; }, exports: function () { return [CalendarHeaderComponent,
        CalendarFooterComponent,
        DateTableComponent,
        YearPanelComponent,
        MonthTableComponent,
        MonthPanelComponent,
        DecadePanelComponent,
        InnerPopupComponent,
        DatePopupComponent,
        DateTableCellComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LibPackerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ThyButtonModule, ThyIconModule, ThyTimePickerModule],
                exports: [
                    CalendarHeaderComponent,
                    CalendarFooterComponent,
                    DateTableComponent,
                    YearPanelComponent,
                    MonthTableComponent,
                    MonthPanelComponent,
                    DecadePanelComponent,
                    InnerPopupComponent,
                    DatePopupComponent,
                    DateTableCellComponent
                ],
                declarations: [
                    CalendarHeaderComponent,
                    CalendarFooterComponent,
                    DateTableComponent,
                    YearPanelComponent,
                    MonthTableComponent,
                    MonthPanelComponent,
                    DecadePanelComponent,
                    InnerPopupComponent,
                    DatePopupComponent,
                    DateTableCellComponent
                ]
            }]
    }], null, null); })();
