// ngx-tethys/esm2015/transfer/transfer.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyTransferComponent } from './transfer.component';
import { ThyTransferListComponent } from './transfer-list.component';
import { ThyButtonModule } from 'ngx-tethys/button';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyListModule } from 'ngx-tethys/list';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyFlexibleTextModule } from 'ngx-tethys/flexible-text';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/icon';
import * as ɵngcc3 from 'ngx-tethys/flexible-text';
export class ThyTransferModule {
}
ThyTransferModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTransferModule });
ThyTransferModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTransferModule_Factory(t) { return new (t || ThyTransferModule)(); }, imports: [[CommonModule, ThyButtonModule, DragDropModule, ThyIconModule, ThyListModule, ThySharedModule, ThyFlexibleTextModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTransferModule, { declarations: function () { return [ThyTransferComponent, ThyTransferListComponent]; }, imports: function () { return [CommonModule, ThyButtonModule, DragDropModule, ThyIconModule, ThyListModule, ThySharedModule, ThyFlexibleTextModule]; }, exports: function () { return [ThyTransferComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTransferModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyTransferComponent, ThyTransferListComponent],
                imports: [CommonModule, ThyButtonModule, DragDropModule, ThyIconModule, ThyListModule, ThySharedModule, ThyFlexibleTextModule],
                exports: [ThyTransferComponent]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(ThyTransferComponent, [ThyTransferListComponent, ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass, ɵngcc1.NgTemplateOutlet, ɵngcc3.ThyFlexibleTextComponent], []);
