// ngx-tethys/esm2015/loading/loading.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyLoadingComponent } from './loading.component';
import * as ɵngcc0 from '@angular/core';
export class ThyLoadingModule {
}
ThyLoadingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLoadingModule });
ThyLoadingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLoadingModule_Factory(t) { return new (t || ThyLoadingModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLoadingModule, { declarations: function () { return [ThyLoadingComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyLoadingComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLoadingModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyLoadingComponent
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ThyLoadingComponent
                ]
            }]
    }], null, null); })();
