// ngx-tethys/esm2015/loading/loading.component.js
import { Component, Input, HostBinding, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function ThyLoadingComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "h4");
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 3);
    ɵngcc0.ɵɵelement(5, "span", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵstyleProp("display", ctx_r0.tip ? "inline" : "none");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.tip);
} }
function ThyLoadingComponent_ng_content_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0, 0, ["*ngIf", "isMask || isDone"]);
} }
const _c0 = ["*"];
export class ThyLoadingComponent {
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.loadingClassName = true;
    }
    set thyDone(value) {
        this.isDone = coerceBooleanProperty(value);
        this.changeDetectorRef.markForCheck();
    }
    set thyTip(value) {
        this.tip = value;
        this.changeDetectorRef.markForCheck();
    }
    // 不传或穿false,没有遮罩层，加载完成出现内容
    set thyIsMask(value) {
        this.isMask = coerceBooleanProperty(value);
        this.changeDetectorRef.markForCheck();
    }
}
ThyLoadingComponent.ɵfac = function ThyLoadingComponent_Factory(t) { return new (t || ThyLoadingComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyLoadingComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyLoadingComponent, selectors: [["thy-loading"]], hostVars: 2, hostBindings: function ThyLoadingComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-loading", ctx.loadingClassName);
    } }, inputs: { thyDone: "thyDone", thyTip: "thyTip", thyIsMask: "thyIsMask" }, ngContentSelectors: _c0, decls: 3, vars: 4, consts: [["class", "thy-loading-indicator", 4, "ngIf"], [4, "ngIf"], [1, "thy-loading-indicator"], [1, "thy-loading-ellipsis"], [1, "thy-spot"]], template: function ThyLoadingComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtemplate(1, ThyLoadingComponent_div_1_Template, 6, 3, "div", 0);
        ɵngcc0.ɵɵtemplate(2, ThyLoadingComponent_ng_content_2_Template, 1, 0, "ng-content", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-loading-mask", ctx.isMask && !ctx.isDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isMask || ctx.isDone);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2, changeDetection: 0 });
ThyLoadingComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
ThyLoadingComponent.propDecorators = {
    thyDone: [{ type: Input }],
    thyTip: [{ type: Input }],
    thyIsMask: [{ type: Input }],
    loadingClassName: [{ type: HostBinding, args: ['class.thy-loading',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLoadingComponent, [{
        type: Component,
        args: [{
                selector: 'thy-loading',
                template: "<div [class.thy-loading-mask]=\"isMask && !isDone\">\n    <div class=\"thy-loading-indicator\" *ngIf=\"!isDone\">\n        <h4>\n            <span [style.display]=\"tip ? 'inline' : 'none'\">{{tip}}</span>\n            <div class=\"thy-loading-ellipsis\">\n                <span class=\"thy-spot\"></span>\n            </div>\n        </h4>\n    </div>\n    <ng-content *ngIf=\"isMask || isDone\"></ng-content>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { loadingClassName: [{
            type: HostBinding,
            args: ['class.thy-loading']
        }], thyDone: [{
            type: Input
        }], thyTip: [{
            type: Input
        }], thyIsMask: [{
            type: Input
        }] }); })();
