// ngx-tethys/esm2015/checkbox/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThyCheckboxComponent } from './checkbox.component';
import * as ɵngcc0 from '@angular/core';
export class ThyCheckboxModule {
}
ThyCheckboxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCheckboxModule });
ThyCheckboxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCheckboxModule_Factory(t) { return new (t || ThyCheckboxModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCheckboxModule, { declarations: function () { return [ThyCheckboxComponent]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [ThyCheckboxComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCheckboxModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                declarations: [
                    ThyCheckboxComponent
                ],
                exports: [
                    ThyCheckboxComponent
                ]
            }]
    }], null, null); })();
