// ngx-tethys/esm2015/dialog/confirm/confirm.component.js
import { Component, ChangeDetectionStrategy, ChangeDetectorRef, Inject, Input } from '@angular/core';
import { ThyDialogRef } from '../dialog-ref';
import { THY_CONFIRM_DEFAULT_OPTIONS, THY_CONFIRM_DEFAULT_OPTIONS_VALUE } from '../confirm.config';
import { finalize } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../dialog-ref';
export class ThyConfirmComponent {
    constructor(dialogRef, changeDetectorRef, defaultConfig) {
        this.dialogRef = dialogRef;
        this.changeDetectorRef = changeDetectorRef;
        this.defaultConfig = defaultConfig;
        this.defaultConfig = Object.assign(Object.assign({}, THY_CONFIRM_DEFAULT_OPTIONS_VALUE), this.defaultConfig);
    }
    ngOnInit() {
        this.defaultConfig = Object.assign(Object.assign({}, this.defaultConfig), this.options);
        this.title = this.defaultConfig.title;
        this.content = this.defaultConfig.content;
        this.okText = this.defaultConfig.okText;
        this.okType = this.defaultConfig.okType;
        this.cancelText = this.defaultConfig.cancelText;
        this.okLoadingText = this.defaultConfig.okLoadingText || this.okText;
        this.footerAlign = this.defaultConfig.footerAlign;
    }
    confirm() {
        this.loading = true;
        const result = this.options.onOk();
        if (result && result.subscribe) {
            result
                .pipe(finalize(() => {
                this.loading = false;
                this.changeDetectorRef.markForCheck();
            }))
                .subscribe(success => {
                if (success) {
                    this.close();
                }
            });
        }
        else {
            this.close();
        }
    }
    close() {
        this.dialogRef.close();
    }
    ngOnDestroy() { }
}
ThyConfirmComponent.ɵfac = function ThyConfirmComponent_Factory(t) { return new (t || ThyConfirmComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyDialogRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(THY_CONFIRM_DEFAULT_OPTIONS)); };
ThyConfirmComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyConfirmComponent, selectors: [["thy-confirm-default"]], inputs: { options: "options" }, decls: 11, vars: 9, consts: [[3, "thyTitle", "thyOnClose"], [1, "thy-form"], ["thyForm", "", "name", "thyConfirm", "thyLayout", "vertical"], [1, "form-group"], [3, "innerHTML"], [3, "thyAlign"], [1, "thy-confirm-footer", 3, "ngClass"], [3, "thyButton", "thyLoading", "thyLoadingText", "click"], ["thyButton", "link-secondary", 1, "thy-confirm-cancel", 3, "click"]], template: function ThyConfirmComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "thy-dialog-header", 0);
        ɵngcc0.ɵɵlistener("thyOnClose", function ThyConfirmComponent_Template_thy_dialog_header_thyOnClose_0_listener() { return ctx.close(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(1, "thy-dialog-body", 1);
        ɵngcc0.ɵɵelementStart(2, "form", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelement(4, "p", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "thy-form-group-footer", 5);
        ɵngcc0.ɵɵelementStart(6, "div", 6);
        ɵngcc0.ɵɵelementStart(7, "button", 7);
        ɵngcc0.ɵɵlistener("click", function ThyConfirmComponent_Template_button_click_7_listener() { return ctx.confirm(); });
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "button", 8);
        ɵngcc0.ɵɵlistener("click", function ThyConfirmComponent_Template_button_click_9_listener() { return ctx.close(); });
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("thyTitle", ctx.title || "Install Angular");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("innerHTML", ctx.content, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thyAlign", ctx.footerAlign);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", "thy-confirm-footer-" + ctx.footerAlign);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thyButton", ctx.okType)("thyLoading", ctx.loading)("thyLoadingText", ctx.okLoadingText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.okText);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.cancelText);
    } }, encapsulation: 2, changeDetection: 0 });
ThyConfirmComponent.ctorParameters = () => [
    { type: ThyDialogRef },
    { type: ChangeDetectorRef },
    { type: undefined, decorators: [{ type: Inject, args: [THY_CONFIRM_DEFAULT_OPTIONS,] }] }
];
ThyConfirmComponent.propDecorators = {
    options: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyConfirmComponent, [{
        type: Component,
        args: [{
                selector: 'thy-confirm-default',
                template: "<thy-dialog-header [thyTitle]=\"title || 'Install Angular'\" (thyOnClose)=\"close()\"> </thy-dialog-header>\n<thy-dialog-body class=\"thy-form\">\n  <form thyForm name=\"thyConfirm\" thyLayout=\"vertical\">\n    <div class=\"form-group\">\n      <p [innerHTML]=\"content\"></p>\n    </div>\n    <thy-form-group-footer [thyAlign]=\"footerAlign\">\n      <div class=\"thy-confirm-footer\" [ngClass]=\"'thy-confirm-footer-' + footerAlign\">\n        <button [thyButton]=\"okType\" (click)=\"confirm()\" [thyLoading]=\"loading\" [thyLoadingText]=\"okLoadingText\">{{ okText }}</button>\n        <button thyButton=\"link-secondary\" (click)=\"close()\" class=\"thy-confirm-cancel\">{{ cancelText }}</button>\n      </div>\n    </thy-form-group-footer>\n  </form>\n</thy-dialog-body>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.ThyDialogRef }, { type: ɵngcc0.ChangeDetectorRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_CONFIRM_DEFAULT_OPTIONS]
            }] }]; }, { options: [{
            type: Input
        }] }); })();
