// ngx-tethys/esm2015/shared/option/module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyListOptionComponent } from './list-option/list-option.component';
import { ThyOptionGroupComponent } from './option-group.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyOptionComponent } from './option.component';
import { ThySelectOptionGroupComponent } from './group/option-group.component';
import * as ɵngcc0 from '@angular/core';
export class ThyOptionModule {
}
ThyOptionModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyOptionModule });
ThyOptionModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyOptionModule_Factory(t) { return new (t || ThyOptionModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyOptionModule, { declarations: function () { return [ThyListOptionComponent, ThyOptionGroupComponent, ThyOptionComponent, ThySelectOptionGroupComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyListOptionComponent, ThyOptionGroupComponent, ThyOptionComponent, ThySelectOptionGroupComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyOptionModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ThyIconModule],
                exports: [ThyListOptionComponent, ThyOptionGroupComponent, ThyOptionComponent, ThySelectOptionGroupComponent],
                declarations: [ThyListOptionComponent, ThyOptionGroupComponent, ThyOptionComponent, ThySelectOptionGroupComponent]
            }]
    }], null, null); })();
