// ngx-tethys/esm2015/shared/option/option-group.component.js
import { Component, ViewEncapsulation, ChangeDetectionStrategy } from '@angular/core';
import { mixinDisabled, MixinBase } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
const _MixinBase = mixinDisabled(MixinBase);
export class ThyOptionGroupComponent extends _MixinBase {
    set thyDisabled(value) {
        this.thyDisabled = value;
    }
}
ThyOptionGroupComponent.ɵfac = function ThyOptionGroupComponent_Factory(t) { return ɵThyOptionGroupComponent_BaseFactory(t || ThyOptionGroupComponent); };
ThyOptionGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyOptionGroupComponent, selectors: [["thy-list-option-group"], ["", "thy-list-option-group", ""]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyOptionGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
const ɵThyOptionGroupComponent_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(ThyOptionGroupComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyOptionGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-list-option-group,[thy-list-option-group]',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
