// ngx-tethys/esm2015/store/module.js
import { NgModule, NgModuleRef } from '@angular/core';
import { ROOT_STATE_TOKEN, FEATURE_STATE_TOKEN } from './types';
import { clearInjector, setInjector } from './internals/static-injector';
import * as ɵngcc0 from '@angular/core';
export class ThyRootStoreModule {
    constructor(ngModuleRef) {
        setInjector(ngModuleRef.injector);
        ngModuleRef.onDestroy(clearInjector);
    }
}
ThyRootStoreModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyRootStoreModule });
ThyRootStoreModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyRootStoreModule_Factory(t) { return new (t || ThyRootStoreModule)(ɵngcc0.ɵɵinject(ɵngcc0.NgModuleRef)); } });
ThyRootStoreModule.ctorParameters = () => [
    { type: NgModuleRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRootStoreModule, [{
        type: NgModule
    }], function () { return [{ type: ɵngcc0.NgModuleRef }]; }, null); })();
export class ThyFeatureStoreModule {
}
ThyFeatureStoreModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyFeatureStoreModule });
ThyFeatureStoreModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyFeatureStoreModule_Factory(t) { return new (t || ThyFeatureStoreModule)(); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFeatureStoreModule, [{
        type: NgModule
    }], null, null); })();
export class ThyStoreModule {
    static forRoot(stores = []) {
        return {
            ngModule: ThyRootStoreModule,
            providers: [
                ...stores,
                {
                    provide: ROOT_STATE_TOKEN,
                    useValue: stores
                }
            ]
        };
    }
    static forFeature(stores = []) {
        return {
            ngModule: ThyFeatureStoreModule,
            providers: [
                ...stores,
                {
                    provide: FEATURE_STATE_TOKEN,
                    multi: true,
                    useValue: stores
                }
            ]
        };
    }
}
ThyStoreModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyStoreModule });
ThyStoreModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyStoreModule_Factory(t) { return new (t || ThyStoreModule)(); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStoreModule, [{
        type: NgModule,
        args: [{}]
    }], null, null); })();
