// ngx-tethys/esm2015/date-range/module.js
import { ThyIconModule } from 'ngx-tethys/icon';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThyDateRangeComponent } from './date-range.component';
import { ThyActionMenuModule } from 'ngx-tethys/action-menu';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyDatePickerModule } from 'ngx-tethys/date-picker';
import { OptionalDateRangesComponent } from './optional-dates/optional-dates.component';
import { ThyNavModule } from 'ngx-tethys/nav';
import * as ɵngcc0 from '@angular/core';
export class ThyDateRangeModule {
}
ThyDateRangeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDateRangeModule });
ThyDateRangeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDateRangeModule_Factory(t) { return new (t || ThyDateRangeModule)(); }, imports: [[CommonModule, FormsModule, ThyActionMenuModule, ThySharedModule, ThyIconModule, ThyDatePickerModule, ThyNavModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDateRangeModule, { declarations: function () { return [ThyDateRangeComponent, OptionalDateRangesComponent]; }, imports: function () { return [CommonModule, FormsModule, ThyActionMenuModule, ThySharedModule, ThyIconModule, ThyDatePickerModule, ThyNavModule]; }, exports: function () { return [ThyDateRangeComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDateRangeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ThyActionMenuModule, ThySharedModule, ThyIconModule, ThyDatePickerModule, ThyNavModule],
                declarations: [ThyDateRangeComponent, OptionalDateRangesComponent],
                exports: [ThyDateRangeComponent],
                entryComponents: [OptionalDateRangesComponent]
            }]
    }], null, null); })();
