// ngx-tethys/esm2015/button/button-group.component.js
import { Component, HostBinding, Input, ElementRef, ViewEncapsulation } from '@angular/core';
import { UpdateHostClassService } from 'ngx-tethys/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';

const _c0 = ["*"];
const buttonGroupSizeMap = {
    sm: ['btn-group-sm'],
    md: ['btn-group-md'],
    lg: ['btn-group-lg'],
    xs: ['btn-group-xs']
};
const buttonGroupTypeMap = {
    'outline-primary': ['btn-group-outline-primary'],
    'outline-default': ['btn-group-outline-default']
};
export class ThyButtonGroupComponent {
    constructor(updateHostClassService, elementRef) {
        this.updateHostClassService = updateHostClassService;
        this.elementRef = elementRef;
        this.initialized = false;
        this._isButtonGroup = true;
        this.thyClearMinWidthClassName = false;
        this.updateHostClassService.initializeElement(elementRef.nativeElement);
    }
    set thySize(size) {
        this.size = size;
        if (this.initialized) {
            this.setClasses();
        }
    }
    set thyType(type) {
        this.type = type;
        if (this.initialized) {
            this.setClasses();
        }
    }
    set thyClearMinWidth(value) {
        this.thyClearMinWidthClassName = coerceBooleanProperty(value);
    }
    ngOnInit() {
        this.setClasses();
        this.initialized = true;
    }
    setClasses() {
        let classNames = [];
        if (buttonGroupTypeMap[this.type]) {
            classNames = [...buttonGroupTypeMap[this.type]];
        }
        if (buttonGroupSizeMap[this.size]) {
            classNames.push(buttonGroupSizeMap[this.size]);
        }
        this.updateHostClassService.updateClass(classNames);
    }
}
ThyButtonGroupComponent.ɵfac = function ThyButtonGroupComponent_Factory(t) { return new (t || ThyButtonGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyButtonGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyButtonGroupComponent, selectors: [["thy-button-group"]], hostVars: 4, hostBindings: function ThyButtonGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("btn-group", ctx._isButtonGroup)("btn-group-clear-min-width", ctx.thyClearMinWidthClassName);
    } }, inputs: { thySize: "thySize", thyType: "thyType", thyClearMinWidth: "thyClearMinWidth" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyButtonGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyButtonGroupComponent.ctorParameters = () => [
    { type: UpdateHostClassService },
    { type: ElementRef }
];
ThyButtonGroupComponent.propDecorators = {
    thySize: [{ type: Input }],
    thyType: [{ type: Input }],
    thyClearMinWidth: [{ type: Input }],
    _isButtonGroup: [{ type: HostBinding, args: ['class.btn-group',] }],
    thyClearMinWidthClassName: [{ type: HostBinding, args: [`class.btn-group-clear-min-width`,] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyButtonGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-button-group',
                template: '<ng-content></ng-content>',
                providers: [UpdateHostClassService],
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }]; }, { _isButtonGroup: [{
            type: HostBinding,
            args: ['class.btn-group']
        }], thyClearMinWidthClassName: [{
            type: HostBinding,
            args: [`class.btn-group-clear-min-width`]
        }], thySize: [{
            type: Input
        }], thyType: [{
            type: Input
        }], thyClearMinWidth: [{
            type: Input
        }] }); })();
