// ngx-tethys/esm2015/button/button.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyButtonGroupComponent } from './button-group.component';
import { ThyButtonIconComponent } from './button-icon.component';
import { ThyButtonComponent } from './button.component';
import * as ɵngcc0 from '@angular/core';
export class ThyButtonModule {
}
ThyButtonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyButtonModule });
ThyButtonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyButtonModule_Factory(t) { return new (t || ThyButtonModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyButtonModule, { declarations: function () { return [ThyButtonComponent, ThyButtonIconComponent, ThyButtonGroupComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyButtonComponent, ThyButtonIconComponent, ThyButtonGroupComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyButtonModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyButtonComponent, ThyButtonIconComponent, ThyButtonGroupComponent],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyButtonComponent, ThyButtonIconComponent, ThyButtonGroupComponent]
            }]
    }], null, null); })();
