// ngx-tethys/esm2015/result/result.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyResultComponent } from './result.component';
import * as ɵngcc0 from '@angular/core';
export class ThyResultModule {
}
ThyResultModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyResultModule });
ThyResultModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyResultModule_Factory(t) { return new (t || ThyResultModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyResultModule, { declarations: function () { return [ThyResultComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyResultComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyResultModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyResultComponent],
                imports: [CommonModule],
                exports: [ThyResultComponent]
            }]
    }], null, null); })();
