// ngx-tethys/esm2015/dropdown/dropdown.directive.js
import { Directive, HostBinding, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
const thyDropdown = 'thyDropdown';
const thyDropdownSplit = 'thyDropdownSplit';
export class ThyDropdownDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this._isDropdownClass = true;
        this._isDropdownSplitClass = false;
        const attrs = Array.from(elementRef.nativeElement.attributes);
        if (attrs.find((item) => item.name === thyDropdownSplit.toLocaleLowerCase())) {
            this._isDropdownSplitClass = true;
        }
    }
}
ThyDropdownDirective.ɵfac = function ThyDropdownDirective_Factory(t) { return new (t || ThyDropdownDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyDropdownDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDropdownDirective, selectors: [["", "thyDropdown", ""], ["", "thyDropdownSplit", ""]], hostVars: 4, hostBindings: function ThyDropdownDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-dropdown-toggle", ctx._isDropdownClass)("thy-dropdown-toggle-split", ctx._isDropdownSplitClass);
    } } });
ThyDropdownDirective.ctorParameters = () => [
    { type: ElementRef }
];
ThyDropdownDirective.propDecorators = {
    _isDropdownClass: [{ type: HostBinding, args: ['class.thy-dropdown-toggle',] }],
    _isDropdownSplitClass: [{ type: HostBinding, args: ['class.thy-dropdown-toggle-split',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDropdownDirective, [{
        type: Directive,
        args: [{
                selector: `[${thyDropdown}],[${thyDropdownSplit}]`
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { _isDropdownClass: [{
            type: HostBinding,
            args: ['class.thy-dropdown-toggle']
        }], _isDropdownSplitClass: [{
            type: HostBinding,
            args: ['class.thy-dropdown-toggle-split']
        }] }); })();
