// ngx-tethys/esm2015/dropdown/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyDropdownDirective } from './dropdown.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyDropdownModule {
}
ThyDropdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDropdownModule });
ThyDropdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDropdownModule_Factory(t) { return new (t || ThyDropdownModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDropdownModule, { declarations: function () { return [ThyDropdownDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyDropdownDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDropdownModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    ThyDropdownDirective
                ],
                exports: [
                    ThyDropdownDirective
                ]
            }]
    }], null, null); })();
