// ngx-tethys/esm2015/select/module.js
import { ThyEmptyModule } from 'ngx-tethys/empty';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyInputModule } from 'ngx-tethys/input';
import { ThyLabelModule } from 'ngx-tethys/label';
import { ThyLoadingModule } from 'ngx-tethys/loading';
import { ThyOptionModule, ThySelectCommonModule, ThySharedModule } from 'ngx-tethys/shared';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ThySelectCustomComponent } from './custom-select/custom-select.component';
import { ThySelectComponent } from './select.component';
import { THY_SELECT_SCROLL_STRATEGY_PROVIDER } from './select.config';
import * as ɵngcc0 from '@angular/core';
export class ThySelectModule {
}
ThySelectModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySelectModule });
ThySelectModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySelectModule_Factory(t) { return new (t || ThySelectModule)(); }, providers: [THY_SELECT_SCROLL_STRATEGY_PROVIDER], imports: [[
            CommonModule,
            FormsModule,
            ThyInputModule,
            ThyLabelModule,
            OverlayModule,
            ThyLoadingModule,
            ThySharedModule,
            ThyIconModule,
            ThyEmptyModule,
            ThySelectCommonModule,
            ThyOptionModule
        ], ThyOptionModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySelectModule, { declarations: function () { return [ThySelectComponent, ThySelectCustomComponent]; }, imports: function () { return [CommonModule,
        FormsModule,
        ThyInputModule,
        ThyLabelModule,
        OverlayModule,
        ThyLoadingModule,
        ThySharedModule,
        ThyIconModule,
        ThyEmptyModule,
        ThySelectCommonModule,
        ThyOptionModule]; }, exports: function () { return [ThySelectComponent, ThySelectCustomComponent, ThyOptionModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ThyInputModule,
                    ThyLabelModule,
                    OverlayModule,
                    ThyLoadingModule,
                    ThySharedModule,
                    ThyIconModule,
                    ThyEmptyModule,
                    ThySelectCommonModule,
                    ThyOptionModule
                ],
                declarations: [ThySelectComponent, ThySelectCustomComponent],
                exports: [ThySelectComponent, ThySelectCustomComponent, ThyOptionModule],
                providers: [THY_SELECT_SCROLL_STRATEGY_PROVIDER]
            }]
    }], null, null); })();
