// ngx-tethys/esm2015/select/select.component.js
import { Component, forwardRef, HostBinding, Input, ElementRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { UpdateHostClassService } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from 'ngx-tethys/input';
import * as ɵngcc4 from 'ngx-tethys/icon';
import * as ɵngcc5 from '@angular/common';

function ThySelectComponent_a_3_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 3);
    ɵngcc0.ɵɵlistener("click", function ThySelectComponent_a_3_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.clearSelectValue($event); });
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = ["*"];
const noop = () => { };
const ɵ0 = noop;
export class ThySelectComponent {
    constructor(elementRef, updateHostClassService) {
        this.elementRef = elementRef;
        this.updateHostClassService = updateHostClassService;
        // The internal data model
        this._innerValue = null;
        this._disabled = false;
        this._expandOptions = false;
        this.onTouchedCallback = noop;
        this.onChangeCallback = noop;
        this._isSelect = true;
        this.thyAllowClear = false;
        this.updateHostClassService.initializeElement(elementRef.nativeElement);
    }
    set thySize(value) {
        this._size = value;
    }
    writeValue(obj) {
        if (obj !== this._innerValue) {
            this._innerValue = obj;
        }
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    setDisabledState(isDisabled) {
        this._disabled = isDisabled;
    }
    ngModelChange() {
        this.onChangeCallback(this._innerValue);
    }
    ngOnInit() {
        const classes = this._size ? [`thy-select-${this._size}`] : [];
        this.updateHostClassService.updateClass(classes);
    }
    clearSelectValue(event) {
        event.stopPropagation();
        this._innerValue = '';
        this.onChangeCallback(this._innerValue);
    }
}
ThySelectComponent.ɵfac = function ThySelectComponent_Factory(t) { return new (t || ThySelectComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService)); };
ThySelectComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySelectComponent, selectors: [["thy-select"]], hostVars: 2, hostBindings: function ThySelectComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-select", ctx._isSelect);
    } }, inputs: { thyAllowClear: "thyAllowClear", thySize: "thySize", name: "name" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => ThySelectComponent),
                multi: true
            },
            UpdateHostClassService
        ])], ngContentSelectors: _c0, decls: 4, vars: 6, consts: [["thyInput", "", 3, "thySize", "disabled", "ngModel", "ngModelChange"], ["thyIconName", "angle-down"], ["class", "thy-select-remove remove-link", "href", "javascript:;", 3, "click", 4, "ngIf"], ["href", "javascript:;", 1, "thy-select-remove", "remove-link", 3, "click"], ["thyIconName", "close-circle-bold-fill", 1, "remove-link-icon"]], template: function ThySelectComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "select", 0);
        ɵngcc0.ɵɵlistener("ngModelChange", function ThySelectComponent_Template_select_ngModelChange_0_listener($event) { return ctx._innerValue = $event; })("ngModelChange", function ThySelectComponent_Template_select_ngModelChange_0_listener() { return ctx.ngModelChange(); });
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(2, "thy-icon", 1);
        ɵngcc0.ɵɵtemplate(3, ThySelectComponent_a_3_Template, 2, 0, "a", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-select-selection-allow-clear", ctx.thyAllowClear);
        ɵngcc0.ɵɵproperty("thySize", ctx._size)("disabled", ctx._disabled)("ngModel", ctx._innerValue);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyAllowClear && ctx._innerValue);
    } }, directives: [ɵngcc2.SelectControlValueAccessor, ɵngcc3.ThyInputDirective, ɵngcc2.NgControlStatus, ɵngcc2.NgModel, ɵngcc4.ThyIconComponent, ɵngcc5.NgIf], encapsulation: 2 });
ThySelectComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: UpdateHostClassService }
];
ThySelectComponent.propDecorators = {
    _isSelect: [{ type: HostBinding, args: ['class.thy-select',] }],
    thySize: [{ type: Input }],
    name: [{ type: Input }],
    thyAllowClear: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySelectComponent, [{
        type: Component,
        args: [{
                selector: 'thy-select',
                template: "<select\n  thyInput\n  [thySize]=\"_size\"\n  [disabled]=\"_disabled\"\n  [(ngModel)]=\"_innerValue\"\n  (ngModelChange)=\"ngModelChange()\"\n  [class.thy-select-selection-allow-clear]=\"thyAllowClear\"\n>\n  <ng-content></ng-content>\n</select>\n<thy-icon thyIconName=\"angle-down\"></thy-icon>\n<a\n  class=\"thy-select-remove remove-link\"\n  href=\"javascript:;\"\n  (click)=\"clearSelectValue($event)\"\n  *ngIf=\"thyAllowClear && _innerValue\"\n>\n  <thy-icon class=\"remove-link-icon\" thyIconName=\"close-circle-bold-fill\"></thy-icon>\n</a>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => ThySelectComponent),
                        multi: true
                    },
                    UpdateHostClassService
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.UpdateHostClassService }]; }, { _isSelect: [{
            type: HostBinding,
            args: ['class.thy-select']
        }], thyAllowClear: [{
            type: Input
        }], thySize: [{
            type: Input
        }], name: [{
            type: Input
        }] }); })();
export { ɵ0 };
