// ngx-tethys/esm2015/fullscreen/fullscreen.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyFullscreenComponent } from './fullscreen.component';
import * as ɵngcc0 from '@angular/core';
export class ThyFullscreenModule {
}
ThyFullscreenModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyFullscreenModule });
ThyFullscreenModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyFullscreenModule_Factory(t) { return new (t || ThyFullscreenModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyFullscreenModule, { declarations: function () { return [ThyFullscreenComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyFullscreenComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFullscreenModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyFullscreenComponent],
                imports: [CommonModule],
                exports: [ThyFullscreenComponent]
            }]
    }], null, null); })();
