// ngx-tethys/esm2015/fullscreen/fullscreen.service.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import { ThyFullscreenRef } from './fullscreen-ref';
import { ThyFullscreenMode } from './fullscreen.config';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as ɵngcc0 from '@angular/core';
export class ThyFullscreen {
    constructor(document) {
        this.document = document;
        this.fullscreenRefs = [];
    }
    /**
     * 开始全屏
     * @param config
     */
    launch(config) {
        config.mode = config.mode || ThyFullscreenMode.immersive;
        const fullscreenRef = new ThyFullscreenRef(this.document);
        fullscreenRef.fullscreenConfig = config;
        fullscreenRef.launch();
        this.fullscreenRefs.push(fullscreenRef);
        fullscreenRef.afterExited().subscribe(() => {
            const index = this.fullscreenRefs.indexOf(fullscreenRef);
            if (index > -1) {
                this.fullscreenRefs.splice(index, 1);
            }
        });
        return fullscreenRef;
    }
    /**
     * 退出最后一个 ThyFullscreenRef 的全屏
     * @param config
     */
    exit() {
        const lastFullscreenRef = this.fullscreenRefs[this.fullscreenRefs.length - 1];
        lastFullscreenRef.exit();
    }
}
ThyFullscreen.ɵfac = function ThyFullscreen_Factory(t) { return new (t || ThyFullscreen)(ɵngcc0.ɵɵinject(DOCUMENT)); };
ThyFullscreen.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyFullscreen_Factory() { return new ThyFullscreen(i0.ɵɵinject(i1.DOCUMENT)); }, token: ThyFullscreen, providedIn: "root" });
ThyFullscreen.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFullscreen, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, null); })();
