// ngx-tethys/esm2015/fullscreen/fullscreen.component.js
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { Subject } from 'rxjs';
import { ThyFullscreenMode } from './fullscreen.config';
import { ThyFullscreen } from './fullscreen.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './fullscreen.service';

const _c0 = ["*"];
export class ThyFullscreenComponent {
    constructor(elementRef, service) {
        this.elementRef = elementRef;
        this.service = service;
        this.thyMode = ThyFullscreenMode.immersive;
        this.thyFullscreenChange = new EventEmitter();
        this.ngUnsubscribe$ = new Subject();
        // 点击打开或关闭全屏
        this.handleFullscreen = () => {
            const targetElement = this.elementRef.nativeElement.querySelector('[fullscreen-target]');
            const containerElement = this.elementRef.nativeElement.querySelector('[fullscreen-container]');
            const fullscreen = targetElement.classList.contains('thy-fullscreen-active');
            if (fullscreen) {
                this.service.exit();
            }
            else {
                const fullscreenRef = this.service.launch({
                    mode: this.thyMode,
                    target: targetElement,
                    targetLaunchededClasse: this.thyFullscreenClasses,
                    emulatedContainer: containerElement
                });
                this.thyFullscreenChange.emit(true);
                fullscreenRef.afterExited().subscribe(() => {
                    this.thyFullscreenChange.emit(false);
                });
            }
        };
    }
    ngOnInit() { }
    ngAfterViewInit() {
        const btnLaunch = this.elementRef.nativeElement.querySelector('[fullscreen-launch]');
        if (btnLaunch) {
            btnLaunch.addEventListener('click', this.handleFullscreen);
        }
    }
    ngOnDestroy() {
        this.ngUnsubscribe$.next();
        this.ngUnsubscribe$.complete();
        const btnLaunch = this.elementRef.nativeElement.querySelector('[fullscreen-launch]');
        if (btnLaunch) {
            btnLaunch.removeEventListener('click', this.handleFullscreen);
        }
    }
}
ThyFullscreenComponent.ɵfac = function ThyFullscreenComponent_Factory(t) { return new (t || ThyFullscreenComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyFullscreen)); };
ThyFullscreenComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFullscreenComponent, selectors: [["thy-fullscreen"], ["", "thyFullscreen", ""]], inputs: { thyMode: "thyMode", thyFullscreenClasses: "thyFullscreenClasses" }, outputs: { thyFullscreenChange: "thyFullscreenChange" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyFullscreenComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyFullscreenComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyFullscreen }
];
ThyFullscreenComponent.propDecorators = {
    thyMode: [{ type: Input }],
    thyFullscreenClasses: [{ type: Input }],
    thyFullscreenChange: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFullscreenComponent, [{
        type: Component,
        args: [{
                selector: 'thy-fullscreen, [thyFullscreen]',
                template: "<ng-content> </ng-content>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyFullscreen }]; }, { thyMode: [{
            type: Input
        }], thyFullscreenChange: [{
            type: Output
        }], thyFullscreenClasses: [{
            type: Input
        }] }); })();
