// ngx-tethys/esm2015/statistic/statistic.component.js
import { Component, Input, HostBinding, ElementRef, TemplateRef, ContentChild, Renderer2 } from '@angular/core';
import { hexToRgb } from 'ngx-tethys/util';
import { UpdateHostClassService } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '@angular/common';

const _c0 = ["prefix"];
const _c1 = ["value"];
const _c2 = ["title"];
const _c3 = ["suffix"];
function ThyStatisticComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 7);
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r6 = ɵngcc0.ɵɵreference(9);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r6);
} }
function ThyStatisticComponent_span_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r8.thyPrefix);
} }
function ThyStatisticComponent_span_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 7);
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r9.prefixTemplate);
} }
function ThyStatisticComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 8);
    ɵngcc0.ɵɵtemplate(1, ThyStatisticComponent_span_2_span_1_Template, 2, 1, "span", 4);
    ɵngcc0.ɵɵtemplate(2, ThyStatisticComponent_span_2_ng_container_2_Template, 1, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.thyPrefix);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.prefixTemplate);
} }
function ThyStatisticComponent_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.thyValue);
} }
function ThyStatisticComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 7);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r3.valueTemplate);
} }
function ThyStatisticComponent_span_6_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r10.thySuffix);
} }
function ThyStatisticComponent_span_6_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 7);
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r11.suffixTemplate);
} }
function ThyStatisticComponent_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtemplate(1, ThyStatisticComponent_span_6_span_1_Template, 2, 1, "span", 4);
    ɵngcc0.ɵɵtemplate(2, ThyStatisticComponent_span_6_ng_container_2_Template, 1, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.thySuffix);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.suffixTemplate);
} }
function ThyStatisticComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 7);
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r6 = ɵngcc0.ɵɵreference(9);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r6);
} }
function ThyStatisticComponent_ng_template_8_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r12.thyTitle);
} }
function ThyStatisticComponent_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵtemplate(1, ThyStatisticComponent_ng_template_8_span_1_Template, 2, 1, "span", 4);
    ɵngcc0.ɵɵelementContainer(2, 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r7.thyTitle);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r7.titleTemplate);
} }
export class ThyStatisticComponent {
    constructor(elementRef, updateHostClassService, renderer) {
        this.elementRef = elementRef;
        this.updateHostClassService = updateHostClassService;
        this.renderer = renderer;
        this._initialized = false;
        this.class = true;
        this.thyValueStyle = {};
        this.thyTitlePosition = 'bottom';
        this.updateHostClassService.initializeElement(elementRef.nativeElement);
    }
    set prefix(value) {
        this.prefixTemplate = value;
    }
    set thyPrefixTemplate(value) {
        this.prefixTemplate = value;
    }
    set value(value) {
        this.valueTemplate = value;
    }
    set thyValueTemplate(value) {
        this.valueTemplate = value;
    }
    set title(value) {
        this.titleTemplate = value;
    }
    set thyTitleTemplate(value) {
        this.titleTemplate = value;
    }
    set suffix(value) {
        this.suffixTemplate = value;
    }
    set thySuffixTemplate(value) {
        this.suffixTemplate = value;
    }
    set thyShape(value) {
        this._shape = value;
        if (this._initialized) {
            this._setClassesByType();
        }
    }
    set thySize(value) {
        this._size = value;
        if (this._initialized) {
            this._setClassesByType();
        }
    }
    ngOnInit() {
        this._setClassesByType();
        this._initialized = true;
    }
    setColor(color) {
        this.renderer.setStyle(this.elementRef.nativeElement, 'color', color);
        if (this._shape === 'card') {
            this.renderer.setStyle(this.elementRef.nativeElement, 'border-color', color);
            this.renderer.setStyle(this.elementRef.nativeElement, 'background-color', hexToRgb(color, 0.05));
        }
    }
    _setClassesByType() {
        const classNames = [];
        if (this.thyColor) {
            if (RegExp(/^#([0-9a-fA-F]{6}|[0-9a-fA-F]{3})$/).test(this.thyColor)) {
                this.setColor(this.thyColor);
            }
            else {
                classNames.push(`thy-statistic-${this.thyColor}`);
            }
        }
        if (this._shape) {
            classNames.push(`thy-statistic-${this._shape}`);
        }
        if (!this._size) {
            this._size = 'default';
        }
        classNames.push(`thy-statistic-${this._size}`);
        this.renderer.setStyle(this.elementRef.nativeElement, 'font-size', this.thySize);
        this.updateHostClassService.updateClass(classNames);
    }
}
ThyStatisticComponent.ɵfac = function ThyStatisticComponent_Factory(t) { return new (t || ThyStatisticComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyStatisticComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStatisticComponent, selectors: [["thy-statistic"]], contentQueries: function ThyStatisticComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c2, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c3, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.prefix = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.value = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.title = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.suffix = _t.first);
    } }, hostVars: 2, hostBindings: function ThyStatisticComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-statistic", ctx.class);
    } }, inputs: { thyValueStyle: "thyValueStyle", thyTitlePosition: "thyTitlePosition", thyPrefixTemplate: "thyPrefixTemplate", thyValueTemplate: "thyValueTemplate", thyTitleTemplate: "thyTitleTemplate", thySuffixTemplate: "thySuffixTemplate", thyShape: "thyShape", thySize: "thySize", thyPrefix: "thyPrefix", thyValue: "thyValue", thyTitle: "thyTitle", thySuffix: "thySuffix", thyColor: "thyColor" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], decls: 10, vars: 7, consts: [[3, "ngTemplateOutlet", 4, "ngIf"], [1, "thy-statistic-content", 3, "ngStyle"], ["class", "thy-statistic-content-prefix", 4, "ngIf"], [1, "thy-statistic-content-value"], [4, "ngIf"], ["class", "thy-statistic-content-suffix", 4, "ngIf"], ["title", ""], [3, "ngTemplateOutlet"], [1, "thy-statistic-content-prefix"], [1, "thy-statistic-content-suffix"], [1, "thy-statistic-title"]], template: function ThyStatisticComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyStatisticComponent_ng_container_0_Template, 1, 1, "ng-container", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, ThyStatisticComponent_span_2_Template, 3, 2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "span", 3);
        ɵngcc0.ɵɵtemplate(4, ThyStatisticComponent_span_4_Template, 2, 1, "span", 4);
        ɵngcc0.ɵɵtemplate(5, ThyStatisticComponent_ng_container_5_Template, 1, 1, "ng-container", 0);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, ThyStatisticComponent_span_6_Template, 3, 2, "span", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, ThyStatisticComponent_ng_container_7_Template, 1, 1, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(8, ThyStatisticComponent_ng_template_8_Template, 3, 2, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", (ctx.thyTitle || ctx.titleTemplate) && ctx.thyTitlePosition === "top");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngStyle", ctx.thyValueStyle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyPrefix || ctx.prefixTemplate);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.valueTemplate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.valueTemplate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thySuffix || ctx.suffixTemplate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.thyTitle || ctx.titleTemplate) && ctx.thyTitlePosition === "bottom");
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgStyle, ɵngcc2.NgTemplateOutlet], encapsulation: 2 });
ThyStatisticComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: UpdateHostClassService },
    { type: Renderer2 }
];
ThyStatisticComponent.propDecorators = {
    class: [{ type: HostBinding, args: [`class.thy-statistic`,] }],
    thyValueStyle: [{ type: Input }],
    thyPrefix: [{ type: Input }],
    prefix: [{ type: ContentChild, args: ['prefix', { static: true },] }],
    thyPrefixTemplate: [{ type: Input }],
    thyValue: [{ type: Input }],
    value: [{ type: ContentChild, args: ['value', { static: true },] }],
    thyValueTemplate: [{ type: Input }],
    thyTitle: [{ type: Input }],
    title: [{ type: ContentChild, args: ['title', { static: true },] }],
    thyTitleTemplate: [{ type: Input }],
    thySuffix: [{ type: Input }],
    suffix: [{ type: ContentChild, args: ['suffix', { static: true },] }],
    thySuffixTemplate: [{ type: Input }],
    thyTitlePosition: [{ type: Input }],
    thyShape: [{ type: Input }],
    thyColor: [{ type: Input }],
    thySize: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStatisticComponent, [{
        type: Component,
        args: [{
                selector: 'thy-statistic',
                template: "<ng-container [ngTemplateOutlet]=\"title\" *ngIf=\"(thyTitle || titleTemplate) && thyTitlePosition === 'top'\"></ng-container>\n<div class=\"thy-statistic-content\" [ngStyle]=\"thyValueStyle\">\n  <span *ngIf=\"thyPrefix || prefixTemplate\" class=\"thy-statistic-content-prefix\">\n    <span *ngIf=\"thyPrefix\">{{ thyPrefix }}</span>\n    <ng-container *ngIf=\"prefixTemplate\" [ngTemplateOutlet]=\"prefixTemplate\"></ng-container>\n  </span>\n\n  <span class=\"thy-statistic-content-value\">\n    <span *ngIf=\"!valueTemplate\">{{ thyValue }}</span>\n    <ng-container *ngIf=\"valueTemplate\" [ngTemplateOutlet]=\"valueTemplate\"></ng-container>\n  </span>\n\n  <span *ngIf=\"thySuffix || suffixTemplate\" class=\"thy-statistic-content-suffix\">\n    <span *ngIf=\"thySuffix\">{{ thySuffix }}</span>\n    <ng-container *ngIf=\"suffixTemplate\" [ngTemplateOutlet]=\"suffixTemplate\"></ng-container>\n  </span>\n</div>\n<ng-container [ngTemplateOutlet]=\"title\" *ngIf=\"(thyTitle || titleTemplate) && thyTitlePosition === 'bottom'\"></ng-container>\n\n<ng-template #title>\n  <div class=\"thy-statistic-title\">\n    <span *ngIf=\"thyTitle\">{{ thyTitle }}</span>\n    <ng-container [ngTemplateOutlet]=\"titleTemplate\"></ng-container>\n  </div>\n</ng-template>\n",
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.Renderer2 }]; }, { class: [{
            type: HostBinding,
            args: [`class.thy-statistic`]
        }], thyValueStyle: [{
            type: Input
        }], thyTitlePosition: [{
            type: Input
        }], prefix: [{
            type: ContentChild,
            args: ['prefix', { static: true }]
        }], thyPrefixTemplate: [{
            type: Input
        }], value: [{
            type: ContentChild,
            args: ['value', { static: true }]
        }], thyValueTemplate: [{
            type: Input
        }], title: [{
            type: ContentChild,
            args: ['title', { static: true }]
        }], thyTitleTemplate: [{
            type: Input
        }], suffix: [{
            type: ContentChild,
            args: ['suffix', { static: true }]
        }], thySuffixTemplate: [{
            type: Input
        }], thyShape: [{
            type: Input
        }], thySize: [{
            type: Input
        }], thyPrefix: [{
            type: Input
        }], thyValue: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], thySuffix: [{
            type: Input
        }], thyColor: [{
            type: Input
        }] }); })();
