// ngx-tethys/esm2015/statistic/statistic.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyStatisticComponent } from './statistic.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyStatisticModule {
}
ThyStatisticModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyStatisticModule });
ThyStatisticModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyStatisticModule_Factory(t) { return new (t || ThyStatisticModule)(); }, providers: [], imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyStatisticModule, { declarations: function () { return [ThyStatisticComponent]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyStatisticComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStatisticModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyStatisticComponent],
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyStatisticComponent],
                providers: []
            }]
    }], null, null); })();
