// ngx-tethys/esm2015/form/validator/confirm-validator.directive.js
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
export function confirmValidator(value) {
    return (control) => {
        const isEqual = value === control.value;
        return !isEqual ? { confirm: { value: value, actual: control.value } } : null;
    };
}
export class ThyConfirmValidatorDirective {
    validate(control) {
        return confirmValidator(this.confirm)(control);
    }
}
ThyConfirmValidatorDirective.ɵfac = function ThyConfirmValidatorDirective_Factory(t) { return new (t || ThyConfirmValidatorDirective)(); };
ThyConfirmValidatorDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyConfirmValidatorDirective, selectors: [["", "confirm", ""]], inputs: { confirm: "confirm" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: ThyConfirmValidatorDirective, multi: true }])] });
ThyConfirmValidatorDirective.propDecorators = {
    confirm: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyConfirmValidatorDirective, [{
        type: Directive,
        args: [{
                selector: '[confirm]',
                providers: [{ provide: NG_VALIDATORS, useExisting: ThyConfirmValidatorDirective, multi: true }]
            }]
    }], null, { confirm: [{
            type: Input
        }] }); })();
