// ngx-tethys/esm2015/form/validator/max-validator.directive.js
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS, Validators } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
export class ThyMaxDirective {
    set max(value) {
        this._validator = Validators.max(parseFloat(value));
    }
    validate(control) {
        return this._validator(control);
    }
}
ThyMaxDirective.ɵfac = function ThyMaxDirective_Factory(t) { return new (t || ThyMaxDirective)(); };
ThyMaxDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMaxDirective, selectors: [["", "max", "", "formControlName", ""], ["", "max", "", "formControl", ""], ["", "max", "", "ngModel", ""]], inputs: { max: "max" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => ThyMaxDirective),
                multi: true
            }
        ])] });
ThyMaxDirective.propDecorators = {
    max: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMaxDirective, [{
        type: Directive,
        args: [{
                selector: '[max][formControlName],[max][formControl],[max][ngModel]',
                providers: [
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => ThyMaxDirective),
                        multi: true
                    }
                ]
            }]
    }], null, { max: [{
            type: Input
        }] }); })();
