// ngx-tethys/esm2015/form/validator/unique-validator.directive.js
import { NG_ASYNC_VALIDATORS } from '@angular/forms';
import { Directive, Input, Optional, ElementRef } from '@angular/core';
import { of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { ThyFormDirective } from '../form.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../form.directive';
export class ThyUniqueCheckValidator {
    constructor(elementRef, thyForm) {
        this.elementRef = elementRef;
        this.thyForm = thyForm;
    }
    validate(ctrl) {
        return this.thyUniqueCheck(ctrl.value).pipe(map((failed) => {
            setTimeout(() => {
                if (failed && this.thyForm && this.elementRef.nativeElement.name) {
                    this.thyForm.validator.validateControl(this.elementRef.nativeElement.name);
                }
            });
            return failed ? { thyUniqueCheck: failed } : null;
        }), catchError(() => of(null)));
    }
}
ThyUniqueCheckValidator.ɵfac = function ThyUniqueCheckValidator_Factory(t) { return new (t || ThyUniqueCheckValidator)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyFormDirective, 8)); };
ThyUniqueCheckValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyUniqueCheckValidator, selectors: [["", "thyUniqueCheck", ""]], inputs: { thyUniqueCheck: "thyUniqueCheck" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_ASYNC_VALIDATORS,
                useExisting: ThyUniqueCheckValidator,
                multi: true
            }
        ])] });
ThyUniqueCheckValidator.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyFormDirective, decorators: [{ type: Optional }] }
];
ThyUniqueCheckValidator.propDecorators = {
    thyUniqueCheck: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyUniqueCheckValidator, [{
        type: Directive,
        args: [{
                selector: '[thyUniqueCheck]',
                providers: [
                    {
                        provide: NG_ASYNC_VALIDATORS,
                        useExisting: ThyUniqueCheckValidator,
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyFormDirective, decorators: [{
                type: Optional
            }] }]; }, { thyUniqueCheck: [{
            type: Input
        }] }); })();
