// ngx-tethys/esm2015/switch/switch.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThySwitchComponent } from './switch.component';
import * as ɵngcc0 from '@angular/core';
export class ThySwitchModule {
}
ThySwitchModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySwitchModule });
ThySwitchModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySwitchModule_Factory(t) { return new (t || ThySwitchModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySwitchModule, { declarations: function () { return [ThySwitchComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThySwitchComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySwitchModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThySwitchComponent
                ],
                entryComponents: [
                    ThySwitchComponent
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ThySwitchComponent
                ]
            }]
    }], null, null); })();
