// ngx-tethys/esm2015/alert/alert.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyAlertComponent } from './alert.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyAlertActionItemDirective } from './alert.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyAlertModule {
}
ThyAlertModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAlertModule });
ThyAlertModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAlertModule_Factory(t) { return new (t || ThyAlertModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAlertModule, { declarations: function () { return [ThyAlertComponent, ThyAlertActionItemDirective]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyAlertComponent, ThyAlertActionItemDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAlertModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyAlertComponent, ThyAlertActionItemDirective],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyAlertComponent, ThyAlertActionItemDirective]
            }]
    }], null, null); })();
