// ngx-tethys/esm2015/alert/alert.directive.js
import { Directive, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ThyAlertActionItemDirective {
    constructor() {
        this.class = true;
    }
}
ThyAlertActionItemDirective.ɵfac = function ThyAlertActionItemDirective_Factory(t) { return new (t || ThyAlertActionItemDirective)(); };
ThyAlertActionItemDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyAlertActionItemDirective, selectors: [["", "thyAlertActionItem", ""]], hostVars: 2, hostBindings: function ThyAlertActionItemDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-alert-action", ctx.class);
    } } });
ThyAlertActionItemDirective.ctorParameters = () => [];
ThyAlertActionItemDirective.propDecorators = {
    class: [{ type: HostBinding, args: ['class.thy-alert-action',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAlertActionItemDirective, [{
        type: Directive,
        args: [{ selector: '[thyAlertActionItem]' }]
    }], function () { return []; }, { class: [{
            type: HostBinding,
            args: ['class.thy-alert-action']
        }] }); })();
