// ngx-tethys/esm2015/breadcrumb/breadcrumb-item.component.js
import { Component, ChangeDetectionStrategy, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/icon';

const _c0 = ["*"];
export class ThyBreadcrumbItemComponent {
    constructor() {
        this._isBreadcrumbItem = true;
    }
}
ThyBreadcrumbItemComponent.ɵfac = function ThyBreadcrumbItemComponent_Factory(t) { return new (t || ThyBreadcrumbItemComponent)(); };
ThyBreadcrumbItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyBreadcrumbItemComponent, selectors: [["thy-breadcrumb-item"], ["", "thyBreadcrumbItem", ""]], hostVars: 2, hostBindings: function ThyBreadcrumbItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-breadcrumb-item", ctx._isBreadcrumbItem);
    } }, exportAs: ["ThyBreadcrumbItem"], ngContentSelectors: _c0, decls: 2, vars: 0, consts: [["thyIconName", "angle-right", 1, "default-icon"]], template: function ThyBreadcrumbItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelement(1, "thy-icon", 0);
    } }, directives: [ɵngcc1.ThyIconComponent], encapsulation: 2, changeDetection: 0 });
ThyBreadcrumbItemComponent.propDecorators = {
    _isBreadcrumbItem: [{ type: HostBinding, args: [`class.thy-breadcrumb-item`,] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBreadcrumbItemComponent, [{
        type: Component,
        args: [{
                selector: 'thy-breadcrumb-item,[thyBreadcrumbItem]',
                template: '<ng-content></ng-content><thy-icon class="default-icon" thyIconName="angle-right"></thy-icon>',
                exportAs: 'ThyBreadcrumbItem',
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { _isBreadcrumbItem: [{
            type: HostBinding,
            args: [`class.thy-breadcrumb-item`]
        }] }); })();
