// ngx-tethys/esm2015/breadcrumb/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyBreadcrumbComponent } from './breadcrumb.component';
import { ThyBreadcrumbItemComponent } from './breadcrumb-item.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyBreadcrumbModule {
}
ThyBreadcrumbModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyBreadcrumbModule });
ThyBreadcrumbModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyBreadcrumbModule_Factory(t) { return new (t || ThyBreadcrumbModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyBreadcrumbModule, { declarations: function () { return [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBreadcrumbModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent]
            }]
    }], null, null); })();
