// ngx-tethys/esm2015/drag-drop/drag-handle.directive.js
import { Directive, ElementRef, Input, Optional } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ThyDragDirective } from './drag.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './drag.directive';
export class ThyDragHandleDirective {
    constructor(element, drag) {
        this.element = element;
        this._disabled = false;
        if (drag) {
            drag.dragRef.withHandles(this);
        }
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = coerceBooleanProperty(value);
    }
}
ThyDragHandleDirective.ɵfac = function ThyDragHandleDirective_Factory(t) { return new (t || ThyDragHandleDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyDragDirective, 8)); };
ThyDragHandleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDragHandleDirective, selectors: [["thy-drag-handle"], ["", "thyDragHandle", ""]], inputs: { disabled: ["thyDisabled", "disabled"] } });
ThyDragHandleDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyDragDirective, decorators: [{ type: Optional }] }
];
ThyDragHandleDirective.propDecorators = {
    disabled: [{ type: Input, args: ['thyDisabled',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDragHandleDirective, [{
        type: Directive,
        args: [{
                selector: 'thy-drag-handle,[thyDragHandle]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyDragDirective, decorators: [{
                type: Optional
            }] }]; }, { disabled: [{
            type: Input,
            args: ['thyDisabled']
        }] }); })();
