// ngx-tethys/esm2015/drag-drop/drag-content.directive.js
import { Directive, ElementRef, HostBinding, Optional } from '@angular/core';
import { ThyDragDirective } from './drag.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './drag.directive';
export class ThyDragContentDirective {
    constructor(element, drag) {
        this.element = element;
        this.contentClass = true;
        if (drag) {
            drag.dragRef.withContentElement(this.element);
        }
    }
}
ThyDragContentDirective.ɵfac = function ThyDragContentDirective_Factory(t) { return new (t || ThyDragContentDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyDragDirective, 8)); };
ThyDragContentDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDragContentDirective, selectors: [["thy-drag-content"], ["", "thyDragContent", ""]], hostVars: 2, hostBindings: function ThyDragContentDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-drag-content", ctx.contentClass);
    } } });
ThyDragContentDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyDragDirective, decorators: [{ type: Optional }] }
];
ThyDragContentDirective.propDecorators = {
    contentClass: [{ type: HostBinding, args: ['class.thy-drag-content',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDragContentDirective, [{
        type: Directive,
        args: [{
                selector: 'thy-drag-content,[thyDragContent]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyDragDirective, decorators: [{
                type: Optional
            }] }]; }, { contentClass: [{
            type: HostBinding,
            args: ['class.thy-drag-content']
        }] }); })();
