// ngx-tethys/esm2015/drag-drop/drag.directive.js
import { Directive, NgZone, ElementRef, HostBinding, Input, Inject, Optional } from '@angular/core';
import { DragRef } from './drag-ref';
import { DOCUMENT } from '@angular/common';
import { ThyDragDropService } from './drag-drop.service';
import { mixinUnsubscribe, MixinBase } from 'ngx-tethys/core';
import { THY_DROP_CONTAINER_DIRECTIVE } from './drop-container.class';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './drag-drop.service';
const _MixinBase = mixinUnsubscribe(MixinBase);
export class ThyDragDirective extends _MixinBase {
    constructor(document, container, ngZone, elementRef, service) {
        super();
        this.container = container;
        this.elementRef = elementRef;
        this.isDraggable = true;
        this._disabled = false;
        this.dragRef = new DragRef(elementRef, this, container, service, document, ngZone);
    }
    set dragData(data) {
        this.data = data;
    }
    set disabled(isDisabled) {
        this._disabled = isDisabled;
        if (isDisabled) {
            this.isDraggable = false;
        }
    }
    get disabled() {
        return this._disabled;
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        this.dragRef.dispose();
    }
}
ThyDragDirective.ɵfac = function ThyDragDirective_Factory(t) { return new (t || ThyDragDirective)(ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(THY_DROP_CONTAINER_DIRECTIVE, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyDragDropService)); };
ThyDragDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDragDirective, selectors: [["thy-drag"], ["", "thyDrag", ""]], hostVars: 1, hostBindings: function ThyDragDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("draggable", ctx.isDraggable);
    } }, inputs: { dragData: ["thyDrag", "dragData"], data: ["thyDragData", "data"], disabled: ["thyDragDisabled", "disabled"] }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
ThyDragDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [THY_DROP_CONTAINER_DIRECTIVE,] }] },
    { type: NgZone },
    { type: ElementRef },
    { type: ThyDragDropService }
];
ThyDragDirective.propDecorators = {
    dragData: [{ type: Input, args: ['thyDrag',] }],
    data: [{ type: Input, args: ['thyDragData',] }],
    isDraggable: [{ type: HostBinding, args: ['attr.draggable',] }],
    disabled: [{ type: Input, args: ['thyDragDisabled',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDragDirective, [{
        type: Directive,
        args: [{ selector: 'thy-drag,[thyDrag]' }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [THY_DROP_CONTAINER_DIRECTIVE]
            }] }, { type: ɵngcc0.NgZone }, { type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyDragDropService }]; }, { isDraggable: [{
            type: HostBinding,
            args: ['attr.draggable']
        }], dragData: [{
            type: Input,
            args: ['thyDrag']
        }], data: [{
            type: Input,
            args: ['thyDragData']
        }], disabled: [{
            type: Input,
            args: ['thyDragDisabled']
        }] }); })();
