// ngx-tethys/esm2015/drag-drop/module.js
import { NgModule } from '@angular/core';
import { ThyDragDirective } from './drag.directive';
import { ThyDragHandleDirective } from './drag-handle.directive';
import { ThyDragContentDirective } from './drag-content.directive';
import { ThyDropContainerDirective } from './drop-container.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyDragDropModule {
}
ThyDragDropModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDragDropModule });
ThyDragDropModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDragDropModule_Factory(t) { return new (t || ThyDragDropModule)(); }, providers: [], imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDragDropModule, { declarations: function () { return [ThyDropContainerDirective, ThyDragDirective, ThyDragHandleDirective, ThyDragContentDirective]; }, exports: function () { return [ThyDropContainerDirective, ThyDragDirective, ThyDragHandleDirective, ThyDragContentDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDragDropModule, [{
        type: NgModule,
        args: [{
                imports: [],
                exports: [ThyDropContainerDirective, ThyDragDirective, ThyDragHandleDirective, ThyDragContentDirective],
                declarations: [ThyDropContainerDirective, ThyDragDirective, ThyDragHandleDirective, ThyDragContentDirective],
                providers: []
            }]
    }], null, null); })();
