// ngx-tethys/esm2015/popover/header/popover-header.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { ThyTranslate } from 'ngx-tethys/core';
import { ThyPopover } from '../popover.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '../popover.service';
import * as ɵngcc3 from 'ngx-tethys/icon';
export class ThyPopoverHeaderComponent {
    constructor(translate, thyPopover) {
        this.translate = translate;
        this.thyPopover = thyPopover;
        this.thyClosed = new EventEmitter();
    }
    set thyTitleTranslationKey(key) {
        if (key && !this.thyTitle) {
            this.thyTitle = this.translate.instant(key);
        }
    }
    close(event) {
        this.thyClosed.emit(event);
        this.thyPopover.close();
    }
}
ThyPopoverHeaderComponent.ɵfac = function ThyPopoverHeaderComponent_Factory(t) { return new (t || ThyPopoverHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyPopover)); };
ThyPopoverHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPopoverHeaderComponent, selectors: [["thy-popover-header"]], inputs: { thyTitleTranslationKey: "thyTitleTranslationKey", thyTitle: "thyTitle" }, outputs: { thyClosed: "thyClosed" }, exportAs: ["thyPopoverHeader"], decls: 5, vars: 1, consts: [[1, "thy-popover-header"], [1, "thy-popover-title"], ["type", "button", 1, "close", 3, "click"], ["thyIconName", "close-bold"]], template: function ThyPopoverHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "h5", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "button", 2);
        ɵngcc0.ɵɵlistener("click", function ThyPopoverHeaderComponent_Template_button_click_3_listener($event) { return ctx.close($event); });
        ɵngcc0.ɵɵelement(4, "thy-icon", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyTitle);
    } }, directives: [ɵngcc3.ThyIconComponent], encapsulation: 2 });
ThyPopoverHeaderComponent.ctorParameters = () => [
    { type: ThyTranslate },
    { type: ThyPopover }
];
ThyPopoverHeaderComponent.propDecorators = {
    thyTitle: [{ type: Input }],
    thyTitleTranslationKey: [{ type: Input }],
    thyClosed: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-popover-header',
                template: "<div class=\"thy-popover-header\">\n  <h5 class=\"thy-popover-title\">{{ thyTitle }}</h5>\n  <button type=\"button\" class=\"close\" (click)=\"close($event)\">\n    <thy-icon thyIconName=\"close-bold\"></thy-icon>\n  </button>\n</div>\n",
                exportAs: 'thyPopoverHeader'
            }]
    }], function () { return [{ type: ɵngcc1.ThyTranslate }, { type: ɵngcc2.ThyPopover }]; }, { thyClosed: [{
            type: Output
        }], thyTitleTranslationKey: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }] }); })();
