// ngx-tethys/esm2015/progress/progress.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyTooltipModule } from 'ngx-tethys/tooltip';
import { ThyProgressBarComponent } from './bar/progress-bar.component';
import { ThyProgressComponent } from './progress.component';
import * as ɵngcc0 from '@angular/core';
export class ThyProgressModule {
}
ThyProgressModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyProgressModule });
ThyProgressModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyProgressModule_Factory(t) { return new (t || ThyProgressModule)(); }, imports: [[CommonModule, ThyTooltipModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyProgressModule, { declarations: function () { return [ThyProgressComponent, ThyProgressBarComponent]; }, imports: function () { return [CommonModule, ThyTooltipModule]; }, exports: function () { return [ThyProgressComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyProgressModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyProgressComponent, ThyProgressBarComponent],
                imports: [CommonModule, ThyTooltipModule],
                exports: [ThyProgressComponent]
            }]
    }], null, null); })();
