// ngx-tethys/esm2015/form/form-group-error/form-group-error.component.js
import { __decorate, __metadata } from "tslib";
import { InputBoolean } from 'ngx-tethys/core';
import { Component, HostBinding, ViewEncapsulation, Optional, Input } from '@angular/core';
import { ThyFormDirective } from '../form.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../form.directive';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-tethys/alert';

function ThyFormGroupErrorComponent_thy_alert_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-alert", 2);
} if (rf & 2) {
    const error_r1 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyMessage", error_r1);
} }
const _c0 = function (a0) { return { "col-sm-10 offset-sm-2 col-form-control": a0 }; };
export class ThyFormGroupErrorComponent {
    constructor(thyParentForm) {
        this.thyParentForm = thyParentForm;
        this.thyShowFirst = true;
        this.isHorizontal = true;
    }
    set thyErrors(errors) {
        this.errors = errors;
    }
    get thyErrors() {
        const errors = this.errors || this.thyParentForm.validator.errors;
        return errors && errors.length > 0 && this.thyShowFirst ? [errors[0]] : errors;
    }
    get _isFormGroup() {
        return this.thyErrors && this.thyErrors.length > 0;
    }
    ngOnInit() {
        if (this.thyParentForm) {
            this.isHorizontal = this.thyParentForm.isHorizontal;
        }
    }
}
ThyFormGroupErrorComponent.ɵfac = function ThyFormGroupErrorComponent_Factory(t) { return new (t || ThyFormGroupErrorComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyFormDirective, 8)); };
ThyFormGroupErrorComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFormGroupErrorComponent, selectors: [["thy-form-group-error"]], hostVars: 4, hostBindings: function ThyFormGroupErrorComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("row", ctx.isHorizontal)("form-group", ctx._isFormGroup);
    } }, inputs: { thyShowFirst: "thyShowFirst", thyErrors: "thyErrors" }, decls: 2, vars: 4, consts: [[3, "ngClass"], ["thyType", "danger", 3, "thyMessage", 4, "ngFor", "ngForOf"], ["thyType", "danger", 3, "thyMessage"]], template: function ThyFormGroupErrorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyFormGroupErrorComponent_thy_alert_1_Template, 1, 1, "thy-alert", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c0, ctx.isHorizontal));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.thyErrors);
    } }, directives: [ɵngcc2.NgClass, ɵngcc2.NgForOf, ɵngcc3.ThyAlertComponent], encapsulation: 2 });
ThyFormGroupErrorComponent.ctorParameters = () => [
    { type: ThyFormDirective, decorators: [{ type: Optional }] }
];
ThyFormGroupErrorComponent.propDecorators = {
    thyShowFirst: [{ type: Input }],
    thyErrors: [{ type: Input }],
    _isFormGroup: [{ type: HostBinding, args: ['class.form-group',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.row',] }]
};
__decorate([
    InputBoolean(),
    __metadata("design:type", Object)
], ThyFormGroupErrorComponent.prototype, "thyShowFirst", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupErrorComponent, [{
        type: Component,
        args: [{
                selector: 'thy-form-group-error',
                template: "<div [ngClass]=\"{ 'col-sm-10 offset-sm-2 col-form-control': isHorizontal }\">\n  <thy-alert *ngFor=\"let error of thyErrors\" thyType=\"danger\" [thyMessage]=\"error\"></thy-alert>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc1.ThyFormDirective, decorators: [{
                type: Optional
            }] }]; }, { thyShowFirst: [{
            type: Input
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.row']
        }], thyErrors: [{
            type: Input
        }], _isFormGroup: [{
            type: HostBinding,
            args: ['class.form-group']
        }] }); })();
