// ngx-tethys/esm2015/autocomplete/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThyAutocompleteTriggerDirective } from './autocomplete.trigger.directive';
import { ThyInputModule } from 'ngx-tethys/input';
import { ThyAutocompleteComponent } from './autocomplete.component';
import { ThyLabelModule } from 'ngx-tethys/label';
import { OverlayModule } from '@angular/cdk/overlay';
import { ThyLoadingModule } from 'ngx-tethys/loading';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyEmptyModule } from 'ngx-tethys/empty';
import { ThyAutocompleteContainerComponent } from './overlay/autocomplete-container.component';
import { THY_AUTOCOMPLETE_DEFAULT_CONFIG_PROVIDER } from './overlay/autocomplete.config';
import { ThyOptionModule } from 'ngx-tethys/shared';
import { PortalModule } from '@angular/cdk/portal';
import * as ɵngcc0 from '@angular/core';
export class ThyAutocompleteModule {
}
ThyAutocompleteModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAutocompleteModule });
ThyAutocompleteModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAutocompleteModule_Factory(t) { return new (t || ThyAutocompleteModule)(); }, providers: [THY_AUTOCOMPLETE_DEFAULT_CONFIG_PROVIDER], imports: [[
            CommonModule,
            FormsModule,
            ThyInputModule,
            ThyLabelModule,
            OverlayModule,
            PortalModule,
            ThyLoadingModule,
            ThySharedModule,
            ThyIconModule,
            ThyEmptyModule,
            ThyOptionModule
        ], ThyOptionModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAutocompleteModule, { declarations: function () { return [ThyAutocompleteTriggerDirective, ThyAutocompleteComponent, ThyAutocompleteContainerComponent]; }, imports: function () { return [CommonModule,
        FormsModule,
        ThyInputModule,
        ThyLabelModule,
        OverlayModule,
        PortalModule,
        ThyLoadingModule,
        ThySharedModule,
        ThyIconModule,
        ThyEmptyModule,
        ThyOptionModule]; }, exports: function () { return [ThyAutocompleteTriggerDirective, ThyAutocompleteComponent, ThyAutocompleteContainerComponent, ThyOptionModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAutocompleteModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ThyInputModule,
                    ThyLabelModule,
                    OverlayModule,
                    PortalModule,
                    ThyLoadingModule,
                    ThySharedModule,
                    ThyIconModule,
                    ThyEmptyModule,
                    ThyOptionModule
                ],
                declarations: [ThyAutocompleteTriggerDirective, ThyAutocompleteComponent, ThyAutocompleteContainerComponent],
                entryComponents: [ThyAutocompleteContainerComponent],
                exports: [ThyAutocompleteTriggerDirective, ThyAutocompleteComponent, ThyAutocompleteContainerComponent, ThyOptionModule],
                providers: [THY_AUTOCOMPLETE_DEFAULT_CONFIG_PROVIDER]
            }]
    }], null, null); })();
