// ngx-tethys/esm2015/card/card.component.js
import { Component, HostBinding, Input } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyCardComponent {
    constructor() {
        this.thyCardClass = true;
        this.clearLeftRightPadding = false;
        this._thyDivided = false;
    }
    set thyHasLeftRightPadding(value) {
        this.clearLeftRightPadding = !coerceBooleanProperty(value);
    }
    set thyDivided(value) {
        this._thyDivided = coerceBooleanProperty(value);
    }
}
ThyCardComponent.ɵfac = function ThyCardComponent_Factory(t) { return new (t || ThyCardComponent)(); };
ThyCardComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCardComponent, selectors: [["thy-card"]], hostVars: 6, hostBindings: function ThyCardComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-card", ctx.thyCardClass)("thy-card--clear-left-right-padding", ctx.clearLeftRightPadding)("thy-card--divided", ctx._thyDivided);
    } }, inputs: { thyHasLeftRightPadding: "thyHasLeftRightPadding", thyDivided: "thyDivided" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyCardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyCardComponent.propDecorators = {
    thyCardClass: [{ type: HostBinding, args: ['class.thy-card',] }],
    clearLeftRightPadding: [{ type: HostBinding, args: ['class.thy-card--clear-left-right-padding',] }],
    _thyDivided: [{ type: HostBinding, args: ['class.thy-card--divided',] }],
    thyHasLeftRightPadding: [{ type: Input, args: ['thyHasLeftRightPadding',] }],
    thyDivided: [{ type: Input, args: ['thyDivided',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardComponent, [{
        type: Component,
        args: [{
                selector: 'thy-card',
                template: `
        <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { thyCardClass: [{
            type: HostBinding,
            args: ['class.thy-card']
        }], clearLeftRightPadding: [{
            type: HostBinding,
            args: ['class.thy-card--clear-left-right-padding']
        }], _thyDivided: [{
            type: HostBinding,
            args: ['class.thy-card--divided']
        }], thyHasLeftRightPadding: [{
            type: Input,
            args: ['thyHasLeftRightPadding']
        }], thyDivided: [{
            type: Input,
            args: ['thyDivided']
        }] }); })();
