// ngx-tethys/esm2015/card/card.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyCardComponent } from './card.component';
import { ThyCardHeaderComponent } from './header.component';
import { ThyCardContentComponent } from './content.component';
import * as ɵngcc0 from '@angular/core';
export class ThyCardModule {
}
ThyCardModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCardModule });
ThyCardModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCardModule_Factory(t) { return new (t || ThyCardModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCardModule, { declarations: function () { return [ThyCardComponent,
        ThyCardHeaderComponent,
        ThyCardContentComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyCardComponent,
        ThyCardHeaderComponent,
        ThyCardContentComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyCardComponent,
                    ThyCardHeaderComponent,
                    ThyCardContentComponent,
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ThyCardComponent,
                    ThyCardHeaderComponent,
                    ThyCardContentComponent,
                ]
            }]
    }], null, null); })();
