// ngx-tethys/esm2015/card/content.component.js
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { Component, HostBinding, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyCardContentComponent {
    constructor() {
        this.thyCardContentClass = true;
        this.scrollClassName = false;
        this._thySizeSm = false;
    }
    set thyScroll(value) {
        this.scrollClassName = coerceBooleanProperty(value);
    }
    set thySize(value) {
        this._thySizeSm = value === 'sm';
    }
    ngOnInit() { }
}
ThyCardContentComponent.ɵfac = function ThyCardContentComponent_Factory(t) { return new (t || ThyCardContentComponent)(); };
ThyCardContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCardContentComponent, selectors: [["thy-card-content"]], hostVars: 6, hostBindings: function ThyCardContentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-card-content", ctx.thyCardContentClass)("thy-card-content--scroll", ctx.scrollClassName)("thy-card-content--sm", ctx._thySizeSm);
    } }, inputs: { thyScroll: "thyScroll", thySize: "thySize" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyCardContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyCardContentComponent.propDecorators = {
    thyCardContentClass: [{ type: HostBinding, args: ['class.thy-card-content',] }],
    scrollClassName: [{ type: HostBinding, args: ['class.thy-card-content--scroll',] }],
    thyScroll: [{ type: Input, args: ['thyScroll',] }],
    _thySizeSm: [{ type: HostBinding, args: ['class.thy-card-content--sm',] }],
    thySize: [{ type: Input, args: ['thySize',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardContentComponent, [{
        type: Component,
        args: [{
                selector: 'thy-card-content',
                preserveWhitespaces: false,
                template: `
        <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { thyCardContentClass: [{
            type: HostBinding,
            args: ['class.thy-card-content']
        }], scrollClassName: [{
            type: HostBinding,
            args: ['class.thy-card-content--scroll']
        }], _thySizeSm: [{
            type: HostBinding,
            args: ['class.thy-card-content--sm']
        }], thyScroll: [{
            type: Input,
            args: ['thyScroll']
        }], thySize: [{
            type: Input,
            args: ['thySize']
        }] }); })();
