// ngx-tethys/esm2015/card/header.component.js
import { Component, HostBinding, Input, TemplateRef, ContentChild } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["headerTitle"];
const _c1 = ["headerDescription"];
const _c2 = ["headerOperation"];
function ThyCardHeaderComponent_ng_template_1_Template(rf, ctx) { }
function ThyCardHeaderComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.thyTitle);
} }
function ThyCardHeaderComponent_ng_template_4_Template(rf, ctx) { }
function ThyCardHeaderComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.thyDescription);
} }
function ThyCardHeaderComponent_ng_template_7_Template(rf, ctx) { }
export class ThyCardHeaderComponent {
    constructor() {
        this.thyLayoutHeaderClass = true;
        this._thySizeSm = false;
        this._thySizeLg = false;
    }
    set thySize(value) {
        this._thySizeSm = value === 'sm';
        this._thySizeLg = value === 'lg';
    }
    ngOnInit() { }
}
ThyCardHeaderComponent.ɵfac = function ThyCardHeaderComponent_Factory(t) { return new (t || ThyCardHeaderComponent)(); };
ThyCardHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCardHeaderComponent, selectors: [["thy-card-header"]], contentQueries: function ThyCardHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.descriptionTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationTemplateRef = _t.first);
    } }, hostVars: 6, hostBindings: function ThyCardHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-card-header", ctx.thyLayoutHeaderClass)("thy-card-header--sm", ctx._thySizeSm)("thy-card-header--lg", ctx._thySizeLg);
    } }, inputs: { thySize: "thySize", thyTitle: "thyTitle", thyDescription: "thyDescription" }, decls: 8, vars: 5, consts: [[1, "card-header-title"], [3, "ngTemplateOutlet"], [4, "ngIf"], [1, "card-header-description"], [1, "card-header-operation"]], template: function ThyCardHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyCardHeaderComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(2, ThyCardHeaderComponent_ng_container_2_Template, 3, 1, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyCardHeaderComponent_ng_template_4_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(5, ThyCardHeaderComponent_ng_container_5_Template, 3, 1, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵtemplate(7, ThyCardHeaderComponent_ng_template_7_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.descriptionTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.descriptionTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.operationTemplateRef);
    } }, directives: [ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf], encapsulation: 2 });
ThyCardHeaderComponent.ctorParameters = () => [];
ThyCardHeaderComponent.propDecorators = {
    thyLayoutHeaderClass: [{ type: HostBinding, args: ['class.thy-card-header',] }],
    _thySizeSm: [{ type: HostBinding, args: ['class.thy-card-header--sm',] }],
    _thySizeLg: [{ type: HostBinding, args: ['class.thy-card-header--lg',] }],
    thyTitle: [{ type: Input, args: ['thyTitle',] }],
    thyDescription: [{ type: Input, args: ['thyDescription',] }],
    thySize: [{ type: Input, args: ['thySize',] }],
    titleTemplateRef: [{ type: ContentChild, args: ['headerTitle',] }],
    descriptionTemplateRef: [{ type: ContentChild, args: ['headerDescription',] }],
    operationTemplateRef: [{ type: ContentChild, args: ['headerOperation',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-card-header',
                preserveWhitespaces: false,
                template: "<div class=\"card-header-title\">\n  <ng-template [ngTemplateOutlet]=\"titleTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!titleTemplateRef\">\n    <span>{{ thyTitle }}</span>\n  </ng-container>\n</div>\n<div class=\"card-header-description\">\n  <ng-template [ngTemplateOutlet]=\"descriptionTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!descriptionTemplateRef\">\n    <span>{{ thyDescription }}</span>\n  </ng-container>\n</div>\n<div class=\"card-header-operation\">\n  <ng-template [ngTemplateOutlet]=\"operationTemplateRef\"></ng-template>\n</div>\n"
            }]
    }], function () { return []; }, { thyLayoutHeaderClass: [{
            type: HostBinding,
            args: ['class.thy-card-header']
        }], _thySizeSm: [{
            type: HostBinding,
            args: ['class.thy-card-header--sm']
        }], _thySizeLg: [{
            type: HostBinding,
            args: ['class.thy-card-header--lg']
        }], thySize: [{
            type: Input,
            args: ['thySize']
        }], thyTitle: [{
            type: Input,
            args: ['thyTitle']
        }], thyDescription: [{
            type: Input,
            args: ['thyDescription']
        }], titleTemplateRef: [{
            type: ContentChild,
            args: ['headerTitle']
        }], descriptionTemplateRef: [{
            type: ContentChild,
            args: ['headerDescription']
        }], operationTemplateRef: [{
            type: ContentChild,
            args: ['headerOperation']
        }] }); })();
