// ngx-tethys/esm2015/property-operation/group/property-operation-group.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyPropertyOperationGroupComponent {
    constructor() {
        this._isPropertyOperationGroup = true;
    }
}
ThyPropertyOperationGroupComponent.ɵfac = function ThyPropertyOperationGroupComponent_Factory(t) { return new (t || ThyPropertyOperationGroupComponent)(); };
ThyPropertyOperationGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPropertyOperationGroupComponent, selectors: [["thy-property-operation-group"]], hostVars: 2, hostBindings: function ThyPropertyOperationGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-property-operation-group", ctx._isPropertyOperationGroup);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyPropertyOperationGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyPropertyOperationGroupComponent.ctorParameters = () => [];
ThyPropertyOperationGroupComponent.propDecorators = {
    _isPropertyOperationGroup: [{ type: HostBinding, args: ['class.thy-property-operation-group',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPropertyOperationGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-property-operation-group',
                template: "<ng-content></ng-content>\n"
            }]
    }], function () { return []; }, { _isPropertyOperationGroup: [{
            type: HostBinding,
            args: ['class.thy-property-operation-group']
        }] }); })();
