// ngx-tethys/esm2015/divider/divider.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyDividerComponent } from './divider.component';
import * as ɵngcc0 from '@angular/core';
export class ThyDividerModule {
}
ThyDividerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDividerModule });
ThyDividerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDividerModule_Factory(t) { return new (t || ThyDividerModule)(); }, providers: [], imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDividerModule, { declarations: function () { return [ThyDividerComponent]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyDividerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDividerModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyDividerComponent],
                entryComponents: [],
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyDividerComponent],
                providers: []
            }]
    }], null, null); })();
