// ngx-tethys/esm2015/markdown/module.js
import { NgModule } from '@angular/core';
import { ThyMarkdownParserService } from './thy-markdown-parser.service';
import { ThyMarkdownParserDirective } from './thy-markdown-parser.directive';
import { ThyMarkdownPlanTextParserDirective } from './thy-markdown-text-parser.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyMarkdownModule {
}
ThyMarkdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyMarkdownModule });
ThyMarkdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyMarkdownModule_Factory(t) { return new (t || ThyMarkdownModule)(); }, providers: [ThyMarkdownParserService] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyMarkdownModule, { declarations: function () { return [ThyMarkdownParserDirective, ThyMarkdownPlanTextParserDirective]; }, exports: function () { return [ThyMarkdownParserDirective, ThyMarkdownPlanTextParserDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMarkdownModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyMarkdownParserDirective, ThyMarkdownPlanTextParserDirective],
                exports: [ThyMarkdownParserDirective, ThyMarkdownPlanTextParserDirective],
                providers: [ThyMarkdownParserService]
            }]
    }], null, null); })();
