// ngx-tethys/esm2015/markdown/thy-markdown-parser.service.js
import { Injectable, SecurityContext } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/platform-browser';
export class ThyMarkdownParserService {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    getEmojisRender() {
        return null;
    }
    setHighLightWords() {
        return [];
    }
    filterHTML(html) {
        return html;
    }
    sanitizeHTML(html) {
        return this.sanitizer.sanitize(SecurityContext.HTML, html);
    }
}
ThyMarkdownParserService.ɵfac = function ThyMarkdownParserService_Factory(t) { return new (t || ThyMarkdownParserService)(ɵngcc0.ɵɵinject(ɵngcc1.DomSanitizer)); };
ThyMarkdownParserService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyMarkdownParserService, factory: ThyMarkdownParserService.ɵfac });
ThyMarkdownParserService.ctorParameters = () => [
    { type: DomSanitizer }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMarkdownParserService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DomSanitizer }]; }, null); })();
