// ngx-tethys/esm2015/markdown/thy-markdown-text-parser.directive.js
import { Directive, ElementRef, Input, HostBinding } from '@angular/core';
import { ThyMarkdownParserService } from './thy-markdown-parser.service';
import { liteMarked } from 'ngx-tethys/types';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './thy-markdown-parser.service';
export class ThyMarkdownPlanTextParserDirective {
    constructor(elementRef, thyMarkdownParserService) {
        this.elementRef = elementRef;
        this.thyMarkdownParserService = thyMarkdownParserService;
        this.value = '';
        this.prefix = '';
        this.liteMarkedOptions = {
            gfm: true,
            tables: false,
            breaks: true,
            pedantic: false,
            sanitize: false,
            smartLists: false,
            smartypants: false,
            heading: false,
            link: false,
            list: false,
            wtlink: true,
            wthexcolor: true,
            wthexcolorRender: {
                className: 'msg-inline-color'
            },
            wtat: true,
            wtatRender: {
                memberPrefix: this.prefix,
                className: 'slide-trigger'
            },
            wthash: true,
            wtentity: true,
            wtentityRender: {
                className: 'slide-trigger'
            },
            wthashRender: {
                chlPrefix: '/messages/groups/'
            },
            wtexclamation: true,
            wtemoji: false,
            isParagraphDefault: false,
            isImageDefault: false,
            isBlockquoteDefault: false,
            isHrDefault: false,
            isStrongDefault: false,
            isEmDefault: false,
            isCodespanDefault: false,
            isCodeDefault: false,
            isDelDefault: false,
            isHtmlDefault: false,
            isTextEscape: true
        };
        this.hasClass = true;
    }
    set thyMarkdownPlanText(value) {
        if (value) {
            this.value = value;
            this.translateHTML();
        }
    }
    translateHTML() {
        liteMarked.setOptions(this.liteMarkedOptions);
        let _value = liteMarked.toHTML(this.value, this.liteMarkedOptions.highLightWords);
        _value = this.thyMarkdownParserService.sanitizeHTML(_value);
        this.elementRef.nativeElement.innerHTML = _value;
    }
    ngOnInit() {
        const emojisRender = this.thyMarkdownParserService.getEmojisRender();
        if (emojisRender) {
            this.liteMarkedOptions.wtemoji = true;
            this.liteMarkedOptions.wtemojiRender = emojisRender;
        }
        const _highLightWords = this.thyMarkdownParserService.setHighLightWords();
        if (_highLightWords && _highLightWords instanceof Array) {
            this.liteMarkedOptions.highLightWords = _highLightWords;
        }
        this.translateHTML();
    }
}
ThyMarkdownPlanTextParserDirective.ɵfac = function ThyMarkdownPlanTextParserDirective_Factory(t) { return new (t || ThyMarkdownPlanTextParserDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyMarkdownParserService)); };
ThyMarkdownPlanTextParserDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMarkdownPlanTextParserDirective, selectors: [["", "thyMarkdownPlanText", ""]], hostVars: 2, hostBindings: function ThyMarkdownPlanTextParserDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-markdown-plan-text", ctx.hasClass);
    } }, inputs: { thyMarkdownPlanText: "thyMarkdownPlanText" } });
ThyMarkdownPlanTextParserDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyMarkdownParserService }
];
ThyMarkdownPlanTextParserDirective.propDecorators = {
    thyMarkdownPlanText: [{ type: Input }],
    hasClass: [{ type: HostBinding, args: ['class.thy-markdown-plan-text',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMarkdownPlanTextParserDirective, [{
        type: Directive,
        args: [{
                selector: '[thyMarkdownPlanText]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyMarkdownParserService }]; }, { hasClass: [{
            type: HostBinding,
            args: ['class.thy-markdown-plan-text']
        }], thyMarkdownPlanText: [{
            type: Input
        }] }); })();
