// ngx-tethys/esm2015/action-menu/action-menu.component.js
import { Directive, HostBinding, Input, Component, ViewEncapsulation } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyActionMenuComponent {
    constructor() {
        this.className = true;
        this.themeClassName = false;
        this.styleWidth = '';
    }
    set thyTheme(value) {
        this.themeClassName = value === 'group';
    }
    set thyWidth(value) {
        this.styleWidth = value;
    }
}
ThyActionMenuComponent.ɵfac = function ThyActionMenuComponent_Factory(t) { return new (t || ThyActionMenuComponent)(); };
ThyActionMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyActionMenuComponent, selectors: [["thy-action-menu"]], hostVars: 6, hostBindings: function ThyActionMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", ctx.styleWidth);
        ɵngcc0.ɵɵclassProp("action-menu", ctx.className)("action-menu--group", ctx.themeClassName);
    } }, inputs: { thyTheme: "thyTheme", thyWidth: "thyWidth" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyActionMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyActionMenuComponent.ctorParameters = () => [];
ThyActionMenuComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class.action-menu',] }],
    themeClassName: [{ type: HostBinding, args: ['class.action-menu--group',] }],
    styleWidth: [{ type: HostBinding, args: ['style.width',] }],
    thyTheme: [{ type: Input }],
    thyWidth: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuComponent, [{
        type: Component,
        args: [{
                selector: 'thy-action-menu',
                template: `
        <ng-content></ng-content>
    `,
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.action-menu']
        }], themeClassName: [{
            type: HostBinding,
            args: ['class.action-menu--group']
        }], styleWidth: [{
            type: HostBinding,
            args: ['style.width']
        }], thyTheme: [{
            type: Input
        }], thyWidth: [{
            type: Input
        }] }); })();
export class ThyActionMenuGroupComponent {
    constructor() {
        this.isGroup = true;
    }
    set thyTitle(value) {
        this.groupTitle = value;
    }
}
ThyActionMenuGroupComponent.ɵfac = function ThyActionMenuGroupComponent_Factory(t) { return new (t || ThyActionMenuGroupComponent)(); };
ThyActionMenuGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyActionMenuGroupComponent, selectors: [["thy-action-menu-group"]], hostVars: 2, hostBindings: function ThyActionMenuGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("action-menu-group", ctx.isGroup);
    } }, inputs: { thyTitle: "thyTitle" }, ngContentSelectors: _c0, decls: 3, vars: 1, consts: [[1, "action-menu-group-title"]], template: function ThyActionMenuGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.groupTitle);
    } }, encapsulation: 2 });
ThyActionMenuGroupComponent.ctorParameters = () => [];
ThyActionMenuGroupComponent.propDecorators = {
    isGroup: [{ type: HostBinding, args: ['class.action-menu-group',] }],
    thyTitle: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-action-menu-group',
                template: `
        <div class="action-menu-group-title">{{ groupTitle }}</div>
        <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { isGroup: [{
            type: HostBinding,
            args: ['class.action-menu-group']
        }], thyTitle: [{
            type: Input
        }] }); })();
export class ThyActionMenuDividerComponent {
    constructor() {
        this.className = true;
        this.isCrossing = false;
    }
    set thyTitle(value) {
        this.title = value;
    }
    set thyType(value) {
        this.isCrossing = value === 'crossing';
    }
}
ThyActionMenuDividerComponent.ɵfac = function ThyActionMenuDividerComponent_Factory(t) { return new (t || ThyActionMenuDividerComponent)(); };
ThyActionMenuDividerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyActionMenuDividerComponent, selectors: [["thy-action-menu-divider"]], hostVars: 4, hostBindings: function ThyActionMenuDividerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("action-menu-divider", ctx.className)("action-menu-divider-crossing", ctx.isCrossing);
    } }, inputs: { thyTitle: "thyTitle", thyType: "thyType" }, decls: 2, vars: 1, consts: [[1, "action-menu-divider-title"]], template: function ThyActionMenuDividerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.title);
    } }, encapsulation: 2 });
ThyActionMenuDividerComponent.ctorParameters = () => [];
ThyActionMenuDividerComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class.action-menu-divider',] }],
    isCrossing: [{ type: HostBinding, args: ['class.action-menu-divider-crossing',] }],
    thyTitle: [{ type: Input }],
    thyType: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuDividerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-action-menu-divider',
                template: `
        <div class="action-menu-divider-title">{{ title }}</div>
    `
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.action-menu-divider']
        }], isCrossing: [{
            type: HostBinding,
            args: ['class.action-menu-divider-crossing']
        }], thyTitle: [{
            type: Input
        }], thyType: [{
            type: Input
        }] }); })();
export class ThyActionMenuDividerTitleDirective {
    constructor() {
        this.className = true;
    }
}
ThyActionMenuDividerTitleDirective.ɵfac = function ThyActionMenuDividerTitleDirective_Factory(t) { return new (t || ThyActionMenuDividerTitleDirective)(); };
ThyActionMenuDividerTitleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuDividerTitleDirective, selectors: [["", "thyActionMenuDividerTitle", ""]], hostVars: 2, hostBindings: function ThyActionMenuDividerTitleDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("action-menu-divider-title", ctx.className);
    } } });
ThyActionMenuDividerTitleDirective.ctorParameters = () => [];
ThyActionMenuDividerTitleDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.action-menu-divider-title',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuDividerTitleDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuDividerTitle]'
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.action-menu-divider-title']
        }] }); })();
