// ngx-tethys/esm2015/action-menu/action-menu-toggle.component.js
import { Directive, Input, HostListener, TemplateRef } from '@angular/core';
import { ThyPopover } from 'ngx-tethys/popover';
import { camelCase } from 'ngx-tethys/util';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
export var ActionEnum;
(function (ActionEnum) {
    ActionEnum["click"] = "click";
    ActionEnum["contextmenu"] = "contextmenu";
})(ActionEnum || (ActionEnum = {}));
export class ThyActionMenuToggleDirective {
    constructor(thyPopover) {
        this.thyPopover = thyPopover;
        this._action = ActionEnum.click;
        this._stopPropagation = false;
        this._thyContainerClass = '';
    }
    set thyActionMenuToggle(value) {
        this._templateRef = value;
    }
    set thyPlacement(value) {
        this._placement = value;
    }
    set thyAction(value) {
        this._action = value;
    }
    set thyStopPropagation(value) {
        this._stopPropagation = coerceBooleanProperty(value);
    }
    set thyContainerClass(value) {
        this._thyContainerClass = value;
    }
    onClick(event) {
        if (this._action === ActionEnum.click) {
            this._show.bind(this)(event);
        }
    }
    onRightClick(event) {
        if (this._action === ActionEnum.contextmenu) {
            this._show.bind(this)(event);
            return false;
        }
        return true;
    }
    _show(event) {
        if (this._stopPropagation) {
            event.stopPropagation();
        }
        this.thyPopover.open(this._templateRef, {
            origin: event.currentTarget,
            insideClosable: true,
            placement: this._placement ? camelCase(this._placement.split(' ')) : 'bottomLeft',
            panelClass: this._thyContainerClass,
            originActiveClass: this.thyOriginActiveClass
        });
    }
}
ThyActionMenuToggleDirective.ɵfac = function ThyActionMenuToggleDirective_Factory(t) { return new (t || ThyActionMenuToggleDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover)); };
ThyActionMenuToggleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuToggleDirective, selectors: [["", "thyActionMenuToggle", ""]], hostBindings: function ThyActionMenuToggleDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyActionMenuToggleDirective_click_HostBindingHandler($event) { return ctx.onClick($event); })("contextmenu", function ThyActionMenuToggleDirective_contextmenu_HostBindingHandler($event) { return ctx.onRightClick($event); });
    } }, inputs: { thyActionMenuToggle: "thyActionMenuToggle", thyPlacement: "thyPlacement", thyAction: "thyAction", thyStopPropagation: "thyStopPropagation", thyContainerClass: "thyContainerClass", thyOriginActiveClass: "thyOriginActiveClass" } });
ThyActionMenuToggleDirective.ctorParameters = () => [
    { type: ThyPopover }
];
ThyActionMenuToggleDirective.propDecorators = {
    thyActionMenuToggle: [{ type: Input }],
    thyPlacement: [{ type: Input }],
    thyAction: [{ type: Input }],
    thyStopPropagation: [{ type: Input }],
    thyContainerClass: [{ type: Input }],
    thyOriginActiveClass: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
    onRightClick: [{ type: HostListener, args: ['contextmenu', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuToggleDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuToggle]'
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }]; }, { thyActionMenuToggle: [{
            type: Input
        }], thyPlacement: [{
            type: Input
        }], thyAction: [{
            type: Input
        }], thyStopPropagation: [{
            type: Input
        }], thyContainerClass: [{
            type: Input
        }], onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }], onRightClick: [{
            type: HostListener,
            args: ['contextmenu', ['$event']]
        }], thyOriginActiveClass: [{
            type: Input
        }] }); })();
