// ngx-tethys/esm2015/action-menu/action-menu-sub-item.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import { ThyPositioningService } from 'ngx-tethys/positioning';
import { ThyActionMenuItemDirective } from './action-menu-item.directive';
import { mixinUnsubscribe, MixinBase } from 'ngx-tethys/core';
import { takeUntil } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './action-menu-item.directive';
import * as ɵngcc2 from 'ngx-tethys/positioning';
const _MixinBase = mixinUnsubscribe(MixinBase);
export class ThyActionMenuSubItemDirective extends _MixinBase {
    constructor(actionMenuItem, positioningService) {
        super();
        this.actionMenuItem = actionMenuItem;
        this.positioningService = positioningService;
        this.className = true;
        this.thyActionMenuSubItem = 'right';
    }
    ngOnInit() {
        let direction = this.thyActionMenuSubItem || 'right';
        if (this.thyActionMenuSubItem === 'auto') {
            this.actionMenuItem.updateClass([`action-menu-item-${direction}`]);
            this.actionMenuItem
                .bindMouseenterEvent()
                .pipe(takeUntil(this.ngUnsubscribe$))
                .subscribe(() => {
                const element = this.actionMenuItem.getElement();
                const offset = this.positioningService.offset(element);
                if (document.documentElement.clientWidth < offset.left + offset.width + offset.width) {
                    direction = 'left';
                }
                else {
                    direction = 'right';
                }
                this.actionMenuItem.updateClass([`action-menu-item-${direction}`]);
            });
        }
        else {
            this.actionMenuItem.updateClass([`action-menu-item-${direction}`]);
        }
    }
    ngOnDestroy() {
        super.ngOnDestroy();
    }
}
ThyActionMenuSubItemDirective.ɵfac = function ThyActionMenuSubItemDirective_Factory(t) { return new (t || ThyActionMenuSubItemDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyActionMenuItemDirective), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyPositioningService)); };
ThyActionMenuSubItemDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuSubItemDirective, selectors: [["", "thyActionMenuSubItem", ""]], hostVars: 2, hostBindings: function ThyActionMenuSubItemDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("action-menu-sub-item", ctx.className);
    } }, inputs: { thyActionMenuSubItem: "thyActionMenuSubItem" }, features: [ɵngcc0.ɵɵProvidersFeature([]), ɵngcc0.ɵɵInheritDefinitionFeature] });
ThyActionMenuSubItemDirective.ctorParameters = () => [
    { type: ThyActionMenuItemDirective },
    { type: ThyPositioningService }
];
ThyActionMenuSubItemDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.action-menu-sub-item',] }],
    thyActionMenuSubItem: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuSubItemDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuSubItem]',
                providers: []
            }]
    }], function () { return [{ type: ɵngcc1.ThyActionMenuItemDirective }, { type: ɵngcc2.ThyPositioningService }]; }, { className: [{
            type: HostBinding,
            args: ['class.action-menu-sub-item']
        }], thyActionMenuSubItem: [{
            type: Input
        }] }); })();
