// ngx-tethys/esm2015/action-menu/action-menu.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyActionMenuToggleDirective } from './action-menu-toggle.component';
import { ThyActionMenuComponent, ThyActionMenuDividerComponent, ThyActionMenuDividerTitleDirective, ThyActionMenuGroupComponent } from './action-menu.component';
import { ThyActionMenuItemDirective, ThyActionMenuItemIconDirective, ThyActionMenuItemNameDirective, ThyActionMenuItemExtendIconDirective, ThyActionMenuItemMetaDirective, ThyActionMenuItemInfoDirective, ThyActionMenuItemActiveDirective } from './action-menu-item.directive';
import { ThyActionMenuSubItemDirective } from './action-menu-sub-item.directive';
import { ThyPopoverModule } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
export class ThyActionMenuModule {
}
ThyActionMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyActionMenuModule });
ThyActionMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyActionMenuModule_Factory(t) { return new (t || ThyActionMenuModule)(); }, providers: [], imports: [[CommonModule, ThyPopoverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyActionMenuModule, { declarations: function () { return [ThyActionMenuToggleDirective,
        ThyActionMenuComponent,
        ThyActionMenuDividerComponent,
        ThyActionMenuItemDirective,
        ThyActionMenuItemIconDirective,
        ThyActionMenuItemNameDirective,
        ThyActionMenuItemExtendIconDirective,
        ThyActionMenuItemMetaDirective,
        ThyActionMenuItemInfoDirective,
        ThyActionMenuSubItemDirective,
        ThyActionMenuDividerTitleDirective,
        ThyActionMenuItemActiveDirective,
        ThyActionMenuGroupComponent]; }, imports: function () { return [CommonModule, ThyPopoverModule]; }, exports: function () { return [ThyActionMenuToggleDirective,
        ThyActionMenuComponent,
        ThyActionMenuDividerComponent,
        ThyActionMenuItemDirective,
        ThyActionMenuItemIconDirective,
        ThyActionMenuItemNameDirective,
        ThyActionMenuItemExtendIconDirective,
        ThyActionMenuItemMetaDirective,
        ThyActionMenuItemInfoDirective,
        ThyActionMenuSubItemDirective,
        ThyActionMenuDividerTitleDirective,
        ThyActionMenuItemActiveDirective,
        ThyActionMenuGroupComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyActionMenuToggleDirective,
                    ThyActionMenuComponent,
                    ThyActionMenuDividerComponent,
                    ThyActionMenuItemDirective,
                    ThyActionMenuItemIconDirective,
                    ThyActionMenuItemNameDirective,
                    ThyActionMenuItemExtendIconDirective,
                    ThyActionMenuItemMetaDirective,
                    ThyActionMenuItemInfoDirective,
                    ThyActionMenuSubItemDirective,
                    ThyActionMenuDividerTitleDirective,
                    ThyActionMenuItemActiveDirective,
                    ThyActionMenuGroupComponent
                ],
                imports: [CommonModule, ThyPopoverModule],
                providers: [],
                exports: [
                    ThyActionMenuToggleDirective,
                    ThyActionMenuComponent,
                    ThyActionMenuDividerComponent,
                    ThyActionMenuItemDirective,
                    ThyActionMenuItemIconDirective,
                    ThyActionMenuItemNameDirective,
                    ThyActionMenuItemExtendIconDirective,
                    ThyActionMenuItemMetaDirective,
                    ThyActionMenuItemInfoDirective,
                    ThyActionMenuSubItemDirective,
                    ThyActionMenuDividerTitleDirective,
                    ThyActionMenuItemActiveDirective,
                    ThyActionMenuGroupComponent
                ]
            }]
    }], null, null); })();
