// ngx-tethys/esm2015/action-menu/action-menu-item.directive.js
import { Directive, HostBinding, Input, HostListener, ElementRef } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { fromEvent } from 'rxjs';
import { debounceTime, shareReplay } from 'rxjs/operators';
import { UpdateHostClassService } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
export class ThyActionMenuItemDirective {
    constructor(elementRef, updateHostClassService) {
        this.elementRef = elementRef;
        this.updateHostClassService = updateHostClassService;
        this.className = true;
        this.disabled = false;
        this.danger = false;
        this.success = false;
    }
    set thyDisabled(value) {
        this.disabled = coerceBooleanProperty(value);
    }
    set thyType(value) {
        this[value] = true;
    }
    onClick(event) {
        if (this.disabled) {
            event.stopPropagation();
            event.preventDefault();
        }
    }
    ngOnInit() {
        this.updateHostClassService.initializeElement(this.elementRef);
    }
    updateClass(classes) {
        this.updateHostClassService.updateClass(classes);
    }
    getWidth() {
        return this.elementRef.nativeElement.offsetWidth;
    }
    getElement() {
        return this.elementRef.nativeElement;
    }
    bindMouseenterEvent() {
        return fromEvent(this.elementRef.nativeElement, 'mouseenter').pipe(debounceTime(100), shareReplay());
    }
}
ThyActionMenuItemDirective.ɵfac = function ThyActionMenuItemDirective_Factory(t) { return new (t || ThyActionMenuItemDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService)); };
ThyActionMenuItemDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemDirective, selectors: [["", "thyActionMenuItem", ""]], hostVars: 8, hostBindings: function ThyActionMenuItemDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyActionMenuItemDirective_click_HostBindingHandler($event) { return ctx.onClick($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("action-menu-item", ctx.className)("action-menu-item--disabled", ctx.disabled)("action-menu-item--danger", ctx.danger)("action-menu-item--success", ctx.success);
    } }, inputs: { thyDisabled: "thyDisabled", thyType: "thyType" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])] });
ThyActionMenuItemDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: UpdateHostClassService }
];
ThyActionMenuItemDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.action-menu-item',] }],
    disabled: [{ type: HostBinding, args: ['class.action-menu-item--disabled',] }],
    danger: [{ type: HostBinding, args: ['class.action-menu-item--danger',] }],
    success: [{ type: HostBinding, args: ['class.action-menu-item--success',] }],
    thyDisabled: [{ type: Input }],
    thyType: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItem]',
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.UpdateHostClassService }]; }, { className: [{
            type: HostBinding,
            args: ['class.action-menu-item']
        }], disabled: [{
            type: HostBinding,
            args: ['class.action-menu-item--disabled']
        }], danger: [{
            type: HostBinding,
            args: ['class.action-menu-item--danger']
        }], success: [{
            type: HostBinding,
            args: ['class.action-menu-item--success']
        }], thyDisabled: [{
            type: Input
        }], thyType: [{
            type: Input
        }], onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
export class ThyActionMenuItemIconDirective {
    constructor() {
        this.className = true;
    }
}
ThyActionMenuItemIconDirective.ɵfac = function ThyActionMenuItemIconDirective_Factory(t) { return new (t || ThyActionMenuItemIconDirective)(); };
ThyActionMenuItemIconDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemIconDirective, selectors: [["", "thyActionMenuItemIcon", ""]], hostVars: 2, hostBindings: function ThyActionMenuItemIconDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("icon", ctx.className);
    } } });
ThyActionMenuItemIconDirective.ctorParameters = () => [];
ThyActionMenuItemIconDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.icon',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemIconDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItemIcon]'
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.icon']
        }] }); })();
export class ThyActionMenuItemNameDirective {
    constructor() {
        this.className = true;
    }
}
ThyActionMenuItemNameDirective.ɵfac = function ThyActionMenuItemNameDirective_Factory(t) { return new (t || ThyActionMenuItemNameDirective)(); };
ThyActionMenuItemNameDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemNameDirective, selectors: [["", "thyActionMenuItemName", ""]], hostVars: 2, hostBindings: function ThyActionMenuItemNameDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("name", ctx.className);
    } } });
ThyActionMenuItemNameDirective.ctorParameters = () => [];
ThyActionMenuItemNameDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.name',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemNameDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItemName]'
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.name']
        }] }); })();
export class ThyActionMenuItemMetaDirective {
    constructor() {
        this.className = true;
    }
}
ThyActionMenuItemMetaDirective.ɵfac = function ThyActionMenuItemMetaDirective_Factory(t) { return new (t || ThyActionMenuItemMetaDirective)(); };
ThyActionMenuItemMetaDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemMetaDirective, selectors: [["", "thyActionMenuItemMeta", ""]], hostVars: 2, hostBindings: function ThyActionMenuItemMetaDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("meta", ctx.className);
    } } });
ThyActionMenuItemMetaDirective.ctorParameters = () => [];
ThyActionMenuItemMetaDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.meta',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemMetaDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItemMeta]'
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.meta']
        }] }); })();
export class ThyActionMenuItemInfoDirective {
    constructor() {
        this.className = true;
    }
}
ThyActionMenuItemInfoDirective.ɵfac = function ThyActionMenuItemInfoDirective_Factory(t) { return new (t || ThyActionMenuItemInfoDirective)(); };
ThyActionMenuItemInfoDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemInfoDirective, selectors: [["", "thyActionMenuItemInfo", ""]], hostVars: 2, hostBindings: function ThyActionMenuItemInfoDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("info", ctx.className);
    } } });
ThyActionMenuItemInfoDirective.ctorParameters = () => [];
ThyActionMenuItemInfoDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.info',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemInfoDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItemInfo]'
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.info']
        }] }); })();
export class ThyActionMenuItemExtendIconDirective {
    constructor() {
        this.className = true;
    }
}
ThyActionMenuItemExtendIconDirective.ɵfac = function ThyActionMenuItemExtendIconDirective_Factory(t) { return new (t || ThyActionMenuItemExtendIconDirective)(); };
ThyActionMenuItemExtendIconDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemExtendIconDirective, selectors: [["", "thyActionMenuItemExtendIcon", ""]], hostVars: 2, hostBindings: function ThyActionMenuItemExtendIconDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("extend-icon", ctx.className);
    } } });
ThyActionMenuItemExtendIconDirective.ctorParameters = () => [];
ThyActionMenuItemExtendIconDirective.propDecorators = {
    className: [{ type: HostBinding, args: ['class.extend-icon',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemExtendIconDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItemExtendIcon]'
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class.extend-icon']
        }] }); })();
export class ThyActionMenuItemActiveDirective {
    constructor() {
        this._isActive = false;
    }
    set thyActionMenuItemActive(value) {
        this._isActive = coerceBooleanProperty(value);
    }
}
ThyActionMenuItemActiveDirective.ɵfac = function ThyActionMenuItemActiveDirective_Factory(t) { return new (t || ThyActionMenuItemActiveDirective)(); };
ThyActionMenuItemActiveDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyActionMenuItemActiveDirective, selectors: [["", "thyActionMenuItemActive", ""]], hostVars: 2, hostBindings: function ThyActionMenuItemActiveDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("active", ctx._isActive);
    } }, inputs: { thyActionMenuItemActive: "thyActionMenuItemActive" } });
ThyActionMenuItemActiveDirective.ctorParameters = () => [];
ThyActionMenuItemActiveDirective.propDecorators = {
    _isActive: [{ type: HostBinding, args: ['class.active',] }],
    thyActionMenuItemActive: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyActionMenuItemActiveDirective, [{
        type: Directive,
        args: [{
                selector: '[thyActionMenuItemActive]'
            }]
    }], function () { return []; }, { _isActive: [{
            type: HostBinding,
            args: ['class.active']
        }], thyActionMenuItemActive: [{
            type: Input
        }] }); })();
