// ngx-tethys/esm2015/shared/base-form-check.component.js
import { HostBinding, Input, ChangeDetectorRef, Directive } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { ThyTranslate } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
const noop = () => { };
const ɵ0 = noop;
export class ThyFormCheckBaseComponent {
    constructor(thyTranslate, changeDetectorRef) {
        this.thyTranslate = thyTranslate;
        this.changeDetectorRef = changeDetectorRef;
        // The internal data model
        this._innerValue = null;
        this._disabled = false;
        this.onTouchedCallback = noop;
        this.onChangeCallback = noop;
        this._isFormCheck = true;
        this._isFormCheckInline = false;
        this._isChecked = false;
    }
    get _isNoText() {
        return this._isFormCheckInline && !this._labelText;
    }
    set thyInline(value) {
        this._isFormCheckInline = coerceBooleanProperty(value);
    }
    set thyLabelText(value) {
        this._labelText = value;
    }
    set thyLabelTextTranslateKey(value) {
        if (value) {
            this._labelText = this.thyTranslate.instant(value);
        }
        else {
            this._labelText = '';
        }
    }
    set thyDisabled(value) {
        this.setDisabledState(coerceBooleanProperty(value));
    }
    writeValue(obj) {
        if (obj !== this._innerValue) {
            this._innerValue = obj;
            this._isChecked = !!this._innerValue;
        }
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    setDisabledState(isDisabled) {
        this._disabled = isDisabled;
        this.markForCheck();
    }
    updateValue(value) {
        this._innerValue = value;
        this._isChecked = !!this._innerValue;
        this.onChangeCallback(value);
        this.markForCheck();
    }
    change() {
        this.updateValue(!this._innerValue);
    }
    markForCheck() {
        if (this.changeDetectorRef) {
            this.changeDetectorRef.markForCheck();
        }
    }
}
ThyFormCheckBaseComponent.ɵfac = function ThyFormCheckBaseComponent_Factory(t) { return new (t || ThyFormCheckBaseComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyFormCheckBaseComponent.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyFormCheckBaseComponent, hostVars: 8, hostBindings: function ThyFormCheckBaseComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("form-check", ctx._isFormCheck)("form-check-inline", ctx._isFormCheckInline)("form-check-checked", ctx._isChecked)("form-check-inline-no-label-text", ctx._isNoText);
    } }, inputs: { thyInline: "thyInline", thyLabelText: "thyLabelText", thyLabelTextTranslateKey: "thyLabelTextTranslateKey", thyDisabled: "thyDisabled" } });
ThyFormCheckBaseComponent.ctorParameters = () => [
    { type: ThyTranslate },
    { type: ChangeDetectorRef }
];
ThyFormCheckBaseComponent.propDecorators = {
    _isFormCheck: [{ type: HostBinding, args: ['class.form-check',] }],
    _isFormCheckInline: [{ type: HostBinding, args: ['class.form-check-inline',] }],
    _isChecked: [{ type: HostBinding, args: ['class.form-check-checked',] }],
    _isNoText: [{ type: HostBinding, args: ['class.form-check-inline-no-label-text',] }],
    thyInline: [{ type: Input }],
    thyLabelText: [{ type: Input }],
    thyLabelTextTranslateKey: [{ type: Input }],
    thyDisabled: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormCheckBaseComponent, [{
        type: Directive
    }], function () { return [{ type: ɵngcc1.ThyTranslate }, { type: ɵngcc0.ChangeDetectorRef }]; }, { _isFormCheck: [{
            type: HostBinding,
            args: ['class.form-check']
        }], _isFormCheckInline: [{
            type: HostBinding,
            args: ['class.form-check-inline']
        }], _isChecked: [{
            type: HostBinding,
            args: ['class.form-check-checked']
        }], _isNoText: [{
            type: HostBinding,
            args: ['class.form-check-inline-no-label-text']
        }], thyInline: [{
            type: Input
        }], thyLabelText: [{
            type: Input
        }], thyLabelTextTranslateKey: [{
            type: Input
        }], thyDisabled: [{
            type: Input
        }] }); })();
export { ɵ0 };
