// ngx-tethys/esm2015/shared/shared.module.js
import { NgModule } from '@angular/core';
import { ThyTranscludeDirective } from './ng-transclude.directive';
import { ThyAutofocusDirective } from './directives/thy-autofocus.directive';
import { ThyEnterDirective } from './directives/thy-enter.directive';
import { ThyCtrlEnterDirective } from './directives/thy-ctrl-enter.directive';
import { ThyShowDirective } from './directives/thy-show';
import { ThyStopPropagationDirective } from './directives/thy-stop-propagation.directive';
import { ThyContextMenuDirective } from './directives/thy-contextmenu.directive';
import { ThyScrollDirective } from './directives/thy-scroll.directive';
import { ThyDragDropDirective } from './directives/thy-drag-drop.directive';
import * as ɵngcc0 from '@angular/core';
export class ThySharedModule {
}
ThySharedModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySharedModule });
ThySharedModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySharedModule_Factory(t) { return new (t || ThySharedModule)(); }, providers: [] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySharedModule, { declarations: function () { return [ThyTranscludeDirective,
        ThyAutofocusDirective,
        ThyEnterDirective,
        ThyCtrlEnterDirective,
        ThyShowDirective,
        ThyStopPropagationDirective,
        ThyContextMenuDirective,
        ThyScrollDirective,
        ThyDragDropDirective]; }, exports: function () { return [ThyTranscludeDirective,
        ThyAutofocusDirective,
        ThyEnterDirective,
        ThyCtrlEnterDirective,
        ThyShowDirective,
        ThyStopPropagationDirective,
        ThyContextMenuDirective,
        ThyScrollDirective,
        ThyDragDropDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySharedModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyTranscludeDirective,
                    ThyAutofocusDirective,
                    ThyEnterDirective,
                    ThyCtrlEnterDirective,
                    ThyShowDirective,
                    ThyStopPropagationDirective,
                    ThyContextMenuDirective,
                    ThyScrollDirective,
                    ThyDragDropDirective
                ],
                exports: [
                    ThyTranscludeDirective,
                    ThyAutofocusDirective,
                    ThyEnterDirective,
                    ThyCtrlEnterDirective,
                    ThyShowDirective,
                    ThyStopPropagationDirective,
                    ThyContextMenuDirective,
                    ThyScrollDirective,
                    ThyDragDropDirective
                ],
                providers: []
            }]
    }], null, null); })();
