// ngx-tethys/esm2015/copy/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyCopyDirective } from './copy.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyCopyModule {
}
ThyCopyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCopyModule });
ThyCopyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCopyModule_Factory(t) { return new (t || ThyCopyModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCopyModule, { declarations: function () { return [ThyCopyDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyCopyDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCopyModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyCopyDirective],
                imports: [CommonModule],
                exports: [ThyCopyDirective]
            }]
    }], null, null); })();
