// ngx-tethys/esm2015/progress/bar/progress-bar.component.js
import { Component, Input, HostBinding, ElementRef, ViewEncapsulation, InjectionToken, Inject } from '@angular/core';
import { UpdateHostClassService } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';

const _c0 = ["*"];
export const THY_PROGRESS_COMPONENT = new InjectionToken('THY_PROGRESS_COMPONENT');
export class ThyProgressBarComponent {
    constructor(updateHostClassService, elementRef, progress) {
        this.updateHostClassService = updateHostClassService;
        this.progress = progress;
        this.isProgressBar = true;
        this.percent = 0;
        this.color = '';
        updateHostClassService.initializeElement(elementRef.nativeElement);
    }
    set thyType(type) {
        // this.type = type;
        this.updateHostClassService.updateClass(type ? [`progress-bar-${type}`, `bg-${type}`] : []);
    }
    set thyValue(value) {
        if (!value && value !== 0) {
            return;
        }
        this.value = value;
        this.recalculatePercentage();
    }
    set thyColor(color) {
        this.color = color || '';
    }
    recalculatePercentage() {
        this.percent = +((this.value / this.progress.max) * 100).toFixed(2);
        // if (this.progress && this.progress.bars) {
        //     const totalPercentage = this.progress.bars.reduce(function(total: number, bar): number {
        //         return total + bar.percent;
        //     }, 0);
        //     if (totalPercentage > 100) {
        //         this.percent -= totalPercentage - 100;
        //     }
        // }
    }
}
ThyProgressBarComponent.ɵfac = function ThyProgressBarComponent_Factory(t) { return new (t || ThyProgressBarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(THY_PROGRESS_COMPONENT)); };
ThyProgressBarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyProgressBarComponent, selectors: [["thy-progress-bar"]], hostVars: 6, hostBindings: function ThyProgressBarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", ctx.percent, "%")("background-color", ctx.color);
        ɵngcc0.ɵɵclassProp("progress-bar", ctx.isProgressBar);
    } }, inputs: { thyType: "thyType", thyValue: "thyValue", thyColor: "thyColor", thyTips: "thyTips" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyProgressBarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyProgressBarComponent.ctorParameters = () => [
    { type: UpdateHostClassService },
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Inject, args: [THY_PROGRESS_COMPONENT,] }] }
];
ThyProgressBarComponent.propDecorators = {
    isProgressBar: [{ type: HostBinding, args: [`class.progress-bar`,] }],
    percent: [{ type: HostBinding, args: ['style.width.%',] }],
    color: [{ type: HostBinding, args: ['style.background-color',] }],
    thyTips: [{ type: Input }],
    thyType: [{ type: Input }],
    thyValue: [{ type: Input }],
    thyColor: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyProgressBarComponent, [{
        type: Component,
        args: [{
                selector: 'thy-progress-bar',
                template: "<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None,
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_PROGRESS_COMPONENT]
            }] }]; }, { isProgressBar: [{
            type: HostBinding,
            args: [`class.progress-bar`]
        }], percent: [{
            type: HostBinding,
            args: ['style.width.%']
        }], color: [{
            type: HostBinding,
            args: ['style.background-color']
        }], thyType: [{
            type: Input
        }], thyValue: [{
            type: Input
        }], thyColor: [{
            type: Input
        }], thyTips: [{
            type: Input
        }] }); })();
