// ngx-tethys/esm2015/resizable/module.js
import { NgModule } from '@angular/core';
import { ThyResizableDirective } from './resizable.directive';
import { ThyResizeHandleComponent } from './resize-handle.component';
import { ThyResizeHandlesComponent } from './resize-handles.component';
import { CommonModule } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
export class ThyResizableModule {
}
ThyResizableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyResizableModule });
ThyResizableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyResizableModule_Factory(t) { return new (t || ThyResizableModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyResizableModule, { declarations: function () { return [ThyResizableDirective, ThyResizeHandleComponent, ThyResizeHandlesComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyResizableDirective, ThyResizeHandleComponent, ThyResizeHandlesComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyResizableModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [ThyResizableDirective, ThyResizeHandleComponent, ThyResizeHandlesComponent],
                exports: [ThyResizableDirective, ThyResizeHandleComponent, ThyResizeHandlesComponent]
            }]
    }], null, null); })();
