// ngx-tethys/esm2015/arrow-switcher/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyArrowSwitcherComponent } from './arrow-switcher.component';
import * as ɵngcc0 from '@angular/core';
export class ThyArrowSwitcherModule {
}
ThyArrowSwitcherModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyArrowSwitcherModule });
ThyArrowSwitcherModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyArrowSwitcherModule_Factory(t) { return new (t || ThyArrowSwitcherModule)(); }, imports: [[CommonModule, ThyButtonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyArrowSwitcherModule, { declarations: function () { return [ThyArrowSwitcherComponent]; }, imports: function () { return [CommonModule, ThyButtonModule]; }, exports: function () { return [ThyArrowSwitcherComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyArrowSwitcherModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyArrowSwitcherComponent],
                imports: [CommonModule, ThyButtonModule],
                exports: [ThyArrowSwitcherComponent]
            }]
    }], null, null); })();
