// ngx-tethys/esm2015/menu/item/menu-item.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyMenuItemComponent {
    constructor() {
        this.isThyMenuItem = true;
    }
    ngOnInit() { }
}
ThyMenuItemComponent.ɵfac = function ThyMenuItemComponent_Factory(t) { return new (t || ThyMenuItemComponent)(); };
ThyMenuItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemComponent, selectors: [["thy-menu-item"], ["", "thy-menu-item", ""], ["", "thyMenuItem", ""]], hostVars: 2, hostBindings: function ThyMenuItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item", ctx.isThyMenuItem);
    } }, ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "thy-menu-item-content"]], template: function ThyMenuItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
ThyMenuItemComponent.ctorParameters = () => [];
ThyMenuItemComponent.propDecorators = {
    isThyMenuItem: [{ type: HostBinding, args: ['class.thy-menu-item',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item,[thy-menu-item],[thyMenuItem]',
                template: "<div class=\"thy-menu-item-content\">\n  <ng-content></ng-content>\n</div>\n"
            }]
    }], function () { return []; }, { isThyMenuItem: [{
            type: HostBinding,
            args: ['class.thy-menu-item']
        }] }); })();
