// ngx-tethys/bundles/ngx-tethys-popover.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/cdk/portal'),require('@angular/cdk/overlay'),require('@angular/cdk/platform'),require('ngx-tethys/icon'),require('@angular/cdk/a11y'),require('@angular/common'),exports, require('@angular/core'), require('ngx-tethys/core'), require('ngx-tethys/util'), require('rxjs'), require('rxjs/operators'), require('@angular/cdk/bidi'), require('@angular/cdk/coercion'), require('@angular/cdk/overlay'), require('@angular/cdk/platform'), require('@angular/cdk/portal'), require('@angular/common'), require('@angular/animations'), require('@angular/cdk/scrolling'), require('ngx-tethys/icon'), require('@angular/cdk/a11y')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/popover', ['@angular/core','ngx-tethys/core','@angular/cdk/portal','@angular/cdk/overlay','@angular/cdk/platform','ngx-tethys/icon','@angular/cdk/a11y','@angular/common','exports', '@angular/core', 'ngx-tethys/core', 'ngx-tethys/util', 'rxjs', 'rxjs/operators', '@angular/cdk/bidi', '@angular/cdk/coercion', '@angular/cdk/overlay', '@angular/cdk/platform', '@angular/cdk/portal', '@angular/common', '@angular/animations', '@angular/cdk/scrolling', 'ngx-tethys/icon', '@angular/cdk/a11y'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.cdk.portal,global.ng.cdk.overlay,global.ng.cdk.platform,global.ngxTethys.icon,global.ng.cdk.a11y,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].popover = {}), global.ng.core, global['ngx-tethys'].core, global['ngx-tethys'].util, global.rxjs, global.rxjs.operators, global.ng.cdk.bidi, global.ng.cdk.coercion, global.ng.cdk.overlay, global.ng.cdk.platform, global.ng.cdk.portal, global.ng.common, global.ng.animations, global.ng.cdk.scrolling, global['ngx-tethys'].icon, global.ng.cdk.a11y));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,exports, i0, core, util, rxjs, operators, bidi, coercion, i1, i5, portal, i4, animations, i3, icon, a11y) { 
var _c0 = ["*"];
function ThyPopoverContainerComponent_ng_template_0_Template(rf, ctx) { }
'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    var ThyPopoverBodyComponent = /** @class */ (function () {
        function ThyPopoverBodyComponent() {
            this.isPopoverBody = true;
        }
ThyPopoverBodyComponent.ɵfac = function ThyPopoverBodyComponent_Factory(t) { return new (t || ThyPopoverBodyComponent)(); };
ThyPopoverBodyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPopoverBodyComponent, selectors: [["thy-popover-body"]], hostVars: 2, hostBindings: function ThyPopoverBodyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-popover-body", ctx.isPopoverBody);
    } }, exportAs: ["thyPopoverBody"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyPopoverBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverBodyComponent, [{
        type: i0.Component,
        args: [{
                selector: 'thy-popover-body',
                template: '<ng-content></ng-content>',
                exportAs: 'thyPopoverBody'
            }]
    }], function () { return []; }, { isPopoverBody: [{
            type: i0.HostBinding,
            args: ['class.thy-popover-body']
        }] }); })();
        return ThyPopoverBodyComponent;
    }());
    ThyPopoverBodyComponent.propDecorators = {
        isPopoverBody: [{ type: i0.HostBinding, args: ['class.thy-popover-body',] }]
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ThyPopoverConfig = /** @class */ (function (_super) {
        __extends(ThyPopoverConfig, _super);
        function ThyPopoverConfig() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ThyPopoverConfig;
    }(core.ThyAbstractOverlayConfig));
    var THY_POPOVER_DEFAULT_CONFIG = new i0.InjectionToken('thy-popover-default-config');
    var THY_POPOVER_DEFAULT_CONFIG_VALUE = {
        hasBackdrop: true,
        backdropClass: 'thy-popover-backdrop',
        panelClass: '',
        offset: 4,
        backdropClosable: true,
        closeOnNavigation: true,
        placement: 'bottom',
        insideClosable: false,
        manualClosure: false,
        originActiveClass: 'thy-popover-origin-active'
    };
    var THY_POPOVER_DEFAULT_CONFIG_PROVIDER = {
        provide: THY_POPOVER_DEFAULT_CONFIG,
        useValue: THY_POPOVER_DEFAULT_CONFIG_VALUE
    };
    var THY_POPOVER_SCROLL_STRATEGY = new i0.InjectionToken('thy-popover-scroll-strategy');
    var THY_POPOVER_SCROLL_STRATEGY_FACTORY = function (overlay) {
        return function () { return overlay.scrollStrategies.block(); };
    };
    var THY_POPOVER_SCROLL_STRATEGY_PROVIDER = {
        provide: THY_POPOVER_SCROLL_STRATEGY,
        deps: [i1.Overlay],
        useFactory: THY_POPOVER_SCROLL_STRATEGY_FACTORY
    };

    var animationBody = [
        // Note: The `enter` animation transitions to `transform: none`, because for some reason
        // specifying the transform explicitly, causes IE both to blur the dialog content and
        // decimate the animation performance. Leaving it as `none` solves both issues. translate(-300px,-100px)
        animations.state('void, exit', animations.style({ opacity: 0, transform: 'scale(0.1)' })),
        animations.state('enter', animations.style({ transform: 'none' }))
        // transition('* => enter', animate('100ms cubic-bezier(0, 0, 0.57, 1)', style({ transform: 'none', opacity: 1 }))),
        // transition(
        //     '* => exit',
        //     animate('150ms cubic-bezier(0.24, 0.18, 0.96, 0.52)', style({ transform: 'scale(0.3)', opacity: 0 }))
        // )
    ];
    /**
     * Animations used by ThyPopover.
     * @docs-private
     */
    var thyPopoverAnimations = {
        popoverContainer: animations.trigger('popoverContainer', animationBody)
    };
    // export const setAnimationStart

    var popoverUpperOverlayOptions = {
        name: 'popover',
        animationEnabled: true,
        disposeWhenClose: true
    };

    var ThyPopoverContainerComponent = /** @class */ (function (_super) {
        __extends(ThyPopoverContainerComponent, _super);
        function ThyPopoverContainerComponent(elementRef, document, config, changeDetectorRef, thyClickDispatcher, ngZone) {
            var _this = _super.call(this, popoverUpperOverlayOptions, changeDetectorRef) || this;
            _this.elementRef = elementRef;
            _this.document = document;
            _this.config = config;
            _this.thyClickDispatcher = thyClickDispatcher;
            _this.ngZone = ngZone;
            /** State of the popover animation. */
            _this.animationState = 'enter';
            /** Emits when an animation state changes. */
            _this.animationStateChanged = new i0.EventEmitter();
            _this.insideClicked = new i0.EventEmitter();
            _this.outsideClicked = new i0.EventEmitter();
            _this.animationOpeningDone = _this.animationStateChanged.pipe(operators.filter(function (event) {
                return event.phaseName === 'done' && event.toState === 'enter';
            }));
            _this.animationClosingDone = _this.animationStateChanged.pipe(operators.filter(function (event) {
                return event.phaseName === 'done' && event.toState === 'exit';
            }));
            return _this;
        }
        ThyPopoverContainerComponent.prototype.beforeAttachPortal = function () { };
        ThyPopoverContainerComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.config.outsideClosable && !this.config.hasBackdrop) {
                rxjs.timer().subscribe(function () {
                    _this.thyClickDispatcher
                        .clicked()
                        .pipe(operators.takeUntil(_this.animationClosingDone))
                        .subscribe(function (event) {
                        if (!_this.elementRef.nativeElement.contains(event.target)) {
                            _this.ngZone.run(function () {
                                _this.outsideClicked.emit();
                            });
                        }
                    });
                });
            }
        };
        /** Callback, invoked whenever an animation on the host completes. */
        ThyPopoverContainerComponent.prototype.onAnimationDone = function (event) {
            // if (event.toState === 'void') {
            //     this.trapFocus();
            // } else if (event.toState === 'exit') {
            //     this.restoreFocus();
            // }
            this.animationStateChanged.emit(event);
        };
        /** Callback, invoked when an animation on the host starts. */
        ThyPopoverContainerComponent.prototype.onAnimationStart = function (event) {
            this.animationStateChanged.emit(event);
        };
        ThyPopoverContainerComponent.prototype.startExitAnimation = function () {
            this.animationState = 'exit';
            // Mark the container for check so it can react if the
            // view container is using OnPush change detection.
            this.changeDetectorRef.markForCheck();
        };
        ThyPopoverContainerComponent.prototype.onInsideClick = function () {
            if (this.config.insideClosable) {
                this.insideClicked.emit();
            }
        };
        ThyPopoverContainerComponent.prototype.ngOnDestroy = function () {
            _super.prototype.destroy.call(this);
        };
ThyPopoverContainerComponent.ɵfac = function ThyPopoverContainerComponent_Factory(t) { return new (t || ThyPopoverContainerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(i4.DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ThyPopoverConfig), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyClickDispatcher), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyPopoverContainerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPopoverContainerComponent, selectors: [["thy-popover-container"]], viewQuery: function ThyPopoverContainerComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(portal.CdkPortalOutlet, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.portalOutlet = _t.first);
    } }, hostAttrs: ["tabindex", "-1", 1, "thy-popover-container"], hostVars: 2, hostBindings: function ThyPopoverContainerComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵsyntheticHostListener("@popoverContainer.start", function ThyPopoverContainerComponent_animation_popoverContainer_start_HostBindingHandler($event) { return ctx.onAnimationStart($event); })("@popoverContainer.done", function ThyPopoverContainerComponent_animation_popoverContainer_done_HostBindingHandler($event) { return ctx.onAnimationDone($event); });
        ɵngcc0.ɵɵlistener("click", function ThyPopoverContainerComponent_click_HostBindingHandler() { return ctx.onInsideClick(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("role", "popover");
        ɵngcc0.ɵɵsyntheticHostProperty("@popoverContainer", ctx.animationState);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 0, consts: [["cdkPortalOutlet", ""]], template: function ThyPopoverContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyPopoverContainerComponent_ng_template_0_Template, 0, 0, "ng-template", 0);
    } }, directives: [ɵngcc2.CdkPortalOutlet], encapsulation: 2, data: { animation: [thyPopoverAnimations.popoverContainer] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverContainerComponent, [{
        type: i0.Component,
        args: [{
                selector: 'thy-popover-container',
                template: "<ng-template cdkPortalOutlet></ng-template>\n",
                animations: [thyPopoverAnimations.popoverContainer],
                host: {
                    class: 'thy-popover-container',
                    tabindex: '-1',
                    '[attr.role]': "'popover'",
                    '[@popoverContainer]': 'animationState',
                    '(@popoverContainer.start)': 'onAnimationStart($event)',
                    '(@popoverContainer.done)': 'onAnimationDone($event)'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: undefined, decorators: [{
                type: i0.Inject,
                args: [i4.DOCUMENT]
            }] }, { type: ThyPopoverConfig }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.ThyClickDispatcher }, { type: ɵngcc0.NgZone }]; }, { onInsideClick: [{
            type: i0.HostListener,
            args: ['click', []]
        }], portalOutlet: [{
            type: i0.ViewChild,
            args: [portal.CdkPortalOutlet, { static: true }]
        }] }); })();
        return ThyPopoverContainerComponent;
    }(core.ThyAbstractOverlayContainer));
    ThyPopoverContainerComponent.ctorParameters = function () { return [
        { type: i0.ElementRef },
        { type: undefined, decorators: [{ type: i0.Inject, args: [i4.DOCUMENT,] }] },
        { type: ThyPopoverConfig },
        { type: i0.ChangeDetectorRef },
        { type: core.ThyClickDispatcher },
        { type: i0.NgZone }
    ]; };
    ThyPopoverContainerComponent.propDecorators = {
        portalOutlet: [{ type: i0.ViewChild, args: [portal.CdkPortalOutlet, { static: true },] }],
        onInsideClick: [{ type: i0.HostListener, args: ['click', [],] }]
    };

    var ThyPopoverRef = /** @class */ (function (_super) {
        __extends(ThyPopoverRef, _super);
        function ThyPopoverRef() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ThyPopoverRef;
    }(core.ThyAbstractOverlayRef));
    var ThyInternalPopoverRef = /** @class */ (function (_super) {
        __extends(ThyInternalPopoverRef, _super);
        function ThyInternalPopoverRef(overlayRef, containerInstance, config) {
            var _this = _super.call(this, popoverUpperOverlayOptions, overlayRef, containerInstance, config) || this;
            containerInstance.insideClicked.subscribe(function () {
                _this.close();
            });
            containerInstance.outsideClicked.subscribe(function () {
                _this.close();
            });
            return _this;
        }
        /**
         * Updates the popover's position.
         * @param position New popover position.
         */
        ThyInternalPopoverRef.prototype.updatePosition = function () {
            this.overlayRef.updatePosition();
            return this;
        };
        return ThyInternalPopoverRef;
    }(core.ThyAbstractInternalOverlayRef));

    var ThyPopover = /** @class */ (function (_super) {
        __extends(ThyPopover, _super);
        function ThyPopover(overlay, injector, defaultConfig, scrollStrategy, ngZone, _viewportRuler, _document, _platform, _overlayContainer) {
            var _this = _super.call(this, popoverUpperOverlayOptions, overlay, injector, Object.assign(Object.assign({}, THY_POPOVER_DEFAULT_CONFIG_VALUE), defaultConfig), scrollStrategy) || this;
            _this.ngZone = ngZone;
            _this._viewportRuler = _viewportRuler;
            _this._document = _document;
            _this._platform = _platform;
            _this._overlayContainer = _overlayContainer;
            _this.ngUnsubscribe$ = new rxjs.Subject();
            _this.originInstancesMap = new Map();
            return _this;
        }
        ThyPopover.prototype.buildPositionStrategy = function (config) {
            var _this = this;
            var origin = config.originPosition ? config.originPosition : config.origin;
            // const positionStrategy = this.overlay.position().flexibleConnectedTo(origin);
            var positionStrategy = new i1.FlexibleConnectedPositionStrategy(origin, this._viewportRuler, this._document, this._platform, this._overlayContainer);
            var positions = core.getFlexiblePositions(config.placement, config.offset, 'thy-popover');
            positionStrategy.withPositions(positions);
            positionStrategy.withGrowAfterOpen(true);
            positionStrategy.positionChanges.pipe(operators.takeUntil(this.ngUnsubscribe$)).subscribe(function (change) {
                if (change.scrollableViewProperties.isOverlayClipped) {
                    // After position changes occur and the overlay is clipped by
                    // a parent scrollable then close the tooltip.
                    _this.ngZone.run(function () { return _this.close(); });
                }
            });
            return positionStrategy;
        };
        ThyPopover.prototype.buildScrollStrategy = function (config) {
            if (config.scrollStrategy) {
                return config.scrollStrategy;
            }
            else if (this.scrollStrategy && util.isFunction(this.scrollStrategy)) {
                return this.scrollStrategy();
            }
            else {
                this.overlay.scrollStrategies.block();
            }
        };
        ThyPopover.prototype.buildOverlayConfig = function (config) {
            var positionStrategy = this.buildPositionStrategy(config);
            var overlayConfig = this.buildBaseOverlayConfig(config);
            overlayConfig.positionStrategy = positionStrategy;
            overlayConfig.scrollStrategy = this.buildScrollStrategy(config);
            return overlayConfig;
        };
        ThyPopover.prototype.attachUpperOverlayContainer = function (overlay, config) {
            var userInjector = config && config.viewContainerRef && config.viewContainerRef.injector;
            var injector = i0.Injector.create({
                parent: userInjector || this.injector,
                providers: [{ provide: ThyPopoverConfig, useValue: config }]
            });
            var containerPortal = new portal.ComponentPortal(ThyPopoverContainerComponent, config.viewContainerRef, injector);
            var containerRef = overlay.attach(containerPortal);
            return containerRef.instance;
        };
        ThyPopover.prototype.createUpperOverlayRef = function (overlayRef, containerInstance, config) {
            return new ThyInternalPopoverRef(overlayRef, containerInstance, config);
        };
        ThyPopover.prototype.createInjector = function (config, popoverRef, popoverContainer) {
            var userInjector = config && config.viewContainerRef && config.viewContainerRef.injector;
            var injectionTokens = [
                { provide: ThyPopoverContainerComponent, useValue: popoverContainer },
                {
                    provide: ThyPopoverRef,
                    useValue: popoverRef
                }
            ];
            if (config.direction && (!userInjector || !userInjector.get(bidi.Directionality, null))) {
                injectionTokens.push({
                    provide: bidi.Directionality,
                    useValue: {
                        value: config.direction,
                        change: rxjs.of()
                    }
                });
            }
            return i0.Injector.create({ parent: userInjector || this.injector, providers: injectionTokens });
        };
        ThyPopover.prototype.originElementAddActiveClass = function (config) {
            var _a;
            if (config.originActiveClass) {
                (_a = coercion.coerceElement(config.origin).classList).add.apply(_a, __spread(coercion.coerceArray(config.originActiveClass)));
            }
        };
        ThyPopover.prototype.originElementRemoveActiveClass = function (config) {
            var _a;
            if (config.originActiveClass) {
                (_a = coercion.coerceElement(config.origin).classList).remove.apply(_a, __spread(coercion.coerceArray(config.originActiveClass)));
            }
        };
        ThyPopover.prototype.ensureCloseClosest = function (origin) {
            var closeAndEnd = false;
            this.originInstancesMap.forEach(function (value, key) {
                if (value.config.manualClosure) {
                    if (key === origin) {
                        value.popoverRef.close();
                        closeAndEnd = true;
                    }
                }
                else {
                    if (key === origin) {
                        closeAndEnd = true;
                    }
                    value.popoverRef.close();
                }
            });
            return closeAndEnd;
        };
        ThyPopover.prototype.open = function (componentOrTemplateRef, config) {
            var _this = this;
            var originElement = coercion.coerceElement(config.origin);
            // 默认关闭之前的弹出框
            // 1. 当之前的 Popover 设置 manualClosure 为 true 时, 弹出其他 Popover 时不自动关闭 manualClosure 为 true 的 Popover
            // 2. 当前的 Origin 对应的 Popover 已经弹出，不管 manualClosure 设置为何，直接关闭并返回
            if (this.ensureCloseClosest(originElement)) {
                return;
            }
            var popoverRef = this.openUpperOverlay(componentOrTemplateRef, config);
            config = popoverRef.containerInstance.config;
            popoverRef.afterClosed().subscribe(function () {
                _this.originElementRemoveActiveClass(config);
                _this.originInstancesMap.delete(originElement);
            });
            this.originElementAddActiveClass(config);
            this.originInstancesMap.set(originElement, {
                config: config,
                popoverRef: popoverRef
            });
            return popoverRef;
        };
        ThyPopover.prototype.ngOnDestroy = function () {
            this.dispose();
        };
ThyPopover.ɵfac = function ThyPopover_Factory(t) { return new (t || ThyPopover)(ɵngcc0.ɵɵinject(ɵngcc3.Overlay), ɵngcc0.ɵɵinject(ɵngcc0.Injector), ɵngcc0.ɵɵinject(THY_POPOVER_DEFAULT_CONFIG, 8), ɵngcc0.ɵɵinject(THY_POPOVER_SCROLL_STRATEGY), ɵngcc0.ɵɵinject(ɵngcc0.NgZone), ɵngcc0.ɵɵinject(ɵngcc3.ViewportRuler), ɵngcc0.ɵɵinject(i4.DOCUMENT, 8), ɵngcc0.ɵɵinject(ɵngcc4.Platform), ɵngcc0.ɵɵinject(ɵngcc3.OverlayContainer)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopover, [{
        type: i0.Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc3.Overlay }, { type: ɵngcc0.Injector }, { type: ThyPopoverConfig, decorators: [{
                type: i0.Optional
            }, {
                type: i0.Inject,
                args: [THY_POPOVER_DEFAULT_CONFIG]
            }] }, { type: undefined, decorators: [{
                type: i0.Inject,
                args: [THY_POPOVER_SCROLL_STRATEGY]
            }] }, { type: ɵngcc0.NgZone }, { type: ɵngcc3.ViewportRuler }, { type: undefined, decorators: [{
                type: i0.Optional
            }, {
                type: i0.Inject,
                args: [i4.DOCUMENT]
            }] }, { type: ɵngcc4.Platform }, { type: ɵngcc3.OverlayContainer }]; }, null); })();
        return ThyPopover;
    }(core.ThyAbstractOverlayService));
    ThyPopover.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function ThyPopover_Factory() { return new ThyPopover(i0__namespace.ɵɵinject(i1__namespace.Overlay), i0__namespace.ɵɵinject(i0__namespace.INJECTOR), i0__namespace.ɵɵinject(THY_POPOVER_DEFAULT_CONFIG, 8), i0__namespace.ɵɵinject(THY_POPOVER_SCROLL_STRATEGY), i0__namespace.ɵɵinject(i0__namespace.NgZone), i0__namespace.ɵɵinject(i3__namespace.ViewportRuler), i0__namespace.ɵɵinject(i4__namespace.DOCUMENT, 8), i0__namespace.ɵɵinject(i5__namespace.Platform), i0__namespace.ɵɵinject(i1__namespace.OverlayContainer)); }, token: ThyPopover, providedIn: "root" });
    ThyPopover.ctorParameters = function () { return [
        { type: i1.Overlay },
        { type: i0.Injector },
        { type: ThyPopoverConfig, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [THY_POPOVER_DEFAULT_CONFIG,] }] },
        { type: undefined, decorators: [{ type: i0.Inject, args: [THY_POPOVER_SCROLL_STRATEGY,] }] },
        { type: i0.NgZone },
        { type: i1.ViewportRuler },
        { type: undefined, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [i4.DOCUMENT,] }] },
        { type: i5.Platform },
        { type: i1.OverlayContainer }
    ]; };

    var ThyPopoverHeaderComponent = /** @class */ (function () {
        function ThyPopoverHeaderComponent(translate, thyPopover) {
            this.translate = translate;
            this.thyPopover = thyPopover;
            this.thyClosed = new i0.EventEmitter();
        }
        Object.defineProperty(ThyPopoverHeaderComponent.prototype, "thyTitleTranslationKey", {
            set: function (key) {
                if (key && !this.thyTitle) {
                    this.thyTitle = this.translate.instant(key);
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyPopoverHeaderComponent.prototype.close = function (event) {
            this.thyClosed.emit(event);
            this.thyPopover.close();
        };
ThyPopoverHeaderComponent.ɵfac = function ThyPopoverHeaderComponent_Factory(t) { return new (t || ThyPopoverHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate), ɵngcc0.ɵɵdirectiveInject(ThyPopover)); };
ThyPopoverHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPopoverHeaderComponent, selectors: [["thy-popover-header"]], inputs: { thyTitleTranslationKey: "thyTitleTranslationKey", thyTitle: "thyTitle" }, outputs: { thyClosed: "thyClosed" }, exportAs: ["thyPopoverHeader"], decls: 5, vars: 1, consts: [[1, "thy-popover-header"], [1, "thy-popover-title"], ["type", "button", 1, "close", 3, "click"], ["thyIconName", "close-bold"]], template: function ThyPopoverHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "h5", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "button", 2);
        ɵngcc0.ɵɵlistener("click", function ThyPopoverHeaderComponent_Template_button_click_3_listener($event) { return ctx.close($event); });
        ɵngcc0.ɵɵelement(4, "thy-icon", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyTitle);
    } }, directives: [ɵngcc5.ThyIconComponent], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverHeaderComponent, [{
        type: i0.Component,
        args: [{
                selector: 'thy-popover-header',
                template: "<div class=\"thy-popover-header\">\n  <h5 class=\"thy-popover-title\">{{ thyTitle }}</h5>\n  <button type=\"button\" class=\"close\" (click)=\"close($event)\">\n    <thy-icon thyIconName=\"close-bold\"></thy-icon>\n  </button>\n</div>\n",
                exportAs: 'thyPopoverHeader'
            }]
    }], function () { return [{ type: ɵngcc1.ThyTranslate }, { type: ThyPopover }]; }, { thyClosed: [{
            type: i0.Output
        }], thyTitleTranslationKey: [{
            type: i0.Input
        }], thyTitle: [{
            type: i0.Input
        }] }); })();
        return ThyPopoverHeaderComponent;
    }());
    ThyPopoverHeaderComponent.ctorParameters = function () { return [
        { type: core.ThyTranslate },
        { type: ThyPopover }
    ]; };
    ThyPopoverHeaderComponent.propDecorators = {
        thyTitle: [{ type: i0.Input }],
        thyTitleTranslationKey: [{ type: i0.Input }],
        thyClosed: [{ type: i0.Output }]
    };

    var ThyPopoverDirective = /** @class */ (function (_super) {
        __extends(ThyPopoverDirective, _super);
        function ThyPopoverDirective(elementRef, platform, focusMonitor, ngZone, popover, viewContainerRef) {
            var _this = _super.call(this, elementRef, platform, focusMonitor, ngZone) || this;
            _this.popover = popover;
            _this.viewContainerRef = viewContainerRef;
            _this.popoverOpened = false;
            _this.showDelay = 0;
            _this.hideDelay = 0;
            _this.tooltipPin = true;
            return _this;
        }
        Object.defineProperty(ThyPopoverDirective.prototype, "thyTrigger", {
            set: function (trigger) {
                this.trigger = trigger;
            },
            enumerable: false,
            configurable: true
        });
        ThyPopoverDirective.prototype.ngOnInit = function () {
            this.initialize();
        };
        ThyPopoverDirective.prototype.createOverlay = function () {
            var _this = this;
            var config = Object.assign({
                origin: this.elementRef.nativeElement,
                hasBackdrop: this.trigger === 'click' || this.trigger === 'focus',
                viewContainerRef: this.viewContainerRef,
                placement: this.thyPlacement,
                offset: this.thyOffset
            }, this.thyConfig);
            this.popoverRef = this.popover.open(this.content, config);
            this.popoverRef.afterClosed().subscribe(function () {
                _this.popoverOpened = false;
            });
            return this.popoverRef.getOverlayRef();
        };
        ThyPopoverDirective.prototype.show = function (delay) {
            var _this = this;
            if (delay === void 0) { delay = this.showDelay; }
            if (this.hideTimeoutId) {
                clearTimeout(this.hideTimeoutId);
                this.hideTimeoutId = null;
            }
            if (this.disabled || (this.overlayRef && this.overlayRef.hasAttached())) {
                return;
            }
            if (this.trigger !== 'hover') {
                delay = 0;
            }
            this.showTimeoutId = setTimeout(function () {
                var overlayRef = _this.createOverlay();
                _this.overlayRef = overlayRef;
                _this.popoverOpened = true;
                _this.showTimeoutId = null;
            }, delay);
        };
        ThyPopoverDirective.prototype.hide = function (delay) {
            var _this = this;
            if (delay === void 0) { delay = this.hideDelay; }
            if (this.showTimeoutId) {
                clearTimeout(this.showTimeoutId);
                this.showTimeoutId = null;
            }
            this.hideTimeoutId = setTimeout(function () {
                if (_this.popoverRef) {
                    _this.popoverRef.close();
                }
                _this.hideTimeoutId = null;
            }, delay);
        };
        ThyPopoverDirective.prototype.ngOnDestroy = function () {
            this.dispose();
        };
ThyPopoverDirective.ɵfac = function ThyPopoverDirective_Factory(t) { return new (t || ThyPopoverDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.FocusMonitor), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ThyPopover), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ThyPopoverDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyPopoverDirective, selectors: [["", "thyPopover", ""]], hostVars: 2, hostBindings: function ThyPopoverDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-popover-opened", ctx.popoverOpened);
    } }, inputs: { thyTrigger: "thyTrigger", content: ["thyPopover", "content"], thyPlacement: "thyPlacement", thyOffset: "thyOffset", thyConfig: "thyConfig", showDelay: ["thyShowDelay", "showDelay"], hideDelay: ["thyHideDelay", "hideDelay"] }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverDirective, [{
        type: i0.Directive,
        args: [{
                selector: '[thyPopover]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc4.Platform }, { type: ɵngcc6.FocusMonitor }, { type: ɵngcc0.NgZone }, { type: ThyPopover }, { type: ɵngcc0.ViewContainerRef }]; }, { thyTrigger: [{
            type: i0.Input
        }], popoverOpened: [{
            type: i0.HostBinding,
            args: ["class.thy-popover-opened"]
        }], content: [{
            type: i0.Input,
            args: ['thyPopover']
        }], thyPlacement: [{
            type: i0.Input
        }], thyOffset: [{
            type: i0.Input
        }], thyConfig: [{
            type: i0.Input
        }], showDelay: [{
            type: i0.Input,
            args: ['thyShowDelay']
        }], hideDelay: [{
            type: i0.Input,
            args: ['thyHideDelay']
        }] }); })();
        return ThyPopoverDirective;
    }(core.ThyOverlayDirectiveBase));
    ThyPopoverDirective.ctorParameters = function () { return [
        { type: i0.ElementRef },
        { type: i5.Platform },
        { type: a11y.FocusMonitor },
        { type: i0.NgZone },
        { type: ThyPopover },
        { type: i0.ViewContainerRef }
    ]; };
    ThyPopoverDirective.propDecorators = {
        popoverOpened: [{ type: i0.HostBinding, args: ["class.thy-popover-opened",] }],
        content: [{ type: i0.Input, args: ['thyPopover',] }],
        thyTrigger: [{ type: i0.Input }],
        thyPlacement: [{ type: i0.Input }],
        thyOffset: [{ type: i0.Input }],
        thyConfig: [{ type: i0.Input }],
        showDelay: [{ type: i0.Input, args: ['thyShowDelay',] }],
        hideDelay: [{ type: i0.Input, args: ['thyHideDelay',] }]
    };

    var ThyPopoverModule = /** @class */ (function () {
        function ThyPopoverModule() {
        }
ThyPopoverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyPopoverModule });
ThyPopoverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyPopoverModule_Factory(t) { return new (t || ThyPopoverModule)(); }, providers: [THY_POPOVER_DEFAULT_CONFIG_PROVIDER, THY_POPOVER_SCROLL_STRATEGY_PROVIDER], imports: [[i4.CommonModule, i1.OverlayModule, portal.PortalModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyPopoverModule, { declarations: [ThyPopoverContainerComponent, ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent], imports: [ɵngcc7.CommonModule, ɵngcc3.OverlayModule, ɵngcc2.PortalModule, ɵngcc5.ThyIconModule], exports: [ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverModule, [{
        type: i0.NgModule,
        args: [{
                declarations: [ThyPopoverContainerComponent, ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent],
                entryComponents: [ThyPopoverContainerComponent],
                imports: [i4.CommonModule, i1.OverlayModule, portal.PortalModule, icon.ThyIconModule],
                exports: [ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent],
                providers: [THY_POPOVER_DEFAULT_CONFIG_PROVIDER, THY_POPOVER_SCROLL_STRATEGY_PROVIDER]
            }]
    }], function () { return []; }, null); })();
        return ThyPopoverModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.THY_POPOVER_DEFAULT_CONFIG = THY_POPOVER_DEFAULT_CONFIG;
    exports.THY_POPOVER_DEFAULT_CONFIG_PROVIDER = THY_POPOVER_DEFAULT_CONFIG_PROVIDER;
    exports.THY_POPOVER_DEFAULT_CONFIG_VALUE = THY_POPOVER_DEFAULT_CONFIG_VALUE;
    exports.THY_POPOVER_SCROLL_STRATEGY = THY_POPOVER_SCROLL_STRATEGY;
    exports.THY_POPOVER_SCROLL_STRATEGY_FACTORY = THY_POPOVER_SCROLL_STRATEGY_FACTORY;
    exports.THY_POPOVER_SCROLL_STRATEGY_PROVIDER = THY_POPOVER_SCROLL_STRATEGY_PROVIDER;
    exports.ThyInternalPopoverRef = ThyInternalPopoverRef;
    exports.ThyPopover = ThyPopover;
    exports.ThyPopoverBodyComponent = ThyPopoverBodyComponent;
    exports.ThyPopoverConfig = ThyPopoverConfig;
    exports.ThyPopoverContainerComponent = ThyPopoverContainerComponent;
    exports.ThyPopoverDirective = ThyPopoverDirective;
    exports.ThyPopoverHeaderComponent = ThyPopoverHeaderComponent;
    exports.ThyPopoverModule = ThyPopoverModule;
    exports.ThyPopoverRef = ThyPopoverRef;
    exports.thyPopoverAnimations = thyPopoverAnimations;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
