// ngx-tethys/bundles/ngx-tethys-cascader.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/cdk/overlay'),require('@angular/common'),require('ngx-tethys/input'),require('ngx-tethys/icon'),require('@angular/forms'),exports, require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('@angular/cdk/overlay'), require('ngx-tethys/input'), require('ngx-tethys/core'), require('ngx-tethys/util'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/cascader', ['@angular/core','ngx-tethys/core','@angular/cdk/overlay','@angular/common','ngx-tethys/input','ngx-tethys/icon','@angular/forms','exports', '@angular/core', '@angular/common', '@angular/forms', '@angular/cdk/overlay', 'ngx-tethys/input', 'ngx-tethys/core', 'ngx-tethys/util', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.cdk.overlay,global.ng.common,global.ngxTethys.input,global.ngxTethys.icon,global.ng.forms,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].cascader = {}), global.ng.core, global.ng.common, global.ng.forms, global.ng.cdk.overlay, global['ngx-tethys'].input, global['ngx-tethys'].core, global['ngx-tethys'].util, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,exports, core$1, common, forms, overlay, input, core, util, icon) { 
var _c0 = ["input"];
var _c1 = ["menu"];
function ThyCascaderComponent_ng_container_3_span_5_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "span", 11);
    ɵngcc0.ɵɵlistener("click", function ThyCascaderComponent_ng_container_3_span_5_Template_span_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); var ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9.clearSelection($event); });
    ɵngcc0.ɵɵelement(1, "thy-icon", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r5.clearCls);
} }
function ThyCascaderComponent_ng_container_3_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r6.labelRenderText);
} }
function ThyCascaderComponent_ng_container_3_ng_template_8_ng_template_0_Template(rf, ctx) { }
function ThyCascaderComponent_ng_container_3_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyCascaderComponent_ng_container_3_ng_template_8_ng_template_0_Template, 0, 0, "ng-template", 13);
} if (rf & 2) {
    var ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r8.thyLabelRender)("ngTemplateOutletContext", ctx_r8.labelRenderContext);
} }
function ThyCascaderComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "input", 4, 5);
    ɵngcc0.ɵɵelementStart(3, "span", 6);
    ɵngcc0.ɵɵelement(4, "thy-icon", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, ThyCascaderComponent_ng_container_3_span_5_Template, 2, 1, "span", 8);
    ɵngcc0.ɵɵelementStart(6, "span", 6);
    ɵngcc0.ɵɵtemplate(7, ThyCascaderComponent_ng_container_3_ng_container_7_Template, 2, 1, "ng-container", 9);
    ɵngcc0.ɵɵtemplate(8, ThyCascaderComponent_ng_container_3_ng_template_8_Template, 1, 2, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var _r7 = ɵngcc0.ɵɵreference(9);
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thySize", ctx_r2.thySize)("ngClass", ctx_r2.inputCls)("disabled", ctx_r2.disabled)("placeholder", ctx_r2.showPlaceholder ? ctx_r2.thyPlaceHolder : "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.arrowCls);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.isLoading && ctx_r2.labelRenderText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.labelCls);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.isLabelRenderTemplate)("ngIfElse", _r7);
} }
function ThyCascaderComponent_ng_template_4_ul_2_li_1_Template(rf, ctx) { if (rf & 1) {
    var _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 18);
    ɵngcc0.ɵɵlistener("click", function ThyCascaderComponent_ng_template_4_ul_2_li_1_Template_li_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); var option_r17 = ctx.$implicit; var i_r15 = ɵngcc0.ɵɵnextContext().index; var ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.onOptionClick(option_r17, i_r15, $event); })("mouseover", function ThyCascaderComponent_ng_template_4_ul_2_li_1_Template_li_mouseover_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); var option_r17 = ctx.$implicit; var i_r15 = ɵngcc0.ɵɵnextContext().index; var ctx_r21 = ɵngcc0.ɵɵnextContext(2); return ctx_r21.onOptionMouseover(option_r17, i_r15, $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var option_r17 = ctx.$implicit;
    var i_r15 = ɵngcc0.ɵɵnextContext().index;
    var ctx_r16 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("option", option_r17)("thyLabelProperty", ctx_r16.thyLabelProperty)("active", ctx_r16.isActivedOption(option_r17, i_r15));
} }
function ThyCascaderComponent_ng_template_4_ul_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 6);
    ɵngcc0.ɵɵtemplate(1, ThyCascaderComponent_ng_template_4_ul_2_li_1_Template, 1, 3, "li", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var options_r14 = ctx.$implicit;
    var ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r13.menuColumnCls);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", options_r14);
} }
function ThyCascaderComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    var _r25 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 14, 15);
    ɵngcc0.ɵɵlistener("mouseleave", function ThyCascaderComponent_ng_template_4_Template_div_mouseleave_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r25); var ctx_r24 = ɵngcc0.ɵɵnextContext(); return ctx_r24.onMenuMouseleave($event); });
    ɵngcc0.ɵɵtemplate(2, ThyCascaderComponent_ng_template_4_ul_2_Template, 2, 2, "ul", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.menuCls)("ngStyle", ctx_r3.thyMenuStyle);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.thyColumns);
} }
var _c2 = ["thy-cascader-option", ""];
function ThyCascaderOptionComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 1);
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    function toArray(value) {
        var ret;
        if (value == null) {
            ret = [];
        }
        else if (!Array.isArray(value)) {
            ret = [value];
        }
        else {
            ret = value;
        }
        return ret;
    }
    function arrayEquals(array1, array2) {
        if (!array1 || !array2 || array1.length !== array2.length) {
            return false;
        }
        var len = array1.length;
        for (var i = 0; i < len; i++) {
            if (array1[i] !== array2[i]) {
                return false;
            }
        }
        return true;
    }
    var defaultDisplayRender = function (label) { return label.join(' / '); };
    var ɵ0 = defaultDisplayRender;
    var ThyCascaderComponent = /** @class */ (function () {
        function ThyCascaderComponent(cdr, elementRef, updateHostClassService) {
            this.cdr = cdr;
            this.elementRef = elementRef;
            this.updateHostClassService = updateHostClassService;
            this.changeOnSelect = false;
            this.showInput = true;
            this.prefixCls = 'thy-cascader';
            this.dropDownPosition = 'bottom';
            this.menuVisible = false;
            this.isLoading = false;
            this.isOpening = false;
            this.showSearch = false;
            this._thySize = 'md';
            this.isLabelRenderTemplate = false;
            this.labelRenderContext = {};
            this.onChange = Function.prototype;
            this.onTouched = Function.prototype;
            this.cascaderPositon = __spread(core.EXPANDED_DROPDOWN_POSITIONS);
            this.selectedOptions = [];
            this.activatedOptions = [];
            this.thyColumns = [];
            this.thyValueProperty = 'value';
            this.thyLabelProperty = 'label';
            this.thyPlaceHolder = '请选择';
            this._inputValue = '';
            this.inSearch = false;
            this.thyTriggerAction = ['click'];
            this.thyExpandTriggerAction = ['click'];
            this.disabled = false;
            this.thyChange = new core$1.EventEmitter();
            this.thySelectionChange = new core$1.EventEmitter();
            this.thySelect = new core$1.EventEmitter();
            this.thyClear = new core$1.EventEmitter();
            updateHostClassService.initializeElement(elementRef.nativeElement);
        }
        Object.defineProperty(ThyCascaderComponent.prototype, "thyLabelRender", {
            get: function () {
                return this.labelRenderTpl;
            },
            set: function (value) {
                this.labelRenderTpl = value;
                this.isLabelRenderTemplate = value instanceof core$1.TemplateRef;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "inputValue", {
            get: function () {
                return this._inputValue;
            },
            set: function (inputValue) {
                this._inputValue = inputValue;
                var willBeInSearch = !!inputValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "thyChangeOnSelect", {
            get: function () {
                return this.changeOnSelect;
            },
            set: function (value) {
                this.changeOnSelect = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "thyShowInput", {
            get: function () {
                return this.showInput;
            },
            set: function (value) {
                this.showInput = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "thyOptions", {
            set: function (options) {
                this.oldColumnsHolder = this.thyColumns = options && options.length ? [options] : [];
                if (!this.inSearch) {
                    if (this.defaultValue && this.thyColumns.length) {
                        this.initOptions(0);
                    }
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "thyMenuClassName", {
            get: function () {
                return this.menuClassName;
            },
            set: function (value) {
                this.menuClassName = value;
                this.setMenuClass();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "thyColumnClassName", {
            get: function () {
                return this.columnClassName;
            },
            set: function (value) {
                this.columnClassName = value;
                this.setMenuClass();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderComponent.prototype, "thySize", {
            get: function () {
                return this._thySize;
            },
            set: function (v) {
                this._thySize = v;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.ngOnInit = function () {
            this.setClassMap();
            this.setMenuClass();
            this.setMenuColumnClass();
            this.setArrowClass();
            this.setLabelClass();
            this.setClearClass();
            this.setInputClass();
            this.initPosition();
        };
        ThyCascaderComponent.prototype.initPosition = function () {
            this.cascaderPositon[0].offsetY = 10; // 左下
            this.cascaderPositon[1].offsetY = 10; // 右下
            this.cascaderPositon[2].offsetY = -10; // 右下
            this.cascaderPositon[3].offsetY = -10; // 右下
            this.positions = this.cascaderPositon;
        };
        ThyCascaderComponent.prototype.initOptions = function (index) {
            var _this = this;
            var vs = this.defaultValue;
            var load = function () {
                _this.activateOnInit(index, vs[index]);
                if (index < vs.length - 1) {
                    _this.initOptions(index + 1);
                }
                if (index === vs.length - 1) {
                    _this.afterWriteValue();
                }
            };
            if (this.isLoaded(index) || !this.thyLoadData) {
                load();
            }
            else {
                var node = this.activatedOptions[index - 1] || {};
                this.loadChildren(node, index - 1, load, this.afterWriteValue.bind(this));
            }
        };
        ThyCascaderComponent.prototype.activateOnInit = function (index, value) {
            var _a;
            var option = this.findOption(value, index);
            if (!option) {
                option =
                    typeof value === 'object'
                        ? value
                        : (_a = {},
                            _a["" + (this.thyValueProperty || 'value')] = value,
                            _a["" + (this.thyLabelProperty || 'label')] = value,
                            _a);
            }
            this.setActiveOption(option, index, false, false);
        };
        ThyCascaderComponent.prototype.writeValue = function (value) {
            var vs = (this.defaultValue = toArray(value));
            if (vs.length) {
                this.initOptions(0);
            }
            else {
                this.value = vs;
                this.activatedOptions = [];
                this.afterWriteValue();
            }
        };
        ThyCascaderComponent.prototype.afterWriteValue = function () {
            this.selectedOptions = this.activatedOptions;
            this.value = this.getSubmitValue();
            this.buildDisplayLabel();
        };
        ThyCascaderComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        ThyCascaderComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        ThyCascaderComponent.prototype.onPositionChange = function (position) {
            var newValue = position.connectionPair.originY === 'bottom' ? 'bottom' : 'top';
            if (this.dropDownPosition !== newValue) {
                this.dropDownPosition = newValue;
                this.cdr.detectChanges();
            }
        };
        ThyCascaderComponent.prototype.isLoaded = function (index) {
            return this.thyColumns[index] && this.thyColumns[index].length > 0;
        };
        ThyCascaderComponent.prototype.getOptionLabel = function (option) {
            return option[this.thyLabelProperty || 'label'];
        };
        ThyCascaderComponent.prototype.getOptionValue = function (option) {
            return option[this.thyValueProperty || 'value'];
        };
        ThyCascaderComponent.prototype.hasInput = function () {
            return this.inputValue.length > 0;
        };
        ThyCascaderComponent.prototype.hasValue = function () {
            return this.value && this.value.length > 0;
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "showPlaceholder", {
            get: function () {
                return !(this.hasInput() || this.hasValue());
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.isActivedOption = function (option, index) {
            var activeOpt = this.activatedOptions[index];
            return activeOpt === option;
        };
        ThyCascaderComponent.prototype.findOption = function (option, index) {
            var _this = this;
            var options = this.thyColumns[index];
            if (options) {
                var value_1 = typeof option === 'object' ? this.getOptionValue(option) : option;
                return options.find(function (o) { return value_1 === _this.getOptionValue(o); });
            }
            return null;
        };
        ThyCascaderComponent.prototype.buildDisplayLabel = function () {
            var _this = this;
            var selectedOptions = this.selectedOptions;
            var labels = selectedOptions.map(function (o) { return _this.getOptionLabel(o); });
            if (this.isLabelRenderTemplate) {
                this.labelRenderContext = { labels: labels, selectedOptions: selectedOptions };
            }
            else {
                this.labelRenderText = defaultDisplayRender.call(this, labels, selectedOptions);
            }
        };
        ThyCascaderComponent.prototype.isMenuVisible = function () {
            return this.menuVisible;
        };
        ThyCascaderComponent.prototype.setMenuVisible = function (menuVisible) {
            if (this.menuVisible !== menuVisible) {
                this.menuVisible = menuVisible;
                this.setClassMap();
                this.setArrowClass();
                this.setMenuClass();
            }
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "menuCls", {
            get: function () {
                return this._menuCls;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.setMenuClass = function () {
            var _a;
            this._menuCls = (_a = {},
                _a[this.prefixCls + "-menus"] = true,
                _a[this.prefixCls + "-menus-hidden"] = !this.menuVisible,
                _a["" + this.thyMenuClassName] = this.thyMenuClassName,
                _a);
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "menuColumnCls", {
            get: function () {
                return this._menuColumnCls;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.setMenuColumnClass = function () {
            var _a;
            this._menuColumnCls = (_a = {},
                _a[this.prefixCls + "-menu"] = true,
                _a["" + this.thyColumnClassName] = this.thyColumnClassName,
                _a);
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "arrowCls", {
            get: function () {
                return this._arrowCls;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.setArrowClass = function () {
            var _a;
            this._arrowCls = (_a = {},
                _a[this.prefixCls + "-picker-arrow"] = true,
                _a[this.prefixCls + "-picker-arrow-expand"] = this.menuVisible,
                _a);
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "clearCls", {
            get: function () {
                return this._clearCls;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.setClearClass = function () {
            var _a;
            this._clearCls = (_a = {},
                _a[this.prefixCls + "-picker-clear"] = true,
                _a["remove-link"] = true,
                _a);
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "labelCls", {
            get: function () {
                return this._labelCls;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.setLabelClass = function () {
            var _a;
            this._labelCls = (_a = {},
                _a[this.prefixCls + "-picker-label"] = true,
                _a[this.prefixCls + "-show-search"] = false,
                _a[this.prefixCls + "-focused"] = false,
                _a);
        };
        Object.defineProperty(ThyCascaderComponent.prototype, "inputCls", {
            get: function () {
                return this._inputCls;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderComponent.prototype.setInputClass = function () {
            var _a;
            this._inputCls = (_a = {},
                _a[this.prefixCls + "-input"] = true,
                _a);
        };
        ThyCascaderComponent.prototype.setClassMap = function () {
            var _a;
            var classMap = (_a = {},
                _a["" + this.prefixCls] = true,
                _a[this.prefixCls + "-picker"] = true,
                _a[this.prefixCls + "-" + this.thySize] = true,
                _a[this.prefixCls + "-picker-disabled"] = this.disabled,
                _a[this.prefixCls + "-picker-open"] = this.menuVisible,
                _a);
            this.updateHostClassService.updateClassByMap(classMap);
        };
        ThyCascaderComponent.prototype.isClickTriggerAction = function () {
            if (typeof this.thyTriggerAction === 'string') {
                return this.thyTriggerAction === 'click';
            }
            return this.thyTriggerAction.indexOf('click') !== -1;
        };
        ThyCascaderComponent.prototype.isHoverTriggerAction = function () {
            if (typeof this.thyTriggerAction === 'string') {
                return this.thyTriggerAction === 'hover';
            }
            return this.thyTriggerAction.indexOf('hover') !== -1;
        };
        ThyCascaderComponent.prototype.isHoverExpandTriggerAction = function () {
            if (typeof this.thyExpandTriggerAction === 'string') {
                return this.thyExpandTriggerAction === 'hover';
            }
            return this.thyExpandTriggerAction.indexOf('hover') !== -1;
        };
        ThyCascaderComponent.prototype.trggleClick = function ($event) {
            if (this.disabled) {
                return;
            }
            this.onTouched();
            if (this.isClickTriggerAction()) {
                this.setMenuVisible(!this.menuVisible);
            }
        };
        ThyCascaderComponent.prototype.trggleHover = function ($event) {
            if (this.disabled) {
                return;
            }
            if (this.isHoverTriggerAction()) {
                this.setMenuVisible(!this.menuVisible);
            }
        };
        ThyCascaderComponent.prototype.onOptionClick = function (option, index, event) {
            if (event) {
                event.preventDefault();
            }
            if (option && option.disabled) {
                return;
            }
            this.setActiveOption(option, index, true);
        };
        ThyCascaderComponent.prototype.onOptionMouseover = function (option, index, event) {
            if (event) {
                event.preventDefault();
            }
            if (option && option.disabled) {
                return;
            }
            if (!this.isHoverExpandTriggerAction()) {
                return;
            }
            this.setActiveOption(option, index, false);
        };
        ThyCascaderComponent.prototype.onMenuMouseleave = function (event) {
            if (event) {
                event.preventDefault();
            }
            if (!this.isHoverTriggerAction()) {
                return;
            }
            this.setMenuVisible(!this.menuVisible);
        };
        ThyCascaderComponent.prototype.closeMenu = function () {
            this.setMenuVisible(false);
            this.setArrowClass();
        };
        ThyCascaderComponent.prototype.setActiveOption = function (option, index, select, loadChildren) {
            if (loadChildren === void 0) { loadChildren = true; }
            if (!option || option.disabled) {
                return;
            }
            this.activatedOptions[index] = option;
            for (var i = index - 1; i >= 0; i--) {
                if (!this.activatedOptions[i]) {
                    this.activatedOptions[i] = this.activatedOptions[i + 1].parent;
                }
            }
            if (index < this.activatedOptions.length - 1) {
                this.activatedOptions = this.activatedOptions.slice(0, index + 1);
            }
            if (option.children && option.children.length) {
                option.isLeaf = false;
                option.children.forEach(function (child) { return (child.parent = option); });
                this.setColumnData(option.children, index + 1);
            }
            else if (!option.isLeaf && loadChildren) {
                this.loadChildren(option, index);
            }
            else {
                if (index < this.thyColumns.length - 1) {
                    this.thyColumns = this.thyColumns.slice(0, index + 1);
                }
            }
            if (select) {
                this.onSelectOption(option, index);
            }
        };
        ThyCascaderComponent.prototype.onSelectOption = function (option, index) {
            this.thySelect.emit({ option: option, index: index });
            if (option.isLeaf || this.thyChangeOnSelect || this.shouldPerformSelection(option, index)) {
                this.selectedOptions = this.activatedOptions;
                this.buildDisplayLabel();
                this.onValueChange();
            }
            if (option.isLeaf) {
                this.setMenuVisible(false);
            }
        };
        ThyCascaderComponent.prototype.shouldPerformSelection = function (option, level) {
            return typeof this.thyChangeOn === 'function' ? this.thyChangeOn(option, level) === true : false;
        };
        ThyCascaderComponent.prototype.onValueChange = function () {
            var value = this.getSubmitValue();
            if (!arrayEquals(this.value, value)) {
                this.defaultValue = null;
                this.value = value;
                this.onChange(value);
                if (value.length === 0) {
                    this.thyClear.emit();
                }
                this.thySelectionChange.emit(this.selectedOptions);
                this.thyChange.emit(value);
            }
        };
        ThyCascaderComponent.prototype.clearSelection = function ($event) {
            if ($event) {
                $event.stopPropagation();
                $event.preventDefault();
            }
            this.labelRenderText = '';
            this.labelRenderContext = {};
            this.selectedOptions = [];
            this.activatedOptions = [];
            this.inputValue = '';
            this.setMenuVisible(false);
            this.onValueChange();
        };
        ThyCascaderComponent.prototype.focus = function () { };
        ThyCascaderComponent.prototype.loadChildren = function (option, index, success, failure) {
            var _this = this;
            if (this.thyLoadData) {
                this.isLoading = true;
                this.thyLoadData(option, index).then(function () {
                    option.loading = _this.isLoading = false;
                    if (option.children) {
                        option.children.forEach(function (child) { return (child.parent = index < 0 ? undefined : option); });
                        _this.setColumnData(option.children, index + 1);
                    }
                    if (success) {
                        success();
                    }
                }, function () {
                    option.loading = _this.isLoading = false;
                    option.isLeaf = true;
                    if (failure) {
                        failure();
                    }
                });
            }
        };
        ThyCascaderComponent.prototype.setColumnData = function (options, index) {
            if (!arrayEquals(this.thyColumns[index], options)) {
                this.thyColumns[index] = options;
                if (index < this.thyColumns.length - 1) {
                    this.thyColumns = this.thyColumns.slice(0, index + 1);
                }
            }
        };
        ThyCascaderComponent.prototype.getSubmitValue = function () {
            var _this = this;
            var values = [];
            this.selectedOptions.forEach(function (option) {
                values.push(_this.getOptionValue(option));
            });
            return values;
        };
ThyCascaderComponent.ɵfac = function ThyCascaderComponent_Factory(t) { return new (t || ThyCascaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService)); };
ThyCascaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCascaderComponent, selectors: [["thy-cascader"], ["", "thy-cascader", ""]], viewQuery: function ThyCascaderComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
        ɵngcc0.ɵɵviewQuery(_c1, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.input = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.menu = _t.first);
    } }, hostBindings: function ThyCascaderComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyCascaderComponent_click_HostBindingHandler($event) { return ctx.trggleClick($event); })("mouseover", function ThyCascaderComponent_mouseover_HostBindingHandler($event) { return ctx.trggleHover($event); });
    } }, inputs: { thyValueProperty: "thyValueProperty", thyLabelProperty: "thyLabelProperty", thyPlaceHolder: "thyPlaceHolder", thyTriggerAction: "thyTriggerAction", thyExpandTriggerAction: "thyExpandTriggerAction", disabled: "disabled", thyLabelRender: "thyLabelRender", thyChangeOnSelect: "thyChangeOnSelect", thyShowInput: "thyShowInput", thyOptions: "thyOptions", thyMenuClassName: "thyMenuClassName", thyColumnClassName: "thyColumnClassName", thySize: "thySize", thyLoadData: "thyLoadData", thyMenuStyle: "thyMenuStyle", thyChangeOn: "thyChangeOn" }, outputs: { thyChange: "thyChange", thySelectionChange: "thySelectionChange", thySelect: "thySelect", thyClear: "thyClear" }, features: [ɵngcc0.ɵɵProvidersFeature([
            core.UpdateHostClassService,
            {
                provide: forms.NG_VALUE_ACCESSOR,
                useExisting: core$1.forwardRef(function () { return ThyCascaderComponent; }),
                multi: true
            }
        ])], decls: 5, vars: 4, consts: [["cdkOverlayOrigin", ""], ["origin", "cdkOverlayOrigin", "trigger", ""], [4, "ngIf"], ["cdkConnectedOverlay", "", "cdkConnectedOverlayHasBackdrop", "", "cdkConnectedOverlayBackdropClass", "thy-cascader-backdrop", 3, "cdkConnectedOverlayOrigin", "cdkConnectedOverlayPositions", "cdkConnectedOverlayOpen", "backdropClick", "detach", "positionChange"], ["thyInput", "", "type", "text", "readonly", "", 3, "thySize", "ngClass", "disabled", "placeholder"], ["input", ""], [3, "ngClass"], ["thyIconName", "angle-down", 1, "font-size-base"], [3, "ngClass", "click", 4, "ngIf"], [4, "ngIf", "ngIfElse"], ["labelTemplate", ""], [3, "ngClass", "click"], ["thyIconName", "close-circle-bold-fill", 1, "remove-link-icon", "font-size-base"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [3, "ngClass", "ngStyle", "mouseleave"], ["menu", ""], [3, "ngClass", 4, "ngFor", "ngForOf"], ["thy-cascader-option", "", 3, "option", "thyLabelProperty", "active", "click", "mouseover", 4, "ngFor", "ngForOf"], ["thy-cascader-option", "", 3, "option", "thyLabelProperty", "active", "click", "mouseover"]], template: function ThyCascaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵtemplate(3, ThyCascaderComponent_ng_container_3_Template, 10, 9, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, ThyCascaderComponent_ng_template_4_Template, 3, 3, "ng-template", 3);
        ɵngcc0.ɵɵlistener("backdropClick", function ThyCascaderComponent_Template_ng_template_backdropClick_4_listener() { return ctx.closeMenu(); })("detach", function ThyCascaderComponent_Template_ng_template_detach_4_listener() { return ctx.closeMenu(); })("positionChange", function ThyCascaderComponent_Template_ng_template_positionChange_4_listener($event) { return ctx.onPositionChange($event); });
    } if (rf & 2) {
        var _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowInput);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("cdkConnectedOverlayOrigin", _r0)("cdkConnectedOverlayPositions", ctx.positions)("cdkConnectedOverlayOpen", ctx.menuVisible);
    } }, directives: function () { return [ɵngcc2.CdkOverlayOrigin, ɵngcc3.NgIf, ɵngcc2.CdkConnectedOverlay, ɵngcc4.ThyInputDirective, ɵngcc3.NgClass, ɵngcc5.ThyIconComponent, ɵngcc3.NgTemplateOutlet, ɵngcc3.NgStyle, ɵngcc3.NgForOf, ThyCascaderOptionComponent]; }, styles: [".thy-cascader-menus[_ngcontent-%COMP%] {\n                position: relative;\n            }"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCascaderComponent, [{
        type: core$1.Component,
        args: [{
                selector: 'thy-cascader,[thy-cascader]',
                template: "<div cdkOverlayOrigin #origin=\"cdkOverlayOrigin\" #trigger>\n  <ng-container *ngIf=\"thyShowInput\">\n    <input\n      #input\n      thyInput\n      type=\"text\"\n      [thySize]=\"thySize\"\n      [ngClass]=\"inputCls\"\n      readonly\n      [disabled]=\"disabled\"\n      [placeholder]=\"showPlaceholder ? thyPlaceHolder : ''\"\n    />\n    <span [ngClass]=\"arrowCls\">\n      <thy-icon thyIconName=\"angle-down\" class=\"font-size-base\"></thy-icon>\n    </span>\n    <span [ngClass]=\"clearCls\" *ngIf=\"!isLoading && labelRenderText\" (click)=\"clearSelection($event)\">\n      <thy-icon class=\"remove-link-icon font-size-base\" thyIconName=\"close-circle-bold-fill\"></thy-icon>\n    </span>\n    <span [ngClass]=\"labelCls\">\n      <ng-container *ngIf=\"!isLabelRenderTemplate; else labelTemplate\">{{ labelRenderText }}</ng-container>\n      <ng-template #labelTemplate>\n        <ng-template [ngTemplateOutlet]=\"thyLabelRender\" [ngTemplateOutletContext]=\"labelRenderContext\"></ng-template>\n      </ng-template>\n    </span>\n  </ng-container>\n</div>\n<ng-template\n  cdkConnectedOverlay\n  cdkConnectedOverlayHasBackdrop\n  cdkConnectedOverlayBackdropClass=\"thy-cascader-backdrop\"\n  [cdkConnectedOverlayOrigin]=\"origin\"\n  [cdkConnectedOverlayPositions]=\"positions\"\n  (backdropClick)=\"closeMenu()\"\n  (detach)=\"closeMenu()\"\n  (positionChange)=\"onPositionChange($event)\"\n  [cdkConnectedOverlayOpen]=\"menuVisible\"\n>\n  <div #menu [ngClass]=\"menuCls\" [ngStyle]=\"thyMenuStyle\" (mouseleave)=\"onMenuMouseleave($event)\">\n    <ul *ngFor=\"let options of thyColumns; let i = index\" [ngClass]=\"menuColumnCls\">\n      <li\n        thy-cascader-option\n        *ngFor=\"let option of options\"\n        [option]=\"option\"\n        (click)=\"onOptionClick(option, i, $event)\"\n        (mouseover)=\"onOptionMouseover(option, i, $event)\"\n        [thyLabelProperty]=\"thyLabelProperty\"\n        [active]=\"isActivedOption(option, i)\"\n      ></li>\n    </ul>\n  </div>\n</ng-template>\n",
                providers: [
                    core.UpdateHostClassService,
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core$1.forwardRef(function () { return ThyCascaderComponent; }),
                        multi: true
                    }
                ],
                styles: ["\n            .thy-cascader-menus {\n                position: relative;\n            }\n        "]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.ElementRef }, { type: ɵngcc1.UpdateHostClassService }]; }, { thyValueProperty: [{
            type: core$1.Input
        }], thyLabelProperty: [{
            type: core$1.Input
        }], thyPlaceHolder: [{
            type: core$1.Input
        }], thyTriggerAction: [{
            type: core$1.Input
        }], thyExpandTriggerAction: [{
            type: core$1.Input
        }], disabled: [{
            type: core$1.Input
        }], thyChange: [{
            type: core$1.Output
        }], thySelectionChange: [{
            type: core$1.Output
        }], thySelect: [{
            type: core$1.Output
        }], thyClear: [{
            type: core$1.Output
        }], thyLabelRender: [{
            type: core$1.Input
        }], thyChangeOnSelect: [{
            type: core$1.Input
        }], thyShowInput: [{
            type: core$1.Input
        }], thyOptions: [{
            type: core$1.Input
        }], thyMenuClassName: [{
            type: core$1.Input
        }], thyColumnClassName: [{
            type: core$1.Input
        }], thySize: [{
            type: core$1.Input
        }], trggleClick: [{
            type: core$1.HostListener,
            args: ['click', ['$event']]
        }], trggleHover: [{
            type: core$1.HostListener,
            args: ['mouseover', ['$event']]
        }], thyLoadData: [{
            type: core$1.Input
        }], thyMenuStyle: [{
            type: core$1.Input
        }], thyChangeOn: [{
            type: core$1.Input
        }], input: [{
            type: core$1.ViewChild,
            args: ['input']
        }], menu: [{
            type: core$1.ViewChild,
            args: ['menu']
        }] }); })();
        return ThyCascaderComponent;
    }());
    ThyCascaderComponent.ctorParameters = function () { return [
        { type: core$1.ChangeDetectorRef },
        { type: core$1.ElementRef },
        { type: core.UpdateHostClassService }
    ]; };
    ThyCascaderComponent.propDecorators = {
        thyLabelRender: [{ type: core$1.Input }],
        thyValueProperty: [{ type: core$1.Input }],
        thyLabelProperty: [{ type: core$1.Input }],
        thyPlaceHolder: [{ type: core$1.Input }],
        thyLoadData: [{ type: core$1.Input }],
        thyChangeOnSelect: [{ type: core$1.Input }],
        thyShowInput: [{ type: core$1.Input }],
        thyTriggerAction: [{ type: core$1.Input }],
        thyExpandTriggerAction: [{ type: core$1.Input }],
        thyMenuStyle: [{ type: core$1.Input }],
        thyOptions: [{ type: core$1.Input }],
        thyMenuClassName: [{ type: core$1.Input }],
        thyColumnClassName: [{ type: core$1.Input }],
        disabled: [{ type: core$1.Input }],
        thySize: [{ type: core$1.Input }],
        thyChange: [{ type: core$1.Output }],
        thySelectionChange: [{ type: core$1.Output }],
        thySelect: [{ type: core$1.Output }],
        thyChangeOn: [{ type: core$1.Input }],
        thyClear: [{ type: core$1.Output }],
        input: [{ type: core$1.ViewChild, args: ['input',] }],
        menu: [{ type: core$1.ViewChild, args: ['menu',] }],
        trggleClick: [{ type: core$1.HostListener, args: ['click', ['$event'],] }],
        trggleHover: [{ type: core$1.HostListener, args: ['mouseover', ['$event'],] }]
    };

    var ThyCascaderOptionComponent = /** @class */ (function () {
        function ThyCascaderOptionComponent() {
            this.item = true;
            this.active = false;
        }
        Object.defineProperty(ThyCascaderOptionComponent.prototype, "disabled", {
            get: function () {
                return this.option.disabled;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCascaderOptionComponent.prototype, "expand", {
            get: function () {
                return this.option && !this.option.isLeaf;
            },
            enumerable: false,
            configurable: true
        });
        ThyCascaderOptionComponent.prototype.getOptionLabel = function () {
            return this.option ? this.option[this.thyLabelProperty || 'label'] : '';
        };
        ThyCascaderOptionComponent.prototype.ngOnInit = function () { };
ThyCascaderOptionComponent.ɵfac = function ThyCascaderOptionComponent_Factory(t) { return new (t || ThyCascaderOptionComponent)(); };
ThyCascaderOptionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCascaderOptionComponent, selectors: [["", "thy-cascader-option", ""]], hostVars: 8, hostBindings: function ThyCascaderOptionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-cascader-menu-item", ctx.item)("thy-cascader-menu-item-active", ctx.active)("thy-cascader-menu-item-disabled", ctx.disabled)("thy-cascader-menu-item-expand", ctx.expand);
    } }, inputs: { active: "active", option: "option", thyLabelProperty: "thyLabelProperty" }, features: [ɵngcc0.ɵɵProvidersFeature([core.UpdateHostClassService])], attrs: _c2, decls: 3, vars: 2, consts: [["class", "wtf wtf-angle-right thy-cascader-menu-item-expand-icon", 4, "ngIf"], [1, "wtf", "wtf-angle-right", "thy-cascader-menu-item-expand-icon"]], template: function ThyCascaderOptionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(2, ThyCascaderOptionComponent_span_2_Template, 1, 0, "span", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.getOptionLabel(), "\n");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.option.isLeaf || ctx.option.children && ctx.option.children.length || ctx.option.loading);
    } }, directives: [ɵngcc3.NgIf], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCascaderOptionComponent, [{
        type: core$1.Component,
        args: [{
                changeDetection: core$1.ChangeDetectionStrategy.OnPush,
                encapsulation: core$1.ViewEncapsulation.None,
                selector: '[thy-cascader-option]',
                template: "<ng-container>\n    {{ getOptionLabel() }}\n</ng-container>\n<span *ngIf=\"!option.isLeaf || option.children && option.children.length || option.loading\" class=\"wtf wtf-angle-right thy-cascader-menu-item-expand-icon\">\n</span>",
                providers: [core.UpdateHostClassService]
            }]
    }], function () { return []; }, { item: [{
            type: core$1.HostBinding,
            args: ['class.thy-cascader-menu-item']
        }], active: [{
            type: core$1.HostBinding,
            args: ['class.thy-cascader-menu-item-active']
        }, {
            type: core$1.Input
        }], disabled: [{
            type: core$1.HostBinding,
            args: ['class.thy-cascader-menu-item-disabled']
        }], expand: [{
            type: core$1.HostBinding,
            args: ['class.thy-cascader-menu-item-expand']
        }], option: [{
            type: core$1.Input
        }], thyLabelProperty: [{
            type: core$1.Input
        }] }); })();
        return ThyCascaderOptionComponent;
    }());
    ThyCascaderOptionComponent.ctorParameters = function () { return []; };
    ThyCascaderOptionComponent.propDecorators = {
        option: [{ type: core$1.Input }],
        item: [{ type: core$1.HostBinding, args: ['class.thy-cascader-menu-item',] }],
        active: [{ type: core$1.HostBinding, args: ['class.thy-cascader-menu-item-active',] }, { type: core$1.Input }],
        disabled: [{ type: core$1.HostBinding, args: ['class.thy-cascader-menu-item-disabled',] }],
        expand: [{ type: core$1.HostBinding, args: ['class.thy-cascader-menu-item-expand',] }],
        thyLabelProperty: [{ type: core$1.Input }]
    };

    var ThyCascaderModule = /** @class */ (function () {
        function ThyCascaderModule() {
        }
ThyCascaderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCascaderModule });
ThyCascaderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCascaderModule_Factory(t) { return new (t || ThyCascaderModule)(); }, imports: [[common.CommonModule, forms.FormsModule, overlay.OverlayModule, input.ThyInputModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCascaderModule, { declarations: [ThyCascaderComponent, ThyCascaderOptionComponent], imports: [ɵngcc3.CommonModule, ɵngcc6.FormsModule, ɵngcc2.OverlayModule, ɵngcc4.ThyInputModule, ɵngcc5.ThyIconModule], exports: [ThyCascaderComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCascaderModule, [{
        type: core$1.NgModule,
        args: [{
                imports: [common.CommonModule, forms.FormsModule, overlay.OverlayModule, input.ThyInputModule, icon.ThyIconModule],
                declarations: [ThyCascaderComponent, ThyCascaderOptionComponent],
                exports: [ThyCascaderComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyCascaderModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyCascaderComponent = ThyCascaderComponent;
    exports.ThyCascaderModule = ThyCascaderModule;
    exports.ThyCascaderOptionComponent = ThyCascaderOptionComponent;
    exports.ɵ0 = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
