// ngx-tethys/bundles/ngx-tethys-back-top.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/cdk/platform'),require('@angular/common'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/icon'), require('rxjs'), require('@angular/cdk/platform'), require('rxjs/operators'), require('ngx-tethys/core')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/back-top', ['@angular/core','ngx-tethys/core','@angular/cdk/platform','@angular/common','ngx-tethys/icon','exports', '@angular/core', '@angular/common', 'ngx-tethys/icon', 'rxjs', '@angular/cdk/platform', 'rxjs/operators', 'ngx-tethys/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.cdk.platform,global.ng.common,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys']['back-top'] = {}), global.ng.core, global.ng.common, global['ngx-tethys'].icon, global.rxjs, global.ng.cdk.platform, global.rxjs.operators, global['ngx-tethys'].core));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,exports, core, common, icon, rxjs, platform, operators, core$1) { 
function ThyBackTopComponent_div_0_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelement(1, "thy-icon", 5);
    ɵngcc0.ɵɵelementEnd();
} }
function ThyBackTopComponent_div_0_ng_template_3_Template(rf, ctx) { }
function ThyBackTopComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    var _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵlistener("click", function ThyBackTopComponent_div_0_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r5); var ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.clickBackTop(); });
    ɵngcc0.ɵɵtemplate(1, ThyBackTopComponent_div_0_ng_template_1_Template, 2, 0, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(3, ThyBackTopComponent_div_0_ng_template_3_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r1 = ɵngcc0.ɵɵreference(2);
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@fadeMotion", undefined);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.thyTemplate || _r1);
} }
'use strict';

    var ThyBackTopComponent = /** @class */ (function () {
        function ThyBackTopComponent(doc, thyScrollService, platform, cdr, zone) {
            this.doc = doc;
            this.thyScrollService = thyScrollService;
            this.platform = platform;
            this.cdr = cdr;
            this.zone = zone;
            this.classNames = true;
            this.thyVisibilityHeight = 400;
            this.thyClick = new core.EventEmitter();
            this.visibleChange = new core.EventEmitter();
            this.visible = false;
            this.scrollListenerDestroy$ = new rxjs.Subject();
            this.target = null;
        }
        ThyBackTopComponent.prototype.ngOnInit = function () {
            this.registerScrollEvent();
        };
        ThyBackTopComponent.prototype.clickBackTop = function () {
            this.thyScrollService.scrollTo(this.getTarget(), 0);
            this.thyClick.emit(true);
        };
        ThyBackTopComponent.prototype.getTarget = function () {
            return this.target || window;
        };
        ThyBackTopComponent.prototype.handleScroll = function () {
            var _this = this;
            if (this.visible === this.thyScrollService.getScroll(this.getTarget()) > this.thyVisibilityHeight) {
                return;
            }
            this.visible = !this.visible;
            this.cdr.detectChanges();
            this.zone.run(function () {
                _this.visibleChange.emit(_this.visible);
            });
        };
        ThyBackTopComponent.prototype.registerScrollEvent = function () {
            var _this = this;
            if (!this.platform.isBrowser) {
                return;
            }
            this.scrollListenerDestroy$.next();
            this.handleScroll();
            this.zone.runOutsideAngular(function () {
                rxjs.fromEvent(_this.getTarget(), 'scroll')
                    .pipe(operators.throttleTime(50), operators.takeUntil(_this.scrollListenerDestroy$))
                    .subscribe(function () { return _this.handleScroll(); });
            });
        };
        ThyBackTopComponent.prototype.ngOnDestroy = function () {
            this.scrollListenerDestroy$.next();
            this.scrollListenerDestroy$.complete();
        };
        ThyBackTopComponent.prototype.ngOnChanges = function (changes) {
            var thyContainer = changes.thyContainer;
            if (thyContainer) {
                this.target = typeof this.thyContainer === 'string' ? this.doc.querySelector(this.thyContainer) : this.thyContainer;
                this.registerScrollEvent();
            }
        };
ThyBackTopComponent.ɵfac = function ThyBackTopComponent_Factory(t) { return new (t || ThyBackTopComponent)(ɵngcc0.ɵɵdirectiveInject(common.DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyScrollService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyBackTopComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyBackTopComponent, selectors: [["thy-back-top"], ["", "thyBackTop", ""]], hostVars: 2, hostBindings: function ThyBackTopComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-back-top-container", ctx.classNames);
    } }, inputs: { thyVisibilityHeight: "thyVisibilityHeight", thyTemplate: "thyTemplate", thyContainer: "thyContainer" }, outputs: { thyClick: "thyClick", visibleChange: "visibleChange" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [["class", "thy-back-top", 3, "click", 4, "ngIf"], [1, "thy-back-top", 3, "click"], ["defaultContent", ""], [3, "ngTemplateOutlet"], [1, "thy-back-top-content"], ["thyIconName", "arrow-up", 1, "back-top-icon", "text-muted", "font-size-lg"]], template: function ThyBackTopComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyBackTopComponent_div_0_Template, 4, 2, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.visible);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet, ɵngcc4.ThyIconComponent], encapsulation: 2, data: { animation: [core$1.fadeMotion] }, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBackTopComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-back-top,[thyBackTop]',
                template: "<div class=\"thy-back-top\" (click)=\"clickBackTop()\" @fadeMotion *ngIf=\"visible\">\n  <ng-template #defaultContent>\n    <div class=\"thy-back-top-content\">\n      <thy-icon class=\"back-top-icon text-muted font-size-lg\" thyIconName=\"arrow-up\"></thy-icon>\n    </div>\n  </ng-template>\n  <ng-template [ngTemplateOutlet]=\"thyTemplate || defaultContent\"></ng-template>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                encapsulation: core.ViewEncapsulation.None,
                animations: [core$1.fadeMotion]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: core.Inject,
                args: [common.DOCUMENT]
            }] }, { type: ɵngcc1.ThyScrollService }, { type: ɵngcc2.Platform }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.NgZone }]; }, { classNames: [{
            type: core.HostBinding,
            args: ['class.thy-back-top-container']
        }], thyVisibilityHeight: [{
            type: core.Input
        }], thyClick: [{
            type: core.Output
        }], visibleChange: [{
            type: core.Output
        }], thyTemplate: [{
            type: core.Input
        }], thyContainer: [{
            type: core.Input
        }] }); })();
        return ThyBackTopComponent;
    }());
    ThyBackTopComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
        { type: core$1.ThyScrollService },
        { type: platform.Platform },
        { type: core.ChangeDetectorRef },
        { type: core.NgZone }
    ]; };
    ThyBackTopComponent.propDecorators = {
        classNames: [{ type: core.HostBinding, args: ['class.thy-back-top-container',] }],
        thyTemplate: [{ type: core.Input }],
        thyVisibilityHeight: [{ type: core.Input }],
        thyContainer: [{ type: core.Input }],
        thyClick: [{ type: core.Output }],
        visibleChange: [{ type: core.Output }]
    };

    var ThyBackTopModule = /** @class */ (function () {
        function ThyBackTopModule() {
        }
ThyBackTopModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyBackTopModule });
ThyBackTopModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyBackTopModule_Factory(t) { return new (t || ThyBackTopModule)(); }, imports: [[common.CommonModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyBackTopModule, { declarations: [ThyBackTopComponent], imports: [ɵngcc3.CommonModule, ɵngcc4.ThyIconModule], exports: [ThyBackTopComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBackTopModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyBackTopComponent],
                imports: [common.CommonModule, icon.ThyIconModule],
                exports: [ThyBackTopComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyBackTopModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyBackTopComponent = ThyBackTopComponent;
    exports.ThyBackTopModule = ThyBackTopModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
