// ngx-tethys/bundles/ngx-tethys-vote.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/common'),require('ngx-tethys/icon'),require('ngx-tethys/shared'),exports, require('@angular/common'), require('@angular/core'), require('ngx-tethys/shared'), require('ngx-tethys/util'), require('ngx-tethys/core'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/vote', ['@angular/core','ngx-tethys/core','@angular/common','ngx-tethys/icon','ngx-tethys/shared','exports', '@angular/common', '@angular/core', 'ngx-tethys/shared', 'ngx-tethys/util', 'ngx-tethys/core', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.common,global.ngxTethys.icon,global.ngxTethys.shared,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].vote = {}), global.ng.common, global.ng.core, global['ngx-tethys'].shared, global['ngx-tethys'].util, global['ngx-tethys'].core, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,exports, common, core, shared, util, core$1, icon) { 
var _c0 = ["voteIcon"];
function ThyVoteComponent_thy_icon_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 2);
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.thyIcon)("thyIconType", ctx_r0._hasVoted ? "fill" : "");
} }
function ThyVoteComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 3);
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.voteIcon);
} }
'use strict';

    var ThyVoteComponent = /** @class */ (function () {
        function ThyVoteComponent(elementRef, updateHostClassService) {
            this.elementRef = elementRef;
            this.updateHostClassService = updateHostClassService;
            this._initialized = false;
            this._isRound = false;
            this.class = true;
            this._hasVoted = true;
            this.thyIcon = 'thumb-up';
            this.updateHostClassService.initializeElement(elementRef.nativeElement);
        }
        Object.defineProperty(ThyVoteComponent.prototype, "thySize", {
            set: function (value) {
                this._size = value;
                if (this._initialized) {
                    this._setClassesByType();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyVoteComponent.prototype, "thyVote", {
            set: function (value) {
                this._type = value;
                if (this._initialized) {
                    this._setClassesByType();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyVoteComponent.prototype, "thyRound", {
            set: function (value) {
                this._isRound = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyVoteComponent.prototype, "thyLayout", {
            set: function (value) {
                this._layout = value;
                if (this._initialized) {
                    this._setClassesByType();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyVoteComponent.prototype, "thyHasVoted", {
            set: function (value) {
                this._hasVoted = util.coerceBooleanProperty(value);
                if (this._initialized) {
                    this._setClassesByType();
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyVoteComponent.prototype.ngOnInit = function () {
            this._setClassesByType();
            this._initialized = true;
        };
        ThyVoteComponent.prototype._setClassesByType = function () {
            var classNames = [];
            if (!this._type) {
                this._type = 'primary';
            }
            if (!this._layout) {
                this._layout = 'horizontal';
            }
            if (!this._size) {
                this._size = 'default';
            }
            if (this._isRound) {
                classNames.push('thy-vote-round');
            }
            classNames.push("thy-vote-" + this._type);
            classNames.push("thy-vote-" + this._layout);
            classNames.push("thy-vote-" + this._layout + "-size-" + this._size);
            this.updateHostClassService.updateClass(classNames);
        };
ThyVoteComponent.ɵfac = function ThyVoteComponent_Factory(t) { return new (t || ThyVoteComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService)); };
ThyVoteComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyVoteComponent, selectors: [["thy-vote"], ["", "thyVote", ""]], contentQueries: function ThyVoteComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.voteIcon = _t.first);
    } }, hostVars: 4, hostBindings: function ThyVoteComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-vote", ctx.class)("has-voted", ctx._hasVoted);
    } }, inputs: { thyIcon: "thyIcon", thySize: "thySize", thyVote: "thyVote", thyRound: "thyRound", thyLayout: "thyLayout", thyHasVoted: "thyHasVoted", thyVoteCount: "thyVoteCount" }, features: [ɵngcc0.ɵɵProvidersFeature([core$1.UpdateHostClassService])], decls: 4, vars: 3, consts: [[3, "thyIconName", "thyIconType", 4, "ngIf"], [3, "ngTemplateOutlet", 4, "ngIf"], [3, "thyIconName", "thyIconType"], [3, "ngTemplateOutlet"]], template: function ThyVoteComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyVoteComponent_thy_icon_0_Template, 1, 2, "thy-icon", 0);
        ɵngcc0.ɵɵtemplate(1, ThyVoteComponent_ng_container_1_Template, 1, 1, "ng-container", 1);
        ɵngcc0.ɵɵelementStart(2, "div");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.voteIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.voteIcon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyVoteCount);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.ThyIconComponent, ɵngcc2.NgTemplateOutlet], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyVoteComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-vote,[thyVote]',
                template: "<thy-icon *ngIf=\"!voteIcon\" [thyIconName]=\"thyIcon\" [thyIconType]=\"_hasVoted ? 'fill' : ''\"></thy-icon>\n<ng-container *ngIf=\"voteIcon\" [ngTemplateOutlet]=\"voteIcon\"></ng-container>\n<div>{{ thyVoteCount }}</div>\n",
                providers: [core$1.UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.UpdateHostClassService }]; }, { class: [{
            type: core.HostBinding,
            args: ["class.thy-vote"]
        }], _hasVoted: [{
            type: core.HostBinding,
            args: ["class.has-voted"]
        }], thyIcon: [{
            type: core.Input
        }], thySize: [{
            type: core.Input
        }], thyVote: [{
            type: core.Input
        }], thyRound: [{
            type: core.Input
        }], thyLayout: [{
            type: core.Input
        }], thyHasVoted: [{
            type: core.Input
        }], thyVoteCount: [{
            type: core.Input
        }], voteIcon: [{
            type: core.ContentChild,
            args: ['voteIcon']
        }] }); })();
        return ThyVoteComponent;
    }());
    ThyVoteComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core$1.UpdateHostClassService }
    ]; };
    ThyVoteComponent.propDecorators = {
        class: [{ type: core.HostBinding, args: ["class.thy-vote",] }],
        _hasVoted: [{ type: core.HostBinding, args: ["class.has-voted",] }],
        thySize: [{ type: core.Input }],
        thyVote: [{ type: core.Input }],
        thyRound: [{ type: core.Input }],
        thyLayout: [{ type: core.Input }],
        thyVoteCount: [{ type: core.Input }],
        thyIcon: [{ type: core.Input }],
        thyHasVoted: [{ type: core.Input }],
        voteIcon: [{ type: core.ContentChild, args: ['voteIcon',] }]
    };

    var ThyVoteModule = /** @class */ (function () {
        function ThyVoteModule() {
        }
ThyVoteModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyVoteModule });
ThyVoteModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyVoteModule_Factory(t) { return new (t || ThyVoteModule)(); }, providers: [], imports: [[common.CommonModule, shared.ThySharedModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyVoteModule, { declarations: [ThyVoteComponent], imports: [ɵngcc2.CommonModule, ɵngcc4.ThySharedModule, ɵngcc3.ThyIconModule], exports: [ThyVoteComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyVoteModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyVoteComponent],
                imports: [common.CommonModule, shared.ThySharedModule, icon.ThyIconModule],
                exports: [ThyVoteComponent],
                providers: []
            }]
    }], function () { return []; }, null); })();
        return ThyVoteModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyVoteComponent = ThyVoteComponent;
    exports.ThyVoteModule = ThyVoteModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
