// ngx-tethys/bundles/ngx-tethys-transfer.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/icon'),require('@angular/common'),require('ngx-tethys/flexible-text'),require('@angular/cdk/drag-drop'),require('ngx-tethys/list'),require('ngx-tethys/shared'),require('ngx-tethys/button'),exports, require('@angular/core'), require('@angular/common'), require('@angular/cdk/drag-drop'), require('ngx-tethys/button'), require('ngx-tethys/icon'), require('ngx-tethys/list'), require('ngx-tethys/shared'), require('ngx-tethys/flexible-text')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/transfer', ['@angular/core','ngx-tethys/icon','@angular/common','ngx-tethys/flexible-text','@angular/cdk/drag-drop','ngx-tethys/list','ngx-tethys/shared','ngx-tethys/button','exports', '@angular/core', '@angular/common', '@angular/cdk/drag-drop', 'ngx-tethys/button', 'ngx-tethys/icon', 'ngx-tethys/list', 'ngx-tethys/shared', 'ngx-tethys/flexible-text'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.icon,global.ng.common,global.ngxTethys.flexibleText,global.ng.cdk.dragDrop,global.ngxTethys.list,global.ngxTethys.shared,global.ngxTethys.button,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].transfer = {}), global.ng.core, global.ng.common, global.ng.cdk.dragDrop, global['ngx-tethys'].button, global['ngx-tethys'].icon, global['ngx-tethys'].list, global['ngx-tethys'].shared, global['ngx-tethys']['flexible-text']));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,exports, core, common, dragDrop, button, icon, list, shared, flexibleText) { 
var _c0 = ["renderTemplate"];
var _c1 = ["renderLeftTemplate"];
var _c2 = ["renderRightTemplate"];
function ThyTransferComponent_ng_template_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 14);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyTransferComponent_ng_template_1_ng_template_3_Template(rf, ctx) { }
function ThyTransferComponent_ng_template_1_thy_icon_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 15);
} }
var _c3 = function (a0) { return { "un-select-item": a0 }; };
var _c4 = function (a0) { return { $implicit: a0 }; };
function ThyTransferComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    var _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵlistener("click", function ThyTransferComponent_ng_template_1_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r11); var item_r6 = ctx.$implicit; var ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.onSelect(item_r6.direction === "left" ? "left" : "right", { item: item_r6 }); });
    ɵngcc0.ɵɵtemplate(1, ThyTransferComponent_ng_template_1_ng_container_1_Template, 2, 0, "ng-container", 10);
    ɵngcc0.ɵɵelementStart(2, "div", 11);
    ɵngcc0.ɵɵtemplate(3, ThyTransferComponent_ng_template_1_ng_template_3_Template, 0, 0, "ng-template", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, ThyTransferComponent_ng_template_1_thy_icon_4_Template, 1, 0, "thy-icon", 13);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r6 = ctx.$implicit;
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    var _r4 = ɵngcc0.ɵɵreference(12);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.thyLeftDraggable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c3, item_r6.direction !== "right"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r4)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(7, _c4, item_r6));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r6.direction === "right");
} }
function ThyTransferComponent_ng_template_9_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 14);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyTransferComponent_ng_template_9_ng_template_2_Template(rf, ctx) { }
function ThyTransferComponent_ng_template_9_a_3_Template(rf, ctx) { if (rf & 1) {
    var _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 18);
    ɵngcc0.ɵɵlistener("click", function ThyTransferComponent_ng_template_9_a_3_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r18); var item_r12 = ɵngcc0.ɵɵnextContext().$implicit; var ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.onSelect("right", { item: item_r12 }); });
    ɵngcc0.ɵɵelement(1, "thy-icon", 19);
    ɵngcc0.ɵɵelementEnd();
} }
function ThyTransferComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTransferComponent_ng_template_9_ng_container_0_Template, 2, 0, "ng-container", 10);
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵtemplate(2, ThyTransferComponent_ng_template_9_ng_template_2_Template, 0, 0, "ng-template", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, ThyTransferComponent_ng_template_9_a_3_Template, 2, 0, "a", 17);
} if (rf & 2) {
    var item_r12 = ctx.$implicit;
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    var _r4 = ɵngcc0.ɵɵreference(12);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.thyRightDraggable);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r4)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(4, _c4, item_r12));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r12.direction === "right" && !item_r12.isFixed);
} }
function ThyTransferComponent_ng_template_11_0_ng_template_0_Template(rf, ctx) { }
function ThyTransferComponent_ng_template_11_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTransferComponent_ng_template_11_0_ng_template_0_Template, 0, 0, "ng-template", 12);
} if (rf & 2) {
    var item_r19 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r20 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r20.templateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c4, item_r19));
} }
function ThyTransferComponent_ng_template_11_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "thy-flexible-text", 20);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r19 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyTooltipContent", item_r19.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r19.title, " ");
} }
function ThyTransferComponent_ng_template_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTransferComponent_ng_template_11_0_Template, 1, 4, undefined, 10);
    ɵngcc0.ɵɵtemplate(1, ThyTransferComponent_ng_template_11_ng_container_1_Template, 3, 2, "ng-container", 10);
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.templateRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r5.templateRef);
} }
function ThyTransferListComponent_ng_container_3_ng_container_2_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 11);
} }
function ThyTransferListComponent_ng_container_3_ng_container_2_thy_list_item_5_ng_template_1_Template(rf, ctx) { }
var _c5 = function (a0) { return { active: a0 }; };
function ThyTransferListComponent_ng_container_3_ng_container_2_thy_list_item_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "thy-list-item", 12);
    ɵngcc0.ɵɵtemplate(1, ThyTransferListComponent_ng_container_3_ng_container_2_thy_list_item_5_ng_template_1_Template, 0, 0, "ng-template", 13);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r8 = ctx.$implicit;
    var ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("cdkDragData", item_r8)("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c5, item_r8.checked));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r7.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(6, _c4, item_r8));
} }
function ThyTransferListComponent_ng_container_3_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    var _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 9);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "thy-list", 10);
    ɵngcc0.ɵɵlistener("cdkDropListDropped", function ThyTransferListComponent_ng_container_3_ng_container_2_Template_thy_list_cdkDropListDropped_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); var ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.drop($event); });
    ɵngcc0.ɵɵtemplate(4, ThyTransferListComponent_ng_container_3_ng_container_2_div_4_Template, 1, 0, "div", 7);
    ɵngcc0.ɵɵtemplate(5, ThyTransferListComponent_ng_container_3_ng_container_2_thy_list_item_5_Template, 2, 8, "thy-list-item", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 9);
    ɵngcc0.ɵɵtext(7, "\u672A\u9501\u5B9A");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1("\u9501\u5B9A (\u4E0A\u9650", ctx_r3.maxLock, "\u4E2A)");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkDropListData", ctx_r3.lockItems)("cdkDropListEnterPredicate", ctx_r3.lockListEnterPredicate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.lockItems.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.lockItems);
} }
function ThyTransferListComponent_ng_container_3_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 11);
} }
function ThyTransferListComponent_ng_container_3_thy_list_item_5_ng_template_1_Template(rf, ctx) { }
function ThyTransferListComponent_ng_container_3_thy_list_item_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "thy-list-item", 12);
    ɵngcc0.ɵɵtemplate(1, ThyTransferListComponent_ng_container_3_thy_list_item_5_ng_template_1_Template, 0, 0, "ng-template", 13);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r12 = ctx.$implicit;
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("cdkDragData", item_r12)("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c5, item_r12.checked));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r5.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(6, _c4, item_r12));
} }
function ThyTransferListComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    var _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 4);
    ɵngcc0.ɵɵtemplate(2, ThyTransferListComponent_ng_container_3_ng_container_2_Template, 8, 5, "ng-container", 5);
    ɵngcc0.ɵɵelementStart(3, "thy-list", 6);
    ɵngcc0.ɵɵlistener("cdkDropListDropped", function ThyTransferListComponent_ng_container_3_Template_thy_list_cdkDropListDropped_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r15); var ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.drop($event); });
    ɵngcc0.ɵɵtemplate(4, ThyTransferListComponent_ng_container_3_div_4_Template, 1, 0, "div", 7);
    ɵngcc0.ɵɵtemplate(5, ThyTransferListComponent_ng_container_3_thy_list_item_5_Template, 2, 8, "thy-list-item", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkDropListGroupDisabled", !ctx_r0.draggable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.canLock);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkDropListDisabled", !ctx_r0.draggable)("cdkDropListData", ctx_r0.unlockItems);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.unlockItems.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.unlockItems);
} }
function ThyTransferListComponent_ng_template_4_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var _c6 = function (a0, a1, a2) { return { $implicit: a0, onSelectItem: a1, onUnselectItem: a2 }; };
function ThyTransferListComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵtemplate(1, ThyTransferListComponent_ng_template_4_ng_container_1_Template, 1, 0, "ng-container", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r2.contentRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(2, _c6, ctx_r2.items, ctx_r2.onSelectItem, ctx_r2.onUnselectItem));
} }
'use strict';

    exports.TransferDirection = void 0;
    (function (TransferDirection) {
        TransferDirection["left"] = "left";
        TransferDirection["right"] = "right";
    })(exports.TransferDirection || (exports.TransferDirection = {}));

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ThyTransferComponent = /** @class */ (function () {
        function ThyTransferComponent() {
            this.hostClass = 'thy-transfer';
            this.leftDataSource = [];
            this.rightDataSource = [];
            this.allDataSource = [];
            this.rightDraggable = false;
            this._autoMove = true;
            this.thyDraggableUpdate = new core.EventEmitter();
            this.thyChange = new core.EventEmitter();
        }
        Object.defineProperty(ThyTransferComponent.prototype, "thyData", {
            set: function (value) {
                if (value) {
                    this.initializeTransferData(value);
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTransferComponent.prototype, "thyTitles", {
            // @Input() transferToRight
            set: function (value) {
                this.leftTitle = value[0] || '';
                this.rightTitle = value[1] || '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTransferComponent.prototype, "thyAutoMove", {
            // Currently not implemented, in order to support the selections move
            set: function (value) {
                this._autoMove = value;
            },
            enumerable: false,
            configurable: true
        });
        ThyTransferComponent.prototype.ngOnInit = function () { };
        ThyTransferComponent.prototype.initializeTransferData = function (data) {
            var _this = this;
            if (data === void 0) { data = []; }
            this.allDataSource = [];
            this.leftDataSource = [];
            this.rightDataSource = [];
            data.forEach(function (item) {
                _this.allDataSource.push(item);
                if (item.direction === exports.TransferDirection.left) {
                    _this.leftDataSource.push(item);
                }
                if (item.direction === exports.TransferDirection.right) {
                    _this.rightDataSource.push(item);
                }
            });
        };
        ThyTransferComponent.prototype.onSelect = function (from, event) {
            if (event.item.isFixed) {
                return;
            }
            var to = from === exports.TransferDirection.left ? exports.TransferDirection.right : exports.TransferDirection.left;
            event.item.checked = !event.item.checked;
            if (this._autoMove) {
                this.onMove(to);
            }
        };
        ThyTransferComponent.prototype.selectItem = function (event) {
            this.onSelect(exports.TransferDirection.left, event);
        };
        ThyTransferComponent.prototype.unselectItem = function (event) {
            this.onSelect(exports.TransferDirection.right, event);
        };
        ThyTransferComponent.prototype.groupListByIsLock = function (list) {
            if (list === void 0) { list = []; }
            var lock = [], unlock = [];
            list.forEach(function (item) {
                if (item.isLock) {
                    lock.push(item);
                }
                else {
                    unlock.push(item);
                }
            });
            return { lock: lock, unlock: unlock };
        };
        ThyTransferComponent.prototype.onMove = function (to) {
            var fromDataSource = to === exports.TransferDirection.right ? this.leftDataSource : this.rightDataSource;
            var toDataSource = to === exports.TransferDirection.right ? this.rightDataSource : this.leftDataSource;
            var selections = fromDataSource.filter(function (item) { return item.checked; });
            var changeEvent = {
                from: to === exports.TransferDirection.right ? exports.TransferDirection.left : exports.TransferDirection.right,
                to: to,
                items: __spread(selections)
            };
            selections.forEach(function (item) {
                var index = fromDataSource.indexOf(item);
                var removed = fromDataSource.splice(index, 1)[0];
                removed.checked = !removed.checked;
                removed.direction = to;
                toDataSource.push(removed);
            });
            this.thyChange.emit(Object.assign(Object.assign({}, changeEvent), { left: this.groupListByIsLock(this.leftDataSource), right: this.groupListByIsLock(this.rightDataSource) }));
        };
        ThyTransferComponent.prototype.onDragUpdate = function (direction, event) {
            var otherDirectionData = direction === exports.TransferDirection.left ? this.rightDataSource : this.leftDataSource;
            var otherListData = this.groupListByIsLock(otherDirectionData);
            this.thyDraggableUpdate.emit(Object.assign(Object.assign({}, event.dragEvent), { left: direction === exports.TransferDirection.left ? event.listData : otherListData, right: direction === exports.TransferDirection.right ? event.listData : otherListData }));
        };
ThyTransferComponent.ɵfac = function ThyTransferComponent_Factory(t) { return new (t || ThyTransferComponent)(); };
ThyTransferComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTransferComponent, selectors: [["thy-transfer"]], contentQueries: function ThyTransferComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.leftContentRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.rightContentRef = _t.first);
    } }, hostVars: 2, hostBindings: function ThyTransferComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.hostClass);
    } }, inputs: { thyData: "thyData", thyTitles: "thyTitles", thyAutoMove: "thyAutoMove", thyrenderLeftTemplateRef: "thyrenderLeftTemplateRef", thyrenderRightTemplateRef: "thyrenderRightTemplateRef", thyRightCanLock: "thyRightCanLock", thyRightLockMax: "thyRightLockMax", thyLeftDraggable: "thyLeftDraggable", thyRightDraggable: "thyRightDraggable" }, outputs: { thyDraggableUpdate: "thyDraggableUpdate", thyChange: "thyChange" }, decls: 13, vars: 12, consts: [[3, "items", "title", "draggable", "renderContentRef", "template", "draggableUpdate", "selectItem", "unselectItem"], ["leftTemplateRef", ""], [1, "thy-transfer-operation"], [1, "thy-transfer-operation-body"], ["thyIconName", "angle-right", 1, "operation-link"], ["thyIconName", "angle-left", 1, "operation-link"], [3, "items", "title", "draggable", "canLock", "maxLock", "template", "renderContentRef", "draggableUpdate", "selectItem", "unselectItem"], ["rightTemplateRef", ""], ["thyTransferListItem", ""], [1, "list-item-content", 3, "click"], [4, "ngIf"], [1, "item-name", 3, "ngClass"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["class", "text-primary item-status", "thyIconName", "check", 4, "ngIf"], ["thyIconName", "drag", 1, "drag-handle"], ["thyIconName", "check", 1, "text-primary", "item-status"], [1, "item-name"], ["class", "link-secondary", 3, "click", 4, "ngIf"], [1, "link-secondary", 3, "click"], ["thyIconName", "close", 1, "item-status"], [3, "thyTooltipContent"]], template: function ThyTransferComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "thy-transfer-list", 0);
        ɵngcc0.ɵɵlistener("draggableUpdate", function ThyTransferComponent_Template_thy_transfer_list_draggableUpdate_0_listener($event) { return ctx.onDragUpdate("left", $event); })("selectItem", function ThyTransferComponent_Template_thy_transfer_list_selectItem_0_listener($event) { return ctx.selectItem($event); })("unselectItem", function ThyTransferComponent_Template_thy_transfer_list_unselectItem_0_listener($event) { return ctx.unselectItem($event); });
        ɵngcc0.ɵɵtemplate(1, ThyTransferComponent_ng_template_1_Template, 5, 9, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelement(5, "thy-icon", 4);
        ɵngcc0.ɵɵelement(6, "br");
        ɵngcc0.ɵɵelement(7, "thy-icon", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "thy-transfer-list", 6);
        ɵngcc0.ɵɵlistener("draggableUpdate", function ThyTransferComponent_Template_thy_transfer_list_draggableUpdate_8_listener($event) { return ctx.onDragUpdate("right", $event); })("selectItem", function ThyTransferComponent_Template_thy_transfer_list_selectItem_8_listener($event) { return ctx.selectItem($event); })("unselectItem", function ThyTransferComponent_Template_thy_transfer_list_unselectItem_8_listener($event) { return ctx.unselectItem($event); });
        ɵngcc0.ɵɵtemplate(9, ThyTransferComponent_ng_template_9_Template, 4, 6, "ng-template", null, 7, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, ThyTransferComponent_ng_template_11_Template, 2, 2, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        var _r0 = ɵngcc0.ɵɵreference(2);
        var _r2 = ɵngcc0.ɵɵreference(10);
        ɵngcc0.ɵɵproperty("items", ctx.allDataSource)("title", ctx.leftTitle)("draggable", ctx.thyLeftDraggable)("renderContentRef", ctx.leftContentRef)("template", _r0);
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵproperty("items", ctx.rightDataSource)("title", ctx.rightTitle)("draggable", ctx.thyRightDraggable)("canLock", ctx.thyRightCanLock)("maxLock", ctx.thyRightLockMax)("template", _r2)("renderContentRef", ctx.rightContentRef);
    } }, directives: function () { return [ThyTransferListComponent, ɵngcc1.ThyIconComponent, ɵngcc2.NgIf, ɵngcc2.NgClass, ɵngcc2.NgTemplateOutlet, ɵngcc3.ThyFlexibleTextComponent]; }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTransferComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-transfer',
                template: "<thy-transfer-list\n  [items]=\"allDataSource\"\n  [title]=\"leftTitle\"\n  [draggable]=\"thyLeftDraggable\"\n  [renderContentRef]=\"leftContentRef\"\n  (draggableUpdate)=\"onDragUpdate('left', $event)\"\n  [template]=\"leftTemplateRef\"\n  (selectItem)=\"selectItem($event)\"\n  (unselectItem)=\"unselectItem($event)\"\n>\n  <ng-template #leftTemplateRef let-item>\n    <div (click)=\"onSelect(item.direction === 'left' ? 'left' : 'right', { item: item })\" class=\"list-item-content\">\n      <ng-container *ngIf=\"thyLeftDraggable\">\n        <thy-icon thyIconName=\"drag\" class=\"drag-handle\"></thy-icon>\n      </ng-container>\n      <div class=\"item-name\" [ngClass]=\"{ 'un-select-item': item.direction !== 'right' }\">\n        <ng-template [ngTemplateOutlet]=\"thyTransferListItem\" [ngTemplateOutletContext]=\"{ $implicit: item }\"></ng-template>\n      </div>\n      <thy-icon *ngIf=\"item.direction === 'right'\" class=\"text-primary item-status\" thyIconName=\"check\"></thy-icon>\n    </div>\n  </ng-template>\n</thy-transfer-list>\n\n<div class=\"thy-transfer-operation\">\n  <div class=\"thy-transfer-operation-body\">\n    <thy-icon thyIconName=\"angle-right\" class=\"operation-link\"></thy-icon>\n    <br />\n    <thy-icon thyIconName=\"angle-left\" class=\"operation-link\"></thy-icon>\n  </div>\n</div>\n\n<thy-transfer-list\n  [items]=\"rightDataSource\"\n  [title]=\"rightTitle\"\n  [draggable]=\"thyRightDraggable\"\n  (draggableUpdate)=\"onDragUpdate('right', $event)\"\n  [canLock]=\"thyRightCanLock\"\n  [maxLock]=\"thyRightLockMax\"\n  [template]=\"rightTemplateRef\"\n  [renderContentRef]=\"rightContentRef\"\n  (selectItem)=\"selectItem($event)\"\n  (unselectItem)=\"unselectItem($event)\"\n>\n  <ng-template #rightTemplateRef let-item>\n    <ng-container *ngIf=\"thyRightDraggable\">\n      <thy-icon thyIconName=\"drag\" class=\"drag-handle\"></thy-icon>\n    </ng-container>\n    <div class=\"item-name\">\n      <ng-template [ngTemplateOutlet]=\"thyTransferListItem\" [ngTemplateOutletContext]=\"{ $implicit: item }\"></ng-template>\n    </div>\n    <a class=\"link-secondary\" *ngIf=\"item.direction === 'right' && !item.isFixed\" (click)=\"onSelect('right', { item: item })\">\n      <thy-icon class=\"item-status\" thyIconName=\"close\"></thy-icon>\n    </a>\n  </ng-template>\n</thy-transfer-list>\n\n<ng-template #thyTransferListItem let-item>\n  <ng-template *ngIf=\"templateRef\" [ngTemplateOutlet]=\"templateRef\" [ngTemplateOutletContext]=\"{ $implicit: item }\"></ng-template>\n  <ng-container *ngIf=\"!templateRef\">\n    <thy-flexible-text [thyTooltipContent]=\"item.title\">\n      {{ item.title }}\n    </thy-flexible-text>\n  </ng-container>\n</ng-template>\n",
                encapsulation: core.ViewEncapsulation.None
            }]
    }], function () { return []; }, { hostClass: [{
            type: core.HostBinding,
            args: ['class']
        }], thyDraggableUpdate: [{
            type: core.Output
        }], thyChange: [{
            type: core.Output
        }], thyData: [{
            type: core.Input
        }], thyTitles: [{
            type: core.Input
        }], thyAutoMove: [{
            type: core.Input
        }], thyrenderLeftTemplateRef: [{
            type: core.Input
        }], thyrenderRightTemplateRef: [{
            type: core.Input
        }], thyRightCanLock: [{
            type: core.Input
        }], thyRightLockMax: [{
            type: core.Input
        }], thyLeftDraggable: [{
            type: core.Input
        }], thyRightDraggable: [{
            type: core.Input
        }], templateRef: [{
            type: core.ContentChild,
            args: ['renderTemplate']
        }], leftContentRef: [{
            type: core.ContentChild,
            args: ['renderLeftTemplate']
        }], rightContentRef: [{
            type: core.ContentChild,
            args: ['renderRightTemplate']
        }] }); })();
        return ThyTransferComponent;
    }());
    ThyTransferComponent.propDecorators = {
        hostClass: [{ type: core.HostBinding, args: ['class',] }],
        thyData: [{ type: core.Input }],
        thyrenderLeftTemplateRef: [{ type: core.Input }],
        thyrenderRightTemplateRef: [{ type: core.Input }],
        thyTitles: [{ type: core.Input }],
        thyRightCanLock: [{ type: core.Input }],
        thyRightLockMax: [{ type: core.Input }],
        thyAutoMove: [{ type: core.Input }],
        thyLeftDraggable: [{ type: core.Input }],
        thyRightDraggable: [{ type: core.Input }],
        thyDraggableUpdate: [{ type: core.Output }],
        thyChange: [{ type: core.Output }],
        templateRef: [{ type: core.ContentChild, args: ['renderTemplate',] }],
        leftContentRef: [{ type: core.ContentChild, args: ['renderLeftTemplate',] }],
        rightContentRef: [{ type: core.ContentChild, args: ['renderRightTemplate',] }]
    };

    var ThyTransferListComponent = /** @class */ (function () {
        function ThyTransferListComponent(root, differs) {
            var _this = this;
            this.root = root;
            this.differs = differs;
            this.lockItems = [];
            this.unlockItems = [];
            this.draggableUpdate = new core.EventEmitter();
            this.selectItem = new core.EventEmitter();
            this.unselectItem = new core.EventEmitter();
            this.hostClass = 'thy-transfer-list';
            this.lockListEnterPredicate = function () {
                return _this.lockItems.length < _this.maxLock;
            };
            this.onSelectItem = function (item) {
                _this.selectItem.emit({ item: item });
            };
            this.onUnselectItem = function (item) {
                _this.unselectItem.emit({ item: item });
            };
        }
        ThyTransferListComponent.prototype.ngOnInit = function () {
            this._combineTransferData();
            if (this.canLock) {
                this._lockDiff = this.differs.find(this.lockItems).create();
                this._unlockDiff = this.differs.find(this.unlockItems).create();
            }
            else {
                this._unlockDiff = this.differs.find(this.unlockItems).create();
            }
            this._diff = this.differs.find(this.items).create();
        };
        ThyTransferListComponent.prototype._combineTransferData = function () {
            var _this = this;
            this.lockItems = [];
            this.unlockItems = [];
            if (this.canLock) {
                (this.items || []).forEach(function (item) {
                    if (item.isLock) {
                        _this.lockItems.push(item);
                    }
                    else {
                        _this.unlockItems.push(item);
                    }
                });
            }
            else {
                this.unlockItems = this.items;
            }
        };
        ThyTransferListComponent.prototype._afterChangeItems = function (changes, items) {
            // 数据发生变化时，更改order值
            changes.forEachAddedItem(function (record) {
                record.item.order = record.currentIndex;
            });
            changes.forEachRemovedItem(function () {
                items.forEach(function (item, index) {
                    item.order = index;
                });
            });
            changes.forEachMovedItem(function () {
                items.forEach(function (item, index) {
                    item.order = index;
                });
            });
        };
        ThyTransferListComponent.prototype.ngDoCheck = function () {
            var changes = this._diff.diff(this.items);
            if (changes) {
                this._afterChangeItems(changes, this.items);
                this._combineTransferData();
            }
            if (this._lockDiff) {
                var lockChanges = this._lockDiff.diff(this.lockItems);
                if (lockChanges) {
                    this._afterChangeItems(lockChanges, this.lockItems);
                }
            }
            var unlockChanges = this._unlockDiff.diff(this.unlockItems);
            if (unlockChanges) {
                this._afterChangeItems(unlockChanges, this.unlockItems);
            }
        };
        ThyTransferListComponent.prototype.drop = function (event) {
            if (event.previousContainer === event.container) {
                dragDrop.moveItemInArray(event.container.data, event.previousIndex, event.currentIndex);
            }
            else {
                dragDrop.transferArrayItem(event.previousContainer.data, event.container.data, event.previousIndex, event.currentIndex);
                (event.previousContainer.data || []).forEach(function (item) {
                    item.isLock = event.previousContainer.id === 'lock';
                });
                (event.container.data || []).forEach(function (item) {
                    item.isLock = event.container.id === 'lock';
                });
            }
            var dragEvent = {
                model: event.item.data,
                models: event.container.data,
                oldIndex: event.previousIndex,
                newIndex: event.currentIndex
            };
            this.draggableUpdate.emit({
                dragEvent: dragEvent,
                listData: { lock: this.lockItems, unlock: this.unlockItems }
            });
        };
ThyTransferListComponent.ɵfac = function ThyTransferListComponent_Factory(t) { return new (t || ThyTransferListComponent)(ɵngcc0.ɵɵdirectiveInject(ThyTransferComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.IterableDiffers)); };
ThyTransferListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTransferListComponent, selectors: [["thy-transfer-list"]], hostVars: 2, hostBindings: function ThyTransferListComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.hostClass);
    } }, inputs: { title: "title", items: "items", draggable: "draggable", canLock: "canLock", maxLock: "maxLock", template: "template", contentRef: ["renderContentRef", "contentRef"] }, outputs: { draggableUpdate: "draggableUpdate", selectItem: "selectItem", unselectItem: "unselectItem" }, decls: 6, vars: 4, consts: [[1, "thy-transfer-list-header"], [1, "thy-transfer-list-header-title"], [4, "ngIf", "ngIfElse"], ["renderContent", ""], ["cdkDropListGroup", "", 1, "thy-transfer-list-body", 3, "cdkDropListGroupDisabled"], [4, "ngIf"], ["cdkDropList", "", "thyDragDrop", "", "id", "unlock", 1, "thy-transfer-list-content", 3, "cdkDropListDisabled", "cdkDropListData", "cdkDropListDropped"], ["class", "cdk-drop-list-empty", 4, "ngIf"], ["class", "thy-transfer-list-content-item text-truncate", "cdkDrag", "", 3, "cdkDragData", "ngClass", 4, "ngFor", "ngForOf"], [1, "thy-transfer-list-group-name"], ["cdkDropList", "", "thyDragDrop", "", "id", "lock", 1, "thy-transfer-list-content", 3, "cdkDropListData", "cdkDropListEnterPredicate", "cdkDropListDropped"], [1, "cdk-drop-list-empty"], ["cdkDrag", "", 1, "thy-transfer-list-content-item", "text-truncate", 3, "cdkDragData", "ngClass"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "thy-transfer-list-body"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"]], template: function ThyTransferListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "span", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ThyTransferListComponent_ng_container_3_Template, 6, 6, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(4, ThyTransferListComponent_ng_template_4_Template, 2, 6, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        var _r1 = ɵngcc0.ɵɵreference(5);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate2("", ctx.title, " \u00B7 ", ctx.items == null ? null : ctx.items.length, "");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.contentRef)("ngIfElse", _r1);
    } }, directives: [ɵngcc2.NgIf, ɵngcc4.CdkDropListGroup, ɵngcc5.ThyListComponent, ɵngcc4.CdkDropList, ɵngcc6.ThyDragDropDirective, ɵngcc2.NgForOf, ɵngcc5.ThyListItemComponent, ɵngcc4.CdkDrag, ɵngcc2.NgClass, ɵngcc2.NgTemplateOutlet], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTransferListComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-transfer-list',
                template: "<div class=\"thy-transfer-list-header\">\n  <span class=\"thy-transfer-list-header-title\">{{ title }} \u00B7 {{ items?.length }}</span>\n</div>\n<ng-container *ngIf=\"!contentRef; else renderContent\">\n  <div class=\"thy-transfer-list-body\" cdkDropListGroup [cdkDropListGroupDisabled]=\"!draggable\">\n    <ng-container *ngIf=\"canLock\">\n      <div class=\"thy-transfer-list-group-name\">\u9501\u5B9A (\u4E0A\u9650{{ maxLock }}\u4E2A)</div>\n      <thy-list\n        class=\"thy-transfer-list-content\"\n        cdkDropList\n        thyDragDrop\n        id=\"lock\"\n        [cdkDropListData]=\"lockItems\"\n        (cdkDropListDropped)=\"drop($event)\"\n        [cdkDropListEnterPredicate]=\"lockListEnterPredicate\"\n      >\n        <div class=\"cdk-drop-list-empty\" *ngIf=\"lockItems.length === 0\"></div>\n        <thy-list-item\n          class=\"thy-transfer-list-content-item text-truncate\"\n          cdkDrag\n          [cdkDragData]=\"item\"\n          *ngFor=\"let item of lockItems\"\n          [ngClass]=\"{ active: item.checked }\"\n        >\n          <ng-template [ngTemplateOutlet]=\"template\" [ngTemplateOutletContext]=\"{ $implicit: item }\"></ng-template>\n        </thy-list-item>\n      </thy-list>\n\n      <div class=\"thy-transfer-list-group-name\">\u672A\u9501\u5B9A</div>\n    </ng-container>\n    <thy-list\n      class=\"thy-transfer-list-content\"\n      cdkDropList\n      thyDragDrop\n      [cdkDropListDisabled]=\"!draggable\"\n      id=\"unlock\"\n      [cdkDropListData]=\"unlockItems\"\n      (cdkDropListDropped)=\"drop($event)\"\n    >\n      <div class=\"cdk-drop-list-empty\" *ngIf=\"unlockItems.length === 0\"></div>\n      <thy-list-item\n        class=\"thy-transfer-list-content-item text-truncate\"\n        cdkDrag\n        [cdkDragData]=\"item\"\n        *ngFor=\"let item of unlockItems\"\n        [ngClass]=\"{ active: item.checked }\"\n      >\n        <ng-template [ngTemplateOutlet]=\"template\" [ngTemplateOutletContext]=\"{ $implicit: item }\"></ng-template>\n      </thy-list-item>\n    </thy-list>\n  </div>\n</ng-container>\n\n<ng-template #renderContent>\n  <div class=\"thy-transfer-list-body\">\n    <ng-container\n      *ngTemplateOutlet=\"\n        contentRef;\n        context: {\n          $implicit: items,\n          onSelectItem: onSelectItem,\n          onUnselectItem: onUnselectItem\n        }\n      \"\n    ></ng-container>\n  </div>\n</ng-template>\n",
                encapsulation: core.ViewEncapsulation.None
            }]
    }], function () { return [{ type: ThyTransferComponent }, { type: ɵngcc0.IterableDiffers }]; }, { draggableUpdate: [{
            type: core.Output
        }], selectItem: [{
            type: core.Output
        }], unselectItem: [{
            type: core.Output
        }], hostClass: [{
            type: core.HostBinding,
            args: ['class']
        }], title: [{
            type: core.Input
        }], items: [{
            type: core.Input
        }], draggable: [{
            type: core.Input
        }], canLock: [{
            type: core.Input
        }], maxLock: [{
            type: core.Input
        }], template: [{
            type: core.Input
        }], contentRef: [{
            type: core.Input,
            args: ['renderContentRef']
        }] }); })();
        return ThyTransferListComponent;
    }());
    ThyTransferListComponent.ctorParameters = function () { return [
        { type: ThyTransferComponent },
        { type: core.IterableDiffers }
    ]; };
    ThyTransferListComponent.propDecorators = {
        title: [{ type: core.Input }],
        items: [{ type: core.Input }],
        draggable: [{ type: core.Input }],
        canLock: [{ type: core.Input }],
        maxLock: [{ type: core.Input }],
        template: [{ type: core.Input }],
        contentRef: [{ type: core.Input, args: ['renderContentRef',] }],
        draggableUpdate: [{ type: core.Output }],
        selectItem: [{ type: core.Output }],
        unselectItem: [{ type: core.Output }],
        hostClass: [{ type: core.HostBinding, args: ['class',] }]
    };

    var ThyTransferModule = /** @class */ (function () {
        function ThyTransferModule() {
        }
ThyTransferModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTransferModule });
ThyTransferModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTransferModule_Factory(t) { return new (t || ThyTransferModule)(); }, imports: [[common.CommonModule, button.ThyButtonModule, dragDrop.DragDropModule, icon.ThyIconModule, list.ThyListModule, shared.ThySharedModule, flexibleText.ThyFlexibleTextModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTransferModule, { declarations: [ThyTransferComponent, ThyTransferListComponent], imports: [ɵngcc2.CommonModule, ɵngcc7.ThyButtonModule, ɵngcc4.DragDropModule, ɵngcc1.ThyIconModule, ɵngcc5.ThyListModule, ɵngcc6.ThySharedModule, ɵngcc3.ThyFlexibleTextModule], exports: [ThyTransferComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTransferModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyTransferComponent, ThyTransferListComponent],
                imports: [common.CommonModule, button.ThyButtonModule, dragDrop.DragDropModule, icon.ThyIconModule, list.ThyListModule, shared.ThySharedModule, flexibleText.ThyFlexibleTextModule],
                exports: [ThyTransferComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyTransferModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyTransferComponent = ThyTransferComponent;
    exports.ThyTransferListComponent = ThyTransferListComponent;
    exports.ThyTransferModule = ThyTransferModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
