// ngx-tethys/bundles/ngx-tethys-stepper.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/stepper', ['@angular/core','@angular/common','exports', '@angular/core', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].stepper = {}), global.ng.core, global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, common) { 
function ThyStepComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
var _c0 = ["*"];
function ThyStepperComponent_div_0_ng_container_1_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 7);
} if (rf & 2) {
    var i_r3 = ɵngcc0.ɵɵnextContext().index;
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("thy-stepper-line-active", i_r3 < ctx_r5.selectedIndex);
} }
function ThyStepperComponent_div_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-step-header", 5);
    ɵngcc0.ɵɵtemplate(2, ThyStepperComponent_div_0_ng_container_1_div_2_Template, 1, 2, "div", 6);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var step_r2 = ctx.$implicit;
    var i_r3 = ctx.index;
    var isLast_r4 = ctx.last;
    var ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("thy-stepper-header-active", i_r3 <= ctx_r1.selectedIndex);
    ɵngcc0.ɵɵproperty("index", i_r3)("label", step_r2.thyLabel || step_r2.label)("selected", ctx_r1.selectedIndex === i_r3)("active", i_r3 <= ctx_r1.selectedIndex);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !isLast_r4);
} }
function ThyStepperComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵtemplate(1, ThyStepperComponent_div_0_ng_container_1_Template, 3, 7, "ng-container", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.steps);
} }
'use strict';

    var ThyStepHeaderComponent = /** @class */ (function () {
        function ThyStepHeaderComponent() {
            this.thyStepHeader = true;
        }
ThyStepHeaderComponent.ɵfac = function ThyStepHeaderComponent_Factory(t) { return new (t || ThyStepHeaderComponent)(); };
ThyStepHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStepHeaderComponent, selectors: [["thy-step-header"]], hostVars: 4, hostBindings: function ThyStepHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-stepper-header", ctx.thyStepHeader)("thy-stepper-header-selected", ctx.selected);
    } }, inputs: { label: "label", index: "index", selected: "selected", active: "active" }, decls: 5, vars: 2, consts: [[1, "thy-step-number"], [1, "thy-step-label"]], template: function ThyStepHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "span");
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 1);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.index + 1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.label);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepHeaderComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-step-header',
                template: "<div class=\"thy-step-number\"> <span>{{index + 1}}</span></div>\n<div class=\"thy-step-label\">{{label}}</div>"
            }]
    }], function () { return []; }, { thyStepHeader: [{
            type: core.HostBinding,
            args: ['class.thy-stepper-header']
        }], label: [{
            type: core.Input
        }], index: [{
            type: core.Input
        }], selected: [{
            type: core.HostBinding,
            args: ['class.thy-stepper-header-selected']
        }, {
            type: core.Input
        }], active: [{
            type: core.Input
        }] }); })();
        return ThyStepHeaderComponent;
    }());
    ThyStepHeaderComponent.propDecorators = {
        label: [{ type: core.Input }],
        index: [{ type: core.Input }],
        selected: [{ type: core.HostBinding, args: ['class.thy-stepper-header-selected',] }, { type: core.Input }],
        active: [{ type: core.Input }],
        thyStepHeader: [{ type: core.HostBinding, args: ['class.thy-stepper-header',] }]
    };

    var THY_STEPPER_COMPONENT = new core.InjectionToken('THY_STEPPER_COMPONENT');
    var ThyStepComponent = /** @class */ (function () {
        function ThyStepComponent(stepper) {
            this.stepper = stepper;
        }
        ThyStepComponent.prototype.select = function () {
            this.stepper.selected = this;
        };
ThyStepComponent.ɵfac = function ThyStepComponent_Factory(t) { return new (t || ThyStepComponent)(ɵngcc0.ɵɵdirectiveInject(THY_STEPPER_COMPONENT, 8)); };
ThyStepComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStepComponent, selectors: [["thy-step"]], viewQuery: function ThyStepComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(core.TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { label: "label", thyLabel: "thyLabel" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyStepComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThyStepComponent_ng_template_0_Template, 1, 0, "ng-template");
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-step',
                template: "<ng-template><ng-content></ng-content></ng-template>\n"
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: core.Optional
            }, {
                type: core.Inject,
                args: [THY_STEPPER_COMPONENT]
            }] }]; }, { content: [{
            type: core.ViewChild,
            args: [core.TemplateRef, { static: true }]
        }], label: [{
            type: core.Input
        }], thyLabel: [{
            type: core.Input
        }] }); })();
        return ThyStepComponent;
    }());
    ThyStepComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [THY_STEPPER_COMPONENT,] }] }
    ]; };
    ThyStepComponent.propDecorators = {
        content: [{ type: core.ViewChild, args: [core.TemplateRef, { static: true },] }],
        label: [{ type: core.Input }],
        thyLabel: [{ type: core.Input }]
    };

    var ThyStepperComponent = /** @class */ (function () {
        function ThyStepperComponent() {
            this.thyShowStepHeader = true;
            this._selectedIndex = 0;
            this.selectionChange = new core.EventEmitter();
            this.thyStepper = true;
        }
        Object.defineProperty(ThyStepperComponent.prototype, "thySelectedIndex", {
            set: function (value) {
                this.selectedIndex = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStepperComponent.prototype, "thySelected", {
            set: function (value) {
                this.selected = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStepperComponent.prototype, "selected", {
            get: function () {
                return this.steps ? this.steps.toArray()[this.selectedIndex] : null;
            },
            set: function (step) {
                this.selectedIndex = this.steps ? this.steps.toArray().indexOf(step) : -1;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStepperComponent.prototype, "selectedIndex", {
            get: function () {
                return this._selectedIndex;
            },
            set: function (index) {
                if (this.steps) {
                    this._updateSelectedItemIndex(index);
                }
                else {
                    this._selectedIndex = index;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyStepperComponent.prototype._updateSelectedItemIndex = function (newIndex) {
            var stepsArray = this.steps.toArray();
            this.selectionChange.emit({
                selectedIndex: newIndex,
                previouslySelectedIndex: this._selectedIndex,
                selectedStep: stepsArray[newIndex],
                previouslySelectedStep: stepsArray[this._selectedIndex]
            });
            this._selectedIndex = newIndex;
        };
        ThyStepperComponent.prototype.to = function (index) {
            this.selectedIndex = Math.min(index, this.steps.length - 1);
        };
        ThyStepperComponent.prototype.next = function () {
            this.selectedIndex = Math.min(this._selectedIndex + 1, this.steps.length - 1);
        };
        /** Selects and focuses the previous step in list. */
        ThyStepperComponent.prototype.previous = function () {
            this.selectedIndex = Math.max(this._selectedIndex - 1, 0);
        };
ThyStepperComponent.ɵfac = function ThyStepperComponent_Factory(t) { return new (t || ThyStepperComponent)(); };
ThyStepperComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStepperComponent, selectors: [["thy-stepper"]], contentQueries: function ThyStepperComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyStepComponent, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.steps = _t);
    } }, viewQuery: function ThyStepperComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(ThyStepHeaderComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepHeaders = _t);
    } }, hostVars: 2, hostBindings: function ThyStepperComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-stepper", ctx.thyStepper);
    } }, inputs: { thyShowStepHeader: "thyShowStepHeader", thySelectedIndex: "thySelectedIndex", thySelected: "thySelected" }, outputs: { selectionChange: "selectionChange" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: THY_STEPPER_COMPONENT,
                useExisting: ThyStepperComponent
            }
        ])], decls: 3, vars: 2, consts: [["class", "thy-stepper-header-container", 4, "ngIf"], [1, "thy-stepper-content-container"], [3, "ngTemplateOutlet"], [1, "thy-stepper-header-container"], [4, "ngFor", "ngForOf"], [3, "index", "label", "selected", "active"], ["class", "thy-stepper-line", 3, "thy-stepper-line-active", 4, "ngIf"], [1, "thy-stepper-line"]], template: function ThyStepperComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyStepperComponent_div_0_Template, 2, 1, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementContainer(2, 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowStepHeader);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.selected == null ? null : ctx.selected.content);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgForOf, ThyStepHeaderComponent], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-stepper',
                template: "<div *ngIf=\"thyShowStepHeader\" class=\"thy-stepper-header-container\">\n  <ng-container *ngFor=\"let step of steps; let i = index; let isLast = last\">\n    <thy-step-header\n      [index]=\"i\"\n      [label]=\"step.thyLabel || step.label\"\n      [selected]=\"selectedIndex === i\"\n      [class.thy-stepper-header-active]=\"i <= selectedIndex\"\n      [active]=\"i <= selectedIndex\"\n    >\n    </thy-step-header>\n    <div *ngIf=\"!isLast\" class=\"thy-stepper-line\" [class.thy-stepper-line-active]=\"i < selectedIndex\"></div>\n  </ng-container>\n</div>\n\n<div class=\"thy-stepper-content-container\">\n  <ng-container [ngTemplateOutlet]=\"selected?.content\"></ng-container>\n</div>\n",
                providers: [
                    {
                        provide: THY_STEPPER_COMPONENT,
                        useExisting: ThyStepperComponent
                    }
                ]
            }]
    }], function () { return []; }, { thyShowStepHeader: [{
            type: core.Input
        }], selectionChange: [{
            type: core.Output
        }], thyStepper: [{
            type: core.HostBinding,
            args: ['class.thy-stepper']
        }], thySelectedIndex: [{
            type: core.Input
        }], thySelected: [{
            type: core.Input
        }], stepHeaders: [{
            type: core.ViewChildren,
            args: [ThyStepHeaderComponent]
        }], steps: [{
            type: core.ContentChildren,
            args: [ThyStepComponent]
        }] }); })();
        return ThyStepperComponent;
    }());
    ThyStepperComponent.propDecorators = {
        thySelectedIndex: [{ type: core.Input }],
        thySelected: [{ type: core.Input }],
        thyShowStepHeader: [{ type: core.Input }],
        selectionChange: [{ type: core.Output }],
        stepHeaders: [{ type: core.ViewChildren, args: [ThyStepHeaderComponent,] }],
        steps: [{ type: core.ContentChildren, args: [ThyStepComponent,] }],
        thyStepper: [{ type: core.HostBinding, args: ['class.thy-stepper',] }]
    };

    /** Button that moves to the next step in a stepper workflow. */
    var ThyStepperNextDirective = /** @class */ (function () {
        function ThyStepperNextDirective(stepper) {
            this.stepper = stepper;
        }
        ThyStepperNextDirective.prototype.click = function ($event) {
            this.stepper.next();
        };
ThyStepperNextDirective.ɵfac = function ThyStepperNextDirective_Factory(t) { return new (t || ThyStepperNextDirective)(ɵngcc0.ɵɵdirectiveInject(ThyStepperComponent)); };
ThyStepperNextDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyStepperNextDirective, selectors: [["", "thyStepperNext", ""]], hostBindings: function ThyStepperNextDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyStepperNextDirective_click_HostBindingHandler($event) { return ctx.click($event); });
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperNextDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyStepperNext]'
            }]
    }], function () { return [{ type: ThyStepperComponent }]; }, { click: [{
            type: core.HostListener,
            args: ['click', ['$event']]
        }] }); })();
        return ThyStepperNextDirective;
    }());
    ThyStepperNextDirective.ctorParameters = function () { return [
        { type: ThyStepperComponent }
    ]; };
    ThyStepperNextDirective.propDecorators = {
        click: [{ type: core.HostListener, args: ['click', ['$event'],] }]
    };
    /** Button that moves to the previous step in a stepper workflow. */
    var ThyStepperPreviousDirective = /** @class */ (function () {
        function ThyStepperPreviousDirective(stepper) {
            this.stepper = stepper;
        }
        ThyStepperPreviousDirective.prototype.click = function ($event) {
            this.stepper.previous();
        };
ThyStepperPreviousDirective.ɵfac = function ThyStepperPreviousDirective_Factory(t) { return new (t || ThyStepperPreviousDirective)(ɵngcc0.ɵɵdirectiveInject(ThyStepperComponent)); };
ThyStepperPreviousDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyStepperPreviousDirective, selectors: [["", "thyStepperPrevious", ""]], hostBindings: function ThyStepperPreviousDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyStepperPreviousDirective_click_HostBindingHandler($event) { return ctx.click($event); });
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperPreviousDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyStepperPrevious]'
            }]
    }], function () { return [{ type: ThyStepperComponent }]; }, { click: [{
            type: core.HostListener,
            args: ['click', ['$event']]
        }] }); })();
        return ThyStepperPreviousDirective;
    }());
    ThyStepperPreviousDirective.ctorParameters = function () { return [
        { type: ThyStepperComponent }
    ]; };
    ThyStepperPreviousDirective.propDecorators = {
        click: [{ type: core.HostListener, args: ['click', ['$event'],] }]
    };

    var ThyStepperModule = /** @class */ (function () {
        function ThyStepperModule() {
        }
ThyStepperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyStepperModule });
ThyStepperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyStepperModule_Factory(t) { return new (t || ThyStepperModule)(); }, providers: [], imports: [[common.CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyStepperModule, { declarations: [ThyStepHeaderComponent, ThyStepperComponent, ThyStepComponent, ThyStepperNextDirective, ThyStepperPreviousDirective], imports: [ɵngcc1.CommonModule], exports: [ThyStepperComponent, ThyStepComponent, ThyStepHeaderComponent, ThyStepperNextDirective, ThyStepperPreviousDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyStepHeaderComponent, ThyStepperComponent, ThyStepComponent, ThyStepperNextDirective, ThyStepperPreviousDirective],
                entryComponents: [],
                imports: [common.CommonModule],
                exports: [ThyStepperComponent, ThyStepComponent, ThyStepHeaderComponent, ThyStepperNextDirective, ThyStepperPreviousDirective],
                providers: []
            }]
    }], function () { return []; }, null); })();
        return ThyStepperModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.THY_STEPPER_COMPONENT = THY_STEPPER_COMPONENT;
    exports.ThyStepComponent = ThyStepComponent;
    exports.ThyStepHeaderComponent = ThyStepHeaderComponent;
    exports.ThyStepperComponent = ThyStepperComponent;
    exports.ThyStepperModule = ThyStepperModule;
    exports.ThyStepperNextDirective = ThyStepperNextDirective;
    exports.ThyStepperPreviousDirective = ThyStepperPreviousDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
