// ngx-tethys/bundles/ngx-tethys-calendar.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/date-range'),require('@angular/forms'),require('@angular/common'),require('ngx-tethys/button'),require('ngx-tethys/date-picker'),require('ngx-tethys/shared'),require('ngx-tethys/icon'),require('ngx-tethys/select'),require('ngx-tethys/radio'),exports, require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('ngx-tethys/shared'), require('ngx-tethys/icon'), require('date-fns'), require('ngx-tethys/util'), require('ngx-tethys/core'), require('ngx-tethys/date-picker'), require('ngx-tethys/select'), require('ngx-tethys/radio'), require('ngx-tethys/button'), require('ngx-tethys/date-range')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/calendar', ['@angular/core','ngx-tethys/date-range','@angular/forms','@angular/common','ngx-tethys/button','ngx-tethys/date-picker','ngx-tethys/shared','ngx-tethys/icon','ngx-tethys/select','ngx-tethys/radio','exports', '@angular/core', '@angular/common', '@angular/forms', 'ngx-tethys/shared', 'ngx-tethys/icon', 'date-fns', 'ngx-tethys/util', 'ngx-tethys/core', 'ngx-tethys/date-picker', 'ngx-tethys/select', 'ngx-tethys/radio', 'ngx-tethys/button', 'ngx-tethys/date-range'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.dateRange,global.ng.forms,global.ng.common,global.ngxTethys.button,global.ngxTethys.datePicker,global.ngxTethys.shared,global.ngxTethys.icon,global.ngxTethys.select,global.ngxTethys.radio,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].calendar = {}), global.ng.core, global.ng.common, global.ng.forms, global['ngx-tethys'].shared, global['ngx-tethys'].icon, global.dateFns, global['ngx-tethys'].util, global['ngx-tethys'].core, global['ngx-tethys']['date-picker'], global['ngx-tethys'].select, global['ngx-tethys'].radio, global['ngx-tethys'].button, global['ngx-tethys']['date-range']));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,exports, core, common, forms, shared, icon, dateFns, util, core$1, datePicker, select, radio, button, dateRange) { 
function ThyCalendarHeaderComponent_button_2_Template(rf, ctx) { if (rf & 1) {
    var _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 5);
    ɵngcc0.ɵɵlistener("click", function ThyCalendarHeaderComponent_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.backToday(); });
    ɵngcc0.ɵɵtext(1, " \u4ECA\u5929 ");
    ɵngcc0.ɵɵelementEnd();
} }
function ThyCalendarHeaderComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, ctx_r1.operationRender));
} }
var _c0 = function (a0) { return { $implicit: a0 }; };
function ThyCalendarComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThyCalendarComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "date-table", 6);
    ɵngcc0.ɵɵlistener("valueChange", function ThyCalendarComponent_ng_template_5_Template_date_table_valueChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); var ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.onDateSelect($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("prefixCls", ctx_r2.prefixCls)("value", ctx_r2.currentDate)("selectedValue", ctx_r2.currentDate)("dateCellRender", ctx_r2.dateCell)("disabledDate", ctx_r2.thyDisabledDate);
} }
function ThyCalendarComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    var _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "month-table", 7);
    ɵngcc0.ɵɵlistener("valueChange", function ThyCalendarComponent_ng_template_7_Template_month_table_valueChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.onDateSelect($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("value", ctx_r4.currentDate);
} }
'use strict';

    var ThyCalendarHeaderComponent = /** @class */ (function () {
        function ThyCalendarHeaderComponent(cdr) {
            this.cdr = cdr;
            this.className = true;
            // @Output() readonly modeChange: EventEmitter<'month' | 'year'> = new EventEmitter();
            this.yearChange = new core.EventEmitter();
            this.monthChange = new core.EventEmitter();
            this.dateRangeChange = new core.EventEmitter();
            this.dateRanges = [
                {
                    key: 'month',
                    text: dateFns.getYear(new Date()) + '年' + (dateFns.getMonth(new Date()) + 1) + '月',
                    begin: util.getUnixTime(util.startOfMonth(new Date())),
                    end: util.getUnixTime(util.endOfMonth(new Date())),
                    timestamp: {
                        interval: 1,
                        unit: 'month'
                    }
                }
            ];
        }
        Object.defineProperty(ThyCalendarHeaderComponent.prototype, "currentDate", {
            // @Input() mode: 'month' | 'year' = 'month';
            // @Input() fullscreen = true;
            set: function (value) {
                this.setDate(value);
            },
            enumerable: false,
            configurable: true
        });
        ThyCalendarHeaderComponent.prototype.ngOnInit = function () { };
        ThyCalendarHeaderComponent.prototype.onChangeMonth = function (month) {
            var currentMonth = dateFns.fromUnixTime(month.begin).getMonth();
            this.monthChange.emit(currentMonth);
        };
        ThyCalendarHeaderComponent.prototype.onChangeYear = function (year) {
            var currentYear = dateFns.fromUnixTime(year.begin).getFullYear();
            this.yearChange.emit(currentYear);
        };
        ThyCalendarHeaderComponent.prototype.onChangeRange = function (dateRange) {
            this.isCurrentDate(this._currentDate);
            this.onChangeYear(dateRange);
            this.onChangeMonth(dateRange);
            this.dateRangeChange.emit(dateRange);
        };
        ThyCalendarHeaderComponent.prototype.backToday = function () {
            this._currentDate = new util.TinyDate();
            this.date = Object.assign({}, this.dateRanges[0]);
            this.onChangeRange(this.date);
            this.cdr.detectChanges();
        };
        ThyCalendarHeaderComponent.prototype.setDate = function (value) {
            this.isCurrentDate(value);
            if (this.isCurrent) {
                this._currentDate = value;
                var dateRange = Object.assign(Object.assign({}, this.dateRanges[0]), { key: 'exception', text: dateFns.getYear(this._currentDate.nativeDate) + '年' + (dateFns.getMonth(this._currentDate.nativeDate) + 1) + '月', begin: util.getUnixTime(util.startOfMonth(this._currentDate.nativeDate)), end: util.getUnixTime(util.endOfMonth(this._currentDate.nativeDate)) });
                this.date = dateRange;
            }
            else {
                this._currentDate = new util.TinyDate();
                this.date = Object.assign({}, this.dateRanges[0]);
            }
        };
        ThyCalendarHeaderComponent.prototype.isCurrentDate = function (currentDate) {
            this.isCurrent = currentDate.getMonth() !== dateFns.getMonth(new Date()) || currentDate.getYear() !== dateFns.getYear(new Date());
        };
ThyCalendarHeaderComponent.ɵfac = function ThyCalendarHeaderComponent_Factory(t) { return new (t || ThyCalendarHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyCalendarHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCalendarHeaderComponent, selectors: [["thy-calendar-header"]], hostVars: 2, hostBindings: function ThyCalendarHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-calendar-full-header-container", ctx.className);
    } }, inputs: { currentDate: "currentDate", operationRender: "operationRender" }, outputs: { yearChange: "yearChange", monthChange: "monthChange", dateRangeChange: "dateRangeChange" }, decls: 5, vars: 9, consts: [[1, "thy-calendar-full-header-left"], [1, "select-date-range", 3, "ngModel", "thyHiddenMenu", "thyCustomKey", "thyPickerFormat", "ngModelChange"], ["thyButton", "outline-default-square", "thySize", "md", 3, "click", 4, "ngIf"], [1, "thy-calendar-full-header-right"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], ["thyButton", "outline-default-square", "thySize", "md", 3, "click"]], template: function ThyCalendarHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "thy-date-range", 1);
        ɵngcc0.ɵɵlistener("ngModelChange", function ThyCalendarHeaderComponent_Template_thy_date_range_ngModelChange_1_listener($event) { return ctx.onChangeRange($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, ThyCalendarHeaderComponent_button_2_Template, 2, 0, "button", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyCalendarHeaderComponent_ng_container_4_Template, 4, 3, "ng-container", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngModel", ctx.date)("thyHiddenMenu", true)("thyCustomKey", "exception")("thyPickerFormat", "yyyy\u5E74MM\u6708");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isCurrent);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.operationRender)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(7, _c0, ctx.operationRender));
    } }, directives: [ɵngcc1.ThyDateRangeComponent, ɵngcc2.NgControlStatus, ɵngcc2.NgModel, ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet, ɵngcc4.ThyButtonComponent], pipes: [ɵngcc3.JsonPipe], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCalendarHeaderComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-calendar-header',
                template: "<div class=\"thy-calendar-full-header-left\">\n  <thy-date-range\n    class=\"select-date-range\"\n    [ngModel]=\"date\"\n    [thyHiddenMenu]=\"true\"\n    [thyCustomKey]=\"'exception'\"\n    [thyPickerFormat]=\"'yyyy\u5E74MM\u6708'\"\n    (ngModelChange)=\"onChangeRange($event)\"\n  ></thy-date-range>\n  <button *ngIf=\"isCurrent\" thyButton=\"outline-default-square\" thySize=\"md\" (click)=\"backToday()\">\n    \u4ECA\u5929\n  </button>\n</div>\n<div class=\"thy-calendar-full-header-right\">\n  <ng-container *ngTemplateOutlet=\"$any(operationRender); context: { $implicit: operationRender }\">\n    <span>{{ operationRender | json }}</span>\n  </ng-container>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { className: [{
            type: core.HostBinding,
            args: ['class.thy-calendar-full-header-container']
        }], yearChange: [{
            type: core.Output
        }], monthChange: [{
            type: core.Output
        }], dateRangeChange: [{
            type: core.Output
        }], currentDate: [{
            type: core.Input
        }], operationRender: [{
            type: core.Input
        }] }); })();
        return ThyCalendarHeaderComponent;
    }());
    ThyCalendarHeaderComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef }
    ]; };
    ThyCalendarHeaderComponent.propDecorators = {
        className: [{ type: core.HostBinding, args: ['class.thy-calendar-full-header-container',] }],
        currentDate: [{ type: core.Input }],
        operationRender: [{ type: core.Input }],
        yearChange: [{ type: core.Output }],
        monthChange: [{ type: core.Output }],
        dateRangeChange: [{ type: core.Output }]
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ThyDateCellDirective = /** @class */ (function () {
        function ThyDateCellDirective() {
        }
ThyDateCellDirective.ɵfac = function ThyDateCellDirective_Factory(t) { return new (t || ThyDateCellDirective)(); };
ThyDateCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDateCellDirective, selectors: [["", "thyDateCell", ""]], exportAs: ["thyDateCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDateCellDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyDateCell]',
                exportAs: 'thyDateCell'
            }]
    }], function () { return []; }, null); })();
        return ThyDateCellDirective;
    }());
    var ThyMonthCellDirective = /** @class */ (function () {
        function ThyMonthCellDirective() {
        }
ThyMonthCellDirective.ɵfac = function ThyMonthCellDirective_Factory(t) { return new (t || ThyMonthCellDirective)(); };
ThyMonthCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMonthCellDirective, selectors: [["", "thyMonthCell", ""]], exportAs: ["thyMonthCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMonthCellDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyMonthCell]',
                exportAs: 'thyMonthCell'
            }]
    }], function () { return []; }, null); })();
        return ThyMonthCellDirective;
    }());
    var ThyDateFullCellDirective = /** @class */ (function () {
        function ThyDateFullCellDirective() {
        }
ThyDateFullCellDirective.ɵfac = function ThyDateFullCellDirective_Factory(t) { return new (t || ThyDateFullCellDirective)(); };
ThyDateFullCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDateFullCellDirective, selectors: [["", "thyDateFullCell", ""]], exportAs: ["thyDateFullCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDateFullCellDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyDateFullCell]',
                exportAs: 'thyDateFullCell'
            }]
    }], function () { return []; }, null); })();
        return ThyDateFullCellDirective;
    }());
    var ThyMonthFullCellDirective = /** @class */ (function () {
        function ThyMonthFullCellDirective() {
        }
ThyMonthFullCellDirective.ɵfac = function ThyMonthFullCellDirective_Factory(t) { return new (t || ThyMonthFullCellDirective)(); };
ThyMonthFullCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMonthFullCellDirective, selectors: [["", "thyMonthFullCell", ""]], exportAs: ["thyMonthFullCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMonthFullCellDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyMonthFullCell]',
                exportAs: 'thyMonthFullCell'
            }]
    }], function () { return []; }, null); })();
        return ThyMonthFullCellDirective;
    }());
    var ThyCalendarHeaderOperationDirective = /** @class */ (function () {
        function ThyCalendarHeaderOperationDirective() {
        }
ThyCalendarHeaderOperationDirective.ɵfac = function ThyCalendarHeaderOperationDirective_Factory(t) { return new (t || ThyCalendarHeaderOperationDirective)(); };
ThyCalendarHeaderOperationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyCalendarHeaderOperationDirective, selectors: [["", "thyCalendarHeaderOperation", ""]], exportAs: ["thyCalendarHeaderOperation"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCalendarHeaderOperationDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyCalendarHeaderOperation]',
                exportAs: 'thyCalendarHeaderOperation'
            }]
    }], function () { return []; }, null); })();
        return ThyCalendarHeaderOperationDirective;
    }());

    var ThyCalendarComponent = /** @class */ (function () {
        function ThyCalendarComponent(cdr) {
            this.cdr = cdr;
            this.className = true;
            this.className1 = true;
            // @HostBinding('class.thy-pick-calendar-mini') className2 = !thyFullscreen;
            this.thyMode = 'month';
            this.thyModeChange = new core.EventEmitter();
            this.thyPanelChange = new core.EventEmitter();
            this.thySelectChange = new core.EventEmitter();
            this.thyValueChange = new core.EventEmitter();
            this.thyDateRangeChange = new core.EventEmitter();
            this.thyFullscreen = true;
            this.currentDate = new util.TinyDate();
            this.prefixCls = 'thy-calendar-full';
            this.onChangeFn = function () { };
            this.onTouchFn = function () { };
        }
        Object.defineProperty(ThyCalendarComponent.prototype, "dateCell", {
            get: function () {
                return (this.thyDateCell || this.thyDateCellChild);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCalendarComponent.prototype, "headerOperation", {
            get: function () {
                return (this.thyCalendarHeaderOperation || this.thyCalendarHeaderOperationChild);
            },
            enumerable: false,
            configurable: true
        });
        ThyCalendarComponent.prototype.ngOnInit = function () { };
        // onModeChange(mode: CalendarMode): void {
        //     this.thyModeChange.emit(mode);
        //     this.thyPanelChange.emit({ date: this.currentDate.nativeDate, mode });
        // }
        ThyCalendarComponent.prototype.onYearSelect = function (year) {
            var date = this.currentDate.setYear(year);
            this.updateDate(date);
        };
        ThyCalendarComponent.prototype.onMonthSelect = function (month) {
            var date = this.currentDate.setMonth(month);
            this.updateDate(date);
        };
        ThyCalendarComponent.prototype.onDateSelect = function (date) {
            // Only currentDate is enough in calendar
            // this.value = date;
            this.updateDate(date);
        };
        ThyCalendarComponent.prototype.onDateRangeSelect = function (date) {
            this.thyDateRangeChange.emit(date);
        };
        ThyCalendarComponent.prototype.writeValue = function (value) {
            this.updateDate(new util.TinyDate(value), false);
            this.cdr.markForCheck();
        };
        ThyCalendarComponent.prototype.registerOnChange = function (fn) {
            this.onChangeFn = fn;
        };
        ThyCalendarComponent.prototype.registerOnTouched = function (fn) {
            this.onTouchFn = fn;
        };
        ThyCalendarComponent.prototype.updateDate = function (date, touched) {
            if (touched === void 0) { touched = true; }
            this.currentDate = date;
            if (touched) {
                this.onChangeFn(date.nativeDate);
                this.onTouchFn();
                this.thySelectChange.emit(date.nativeDate);
                this.thyValueChange.emit(date.nativeDate);
            }
        };
        ThyCalendarComponent.prototype.ngOnChanges = function (changes) {
            if (changes.thyValue) {
                this.updateDate(new util.TinyDate(this.thyValue), false);
            }
        };
ThyCalendarComponent.ɵfac = function ThyCalendarComponent_Factory(t) { return new (t || ThyCalendarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyCalendarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCalendarComponent, selectors: [["thy-calendar"]], contentQueries: function ThyCalendarComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyDateCellDirective, true, core.TemplateRef);
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyCalendarHeaderOperationDirective, true, core.TemplateRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.thyDateCellChild = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.thyCalendarHeaderOperationChild = _t.first);
    } }, hostVars: 4, hostBindings: function ThyCalendarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-calendar-container", ctx.className)("thy-calendar-full", ctx.className1);
    } }, inputs: { thyMode: "thyMode", thyFullscreen: "thyFullscreen", thyValue: "thyValue", thyDisabledDate: "thyDisabledDate", thyDateCell: "thyDateCell", thyCalendarHeaderOperation: "thyCalendarHeaderOperation" }, outputs: { thyModeChange: "thyModeChange", thyPanelChange: "thyPanelChange", thySelectChange: "thySelectChange", thyValueChange: "thyValueChange", thyDateRangeChange: "thyDateRangeChange" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: forms.NG_VALUE_ACCESSOR, useExisting: core.forwardRef(function () { return ThyCalendarComponent; }), multi: true }]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 9, vars: 8, consts: [[3, "operationRender", "currentDate", "monthChange", "yearChange", "dateRangeChange"], [1, "thy-calendar-panel"], [1, "thy-calendar-body"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["monthModeTable", ""], ["yearModeTable", ""], [3, "prefixCls", "value", "selectedValue", "dateCellRender", "disabledDate", "valueChange"], [3, "value", "valueChange"]], template: function ThyCalendarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "thy-calendar-header", 0);
        ɵngcc0.ɵɵlistener("monthChange", function ThyCalendarComponent_Template_thy_calendar_header_monthChange_0_listener($event) { return ctx.onMonthSelect($event); })("yearChange", function ThyCalendarComponent_Template_thy_calendar_header_yearChange_0_listener($event) { return ctx.onYearSelect($event); })("dateRangeChange", function ThyCalendarComponent_Template_thy_calendar_header_dateRangeChange_0_listener($event) { return ctx.onDateRangeSelect($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div");
        ɵngcc0.ɵɵelementStart(3, "div", 2);
        ɵngcc0.ɵɵtemplate(4, ThyCalendarComponent_ng_container_4_Template, 1, 0, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ThyCalendarComponent_ng_template_5_Template, 1, 5, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(7, ThyCalendarComponent_ng_template_7_Template, 1, 1, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        var _r1 = ɵngcc0.ɵɵreference(6);
        var _r3 = ɵngcc0.ɵɵreference(8);
        ɵngcc0.ɵɵproperty("operationRender", ctx.headerOperation)("currentDate", ctx.currentDate);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵclassMapInterpolate1("thy-calendar-", ctx.thyMode === "month" ? "date" : "month", "-panel");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyMode === "month")("ngIfThen", _r1)("ngIfElse", _r3);
    } }, directives: [ThyCalendarHeaderComponent, ɵngcc3.NgIf, ɵngcc5.DateTableComponent, ɵngcc5.MonthTableComponent], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCalendarComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-calendar',
                template: "<thy-calendar-header\n  [operationRender]=\"headerOperation\"\n  (monthChange)=\"onMonthSelect($event)\"\n  (yearChange)=\"onYearSelect($event)\"\n  (dateRangeChange)=\"onDateRangeSelect($event)\"\n  [currentDate]=\"currentDate\"\n></thy-calendar-header>\n\n<div class=\"thy-calendar-panel\">\n  <div class=\"thy-calendar-{{ thyMode === 'month' ? 'date' : 'month' }}-panel\">\n    <div class=\"thy-calendar-body\">\n      <ng-container *ngIf=\"thyMode === 'month'; then monthModeTable; else yearModeTable\"></ng-container>\n    </div>\n  </div>\n</div>\n<ng-template #monthModeTable>\n  <date-table\n    [prefixCls]=\"prefixCls\"\n    [value]=\"currentDate\"\n    [selectedValue]=\"currentDate\"\n    [dateCellRender]=\"dateCell\"\n    [disabledDate]=\"thyDisabledDate\"\n    (valueChange)=\"onDateSelect($event)\"\n  ></date-table>\n</ng-template>\n\n<ng-template #yearModeTable>\n  <month-table [value]=\"currentDate\" (valueChange)=\"onDateSelect($event)\"></month-table>\n</ng-template>\n",
                encapsulation: core.ViewEncapsulation.None,
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [{ provide: forms.NG_VALUE_ACCESSOR, useExisting: core.forwardRef(function () { return ThyCalendarComponent; }), multi: true }]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { className: [{
            type: core.HostBinding,
            args: ['class.thy-calendar-container']
        }], className1: [{
            type: core.HostBinding,
            args: ['class.thy-calendar-full']
        }], thyMode: [{
            type: core.Input
        }], thyModeChange: [{
            type: core.Output
        }], thyPanelChange: [{
            type: core.Output
        }], thySelectChange: [{
            type: core.Output
        }], thyValueChange: [{
            type: core.Output
        }], thyDateRangeChange: [{
            type: core.Output
        }], thyFullscreen: [{
            type: core.Input
        }], thyValue: [{
            type: core.Input
        }], thyDisabledDate: [{
            type: core.Input
        }], thyDateCell: [{
            type: core.Input
        }], thyDateCellChild: [{
            type: core.ContentChild,
            args: [ThyDateCellDirective, { read: core.TemplateRef }]
        }], thyCalendarHeaderOperation: [{
            type: core.Input
        }], thyCalendarHeaderOperationChild: [{
            type: core.ContentChild,
            args: [ThyCalendarHeaderOperationDirective, { read: core.TemplateRef }]
        }] }); })();
        return ThyCalendarComponent;
    }());
    ThyCalendarComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef }
    ]; };
    ThyCalendarComponent.propDecorators = {
        className: [{ type: core.HostBinding, args: ['class.thy-calendar-container',] }],
        className1: [{ type: core.HostBinding, args: ['class.thy-calendar-full',] }],
        thyMode: [{ type: core.Input }],
        thyValue: [{ type: core.Input }],
        thyDisabledDate: [{ type: core.Input }],
        thyModeChange: [{ type: core.Output }],
        thyPanelChange: [{ type: core.Output }],
        thySelectChange: [{ type: core.Output }],
        thyValueChange: [{ type: core.Output }],
        thyDateRangeChange: [{ type: core.Output }],
        thyDateCell: [{ type: core.Input }],
        thyDateCellChild: [{ type: core.ContentChild, args: [ThyDateCellDirective, { read: core.TemplateRef },] }],
        thyCalendarHeaderOperation: [{ type: core.Input }],
        thyCalendarHeaderOperationChild: [{ type: core.ContentChild, args: [ThyCalendarHeaderOperationDirective, { read: core.TemplateRef },] }],
        thyFullscreen: [{ type: core.Input }]
    };
    __decorate([
        core$1.InputBoolean(),
        __metadata("design:type", Object)
    ], ThyCalendarComponent.prototype, "thyFullscreen", void 0);

    var ThyCalendarModule = /** @class */ (function () {
        function ThyCalendarModule() {
        }
ThyCalendarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCalendarModule });
ThyCalendarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCalendarModule_Factory(t) { return new (t || ThyCalendarModule)(); }, imports: [[
            common.CommonModule,
            forms.FormsModule,
            shared.ThySharedModule,
            icon.ThyIconModule,
            datePicker.LibPackerModule,
            select.ThySelectModule,
            radio.ThyRadioModule,
            button.ThyButtonModule,
            dateRange.ThyDateRangeModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCalendarModule, { declarations: [ThyCalendarHeaderComponent, ThyCalendarComponent, ThyDateCellDirective, ThyDateFullCellDirective, ThyMonthCellDirective, ThyMonthFullCellDirective, ThyCalendarHeaderOperationDirective], imports: [ɵngcc3.CommonModule, ɵngcc2.FormsModule, ɵngcc6.ThySharedModule, ɵngcc7.ThyIconModule, ɵngcc5.LibPackerModule, ɵngcc8.ThySelectModule, ɵngcc9.ThyRadioModule, ɵngcc4.ThyButtonModule, ɵngcc1.ThyDateRangeModule], exports: [ThyCalendarHeaderComponent, ThyCalendarComponent, ThyDateCellDirective, ThyDateFullCellDirective, ThyMonthCellDirective, ThyMonthFullCellDirective, ThyCalendarHeaderOperationDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCalendarModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    shared.ThySharedModule,
                    icon.ThyIconModule,
                    datePicker.LibPackerModule,
                    select.ThySelectModule,
                    radio.ThyRadioModule,
                    button.ThyButtonModule,
                    dateRange.ThyDateRangeModule
                ],
                declarations: [
                    ThyCalendarHeaderComponent,
                    ThyCalendarComponent,
                    ThyDateCellDirective,
                    ThyDateFullCellDirective,
                    ThyMonthCellDirective,
                    ThyMonthFullCellDirective,
                    ThyCalendarHeaderOperationDirective
                ],
                exports: [
                    ThyCalendarHeaderComponent,
                    ThyCalendarComponent,
                    ThyDateCellDirective,
                    ThyDateFullCellDirective,
                    ThyMonthCellDirective,
                    ThyMonthFullCellDirective,
                    ThyCalendarHeaderOperationDirective
                ]
            }]
    }], function () { return []; }, null); })();
        return ThyCalendarModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyCalendarComponent = ThyCalendarComponent;
    exports.ThyCalendarHeaderComponent = ThyCalendarHeaderComponent;
    exports.ThyCalendarHeaderOperationDirective = ThyCalendarHeaderOperationDirective;
    exports.ThyCalendarModule = ThyCalendarModule;
    exports.ThyDateCellDirective = ThyDateCellDirective;
    exports.ThyDateFullCellDirective = ThyDateFullCellDirective;
    exports.ThyMonthCellDirective = ThyMonthCellDirective;
    exports.ThyMonthFullCellDirective = ThyMonthFullCellDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
