// ngx-tethys/bundles/ngx-tethys-table.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/cdk/overlay'),require('ngx-tethys/core'),require('@angular/common'),require('ngx-tethys/loading'),require('ngx-tethys/shared'),require('@angular/cdk/drag-drop'),require('ngx-tethys/icon'),require('@angular/forms'),require('ngx-tethys/switch'),require('ngx-tethys/empty'),require('ngx-tethys/pagination'),exports, require('ngx-tethys/core'), require('ngx-tethys/util'), require('rxjs'), require('rxjs/operators'), require('@angular/cdk/drag-drop'), require('@angular/cdk/overlay'), require('@angular/common'), require('@angular/core'), require('@angular/forms'), require('ngx-tethys/pagination'), require('ngx-tethys/switch'), require('ngx-tethys/loading'), require('ngx-tethys/empty'), require('ngx-tethys/shared'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/table', ['@angular/core','@angular/cdk/overlay','ngx-tethys/core','@angular/common','ngx-tethys/loading','ngx-tethys/shared','@angular/cdk/drag-drop','ngx-tethys/icon','@angular/forms','ngx-tethys/switch','ngx-tethys/empty','ngx-tethys/pagination','exports', 'ngx-tethys/core', 'ngx-tethys/util', 'rxjs', 'rxjs/operators', '@angular/cdk/drag-drop', '@angular/cdk/overlay', '@angular/common', '@angular/core', '@angular/forms', 'ngx-tethys/pagination', 'ngx-tethys/switch', 'ngx-tethys/loading', 'ngx-tethys/empty', 'ngx-tethys/shared', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.cdk.overlay,global.ngxTethys.core,global.ng.common,global.ngxTethys.loading,global.ngxTethys.shared,global.ng.cdk.dragDrop,global.ngxTethys.icon,global.ng.forms,global.ngxTethys.switch,global.ngxTethys.empty,global.ngxTethys.pagination,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].table = {}), global['ngx-tethys'].core, global['ngx-tethys'].util, global.rxjs, global.rxjs.operators, global.ng.cdk.dragDrop, global.ng.cdk.overlay, global.ng.common, global.ng.core, global.ng.forms, global['ngx-tethys'].pagination, global['ngx-tethys'].switch, global['ngx-tethys'].loading, global['ngx-tethys'].empty, global['ngx-tethys'].shared, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,ɵngcc10,ɵngcc11,exports, core$1, util, rxjs, operators, dragDrop, overlay, common, core, forms, pagination, _switch, loading, empty, shared, icon) { 
var _c0 = ["header"];
var _c1 = ["cell"];
var _c2 = ["*"];
var _c3 = ["empty"];
var _c4 = ["group"];
var _c5 = ["table"];
function ThyTableComponent_col_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "col", 8);
} if (rf & 2) {
    var column_r7 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("width", column_r7.width);
} }
function ThyTableComponent_thead_4_th_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", column_r9.title, " ");
} }
function ThyTableComponent_thead_4_th_2_ng_container_2_ng_template_1_Template(rf, ctx) { }
var _c6 = function (a0) { return { $implicit: a0 }; };
function ThyTableComponent_thead_4_th_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_thead_4_th_2_ng_container_2_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", column_r9.headerTemplateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c6, column_r9));
} }
function ThyTableComponent_thead_4_th_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 10);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_thead_4_th_2_ng_container_1_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_thead_4_th_2_ng_container_2_Template, 2, 4, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var column_r9 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngClass", column_r9.headerClassName);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !column_r9.headerTemplateRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r9.headerTemplateRef);
} }
function ThyTableComponent_thead_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "thead");
    ɵngcc0.ɵɵelementStart(1, "tr");
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_thead_4_th_2_Template, 3, 3, "th", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.columns);
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_ng_template_1_Template(rf, ctx) { }
var _c7 = function (a0, a1) { return { $implicit: a0, group: a1 }; };
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var group_r23 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r24 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r24.groupTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c7, group_r23.origin, group_r23.origin));
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_ng_template_1_Template(rf, ctx) { }
var _c8 = function (a0, a1) { return { row: a0, index: a1, level: 0 }; };
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var row_r29 = ctx.$implicit;
    var i_r30 = ctx.index;
    ɵngcc0.ɵɵnextContext(4);
    var _r18 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r18)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c8, row_r29, i_r30));
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_Template, 2, 5, "ng-container", 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var group_r23 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r25 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", group_r23.children)("ngForTrackBy", ctx_r25.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    var _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "tr", 16);
    ɵngcc0.ɵɵlistener("click", function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_Template_tr_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r34); var group_r23 = ctx.$implicit; var ctx_r33 = ɵngcc0.ɵɵnextContext(3); return ctx_r33.expandGroup(group_r23); });
    ɵngcc0.ɵɵelementStart(2, "td");
    ɵngcc0.ɵɵelementStart(3, "div", 17);
    ɵngcc0.ɵɵelement(4, "thy-icon", 18);
    ɵngcc0.ɵɵtemplate(5, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_Template, 2, 5, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var group_r23 = ctx.$implicit;
    var ctx_r22 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("colspan", ctx_r22.columns.length);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("thyIconName", group_r23.expand ? "angle-down" : "angle-right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r22.groupTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (group_r23.children == null ? null : group_r23.children.length) && group_r23.expand);
} }
function ThyTableComponent_tbody_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_Template, 7, 4, "ng-container", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r15.groups);
} }
function ThyTableComponent_tbody_5_ng_container_2_tr_1_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template(rf, ctx) { if (rf & 1) {
    var _r40 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 21);
    ɵngcc0.ɵɵlistener("cdkDragStarted", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_cdkDragStarted_0_listener() { ɵngcc0.ɵɵrestoreView(_r40); var ctx_r39 = ɵngcc0.ɵɵnextContext(3); return ctx_r39.onDragStarted(); })("mousedown", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_mousedown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); var ctx_r41 = ɵngcc0.ɵɵnextContext(3); return ctx_r41.draggableStopPropagation($event); })("touchstart", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_touchstart_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); var ctx_r42 = ɵngcc0.ɵɵnextContext(3); return ctx_r42.draggableStopPropagation($event); })("pointerdown", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_pointerdown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); var ctx_r43 = ɵngcc0.ɵɵnextContext(3); return ctx_r43.draggableStopPropagation($event); })("click", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); var row_r36 = ctx.$implicit; var ctx_r44 = ɵngcc0.ɵɵnextContext(3); return ctx_r44.onRowClick($event, row_r36); })("thyContextMenu", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_thyContextMenu_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); var row_r36 = ctx.$implicit; var ctx_r45 = ɵngcc0.ɵɵnextContext(3); return ctx_r45.onRowContextMenu($event, row_r36); });
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_2_tr_1_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var row_r36 = ctx.$implicit;
    var i_r37 = ctx.index;
    ɵngcc0.ɵɵnextContext(2);
    var _r20 = ɵngcc0.ɵɵreference(7);
    var ctx_r35 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDragData", row_r36)("ngClass", ctx_r35.renderRowClassName(row_r36, i_r37));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r20)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(4, _c8, row_r36, i_r37));
} }
function ThyTableComponent_tbody_5_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_2_tr_1_Template, 2, 7, "tr", 20);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r16 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r16.model)("ngForTrackBy", ctx_r16.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_container_3_ng_container_1_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_3_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var row_r47 = ctx.$implicit;
    var i_r48 = ctx.index;
    ɵngcc0.ɵɵnextContext(2);
    var _r18 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r18)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c8, row_r47, i_r48));
} }
function ThyTableComponent_tbody_5_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_3_ng_container_1_Template, 2, 5, "ng-container", 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r17 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r17.model)("ngForTrackBy", ctx_r17.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_template_4_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_ng_template_0_Template(rf, ctx) { }
var _c9 = function (a0, a1, a2) { return { row: a0, index: a1, level: a2 }; };
function ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_ng_template_0_Template, 0, 0, "ng-template", 11);
} if (rf & 2) {
    var child_r56 = ctx.$implicit;
    var j_r57 = ctx.index;
    var level_r52 = ɵngcc0.ɵɵnextContext(2).level;
    ɵngcc0.ɵɵnextContext();
    var _r18 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r18)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(2, _c9, child_r56, j_r57, level_r52 + 1));
} }
function ThyTableComponent_tbody_5_ng_template_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_Template, 1, 6, undefined, 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var row_r50 = ɵngcc0.ɵɵnextContext().row;
    var ctx_r54 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", row_r50[ctx_r54.thyChildrenKey])("ngForTrackBy", ctx_r54.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    var _r62 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 21);
    ɵngcc0.ɵɵlistener("cdkDragStarted", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_cdkDragStarted_0_listener() { ɵngcc0.ɵɵrestoreView(_r62); var ctx_r61 = ɵngcc0.ɵɵnextContext(2); return ctx_r61.onDragStarted(); })("mousedown", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_mousedown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); var ctx_r63 = ɵngcc0.ɵɵnextContext(2); return ctx_r63.draggableStopPropagation($event); })("touchstart", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_touchstart_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); var ctx_r64 = ɵngcc0.ɵɵnextContext(2); return ctx_r64.draggableStopPropagation($event); })("pointerdown", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_pointerdown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); var ctx_r65 = ɵngcc0.ɵɵnextContext(2); return ctx_r65.draggableStopPropagation($event); })("click", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); var row_r50 = ctx.row; var ctx_r66 = ɵngcc0.ɵɵnextContext(2); return ctx_r66.onRowClick($event, row_r50); })("thyContextMenu", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_thyContextMenu_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); var row_r50 = ctx.row; var ctx_r67 = ɵngcc0.ɵɵnextContext(2); return ctx_r67.onRowContextMenu($event, row_r50); });
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_4_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_4_ng_container_2_Template, 2, 2, "ng-container", 3);
} if (rf & 2) {
    var row_r50 = ctx.row;
    var i_r51 = ctx.index;
    var level_r52 = ctx.level;
    ɵngcc0.ɵɵnextContext();
    var _r20 = ɵngcc0.ɵɵreference(7);
    var ctx_r19 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDragData", row_r50)("ngClass", ctx_r19.renderRowClassName(row_r50, i_r51));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r20)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(5, _c9, row_r50, i_r51, level_r52 + 1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r19.mode === "tree" && ctx_r19.isExpanded(row_r50));
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 24);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    var _r81 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 26);
    ɵngcc0.ɵɵlistener("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_thy_icon_1_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r81); var row_r68 = ɵngcc0.ɵɵnextContext(3).row; var ctx_r79 = ɵngcc0.ɵɵnextContext(2); return ctx_r79.expandChildren($event, row_r68); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r82 = ɵngcc0.ɵɵnextContext(3);
    var row_r68 = ctx_r82.row;
    var level_r70 = ctx_r82.level;
    var ctx_r77 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵstyleProp("visibility", ctx_r77.showExpand(row_r68) ? "visible" : "hidden")("left", ctx_r77.iconIndentComputed(level_r70), "px");
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r77.isExpanded(row_r68) ? "angle-down" : "angle-right");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_ng_template_2_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_thy_icon_1_Template, 1, 5, "thy-icon", 25);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_ng_template_2_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext().$implicit;
    var row_r68 = ɵngcc0.ɵɵnextContext().row;
    var ctx_r75 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r75.mode === "tree" && column_r72.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", column_r72.templateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c6, row_r68));
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    var _r96 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 26);
    ɵngcc0.ɵɵlistener("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_thy_icon_1_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r96); var row_r68 = ɵngcc0.ɵɵnextContext(4).row; var ctx_r94 = ɵngcc0.ɵɵnextContext(2); return ctx_r94.expandChildren($event, row_r68); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r97 = ɵngcc0.ɵɵnextContext(4);
    var row_r68 = ctx_r97.row;
    var level_r70 = ctx_r97.level;
    var ctx_r90 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵstyleProp("visibility", ctx_r90.showExpand(row_r68) ? "visible" : "hidden")("margin-left", ctx_r90.iconIndentComputed(level_r70), "px");
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r90.isExpanded(row_r68) ? "angle-down" : "angle-right");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext(3).$implicit;
    var row_r68 = ɵngcc0.ɵɵnextContext().row;
    var ctx_r91 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r91.getModelValue(row_r68, column_r72.model), " ");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 30);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", column_r72.defaultText, " ");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_thy_icon_1_Template, 1, 5, "thy-icon", 27);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_container_2_Template, 2, 1, "ng-container", 28);
    ɵngcc0.ɵɵpipe(3, "isValidModelValue");
    ɵngcc0.ɵɵtemplate(4, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_template_4_Template, 2, 1, "ng-template", null, 29, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var _r92 = ɵngcc0.ɵɵreference(5);
    var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var row_r68 = ɵngcc0.ɵɵnextContext().row;
    var ctx_r85 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r85.mode === "tree" && column_r72.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(3, 3, ctx_r85.getModelValue(row_r68, column_r72.model)))("ngIfElse", _r92);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var i_r69 = ɵngcc0.ɵɵnextContext(3).index;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", i_r69 + 1, " ");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    var _r106 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "input", 31);
    ɵngcc0.ɵɵlistener("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r106); var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; var row_r68 = ɵngcc0.ɵɵnextContext().row; return (row_r68[column_r72.key] = $event); })("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template_input_ngModelChange_1_listener() { ɵngcc0.ɵɵrestoreView(_r106); var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; var row_r68 = ɵngcc0.ɵɵnextContext().row; var ctx_r108 = ɵngcc0.ɵɵnextContext(2); return ctx_r108.onCheckboxChange(row_r68, column_r72); })("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template_input_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r106); var ctx_r111 = ɵngcc0.ɵɵnextContext(5); return ctx_r111.onStopPropagation($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var row_r68 = ɵngcc0.ɵɵnextContext().row;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", row_r68[column_r72.key])("disabled", column_r72.disabled);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    var _r115 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "input", 32);
    ɵngcc0.ɵɵlistener("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r115); var ctx_r114 = ɵngcc0.ɵɵnextContext(5); return ctx_r114.selectedRadioRow = $event; })("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template_input_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r115); var ctx_r116 = ɵngcc0.ɵɵnextContext(5); return ctx_r116.onStopPropagation($event); })("change", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template_input_change_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r115); var row_r68 = ɵngcc0.ɵɵnextContext(3).row; var ctx_r117 = ɵngcc0.ɵɵnextContext(2); return ctx_r117.onRadioSelectChange($event, row_r68); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var row_r68 = ɵngcc0.ɵɵnextContext().row;
    var ctx_r88 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r88.selectedRadioRow)("value", row_r68)("disabled", column_r72.disabled);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    var _r123 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "thy-switch", 33);
    ɵngcc0.ɵɵlistener("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template_thy_switch_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r123); var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; var row_r68 = ɵngcc0.ɵɵnextContext().row; return (row_r68[column_r72.key] = $event); })("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template_thy_switch_ngModelChange_1_listener() { ɵngcc0.ɵɵrestoreView(_r123); var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; var row_r68 = ɵngcc0.ɵɵnextContext().row; var ctx_r125 = ɵngcc0.ɵɵnextContext(2); return ctx_r125.onModelChange(row_r68, column_r72); })("thyChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template_thy_switch_thyChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r123); var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; var row_r68 = ɵngcc0.ɵɵnextContext().row; var ctx_r128 = ɵngcc0.ɵɵnextContext(2); return ctx_r128.onSwitchChange($event, row_r68, column_r72); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var row_r68 = ɵngcc0.ɵɵnextContext().row;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", row_r68[column_r72.key])("disabled", column_r72.disabled);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_Template, 6, 5, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_2_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(4, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template, 2, 3, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(5, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r72 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r76 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.model && !column_r72.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.index);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.checkbox);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.radio);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.switch);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 23);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_1_Template, 2, 0, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_Template, 3, 5, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_Template, 6, 5, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var column_r72 = ctx.$implicit;
    var j_r73 = ctx.index;
    var level_r70 = ɵngcc0.ɵɵnextContext().level;
    var ctx_r71 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("thy-table-checkbox-column", column_r72.type === ctx_r71.customType.checkbox);
    ɵngcc0.ɵɵproperty("ngClass", column_r72.className)("ngStyle", ctx_r71.mode === "tree" && column_r72.expand ? ctx_r71.tdIndentComputed(level_r70) : null);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", j_r73 === 0 && ctx_r71.draggable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r71.isTemplateRef(column_r72.templateRef));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r71.isTemplateRef(column_r72.templateRef));
} }
function ThyTableComponent_tbody_5_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTableComponent_tbody_5_ng_template_6_td_0_Template, 4, 7, "td", 22);
} if (rf & 2) {
    var ctx_r21 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r21.columns);
} }
function ThyTableComponent_tbody_5_Template(rf, ctx) { if (rf & 1) {
    var _r136 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tbody", 12);
    ɵngcc0.ɵɵlistener("cdkDropListDropped", function ThyTableComponent_tbody_5_Template_tbody_cdkDropListDropped_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r136); var ctx_r135 = ɵngcc0.ɵɵnextContext(); return ctx_r135.onDragDropped($event); });
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_container_2_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTableComponent_tbody_5_ng_container_3_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(4, ThyTableComponent_tbody_5_ng_template_4_Template, 3, 9, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(6, ThyTableComponent_tbody_5_ng_template_6_Template, 1, 1, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDropListDisabled", !ctx_r3.draggable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "group");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "list");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "tree");
} }
function ThyTableComponent_div_6_ng_container_1_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_div_6_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_div_6_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 36);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r137 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r137.emptyTemplate);
} }
function ThyTableComponent_div_6_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-empty", 37);
} if (rf & 2) {
    var ctx_r139 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyMessage", ctx_r139.emptyOptions.message)("thyTranslationKey", ctx_r139.emptyOptions.translationKey)("thyTranslationValues", ctx_r139.emptyOptions.translationValues)("thyEntityName", ctx_r139.emptyOptions.entityName)("thyEntityNameTranslateKey", ctx_r139.emptyOptions.entityNameTranslateKey)("thyIconName", ctx_r139.emptyOptions.iconName)("thySize", ctx_r139.emptyOptions.size)("thyMarginTop", ctx_r139.emptyOptions.marginTop)("thyTopAuto", ctx_r139.emptyOptions.topAuto)("thyContainer", ctx_r139.emptyOptions.container);
} }
function ThyTableComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 34);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_div_6_ng_container_1_Template, 2, 1, "ng-container", 28);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_div_6_ng_template_2_Template, 1, 10, "ng-template", null, 35, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r138 = ɵngcc0.ɵɵreference(3);
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.emptyTemplate)("ngIfElse", _r138);
} }
function ThyTableComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    var _r142 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵelementStart(1, "thy-pagination", 39);
    ɵngcc0.ɵɵlistener("thyPageChanged", function ThyTableComponent_div_8_Template_thy_pagination_thyPageChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r142); var ctx_r141 = ɵngcc0.ɵɵnextContext(); return ctx_r141.onPageChange($event); })("thyPageIndexChange", function ThyTableComponent_div_8_Template_thy_pagination_thyPageIndexChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r142); var ctx_r143 = ɵngcc0.ɵɵnextContext(); return ctx_r143.onPageIndexChange($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("thy-table-footer-has-padding", ctx_r5.theme === "default");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyPageIndex", ctx_r5.pagination.index)("thyTotal", ctx_r5.pagination.total)("thyPageSize", ctx_r5.pagination.size)("thyShowTotal", ctx_r5.showTotal);
} }
function ThyTableComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
var _c10 = function (a0) { return [a0]; };
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * Injection token used to provide the parent component to options.
     */
    var THY_TABLE_COLUMN_PARENT_COMPONENT = new core.InjectionToken('THY_TABLE_COLUMN_PARENT_COMPONENT');
    /**
     * @deprecated THY_GRID_COLUMN_PARENT_COMPONENT is deprecated, please use THY_TABLE_COLUMN_PARENT_COMPONENT
     */
    var THY_GRID_COLUMN_PARENT_COMPONENT = THY_TABLE_COLUMN_PARENT_COMPONENT;
    var ThyTableColumnComponent = /** @class */ (function () {
        function ThyTableColumnComponent(el, parent) {
            this.el = el;
            this.parent = parent;
            this.thyModelKey = '';
            this.thyTitle = '';
            this.thyType = '';
            this.thyWidth = '';
            this.thyClassName = '';
            this.thyHeaderClassName = '';
            this.thyDisabled = false;
            this.thyDefaultText = '';
            this.thyExpand = false;
            this.selections = [];
            this._firstChange = true;
            this.expand = false;
        }
        Object.defineProperty(ThyTableColumnComponent.prototype, "thySelections", {
            set: function (value) {
                if (value) {
                    if (Array.isArray(value)) {
                        this.selections = value;
                    }
                    else {
                        this.selections = [value];
                    }
                    if (!this._firstChange) {
                        this.parent.updateColumnSelections(this.key, this.selections);
                    }
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableColumnComponent.prototype, "templateRef", {
            set: function (value) {
                if (value) {
                    if (!this.headerTemplateRef && !this.cellTemplateRef) {
                        this.cellTemplateRef = value;
                    }
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyTableColumnComponent.prototype.ngOnInit = function () {
            this.key = this._generateKey();
            this.model = this.thyModelKey;
            this.title = this.thyTitle;
            this.type = this.thyType;
            this.width = this.thyWidth;
            this.className = this.thyClassName;
            this.headerClassName = this.thyHeaderClassName;
            this.disabled = this.thyDisabled;
            this.defaultText = this.thyDefaultText;
            this._firstChange = false;
            this.expand = this.thyExpand;
        };
        ThyTableColumnComponent.prototype._generateKey = function () {
            return ('[$$column]' +
                Math.random()
                    .toString(16)
                    .substr(2, 8));
        };
ThyTableColumnComponent.ɵfac = function ThyTableColumnComponent_Factory(t) { return new (t || ThyTableColumnComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(THY_GRID_COLUMN_PARENT_COMPONENT, 8)); };
ThyTableColumnComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTableColumnComponent, selectors: [["thy-grid-column"], ["thy-table-column"]], contentQueries: function ThyTableColumnComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, core.TemplateRef, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.cellTemplateRef = _t.first);
    } }, inputs: { thyModelKey: "thyModelKey", thyTitle: "thyTitle", thyType: "thyType", thyWidth: "thyWidth", thyClassName: "thyClassName", thyHeaderClassName: "thyHeaderClassName", thyDisabled: "thyDisabled", thyDefaultText: "thyDefaultText", thyExpand: "thyExpand", thySelections: "thySelections" }, ngContentSelectors: _c2, decls: 1, vars: 0, template: function ThyTableColumnComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableColumnComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-grid-column,thy-table-column',
                template: '<ng-content></ng-content>',
                encapsulation: core.ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: undefined, decorators: [{
                type: core.Optional
            }, {
                type: core.Inject,
                args: [THY_GRID_COLUMN_PARENT_COMPONENT]
            }] }]; }, { thyModelKey: [{
            type: core.Input
        }], thyTitle: [{
            type: core.Input
        }], thyType: [{
            type: core.Input
        }], thyWidth: [{
            type: core.Input
        }], thyClassName: [{
            type: core.Input
        }], thyHeaderClassName: [{
            type: core.Input
        }], thyDisabled: [{
            type: core.Input
        }], thyDefaultText: [{
            type: core.Input
        }], thyExpand: [{
            type: core.Input
        }], thySelections: [{
            type: core.Input
        }], templateRef: [{
            type: core.ContentChild,
            args: [core.TemplateRef, { static: true }]
        }], headerTemplateRef: [{
            type: core.ContentChild,
            args: ['header', { static: true }]
        }], cellTemplateRef: [{
            type: core.ContentChild,
            args: ['cell', { static: true }]
        }] }); })();
        return ThyTableColumnComponent;
    }());
    ThyTableColumnComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [THY_GRID_COLUMN_PARENT_COMPONENT,] }] }
    ]; };
    ThyTableColumnComponent.propDecorators = {
        thyModelKey: [{ type: core.Input }],
        thyTitle: [{ type: core.Input }],
        thyType: [{ type: core.Input }],
        thyWidth: [{ type: core.Input }],
        thyClassName: [{ type: core.Input }],
        thyHeaderClassName: [{ type: core.Input }],
        thyDisabled: [{ type: core.Input }],
        thySelections: [{ type: core.Input }],
        thyDefaultText: [{ type: core.Input }],
        thyExpand: [{ type: core.Input }],
        headerTemplateRef: [{ type: core.ContentChild, args: ['header', { static: true },] }],
        cellTemplateRef: [{ type: core.ContentChild, args: ['cell', { static: true },] }],
        templateRef: [{ type: core.ContentChild, args: [core.TemplateRef, { static: true },] }]
    };
    /**
     * @deprecated ThyTableColumnComponent is deprecated, please use ThyTableColumnComponent
     */
    var ThyGridColumnComponent = ThyTableColumnComponent;

    var tableThemeMap = {
        default: 'table-default',
        bordered: 'table-bordered'
    };
    var tableSizeMap = {
        sm: 'table-sm'
    };
    var customType = {
        index: 'index',
        checkbox: 'checkbox',
        radio: 'radio',
        switch: 'switch'
    };
    var _MixinBase = core$1.mixinUnsubscribe(core$1.MixinBase);
    var ThyTableComponent = /** @class */ (function (_super) {
        __extends(ThyTableComponent, _super);
        function ThyTableComponent(elementRef, _differs, viewportRuler, updateHostClassService, document) {
            var _this = _super.call(this) || this;
            _this.elementRef = elementRef;
            _this._differs = _differs;
            _this.viewportRuler = viewportRuler;
            _this.updateHostClassService = updateHostClassService;
            _this.document = document;
            _this.customType = customType;
            _this.model = [];
            _this.groups = [];
            _this.rowKey = '_id';
            _this.mode = 'list';
            _this.columns = [];
            _this.theme = 'default';
            _this.className = '';
            _this.size = 'default';
            _this.loadingDone = true;
            _this.emptyOptions = {};
            _this.draggable = false;
            _this.selectedRadioRow = null;
            _this.pagination = { index: 1, size: 20, total: 0 };
            _this.wholeRowSelect = false;
            _this._filter = null;
            _this.initialized = false;
            _this._oldThyClassName = '';
            _this.thyShowHeader = true;
            _this.showTotal = false;
            _this.thyIndent = 20;
            _this.thyChildrenKey = 'children';
            _this.thyOnSwitchChange = new core.EventEmitter();
            _this.thyOnPageChange = new core.EventEmitter();
            _this.thyOnPageIndexChange = new core.EventEmitter();
            _this.thyOnMultiSelectChange = new core.EventEmitter();
            _this.thyOnRadioSelectChange = new core.EventEmitter();
            _this.thyOnDraggableChange = new core.EventEmitter();
            _this.thyOnRowClick = new core.EventEmitter();
            _this.thyOnRowContextMenu = new core.EventEmitter();
            _this.isTableClass = true;
            // 数据的折叠展开状态
            _this.expandStatusMap = {};
            _this._bindTrackFn();
            return _this;
        }
        Object.defineProperty(ThyTableComponent.prototype, "thyMode", {
            set: function (value) {
                this.mode = value || this.mode;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyGroupBy", {
            set: function (value) {
                this.groupBy = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyRowKey", {
            set: function (value) {
                this.rowKey = value || this.rowKey;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyGroups", {
            set: function (value) {
                if (this.mode === 'group') {
                    this.buildGroups(value);
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyModel", {
            set: function (value) {
                this.model = value || [];
                this._diff = this._differs.find(this.model).create();
                this._initializeDataModel();
                if (this.mode === 'group') {
                    this.buildModel();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyTheme", {
            set: function (value) {
                this.theme = value || this.theme;
                this._setClass();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thySize", {
            set: function (value) {
                this.size = value || this.size;
                this._setClass();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyClassName", {
            set: function (value) {
                var _this = this;
                var list = this.className.split(' ').filter(function (a) { return a.trim(); });
                var index = list.findIndex(function (item) { return item === _this._oldThyClassName; });
                if (index !== -1) {
                    list.splice(index, 1, value);
                }
                else {
                    list.push(value);
                }
                this._oldThyClassName = value;
                this.className = list.join(' ');
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyRowClassName", {
            set: function (value) {
                this.rowClassName = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyLoadingDone", {
            set: function (value) {
                this.loadingDone = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyLoadingText", {
            set: function (value) {
                this.loadingText = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyEmptyOptions", {
            set: function (value) {
                this.emptyOptions = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyDraggable", {
            set: function (value) {
                this.draggable = util.coerceBooleanProperty(value);
                if (this.mode !== 'list' && this.draggable) {
                    throw new Error('Only list mode sorting is supported');
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyFilter", {
            set: function (value) {
                this._filter = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyPageIndex", {
            set: function (value) {
                this.pagination.index = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyPageSize", {
            set: function (value) {
                this.pagination.size = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyPageTotal", {
            set: function (value) {
                this.pagination.total = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "thyWholeRowSelect", {
            set: function (value) {
                if (value) {
                    this.className += ' table-hover';
                }
                this.wholeRowSelect = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTableComponent.prototype, "listOfColumnComponents", {
            set: function (components) {
                if (components) {
                    this._listOfColumnComponents = components;
                    this._initializeColumns();
                    this._initializeDataModel();
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyTableComponent.prototype._getSelectionKeys = function (selections) {
            var _this = this;
            return selections.map(function (item) {
                if (typeof item === 'number' || typeof item === 'string') {
                    return item;
                }
                else {
                    return item[_this.rowKey];
                }
            });
        };
        ThyTableComponent.prototype._initializeColumns = function () {
            var _this = this;
            var components = this._listOfColumnComponents ? this._listOfColumnComponents.toArray() : [];
            var hasExpand = components.some(function (item) { return item.expand === true; });
            this.columns = components.map(function (component, i) {
                var selections = _this._getSelectionKeys(component.selections);
                return {
                    key: component.key,
                    model: component.model,
                    title: component.title,
                    type: component.type,
                    selections: selections,
                    width: component.width,
                    className: component.className,
                    headerClassName: component.headerClassName,
                    disabled: component.disabled,
                    defaultText: component.defaultText,
                    expand: hasExpand ? component.expand : i === 0,
                    templateRef: component.cellTemplateRef,
                    headerTemplateRef: component.headerTemplateRef
                };
            });
        };
        ThyTableComponent.prototype._initializeDataModel = function () {
            var _this = this;
            this.model.forEach(function (row) {
                _this.columns.forEach(function (column) {
                    _this._initialSelections(row, column);
                    _this._initialCustomModelValue(row, column);
                });
            });
        };
        ThyTableComponent.prototype._initialSelections = function (row, column) {
            if (column.selections) {
                if (column.type === 'checkbox') {
                    row[column.key] = column.selections.includes(row[this.rowKey]);
                    this.onModelChange(row, column);
                }
                if (column.type === 'radio') {
                    if (column.selections.includes(row[this.rowKey])) {
                        this.selectedRadioRow = row;
                    }
                }
            }
        };
        ThyTableComponent.prototype._initialCustomModelValue = function (row, column) {
            if (column.type === customType.switch) {
                row[column.key] = util.get(row, column.model);
            }
        };
        ThyTableComponent.prototype._refreshCustomModelValue = function (row) {
            var _this = this;
            this.columns.forEach(function (column) {
                _this._initialCustomModelValue(row, column);
            });
        };
        ThyTableComponent.prototype._applyDiffChanges = function (changes) {
            var _this = this;
            if (changes) {
                changes.forEachAddedItem(function (record) {
                    _this._refreshCustomModelValue(record.item);
                });
            }
        };
        ThyTableComponent.prototype._bindTrackFn = function () {
            this.trackByFn = function (index, row) {
                return row && this.rowKey ? row[this.rowKey] : index;
            }.bind(this);
        };
        ThyTableComponent.prototype._destroyInvalidAttribute = function () {
            this.model.forEach(function (row) {
                for (var key in row) {
                    if (key.includes('[$$column]')) {
                        delete row[key];
                    }
                }
            });
        };
        ThyTableComponent.prototype._setClass = function (first) {
            if (first === void 0) { first = false; }
            if (!first && !this.initialized) {
                return;
            }
            var classNames = [];
            if (tableSizeMap[this.size]) {
                classNames.push(tableSizeMap[this.size]);
            }
            if (tableThemeMap[this.theme]) {
                classNames.push(tableThemeMap[this.theme]);
            }
            if (tableSizeMap[this.size] === tableSizeMap['sm'] && tableThemeMap[this.theme] === tableThemeMap['default']) {
                classNames.push('table-default-sm-bottom-padding');
            }
            this.updateHostClassService.updateClass(classNames);
        };
        ThyTableComponent.prototype._filterModel = function () {
            if (this.model && this.model.length > 0) {
                if (this._filter) {
                }
            }
        };
        ThyTableComponent.prototype.updateColumnSelections = function (key, selections) {
            var _this = this;
            var column = this.columns.find(function (item) { return item.key === key; });
            column.selections = this._getSelectionKeys(selections);
            this.model.forEach(function (row) {
                _this._initialSelections(row, column);
            });
        };
        ThyTableComponent.prototype.isTemplateRef = function (ref) {
            return ref instanceof core.TemplateRef;
        };
        ThyTableComponent.prototype.getModelValue = function (row, path) {
            return util.get(row, path);
        };
        ThyTableComponent.prototype.renderRowClassName = function (row, index) {
            if (!this.rowClassName) {
                return null;
            }
            if (util.isString(this.rowClassName)) {
                return this.rowClassName;
            }
            else {
                return this.rowClassName(row, index);
            }
        };
        ThyTableComponent.prototype.onModelChange = function (row, column) {
            if (column.model) {
                util.set(row, column.model, row[column.key]);
            }
        };
        ThyTableComponent.prototype.onStopPropagation = function (event) {
            if (this.wholeRowSelect) {
                event.stopPropagation();
            }
        };
        ThyTableComponent.prototype.onPageChange = function (event) {
            this.thyOnPageChange.emit(event);
        };
        ThyTableComponent.prototype.onPageIndexChange = function (event) {
            this.thyOnPageIndexChange.emit(event);
        };
        ThyTableComponent.prototype.onCheckboxChange = function (row, column) {
            this.onModelChange(row, column);
            this.onMultiSelectChange(null, row, column);
        };
        ThyTableComponent.prototype.onMultiSelectChange = function (event, row, column) {
            var rows = this.model.filter(function (item) {
                return item[column.key];
            });
            var multiSelectEvent = {
                event: event,
                row: row,
                rows: rows
            };
            this.thyOnMultiSelectChange.emit(multiSelectEvent);
        };
        ThyTableComponent.prototype.onRadioSelectChange = function (event, row) {
            var radioSelectEvent = {
                event: event,
                row: row
            };
            this.thyOnRadioSelectChange.emit(radioSelectEvent);
        };
        ThyTableComponent.prototype.onSwitchChange = function (event, row, column) {
            var switchEvent = {
                event: event,
                row: row,
                refresh: function (value) {
                    value = value || row;
                    setTimeout(function () {
                        value[column.key] = util.get(value, column.model);
                    });
                }
            };
            this.thyOnSwitchChange.emit(switchEvent);
        };
        ThyTableComponent.prototype.showExpand = function (row) {
            return row[this.thyChildrenKey] && row[this.thyChildrenKey].length > 0;
        };
        ThyTableComponent.prototype.isExpanded = function (row) {
            return this.expandStatusMap[row[this.rowKey]];
        };
        ThyTableComponent.prototype.iconIndentComputed = function (level) {
            if (this.mode === 'tree') {
                return level * this.thyIndent - 10;
            }
        };
        ThyTableComponent.prototype.tdIndentComputed = function (level) {
            return {
                position: 'relative',
                paddingLeft: (level + 1) * this.thyIndent - 10 + "px"
            };
        };
        ThyTableComponent.prototype.expandChildren = function (event, row) {
            event.stopPropagation();
            if (this.isExpanded(row)) {
                this.expandStatusMap[row[this.rowKey]] = false;
            }
            else {
                this.expandStatusMap[row[this.rowKey]] = true;
            }
        };
        ThyTableComponent.prototype.onDragStarted = function () {
            var _this = this;
            setTimeout(function () {
                var preview = _this.document.getElementsByClassName('cdk-drag-preview')[0];
                if (preview) {
                    preview.classList.add('thy-table-drag-preview');
                }
            });
        };
        ThyTableComponent.prototype.onDragDropped = function (event) {
            var dragEvent = {
                model: event.item,
                models: this.model,
                oldIndex: event.previousIndex,
                newIndex: event.currentIndex
            };
            dragDrop.moveItemInArray(this.model, event.previousIndex, event.currentIndex);
            this.thyOnDraggableChange.emit(dragEvent);
        };
        ThyTableComponent.prototype.onRowClick = function (event, row) {
            if (this.wholeRowSelect) {
                var column = this.columns.find(function (item) {
                    return item.type === customType.checkbox || item.type === customType.radio;
                });
                if (!column.disabled) {
                    if (column.type === customType.checkbox) {
                        row[column.key] = !row[column.key];
                        this.onModelChange(row, column);
                        this.onMultiSelectChange(event, row, column);
                    }
                    if (column.type === customType.radio) {
                        this.selectedRadioRow = row;
                        this.onRadioSelectChange(event, row);
                    }
                }
            }
            var rowEvent = {
                event: event,
                row: row
            };
            this.thyOnRowClick.emit(rowEvent);
        };
        ThyTableComponent.prototype.onRowContextMenu = function (event, row) {
            var contextMenuEvent = {
                event: event,
                row: row
            };
            this.thyOnRowContextMenu.emit(contextMenuEvent);
        };
        ThyTableComponent.prototype.draggableStopPropagation = function (event) {
            if (!this.draggable) {
                event.stopPropagation();
            }
        };
        ThyTableComponent.prototype._refreshColumns = function () {
            var components = this._listOfColumnComponents ? this._listOfColumnComponents.toArray() : [];
            var _columns = components.map(function (component) {
                return {
                    width: component.width,
                    className: component.className
                };
            });
            this.columns.forEach(function (n, i) {
                Object.assign(n, _columns[i]);
            });
        };
        ThyTableComponent.prototype.buildGroups = function (originGroups) {
            var _this = this;
            var collapsedIds = this.groups.filter(function (group) { return !group.expand; }).map(function (group) { return group.id; });
            this.groups = [];
            originGroups.forEach(function (origin) {
                var group = { id: origin[_this.rowKey], children: [], origin: origin };
                group.expand = !collapsedIds.includes(group.id);
                _this.groups.push(group);
            });
        };
        ThyTableComponent.prototype.buildModel = function () {
            var _this = this;
            var groupsMap = util.keyBy(this.groups, 'id');
            this.model.forEach(function (row) {
                var group = groupsMap[row[_this.groupBy]];
                if (group) {
                    group.children.push(row);
                }
            });
        };
        ThyTableComponent.prototype.expandGroup = function (group) {
            group.expand = !group.expand;
        };
        ThyTableComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.updateHostClassService.initializeElement(this.tableElementRef.nativeElement);
            this._setClass(true);
            this.initialized = true;
            rxjs.merge(this.viewportRuler.change(200), rxjs.of(null).pipe(operators.delay(200)))
                .pipe(operators.takeUntil(this.ngUnsubscribe$))
                .subscribe(function () {
                _this._refreshColumns();
            });
        };
        ThyTableComponent.prototype.ngOnChanges = function (simpleChangs) {
            var modeChange = simpleChangs.thyMode;
            var thyGroupsChange = simpleChangs.thyGroups;
            var isGroupMode = modeChange && modeChange.currentValue === 'group';
            if (isGroupMode && thyGroupsChange && thyGroupsChange.firstChange) {
                this.buildGroups(thyGroupsChange.currentValue);
                this.buildModel();
            }
        };
        ThyTableComponent.prototype.ngDoCheck = function () {
            if (this._diff) {
                var changes = this._diff.diff(this.model);
                this._applyDiffChanges(changes);
            }
        };
        ThyTableComponent.prototype.ngOnDestroy = function () {
            _super.prototype.ngOnDestroy.call(this);
            this._destroyInvalidAttribute();
        };
ThyTableComponent.ɵfac = function ThyTableComponent_Factory(t) { return new (t || ThyTableComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.IterableDiffers), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ViewportRuler), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(common.DOCUMENT)); };
ThyTableComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTableComponent, selectors: [["thy-grid"], ["thy-table"]], contentQueries: function ThyTableComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c3, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c4, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyTableColumnComponent, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.emptyTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.groupTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.listOfColumnComponents = _t);
    } }, viewQuery: function ThyTableComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c5, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tableElementRef = _t.first);
    } }, hostVars: 4, hostBindings: function ThyTableComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-table-hover-display-operation", ctx.thyHoverDisplayOperation)("thy-table", ctx.isTableClass);
    } }, inputs: { thyMode: "thyMode", thyGroupBy: "thyGroupBy", thyRowKey: "thyRowKey", thyGroups: "thyGroups", thyModel: "thyModel", thyTheme: "thyTheme", thySize: "thySize", thyClassName: "thyClassName", thyRowClassName: "thyRowClassName", thyLoadingDone: "thyLoadingDone", thyLoadingText: "thyLoadingText", thyEmptyOptions: "thyEmptyOptions", thyDraggable: "thyDraggable", thyFilter: "thyFilter", thyPageIndex: "thyPageIndex", thyPageSize: "thyPageSize", thyPageTotal: "thyPageTotal", thyWholeRowSelect: "thyWholeRowSelect", thyShowHeader: "thyShowHeader", showTotal: ["thyShowTotal", "showTotal"], thyIndent: "thyIndent", thyChildrenKey: "thyChildrenKey", thyHoverDisplayOperation: "thyHoverDisplayOperation" }, outputs: { thyOnSwitchChange: "thyOnSwitchChange", thyOnPageChange: "thyOnPageChange", thyOnPageIndexChange: "thyOnPageIndexChange", thyOnMultiSelectChange: "thyOnMultiSelectChange", thyOnRadioSelectChange: "thyOnRadioSelectChange", thyOnDraggableChange: "thyOnDraggableChange", thyOnRowClick: "thyOnRowClick", thyOnRowContextMenu: "thyOnRowContextMenu" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: THY_TABLE_COLUMN_PARENT_COMPONENT,
                useExisting: ThyTableComponent
            },
            core$1.UpdateHostClassService
        ]), ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c2, decls: 10, vars: 14, consts: [[1, "table", 3, "ngClass"], ["table", ""], [3, "width", 4, "ngFor", "ngForOf"], [4, "ngIf"], ["thyDragDrop", "", "cdkDropList", "", 3, "cdkDropListDisabled", "cdkDropListDropped", 4, "ngIf"], ["class", "thy-table-empty clear-hover", 4, "ngIf"], [3, "thyDone", "thyTip"], ["class", "thy-table-footer", 3, "thy-table-footer-has-padding", 4, "ngIf"], [3, "width"], [3, "ngClass", 4, "ngFor", "ngForOf"], [3, "ngClass"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["thyDragDrop", "", "cdkDropList", "", 3, "cdkDropListDisabled", "cdkDropListDropped"], ["trTemplate", ""], ["tdsTemplate", ""], [4, "ngFor", "ngForOf"], [1, "thy-table-group", 3, "click"], [1, "thy-table-group-container"], [1, "expand-icon", 3, "thyIconName"], [4, "ngFor", "ngForOf", "ngForTrackBy"], ["class", "thy-sortable-item", "cdkDrag", "", 3, "cdkDragData", "ngClass", "cdkDragStarted", "mousedown", "touchstart", "pointerdown", "click", "thyContextMenu", 4, "ngFor", "ngForOf", "ngForTrackBy"], ["cdkDrag", "", 1, "thy-sortable-item", 3, "cdkDragData", "ngClass", "cdkDragStarted", "mousedown", "touchstart", "pointerdown", "click", "thyContextMenu"], [3, "ngClass", "thy-table-checkbox-column", "ngStyle", 4, "ngFor", "ngForOf"], [3, "ngClass", "ngStyle"], ["thyIconName", "drag", 1, "table-draggable-icon"], ["class", "tree-expand-icon cursor-pointer", 3, "thyIconName", "visibility", "left", "click", 4, "ngIf"], [1, "tree-expand-icon", "cursor-pointer", 3, "thyIconName", "click"], ["class", "tree-expand-icon cursor-pointer", 3, "thyIconName", "visibility", "marginLeft", "click", 4, "ngIf"], [4, "ngIf", "ngIfElse"], ["default", ""], [1, "text-desc"], ["type", "checkbox", 3, "ngModel", "disabled", "ngModelChange", "click"], ["type", "radio", 3, "ngModel", "value", "disabled", "ngModelChange", "click", "change"], [3, "ngModel", "disabled", "ngModelChange", "thyChange"], [1, "thy-table-empty", "clear-hover"], ["defaultEmpty", ""], [3, "ngTemplateOutlet"], [3, "thyMessage", "thyTranslationKey", "thyTranslationValues", "thyEntityName", "thyEntityNameTranslateKey", "thyIconName", "thySize", "thyMarginTop", "thyTopAuto", "thyContainer"], [1, "thy-table-footer"], [3, "thyPageIndex", "thyTotal", "thyPageSize", "thyShowTotal", "thyPageChanged", "thyPageIndexChange"]], template: function ThyTableComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "table", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "colgroup");
        ɵngcc0.ɵɵtemplate(3, ThyTableComponent_col_3_Template, 1, 1, "col", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, ThyTableComponent_thead_4_Template, 3, 1, "thead", 3);
        ɵngcc0.ɵɵtemplate(5, ThyTableComponent_tbody_5_Template, 8, 4, "tbody", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, ThyTableComponent_div_6_Template, 4, 2, "div", 5);
        ɵngcc0.ɵɵelement(7, "thy-loading", 6);
        ɵngcc0.ɵɵtemplate(8, ThyTableComponent_div_8_Template, 2, 6, "div", 7);
        ɵngcc0.ɵɵtemplate(9, ThyTableComponent_ng_template_9_Template, 1, 0, "ng-template");
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("table-draggable", ctx.draggable)("table-group", ctx.mode === "group");
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(12, _c10, ctx.className));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.columns);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowHeader);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingDone && ctx.model.length === 0 && ctx.groups.length === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thyDone", ctx.loadingDone)("thyTip", ctx.loadingText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.pagination.total > ctx.pagination.size);
    } }, directives: function () { return [ɵngcc3.NgClass, ɵngcc3.NgForOf, ɵngcc3.NgIf, ɵngcc4.ThyLoadingComponent, ɵngcc3.NgTemplateOutlet, ɵngcc5.ThyDragDropDirective, ɵngcc6.CdkDropList, ɵngcc7.ThyIconComponent, ɵngcc6.CdkDrag, ɵngcc5.ThyContextMenuDirective, ɵngcc3.NgStyle, ɵngcc8.CheckboxControlValueAccessor, ɵngcc8.NgControlStatus, ɵngcc8.NgModel, ɵngcc8.RadioControlValueAccessor, ɵngcc8.DefaultValueAccessor, ɵngcc9.ThySwitchComponent, ɵngcc10.ThyEmptyComponent, ɵngcc11.ThyPaginationComponent]; }, pipes: function () { return [TableIsValidModelValuePipe]; }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-grid,thy-table',
                template: "<table #table class=\"table\" [ngClass]=\"[className]\" [class.table-draggable]=\"draggable\" [class.table-group]=\"mode === 'group'\">\n  <colgroup>\n    <col *ngFor=\"let column of columns\" [width]=\"column.width\" />\n  </colgroup>\n\n  <thead *ngIf=\"thyShowHeader\">\n    <tr>\n      <th *ngFor=\"let column of columns\" [ngClass]=\"column.headerClassName\">\n        <ng-container *ngIf=\"!column.headerTemplateRef\">\n          {{ column.title }}\n        </ng-container>\n        <ng-container *ngIf=\"column.headerTemplateRef\">\n          <ng-template [ngTemplateOutlet]=\"column.headerTemplateRef\" [ngTemplateOutletContext]=\"{ $implicit: column }\"></ng-template>\n        </ng-container>\n      </th>\n    </tr>\n  </thead>\n\n  <tbody *ngIf=\"loadingDone\" thyDragDrop cdkDropList [cdkDropListDisabled]=\"!draggable\" (cdkDropListDropped)=\"onDragDropped($event)\">\n    <!-- group -->\n    <ng-container *ngIf=\"mode === 'group'\">\n      <ng-container *ngFor=\"let group of groups\">\n        <tr class=\"thy-table-group\" (click)=\"expandGroup(group)\">\n          <td [attr.colspan]=\"columns.length\">\n            <div class=\"thy-table-group-container\">\n              <thy-icon class=\"expand-icon\" [thyIconName]=\"group.expand ? 'angle-down' : 'angle-right'\"></thy-icon>\n              <ng-container *ngIf=\"groupTemplate\">\n                <ng-template\n                  [ngTemplateOutlet]=\"groupTemplate\"\n                  [ngTemplateOutletContext]=\"{ $implicit: group.origin, group: group.origin }\"\n                ></ng-template>\n              </ng-container>\n            </div>\n          </td>\n        </tr>\n        <ng-container *ngIf=\"group.children?.length && group.expand\">\n          <ng-container *ngFor=\"let row of group.children; trackBy: trackByFn; let i = index\">\n            <ng-template [ngTemplateOutlet]=\"trTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: 0 }\"></ng-template>\n          </ng-container>\n        </ng-container>\n      </ng-container>\n    </ng-container>\n\n    <!-- list\u6A21\u5F0F\u4E0B\u6CA1\u6709\u76F4\u63A5\u4F7F\u7528item\u6A21\u677F\uFF0C\u662F\u56E0\u4E3A\u4F7F\u7528\u4E86\u6A21\u677F\u5F71\u54CD\u4E86Angular\u7684\u4F9D\u8D56\u6811\uFF0C\u5BFC\u81F4CdkDragDrop\u6392\u5E8F\u5931\u6548\uFF0C\u76EE\u524D\u8FD9\u6837\u4FEE\u6539\u53EF\u4EE5\u6682\u65F6\u89E3\u51B3\u7EBF\u4E0A\u4E0D\u80FD\u6392\u5E8F\u7684\u95EE\u9898\uFF0CTree\u6A21\u5F0F\u7684\u6392\u5E8F\u6682\u65F6\u4E0D\u652F\u6301 -->\n    <ng-container *ngIf=\"mode === 'list'\">\n      <tr\n        *ngFor=\"let row of model; trackBy: trackByFn; let i = index\"\n        class=\"thy-sortable-item\"\n        cdkDrag\n        [cdkDragData]=\"row\"\n        (cdkDragStarted)=\"onDragStarted()\"\n        [ngClass]=\"renderRowClassName(row, i)\"\n        (mousedown)=\"draggableStopPropagation($event)\"\n        (touchstart)=\"draggableStopPropagation($event)\"\n        (pointerdown)=\"draggableStopPropagation($event)\"\n        (click)=\"onRowClick($event, row)\"\n        (thyContextMenu)=\"onRowContextMenu($event, row)\"\n      >\n        <ng-template [ngTemplateOutlet]=\"tdsTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: 0 }\"></ng-template>\n      </tr>\n    </ng-container>\n\n    <ng-container *ngIf=\"mode === 'tree'\">\n      <ng-container *ngFor=\"let row of model; trackBy: trackByFn; let i = index\">\n        <ng-template [ngTemplateOutlet]=\"trTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: 0 }\"></ng-template>\n      </ng-container>\n    </ng-container>\n\n    <ng-template #trTemplate let-row=\"row\" let-i=\"index\" let-level=\"level\">\n      <tr\n        class=\"thy-sortable-item\"\n        cdkDrag\n        [cdkDragData]=\"row\"\n        (cdkDragStarted)=\"onDragStarted()\"\n        [ngClass]=\"renderRowClassName(row, i)\"\n        (mousedown)=\"draggableStopPropagation($event)\"\n        (touchstart)=\"draggableStopPropagation($event)\"\n        (pointerdown)=\"draggableStopPropagation($event)\"\n        (click)=\"onRowClick($event, row)\"\n        (thyContextMenu)=\"onRowContextMenu($event, row)\"\n      >\n        <ng-template [ngTemplateOutlet]=\"tdsTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: level + 1 }\"></ng-template>\n      </tr>\n\n      <ng-container *ngIf=\"mode === 'tree' && isExpanded(row)\">\n        <ng-template\n          *ngFor=\"let child of row[thyChildrenKey]; trackBy: trackByFn; let j = index\"\n          [ngTemplateOutlet]=\"trTemplate\"\n          [ngTemplateOutletContext]=\"{ row: child, index: j, level: level + 1 }\"\n        ></ng-template>\n      </ng-container>\n    </ng-template>\n\n    <ng-template #tdsTemplate let-row=\"row\" let-i=\"index\" let-level=\"level\">\n      <td\n        *ngFor=\"let column of columns; let j = index\"\n        [ngClass]=\"column.className\"\n        [class.thy-table-checkbox-column]=\"column.type === customType.checkbox\"\n        [ngStyle]=\"mode === 'tree' && column.expand ? tdIndentComputed(level) : null\"\n      >\n        <ng-container *ngIf=\"j === 0 && draggable\">\n          <thy-icon class=\"table-draggable-icon\" thyIconName=\"drag\"></thy-icon>\n        </ng-container>\n\n        <!--template-->\n        <ng-container *ngIf=\"isTemplateRef(column.templateRef)\">\n          <thy-icon\n            class=\"tree-expand-icon cursor-pointer\"\n            *ngIf=\"mode === 'tree' && column.expand\"\n            [thyIconName]=\"isExpanded(row) ? 'angle-down' : 'angle-right'\"\n            [style.visibility]=\"showExpand(row) ? 'visible' : 'hidden'\"\n            [style.left.px]=\"iconIndentComputed(level)\"\n            (click)=\"expandChildren($event, row)\"\n          ></thy-icon>\n          <ng-template [ngTemplateOutlet]=\"column.templateRef\" [ngTemplateOutletContext]=\"{ $implicit: row }\"></ng-template>\n        </ng-container>\n\n        <!--not template-->\n        <ng-container *ngIf=\"!isTemplateRef(column.templateRef)\">\n          <!-- default -->\n          <ng-container *ngIf=\"column.model && !column.type\">\n            <thy-icon\n              class=\"tree-expand-icon cursor-pointer\"\n              *ngIf=\"mode === 'tree' && column.expand\"\n              [thyIconName]=\"isExpanded(row) ? 'angle-down' : 'angle-right'\"\n              [style.visibility]=\"showExpand(row) ? 'visible' : 'hidden'\"\n              [style.marginLeft.px]=\"iconIndentComputed(level)\"\n              (click)=\"expandChildren($event, row)\"\n            ></thy-icon>\n            <ng-container *ngIf=\"getModelValue(row, column.model) | isValidModelValue; else default\">\n              {{ getModelValue(row, column.model) }}\n            </ng-container>\n            <ng-template #default>\n              <div class=\"text-desc\">\n                {{ column.defaultText }}\n              </div>\n            </ng-template>\n          </ng-container>\n\n          <!-- index -->\n          <ng-container *ngIf=\"column.type === customType.index\">\n            {{ i + 1 }}\n          </ng-container>\n\n          <!-- checkbox -->\n          <ng-container *ngIf=\"column.type === customType.checkbox\">\n            <input\n              type=\"checkbox\"\n              [(ngModel)]=\"row[column.key]\"\n              (ngModelChange)=\"onCheckboxChange(row, column)\"\n              (click)=\"onStopPropagation($event)\"\n              [disabled]=\"column.disabled\"\n            />\n          </ng-container>\n\n          <!-- radio -->\n          <ng-container *ngIf=\"column.type === customType.radio\">\n            <input\n              type=\"radio\"\n              [(ngModel)]=\"selectedRadioRow\"\n              [value]=\"row\"\n              [disabled]=\"column.disabled\"\n              (click)=\"onStopPropagation($event)\"\n              (change)=\"onRadioSelectChange($event, row)\"\n            />\n          </ng-container>\n\n          <!-- switch -->\n          <ng-container *ngIf=\"column.type === customType.switch\">\n            <thy-switch\n              [(ngModel)]=\"row[column.key]\"\n              (ngModelChange)=\"onModelChange(row, column)\"\n              [disabled]=\"column.disabled\"\n              (thyChange)=\"onSwitchChange($event, row, column)\"\n            ></thy-switch>\n          </ng-container>\n        </ng-container>\n      </td>\n    </ng-template>\n  </tbody>\n</table>\n\n<div *ngIf=\"loadingDone && model.length === 0 && groups.length === 0\" class=\"thy-table-empty clear-hover\">\n  <ng-container *ngIf=\"emptyTemplate; else defaultEmpty\">\n    <ng-template [ngTemplateOutlet]=\"emptyTemplate\"></ng-template>\n  </ng-container>\n  <ng-template #defaultEmpty>\n    <thy-empty\n      [thyMessage]=\"emptyOptions.message\"\n      [thyTranslationKey]=\"emptyOptions.translationKey\"\n      [thyTranslationValues]=\"emptyOptions.translationValues\"\n      [thyEntityName]=\"emptyOptions.entityName\"\n      [thyEntityNameTranslateKey]=\"emptyOptions.entityNameTranslateKey\"\n      [thyIconName]=\"emptyOptions.iconName\"\n      [thySize]=\"emptyOptions.size\"\n      [thyMarginTop]=\"emptyOptions.marginTop\"\n      [thyTopAuto]=\"emptyOptions.topAuto\"\n      [thyContainer]=\"emptyOptions.container\"\n    ></thy-empty>\n  </ng-template>\n</div>\n\n<thy-loading [thyDone]=\"loadingDone\" [thyTip]=\"loadingText\"></thy-loading>\n\n<div class=\"thy-table-footer\" *ngIf=\"pagination.total > pagination.size\" [class.thy-table-footer-has-padding]=\"theme === 'default'\">\n  <thy-pagination\n    [thyPageIndex]=\"pagination.index\"\n    [thyTotal]=\"pagination.total\"\n    [thyPageSize]=\"pagination.size\"\n    [thyShowTotal]=\"showTotal\"\n    (thyPageChanged)=\"onPageChange($event)\"\n    (thyPageIndexChange)=\"onPageIndexChange($event)\"\n  ></thy-pagination>\n</div>\n\n<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                providers: [
                    {
                        provide: THY_TABLE_COLUMN_PARENT_COMPONENT,
                        useExisting: ThyTableComponent
                    },
                    core$1.UpdateHostClassService
                ],
                encapsulation: core.ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.IterableDiffers }, { type: ɵngcc1.ViewportRuler }, { type: ɵngcc2.UpdateHostClassService }, { type: undefined, decorators: [{
                type: core.Inject,
                args: [common.DOCUMENT]
            }] }]; }, { thyMode: [{
            type: core.Input
        }], thyGroupBy: [{
            type: core.Input
        }], thyRowKey: [{
            type: core.Input
        }], thyGroups: [{
            type: core.Input
        }], thyModel: [{
            type: core.Input
        }], thyTheme: [{
            type: core.Input
        }], thySize: [{
            type: core.Input
        }], thyClassName: [{
            type: core.Input
        }], thyRowClassName: [{
            type: core.Input
        }], thyLoadingDone: [{
            type: core.Input
        }], thyLoadingText: [{
            type: core.Input
        }], thyEmptyOptions: [{
            type: core.Input
        }], thyDraggable: [{
            type: core.Input
        }], thyFilter: [{
            type: core.Input
        }], thyPageIndex: [{
            type: core.Input
        }], thyPageSize: [{
            type: core.Input
        }], thyPageTotal: [{
            type: core.Input
        }], thyWholeRowSelect: [{
            type: core.Input
        }], listOfColumnComponents: [{
            type: core.ContentChildren,
            args: [ThyTableColumnComponent]
        }], emptyTemplate: [{
            type: core.ContentChild,
            args: ['empty']
        }], tableElementRef: [{
            type: core.ViewChild,
            args: ['table', { static: true }]
        }], thyShowHeader: [{
            type: core.Input
        }], showTotal: [{
            type: core.Input,
            args: ['thyShowTotal']
        }], thyIndent: [{
            type: core.Input
        }], thyChildrenKey: [{
            type: core.Input
        }], thyHoverDisplayOperation: [{
            type: core.HostBinding,
            args: ['class.thy-table-hover-display-operation']
        }, {
            type: core.Input
        }], thyOnSwitchChange: [{
            type: core.Output
        }], thyOnPageChange: [{
            type: core.Output
        }], thyOnPageIndexChange: [{
            type: core.Output
        }], thyOnMultiSelectChange: [{
            type: core.Output
        }], thyOnRadioSelectChange: [{
            type: core.Output
        }], thyOnDraggableChange: [{
            type: core.Output
        }], thyOnRowClick: [{
            type: core.Output
        }], thyOnRowContextMenu: [{
            type: core.Output
        }], groupTemplate: [{
            type: core.ContentChild,
            args: ['group', { static: true }]
        }], isTableClass: [{
            type: core.HostBinding,
            args: ['class.thy-table']
        }] }); })();
        return ThyTableComponent;
    }(_MixinBase));
    ThyTableComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.IterableDiffers },
        { type: overlay.ViewportRuler },
        { type: core$1.UpdateHostClassService },
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
    ]; };
    ThyTableComponent.propDecorators = {
        emptyTemplate: [{ type: core.ContentChild, args: ['empty',] }],
        tableElementRef: [{ type: core.ViewChild, args: ['table', { static: true },] }],
        thyMode: [{ type: core.Input }],
        thyGroupBy: [{ type: core.Input }],
        thyRowKey: [{ type: core.Input }],
        thyGroups: [{ type: core.Input }],
        thyModel: [{ type: core.Input }],
        thyTheme: [{ type: core.Input }],
        thySize: [{ type: core.Input }],
        thyClassName: [{ type: core.Input }],
        thyRowClassName: [{ type: core.Input }],
        thyLoadingDone: [{ type: core.Input }],
        thyLoadingText: [{ type: core.Input }],
        thyEmptyOptions: [{ type: core.Input }],
        thyDraggable: [{ type: core.Input }],
        thyFilter: [{ type: core.Input }],
        thyPageIndex: [{ type: core.Input }],
        thyPageSize: [{ type: core.Input }],
        thyPageTotal: [{ type: core.Input }],
        thyWholeRowSelect: [{ type: core.Input }],
        thyShowHeader: [{ type: core.Input }],
        showTotal: [{ type: core.Input, args: ['thyShowTotal',] }],
        thyIndent: [{ type: core.Input }],
        thyChildrenKey: [{ type: core.Input }],
        thyHoverDisplayOperation: [{ type: core.HostBinding, args: ['class.thy-table-hover-display-operation',] }, { type: core.Input }],
        thyOnSwitchChange: [{ type: core.Output }],
        thyOnPageChange: [{ type: core.Output }],
        thyOnPageIndexChange: [{ type: core.Output }],
        thyOnMultiSelectChange: [{ type: core.Output }],
        thyOnRadioSelectChange: [{ type: core.Output }],
        thyOnDraggableChange: [{ type: core.Output }],
        thyOnRowClick: [{ type: core.Output }],
        thyOnRowContextMenu: [{ type: core.Output }],
        groupTemplate: [{ type: core.ContentChild, args: ['group', { static: true },] }],
        listOfColumnComponents: [{ type: core.ContentChildren, args: [ThyTableColumnComponent,] }],
        isTableClass: [{ type: core.HostBinding, args: ['class.thy-table',] }]
    };
    /**
     * @deprecated ThyGridComponent is deprecated, please use ThyTableComponent
     */
    var ThyGridComponent = ThyTableComponent;

    var TableIsValidModelValuePipe = /** @class */ (function () {
        function TableIsValidModelValuePipe() {
        }
        TableIsValidModelValuePipe.prototype.transform = function (value) {
            return value !== '' && value !== undefined && value !== null;
        };
TableIsValidModelValuePipe.ɵfac = function TableIsValidModelValuePipe_Factory(t) { return new (t || TableIsValidModelValuePipe)(); };
TableIsValidModelValuePipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "isValidModelValue", type: TableIsValidModelValuePipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TableIsValidModelValuePipe, [{
        type: core.Pipe,
        args: [{ name: 'isValidModelValue' }]
    }], function () { return []; }, null); })();
        return TableIsValidModelValuePipe;
    }());

    var ThyTableModule = /** @class */ (function () {
        function ThyTableModule() {
        }
ThyTableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTableModule });
ThyTableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTableModule_Factory(t) { return new (t || ThyTableModule)(); }, imports: [[
            common.CommonModule,
            forms.FormsModule,
            pagination.ThyPaginationModule,
            _switch.ThySwitchModule,
            loading.ThyLoadingModule,
            empty.ThyEmptyModule,
            shared.ThySharedModule,
            icon.ThyIconModule,
            dragDrop.DragDropModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTableModule, { declarations: [ThyTableComponent, ThyTableColumnComponent, TableIsValidModelValuePipe], imports: [ɵngcc3.CommonModule, ɵngcc8.FormsModule, ɵngcc11.ThyPaginationModule, ɵngcc9.ThySwitchModule, ɵngcc4.ThyLoadingModule, ɵngcc10.ThyEmptyModule, ɵngcc5.ThySharedModule, ɵngcc7.ThyIconModule, ɵngcc6.DragDropModule], exports: [ThyTableComponent, ThyTableColumnComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyTableComponent, ThyTableColumnComponent, TableIsValidModelValuePipe],
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    pagination.ThyPaginationModule,
                    _switch.ThySwitchModule,
                    loading.ThyLoadingModule,
                    empty.ThyEmptyModule,
                    shared.ThySharedModule,
                    icon.ThyIconModule,
                    dragDrop.DragDropModule
                ],
                exports: [ThyTableComponent, ThyTableColumnComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyTableModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.THY_GRID_COLUMN_PARENT_COMPONENT = THY_GRID_COLUMN_PARENT_COMPONENT;
    exports.THY_TABLE_COLUMN_PARENT_COMPONENT = THY_TABLE_COLUMN_PARENT_COMPONENT;
    exports.TableIsValidModelValuePipe = TableIsValidModelValuePipe;
    exports.ThyGridColumnComponent = ThyGridColumnComponent;
    exports.ThyGridComponent = ThyGridComponent;
    exports.ThyTableColumnComponent = ThyTableColumnComponent;
    exports.ThyTableComponent = ThyTableComponent;
    exports.ThyTableModule = ThyTableModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
