// ngx-tethys/bundles/ngx-tethys-menu.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/popover'),require('ngx-tethys/icon'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('@angular/animations'), require('ngx-tethys/popover'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/menu', ['@angular/core','ngx-tethys/popover','ngx-tethys/icon','@angular/common','exports', '@angular/core', '@angular/common', '@angular/animations', 'ngx-tethys/popover', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.popover,global.ngxTethys.icon,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].menu = {}), global.ng.core, global.ng.common, global.ng.animations, global['ngx-tethys'].popover, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, animations, popover, icon) { 
var _c0 = ["*"];
var _c1 = ["thyMenuGroup"];
function ThyMenuGroupComponent_thy_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 7);
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.iconClass);
} }
function ThyMenuGroupComponent_thy_icon_5_Template(rf, ctx) { if (rf & 1) {
    var _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 8);
    ɵngcc0.ɵɵlistener("click", function ThyMenuGroupComponent_thy_icon_5_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); var ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onActionClick($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r1.rightIconClass);
} }
'use strict';

    var ThyMenuComponent = /** @class */ (function () {
        function ThyMenuComponent() {
            this.isThyMenu = true;
        }
        ThyMenuComponent.prototype.ngOnInit = function () { };
ThyMenuComponent.ɵfac = function ThyMenuComponent_Factory(t) { return new (t || ThyMenuComponent)(); };
ThyMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuComponent, selectors: [["thy-menu"], ["", "thy-menu", ""], ["", "thyMenu", ""]], hostVars: 2, hostBindings: function ThyMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu", ctx.isThyMenu);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu,[thy-menu],[thyMenu]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenu: [{
            type: core.HostBinding,
            args: ['class.thy-menu']
        }] }); })();
        return ThyMenuComponent;
    }());
    ThyMenuComponent.ctorParameters = function () { return []; };
    ThyMenuComponent.propDecorators = {
        isThyMenu: [{ type: core.HostBinding, args: ['class.thy-menu',] }]
    };

    var ThyMenuItemComponent = /** @class */ (function () {
        function ThyMenuItemComponent() {
            this.isThyMenuItem = true;
        }
        ThyMenuItemComponent.prototype.ngOnInit = function () { };
ThyMenuItemComponent.ɵfac = function ThyMenuItemComponent_Factory(t) { return new (t || ThyMenuItemComponent)(); };
ThyMenuItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemComponent, selectors: [["thy-menu-item"], ["", "thy-menu-item", ""], ["", "thyMenuItem", ""]], hostVars: 2, hostBindings: function ThyMenuItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item", ctx.isThyMenuItem);
    } }, ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "thy-menu-item-content"]], template: function ThyMenuItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu-item,[thy-menu-item],[thyMenuItem]',
                template: "<div class=\"thy-menu-item-content\">\n  <ng-content></ng-content>\n</div>\n"
            }]
    }], function () { return []; }, { isThyMenuItem: [{
            type: core.HostBinding,
            args: ['class.thy-menu-item']
        }] }); })();
        return ThyMenuItemComponent;
    }());
    ThyMenuItemComponent.ctorParameters = function () { return []; };
    ThyMenuItemComponent.propDecorators = {
        isThyMenuItem: [{ type: core.HostBinding, args: ['class.thy-menu-item',] }]
    };

    var ThyMenuGroupComponent = /** @class */ (function () {
        function ThyMenuGroupComponent(popover) {
            this.popover = popover;
            this.rightIconClass = 'more';
            this.iconClass = 'folder-bold';
            this.groupHeaderPaddingLeft = 0;
            this.isThyMenuGroup = true;
            this.showIcon = false;
            this.isCollapsed = true;
            this.thyTitle = '';
            this.thyShowAction = false;
            this.thyActionStopPropagation = true;
            this.thyOnActionClick = new core.EventEmitter();
        }
        Object.defineProperty(ThyMenuGroupComponent.prototype, "thyExpand", {
            set: function (value) {
                this.isCollapsed = !!!value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyMenuGroupComponent.prototype, "thyShowIcon", {
            set: function (value) {
                this.showIcon = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyMenuGroupComponent.prototype, "thyIcon", {
            set: function (value) {
                this.iconClass = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyMenuGroupComponent.prototype, "thyActionIcon", {
            set: function (value) {
                this.rightIconClass = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyMenuGroupComponent.prototype, "thyActionMenu", {
            set: function (value) {
                this._actionMenu = value;
            },
            enumerable: false,
            configurable: true
        });
        ThyMenuGroupComponent.prototype.ngOnInit = function () { };
        ThyMenuGroupComponent.prototype.collapseGroup = function () {
            this.isCollapsed = !this.isCollapsed;
        };
        ThyMenuGroupComponent.prototype.onActionClick = function (event) {
            if (this.thyActionStopPropagation) {
                event.stopPropagation();
            }
            if (this._actionMenu) {
                this.popover.open(this._actionMenu, {
                    origin: event.currentTarget,
                    insideClosable: true,
                    placement: 'bottom'
                });
            }
            else {
                this.thyOnActionClick.emit(event);
            }
        };
ThyMenuGroupComponent.ɵfac = function ThyMenuGroupComponent_Factory(t) { return new (t || ThyMenuGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover)); };
ThyMenuGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuGroupComponent, selectors: [["thy-menu-group"], ["", "thy-menu-group", ""], ["", "thyMenuGroup", ""]], viewQuery: function ThyMenuGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c1, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._thyMenuGroup = _t.first);
    } }, hostVars: 6, hostBindings: function ThyMenuGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-group", ctx.isThyMenuGroup)("has-icon", ctx.showIcon)("collapsed", ctx.isCollapsed);
    } }, inputs: { thyTitle: "thyTitle", thyShowAction: "thyShowAction", thyActionStopPropagation: "thyActionStopPropagation", thyExpand: "thyExpand", thyShowIcon: "thyShowIcon", thyIcon: "thyIcon", thyActionIcon: "thyActionIcon", thyActionMenu: "thyActionMenu" }, outputs: { thyOnActionClick: "thyOnActionClick" }, ngContentSelectors: _c0, decls: 9, vars: 5, consts: [[1, "thy-menu-group-header", 3, "click"], ["thyIconName", "caret-right-down", 1, "thy-menu-group-arrow"], ["class", "thy-menu-group-folder", 3, "thyIconName", 4, "ngIf"], [1, "thy-menu-group-title", 3, "title"], ["class", "thy-menu-group-more", 3, "thyIconName", "click", 4, "ngIf"], [1, "thy-menu-group-body"], ["thyMenuGroup", ""], [1, "thy-menu-group-folder", 3, "thyIconName"], [1, "thy-menu-group-more", 3, "thyIconName", "click"]], template: function ThyMenuGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("click", function ThyMenuGroupComponent_Template_div_click_0_listener() { return ctx.collapseGroup(); });
        ɵngcc0.ɵɵelement(1, "thy-icon", 1);
        ɵngcc0.ɵɵtemplate(2, ThyMenuGroupComponent_thy_icon_2_Template, 1, 1, "thy-icon", 2);
        ɵngcc0.ɵɵelementStart(3, "span", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ThyMenuGroupComponent_thy_icon_5_Template, 1, 1, "thy-icon", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5, 6);
        ɵngcc0.ɵɵprojection(8);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("title", ctx.thyTitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyTitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowAction);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("@detailsContentAnimation", ctx.isCollapsed);
    } }, directives: [ɵngcc2.ThyIconComponent, ɵngcc3.NgIf], encapsulation: 2, data: { animation: [
            animations.trigger('detailsContentAnimation', [
                animations.state('void', animations.style({
                    height: '*'
                })),
                animations.state('1', animations.style({
                    height: 0,
                    overflow: 'hidden'
                })),
                animations.state('0', animations.style({
                    height: '*'
                })),
                animations.transition('* => *', animations.animate('0ms ease-out'))
            ])
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuGroupComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu-group, [thy-menu-group],[thyMenuGroup]',
                template: "<div class=\"thy-menu-group-header\" (click)=\"collapseGroup()\">\n  <thy-icon thyIconName=\"caret-right-down\" class=\"thy-menu-group-arrow\"></thy-icon>\n  <thy-icon *ngIf=\"showIcon\" class=\"thy-menu-group-folder\" [thyIconName]=\"iconClass\"></thy-icon>\n  <span class=\"thy-menu-group-title\" [title]=\"thyTitle\">{{ thyTitle }}</span>\n  <thy-icon\n    *ngIf=\"thyShowAction\"\n    class=\"thy-menu-group-more\"\n    [thyIconName]=\"rightIconClass\"\n    (click)=\"onActionClick($event)\"\n  ></thy-icon>\n</div>\n<div class=\"thy-menu-group-body\" #thyMenuGroup [@detailsContentAnimation]=\"isCollapsed\">\n  <ng-content></ng-content>\n</div>\n",
                animations: [
                    animations.trigger('detailsContentAnimation', [
                        animations.state('void', animations.style({
                            height: '*'
                        })),
                        animations.state('1', animations.style({
                            height: 0,
                            overflow: 'hidden'
                        })),
                        animations.state('0', animations.style({
                            height: '*'
                        })),
                        animations.transition('* => *', animations.animate('0ms ease-out'))
                    ])
                ]
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }]; }, { isThyMenuGroup: [{
            type: core.HostBinding,
            args: ['class.thy-menu-group']
        }], showIcon: [{
            type: core.HostBinding,
            args: ['class.has-icon']
        }], isCollapsed: [{
            type: core.HostBinding,
            args: ['class.collapsed']
        }], thyTitle: [{
            type: core.Input
        }], thyShowAction: [{
            type: core.Input
        }], thyActionStopPropagation: [{
            type: core.Input
        }], thyOnActionClick: [{
            type: core.Output
        }], thyExpand: [{
            type: core.Input,
            args: ['thyExpand']
        }], thyShowIcon: [{
            type: core.Input,
            args: ['thyShowIcon']
        }], thyIcon: [{
            type: core.Input,
            args: ['thyIcon']
        }], thyActionIcon: [{
            type: core.Input,
            args: ['thyActionIcon']
        }], thyActionMenu: [{
            type: core.Input
        }], _thyMenuGroup: [{
            type: core.ViewChild,
            args: ['thyMenuGroup', { static: true }]
        }] }); })();
        return ThyMenuGroupComponent;
    }());
    ThyMenuGroupComponent.ctorParameters = function () { return [
        { type: popover.ThyPopover }
    ]; };
    ThyMenuGroupComponent.propDecorators = {
        _thyMenuGroup: [{ type: core.ViewChild, args: ['thyMenuGroup', { static: true },] }],
        isThyMenuGroup: [{ type: core.HostBinding, args: ['class.thy-menu-group',] }],
        showIcon: [{ type: core.HostBinding, args: ['class.has-icon',] }],
        isCollapsed: [{ type: core.HostBinding, args: ['class.collapsed',] }],
        thyTitle: [{ type: core.Input }],
        thyExpand: [{ type: core.Input, args: ['thyExpand',] }],
        thyShowIcon: [{ type: core.Input, args: ['thyShowIcon',] }],
        thyIcon: [{ type: core.Input, args: ['thyIcon',] }],
        thyActionIcon: [{ type: core.Input, args: ['thyActionIcon',] }],
        thyShowAction: [{ type: core.Input }],
        thyActionStopPropagation: [{ type: core.Input }],
        thyOnActionClick: [{ type: core.Output }],
        thyActionMenu: [{ type: core.Input }]
    };

    var ThyMenuItemNameComponent = /** @class */ (function () {
        function ThyMenuItemNameComponent() {
            this.isThyMenuItemName = true;
            this.thyOverflowEllipsis = true;
        }
        ThyMenuItemNameComponent.prototype.ngOnInit = function () { };
ThyMenuItemNameComponent.ɵfac = function ThyMenuItemNameComponent_Factory(t) { return new (t || ThyMenuItemNameComponent)(); };
ThyMenuItemNameComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemNameComponent, selectors: [["thy-menu-item-name"], ["", "thy-menu-item-name", ""], ["", "thyMenuItemName", ""]], hostVars: 4, hostBindings: function ThyMenuItemNameComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-name", ctx.isThyMenuItemName)("thy-menu-item-name-ellipsis", ctx.thyOverflowEllipsis);
    } }, inputs: { thyOverflowEllipsis: "thyOverflowEllipsis" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemNameComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemNameComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu-item-name,[thy-menu-item-name],[thyMenuItemName]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenuItemName: [{
            type: core.HostBinding,
            args: ['class.thy-menu-item-name']
        }], thyOverflowEllipsis: [{
            type: core.HostBinding,
            args: ['class.thy-menu-item-name-ellipsis']
        }, {
            type: core.Input
        }] }); })();
        return ThyMenuItemNameComponent;
    }());
    ThyMenuItemNameComponent.ctorParameters = function () { return []; };
    ThyMenuItemNameComponent.propDecorators = {
        isThyMenuItemName: [{ type: core.HostBinding, args: ['class.thy-menu-item-name',] }],
        thyOverflowEllipsis: [{ type: core.HostBinding, args: ['class.thy-menu-item-name-ellipsis',] }, { type: core.Input }]
    };

    var ThyMenuItemIconComponent = /** @class */ (function () {
        function ThyMenuItemIconComponent(elementRef, renderer) {
            this.elementRef = elementRef;
            this.renderer = renderer;
            this.isThyMenuItemIcon = true;
        }
        Object.defineProperty(ThyMenuItemIconComponent.prototype, "thyColor", {
            set: function (value) {
                if (value) {
                    this.renderer.setStyle(this.elementRef.nativeElement, 'color', value);
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyMenuItemIconComponent.prototype.ngOnInit = function () { };
ThyMenuItemIconComponent.ɵfac = function ThyMenuItemIconComponent_Factory(t) { return new (t || ThyMenuItemIconComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyMenuItemIconComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemIconComponent, selectors: [["thy-menu-item-icon"], ["", "thy-menu-item-icon", ""], ["", "thyMenuItemIcon", ""]], hostVars: 2, hostBindings: function ThyMenuItemIconComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-icon", ctx.isThyMenuItemIcon);
    } }, inputs: { thyColor: "thyColor" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemIconComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemIconComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu-item-icon,[thy-menu-item-icon],[thyMenuItemIcon]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { isThyMenuItemIcon: [{
            type: core.HostBinding,
            args: ['class.thy-menu-item-icon']
        }], thyColor: [{
            type: core.Input
        }] }); })();
        return ThyMenuItemIconComponent;
    }());
    ThyMenuItemIconComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.Renderer2 }
    ]; };
    ThyMenuItemIconComponent.propDecorators = {
        isThyMenuItemIcon: [{ type: core.HostBinding, args: ['class.thy-menu-item-icon',] }],
        thyColor: [{ type: core.Input }]
    };

    var ThyMenuItemActionComponent = /** @class */ (function () {
        function ThyMenuItemActionComponent(popover, renderer, elementRef) {
            this.popover = popover;
            this.renderer = renderer;
            this.elementRef = elementRef;
            this._boundEvent = false;
            this.isThyMenuItemIconMore = true;
            this.thyStopPropagation = true;
        }
        Object.defineProperty(ThyMenuItemActionComponent.prototype, "thyActionMenu", {
            set: function (value) {
                this._actionMenu = value;
                if (this._actionMenu) {
                    this.bindClickEvent();
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyMenuItemActionComponent.prototype.bindClickEvent = function () {
            var _this = this;
            if (this._boundEvent) {
                return;
            }
            this._boundEvent = true;
            this.renderer.listen(this.elementRef.nativeElement, 'click', function (event) {
                if (_this.thyStopPropagation) {
                    event.stopPropagation();
                }
                if (_this._actionMenu) {
                    _this.popover.open(_this._actionMenu, {
                        origin: event.currentTarget,
                        insideClosable: true,
                        placement: 'bottom'
                    });
                }
            });
        };
        ThyMenuItemActionComponent.prototype.ngOnInit = function () { };
ThyMenuItemActionComponent.ɵfac = function ThyMenuItemActionComponent_Factory(t) { return new (t || ThyMenuItemActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyMenuItemActionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemActionComponent, selectors: [["thy-menu-item-action"], ["", "thy-menu-item-action", ""], ["", "thyMenuItemAction", ""]], hostVars: 2, hostBindings: function ThyMenuItemActionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-action", ctx.isThyMenuItemIconMore);
    } }, inputs: { thyStopPropagation: "thyStopPropagation", thyActionMenu: "thyActionMenu" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemActionComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu-item-action,[thy-menu-item-action],[thyMenuItemAction]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { isThyMenuItemIconMore: [{
            type: core.HostBinding,
            args: ['class.thy-menu-item-action']
        }], thyStopPropagation: [{
            type: core.Input
        }], thyActionMenu: [{
            type: core.Input
        }] }); })();
        return ThyMenuItemActionComponent;
    }());
    ThyMenuItemActionComponent.ctorParameters = function () { return [
        { type: popover.ThyPopover },
        { type: core.Renderer2 },
        { type: core.ElementRef }
    ]; };
    ThyMenuItemActionComponent.propDecorators = {
        isThyMenuItemIconMore: [{ type: core.HostBinding, args: ['class.thy-menu-item-action',] }],
        thyActionMenu: [{ type: core.Input }],
        thyStopPropagation: [{ type: core.Input }]
    };

    var ThyMenuDividerComponent = /** @class */ (function () {
        function ThyMenuDividerComponent() {
            this.isThyMenuDivider = true;
        }
        ThyMenuDividerComponent.prototype.ngOnInit = function () { };
ThyMenuDividerComponent.ɵfac = function ThyMenuDividerComponent_Factory(t) { return new (t || ThyMenuDividerComponent)(); };
ThyMenuDividerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuDividerComponent, selectors: [["thy-menu-divider"], ["", "thy-menu-divider", ""], ["", "thyMenuDivider", ""]], hostVars: 2, hostBindings: function ThyMenuDividerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-divider", ctx.isThyMenuDivider);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuDividerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuDividerComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-menu-divider,[thy-menu-divider],[thyMenuDivider]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenuDivider: [{
            type: core.HostBinding,
            args: ['class.thy-menu-divider']
        }] }); })();
        return ThyMenuDividerComponent;
    }());
    ThyMenuDividerComponent.ctorParameters = function () { return []; };
    ThyMenuDividerComponent.propDecorators = {
        isThyMenuDivider: [{ type: core.HostBinding, args: ['class.thy-menu-divider',] }]
    };

    var ThyMenuModule = /** @class */ (function () {
        function ThyMenuModule() {
        }
ThyMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyMenuModule });
ThyMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyMenuModule_Factory(t) { return new (t || ThyMenuModule)(); }, imports: [[common.CommonModule, icon.ThyIconModule, popover.ThyPopoverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyMenuModule, { declarations: [ThyMenuComponent, ThyMenuGroupComponent, ThyMenuItemComponent, ThyMenuItemNameComponent, ThyMenuItemIconComponent, ThyMenuItemActionComponent, ThyMenuDividerComponent], imports: [ɵngcc3.CommonModule, ɵngcc2.ThyIconModule, ɵngcc1.ThyPopoverModule], exports: [ThyMenuComponent, ThyMenuGroupComponent, ThyMenuItemComponent, ThyMenuItemNameComponent, ThyMenuItemIconComponent, ThyMenuItemActionComponent, ThyMenuDividerComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    ThyMenuComponent,
                    ThyMenuGroupComponent,
                    ThyMenuItemComponent,
                    ThyMenuItemNameComponent,
                    ThyMenuItemIconComponent,
                    ThyMenuItemActionComponent,
                    ThyMenuDividerComponent
                ],
                imports: [common.CommonModule, icon.ThyIconModule, popover.ThyPopoverModule],
                exports: [
                    ThyMenuComponent,
                    ThyMenuGroupComponent,
                    ThyMenuItemComponent,
                    ThyMenuItemNameComponent,
                    ThyMenuItemIconComponent,
                    ThyMenuItemActionComponent,
                    ThyMenuDividerComponent
                ]
            }]
    }], function () { return []; }, null); })();
        return ThyMenuModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyMenuComponent = ThyMenuComponent;
    exports.ThyMenuDividerComponent = ThyMenuDividerComponent;
    exports.ThyMenuGroupComponent = ThyMenuGroupComponent;
    exports.ThyMenuItemActionComponent = ThyMenuItemActionComponent;
    exports.ThyMenuItemComponent = ThyMenuItemComponent;
    exports.ThyMenuItemIconComponent = ThyMenuItemIconComponent;
    exports.ThyMenuItemNameComponent = ThyMenuItemNameComponent;
    exports.ThyMenuModule = ThyMenuModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
