// ngx-tethys/bundles/ngx-tethys-layout.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('ngx-tethys/icon'),require('@angular/cdk/drag-drop'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/util'), require('@angular/cdk/drag-drop'), require('ngx-tethys/icon')) :
  typeof define === 'function' && define.amd ? define('ngx-tethys/layout', ['@angular/core','@angular/common','ngx-tethys/icon','@angular/cdk/drag-drop','exports', '@angular/core', '@angular/common', 'ngx-tethys/util', '@angular/cdk/drag-drop', 'ngx-tethys/icon'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ngxTethys.icon,global.ng.cdk.dragDrop,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].layout = {}), global.ng.core, global.ng.common, global['ngx-tethys'].util, global.ng.cdk.dragDrop, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, util, dragDrop, icon) { 
var _c0 = ["*"];
var _c1 = ["headerTitle"];
var _c2 = ["headerContent"];
var _c3 = ["headerOperation"];
function ThyHeaderComponent_ng_template_1_Template(rf, ctx) { }
function ThyHeaderComponent_ng_container_2_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 8);
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r4.svgIconName);
} }
function ThyHeaderComponent_ng_container_2_ng_template_2_i_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 10);
} if (rf & 2) {
    var ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r7.iconClass);
} }
function ThyHeaderComponent_ng_container_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyHeaderComponent_ng_container_2_ng_template_2_i_0_Template, 1, 1, "i", 9);
} if (rf & 2) {
    var ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.iconClass);
} }
function ThyHeaderComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyHeaderComponent_ng_container_2_thy_icon_1_Template, 1, 1, "thy-icon", 5);
    ɵngcc0.ɵɵtemplate(2, ThyHeaderComponent_ng_container_2_ng_template_2_Template, 1, 1, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(4, "span", 7);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var _r5 = ɵngcc0.ɵɵreference(3);
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.svgIconName)("ngIfElse", _r5);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.thyTitle);
} }
function ThyHeaderComponent_ng_template_4_Template(rf, ctx) { }
function ThyHeaderComponent_ng_template_6_Template(rf, ctx) { }
var _c4 = ["dragRef"];
'use strict';

  var ThyLayoutComponent = /** @class */ (function () {
      function ThyLayoutComponent() {
          this.thyLayoutClass = true;
          this.hasSidebar = false;
      }
ThyLayoutComponent.ɵfac = function ThyLayoutComponent_Factory(t) { return new (t || ThyLayoutComponent)(); };
ThyLayoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyLayoutComponent, selectors: [["thy-layout"]], hostVars: 4, hostBindings: function ThyLayoutComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout", ctx.thyLayoutClass)("thy-layout--has-sidebar", ctx.hasSidebar);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyLayoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLayoutComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-layout',
                template: "\n    <ng-content></ng-content>\n  "
            }]
    }], function () { return []; }, { thyLayoutClass: [{
            type: core.HostBinding,
            args: ['class.thy-layout']
        }], hasSidebar: [{
            type: core.HostBinding,
            args: ['class.thy-layout--has-sidebar']
        }] }); })();
      return ThyLayoutComponent;
  }());
  ThyLayoutComponent.propDecorators = {
      thyLayoutClass: [{ type: core.HostBinding, args: ['class.thy-layout',] }],
      hasSidebar: [{ type: core.HostBinding, args: ['class.thy-layout--has-sidebar',] }]
  };

  var ThyHeaderComponent = /** @class */ (function () {
      function ThyHeaderComponent() {
          this.thyLayoutHeaderClass = true;
          this._thyHasBorder = false;
          this._thySizeSm = false;
          this.thyIconPrefix = 'wtf';
      }
      Object.defineProperty(ThyHeaderComponent.prototype, "thyHasBorder", {
          set: function (value) {
              this._thyHasBorder = util.coerceBooleanProperty(value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(ThyHeaderComponent.prototype, "thySize", {
          set: function (value) {
              this._thySizeSm = value === 'sm';
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(ThyHeaderComponent.prototype, "thyIcon", {
          set: function (value) {
              if (value) {
                  if (value.includes('wtf')) {
                      this.iconClass = this.thyIconPrefix + " " + value;
                  }
                  else {
                      this.svgIconName = value;
                  }
              }
              else {
                  this.iconClass = null;
                  this.svgIconName = null;
              }
          },
          enumerable: false,
          configurable: true
      });
      ThyHeaderComponent.prototype.ngOnInit = function () { };
ThyHeaderComponent.ɵfac = function ThyHeaderComponent_Factory(t) { return new (t || ThyHeaderComponent)(); };
ThyHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyHeaderComponent, selectors: [["thy-header"]], contentQueries: function ThyHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c3, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationTemplateRef = _t.first);
    } }, hostVars: 6, hostBindings: function ThyHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-header", ctx.thyLayoutHeaderClass)("header-has-border", ctx._thyHasBorder)("thy-layout-header-sm", ctx._thySizeSm);
    } }, inputs: { thyIconPrefix: "thyIconPrefix", thyHasBorder: "thyHasBorder", thySize: "thySize", thyIcon: "thyIcon", thyTitle: "thyTitle" }, decls: 7, vars: 4, consts: [[1, "layout-header-title"], [3, "ngTemplateOutlet"], [4, "ngIf"], [1, "layout-header-content"], [1, "layout-header-operation"], ["class", "prefix-icon", 3, "thyIconName", 4, "ngIf", "ngIfElse"], ["iconFont", ""], [1, "title-name"], [1, "prefix-icon", 3, "thyIconName"], ["class", "prefix-icon", 3, "ngClass", 4, "ngIf"], [1, "prefix-icon", 3, "ngClass"]], template: function ThyHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyHeaderComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(2, ThyHeaderComponent_ng_container_2_Template, 6, 3, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyHeaderComponent_ng_template_4_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵtemplate(6, ThyHeaderComponent_ng_template_6_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.contentTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.operationTemplateRef);
    } }, directives: [ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyHeaderComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-header',
                preserveWhitespaces: false,
                template: "<div class=\"layout-header-title\">\n  <ng-template [ngTemplateOutlet]=\"titleTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!titleTemplateRef\">\n    <thy-icon *ngIf=\"svgIconName; else iconFont\" class=\"prefix-icon\" [thyIconName]=\"svgIconName\"></thy-icon>\n    <ng-template #iconFont>\n      <i *ngIf=\"iconClass\" class=\"prefix-icon\" [ngClass]=\"iconClass\"></i>\n    </ng-template>\n    <span class=\"title-name\">{{ thyTitle }}</span></ng-container\n  >\n</div>\n<div class=\"layout-header-content\">\n  <ng-template [ngTemplateOutlet]=\"contentTemplateRef\"></ng-template>\n</div>\n<div class=\"layout-header-operation\">\n  <ng-template [ngTemplateOutlet]=\"operationTemplateRef\"></ng-template>\n</div>\n"
            }]
    }], function () { return []; }, { thyLayoutHeaderClass: [{
            type: core.HostBinding,
            args: ['class.thy-layout-header']
        }], _thyHasBorder: [{
            type: core.HostBinding,
            args: ['class.header-has-border']
        }], _thySizeSm: [{
            type: core.HostBinding,
            args: ['class.thy-layout-header-sm']
        }], thyIconPrefix: [{
            type: core.Input
        }], thyHasBorder: [{
            type: core.Input,
            args: ['thyHasBorder']
        }], thySize: [{
            type: core.Input,
            args: ['thySize']
        }], thyIcon: [{
            type: core.Input,
            args: ['thyIcon']
        }], thyTitle: [{
            type: core.Input
        }], titleTemplateRef: [{
            type: core.ContentChild,
            args: ['headerTitle']
        }], contentTemplateRef: [{
            type: core.ContentChild,
            args: ['headerContent']
        }], operationTemplateRef: [{
            type: core.ContentChild,
            args: ['headerOperation']
        }] }); })();
      return ThyHeaderComponent;
  }());
  ThyHeaderComponent.ctorParameters = function () { return []; };
  ThyHeaderComponent.propDecorators = {
      thyLayoutHeaderClass: [{ type: core.HostBinding, args: ['class.thy-layout-header',] }],
      _thyHasBorder: [{ type: core.HostBinding, args: ['class.header-has-border',] }],
      _thySizeSm: [{ type: core.HostBinding, args: ['class.thy-layout-header-sm',] }],
      thyHasBorder: [{ type: core.Input, args: ['thyHasBorder',] }],
      thySize: [{ type: core.Input, args: ['thySize',] }],
      thyTitle: [{ type: core.Input }],
      thyIconPrefix: [{ type: core.Input }],
      thyIcon: [{ type: core.Input, args: ['thyIcon',] }],
      titleTemplateRef: [{ type: core.ContentChild, args: ['headerTitle',] }],
      contentTemplateRef: [{ type: core.ContentChild, args: ['headerContent',] }],
      operationTemplateRef: [{ type: core.ContentChild, args: ['headerOperation',] }]
  };

  var ThyContentComponent = /** @class */ (function () {
      function ThyContentComponent() {
          this._isLayoutContent = true;
      }
ThyContentComponent.ɵfac = function ThyContentComponent_Factory(t) { return new (t || ThyContentComponent)(); };
ThyContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentComponent, selectors: [["thy-content"]], hostVars: 2, hostBindings: function ThyContentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content", ctx._isLayoutContent);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-content',
                preserveWhitespaces: false,
                template: "\n    <ng-content></ng-content>\n    "
            }]
    }], function () { return []; }, { _isLayoutContent: [{
            type: core.HostBinding,
            args: ['class.thy-layout-content']
        }] }); })();
      return ThyContentComponent;
  }());
  ThyContentComponent.propDecorators = {
      _isLayoutContent: [{ type: core.HostBinding, args: ['class.thy-layout-content',] }]
  };

  var LG_WIDTH = 300;
  var ThySidebarComponent = /** @class */ (function () {
      function ThySidebarComponent(thyLayoutComponent, renderer, elementRef, ngZone) {
          this.thyLayoutComponent = thyLayoutComponent;
          this.renderer = renderer;
          this.elementRef = elementRef;
          this.ngZone = ngZone;
          this.thyLayoutSidebarClass = true;
          this.thyLayoutSidebarClearBorderRightClass = false;
          this.sidebarIsolated = false;
      }
      Object.defineProperty(ThySidebarComponent.prototype, "thyWidth", {
          set: function (value) {
              if (value === 'lg') {
                  value = LG_WIDTH;
              }
              this.thyLayoutSidebarWidth = value;
              this.widthPassive = value;
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(ThySidebarComponent.prototype, "thyHasBorderRight", {
          set: function (value) {
              this.thyLayoutSidebarClearBorderRightClass = !util.coerceBooleanProperty(value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(ThySidebarComponent.prototype, "thyIsolated", {
          set: function (value) {
              this.sidebarIsolated = util.coerceBooleanProperty(value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(ThySidebarComponent.prototype, "thyIsDraggableWidth", {
          set: function (value) {
              if (util.coerceBooleanProperty(value)) {
                  this.thyLayoutSidebarPointerEvents = 'all';
              }
          },
          enumerable: false,
          configurable: true
      });
      ThySidebarComponent.prototype.ngOnInit = function () {
          var _this = this;
          if (this.thyLayoutComponent) {
              this.thyLayoutComponent.hasSidebar = true;
          }
          this.ngZone.runOutsideAngular(function () {
              setTimeout(function () {
                  _this.widthPassive = _this.elementRef.nativeElement.clientWidth;
                  _this.renderer.setStyle(_this.dragRef.nativeElement, 'left', _this.numberConvertToFloor(_this.widthPassive) + 'px');
              }, 0);
          });
      };
      ThySidebarComponent.prototype.dragStartedHandler = function () {
          this.dragStartedX = this.dragRef.nativeElement.getBoundingClientRect().x;
      };
      ThySidebarComponent.prototype.dragEndedHandler = function () {
          var x = this.dragRef.nativeElement.getBoundingClientRect().x;
          this.widthPassive = this.numberConvertToFloor(this.widthPassive * 1 + (x - this.dragStartedX));
          this.thyLayoutSidebarWidth = this.widthPassive;
      };
      ThySidebarComponent.prototype.numberConvertToFloor = function (value) {
          var result = Math.floor(value * 1);
          if (result < 1) {
              result = 1;
          }
          return result;
      };
ThySidebarComponent.ɵfac = function ThySidebarComponent_Factory(t) { return new (t || ThySidebarComponent)(ɵngcc0.ɵɵdirectiveInject(ThyLayoutComponent, 9), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThySidebarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySidebarComponent, selectors: [["thy-sidebar"]], viewQuery: function ThySidebarComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c4, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dragRef = _t.first);
    } }, hostVars: 10, hostBindings: function ThySidebarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", ctx.thyLayoutSidebarWidth, "px")("pointer-events", ctx.thyLayoutSidebarPointerEvents);
        ɵngcc0.ɵɵclassProp("thy-layout-sidebar", ctx.thyLayoutSidebarClass)("thy-layout-sidebar--clear-border-right", ctx.thyLayoutSidebarClearBorderRightClass)("thy-layout-sidebar-isolated", ctx.sidebarIsolated);
    } }, inputs: { thyWidth: "thyWidth", thyHasBorderRight: "thyHasBorderRight", thyIsolated: "thyIsolated", thyIsDraggableWidth: "thyIsDraggableWidth" }, ngContentSelectors: _c0, decls: 3, vars: 0, consts: [["cdkDrag", "", "cdkDragLockAxis", "x", "cdkDragBoundary", ".thy-layout", 1, "sidebar-drag", 3, "cdkDragStarted", "cdkDragEnded"], ["dragRef", ""]], template: function ThySidebarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0, 1);
        ɵngcc0.ɵɵlistener("cdkDragStarted", function ThySidebarComponent_Template_div_cdkDragStarted_1_listener() { return ctx.dragStartedHandler(); })("cdkDragEnded", function ThySidebarComponent_Template_div_cdkDragEnded_1_listener() { return ctx.dragEndedHandler(); });
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc3.CdkDrag], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySidebarComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-sidebar',
                preserveWhitespaces: false,
                template: "\n        <ng-content></ng-content>\n        <div\n            class=\"sidebar-drag\"\n            #dragRef\n            cdkDrag\n            cdkDragLockAxis=\"x\"\n            cdkDragBoundary=\".thy-layout\"\n            (cdkDragStarted)=\"dragStartedHandler()\"\n            (cdkDragEnded)=\"dragEndedHandler()\"\n        ></div>\n    "
            }]
    }], function () { return [{ type: ThyLayoutComponent, decorators: [{
                type: core.Optional
            }, {
                type: core.Host
            }] }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }]; }, { thyLayoutSidebarClass: [{
            type: core.HostBinding,
            args: ['class.thy-layout-sidebar']
        }], thyLayoutSidebarClearBorderRightClass: [{
            type: core.HostBinding,
            args: ['class.thy-layout-sidebar--clear-border-right']
        }], sidebarIsolated: [{
            type: core.HostBinding,
            args: ['class.thy-layout-sidebar-isolated']
        }], thyWidth: [{
            type: core.Input,
            args: ['thyWidth']
        }], thyHasBorderRight: [{
            type: core.Input,
            args: ['thyHasBorderRight']
        }], thyIsolated: [{
            type: core.Input,
            args: ['thyIsolated']
        }], thyIsDraggableWidth: [{
            type: core.Input,
            args: ['thyIsDraggableWidth']
        }], thyLayoutSidebarWidth: [{
            type: core.HostBinding,
            args: ['style.width.px']
        }], thyLayoutSidebarPointerEvents: [{
            type: core.HostBinding,
            args: ['style.pointer-events']
        }], dragRef: [{
            type: core.ViewChild,
            args: ['dragRef', { static: true }]
        }] }); })();
      return ThySidebarComponent;
  }());
  ThySidebarComponent.ctorParameters = function () { return [
      { type: ThyLayoutComponent, decorators: [{ type: core.Optional }, { type: core.Host }] },
      { type: core.Renderer2 },
      { type: core.ElementRef },
      { type: core.NgZone }
  ]; };
  ThySidebarComponent.propDecorators = {
      thyLayoutSidebarClass: [{ type: core.HostBinding, args: ['class.thy-layout-sidebar',] }],
      thyLayoutSidebarClearBorderRightClass: [{ type: core.HostBinding, args: ['class.thy-layout-sidebar--clear-border-right',] }],
      thyLayoutSidebarWidth: [{ type: core.HostBinding, args: ['style.width.px',] }],
      thyLayoutSidebarPointerEvents: [{ type: core.HostBinding, args: ['style.pointer-events',] }],
      sidebarIsolated: [{ type: core.HostBinding, args: ['class.thy-layout-sidebar-isolated',] }],
      thyWidth: [{ type: core.Input, args: ['thyWidth',] }],
      thyHasBorderRight: [{ type: core.Input, args: ['thyHasBorderRight',] }],
      thyIsolated: [{ type: core.Input, args: ['thyIsolated',] }],
      thyIsDraggableWidth: [{ type: core.Input, args: ['thyIsDraggableWidth',] }],
      dragRef: [{ type: core.ViewChild, args: ['dragRef', { static: true },] }]
  };

  var ThyContentSectionComponent = /** @class */ (function () {
      function ThyContentSectionComponent() {
          this.thyLayoutContentSectionClass = true;
      }
ThyContentSectionComponent.ɵfac = function ThyContentSectionComponent_Factory(t) { return new (t || ThyContentSectionComponent)(); };
ThyContentSectionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentSectionComponent, selectors: [["thy-content-section"]], hostVars: 2, hostBindings: function ThyContentSectionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content-section", ctx.thyLayoutContentSectionClass);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentSectionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentSectionComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-content-section',
                preserveWhitespaces: false,
                template: "\n    <ng-content></ng-content>\n    "
            }]
    }], function () { return []; }, { thyLayoutContentSectionClass: [{
            type: core.HostBinding,
            args: ['class.thy-layout-content-section']
        }] }); })();
      return ThyContentSectionComponent;
  }());
  ThyContentSectionComponent.propDecorators = {
      thyLayoutContentSectionClass: [{ type: core.HostBinding, args: ['class.thy-layout-content-section',] }]
  };

  var ThyContentMainComponent = /** @class */ (function () {
      function ThyContentMainComponent() {
          this.thyLayoutContentClass = true;
      }
ThyContentMainComponent.ɵfac = function ThyContentMainComponent_Factory(t) { return new (t || ThyContentMainComponent)(); };
ThyContentMainComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentMainComponent, selectors: [["thy-content-main"]], hostVars: 2, hostBindings: function ThyContentMainComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content-main", ctx.thyLayoutContentClass);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentMainComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentMainComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-content-main',
                preserveWhitespaces: false,
                template: "\n    <ng-content></ng-content>\n    "
            }]
    }], function () { return []; }, { thyLayoutContentClass: [{
            type: core.HostBinding,
            args: ['class.thy-layout-content-main']
        }] }); })();
      return ThyContentMainComponent;
  }());
  ThyContentMainComponent.propDecorators = {
      thyLayoutContentClass: [{ type: core.HostBinding, args: ['class.thy-layout-content-main',] }]
  };

  var ThyLayoutModule = /** @class */ (function () {
      function ThyLayoutModule() {
      }
ThyLayoutModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLayoutModule });
ThyLayoutModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLayoutModule_Factory(t) { return new (t || ThyLayoutModule)(); }, imports: [[common.CommonModule, dragDrop.DragDropModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLayoutModule, { declarations: [ThyLayoutComponent, ThyHeaderComponent, ThyContentComponent, ThySidebarComponent, ThyContentSectionComponent, ThyContentMainComponent], imports: [ɵngcc1.CommonModule, ɵngcc3.DragDropModule, ɵngcc2.ThyIconModule], exports: [ThyLayoutComponent, ThyHeaderComponent, ThyContentComponent, ThySidebarComponent, ThyContentSectionComponent, ThyContentMainComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLayoutModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    ThyLayoutComponent,
                    ThyHeaderComponent,
                    ThyContentComponent,
                    ThySidebarComponent,
                    ThyContentSectionComponent,
                    ThyContentMainComponent
                ],
                imports: [common.CommonModule, dragDrop.DragDropModule, icon.ThyIconModule],
                exports: [
                    ThyLayoutComponent,
                    ThyHeaderComponent,
                    ThyContentComponent,
                    ThySidebarComponent,
                    ThyContentSectionComponent,
                    ThyContentMainComponent
                ]
            }]
    }], function () { return []; }, null); })();
      return ThyLayoutModule;
  }());

  /**
   * Generated bundle index. Do not edit.
   */

  exports.ThyContentComponent = ThyContentComponent;
  exports.ThyContentMainComponent = ThyContentMainComponent;
  exports.ThyContentSectionComponent = ThyContentSectionComponent;
  exports.ThyHeaderComponent = ThyHeaderComponent;
  exports.ThyLayoutComponent = ThyLayoutComponent;
  exports.ThyLayoutModule = ThyLayoutModule;
  exports.ThySidebarComponent = ThySidebarComponent;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
