// ngx-tethys/bundles/ngx-tethys-dropdown.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/dropdown', ['@angular/core','@angular/common','exports', '@angular/core', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].dropdown = {}), global.ng.core, global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, common) { 'use strict';

    var thyDropdown = 'thyDropdown';
    var thyDropdownSplit = 'thyDropdownSplit';
    var ThyDropdownDirective = /** @class */ (function () {
        function ThyDropdownDirective(elementRef) {
            this.elementRef = elementRef;
            this._isDropdownClass = true;
            this._isDropdownSplitClass = false;
            var attrs = Array.from(elementRef.nativeElement.attributes);
            if (attrs.find(function (item) { return item.name === thyDropdownSplit.toLocaleLowerCase(); })) {
                this._isDropdownSplitClass = true;
            }
        }
ThyDropdownDirective.ɵfac = function ThyDropdownDirective_Factory(t) { return new (t || ThyDropdownDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyDropdownDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDropdownDirective, selectors: [["", "thyDropdown", ""], ["", "thyDropdownSplit", ""]], hostVars: 4, hostBindings: function ThyDropdownDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-dropdown-toggle", ctx._isDropdownClass)("thy-dropdown-toggle-split", ctx._isDropdownSplitClass);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDropdownDirective, [{
        type: core.Directive,
        args: [{
                selector: "[" + thyDropdown + "],[" + thyDropdownSplit + "]"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { _isDropdownClass: [{
            type: core.HostBinding,
            args: ['class.thy-dropdown-toggle']
        }], _isDropdownSplitClass: [{
            type: core.HostBinding,
            args: ['class.thy-dropdown-toggle-split']
        }] }); })();
        return ThyDropdownDirective;
    }());
    ThyDropdownDirective.ctorParameters = function () { return [
        { type: core.ElementRef }
    ]; };
    ThyDropdownDirective.propDecorators = {
        _isDropdownClass: [{ type: core.HostBinding, args: ['class.thy-dropdown-toggle',] }],
        _isDropdownSplitClass: [{ type: core.HostBinding, args: ['class.thy-dropdown-toggle-split',] }]
    };

    var ThyDropdownModule = /** @class */ (function () {
        function ThyDropdownModule() {
        }
ThyDropdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDropdownModule });
ThyDropdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDropdownModule_Factory(t) { return new (t || ThyDropdownModule)(); }, imports: [[
            common.CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDropdownModule, { declarations: [ThyDropdownDirective], imports: [ɵngcc1.CommonModule], exports: [ThyDropdownDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDropdownModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    common.CommonModule
                ],
                declarations: [
                    ThyDropdownDirective
                ],
                exports: [
                    ThyDropdownDirective
                ]
            }]
    }], function () { return []; }, null); })();
        return ThyDropdownModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyDropdownDirective = ThyDropdownDirective;
    exports.ThyDropdownModule = ThyDropdownModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
