// ngx-tethys/bundles/ngx-tethys-pagination.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/common'),require('ngx-tethys/icon'),require('ngx-tethys/shared'),require('@angular/platform-browser'),exports, require('@angular/core'), require('ngx-tethys/core'), require('ngx-tethys/util'), require('@angular/common'), require('@angular/platform-browser'), require('ngx-tethys/shared'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/pagination', ['@angular/core','ngx-tethys/core','@angular/common','ngx-tethys/icon','ngx-tethys/shared','@angular/platform-browser','exports', '@angular/core', 'ngx-tethys/core', 'ngx-tethys/util', '@angular/common', '@angular/platform-browser', 'ngx-tethys/shared', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.common,global.ngxTethys.icon,global.ngxTethys.shared,global.ng.platformBrowser,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].pagination = {}), global.ng.core, global['ngx-tethys'].core, global['ngx-tethys'].util, global.ng.common, global.ng.platformBrowser, global['ngx-tethys'].shared, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,exports, core, core$1, util, common, platformBrowser, shared, icon) { 
function ThyPaginationComponent_ng_container_0_div_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 10);
    ɵngcc0.ɵɵtext(2, " \u7B2C");
    ɵngcc0.ɵɵelementStart(3, "span", 11);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(5, "\u6761 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div");
    ɵngcc0.ɵɵtext(7, " \u5171\u8BA1");
    ɵngcc0.ɵɵelementStart(8, "span", 11);
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(10, "\u6761 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r9 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate2(" ", ctx_r9.range.from, "-", ctx_r9.range.to, " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r9.total, " ");
} }
function ThyPaginationComponent_ng_container_0_div_1_ng_template_2_ng_template_0_Template(rf, ctx) { }
var _c0 = function (a0, a1) { return { $implicit: a0, range: a1 }; };
function ThyPaginationComponent_ng_container_0_div_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyPaginationComponent_ng_container_0_div_1_ng_template_2_ng_template_0_Template, 0, 0, "ng-template", 12);
} if (rf & 2) {
    var ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r11.showTotal)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c0, ctx_r11.total, ctx_r11.range));
} }
function ThyPaginationComponent_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵtemplate(1, ThyPaginationComponent_ng_container_0_div_1_ng_container_1_Template, 11, 3, "ng-container", 8);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_div_1_ng_template_2_Template, 1, 5, "ng-template", null, 9, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r10 = ɵngcc0.ɵɵreference(3);
    var ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.isTemplateRef(ctx_r1.showTotal))("ngIfElse", _r10);
} }
function ThyPaginationComponent_ng_container_0_li_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r13 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r13.config.firstText);
} }
function ThyPaginationComponent_ng_container_0_li_4_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r14 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r14.config.firstIcon);
} }
function ThyPaginationComponent_ng_container_0_li_4_Template(rf, ctx) { if (rf & 1) {
    var _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_4_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); var ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.selectPage(ctx_r15.firstIndex); });
    ɵngcc0.ɵɵelementStart(1, "a", 14);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_4_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_4_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r2.disabled || ctx_r2.pageIndex === ctx_r2.firstIndex);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.config.firstText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.config.firstIcon);
} }
function ThyPaginationComponent_ng_container_0_li_5_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r17 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r17.config.previousText);
} }
function ThyPaginationComponent_ng_container_0_li_5_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r18 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r18.config.previousIcon);
} }
function ThyPaginationComponent_ng_container_0_li_5_Template(rf, ctx) { if (rf & 1) {
    var _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_5_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r20); var ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.selectPage(ctx_r19.pageIndex - 1); });
    ɵngcc0.ɵɵelementStart(1, "a", 16);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_5_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_5_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r3.disabled || ctx_r3.pageIndex === ctx_r3.firstIndex);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.config.previousText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.config.previousIcon);
} }
function ThyPaginationComponent_ng_container_0_li_6_Template(rf, ctx) { if (rf & 1) {
    var _r23 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_6_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r23); var page_r21 = ctx.$implicit; var ctx_r22 = ɵngcc0.ɵɵnextContext(2); return ctx_r22.selectPage(page_r21.index); });
    ɵngcc0.ɵɵelementStart(1, "a", 17);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var page_r21 = ctx.$implicit;
    var ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("active", page_r21.active)("disabled", ctx_r4.disabled);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(page_r21.text);
} }
function ThyPaginationComponent_ng_container_0_li_7_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r24 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r24.config.nextText);
} }
function ThyPaginationComponent_ng_container_0_li_7_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r25 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r25.config.nextIcon);
} }
function ThyPaginationComponent_ng_container_0_li_7_Template(rf, ctx) { if (rf & 1) {
    var _r27 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 18);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_7_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r27); var ctx_r26 = ɵngcc0.ɵɵnextContext(2); return ctx_r26.selectPage(ctx_r26.pageIndex + 1); });
    ɵngcc0.ɵɵelementStart(1, "a", 19);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_7_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_7_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r5.disabled || ctx_r5.pageIndex === ctx_r5.pageCount);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.config.nextText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.config.nextIcon);
} }
function ThyPaginationComponent_ng_container_0_li_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r28 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r28.config.lastText);
} }
function ThyPaginationComponent_ng_container_0_li_8_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r29 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r29.config.lastIcon);
} }
function ThyPaginationComponent_ng_container_0_li_8_Template(rf, ctx) { if (rf & 1) {
    var _r31 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_8_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r31); var ctx_r30 = ɵngcc0.ɵɵnextContext(2); return ctx_r30.selectPage(ctx_r30.pageCount); });
    ɵngcc0.ɵɵelementStart(1, "a", 20);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_8_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_8_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r6.disabled || ctx_r6.pageIndex === ctx_r6.pageCount);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.config.lastText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.config.lastIcon);
} }
function ThyPaginationComponent_ng_container_0_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "div", 21);
    ɵngcc0.ɵɵpipe(2, "paginationTotalPagesFormat");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind2(2, 1, ctx_r7.pageCount, ctx_r7.config.totalPagesFormat), ɵngcc0.ɵɵsanitizeHtml);
} }
function ThyPaginationComponent_ng_container_0_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    var _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 22);
    ɵngcc0.ɵɵelementStart(2, "div", 23);
    ɵngcc0.ɵɵtext(3, " \u5230 ");
    ɵngcc0.ɵɵelementStart(4, "input", 24, 25);
    ɵngcc0.ɵɵlistener("thyEnter", function ThyPaginationComponent_ng_container_0_ng_container_10_Template_input_thyEnter_4_listener() { ɵngcc0.ɵɵrestoreView(_r34); var _r32 = ɵngcc0.ɵɵreference(5); var ctx_r33 = ɵngcc0.ɵɵnextContext(2); return ctx_r33.jumpPage(_r32); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(6, " \u9875 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "button", 26);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_ng_container_10_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r34); var _r32 = ɵngcc0.ɵɵreference(5); var ctx_r35 = ɵngcc0.ɵɵnextContext(2); return ctx_r35.jumpPage(_r32); });
    ɵngcc0.ɵɵtext(8, "\u786E\u5B9A");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("disabled", ctx_r8.disabled);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("disabled", ctx_r8.disabled);
} }
function ThyPaginationComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyPaginationComponent_ng_container_0_div_1_Template, 4, 2, "div", 1);
    ɵngcc0.ɵɵelementStart(2, "div", 2);
    ɵngcc0.ɵɵelementStart(3, "ul", 3);
    ɵngcc0.ɵɵtemplate(4, ThyPaginationComponent_ng_container_0_li_4_Template, 4, 4, "li", 4);
    ɵngcc0.ɵɵtemplate(5, ThyPaginationComponent_ng_container_0_li_5_Template, 4, 4, "li", 4);
    ɵngcc0.ɵɵtemplate(6, ThyPaginationComponent_ng_container_0_li_6_Template, 3, 5, "li", 5);
    ɵngcc0.ɵɵtemplate(7, ThyPaginationComponent_ng_container_0_li_7_Template, 4, 4, "li", 6);
    ɵngcc0.ɵɵtemplate(8, ThyPaginationComponent_ng_container_0_li_8_Template, 4, 4, "li", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(9, ThyPaginationComponent_ng_container_0_ng_container_9_Template, 3, 4, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(10, ThyPaginationComponent_ng_container_0_ng_container_10_Template, 9, 2, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.showTotal);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.boundaryLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.directionLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.pages);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.directionLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.boundaryLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.totalPagesFormat);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.showQuickJumper);
} }
'use strict';

    var DEFAULT_RANGE_COUNT = 5;
    var PaginationDefaultConfig = {
        boundaryLinks: false,
        directionLinks: true,
        pageSize: 20,
        maxCount: 7,
        rangeCount: DEFAULT_RANGE_COUNT,
        showQuickJumper: true,
        firstText: '第一页',
        lastText: '最后一页',
        previousText: '',
        nextText: '',
        firstIcon: '',
        lastIcon: '',
        previousIcon: 'angle-left',
        nextIcon: 'angle-right',
        totalPagesFormat: '共{total}页'
    };
    var THY_PAGINATION_CONFIG = new core.InjectionToken('THY_PAGINATION_CONFIG', {
        providedIn: 'root',
        factory: function () {
            return {};
        }
    });

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ThyPaginationComponent = /** @class */ (function () {
        function ThyPaginationComponent(paginationConfig, updateHostClassService, elementRef, cdr) {
            this.paginationConfig = paginationConfig;
            this.updateHostClassService = updateHostClassService;
            this.elementRef = elementRef;
            this.cdr = cdr;
            this.isTemplateRef = util.isTemplateRef;
            this.config = Object.assign({}, PaginationDefaultConfig, this.paginationConfig.main);
            this.disabled = false;
            this.pageIndexChange = new core.EventEmitter();
            this.pageChanged = new core.EventEmitter();
            this.pages = [];
            this.pageIndex = 1;
            this.range = { from: 0, to: 0 };
            this.firstIndex = 1;
            this.isHideOnSinglePage = false;
            this.initialized = false;
            this.isPaginationClass = true;
            // 是否显示范围和total
            this.showTotal = false;
            this.updateHostClassService.initializeElement(this.elementRef.nativeElement);
        }
        Object.defineProperty(ThyPaginationComponent.prototype, "thyPageIndex", {
            set: function (pageIndex) {
                this.pageIndex = pageIndex;
                if (this.initialized) {
                    this.setPageIndex(pageIndex);
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyPaginationComponent.prototype, "thyPageSize", {
            set: function (pageSize) {
                this.pageSize = pageSize;
                if (this.initialized) {
                    this.calculatePageCount();
                    this.initializePages(this.pageIndex, this.pageCount);
                    this.cdr.markForCheck();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyPaginationComponent.prototype, "thyTotal", {
            set: function (total) {
                this.total = total;
                if (this.initialized) {
                    this.calculatePageCount();
                    this.initializePages(this.pageIndex, this.pageCount);
                    this.cdr.markForCheck();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyPaginationComponent.prototype, "showQuickJumper", {
            set: function (value) {
                this.config.showQuickJumper = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyPaginationComponent.prototype, "size", {
            set: function (size) {
                this.updateHostClassService.addClass("thy-pagination-" + size);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyPaginationComponent.prototype, "maxCount", {
            set: function (value) {
                this.config.maxCount = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyPaginationComponent.prototype, "thyRangeCount", {
            set: function (value) {
                if (Number.isInteger(value)) {
                    this.config.rangeCount = value;
                    if (this.initialized) {
                        this.setMarginalCount(value);
                    }
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyPaginationComponent.prototype.ngOnInit = function () {
            this.setMarginalCount(this.config.rangeCount);
            this.calculatePageCount();
            this.setPageIndex(this.pageIndex);
            this.initialized = true;
        };
        ThyPaginationComponent.prototype.setMarginalCount = function (range) {
            if (!this.marginalCount) {
                this.marginalCount = range <= DEFAULT_RANGE_COUNT ? 1 : 2;
            }
        };
        ThyPaginationComponent.prototype.setPageIndex = function (pageIndex) {
            this.pageIndex = pageIndex > this.pageCount ? this.pageCount : pageIndex || 1;
            this.range = {
                from: (this.pageIndex - 1) * this.pageSize + 1,
                to: this.pageIndex * this.pageSize
            };
            this.initializePages(this.pageIndex, this.pageCount);
            this.cdr.markForCheck();
        };
        ThyPaginationComponent.prototype.calculatePageCount = function () {
            var pageCount = this.pageSize < 1 ? 1 : Math.ceil(this.total / this.pageSize);
            this.pageCount = Math.max(pageCount || 0, 1);
        };
        ThyPaginationComponent.prototype.makePage = function (index, text, active) {
            return { index: index, text: text, active: active };
        };
        ThyPaginationComponent.prototype.initializePages = function (pageIndex, pageCount) {
            var marginalCount = this.marginalCount;
            var rangeCount = this.config.rangeCount;
            var maxCount = this.config.maxCount;
            var pages = [];
            var isMaxSized = pageCount > maxCount;
            if (isMaxSized) {
                var beforePages = [];
                var afterPages = [];
                // beforePages
                for (var i = 1; i <= marginalCount; i++) {
                    beforePages.push(this.makePage(i, i.toString(), i === pageIndex));
                }
                if (pageIndex - Math.ceil(rangeCount / 2) > this.firstIndex) {
                    beforePages.push(this.makePage(pageIndex - rangeCount, '...', null));
                }
                // afterPages
                if (pageIndex + Math.ceil(rangeCount / 2) < pageCount) {
                    afterPages.push(this.makePage(pageIndex + rangeCount, '...', null));
                }
                for (var i = pageCount - marginalCount + 1; i <= pageCount; i++) {
                    afterPages.push(this.makePage(i, i.toString(), i === pageIndex));
                }
                // mainPages
                var start = Math.max(marginalCount + 1, pageIndex - (rangeCount - 1) / 2);
                var end = Math.min(pageIndex + (rangeCount - 1) / 2, pageCount - marginalCount);
                if (pageIndex - 1 <= marginalCount) {
                    end = rangeCount;
                }
                if (pageCount - pageIndex <= marginalCount) {
                    start = pageCount - rangeCount + 1;
                }
                for (var i = start; i <= end; i++) {
                    pages.push({
                        index: i,
                        text: i.toString(),
                        active: i === +pageIndex
                    });
                }
                pages = __spread(beforePages, pages, afterPages);
            }
            else {
                for (var i = 1; i <= pageCount; i++) {
                    pages.push({
                        index: i,
                        text: i.toString(),
                        active: i === +pageIndex
                    });
                }
            }
            this.pages = pages;
        };
        ThyPaginationComponent.prototype.pageChange = function (pageIndex) {
            this.pageIndexChange.emit(pageIndex);
            this.pageChanged.emit({ page: pageIndex });
        };
        ThyPaginationComponent.prototype.selectPage = function (pageIndex) {
            if (this.disabled || pageIndex === this.firstIndex - 1 || pageIndex === this.pageCount + 1) {
                return;
            }
            this.setPageIndex(pageIndex);
            this.pageChange(this.pageIndex);
        };
        ThyPaginationComponent.prototype.jumpPage = function (input) {
            var pageIndex = +input.value;
            if (Number.isInteger(pageIndex)) {
                this.selectPage(pageIndex);
            }
            input.value = '';
        };
ThyPaginationComponent.ɵfac = function ThyPaginationComponent_Factory(t) { return new (t || ThyPaginationComponent)(ɵngcc0.ɵɵdirectiveInject(THY_PAGINATION_CONFIG, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyPaginationComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPaginationComponent, selectors: [["thy-pagination"]], hostVars: 4, hostBindings: function ThyPaginationComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-pagination", ctx.isPaginationClass)("thy-pagination-has-total", ctx.showTotal);
    } }, inputs: { disabled: ["thyDisabled", "disabled"], showTotal: ["thyShowTotal", "showTotal"], thyPageIndex: "thyPageIndex", thyPageSize: "thyPageSize", thyTotal: "thyTotal", showQuickJumper: ["thyShowQuickJumper", "showQuickJumper"], size: ["thySize", "size"], maxCount: ["thyMaxCount", "maxCount"], thyRangeCount: "thyRangeCount", marginalCount: ["thyMarginalCount", "marginalCount"], hideOnSinglePage: ["thyHideOnSinglePage", "hideOnSinglePage"] }, outputs: { pageIndexChange: "thyPageIndexChange", pageChanged: "thyPageChanged" }, features: [ɵngcc0.ɵɵProvidersFeature([core$1.UpdateHostClassService])], decls: 1, vars: 1, consts: [[4, "ngIf"], ["class", "thy-pagination-total", 4, "ngIf"], [1, "thy-pagination-content"], [1, "thy-pagination-pages"], ["class", "thy-page-item", 3, "disabled", "click", 4, "ngIf"], ["class", "thy-page-item", 3, "active", "disabled", "click", 4, "ngFor", "ngForOf"], ["class", "thy-page-item page-item-next", 3, "disabled", "click", 4, "ngIf"], [1, "thy-pagination-total"], [4, "ngIf", "ngIfElse"], ["totalTemplate", ""], [1, "mr-3"], [1, "number"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "thy-page-item", 3, "click"], ["href", "javascript:;", 1, "thy-page-link", "thy-page-link-first"], [3, "thyIconName"], ["href", "javascript:;", 1, "thy-page-link", "page-link-pre"], ["href", "javascript:;", 1, "thy-page-link"], [1, "thy-page-item", "page-item-next", 3, "click"], ["href", "javascript:;", 1, "thy-page-link", "thy-page-link-next"], ["href", "javascript:;", 1, "thy-page-link", "thy-page-link-last"], [1, "thy-pagination-count", 3, "innerHTML"], [1, "thy-pagination-jumper"], [1, "thy-pagination-jumper-input"], ["type", "text", 3, "disabled", "thyEnter"], ["jumperInput", ""], ["type", "button", 3, "disabled", "click"]], template: function ThyPaginationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyPaginationComponent_ng_container_0_Template, 11, 8, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.hideOnSinglePage && ctx.total > ctx.pageSize || !ctx.hideOnSinglePage);
    } }, directives: function () { return [ɵngcc2.NgIf, ɵngcc2.NgForOf, ɵngcc2.NgTemplateOutlet, ɵngcc3.ThyIconComponent, ɵngcc4.ThyEnterDirective]; }, pipes: function () { return [PaginationTotalCountFormat]; }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPaginationComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-pagination',
                template: "<ng-container *ngIf=\"(hideOnSinglePage && total > pageSize) || !hideOnSinglePage\">\n  <div class=\"thy-pagination-total\" *ngIf=\"showTotal\">\n    <ng-container *ngIf=\"!isTemplateRef(showTotal); else totalTemplate\">\n      <div class=\"mr-3\">\n        \u7B2C<span class=\"number\"> {{ range.from }}-{{ range.to }} </span>\u6761\n      </div>\n      <div>\n        \u5171\u8BA1<span class=\"number\"> {{ total }} </span>\u6761\n      </div>\n    </ng-container>\n    <ng-template #totalTemplate>\n      <ng-template\n        [ngTemplateOutlet]=\"showTotal\"\n        [ngTemplateOutletContext]=\"{ $implicit: total, range: range }\"\n      ></ng-template>\n    </ng-template>\n  </div>\n  <div class=\"thy-pagination-content\">\n    <ul class=\"thy-pagination-pages\">\n      <li\n        class=\"thy-page-item\"\n        [class.disabled]=\"disabled || pageIndex === firstIndex\"\n        *ngIf=\"config.boundaryLinks\"\n        (click)=\"selectPage(firstIndex)\"\n      >\n        <a class=\"thy-page-link thy-page-link-first\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.firstText\">{{ config.firstText }}</ng-container>\n          <ng-container *ngIf=\"config.firstIcon\">\n            <thy-icon [thyIconName]=\"config.firstIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n      <li\n        class=\"thy-page-item\"\n        [class.disabled]=\"disabled || pageIndex === firstIndex\"\n        *ngIf=\"config.directionLinks\"\n        (click)=\"selectPage(pageIndex - 1)\"\n      >\n        <a class=\"thy-page-link  page-link-pre\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.previousText\">{{ config.previousText }}</ng-container>\n          <ng-container *ngIf=\"config.previousIcon\">\n            <thy-icon [thyIconName]=\"config.previousIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n      <li\n        class=\"thy-page-item\"\n        [class.active]=\"page.active\"\n        [class.disabled]=\"disabled\"\n        (click)=\"selectPage(page.index)\"\n        *ngFor=\"let page of pages\"\n      >\n        <a class=\"thy-page-link\" href=\"javascript:;\">{{ page.text }}</a>\n      </li>\n      <li\n        class=\"thy-page-item page-item-next\"\n        [class.disabled]=\"disabled || pageIndex === pageCount\"\n        *ngIf=\"config.directionLinks\"\n        (click)=\"selectPage(pageIndex + 1)\"\n      >\n        <a class=\"thy-page-link thy-page-link-next\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.nextText\">{{ config.nextText }}</ng-container>\n          <ng-container *ngIf=\"config.nextIcon\">\n            <thy-icon [thyIconName]=\"config.nextIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n      <li\n        class=\"thy-page-item\"\n        [class.disabled]=\"disabled || pageIndex === pageCount\"\n        *ngIf=\"config.boundaryLinks\"\n        (click)=\"selectPage(pageCount)\"\n      >\n        <a class=\"thy-page-link thy-page-link-last\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.lastText\">{{ config.lastText }}</ng-container>\n          <ng-container *ngIf=\"config.lastIcon\">\n            <thy-icon [thyIconName]=\"config.lastIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n    </ul>\n    <ng-container *ngIf=\"config.totalPagesFormat\">\n      <div\n        class=\"thy-pagination-count\"\n        [innerHTML]=\"pageCount | paginationTotalPagesFormat: config.totalPagesFormat\"\n      ></div>\n    </ng-container>\n    <ng-container *ngIf=\"config.showQuickJumper\">\n      <div class=\"thy-pagination-jumper\">\n        <div class=\"thy-pagination-jumper-input\">\n          \u5230 <input type=\"text\" [disabled]=\"disabled\" #jumperInput (thyEnter)=\"jumpPage(jumperInput)\" /> \u9875\n        </div>\n        <button type=\"button\" (click)=\"jumpPage(jumperInput)\" [disabled]=\"disabled\">\u786E\u5B9A</button>\n      </div>\n    </ng-container>\n  </div>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [core$1.UpdateHostClassService]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: core.Optional
            }, {
                type: core.Inject,
                args: [THY_PAGINATION_CONFIG]
            }] }, { type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }]; }, { disabled: [{
            type: core.Input,
            args: ['thyDisabled']
        }], pageIndexChange: [{
            type: core.Output,
            args: ['thyPageIndexChange']
        }], pageChanged: [{
            type: core.Output,
            args: ['thyPageChanged']
        }], isPaginationClass: [{
            type: core.HostBinding,
            args: ['class.thy-pagination']
        }], showTotal: [{
            type: core.HostBinding,
            args: ['class.thy-pagination-has-total']
        }, {
            type: core.Input,
            args: ['thyShowTotal']
        }], thyPageIndex: [{
            type: core.Input
        }], thyPageSize: [{
            type: core.Input
        }], thyTotal: [{
            type: core.Input
        }], showQuickJumper: [{
            type: core.Input,
            args: ['thyShowQuickJumper']
        }], size: [{
            type: core.Input,
            args: ['thySize']
        }], maxCount: [{
            type: core.Input,
            args: ['thyMaxCount']
        }], thyRangeCount: [{
            type: core.Input
        }], marginalCount: [{
            type: core.Input,
            args: ['thyMarginalCount']
        }], hideOnSinglePage: [{
            type: core.Input,
            args: ['thyHideOnSinglePage']
        }] }); })();
        return ThyPaginationComponent;
    }());
    ThyPaginationComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [THY_PAGINATION_CONFIG,] }] },
        { type: core$1.UpdateHostClassService },
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };
    ThyPaginationComponent.propDecorators = {
        thyPageIndex: [{ type: core.Input }],
        thyPageSize: [{ type: core.Input }],
        thyTotal: [{ type: core.Input }],
        disabled: [{ type: core.Input, args: ['thyDisabled',] }],
        showQuickJumper: [{ type: core.Input, args: ['thyShowQuickJumper',] }],
        size: [{ type: core.Input, args: ['thySize',] }],
        maxCount: [{ type: core.Input, args: ['thyMaxCount',] }],
        marginalCount: [{ type: core.Input, args: ['thyMarginalCount',] }],
        thyRangeCount: [{ type: core.Input }],
        hideOnSinglePage: [{ type: core.Input, args: ['thyHideOnSinglePage',] }],
        pageIndexChange: [{ type: core.Output, args: ['thyPageIndexChange',] }],
        pageChanged: [{ type: core.Output, args: ['thyPageChanged',] }],
        isPaginationClass: [{ type: core.HostBinding, args: ['class.thy-pagination',] }],
        showTotal: [{ type: core.HostBinding, args: ['class.thy-pagination-has-total',] }, { type: core.Input, args: ['thyShowTotal',] }]
    };

    var PaginationTotalCountFormat = /** @class */ (function () {
        function PaginationTotalCountFormat(sanitizer) {
            this.sanitizer = sanitizer;
        }
        PaginationTotalCountFormat.prototype.transform = function (count, format) {
            if (count && format) {
                return this.sanitizer.bypassSecurityTrustHtml(format.replace(/\{(.+?)\}/g, "<span>" + count + "</span>"));
            }
            else {
                return '';
            }
        };
PaginationTotalCountFormat.ɵfac = function PaginationTotalCountFormat_Factory(t) { return new (t || PaginationTotalCountFormat)(ɵngcc0.ɵɵdirectiveInject(ɵngcc5.DomSanitizer)); };
PaginationTotalCountFormat.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "paginationTotalPagesFormat", type: PaginationTotalCountFormat, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PaginationTotalCountFormat, [{
        type: core.Pipe,
        args: [{
                name: 'paginationTotalPagesFormat'
            }]
    }], function () { return [{ type: ɵngcc5.DomSanitizer }]; }, null); })();
        return PaginationTotalCountFormat;
    }());
    PaginationTotalCountFormat.ctorParameters = function () { return [
        { type: platformBrowser.DomSanitizer }
    ]; };

    var ThyPaginationModule = /** @class */ (function () {
        function ThyPaginationModule() {
        }
ThyPaginationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyPaginationModule });
ThyPaginationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyPaginationModule_Factory(t) { return new (t || ThyPaginationModule)(); }, imports: [[common.CommonModule, shared.ThySharedModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyPaginationModule, { declarations: [ThyPaginationComponent, PaginationTotalCountFormat], imports: [ɵngcc2.CommonModule, ɵngcc4.ThySharedModule, ɵngcc3.ThyIconModule], exports: [ThyPaginationComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPaginationModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule, shared.ThySharedModule, icon.ThyIconModule],
                exports: [ThyPaginationComponent],
                declarations: [ThyPaginationComponent, PaginationTotalCountFormat]
            }]
    }], function () { return []; }, null); })();
        return ThyPaginationModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.DEFAULT_RANGE_COUNT = DEFAULT_RANGE_COUNT;
    exports.PaginationDefaultConfig = PaginationDefaultConfig;
    exports.PaginationTotalCountFormat = PaginationTotalCountFormat;
    exports.THY_PAGINATION_CONFIG = THY_PAGINATION_CONFIG;
    exports.ThyPaginationComponent = ThyPaginationComponent;
    exports.ThyPaginationModule = ThyPaginationModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
