// ngx-tethys/bundles/ngx-tethys-label.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/util'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/label', ['@angular/core','@angular/common','ngx-tethys/icon','exports', '@angular/core', '@angular/common', 'ngx-tethys/util', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].label = {}), global.ng.core, global.ng.common, global['ngx-tethys'].util, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, util, icon) { 
var _c0 = ["thyLabel", ""];
var _c1 = function (a0) { return { color: a0 }; };
function ThyLabelComponent_thy_icon_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 4);
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.beforeIconName)("ngStyle", ɵngcc0.ɵɵpureFunction1(2, _c1, ctx_r0._color));
} }
function ThyLabelComponent_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 5);
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.beforeIconClass)("ngStyle", ɵngcc0.ɵɵpureFunction1(2, _c1, ctx_r1._color));
} }
function ThyLabelComponent_thy_icon_4_Template(rf, ctx) { if (rf & 1) {
    var _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 6);
    ɵngcc0.ɵɵlistener("click", function ThyLabelComponent_thy_icon_4_Template_thy_icon_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r5); var ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.remove(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r2.afterIconName)("ngStyle", ɵngcc0.ɵɵpureFunction1(2, _c1, ctx_r2._color));
} }
function ThyLabelComponent_i_5_Template(rf, ctx) { if (rf & 1) {
    var _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 7);
    ɵngcc0.ɵɵlistener("click", function ThyLabelComponent_i_5_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); var ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.remove(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.afterIconClass);
} }
var _c2 = ["*", "*"];
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var labelTypeClassesMap = {
        default: ['thy-label', 'thy-label-default'],
        primary: ['thy-label', 'thy-label-primary'],
        success: ['thy-label', 'thy-label-success'],
        info: ['thy-label', 'thy-label-info'],
        warning: ['thy-label', 'thy-label-warning'],
        danger: ['thy-label', 'thy-label-danger']
    };
    var ThyLabelComponent = /** @class */ (function () {
        function ThyLabelComponent(el, renderer) {
            this.el = el;
            this.renderer = renderer;
            this._thyHasHover = false;
            this._classNameSM = false;
            this._classNameDM = false;
            this._classNameLG = false;
            this._typeClassNames = [];
            this._backgroundOpacity = 0.1;
            this.thyOnRemove = new core.EventEmitter();
            this.nativeElement = this.el.nativeElement;
        }
        Object.defineProperty(ThyLabelComponent.prototype, "thySize", {
            set: function (value) {
                this._classNameSM = value === 'sm';
                this._classNameDM = value === 'md';
                this._classNameLG = value === 'lg';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyHasHover", {
            set: function (value) {
                this._thyHasHover = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyLabel", {
            set: function (value) {
                this._type = value;
                this._setClassesByType();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyLabelColor", {
            set: function (color) {
                this._color = color;
                this._setLabelCustomColor();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyBackgroundOpacity", {
            set: function (opacity) {
                if (opacity && opacity > 0 && this._backgroundOpacity !== opacity) {
                    this._backgroundOpacity = opacity;
                    this._setLabelCustomColor();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyLabelType", {
            set: function (labelType) {
                this._labelType = labelType;
                this._setClassesByType();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyBeforeIcon", {
            set: function (icon) {
                this._icon = icon;
                if (this._icon) {
                    if (this._icon.includes('wtf')) {
                        var iconPrefix = this.thyIconPrefix || 'wtf';
                        this.beforeIconClass = [iconPrefix, "" + this._icon];
                    }
                    else {
                        this.beforeIconName = "" + this._icon;
                    }
                }
                else {
                    this.beforeIconClass = null;
                    this.beforeIconName = null;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLabelComponent.prototype, "thyAfterIcon", {
            set: function (icon) {
                this._icon = icon;
                if (this._icon) {
                    if (this._icon.includes('wtf')) {
                        var iconPrefix = this.thyIconPrefix || 'wtf';
                        this.afterIconClass = [iconPrefix, "" + this._icon];
                    }
                    else {
                        this.afterIconName = "" + this._icon;
                    }
                }
                else {
                    this.afterIconClass = null;
                    this.afterIconName = null;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyLabelComponent.prototype._setClassesByType = function () {
            var _this = this;
            var classNames = null;
            if (labelTypeClassesMap[this._type]) {
                classNames = labelTypeClassesMap[this._type];
            }
            else {
                classNames = ['thy-label'];
                classNames.push("thy-label-" + this._type);
            }
            if (this._labelType) {
                classNames = __spread(classNames, ["thy-label-" + this._labelType]);
            }
            // remove old classes
            this._typeClassNames.forEach(function (className) {
                _this._removeClass(className);
            });
            // add new classes
            this._typeClassNames = classNames;
            this._typeClassNames.forEach(function (className) {
                _this._addClass(className);
            });
        };
        ThyLabelComponent.prototype._setLabelCustomColor = function () {
            if (this._color) {
                if (this._type.indexOf('emboss') > -1) {
                    if (this._type === 'emboss-status') {
                        this.el.nativeElement.style.color = '#333';
                    }
                    else {
                        this.el.nativeElement.style.color = this._color;
                    }
                    this.el.nativeElement.style.backgroundColor = util.helpers.hexToRgb(this._color, this._backgroundOpacity);
                }
                else if (this._type.indexOf('outline') > -1) {
                    this.el.nativeElement.style.color = this._color;
                    this.el.nativeElement.style.borderColor = this._color;
                }
                else {
                    this.el.nativeElement.style.backgroundColor = this._color;
                }
            }
        };
        ThyLabelComponent.prototype._addClass = function (className) {
            this.renderer.addClass(this.nativeElement, className);
        };
        ThyLabelComponent.prototype._removeClass = function (className) {
            this.renderer.removeClass(this.nativeElement, className);
        };
        ThyLabelComponent.prototype.remove = function () {
            this.thyOnRemove.emit();
        };
ThyLabelComponent.ɵfac = function ThyLabelComponent_Factory(t) { return new (t || ThyLabelComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyLabelComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyLabelComponent, selectors: [["", "thyLabel", ""]], hostVars: 8, hostBindings: function ThyLabelComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("label-has-hover", ctx._thyHasHover)("thy-label--sm", ctx._classNameSM)("thy-label--md", ctx._classNameDM)("thy-label--lg", ctx._classNameLG);
    } }, inputs: { thySize: "thySize", thyHasHover: "thyHasHover", thyLabel: "thyLabel", thyLabelColor: "thyLabelColor", thyBackgroundOpacity: "thyBackgroundOpacity", thyLabelType: "thyLabelType", thyBeforeIcon: "thyBeforeIcon", thyAfterIcon: "thyAfterIcon", thyIconPrefix: "thyIconPrefix" }, outputs: { thyOnRemove: "thyOnRemove" }, attrs: _c0, ngContentSelectors: _c2, decls: 6, vars: 4, consts: [["class", "mr-1", 3, "thyIconName", "ngStyle", 4, "ngIf"], ["class", "mr-1", 3, "ngClass", "ngStyle", 4, "ngIf"], ["class", "ml-1", 3, "thyIconName", "ngStyle", "click", 4, "ngIf"], ["class", "ml-1", 3, "ngClass", "click", 4, "ngIf"], [1, "mr-1", 3, "thyIconName", "ngStyle"], [1, "mr-1", 3, "ngClass", "ngStyle"], [1, "ml-1", 3, "thyIconName", "ngStyle", "click"], [1, "ml-1", 3, "ngClass", "click"]], template: function ThyLabelComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c2);
        ɵngcc0.ɵɵtemplate(0, ThyLabelComponent_thy_icon_0_Template, 1, 4, "thy-icon", 0);
        ɵngcc0.ɵɵtemplate(1, ThyLabelComponent_i_1_Template, 1, 4, "i", 1);
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵprojection(3, 1);
        ɵngcc0.ɵɵtemplate(4, ThyLabelComponent_thy_icon_4_Template, 1, 4, "thy-icon", 2);
        ɵngcc0.ɵɵtemplate(5, ThyLabelComponent_i_5_Template, 1, 1, "i", 3);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.beforeIconName);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.beforeIconClass);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.afterIconName);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.afterIconClass);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgStyle, ɵngcc1.NgClass], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLabelComponent, [{
        type: core.Component,
        args: [{
                selector: '[thyLabel]',
                template: "<thy-icon class=\"mr-1\" [thyIconName]=\"beforeIconName\" [ngStyle]=\"{ color: _color }\" *ngIf=\"beforeIconName\"></thy-icon>\n<i class=\"mr-1\" [ngClass]=\"beforeIconClass\" [ngStyle]=\"{ color: _color }\" *ngIf=\"beforeIconClass\"></i>\n<ng-content></ng-content>\n<ng-content></ng-content>\n<thy-icon\n  class=\"ml-1\"\n  [thyIconName]=\"afterIconName\"\n  (click)=\"remove()\"\n  [ngStyle]=\"{ color: _color }\"\n  *ngIf=\"afterIconName\"\n></thy-icon>\n<i class=\"ml-1\" [ngClass]=\"afterIconClass\" *ngIf=\"afterIconClass\" (click)=\"remove()\"></i>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { _thyHasHover: [{
            type: core.HostBinding,
            args: ['class.label-has-hover']
        }], _classNameSM: [{
            type: core.HostBinding,
            args: ['class.thy-label--sm']
        }], _classNameDM: [{
            type: core.HostBinding,
            args: ['class.thy-label--md']
        }], _classNameLG: [{
            type: core.HostBinding,
            args: ['class.thy-label--lg']
        }], thyOnRemove: [{
            type: core.Output
        }], thySize: [{
            type: core.Input,
            args: ['thySize']
        }], thyHasHover: [{
            type: core.Input,
            args: ['thyHasHover']
        }], thyLabel: [{
            type: core.Input
        }], thyLabelColor: [{
            type: core.Input
        }], thyBackgroundOpacity: [{
            type: core.Input
        }], thyLabelType: [{
            type: core.Input
        }], thyBeforeIcon: [{
            type: core.Input
        }], thyAfterIcon: [{
            type: core.Input
        }], thyIconPrefix: [{
            type: core.Input
        }] }); })();
        return ThyLabelComponent;
    }());
    ThyLabelComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.Renderer2 }
    ]; };
    ThyLabelComponent.propDecorators = {
        _thyHasHover: [{ type: core.HostBinding, args: ['class.label-has-hover',] }],
        _classNameSM: [{ type: core.HostBinding, args: ['class.thy-label--sm',] }],
        _classNameDM: [{ type: core.HostBinding, args: ['class.thy-label--md',] }],
        _classNameLG: [{ type: core.HostBinding, args: ['class.thy-label--lg',] }],
        thyIconPrefix: [{ type: core.Input }],
        thySize: [{ type: core.Input, args: ['thySize',] }],
        thyOnRemove: [{ type: core.Output }],
        thyHasHover: [{ type: core.Input, args: ['thyHasHover',] }],
        thyLabel: [{ type: core.Input }],
        thyLabelColor: [{ type: core.Input }],
        thyBackgroundOpacity: [{ type: core.Input }],
        thyLabelType: [{ type: core.Input }],
        thyBeforeIcon: [{ type: core.Input }],
        thyAfterIcon: [{ type: core.Input }]
    };

    var ThyLabelModule = /** @class */ (function () {
        function ThyLabelModule() {
        }
ThyLabelModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLabelModule });
ThyLabelModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLabelModule_Factory(t) { return new (t || ThyLabelModule)(); }, imports: [[common.CommonModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLabelModule, { declarations: [ThyLabelComponent], imports: [ɵngcc1.CommonModule, ɵngcc2.ThyIconModule], exports: [ThyLabelComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLabelModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyLabelComponent],
                imports: [common.CommonModule, icon.ThyIconModule],
                exports: [ThyLabelComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyLabelModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyLabelComponent = ThyLabelComponent;
    exports.ThyLabelModule = ThyLabelModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
