// ngx-tethys/bundles/ngx-tethys-divider.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('ngx-tethys/shared'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/shared'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/divider', ['@angular/core','@angular/common','ngx-tethys/shared','ngx-tethys/icon','exports', '@angular/core', '@angular/common', 'ngx-tethys/shared', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ngxTethys.shared,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].divider = {}), global.ng.core, global.ng.common, global['ngx-tethys'].shared, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, shared, icon) { 
function ThyDividerComponent_div_0_1_ng_template_0_Template(rf, ctx) { }
function ThyDividerComponent_div_0_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyDividerComponent_div_0_1_ng_template_0_Template, 0, 0, "ng-template");
} }
function ThyDividerComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵtemplate(1, ThyDividerComponent_div_0_1_Template, 1, 0, undefined, 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.templateContent);
} }
function ThyDividerComponent_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 4);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.textContent);
} }
'use strict';

    var ThyDividerComponent = /** @class */ (function () {
        function ThyDividerComponent() {
            this.thyStyle = 'solid';
            this.thyTextDirection = 'center';
        }
        Object.defineProperty(ThyDividerComponent.prototype, "thyText", {
            set: function (value) {
                if (value instanceof core.TemplateRef) {
                    this.templateContent = value;
                }
                else {
                    this.textContent = value;
                }
            },
            enumerable: false,
            configurable: true
        });
ThyDividerComponent.ɵfac = function ThyDividerComponent_Factory(t) { return new (t || ThyDividerComponent)(); };
ThyDividerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyDividerComponent, selectors: [["thy-divider"]], hostVars: 18, hostBindings: function ThyDividerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-divider", true)("thy-divider-horizontal", !ctx.thyVertical)("thy-divider-vertical", ctx.thyVertical)("thy-divider-with-content", ctx.textContent || ctx.templateContent)("thy-divider-with-content-left", (ctx.textContent || ctx.templateContent) && ctx.thyTextDirection === "left")("thy-divider-with-content-right", (ctx.textContent || ctx.templateContent) && ctx.thyTextDirection === "right")("thy-divider-with-content-center", (ctx.textContent || ctx.templateContent) && ctx.thyTextDirection === "center")("thy-divider-solid", ctx.thyStyle === "solid")("thy-divider-dashed", ctx.thyStyle === "dashed");
    } }, inputs: { thyStyle: "thyStyle", thyTextDirection: "thyTextDirection", thyText: "thyText", thyVertical: "thyVertical" }, decls: 2, vars: 2, consts: [["class", "thy-divider-inner-template", 4, "ngIf"], ["class", "thy-divider-inner-text", 4, "ngIf"], [1, "thy-divider-inner-template"], [4, "ngTemplateOutlet"], [1, "thy-divider-inner-text"]], template: function ThyDividerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyDividerComponent_div_0_Template, 2, 1, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyDividerComponent_span_1_Template, 2, 1, "span", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.templateContent);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.textContent);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDividerComponent, [{
        type: core.Component,
        args: [{
                preserveWhitespaces: false,
                encapsulation: core.ViewEncapsulation.None,
                selector: 'thy-divider',
                template: "\n        <div *ngIf=\"templateContent\" class=\"thy-divider-inner-template\">\n            <ng-template *ngTemplateOutlet=\"templateContent\"></ng-template>\n        </div>\n\n        <span *ngIf=\"textContent\" class=\"thy-divider-inner-text\">{{ textContent }}</span>\n    ",
                host: {
                    '[class.thy-divider]': "true",
                    '[class.thy-divider-horizontal]': "!thyVertical",
                    '[class.thy-divider-vertical]': "thyVertical",
                    '[class.thy-divider-with-content]': "textContent || templateContent",
                    '[class.thy-divider-with-content-left]': "(textContent || templateContent) && thyTextDirection === 'left'",
                    '[class.thy-divider-with-content-right]': "(textContent || templateContent) && thyTextDirection === 'right'",
                    '[class.thy-divider-with-content-center]': "(textContent || templateContent) && thyTextDirection === 'center'",
                    '[class.thy-divider-solid]': "thyStyle === 'solid'",
                    '[class.thy-divider-dashed]': "thyStyle === 'dashed'"
                }
            }]
    }], function () { return []; }, { thyStyle: [{
            type: core.Input
        }], thyTextDirection: [{
            type: core.Input
        }], thyText: [{
            type: core.Input
        }], thyVertical: [{
            type: core.Input
        }] }); })();
        return ThyDividerComponent;
    }());
    ThyDividerComponent.ctorParameters = function () { return []; };
    ThyDividerComponent.propDecorators = {
        thyText: [{ type: core.Input }],
        thyVertical: [{ type: core.Input }],
        thyStyle: [{ type: core.Input }],
        thyTextDirection: [{ type: core.Input }]
    };

    var ThyDividerModule = /** @class */ (function () {
        function ThyDividerModule() {
        }
ThyDividerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDividerModule });
ThyDividerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDividerModule_Factory(t) { return new (t || ThyDividerModule)(); }, providers: [], imports: [[common.CommonModule, shared.ThySharedModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDividerModule, { declarations: [ThyDividerComponent], imports: [ɵngcc1.CommonModule, ɵngcc2.ThySharedModule, ɵngcc3.ThyIconModule], exports: [ThyDividerComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDividerModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyDividerComponent],
                entryComponents: [],
                imports: [common.CommonModule, shared.ThySharedModule, icon.ThyIconModule],
                exports: [ThyDividerComponent],
                providers: []
            }]
    }], function () { return []; }, null); })();
        return ThyDividerModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyDividerComponent = ThyDividerComponent;
    exports.ThyDividerModule = ThyDividerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
