// ngx-tethys/bundles/ngx-tethys-card.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/util')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/card', ['@angular/core','@angular/common','exports', '@angular/core', '@angular/common', 'ngx-tethys/util'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].card = {}), global.ng.core, global.ng.common, global['ngx-tethys'].util));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, common, util) { 
var _c0 = ["headerTitle"];
var _c1 = ["headerDescription"];
var _c2 = ["headerOperation"];
function ThyCardHeaderComponent_ng_template_1_Template(rf, ctx) { }
function ThyCardHeaderComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.thyTitle);
} }
function ThyCardHeaderComponent_ng_template_4_Template(rf, ctx) { }
function ThyCardHeaderComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.thyDescription);
} }
function ThyCardHeaderComponent_ng_template_7_Template(rf, ctx) { }
var _c3 = ["*"];
'use strict';

    var ThyCardHeaderComponent = /** @class */ (function () {
        function ThyCardHeaderComponent() {
            this.thyLayoutHeaderClass = true;
            this._thySizeSm = false;
            this._thySizeLg = false;
        }
        Object.defineProperty(ThyCardHeaderComponent.prototype, "thySize", {
            set: function (value) {
                this._thySizeSm = value === 'sm';
                this._thySizeLg = value === 'lg';
            },
            enumerable: false,
            configurable: true
        });
        ThyCardHeaderComponent.prototype.ngOnInit = function () { };
ThyCardHeaderComponent.ɵfac = function ThyCardHeaderComponent_Factory(t) { return new (t || ThyCardHeaderComponent)(); };
ThyCardHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCardHeaderComponent, selectors: [["thy-card-header"]], contentQueries: function ThyCardHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.descriptionTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationTemplateRef = _t.first);
    } }, hostVars: 6, hostBindings: function ThyCardHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-card-header", ctx.thyLayoutHeaderClass)("thy-card-header--sm", ctx._thySizeSm)("thy-card-header--lg", ctx._thySizeLg);
    } }, inputs: { thySize: "thySize", thyTitle: "thyTitle", thyDescription: "thyDescription" }, decls: 8, vars: 5, consts: [[1, "card-header-title"], [3, "ngTemplateOutlet"], [4, "ngIf"], [1, "card-header-description"], [1, "card-header-operation"]], template: function ThyCardHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyCardHeaderComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(2, ThyCardHeaderComponent_ng_container_2_Template, 3, 1, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyCardHeaderComponent_ng_template_4_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(5, ThyCardHeaderComponent_ng_container_5_Template, 3, 1, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵtemplate(7, ThyCardHeaderComponent_ng_template_7_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.descriptionTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.descriptionTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.operationTemplateRef);
    } }, directives: [ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardHeaderComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-card-header',
                preserveWhitespaces: false,
                template: "<div class=\"card-header-title\">\n  <ng-template [ngTemplateOutlet]=\"titleTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!titleTemplateRef\">\n    <span>{{ thyTitle }}</span>\n  </ng-container>\n</div>\n<div class=\"card-header-description\">\n  <ng-template [ngTemplateOutlet]=\"descriptionTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!descriptionTemplateRef\">\n    <span>{{ thyDescription }}</span>\n  </ng-container>\n</div>\n<div class=\"card-header-operation\">\n  <ng-template [ngTemplateOutlet]=\"operationTemplateRef\"></ng-template>\n</div>\n"
            }]
    }], function () { return []; }, { thyLayoutHeaderClass: [{
            type: core.HostBinding,
            args: ['class.thy-card-header']
        }], _thySizeSm: [{
            type: core.HostBinding,
            args: ['class.thy-card-header--sm']
        }], _thySizeLg: [{
            type: core.HostBinding,
            args: ['class.thy-card-header--lg']
        }], thySize: [{
            type: core.Input,
            args: ['thySize']
        }], thyTitle: [{
            type: core.Input,
            args: ['thyTitle']
        }], thyDescription: [{
            type: core.Input,
            args: ['thyDescription']
        }], titleTemplateRef: [{
            type: core.ContentChild,
            args: ['headerTitle']
        }], descriptionTemplateRef: [{
            type: core.ContentChild,
            args: ['headerDescription']
        }], operationTemplateRef: [{
            type: core.ContentChild,
            args: ['headerOperation']
        }] }); })();
        return ThyCardHeaderComponent;
    }());
    ThyCardHeaderComponent.ctorParameters = function () { return []; };
    ThyCardHeaderComponent.propDecorators = {
        thyLayoutHeaderClass: [{ type: core.HostBinding, args: ['class.thy-card-header',] }],
        _thySizeSm: [{ type: core.HostBinding, args: ['class.thy-card-header--sm',] }],
        _thySizeLg: [{ type: core.HostBinding, args: ['class.thy-card-header--lg',] }],
        thyTitle: [{ type: core.Input, args: ['thyTitle',] }],
        thyDescription: [{ type: core.Input, args: ['thyDescription',] }],
        thySize: [{ type: core.Input, args: ['thySize',] }],
        titleTemplateRef: [{ type: core.ContentChild, args: ['headerTitle',] }],
        descriptionTemplateRef: [{ type: core.ContentChild, args: ['headerDescription',] }],
        operationTemplateRef: [{ type: core.ContentChild, args: ['headerOperation',] }]
    };

    var ThyCardComponent = /** @class */ (function () {
        function ThyCardComponent() {
            this.thyCardClass = true;
            this.clearLeftRightPadding = false;
            this._thyDivided = false;
        }
        Object.defineProperty(ThyCardComponent.prototype, "thyHasLeftRightPadding", {
            set: function (value) {
                this.clearLeftRightPadding = !util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCardComponent.prototype, "thyDivided", {
            set: function (value) {
                this._thyDivided = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
ThyCardComponent.ɵfac = function ThyCardComponent_Factory(t) { return new (t || ThyCardComponent)(); };
ThyCardComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCardComponent, selectors: [["thy-card"]], hostVars: 6, hostBindings: function ThyCardComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-card", ctx.thyCardClass)("thy-card--clear-left-right-padding", ctx.clearLeftRightPadding)("thy-card--divided", ctx._thyDivided);
    } }, inputs: { thyHasLeftRightPadding: "thyHasLeftRightPadding", thyDivided: "thyDivided" }, ngContentSelectors: _c3, decls: 1, vars: 0, template: function ThyCardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-card',
                template: "\n        <ng-content></ng-content>\n    "
            }]
    }], function () { return []; }, { thyCardClass: [{
            type: core.HostBinding,
            args: ['class.thy-card']
        }], clearLeftRightPadding: [{
            type: core.HostBinding,
            args: ['class.thy-card--clear-left-right-padding']
        }], _thyDivided: [{
            type: core.HostBinding,
            args: ['class.thy-card--divided']
        }], thyHasLeftRightPadding: [{
            type: core.Input,
            args: ['thyHasLeftRightPadding']
        }], thyDivided: [{
            type: core.Input,
            args: ['thyDivided']
        }] }); })();
        return ThyCardComponent;
    }());
    ThyCardComponent.propDecorators = {
        thyCardClass: [{ type: core.HostBinding, args: ['class.thy-card',] }],
        clearLeftRightPadding: [{ type: core.HostBinding, args: ['class.thy-card--clear-left-right-padding',] }],
        _thyDivided: [{ type: core.HostBinding, args: ['class.thy-card--divided',] }],
        thyHasLeftRightPadding: [{ type: core.Input, args: ['thyHasLeftRightPadding',] }],
        thyDivided: [{ type: core.Input, args: ['thyDivided',] }]
    };

    var ThyCardContentComponent = /** @class */ (function () {
        function ThyCardContentComponent() {
            this.thyCardContentClass = true;
            this.scrollClassName = false;
            this._thySizeSm = false;
        }
        Object.defineProperty(ThyCardContentComponent.prototype, "thyScroll", {
            set: function (value) {
                this.scrollClassName = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyCardContentComponent.prototype, "thySize", {
            set: function (value) {
                this._thySizeSm = value === 'sm';
            },
            enumerable: false,
            configurable: true
        });
        ThyCardContentComponent.prototype.ngOnInit = function () { };
ThyCardContentComponent.ɵfac = function ThyCardContentComponent_Factory(t) { return new (t || ThyCardContentComponent)(); };
ThyCardContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCardContentComponent, selectors: [["thy-card-content"]], hostVars: 6, hostBindings: function ThyCardContentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-card-content", ctx.thyCardContentClass)("thy-card-content--scroll", ctx.scrollClassName)("thy-card-content--sm", ctx._thySizeSm);
    } }, inputs: { thyScroll: "thyScroll", thySize: "thySize" }, ngContentSelectors: _c3, decls: 1, vars: 0, template: function ThyCardContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardContentComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-card-content',
                preserveWhitespaces: false,
                template: "\n        <ng-content></ng-content>\n    "
            }]
    }], function () { return []; }, { thyCardContentClass: [{
            type: core.HostBinding,
            args: ['class.thy-card-content']
        }], scrollClassName: [{
            type: core.HostBinding,
            args: ['class.thy-card-content--scroll']
        }], _thySizeSm: [{
            type: core.HostBinding,
            args: ['class.thy-card-content--sm']
        }], thyScroll: [{
            type: core.Input,
            args: ['thyScroll']
        }], thySize: [{
            type: core.Input,
            args: ['thySize']
        }] }); })();
        return ThyCardContentComponent;
    }());
    ThyCardContentComponent.propDecorators = {
        thyCardContentClass: [{ type: core.HostBinding, args: ['class.thy-card-content',] }],
        scrollClassName: [{ type: core.HostBinding, args: ['class.thy-card-content--scroll',] }],
        thyScroll: [{ type: core.Input, args: ['thyScroll',] }],
        _thySizeSm: [{ type: core.HostBinding, args: ['class.thy-card-content--sm',] }],
        thySize: [{ type: core.Input, args: ['thySize',] }]
    };

    var ThyCardModule = /** @class */ (function () {
        function ThyCardModule() {
        }
ThyCardModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCardModule });
ThyCardModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCardModule_Factory(t) { return new (t || ThyCardModule)(); }, imports: [[
            common.CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCardModule, { declarations: [ThyCardComponent, ThyCardHeaderComponent, ThyCardContentComponent], imports: [ɵngcc1.CommonModule], exports: [ThyCardComponent, ThyCardHeaderComponent, ThyCardContentComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCardModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    ThyCardComponent,
                    ThyCardHeaderComponent,
                    ThyCardContentComponent,
                ],
                imports: [
                    common.CommonModule
                ],
                exports: [
                    ThyCardComponent,
                    ThyCardHeaderComponent,
                    ThyCardContentComponent,
                ]
            }]
    }], function () { return []; }, null); })();
        return ThyCardModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyCardComponent = ThyCardComponent;
    exports.ThyCardContentComponent = ThyCardContentComponent;
    exports.ThyCardHeaderComponent = ThyCardHeaderComponent;
    exports.ThyCardModule = ThyCardModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
