// ngx-tethys/bundles/ngx-tethys-strength.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),exports, require('@angular/core'), require('@angular/forms'), require('ngx-tethys/core')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/strength', ['@angular/core','ngx-tethys/core','exports', '@angular/core', '@angular/forms', 'ngx-tethys/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].strength = {}), global.ng.core, global.ng.forms, global['ngx-tethys'].core));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, forms, core$1) { 'use strict';

    var _a;
    var ThyStrengthEnum;
    (function (ThyStrengthEnum) {
        ThyStrengthEnum[ThyStrengthEnum["highest"] = 4] = "highest";
        ThyStrengthEnum[ThyStrengthEnum["high"] = 3] = "high";
        ThyStrengthEnum[ThyStrengthEnum["average"] = 2] = "average";
        ThyStrengthEnum[ThyStrengthEnum["low"] = 1] = "low";
    })(ThyStrengthEnum || (ThyStrengthEnum = {}));
    var strengthMap = (_a = {},
        _a[ThyStrengthEnum.highest] = {
            level: 'highest',
            text: '最高'
        },
        _a[ThyStrengthEnum.high] = {
            level: 'high',
            text: '高'
        },
        _a[ThyStrengthEnum.average] = {
            level: 'average',
            text: '中'
        },
        _a[ThyStrengthEnum.low] = {
            level: 'low',
            text: '低'
        },
        _a);
    var ThyStrengthComponent = /** @class */ (function () {
        function ThyStrengthComponent(translate) {
            this.translate = translate;
            this.styleClass = true;
            this.strengthMap = JSON.parse(JSON.stringify(strengthMap));
            this._onChange = Function.prototype;
            this._onTouched = Function.prototype;
        }
        Object.defineProperty(ThyStrengthComponent.prototype, "titleKey", {
            set: function (value) {
                this.strengthTitle = this.translate.instant(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStrengthComponent.prototype, "highestKey", {
            set: function (value) {
                this.strengthMap[ThyStrengthEnum.highest].text = this.translate.instant(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStrengthComponent.prototype, "highKey", {
            set: function (value) {
                this.strengthMap[ThyStrengthEnum.high].text = this.translate.instant(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStrengthComponent.prototype, "averageKey", {
            set: function (value) {
                this.strengthMap[ThyStrengthEnum.average].text = this.translate.instant(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyStrengthComponent.prototype, "lowKey", {
            set: function (value) {
                this.strengthMap[ThyStrengthEnum.low].text = this.translate.instant(value);
            },
            enumerable: false,
            configurable: true
        });
        ThyStrengthComponent.prototype.ngOnInit = function () { };
        ThyStrengthComponent.prototype.writeValue = function (value) {
            this.strength = value;
        };
        ThyStrengthComponent.prototype.registerOnChange = function (fn) {
            this._onChange = fn;
        };
        ThyStrengthComponent.prototype.registerOnTouched = function (fn) {
            this._onTouched = fn;
        };
ThyStrengthComponent.ɵfac = function ThyStrengthComponent_Factory(t) { return new (t || ThyStrengthComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate)); };
ThyStrengthComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStrengthComponent, selectors: [["thy-strength"]], hostVars: 2, hostBindings: function ThyStrengthComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("password-strength-container", ctx.styleClass);
    } }, inputs: { titleKey: "titleKey", highestKey: "highestKey", highKey: "highKey", averageKey: "averageKey", lowKey: "lowKey" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: forms.NG_VALUE_ACCESSOR,
                useExisting: core.forwardRef(function () { return ThyStrengthComponent; }),
                multi: true
            }
        ])], decls: 6, vars: 8, template: function ThyStrengthComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementStart(2, "span");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div");
        ɵngcc0.ɵɵelement(5, "span");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1("", ctx.strengthTitle, " ");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMapInterpolate1("strength-text strength-text-", ctx.strengthMap[ctx.strength] == null ? null : ctx.strengthMap[ctx.strength].level, "");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.strengthMap[ctx.strength] == null ? null : ctx.strengthMap[ctx.strength].text);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMapInterpolate1("strength-level strength-level-", ctx.strengthMap[ctx.strength] == null ? null : ctx.strengthMap[ctx.strength].level, "");
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStrengthComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-strength',
                template: "<div>{{strengthTitle}} <span class=\"strength-text strength-text-{{strengthMap[strength]?.level}}\">{{strengthMap[strength]?.text}}</span></div>\n<div class=\"strength-level strength-level-{{strengthMap[strength]?.level}}\">\n    <span></span>\n</div>\n",
                providers: [
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core.forwardRef(function () { return ThyStrengthComponent; }),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.ThyTranslate }]; }, { styleClass: [{
            type: core.HostBinding,
            args: ['class.password-strength-container']
        }], titleKey: [{
            type: core.Input
        }], highestKey: [{
            type: core.Input
        }], highKey: [{
            type: core.Input
        }], averageKey: [{
            type: core.Input
        }], lowKey: [{
            type: core.Input
        }] }); })();
        return ThyStrengthComponent;
    }());
    ThyStrengthComponent.ctorParameters = function () { return [
        { type: core$1.ThyTranslate }
    ]; };
    ThyStrengthComponent.propDecorators = {
        styleClass: [{ type: core.HostBinding, args: ['class.password-strength-container',] }],
        titleKey: [{ type: core.Input }],
        highestKey: [{ type: core.Input }],
        highKey: [{ type: core.Input }],
        averageKey: [{ type: core.Input }],
        lowKey: [{ type: core.Input }]
    };

    var ThyStrengthModule = /** @class */ (function () {
        function ThyStrengthModule() {
        }
ThyStrengthModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyStrengthModule });
ThyStrengthModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyStrengthModule_Factory(t) { return new (t || ThyStrengthModule)(); }, providers: [], imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyStrengthModule, { declarations: [ThyStrengthComponent], exports: [ThyStrengthComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStrengthModule, [{
        type: core.NgModule,
        args: [{
                imports: [],
                exports: [ThyStrengthComponent],
                declarations: [ThyStrengthComponent],
                providers: []
            }]
    }], function () { return []; }, null); })();
        return ThyStrengthModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyStrengthComponent = ThyStrengthComponent;
    exports.ThyStrengthModule = ThyStrengthModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
