// ngx-tethys/bundles/ngx-tethys-avatar.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/platform-browser'),require('@angular/common'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/util'), require('ngx-tethys/core'), require('@angular/platform-browser'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/avatar', ['@angular/core','ngx-tethys/core','@angular/platform-browser','@angular/common','ngx-tethys/icon','exports', '@angular/core', '@angular/common', 'ngx-tethys/util', 'ngx-tethys/core', '@angular/platform-browser', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.platformBrowser,global.ng.common,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].avatar = {}), global.ng.core, global.ng.common, global['ngx-tethys'].util, global['ngx-tethys'].core, global.ng.platformBrowser, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,exports, core, common, util, core$1, platformBrowser, icon) { 
function ThyAvatarComponent_img_0_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "img", 5);
    ɵngcc0.ɵɵlistener("error", function ThyAvatarComponent_img_0_Template_img_error_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); var ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.avatarImgError($event); });
    ɵngcc0.ɵɵpipe(1, "thyAvatarSrc");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("alt", ctx_r0.avatarName || "");
    ɵngcc0.ɵɵproperty("src", ɵngcc0.ɵɵpipeBind2(1, 3, ctx_r0._src, ctx_r0._size), ɵngcc0.ɵɵsanitizeUrl)("ngClass", ctx_r0.thyImgClass);
} }
function ThyAvatarComponent_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 6);
    ɵngcc0.ɵɵpipe(1, "avatarBgColor");
    ɵngcc0.ɵɵelementStart(2, "div");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "avatarShortName");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r1.avatarName));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r1.avatarName));
} }
function ThyAvatarComponent_ng_container_2_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r7.avatarName);
} }
function ThyAvatarComponent_ng_container_2_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 10);
} if (rf & 2) {
    var ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("innerHtml", ctx_r8.avatarNameSafeHtml, ɵngcc0.ɵɵsanitizeHtml);
} }
function ThyAvatarComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyAvatarComponent_ng_container_2_div_1_Template, 2, 1, "div", 7);
    ɵngcc0.ɵɵtemplate(2, ThyAvatarComponent_ng_container_2_div_2_Template, 1, 1, "div", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.avatarNameSafeHtml);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.avatarNameSafeHtml);
} }
function ThyAvatarComponent_a_3_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 11);
    ɵngcc0.ɵɵlistener("click", function ThyAvatarComponent_a_3_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); var ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.remove($event); });
    ɵngcc0.ɵɵelement(1, "thy-icon", 12);
    ɵngcc0.ɵɵelementEnd();
} }
function ThyAvatarComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelement(1, "thy-icon", 14);
    ɵngcc0.ɵɵelementEnd();
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ThyAvatarService = /** @class */ (function () {
        function ThyAvatarService() {
            this.ignoreAvatarSrcPaths = ["default.png"];
        }
        return ThyAvatarService;
    }());
    var ThyDefaultAvatarService = /** @class */ (function (_super) {
        __extends(ThyDefaultAvatarService, _super);
        function ThyDefaultAvatarService() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ThyDefaultAvatarService.prototype.avatarSrcTransform = function (src, size) {
            return src;
        };
        ThyDefaultAvatarService.prototype.nameTransform = function (name) {
            return name;
        };
ThyDefaultAvatarService.ɵfac = function ThyDefaultAvatarService_Factory(t) { return ɵThyDefaultAvatarService_BaseFactory(t || ThyDefaultAvatarService); };
ThyDefaultAvatarService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyDefaultAvatarService, factory: function (t) { return ThyDefaultAvatarService.ɵfac(t); } });
var ɵThyDefaultAvatarService_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(ThyDefaultAvatarService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDefaultAvatarService, [{
        type: core.Injectable
    }], null, null); })();
        return ThyDefaultAvatarService;
    }(ThyAvatarService));

    var sizeArray = [22, 24, 28, 32, 36, 48, 68, 110, 160];
    var DEFAULT_SIZE = 36;
    var thyAvatarSizeMap = {
        xxs: 22,
        xs: 24,
        sm: 32,
        md: 36,
        lg: 48
    };
    var ThyAvatarComponent = /** @class */ (function () {
        function ThyAvatarComponent(updateHostClassService, elementRef, thyAvatarService, domSanitizer) {
            this.updateHostClassService = updateHostClassService;
            this.elementRef = elementRef;
            this.thyAvatarService = thyAvatarService;
            this.domSanitizer = domSanitizer;
            this._showRemove = false;
            this._isAvatar = true;
            this.thyOnRemove = new core.EventEmitter();
            this.thyError = new core.EventEmitter();
            updateHostClassService.initializeElement(elementRef.nativeElement);
        }
        Object.defineProperty(ThyAvatarComponent.prototype, "thySrc", {
            set: function (value) {
                this._setAvatarSrc(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyAvatarComponent.prototype, "thyName", {
            set: function (value) {
                // this._name = value;
                this._setAvatarName(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyAvatarComponent.prototype, "thySize", {
            set: function (value) {
                if (thyAvatarSizeMap[value]) {
                    this._setAvatarSize(thyAvatarSizeMap[value]);
                }
                else {
                    this._setAvatarSize(value * 1);
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyAvatarComponent.prototype, "thyShowRemove", {
            set: function (value) {
                this._showRemove = util.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        ThyAvatarComponent.prototype._setAvatarSize = function (size) {
            if (sizeArray.indexOf(size) > -1) {
                this._size = size;
            }
            else {
                this._size = this.findClosestSize(sizeArray, size);
            }
        };
        ThyAvatarComponent.prototype.findClosestSize = function (sizes, value) {
            var left = 0, right = sizes.length - 1, middle, result;
            while (left <= right) {
                middle = Math.floor((left + right) / 2);
                if (right - left <= 1) {
                    result = sizes[right];
                    break;
                }
                result = sizes[middle];
                if (result === value) {
                    return value;
                }
                else if (result > value) {
                    right = middle;
                }
                else {
                    left = middle;
                }
            }
            return value - sizes[left] < sizes[right] - value ? sizes[left] : sizes[right];
        };
        ThyAvatarComponent.prototype._setAvatarSrc = function (src) {
            if (src && this.thyAvatarService.ignoreAvatarSrcPaths.indexOf(src) < 0) {
                this._src = src;
            }
            else {
                this._src = null;
            }
        };
        ThyAvatarComponent.prototype._setAvatarName = function (value) {
            var name = this.thyAvatarService.nameTransform(value);
            if (util.isString(name)) {
                this.avatarName = name;
            }
            else {
                this.avatarName = value;
                this.avatarNameSafeHtml = name;
            }
        };
        ThyAvatarComponent.prototype.ngOnInit = function () {
            if (!this._size) {
                this._setAvatarSize(DEFAULT_SIZE);
            }
            this.updateHostClassService.updateClass(["thy-avatar-" + this._size]);
        };
        ThyAvatarComponent.prototype.remove = function ($event) {
            this.thyOnRemove.emit($event);
        };
        ThyAvatarComponent.prototype.avatarImgError = function ($event) {
            this._setAvatarSrc(null);
            this.thyError.emit($event);
        };
ThyAvatarComponent.ɵfac = function ThyAvatarComponent_Factory(t) { return new (t || ThyAvatarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ThyAvatarService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DomSanitizer)); };
ThyAvatarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyAvatarComponent, selectors: [["thy-avatar"]], hostVars: 2, hostBindings: function ThyAvatarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-avatar", ctx._isAvatar);
    } }, inputs: { thySrc: "thySrc", thyName: "thyName", thySize: "thySize", thyShowRemove: "thyShowRemove", thyShowName: "thyShowName", thyImgClass: "thyImgClass", thyDisabled: "thyDisabled" }, outputs: { thyOnRemove: "thyOnRemove", thyError: "thyError" }, features: [ɵngcc0.ɵɵProvidersFeature([core$1.UpdateHostClassService])], decls: 5, vars: 5, consts: [["class", "avatar-avatar", 3, "src", "ngClass", "alt", "error", 4, "ngIf"], ["class", "avatar-default", 3, "ngStyle", 4, "ngIf"], [4, "ngIf"], ["href", "javascript:;", "class", "remove-link avatar-remove", 3, "click", 4, "ngIf"], ["class", "thy-avatar-disabled", 4, "ngIf"], [1, "avatar-avatar", 3, "src", "ngClass", "alt", "error"], [1, "avatar-default", 3, "ngStyle"], ["class", "avatar-name", 4, "ngIf"], ["class", "avatar-name", 3, "innerHtml", 4, "ngIf"], [1, "avatar-name"], [1, "avatar-name", 3, "innerHtml"], ["href", "javascript:;", 1, "remove-link", "avatar-remove", 3, "click"], ["thyIconName", "close-circle-bold-fill", 1, "remove-link-icon"], [1, "thy-avatar-disabled"], ["thyIconName", "ban", 1, "thy-avatar-disabled-icon"]], template: function ThyAvatarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyAvatarComponent_img_0_Template, 2, 6, "img", 0);
        ɵngcc0.ɵɵtemplate(1, ThyAvatarComponent_span_1_Template, 5, 6, "span", 1);
        ɵngcc0.ɵɵtemplate(2, ThyAvatarComponent_ng_container_2_Template, 3, 2, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(3, ThyAvatarComponent_a_3_Template, 2, 0, "a", 3);
        ɵngcc0.ɵɵtemplate(4, ThyAvatarComponent_div_4_Template, 2, 0, "div", 4);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx._src);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx._src);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowName);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx._showRemove);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyDisabled);
    } }, directives: function () { return [ɵngcc3.NgIf, ɵngcc3.NgClass, ɵngcc3.NgStyle, ɵngcc4.ThyIconComponent]; }, pipes: function () { return [AvatarSrcPipe, AvatarBgColorPipe, AvatarShortNamePipe]; }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAvatarComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-avatar',
                template: "<img\n  *ngIf=\"_src\"\n  [src]=\"_src | thyAvatarSrc: _size\"\n  class=\"avatar-avatar\"\n  [ngClass]=\"thyImgClass\"\n  alt=\"{{ avatarName || '' }}\"\n  (error)=\"avatarImgError($event)\"\n/>\n<span *ngIf=\"!_src\" class=\"avatar-default\" [ngStyle]=\"avatarName | avatarBgColor\">\n  <div>{{ avatarName | avatarShortName }}</div>\n</span>\n<ng-container *ngIf=\"thyShowName\">\n  <div *ngIf=\"!avatarNameSafeHtml\" class=\"avatar-name\">{{ avatarName }}</div>\n  <div *ngIf=\"avatarNameSafeHtml\" class=\"avatar-name\" [innerHtml]=\"avatarNameSafeHtml\"></div>\n</ng-container>\n<a *ngIf=\"_showRemove\" (click)=\"remove($event)\" href=\"javascript:;\" class=\"remove-link avatar-remove\"\n  ><thy-icon class=\"remove-link-icon\" thyIconName=\"close-circle-bold-fill\"></thy-icon\n></a>\n<div *ngIf=\"thyDisabled\" class=\"thy-avatar-disabled\">\n  <thy-icon class=\"thy-avatar-disabled-icon\" thyIconName=\"ban\"></thy-icon>\n</div>\n",
                providers: [core$1.UpdateHostClassService],
                changeDetection: core.ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }, { type: ThyAvatarService }, { type: ɵngcc2.DomSanitizer }]; }, { _isAvatar: [{
            type: core.HostBinding,
            args: ['class.thy-avatar']
        }], thyOnRemove: [{
            type: core.Output
        }], thyError: [{
            type: core.Output
        }], thySrc: [{
            type: core.Input
        }], thyName: [{
            type: core.Input
        }], thySize: [{
            type: core.Input
        }], thyShowRemove: [{
            type: core.Input
        }], thyShowName: [{
            type: core.Input
        }], thyImgClass: [{
            type: core.Input
        }], thyDisabled: [{
            type: core.Input
        }] }); })();
        return ThyAvatarComponent;
    }());
    ThyAvatarComponent.ctorParameters = function () { return [
        { type: core$1.UpdateHostClassService },
        { type: core.ElementRef },
        { type: ThyAvatarService },
        { type: platformBrowser.DomSanitizer }
    ]; };
    ThyAvatarComponent.propDecorators = {
        _isAvatar: [{ type: core.HostBinding, args: ['class.thy-avatar',] }],
        thyOnRemove: [{ type: core.Output }],
        thyError: [{ type: core.Output }],
        thyShowName: [{ type: core.Input }],
        thySrc: [{ type: core.Input }],
        thyName: [{ type: core.Input }],
        thySize: [{ type: core.Input }],
        thyShowRemove: [{ type: core.Input }],
        thyImgClass: [{ type: core.Input }],
        thyDisabled: [{ type: core.Input }]
    };

    var AvatarShortNamePipe = /** @class */ (function () {
        function AvatarShortNamePipe() {
        }
        AvatarShortNamePipe.prototype.transform = function (name) {
            if (!name) {
                return;
            }
            name = name.trim();
            if (/^[\u4e00-\u9fa5]+$/.test(name)) {
                if (name.length > 2) {
                    return name.substr(name.length - 2, 2);
                }
            }
            if (/^[a-zA-Z\/ ]+$/.test(name)) {
                if (name.indexOf(' ') > 0) {
                    var ens = name.split(' ');
                    return (ens[0].substr(0, 1) + ens[1].substr(0, 1)).toUpperCase();
                }
            }
            if (name.length > 2) {
                return name.substr(0, 2).toUpperCase();
            }
            return name.toUpperCase();
        };
AvatarShortNamePipe.ɵfac = function AvatarShortNamePipe_Factory(t) { return new (t || AvatarShortNamePipe)(); };
AvatarShortNamePipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "avatarShortName", type: AvatarShortNamePipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvatarShortNamePipe, [{
        type: core.Pipe,
        args: [{ name: 'avatarShortName' }]
    }], function () { return []; }, null); })();
        return AvatarShortNamePipe;
    }());
    var AvatarBgColorPipe = /** @class */ (function () {
        function AvatarBgColorPipe() {
        }
        AvatarBgColorPipe.prototype.transform = function (name) {
            if (!name) {
                return;
            }
            var colors = [
                '#56abfb',
                '#5dcfff',
                '#84e17e',
                '#73d897',
                '#ff9f73',
                '#fa8888',
                '#fb7fb7',
                '#9a7ef4',
                '#868af6'
            ];
            var nameArray = name.split('');
            var code = name && name.length > 0
                ? nameArray.reduce(function (result, item) {
                    result.value += item.charCodeAt(0);
                    return result;
                }, { value: 0 }).value
                : 0;
            return {
                'background-color': colors[code % 9]
            };
        };
AvatarBgColorPipe.ɵfac = function AvatarBgColorPipe_Factory(t) { return new (t || AvatarBgColorPipe)(); };
AvatarBgColorPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "avatarBgColor", type: AvatarBgColorPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvatarBgColorPipe, [{
        type: core.Pipe,
        args: [{ name: 'avatarBgColor' }]
    }], function () { return []; }, null); })();
        return AvatarBgColorPipe;
    }());
    var AvatarSrcPipe = /** @class */ (function () {
        function AvatarSrcPipe(thyAvatarService) {
            this.thyAvatarService = thyAvatarService;
        }
        AvatarSrcPipe.prototype.transform = function (src, size) {
            return this.thyAvatarService.avatarSrcTransform(src, size);
        };
AvatarSrcPipe.ɵfac = function AvatarSrcPipe_Factory(t) { return new (t || AvatarSrcPipe)(ɵngcc0.ɵɵdirectiveInject(ThyAvatarService)); };
AvatarSrcPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "thyAvatarSrc", type: AvatarSrcPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvatarSrcPipe, [{
        type: core.Pipe,
        args: [{ name: 'thyAvatarSrc' }]
    }], function () { return [{ type: ThyAvatarService }]; }, null); })();
        return AvatarSrcPipe;
    }());
    AvatarSrcPipe.ctorParameters = function () { return [
        { type: ThyAvatarService }
    ]; };
    var AvatarPipes = [AvatarShortNamePipe, AvatarBgColorPipe, AvatarSrcPipe];

    var ThyAvatarModule = /** @class */ (function () {
        function ThyAvatarModule() {
        }
ThyAvatarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAvatarModule });
ThyAvatarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAvatarModule_Factory(t) { return new (t || ThyAvatarModule)(); }, providers: [
        {
            provide: ThyAvatarService,
            useClass: ThyDefaultAvatarService
        }
    ], imports: [[common.CommonModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAvatarModule, { declarations: [ThyAvatarComponent, AvatarShortNamePipe, AvatarBgColorPipe, AvatarSrcPipe], imports: [ɵngcc3.CommonModule, ɵngcc4.ThyIconModule], exports: [ThyAvatarComponent, AvatarShortNamePipe, AvatarBgColorPipe, AvatarSrcPipe] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAvatarModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyAvatarComponent, AvatarPipes],
                imports: [common.CommonModule, icon.ThyIconModule],
                providers: [
                    {
                        provide: ThyAvatarService,
                        useClass: ThyDefaultAvatarService
                    }
                ],
                exports: [ThyAvatarComponent, AvatarPipes]
            }]
    }], function () { return []; }, null); })();
        return ThyAvatarModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AvatarBgColorPipe = AvatarBgColorPipe;
    exports.AvatarPipes = AvatarPipes;
    exports.AvatarShortNamePipe = AvatarShortNamePipe;
    exports.AvatarSrcPipe = AvatarSrcPipe;
    exports.ThyAvatarComponent = ThyAvatarComponent;
    exports.ThyAvatarModule = ThyAvatarModule;
    exports.ThyAvatarService = ThyAvatarService;
    exports.ThyDefaultAvatarService = ThyDefaultAvatarService;
    exports.thyAvatarSizeMap = thyAvatarSizeMap;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
