// ngx-tethys/bundles/ngx-tethys-notify.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('@angular/animations'), require('ngx-tethys/core'), require('ngx-tethys/store'), require('ngx-tethys/util'), require('rxjs'), require('@angular/cdk/portal'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/notify', ['@angular/core','@angular/common','ngx-tethys/icon','exports', '@angular/core', '@angular/common', '@angular/animations', 'ngx-tethys/core', 'ngx-tethys/store', 'ngx-tethys/util', 'rxjs', '@angular/cdk/portal', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].notify = {}), global.ng.core, global.ng.common, global.ng.animations, global['ngx-tethys'].core, global['ngx-tethys'].store, global['ngx-tethys'].util, global.rxjs, global.ng.cdk.portal, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, animations, core$1, store, util, rxjs, portal, icon) { 
function ThyNotifyComponent_ng_container_0_div_8_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 11);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_div_8_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); var ctx_r5 = ɵngcc0.ɵɵnextContext(3); return ctx_r5.triggerDetail(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r4.option == null ? null : ctx_r4.option.detail["link"]);
} }
var _c0 = function (a0) { return { "thy-notify-content--extend": a0 }; };
function ThyNotifyComponent_ng_container_0_div_8_Template(rf, ctx) { if (rf & 1) {
    var _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_div_8_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.extendContent(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, ThyNotifyComponent_ng_container_0_div_8_a_2_Template, 2, 1, "a", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx_r2.extendContentClass === true));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.option == null ? null : ctx_r2.option.content, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.option == null ? null : ctx_r2.option.detail);
} }
function ThyNotifyComponent_ng_container_0_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.option == null ? null : ctx_r3.option.detail["content"]);
} }
function ThyNotifyComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r10); var ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.closeNotify(); });
    ɵngcc0.ɵɵelement(2, "thy-icon", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 3);
    ɵngcc0.ɵɵelement(4, "thy-icon", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 5);
    ɵngcc0.ɵɵelementStart(6, "div", 6);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(8, ThyNotifyComponent_ng_container_0_div_8_Template, 3, 5, "div", 7);
    ɵngcc0.ɵɵtemplate(9, ThyNotifyComponent_ng_container_0_div_9_Template, 2, 1, "div", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.notifyIconName);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.option == null ? null : ctx_r0.option.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.option == null ? null : ctx_r0.option.detail) || (ctx_r0.option == null ? null : ctx_r0.option.content));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isShowDetail);
} }
function ThyNotifyComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    var _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_1_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.closeNotify(); });
    ɵngcc0.ɵɵelement(2, "thy-icon", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵelement(4, "template", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.option == null ? null : ctx_r1.option.html);
} }
function ThyNotifyContainerComponent_thy_notify_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    var item_r4 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r4);
} }
function ThyNotifyContainerComponent_thy_notify_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    var item_r5 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r5);
} }
function ThyNotifyContainerComponent_thy_notify_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    var item_r6 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r6);
} }
function ThyNotifyContainerComponent_thy_notify_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    var item_r7 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r7);
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var notifyQueueInitialState = {
        topLeftQueue: [],
        topRightQueue: [],
        bottomLeftQueue: [],
        bottomRightQueue: []
    };
    var NotifyQueueStore = /** @class */ (function (_super) {
        __extends(NotifyQueueStore, _super);
        function NotifyQueueStore() {
            return _super.call(this, notifyQueueInitialState) || this;
        }
        NotifyQueueStore.topRightSelector = function (state) {
            return state.topRightQueue;
        };
        NotifyQueueStore.topLeftSelector = function (state) {
            return state.topLeftQueue;
        };
        NotifyQueueStore.bottomLeftSelector = function (state) {
            return state.bottomLeftQueue;
        };
        NotifyQueueStore.bottomRightSelector = function (state) {
            return state.bottomRightQueue;
        };
        NotifyQueueStore.prototype.convertQueueKey = function (placement) {
            var key;
            if (placement === 'topLeft') {
                key = 'topLeftQueue';
            }
            else if (placement === 'topRight') {
                key = 'topRightQueue';
            }
            else if (placement === 'bottomLeft') {
                key = 'bottomLeftQueue';
            }
            else if (placement === 'bottomRight') {
                key = 'bottomRightQueue';
            }
            return key;
        };
        NotifyQueueStore.prototype.addNotify = function (placement, options) {
            var key = this.convertQueueKey(placement);
            var state = this.snapshot;
            if (state[key].length > options.maxStack) {
                state[key].shift();
            }
            state[key].push(options);
            this.next(state);
        };
        NotifyQueueStore.prototype.removeNotify = function (id, placement) {
            var state = this.snapshot;
            if (placement) {
                var queueKey = this.convertQueueKey(placement);
                if (state.hasOwnProperty(queueKey) && state[queueKey].length) {
                    state[queueKey] = state[queueKey].filter(function (item) {
                        return item.id !== id;
                    });
                }
            }
            else {
                Object.keys(state).forEach(function (queueKey) {
                    if (state.hasOwnProperty(queueKey) && state[queueKey].length) {
                        state[queueKey] = state[queueKey].filter(function (item) {
                            return item.id !== id;
                        });
                    }
                });
            }
            this.next(state);
        };
NotifyQueueStore.ɵfac = function NotifyQueueStore_Factory(t) { return new (t || NotifyQueueStore)(); };
NotifyQueueStore.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NotifyQueueStore, factory: function (t) { return NotifyQueueStore.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotifyQueueStore, [{
        type: core.Injectable
    }], function () { return []; }, null); })();
        return NotifyQueueStore;
    }(store.Store));
    NotifyQueueStore.ctorParameters = function () { return []; };
    __decorate([
        store.Action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String, Object]),
        __metadata("design:returntype", void 0)
    ], NotifyQueueStore.prototype, "addNotify", null);
    __decorate([
        store.Action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number, String]),
        __metadata("design:returntype", void 0)
    ], NotifyQueueStore.prototype, "removeNotify", null);

    var ANIMATION_IN_DURATION = 100;
    var ANIMATION_OUT_DURATION = 150;
    var HIDE_STYLE = { transform: 'translateX(0)', opacity: 0, height: 0, paddingTop: 0, paddingBottom: 0, margin: 0 };
    var ThyNotifyComponent = /** @class */ (function () {
        function ThyNotifyComponent(_queueStore, _ngZone) {
            this._queueStore = _queueStore;
            this._ngZone = _ngZone;
            this.className = '';
            this.notifyIconName = '';
            this.extendContentClass = false;
            this.isShowDetail = false;
        }
        Object.defineProperty(ThyNotifyComponent.prototype, "thyOption", {
            set: function (value) {
                this.option = value;
                var type = value.type;
                this.placement = value.placement || 'topRight';
                if (this.placement === 'topLeft' || this.placement === 'bottomLeft') {
                    this.flyInOut = 'flyInOutLeft';
                }
                else {
                    this.flyInOut = 'flyInOutRight';
                }
                this.className = "thy-notify thy-notify-" + type;
            },
            enumerable: false,
            configurable: true
        });
        ThyNotifyComponent.prototype.ngOnInit = function () {
            var iconName = {
                success: 'check-circle-fill',
                info: 'info-circle-fill',
                warning: 'waring-fill',
                error: 'close-circle-fill'
            };
            this.notifyIconName = iconName[this.option.type];
            this._creatCloseTimer();
        };
        ThyNotifyComponent.prototype.ngOnDestroy = function () {
            this._clearCloseTimer();
        };
        ThyNotifyComponent.prototype.extendContent = function () {
            this.extendContentClass = true;
        };
        ThyNotifyComponent.prototype.showDetailToggle = function () {
            this.isShowDetail = !this.isShowDetail;
        };
        ThyNotifyComponent.prototype.closeNotify = function () {
            var _this = this;
            this._ngZone.runOutsideAngular(function () {
                _this.flyInOut = 'componentHide';
                setTimeout(function () {
                    _this._queueStore.removeNotify(_this.option.id, _this.placement);
                }, ANIMATION_OUT_DURATION);
            });
        };
        ThyNotifyComponent.prototype.triggerDetail = function () {
            if (util.helpers.isFunction(this.option.detail.action)) {
                this.option.detail.action();
            }
            if (this.option.detail.content) {
                this.showDetailToggle();
            }
        };
        ThyNotifyComponent.prototype.mouseenter = function () {
            if (this.option.pauseOnHover) {
                this._clearCloseTimer();
            }
        };
        ThyNotifyComponent.prototype.mouseleave = function () {
            if (this.option.pauseOnHover) {
                this._creatCloseTimer();
            }
        };
        ThyNotifyComponent.prototype._creatCloseTimer = function () {
            var _this = this;
            if (this.option.duration) {
                this.closeTimer = setInterval(function () {
                    clearInterval(_this.closeTimer);
                    _this.closeNotify();
                }, this.option.duration);
            }
        };
        ThyNotifyComponent.prototype._clearCloseTimer = function () {
            clearInterval(this.closeTimer);
        };
ThyNotifyComponent.ɵfac = function ThyNotifyComponent_Factory(t) { return new (t || ThyNotifyComponent)(ɵngcc0.ɵɵdirectiveInject(NotifyQueueStore), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyNotifyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNotifyComponent, selectors: [["thy-notify"]], hostVars: 3, hostBindings: function ThyNotifyComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mouseenter", function ThyNotifyComponent_mouseenter_HostBindingHandler() { return ctx.mouseenter(); })("mouseleave", function ThyNotifyComponent_mouseleave_HostBindingHandler() { return ctx.mouseleave(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵsyntheticHostProperty("@flyInOut", ctx.flyInOut);
        ɵngcc0.ɵɵclassMap(ctx.className);
    } }, inputs: { thyOption: "thyOption" }, features: [ɵngcc0.ɵɵProvidersFeature([core$1.UpdateHostClassService])], decls: 2, vars: 2, consts: [[4, "ngIf"], ["href", "javascript:;", 1, "thy-notify-close", 3, "click"], ["thyIconName", "close"], [1, "thy-notify-icon-container"], [3, "thyIconName"], [1, "thy-notify-main"], [1, "thy-notify-title"], ["class", "thy-notify-content", 3, "ngClass", "click", 4, "ngIf"], ["class", "thy-notify-detail", 4, "ngIf"], [1, "thy-notify-content", 3, "ngClass", "click"], ["href", "javascript:;", "class", "link-secondary", 3, "click", 4, "ngIf"], ["href", "javascript:;", 1, "link-secondary", 3, "click"], [1, "thy-notify-detail"], [3, "ngTemplateOutlet"]], template: function ThyNotifyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyNotifyComponent_ng_container_0_Template, 10, 4, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, ThyNotifyComponent_ng_container_1_Template, 5, 1, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.option.html);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.option == null ? null : ctx.option.html);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass, ɵngcc1.NgTemplateOutlet], encapsulation: 2, data: { animation: [
            animations.trigger('flyInOut', [
                animations.state('flyInOutRight', animations.style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                animations.transition('void => flyInOutRight', [
                    animations.style({ transform: 'translateX(100%)', opacity: 0, height: '*' }),
                    animations.animate(ANIMATION_IN_DURATION)
                ]),
                animations.transition('flyInOutRight => componentHide', [animations.animate(ANIMATION_OUT_DURATION, animations.style(HIDE_STYLE))]),
                animations.transition('flyInOutRight => void', [animations.animate(ANIMATION_IN_DURATION, animations.style(HIDE_STYLE))]),
                animations.state('flyInOutLeft', animations.style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                animations.transition('void => flyInOutLeft', [
                    animations.style({ transform: 'translateX(-100%)', opacity: 0, height: '*' }),
                    animations.animate(ANIMATION_IN_DURATION)
                ]),
                animations.transition('flyInOutLeft => componentHide', [animations.animate(ANIMATION_OUT_DURATION, animations.style(HIDE_STYLE))]),
                animations.transition('flyInOutLeft => void', [animations.animate(ANIMATION_IN_DURATION, animations.style(HIDE_STYLE))]),
                animations.state('componentHide', animations.style(HIDE_STYLE))
            ])
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-notify',
                template: "<ng-container *ngIf=\"!option.html\">\n  <a href=\"javascript:;\" class=\"thy-notify-close\" (click)=\"closeNotify()\">\n    <thy-icon thyIconName=\"close\"></thy-icon>\n  </a>\n  <div class=\"thy-notify-icon-container\">\n    <thy-icon [thyIconName]=\"notifyIconName\"></thy-icon>\n  </div>\n  <div class=\"thy-notify-main\">\n    <div class=\"thy-notify-title\">{{ option?.title }}</div>\n    <div\n      *ngIf=\"option?.detail || option?.content\"\n      class=\"thy-notify-content\"\n      [ngClass]=\"{ 'thy-notify-content--extend': extendContentClass === true }\"\n      (click)=\"extendContent()\"\n    >\n      {{ option?.content }}\n      <a href=\"javascript:;\" *ngIf=\"option?.detail\" class=\"link-secondary\" (click)=\"triggerDetail()\">{{ (option?.detail)['link'] }}</a>\n    </div>\n    <div class=\"thy-notify-detail\" *ngIf=\"isShowDetail\">{{ (option?.detail)['content'] }}</div>\n  </div>\n</ng-container>\n<ng-container *ngIf=\"option?.html\">\n  <a href=\"javascript:;\" class=\"thy-notify-close\" (click)=\"closeNotify()\">\n    <thy-icon thyIconName=\"close\"></thy-icon>\n  </a>\n  <div class=\"thy-notify-main\">\n    <template [ngTemplateOutlet]=\"option?.html\"></template>\n  </div>\n</ng-container>\n",
                providers: [core$1.UpdateHostClassService],
                animations: [
                    animations.trigger('flyInOut', [
                        animations.state('flyInOutRight', animations.style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                        animations.transition('void => flyInOutRight', [
                            animations.style({ transform: 'translateX(100%)', opacity: 0, height: '*' }),
                            animations.animate(ANIMATION_IN_DURATION)
                        ]),
                        animations.transition('flyInOutRight => componentHide', [animations.animate(ANIMATION_OUT_DURATION, animations.style(HIDE_STYLE))]),
                        animations.transition('flyInOutRight => void', [animations.animate(ANIMATION_IN_DURATION, animations.style(HIDE_STYLE))]),
                        animations.state('flyInOutLeft', animations.style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                        animations.transition('void => flyInOutLeft', [
                            animations.style({ transform: 'translateX(-100%)', opacity: 0, height: '*' }),
                            animations.animate(ANIMATION_IN_DURATION)
                        ]),
                        animations.transition('flyInOutLeft => componentHide', [animations.animate(ANIMATION_OUT_DURATION, animations.style(HIDE_STYLE))]),
                        animations.transition('flyInOutLeft => void', [animations.animate(ANIMATION_IN_DURATION, animations.style(HIDE_STYLE))]),
                        animations.state('componentHide', animations.style(HIDE_STYLE))
                    ])
                ]
            }]
    }], function () { return [{ type: NotifyQueueStore }, { type: ɵngcc0.NgZone }]; }, { className: [{
            type: core.HostBinding,
            args: ['class']
        }], thyOption: [{
            type: core.Input
        }], mouseenter: [{
            type: core.HostListener,
            args: ['mouseenter']
        }], mouseleave: [{
            type: core.HostListener,
            args: ['mouseleave']
        }], flyInOut: [{
            type: core.HostBinding,
            args: ['@flyInOut']
        }] }); })();
        return ThyNotifyComponent;
    }());
    ThyNotifyComponent.ctorParameters = function () { return [
        { type: NotifyQueueStore },
        { type: core.NgZone }
    ]; };
    ThyNotifyComponent.propDecorators = {
        flyInOut: [{ type: core.HostBinding, args: ['@flyInOut',] }],
        className: [{ type: core.HostBinding, args: ['class',] }],
        thyOption: [{ type: core.Input }],
        mouseenter: [{ type: core.HostListener, args: ['mouseenter',] }],
        mouseleave: [{ type: core.HostListener, args: ['mouseleave',] }]
    };

    var ThyNotifyContainerComponent = /** @class */ (function () {
        function ThyNotifyContainerComponent(queueStore) {
            this.queueStore = queueStore;
            this.className = true;
        }
        ThyNotifyContainerComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.placement = this.initialState.placement;
            var queue$, queueKey;
            if (this.placement === 'bottomRight') {
                this.bottomRight = true;
            }
            else if (this.placement === 'bottomLeft') {
                this.bottomLeft = true;
            }
            else if (this.placement === 'topLeft') {
                this.topLeft = true;
            }
            else {
                this.topRight = true;
            }
            if (this.placement === 'topLeft') {
                queueKey = 'topLeftQueue';
                queue$ = this.queueStore.select(NotifyQueueStore.topLeftSelector);
            }
            else if (this.placement === 'topRight') {
                queueKey = 'topRightQueue';
                queue$ = this.queueStore.select(NotifyQueueStore.topRightSelector);
            }
            else if (this.placement === 'bottomLeft') {
                queueKey = 'bottomLeftQueue';
                queue$ = this.queueStore.select(NotifyQueueStore.bottomLeftSelector);
            }
            else if (this.placement === 'bottomRight') {
                queueKey = 'bottomRightQueue';
                queue$ = this.queueStore.select(NotifyQueueStore.bottomRightSelector);
            }
            queue$.subscribe(function (data) {
                _this[queueKey] = data;
            });
        };
ThyNotifyContainerComponent.ɵfac = function ThyNotifyContainerComponent_Factory(t) { return new (t || ThyNotifyContainerComponent)(ɵngcc0.ɵɵdirectiveInject(NotifyQueueStore)); };
ThyNotifyContainerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNotifyContainerComponent, selectors: [["thy-notify-container"]], hostVars: 10, hostBindings: function ThyNotifyContainerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-notify-root", ctx.className)("thy-notify-bottomRight", ctx.bottomRight)("thy-notify-bottomLeft", ctx.bottomLeft)("thy-notify-topLeft", ctx.topLeft)("thy-notify-topRight", ctx.topRight);
    } }, decls: 4, vars: 4, consts: [[3, "thyOption", 4, "ngFor", "ngForOf"], [3, "thyOption"]], template: function ThyNotifyContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyNotifyContainerComponent_thy_notify_0_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(1, ThyNotifyContainerComponent_thy_notify_1_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(2, ThyNotifyContainerComponent_thy_notify_2_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(3, ThyNotifyContainerComponent_thy_notify_3_Template, 1, 1, "thy-notify", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.topLeftQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.topRightQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bottomLeftQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bottomRightQueue);
    } }, directives: [ɵngcc1.NgForOf, ThyNotifyComponent], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyContainerComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-notify-container',
                template: "<thy-notify *ngFor=\"let item of topLeftQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of topRightQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of bottomLeftQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of bottomRightQueue\" [thyOption]=\"item\"></thy-notify>\n"
            }]
    }], function () { return [{ type: NotifyQueueStore }]; }, { className: [{
            type: core.HostBinding,
            args: ['class.thy-notify-root']
        }], bottomRight: [{
            type: core.HostBinding,
            args: ['class.thy-notify-bottomRight']
        }], bottomLeft: [{
            type: core.HostBinding,
            args: ['class.thy-notify-bottomLeft']
        }], topLeft: [{
            type: core.HostBinding,
            args: ['class.thy-notify-topLeft']
        }], topRight: [{
            type: core.HostBinding,
            args: ['class.thy-notify-topRight']
        }] }); })();
        return ThyNotifyContainerComponent;
    }());
    ThyNotifyContainerComponent.ctorParameters = function () { return [
        { type: NotifyQueueStore }
    ]; };
    ThyNotifyContainerComponent.propDecorators = {
        className: [{ type: core.HostBinding, args: ['class.thy-notify-root',] }],
        bottomRight: [{ type: core.HostBinding, args: ['class.thy-notify-bottomRight',] }],
        bottomLeft: [{ type: core.HostBinding, args: ['class.thy-notify-bottomLeft',] }],
        topLeft: [{ type: core.HostBinding, args: ['class.thy-notify-topLeft',] }],
        topRight: [{ type: core.HostBinding, args: ['class.thy-notify-topRight',] }]
    };

    var CONTAINER_PLACEMENT = new core.InjectionToken('CONTAINER_PLACEMENT');
    var THY_NOTIFY_DEFAULT_OPTIONS = new core.InjectionToken('thy-notify-default-options');
    var ɵ0 = {};
    var THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER = {
        provide: THY_NOTIFY_DEFAULT_OPTIONS,
        useValue: ɵ0
    };

    var NOTIFY_OPTION_DEFAULT = {
        duration: 4500,
        pauseOnHover: true,
        maxStack: 8,
        placement: 'topRight'
    };
    var ThyNotifyService = /** @class */ (function () {
        function ThyNotifyService(injector, componentFactoryResolver, appRef, queueStore, defaultConfig) {
            this.injector = injector;
            this.componentFactoryResolver = componentFactoryResolver;
            this.appRef = appRef;
            this.queueStore = queueStore;
            this.defaultConfig = defaultConfig;
            this.notifyQueue$ = new rxjs.Subject();
            this._lastNotifyId = 0;
        }
        ThyNotifyService.prototype.show = function (options) {
            var notifyConfig = this.formatOptions(options);
            var placement = notifyConfig.placement;
            this.queueStore.addNotify(placement, notifyConfig);
            this._initContainer(placement);
        };
        ThyNotifyService.prototype.success = function (title, content, options) {
            this.show(Object.assign(Object.assign({}, (options || {})), { type: 'success', title: title || (options === null || options === void 0 ? void 0 : options.title) || '成功', content: content || (options === null || options === void 0 ? void 0 : options.content) }));
        };
        ThyNotifyService.prototype.info = function (title, content, options) {
            this.show(Object.assign(Object.assign({}, (options || {})), { type: 'info', title: title || (options === null || options === void 0 ? void 0 : options.title) || '提示', content: content || (options === null || options === void 0 ? void 0 : options.content) }));
        };
        ThyNotifyService.prototype.warning = function (title, content, options) {
            this.show(Object.assign(Object.assign({}, (options || {})), { type: 'warning', title: title || (options === null || options === void 0 ? void 0 : options.title) || '警告', content: content || (options === null || options === void 0 ? void 0 : options.content) }));
        };
        ThyNotifyService.prototype.error = function (title, content, options) {
            var _a, _b;
            var config = util.helpers.isString(options)
                ? { type: 'error', title: title || '错误', content: content, detail: options }
                : Object.assign(Object.assign({}, (options || {})), { type: 'error', title: title || ((_a = options) === null || _a === void 0 ? void 0 : _a.title) || '错误', content: content || ((_b = options) === null || _b === void 0 ? void 0 : _b.content) });
            this.show(config);
        };
        ThyNotifyService.prototype.removeNotifyById = function (id) {
            this.queueStore.removeNotify(id);
        };
        ThyNotifyService.prototype._initContainer = function (placement) {
            if (placement === 'topRight') {
                this.containerRefTopRight = this._loadNotifyContainerComponent(this.containerRefTopRight, placement);
            }
            else if (placement === 'bottomRight') {
                this.containerRefBottomRight = this._loadNotifyContainerComponent(this.containerRefBottomRight, placement);
            }
            else if (placement === 'bottomLeft') {
                this.containerRefBottomLeft = this._loadNotifyContainerComponent(this.containerRefBottomLeft, placement);
            }
            else if (placement === 'topLeft') {
                this.containerRefTopLeft = this._loadNotifyContainerComponent(this.containerRefTopLeft, placement);
            }
        };
        ThyNotifyService.prototype._loadNotifyContainerComponent = function (containerRef, placement) {
            if (!containerRef) {
                var portalOutlet = new portal.DomPortalOutlet(document.body, this.componentFactoryResolver, this.appRef, this.injector);
                var componentPortal = new portal.ComponentPortal(ThyNotifyContainerComponent, null);
                containerRef = portalOutlet.attachComponentPortal(componentPortal);
                Object.assign(containerRef.instance, {
                    initialState: {
                        placement: placement
                    }
                });
                containerRef.changeDetectorRef.detectChanges();
            }
            return containerRef;
        };
        ThyNotifyService.prototype.formatOptions = function (options) {
            if (util.helpers.isString(options.detail)) {
                options = Object.assign(Object.assign({}, options), { detail: { link: '[详情]', content: options.detail } });
            }
            return Object.assign({}, NOTIFY_OPTION_DEFAULT, { id: this._lastNotifyId++ }, this.defaultConfig, options);
        };
ThyNotifyService.ɵfac = function ThyNotifyService_Factory(t) { return new (t || ThyNotifyService)(ɵngcc0.ɵɵinject(ɵngcc0.Injector), ɵngcc0.ɵɵinject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵinject(ɵngcc0.ApplicationRef), ɵngcc0.ɵɵinject(NotifyQueueStore), ɵngcc0.ɵɵinject(THY_NOTIFY_DEFAULT_OPTIONS)); };
ThyNotifyService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyNotifyService, factory: function (t) { return ThyNotifyService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyService, [{
        type: core.Injectable
    }], function () { return [{ type: ɵngcc0.Injector }, { type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc0.ApplicationRef }, { type: NotifyQueueStore }, { type: undefined, decorators: [{
                type: core.Inject,
                args: [THY_NOTIFY_DEFAULT_OPTIONS]
            }] }]; }, null); })();
        return ThyNotifyService;
    }());
    ThyNotifyService.ctorParameters = function () { return [
        { type: core.Injector },
        { type: core.ComponentFactoryResolver },
        { type: core.ApplicationRef },
        { type: NotifyQueueStore },
        { type: undefined, decorators: [{ type: core.Inject, args: [THY_NOTIFY_DEFAULT_OPTIONS,] }] }
    ]; };

    var ThyNotifyModule = /** @class */ (function () {
        function ThyNotifyModule() {
        }
ThyNotifyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyNotifyModule });
ThyNotifyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyNotifyModule_Factory(t) { return new (t || ThyNotifyModule)(); }, providers: [ThyNotifyService, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER], imports: [[common.CommonModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyNotifyModule, { declarations: [ThyNotifyComponent, ThyNotifyContainerComponent], imports: [ɵngcc1.CommonModule, ɵngcc2.ThyIconModule], exports: [ThyNotifyContainerComponent, ThyNotifyComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyNotifyComponent, ThyNotifyContainerComponent],
                entryComponents: [ThyNotifyContainerComponent],
                providers: [ThyNotifyService, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER],
                imports: [common.CommonModule, icon.ThyIconModule],
                exports: [ThyNotifyContainerComponent, ThyNotifyComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyNotifyModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CONTAINER_PLACEMENT = CONTAINER_PLACEMENT;
    exports.NotifyQueueStore = NotifyQueueStore;
    exports.THY_NOTIFY_DEFAULT_OPTIONS = THY_NOTIFY_DEFAULT_OPTIONS;
    exports.THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER = THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER;
    exports.ThyNotifyComponent = ThyNotifyComponent;
    exports.ThyNotifyContainerComponent = ThyNotifyContainerComponent;
    exports.ThyNotifyModule = ThyNotifyModule;
    exports.ThyNotifyService = ThyNotifyService;
    exports.notifyQueueInitialState = notifyQueueInitialState;
    exports.ɵ0 = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
