// ngx-tethys/bundles/ngx-tethys-copy.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/tooltip'),require('ngx-tethys/notify'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('@angular/cdk/coercion'), require('ngx-tethys/tooltip'), require('ngx-tethys/notify')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/copy', ['@angular/core','ngx-tethys/tooltip','ngx-tethys/notify','@angular/common','exports', '@angular/core', '@angular/common', '@angular/cdk/coercion', 'ngx-tethys/tooltip', 'ngx-tethys/notify'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.tooltip,global.ngxTethys.notify,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].copy = {}), global.ng.core, global.ng.common, global.ng.cdk.coercion, global['ngx-tethys'].tooltip, global['ngx-tethys'].notify));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, coercion, tooltip, notify) { 'use strict';

    var ThyCopyDirective = /** @class */ (function () {
        function ThyCopyDirective(document, tooltipService, elementRef, viewContainerRef, notifyService) {
            this.document = document;
            this.tooltipService = tooltipService;
            this.elementRef = elementRef;
            this.viewContainerRef = viewContainerRef;
            this.notifyService = notifyService;
            // 默认为点击标签，可传复制目标标签
            this.thyCopy = new core.EventEmitter();
            this.thyCopySuccessText = '复制成功';
            this.thyCopyTips = '点击复制';
        }
        ThyCopyDirective.prototype.ngOnInit = function () {
            this.tooltipService.attach(this.elementRef, this.viewContainerRef, 'hover');
            this.tooltipService.thyTooltipDirective.content = this.thyCopyTips ? this.thyCopyTips : '点击复制';
        };
        ThyCopyDirective.prototype.getContent = function (event) {
            if (typeof this.thyCopyContent === 'string') {
                return this.thyCopyContent;
            }
            else {
                var target = this.thyCopyContent ? coercion.coerceElement(this.thyCopyContent) : event.target;
                return target.value || target.textContent;
            }
        };
        ThyCopyDirective.prototype.onClick = function (event) {
            var textarea = this.document.createElement('textarea');
            this.document.body.appendChild(textarea);
            textarea.value = this.getContent(event);
            textarea.select();
            try {
                document.execCommand('copy', false, null);
                this.thyCopy.emit({ isSuccess: true, event: event });
                this.notifyService.success(this.thyCopySuccessText);
            }
            catch (err) {
                this.thyCopy.emit({ isSuccess: false, event: event });
                this.notifyService.error('复制失败');
            }
            finally {
                textarea.remove();
            }
        };
        ThyCopyDirective.prototype.ngOnDestroy = function () {
            this.tooltipService.detach();
        };
ThyCopyDirective.ɵfac = function ThyCopyDirective_Factory(t) { return new (t || ThyCopyDirective)(ɵngcc0.ɵɵdirectiveInject(common.DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TooltipService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyNotifyService)); };
ThyCopyDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyCopyDirective, selectors: [["", "thyCopy", ""]], hostBindings: function ThyCopyDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyCopyDirective_click_HostBindingHandler($event) { return ctx.onClick($event); });
    } }, inputs: { thyCopySuccessText: "thyCopySuccessText", thyCopyTips: "thyCopyTips", thyCopyContent: "thyCopyContent" }, outputs: { thyCopy: "thyCopy" }, features: [ɵngcc0.ɵɵProvidersFeature([tooltip.TooltipService, notify.ThyNotifyService])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCopyDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyCopy]',
                providers: [tooltip.TooltipService, notify.ThyNotifyService]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: core.Inject,
                args: [common.DOCUMENT]
            }] }, { type: ɵngcc1.TooltipService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc2.ThyNotifyService }]; }, { thyCopy: [{
            type: core.Output
        }], thyCopySuccessText: [{
            type: core.Input
        }], thyCopyTips: [{
            type: core.Input
        }], onClick: [{
            type: core.HostListener,
            args: ['click', ['$event']]
        }], thyCopyContent: [{
            type: core.Input
        }] }); })();
        return ThyCopyDirective;
    }());
    ThyCopyDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
        { type: tooltip.TooltipService },
        { type: core.ElementRef },
        { type: core.ViewContainerRef },
        { type: notify.ThyNotifyService }
    ]; };
    ThyCopyDirective.propDecorators = {
        thyCopy: [{ type: core.Output }],
        thyCopySuccessText: [{ type: core.Input }],
        thyCopyTips: [{ type: core.Input }],
        thyCopyContent: [{ type: core.Input }],
        onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
    };

    var ThyCopyModule = /** @class */ (function () {
        function ThyCopyModule() {
        }
ThyCopyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCopyModule });
ThyCopyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCopyModule_Factory(t) { return new (t || ThyCopyModule)(); }, imports: [[common.CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCopyModule, { declarations: [ThyCopyDirective], imports: [ɵngcc3.CommonModule], exports: [ThyCopyDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCopyModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyCopyDirective],
                imports: [common.CommonModule],
                exports: [ThyCopyDirective]
            }]
    }], function () { return []; }, null); })();
        return ThyCopyModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyCopyDirective = ThyCopyDirective;
    exports.ThyCopyModule = ThyCopyModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
