// ngx-tethys/bundles/ngx-tethys-loading.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('ngx-tethys/util'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/loading', ['@angular/core','@angular/common','exports', '@angular/core', 'ngx-tethys/util', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].loading = {}), global.ng.core, global['ngx-tethys'].util, global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, util, common) { 
function ThyLoadingComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "h4");
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 3);
    ɵngcc0.ɵɵelement(5, "span", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵstyleProp("display", ctx_r0.tip ? "inline" : "none");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.tip);
} }
function ThyLoadingComponent_ng_content_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0, 0, ["*ngIf", "isMask || isDone"]);
} }
var _c0 = ["*"];
'use strict';

    var ThyLoadingComponent = /** @class */ (function () {
        function ThyLoadingComponent(changeDetectorRef) {
            this.changeDetectorRef = changeDetectorRef;
            this.loadingClassName = true;
        }
        Object.defineProperty(ThyLoadingComponent.prototype, "thyDone", {
            set: function (value) {
                this.isDone = util.coerceBooleanProperty(value);
                this.changeDetectorRef.markForCheck();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLoadingComponent.prototype, "thyTip", {
            set: function (value) {
                this.tip = value;
                this.changeDetectorRef.markForCheck();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyLoadingComponent.prototype, "thyIsMask", {
            // 不传或穿false,没有遮罩层，加载完成出现内容
            set: function (value) {
                this.isMask = util.coerceBooleanProperty(value);
                this.changeDetectorRef.markForCheck();
            },
            enumerable: false,
            configurable: true
        });
ThyLoadingComponent.ɵfac = function ThyLoadingComponent_Factory(t) { return new (t || ThyLoadingComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyLoadingComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyLoadingComponent, selectors: [["thy-loading"]], hostVars: 2, hostBindings: function ThyLoadingComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-loading", ctx.loadingClassName);
    } }, inputs: { thyDone: "thyDone", thyTip: "thyTip", thyIsMask: "thyIsMask" }, ngContentSelectors: _c0, decls: 3, vars: 4, consts: [["class", "thy-loading-indicator", 4, "ngIf"], [4, "ngIf"], [1, "thy-loading-indicator"], [1, "thy-loading-ellipsis"], [1, "thy-spot"]], template: function ThyLoadingComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtemplate(1, ThyLoadingComponent_div_1_Template, 6, 3, "div", 0);
        ɵngcc0.ɵɵtemplate(2, ThyLoadingComponent_ng_content_2_Template, 1, 0, "ng-content", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-loading-mask", ctx.isMask && !ctx.isDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isMask || ctx.isDone);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLoadingComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-loading',
                template: "<div [class.thy-loading-mask]=\"isMask && !isDone\">\n    <div class=\"thy-loading-indicator\" *ngIf=\"!isDone\">\n        <h4>\n            <span [style.display]=\"tip ? 'inline' : 'none'\">{{tip}}</span>\n            <div class=\"thy-loading-ellipsis\">\n                <span class=\"thy-spot\"></span>\n            </div>\n        </h4>\n    </div>\n    <ng-content *ngIf=\"isMask || isDone\"></ng-content>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { loadingClassName: [{
            type: core.HostBinding,
            args: ['class.thy-loading']
        }], thyDone: [{
            type: core.Input
        }], thyTip: [{
            type: core.Input
        }], thyIsMask: [{
            type: core.Input
        }] }); })();
        return ThyLoadingComponent;
    }());
    ThyLoadingComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef }
    ]; };
    ThyLoadingComponent.propDecorators = {
        thyDone: [{ type: core.Input }],
        thyTip: [{ type: core.Input }],
        thyIsMask: [{ type: core.Input }],
        loadingClassName: [{ type: core.HostBinding, args: ['class.thy-loading',] }]
    };

    var ThyLoadingModule = /** @class */ (function () {
        function ThyLoadingModule() {
        }
ThyLoadingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLoadingModule });
ThyLoadingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLoadingModule_Factory(t) { return new (t || ThyLoadingModule)(); }, imports: [[
            common.CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLoadingModule, { declarations: [ThyLoadingComponent], imports: [ɵngcc1.CommonModule], exports: [ThyLoadingComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLoadingModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    ThyLoadingComponent
                ],
                imports: [
                    common.CommonModule
                ],
                exports: [
                    ThyLoadingComponent
                ]
            }]
    }], function () { return []; }, null); })();
        return ThyLoadingModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyLoadingComponent = ThyLoadingComponent;
    exports.ThyLoadingModule = ThyLoadingModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
