// ngx-tethys/bundles/ngx-tethys-nav.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('@angular/common'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/core'), require('ngx-tethys/util'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/nav', ['@angular/core','ngx-tethys/core','@angular/common','ngx-tethys/icon','exports', '@angular/core', '@angular/common', 'ngx-tethys/core', 'ngx-tethys/util', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ng.common,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].nav = {}), global.ng.core, global.ng.common, global['ngx-tethys'].core, global['ngx-tethys'].util, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, core$1, util, icon) { 
var _c0 = ["*"];
var _c1 = ["thyIconNavLink", ""];
function ThyIconNavLinkComponent_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 1);
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.icon);
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var navTypeClassesMap = {
        primary: ['nav-primary'],
        secondary: ['nav-secondary'],
        thirdly: ['nav-thirdly'],
        'secondary-divider': ['nav-secondary-divider']
    };
    var navSizeClassesMap = {
        sm: 'nav-sm'
    };
    var navHorizontalClassesMap = {
        left: '',
        center: 'justify-content-center',
        right: 'justify-content-end'
    };
    var ThyNavComponent = /** @class */ (function () {
        function ThyNavComponent(updateHostClass, elementRef) {
            this.updateHostClass = updateHostClass;
            this.elementRef = elementRef;
            this._initialized = false;
            this._isVertical = false;
            this._isFill = false;
            this.updateHostClass.initializeElement(elementRef.nativeElement);
        }
        Object.defineProperty(ThyNavComponent.prototype, "thyType", {
            set: function (type) {
                this._type = type || 'primary';
                if (this._initialized) {
                    this._updateClasses();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyNavComponent.prototype, "thySize", {
            set: function (size) {
                this._size = size;
                if (this._initialized) {
                    this._updateClasses();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyNavComponent.prototype, "thyHorizontal", {
            set: function (horizontal) {
                this._horizontal = horizontal;
                if (this._initialized) {
                    this._updateClasses();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyNavComponent.prototype, "thyVertical", {
            set: function (value) {
                this._isVertical = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyNavComponent.prototype, "thyFill", {
            set: function (value) {
                this._isFill = value;
            },
            enumerable: false,
            configurable: true
        });
        ThyNavComponent.prototype._updateClasses = function () {
            var classNames = [];
            if (navTypeClassesMap[this._type]) {
                classNames = __spread(navTypeClassesMap[this._type]);
            }
            if (navSizeClassesMap[this._size]) {
                classNames.push(navSizeClassesMap[this._size]);
            }
            if (navHorizontalClassesMap[this._horizontal]) {
                classNames.push(navHorizontalClassesMap[this._horizontal]);
            }
            this.updateHostClass.updateClass(classNames);
        };
        ThyNavComponent.prototype.ngOnInit = function () {
            this._initialized = true;
            this._updateClasses();
        };
ThyNavComponent.ɵfac = function ThyNavComponent_Factory(t) { return new (t || ThyNavComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyNavComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNavComponent, selectors: [["thy-nav"]], hostAttrs: [1, "thy-nav"], hostVars: 4, hostBindings: function ThyNavComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-nav--vertical", ctx._isVertical)("thy-nav--fill", ctx._isFill);
    } }, inputs: { thyType: "thyType", thySize: "thySize", thyHorizontal: "thyHorizontal", thyVertical: "thyVertical", thyFill: "thyFill" }, features: [ɵngcc0.ɵɵProvidersFeature([core$1.UpdateHostClassService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyNavComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNavComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-nav',
                template: "\n        <ng-content></ng-content>\n    ",
                host: {
                    class: 'thy-nav'
                },
                providers: [core$1.UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }]; }, { _isVertical: [{
            type: core.HostBinding,
            args: ['class.thy-nav--vertical']
        }], _isFill: [{
            type: core.HostBinding,
            args: ['class.thy-nav--fill']
        }], thyType: [{
            type: core.Input
        }], thySize: [{
            type: core.Input
        }], thyHorizontal: [{
            type: core.Input
        }], thyVertical: [{
            type: core.Input
        }], thyFill: [{
            type: core.Input
        }] }); })();
        return ThyNavComponent;
    }());
    ThyNavComponent.ctorParameters = function () { return [
        { type: core$1.UpdateHostClassService },
        { type: core.ElementRef }
    ]; };
    ThyNavComponent.propDecorators = {
        thyType: [{ type: core.Input }],
        thySize: [{ type: core.Input }],
        thyHorizontal: [{ type: core.Input }],
        thyVertical: [{ type: core.Input }],
        thyFill: [{ type: core.Input }],
        _isVertical: [{ type: core.HostBinding, args: ['class.thy-nav--vertical',] }],
        _isFill: [{ type: core.HostBinding, args: ['class.thy-nav--fill',] }]
    };

    var ThyNavLinkDirective = /** @class */ (function () {
        function ThyNavLinkDirective() {
            this.navLinkActive = false;
            this.navLinkClass = true;
            // @HostBinding('attr.href') navLinkHref = 'javascript:;';
        }
        Object.defineProperty(ThyNavLinkDirective.prototype, "thyNavLinkActive", {
            set: function (active) {
                this.navLinkActive = util.coerceBooleanProperty(active);
            },
            enumerable: false,
            configurable: true
        });
ThyNavLinkDirective.ɵfac = function ThyNavLinkDirective_Factory(t) { return new (t || ThyNavLinkDirective)(); };
ThyNavLinkDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyNavLinkDirective, selectors: [["", "thyNavLink", ""]], hostVars: 4, hostBindings: function ThyNavLinkDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("active", ctx.navLinkActive)("nav-link", ctx.navLinkClass);
    } }, inputs: { thyNavLinkActive: "thyNavLinkActive" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNavLinkDirective, [{
        type: core.Directive,
        args: [{
                selector: '[thyNavLink]'
            }]
    }], function () { return []; }, { navLinkActive: [{
            type: core.HostBinding,
            args: ['class.active']
        }], navLinkClass: [{
            type: core.HostBinding,
            args: ['class.nav-link']
        }], thyNavLinkActive: [{
            type: core.Input
        }] }); })();
        return ThyNavLinkDirective;
    }());
    ThyNavLinkDirective.propDecorators = {
        thyNavLinkActive: [{ type: core.Input }],
        navLinkActive: [{ type: core.HostBinding, args: ['class.active',] }],
        navLinkClass: [{ type: core.HostBinding, args: ['class.nav-link',] }]
    };

    var ThyIconNavComponent = /** @class */ (function () {
        function ThyIconNavComponent(updateHostClassService, changeDetectorRef, elementRef) {
            this.updateHostClassService = updateHostClassService;
            this.changeDetectorRef = changeDetectorRef;
            this.initialized = false;
            this.isIconNav = true;
            this.updateHostClassService.initializeElement(elementRef);
        }
        Object.defineProperty(ThyIconNavComponent.prototype, "thyType", {
            set: function (type) {
                this.type = type;
                this.updateClasses();
                this.changeDetectorRef.markForCheck();
            },
            enumerable: false,
            configurable: true
        });
        ThyIconNavComponent.prototype.updateClasses = function (bypassInitialized) {
            if (!bypassInitialized && !this.initialized) {
                return;
            }
            this.updateHostClassService.updateClass(this.type ? ["thy-icon-nav-" + this.type] : []);
        };
        ThyIconNavComponent.prototype.ngOnInit = function () {
            this.initialized = true;
            this.updateClasses(true);
        };
ThyIconNavComponent.ɵfac = function ThyIconNavComponent_Factory(t) { return new (t || ThyIconNavComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyIconNavComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyIconNavComponent, selectors: [["thy-icon-nav"]], hostVars: 2, hostBindings: function ThyIconNavComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-icon-nav", ctx.isIconNav);
    } }, inputs: { thyType: "thyType" }, features: [ɵngcc0.ɵɵProvidersFeature([core$1.UpdateHostClassService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyIconNavComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyIconNavComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-icon-nav',
                template: "<ng-content></ng-content>\n",
                providers: [core$1.UpdateHostClassService],
                changeDetection: core.ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.ElementRef }]; }, { isIconNav: [{
            type: core.HostBinding,
            args: ["class.thy-icon-nav"]
        }], thyType: [{
            type: core.Input
        }] }); })();
        return ThyIconNavComponent;
    }());
    ThyIconNavComponent.ctorParameters = function () { return [
        { type: core$1.UpdateHostClassService },
        { type: core.ChangeDetectorRef },
        { type: core.ElementRef }
    ]; };
    ThyIconNavComponent.propDecorators = {
        isIconNav: [{ type: core.HostBinding, args: ["class.thy-icon-nav",] }],
        thyType: [{ type: core.Input }]
    };

    var ThyIconNavLinkComponent = /** @class */ (function () {
        function ThyIconNavLinkComponent() {
            this.navLinkActive = false;
            this.navLinkClass = true;
        }
        Object.defineProperty(ThyIconNavLinkComponent.prototype, "thyIconNavLinkIcon", {
            set: function (icon) {
                this.icon = icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyIconNavLinkComponent.prototype, "thyIconNavLinkActive", {
            set: function (active) {
                this.navLinkActive = util.coerceBooleanProperty(active);
            },
            enumerable: false,
            configurable: true
        });
ThyIconNavLinkComponent.ɵfac = function ThyIconNavLinkComponent_Factory(t) { return new (t || ThyIconNavLinkComponent)(); };
ThyIconNavLinkComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyIconNavLinkComponent, selectors: [["", "thyIconNavLink", ""]], hostVars: 4, hostBindings: function ThyIconNavLinkComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("active", ctx.navLinkActive)("thy-icon-nav-link", ctx.navLinkClass);
    } }, inputs: { thyIconNavLinkIcon: "thyIconNavLinkIcon", thyIconNavLinkActive: "thyIconNavLinkActive" }, attrs: _c1, ngContentSelectors: _c0, decls: 2, vars: 1, consts: [[3, "thyIconName", 4, "ngIf"], [3, "thyIconName"]], template: function ThyIconNavLinkComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵtemplate(1, ThyIconNavLinkComponent_thy_icon_1_Template, 1, 1, "thy-icon", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.icon);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.ThyIconComponent], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyIconNavLinkComponent, [{
        type: core.Component,
        args: [{
                selector: '[thyIconNavLink]',
                template: '<ng-content></ng-content><thy-icon *ngIf="icon" [thyIconName]="icon"></thy-icon>'
            }]
    }], function () { return []; }, { navLinkActive: [{
            type: core.HostBinding,
            args: ['class.active']
        }], navLinkClass: [{
            type: core.HostBinding,
            args: ['class.thy-icon-nav-link']
        }], thyIconNavLinkIcon: [{
            type: core.Input
        }], thyIconNavLinkActive: [{
            type: core.Input
        }] }); })();
        return ThyIconNavLinkComponent;
    }());
    ThyIconNavLinkComponent.propDecorators = {
        navLinkActive: [{ type: core.HostBinding, args: ['class.active',] }],
        navLinkClass: [{ type: core.HostBinding, args: ['class.thy-icon-nav-link',] }],
        thyIconNavLinkIcon: [{ type: core.Input }],
        thyIconNavLinkActive: [{ type: core.Input }]
    };

    var ThyNavModule = /** @class */ (function () {
        function ThyNavModule() {
        }
ThyNavModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyNavModule });
ThyNavModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyNavModule_Factory(t) { return new (t || ThyNavModule)(); }, imports: [[common.CommonModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyNavModule, { declarations: [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent], imports: [ɵngcc2.CommonModule, ɵngcc3.ThyIconModule], exports: [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNavModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent],
                imports: [common.CommonModule, icon.ThyIconModule],
                exports: [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyNavModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyIconNavComponent = ThyIconNavComponent;
    exports.ThyIconNavLinkComponent = ThyIconNavLinkComponent;
    exports.ThyNavComponent = ThyNavComponent;
    exports.ThyNavLinkDirective = ThyNavLinkDirective;
    exports.ThyNavModule = ThyNavModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
