// ngx-tethys/bundles/ngx-tethys-tree.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),require('ngx-tethys/drag-drop'),require('@angular/common'),require('ngx-tethys/icon'),require('ngx-tethys/loading'),require('ngx-tethys/input'),require('ngx-tethys/button'),require('ngx-tethys/shared'),require('@angular/forms'),require('ngx-tethys/list'),require('ngx-tethys/checkbox'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/util'), require('rxjs'), require('@angular/cdk/collections'), require('@angular/forms'), require('ngx-tethys/core'), require('ngx-tethys/drag-drop'), require('rxjs/operators'), require('ngx-tethys/input'), require('ngx-tethys/button'), require('ngx-tethys/shared'), require('ngx-tethys/list'), require('ngx-tethys/loading'), require('ngx-tethys/icon'), require('ngx-tethys/checkbox')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/tree', ['@angular/core','ngx-tethys/core','ngx-tethys/drag-drop','@angular/common','ngx-tethys/icon','ngx-tethys/loading','ngx-tethys/input','ngx-tethys/button','ngx-tethys/shared','@angular/forms','ngx-tethys/list','ngx-tethys/checkbox','exports', '@angular/core', '@angular/common', 'ngx-tethys/util', 'rxjs', '@angular/cdk/collections', '@angular/forms', 'ngx-tethys/core', 'ngx-tethys/drag-drop', 'rxjs/operators', 'ngx-tethys/input', 'ngx-tethys/button', 'ngx-tethys/shared', 'ngx-tethys/list', 'ngx-tethys/loading', 'ngx-tethys/icon', 'ngx-tethys/checkbox'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,global.ngxTethys.dragDrop,global.ng.common,global.ngxTethys.icon,global.ngxTethys.loading,global.ngxTethys.input,global.ngxTethys.button,global.ngxTethys.shared,global.ng.forms,global.ngxTethys.list,global.ngxTethys.checkbox,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].tree = {}), global.ng.core, global.ng.common, global['ngx-tethys'].util, global.rxjs, global.ng.cdk.collections, global.ng.forms, global['ngx-tethys'].core, global['ngx-tethys']['drag-drop'], global.rxjs.operators, global['ngx-tethys'].input, global['ngx-tethys'].button, global['ngx-tethys'].shared, global['ngx-tethys'].list, global['ngx-tethys'].loading, global['ngx-tethys'].icon, global['ngx-tethys'].checkbox));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,ɵngcc10,ɵngcc11,exports, core, common, util, rxjs, collections, forms, core$1, dragDrop, operators, input, button, shared, list, loading, icon, checkbox) { 
var _c0 = ["treeNodeTemplate"];
var _c1 = ["emptyChildrenTemplate"];
function ThyTreeComponent_thy_tree_node_1_Template(rf, ctx) { if (rf & 1) {
    var _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-tree-node", 2);
    ɵngcc0.ɵɵlistener("thyOnClick", function ThyTreeComponent_thy_tree_node_1_Template_thy_tree_node_thyOnClick_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); var ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.thyOnClick.emit($event); })("thyOnExpandChange", function ThyTreeComponent_thy_tree_node_1_Template_thy_tree_node_thyOnExpandChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); var ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.thyOnExpandChange.emit($event); })("thyOnCheckboxChange", function ThyTreeComponent_thy_tree_node_1_Template_thy_tree_node_thyOnCheckboxChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); var ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.thyOnCheckboxChange.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var node_r1 = ctx.$implicit;
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyDrag", node_r1)("node", node_r1)("templateRef", ctx_r0.templateRef)("emptyChildrenTemplateRef", ctx_r0.emptyChildrenTemplate)("thyAsync", ctx_r0.thyAsync)("thyDraggable", ctx_r0.thyDraggable)("thyCheckable", ctx_r0.thyCheckable)("thyMultiple", ctx_r0.thyMultiple)("thyTitleTruncate", ctx_r0.thyTitleTruncate);
} }
var _c2 = ["childrenTree"];
var _c3 = ["title"];
function ThyTreeNodeComponent_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 7);
    ɵngcc0.ɵɵelement(1, "thy-icon", 8);
    ɵngcc0.ɵɵelementEnd();
} }
function ThyTreeNodeComponent_ng_container_3_ng_container_1_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 11);
} if (rf & 2) {
    var ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r7.node.isExpanded ? ctx_r7.root.thyIcons.expand : ctx_r7.root.thyIcons.collapse);
} }
function ThyTreeNodeComponent_ng_container_3_ng_container_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 12);
} if (rf & 2) {
    var ctx_r9 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassProp("collapse", !ctx_r9.node.isExpanded);
} }
function ThyTreeNodeComponent_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_ng_container_3_ng_container_1_thy_icon_1_Template, 1, 1, "thy-icon", 9);
    ɵngcc0.ɵɵtemplate(2, ThyTreeNodeComponent_ng_container_3_ng_container_1_ng_template_2_Template, 1, 2, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var _r8 = ɵngcc0.ɵɵreference(3);
    var ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r6.root.thyIcons == null ? null : ctx_r6.root.thyIcons.expand) && (ctx_r6.root.thyIcons == null ? null : ctx_r6.root.thyIcons.collapse))("ngIfElse", _r8);
} }
function ThyTreeNodeComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_ng_container_3_ng_container_1_Template, 4, 2, "ng-container", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.node.children && ctx_r1.node.children.length > 0 || ctx_r1.thyAsync || ctx_r1.emptyChildrenTemplateRef);
} }
function ThyTreeNodeComponent_input_5_Template(rf, ctx) { if (rf & 1) {
    var _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "input", 13);
    ɵngcc0.ɵɵlistener("click", function ThyTreeNodeComponent_input_5_Template_input_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); var ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.clickNodeCheck($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("form-check-indeterminate", ctx_r2.node.isChecked === ctx_r2.checkState.indeterminate);
    ɵngcc0.ɵɵproperty("checked", ctx_r2.node.isChecked === ctx_r2.checkState.checked)("disabled", ctx_r2.node.isDisabled);
} }
function ThyTreeNodeComponent_ng_container_6_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span", 15);
    ɵngcc0.ɵɵelement(2, "i", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r12.nodeIcon)("ngStyle", ctx_r12.nodeIconStyle);
} }
function ThyTreeNodeComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_ng_container_6_ng_container_1_Template, 3, 2, "ng-container", 3);
    ɵngcc0.ɵɵelementStart(2, "span", 14);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.nodeIcon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("truncate", ctx_r3.thyTitleTruncate);
    ɵngcc0.ɵɵproperty("title", ctx_r3.node.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r3.node.title, "");
} }
function ThyTreeNodeComponent_ng_container_7_ng_template_1_Template(rf, ctx) { }
var _c4 = function (a0, a1, a2) { return { $implicit: a0, node: a1, origin: a2 }; };
function ThyTreeNodeComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_ng_container_7_ng_template_1_Template, 0, 0, "ng-template", 17);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r4.templateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(2, _c4, ctx_r4.node, ctx_r4.node, ctx_r4.node == null ? null : ctx_r4.node.origin));
} }
function ThyTreeNodeComponent_div_8_thy_tree_node_1_Template(rf, ctx) { if (rf & 1) {
    var _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-tree-node", 20);
    ɵngcc0.ɵɵlistener("thyOnClick", function ThyTreeNodeComponent_div_8_thy_tree_node_1_Template_thy_tree_node_thyOnClick_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r19); var ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.thyOnClick.emit($event); })("thyOnCheckboxChange", function ThyTreeNodeComponent_div_8_thy_tree_node_1_Template_thy_tree_node_thyOnCheckboxChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r19); var ctx_r20 = ɵngcc0.ɵɵnextContext(2); return ctx_r20.thyOnCheckboxChange.emit($event); })("thyOnExpandChange", function ThyTreeNodeComponent_div_8_thy_tree_node_1_Template_thy_tree_node_thyOnExpandChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r19); var ctx_r21 = ɵngcc0.ɵɵnextContext(2); return ctx_r21.thyOnExpandChange.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var node_r16 = ctx.$implicit;
    var ctx_r14 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("node", node_r16)("thyDrag", node_r16)("templateRef", ctx_r14.templateRef)("thyAsync", ctx_r14.thyAsync)("thyMultiple", ctx_r14.thyMultiple)("thyDraggable", ctx_r14.thyDraggable)("thyCheckable", ctx_r14.thyCheckable)("thyTitleTruncate", ctx_r14.thyTitleTruncate);
} }
function ThyTreeNodeComponent_div_8_ng_container_2_thy_loading_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-loading", 22);
} if (rf & 2) {
    var ctx_r22 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("thyDone", !ctx_r22.node.isLoading);
} }
function ThyTreeNodeComponent_div_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_div_8_ng_container_2_thy_loading_1_Template, 1, 1, "thy-loading", 21);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r15.node.isLoading);
} }
function ThyTreeNodeComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    var _r24 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵlistener("thyDragDropped", function ThyTreeNodeComponent_div_8_Template_div_thyDragDropped_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r24); var ctx_r23 = ɵngcc0.ɵɵnextContext(); return ctx_r23.root.onDragDrop($event); })("thyDragStarted", function ThyTreeNodeComponent_div_8_Template_div_thyDragStarted_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r24); var ctx_r25 = ɵngcc0.ɵɵnextContext(); return ctx_r25.root.onDragStart($event); });
    ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_div_8_thy_tree_node_1_Template, 1, 8, "thy-tree-node", 19);
    ɵngcc0.ɵɵtemplate(2, ThyTreeNodeComponent_div_8_ng_container_2_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyDropContainer", ctx_r5.node.children)("thyDropContainerDisabled", !ctx_r5.root.thyDraggable)("thyBeforeDragStart", ctx_r5.root.thyBeforeDragStart)("thyBeforeDragOver", ctx_r5.root.beforeDragOver)("thyBeforeDragDrop", ctx_r5.root.thyBeforeDragDrop);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r5.node.children)("ngForTrackBy", ctx_r5.root.trackByFn);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.thyAsync);
} }
var _c5 = ["ThyTreeReplaceRegion", ""];
var _c6 = ["*"];
'use strict';

    exports.ThyTreeNodeCheckState = void 0;
    (function (ThyTreeNodeCheckState) {
        ThyTreeNodeCheckState[ThyTreeNodeCheckState["unchecked"] = 0] = "unchecked";
        ThyTreeNodeCheckState[ThyTreeNodeCheckState["checked"] = 1] = "checked";
        ThyTreeNodeCheckState[ThyTreeNodeCheckState["indeterminate"] = 2] = "indeterminate";
    })(exports.ThyTreeNodeCheckState || (exports.ThyTreeNodeCheckState = {}));
    var ThyTreeIcons = /** @class */ (function () {
        function ThyTreeIcons() {
        }
        return ThyTreeIcons;
    }());

    function checkStateResolve(node) {
        var checkedNodes = node.children.filter(function (n) { return n.isChecked === exports.ThyTreeNodeCheckState.checked; });
        var unCheckedNodes = node.children.filter(function (n) { return n.isChecked === exports.ThyTreeNodeCheckState.unchecked; });
        if (checkedNodes.length === node.children.length) {
            return exports.ThyTreeNodeCheckState.checked;
        }
        else if (unCheckedNodes.length === node.children.length) {
            return exports.ThyTreeNodeCheckState.unchecked;
        }
        else {
            return exports.ThyTreeNodeCheckState.indeterminate;
        }
    }
    var ThyTreeService = /** @class */ (function () {
        function ThyTreeService() {
            this.checkStateResolve = checkStateResolve;
            this.$statusChange = new rxjs.Subject();
        }
        ThyTreeService.prototype._getParallelTreeNodes = function (nodes, list) {
            var _this = this;
            if (list === void 0) { list = []; }
            (nodes || []).forEach(function (node) {
                list.push(node);
                _this._getParallelTreeNodes(node.children || [], list);
            });
            return list;
        };
        ThyTreeService.prototype.setCheckStateResolve = function (resolve) {
            if (resolve === void 0) { resolve = checkStateResolve; }
            this.checkStateResolve = resolve;
        };
        ThyTreeService.prototype.resetSortedTreeNodes = function (treeNodes, parent) {
            var _this = this;
            treeNodes.forEach(function (node) {
                node.level = node.parentNode ? node.parentNode.level + 1 : 0;
                node.origin.children = node.children.map(function (n) { return n.origin; });
                node.parentNode = parent;
                _this.resetSortedTreeNodes(node.children, node);
            });
        };
        ThyTreeService.prototype.getTreeNode = function (key) {
            var allNodes = this._getParallelTreeNodes(this.treeNodes);
            return allNodes.find(function (n) { return n.key === key; });
        };
        ThyTreeService.prototype.getExpandedNodes = function () {
            var allNodes = this._getParallelTreeNodes(this.treeNodes);
            return allNodes.filter(function (n) { return n.isExpanded; });
        };
        ThyTreeService.prototype.getCheckedNodes = function () {
            var allNodes = this._getParallelTreeNodes(this.treeNodes);
            return allNodes.filter(function (n) { return n.isChecked === exports.ThyTreeNodeCheckState.checked; });
        };
        ThyTreeService.prototype.deleteTreeNode = function (node) {
            var children = node.parentNode ? node.parentNode.children : this.treeNodes;
            var index = children.findIndex(function (n) { return n.key === node.key; });
            if (index > -1) {
                children.splice(index, 1);
            }
        };
        ThyTreeService.prototype.expandTreeNodes = function (keyOrKeys) {
            var keys = util.coerceArray(keyOrKeys);
            var needExpandNodes = this._getParallelTreeNodes(this.treeNodes).filter(function (node) {
                return keys.indexOf(node.key) > -1;
            });
            needExpandNodes.forEach(function (node) {
                node.setExpanded(true);
            });
        };
        ThyTreeService.prototype.statusChanged = function () {
            return this.$statusChange.asObservable();
        };
        // 设置节点选中状态
        ThyTreeService.prototype.setNodeChecked = function (node, checked, propagateUp, propagateDown) {
            var _this = this;
            if (propagateUp === void 0) { propagateUp = true; }
            if (propagateDown === void 0) { propagateDown = true; }
            node.isChecked = checked ? exports.ThyTreeNodeCheckState.checked : exports.ThyTreeNodeCheckState.unchecked;
            node.origin.checked = checked;
            if (propagateDown && node.children) {
                node.children.forEach(function (subNode) {
                    _this.setNodeChecked(subNode, checked, false, true);
                });
            }
            if (propagateUp) {
                this.syncNodeCheckState(node.parentNode);
            }
        };
        ThyTreeService.prototype.syncNodeCheckState = function (node) {
            if (node) {
                node.isChecked = this.checkStateResolve(node);
                this.syncNodeCheckState(node.parentNode);
            }
        };
        ThyTreeService.prototype.ngOnDestroy = function () {
            this.$statusChange.complete();
            this.$statusChange = null;
        };
ThyTreeService.ɵfac = function ThyTreeService_Factory(t) { return new (t || ThyTreeService)(); };
ThyTreeService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyTreeService, factory: function (t) { return ThyTreeService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeService, [{
        type: core.Injectable
    }], function () { return []; }, null); })();
        return ThyTreeService;
    }());
    ThyTreeService.ctorParameters = function () { return []; };

    var ThyTreeNode = /** @class */ (function () {
        function ThyTreeNode(node, parent, service) {
            var _this = this;
            if (parent === void 0) { parent = null; }
            this.level = 0;
            this.title = node.title;
            this.key = node.key;
            this.children = [];
            this.parentNode = parent;
            this.level = parent ? parent.level + 1 : this.level;
            this.origin = node;
            this.isDisabled = node.disabled || false;
            this.isExpanded = node.expanded || false;
            this.isChecked = node.checked ? exports.ThyTreeNodeCheckState.checked : exports.ThyTreeNodeCheckState.unchecked;
            this.isLoading = false;
            if (node.children) {
                node.children.forEach(function (childNode) {
                    _this.children.push(new ThyTreeNode(childNode, _this));
                });
            }
            this.service = service;
        }
        Object.defineProperty(ThyTreeNode.prototype, "treeService", {
            get: function () {
                if (this.service) {
                    return this.service;
                }
                else if (this.parentNode) {
                    return this.parentNode.treeService;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyTreeNode.prototype.setKey = function (key) {
            this.origin.key = key;
            this.key = key;
        };
        ThyTreeNode.prototype.setTitle = function (title) {
            this.origin.title = title;
            this.title = title;
        };
        ThyTreeNode.prototype.setExpanded = function (expanded, propagate) {
            if (propagate === void 0) { propagate = false; }
            this.origin.expanded = expanded;
            this.isExpanded = expanded;
            if (propagate && this.children) {
                this.children.forEach(function (n) { return n.setExpanded(expanded, propagate); });
            }
        };
        ThyTreeNode.prototype.setLoading = function (loading) {
            this.isLoading = loading;
        };
        ThyTreeNode.prototype.setChecked = function (checked, propagateUp, propagateDown) {
            if (propagateUp === void 0) { propagateUp = true; }
            if (propagateDown === void 0) { propagateDown = true; }
            this.treeService.setNodeChecked(this, checked, propagateUp, propagateDown);
        };
        ThyTreeNode.prototype.syncNodeCheckState = function () {
            this.treeService.syncNodeCheckState(this);
        };
        ThyTreeNode.prototype.getParentNode = function () {
            return this.parentNode;
        };
        ThyTreeNode.prototype.getChildren = function () {
            return this.children;
        };
        ThyTreeNode.prototype.addChildren = function (children, index) {
            var _this = this;
            if (index === void 0) { index = -1; }
            children = util.helpers.coerceArray(children);
            (children || []).forEach(function (childNode, i) {
                if (index === -1) {
                    _this.children.push(new ThyTreeNode(childNode, _this));
                }
                else {
                    _this.children.splice(index + i, 0, new ThyTreeNode(childNode, _this, _this.treeService));
                }
            });
            this.origin.children = this.getChildren().map(function (n) { return n.origin; });
            this.setLoading(false);
            this.treeService.$statusChange.next({
                eventName: 'addChildren',
                node: this
            });
        };
        return ThyTreeNode;
    }());

    var treeTypeClassMap = {
        default: ['thy-tree-default'],
        especial: ['thy-tree-especial']
    };
    var ThyTreeComponent = /** @class */ (function () {
        function ThyTreeComponent(elementRef, updateHostClassService, thyTreeService) {
            var _this = this;
            this.elementRef = elementRef;
            this.updateHostClassService = updateHostClassService;
            this.thyTreeService = thyTreeService;
            this._draggable = false;
            this.thyShowExpand = true;
            this.thyMultiple = false;
            this.thyAsync = false;
            this._thyType = 'default';
            this.thyIcons = {};
            this.thyTitleTruncate = true;
            this.thyOnClick = new core.EventEmitter();
            this.thyOnCheckboxChange = new core.EventEmitter();
            this.thyOnExpandChange = new core.EventEmitter();
            this.thyOnDragDrop = new core.EventEmitter();
            this.thyTreeClass = true;
            this.beforeDragOver = function (event) {
                return _this.isShowExpand(event.item) || (!_this.isShowExpand(event.item) && event.position !== dragDrop.ThyDropPosition.in);
            };
            this._onTouched = function () { };
            this._onChange = function (_) { };
        }
        Object.defineProperty(ThyTreeComponent.prototype, "thyNodes", {
            set: function (value) {
                var _this = this;
                this._expandedKeys = this.getExpandedNodes().map(function (node) { return node.key; });
                this._selectedKeys = this.getSelectedNodes().map(function (node) { return node.key; });
                this.treeNodes = (value || []).map(function (node) { return new ThyTreeNode(node, null, _this.thyTreeService); });
                this.thyTreeService.treeNodes = this.treeNodes;
                this.thyTreeService.expandTreeNodes(this._expandedKeys);
                this._selectTreeNodes(this._selectedKeys);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeComponent.prototype, "thyDraggable", {
            get: function () {
                return this._draggable;
            },
            set: function (value) {
                this._draggable = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeComponent.prototype, "thyCheckStateResolve", {
            set: function (resolve) {
                this.thyTreeService.setCheckStateResolve(resolve);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeComponent.prototype, "thyType", {
            get: function () {
                return this._thyType;
            },
            set: function (type) {
                this._thyType = type;
                if (type === 'especial') {
                    this.thyIcons = { expand: 'minus-square', collapse: 'plus-square' };
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeComponent.prototype, "thySelectedKeys", {
            set: function (keys) {
                this._selectedKeys = keys;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeComponent.prototype, "templateRef", {
            get: function () {
                return this._templateRef;
            },
            set: function (template) {
                if (template) {
                    this._templateRef = template;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeComponent.prototype, "emptyChildrenTemplateRef", {
            get: function () {
                return this._emptyChildrenTemplateRef;
            },
            set: function (template) {
                if (template) {
                    this._emptyChildrenTemplateRef = template;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyTreeComponent.prototype.ngOnChanges = function (changes) {
            if (changes.thyType && !changes.thyType.isFirstChange()) {
                this._setTreeType();
            }
            if (changes.thyMultiple && !changes.thyMultiple.isFirstChange()) {
                this._instanceSelectionModel();
            }
        };
        ThyTreeComponent.prototype.ngOnInit = function () {
            this.updateHostClassService.initializeElement(this.elementRef.nativeElement);
            this._setTreeType();
            this._setTreeSize();
            this._instanceSelectionModel();
            this._selectTreeNodes(this._selectedKeys);
        };
        ThyTreeComponent.prototype._setTreeType = function () {
            if (this.thyType) {
                this.updateHostClassService.addClass(treeTypeClassMap[this.thyType]);
            }
        };
        ThyTreeComponent.prototype._setTreeSize = function () {
            if (this.thySize) {
                this.updateHostClassService.addClass("thy-tree-" + this.thySize);
            }
        };
        ThyTreeComponent.prototype._instanceSelectionModel = function () {
            this._selectionModel = new collections.SelectionModel(this.thyMultiple);
        };
        ThyTreeComponent.prototype._selectTreeNodes = function (keys) {
            var _this = this;
            (keys || []).forEach(function (key) {
                var node = _this.thyTreeService.getTreeNode(key);
                if (node) {
                    _this.selectTreeNode(_this.thyTreeService.getTreeNode(key));
                }
            });
        };
        ThyTreeComponent.prototype.isSelected = function (node) {
            return this._selectionModel.isSelected(node);
        };
        ThyTreeComponent.prototype.toggleTreeNode = function (node) {
            if (node && !node.isDisabled) {
                this._selectionModel.toggle(node);
            }
        };
        ThyTreeComponent.prototype.trackByFn = function (index, item) {
            return item.key || index;
        };
        ThyTreeComponent.prototype.onDragStart = function (event) {
            if (this.isShowExpand(event.item) && event.item.isExpanded) {
                event.item.setExpanded(false);
            }
        };
        ThyTreeComponent.prototype.onDragDrop = function (event) {
            if (!this.isShowExpand(event.item) && event.position === dragDrop.ThyDropPosition.in) {
                return;
            }
            var parent = event.previousItem.parentNode;
            if (parent) {
                parent.children = parent.children.filter(function (item) { return item !== event.previousItem; });
            }
            else {
                this.treeNodes = this.treeNodes.filter(function (item) { return item !== event.previousItem; });
            }
            switch (event.position) {
                case dragDrop.ThyDropPosition.in:
                    event.item.addChildren(event.previousItem.origin);
                    break;
                case dragDrop.ThyDropPosition.after:
                case dragDrop.ThyDropPosition.before:
                    var targetParent = event.item.parentNode;
                    var index = event.position === dragDrop.ThyDropPosition.before ? 0 : 1;
                    if (targetParent) {
                        targetParent.addChildren(event.previousItem.origin, targetParent.children.indexOf(event.item) + index);
                    }
                    else {
                        this.treeNodes.splice(this.treeNodes.indexOf(event.item) + index, 0, event.previousItem);
                    }
                    break;
            }
            this.thyTreeService.resetSortedTreeNodes(this.treeNodes);
            var afterNode = null;
            var targetNode = null;
            if (event.position === dragDrop.ThyDropPosition.before) {
                afterNode = event.containerItems[event.currentIndex - 1];
                targetNode = event.item.parentNode;
            }
            else if (event.position === dragDrop.ThyDropPosition.after) {
                afterNode = event.containerItems[event.currentIndex];
                targetNode = event.item.parentNode;
            }
            else {
                afterNode = event.item.children[event.item.children.length - 2];
                targetNode = event.item;
            }
            this.thyOnDragDrop.emit({
                event: event,
                currentIndex: event.currentIndex,
                dragNode: event.previousItem,
                targetNode: targetNode,
                afterNode: afterNode
            });
        };
        ThyTreeComponent.prototype.isShowExpand = function (node) {
            if (util.helpers.isFunction(this.thyShowExpand)) {
                return this.thyShowExpand(node);
            }
            else {
                return this.thyShowExpand;
            }
        };
        ThyTreeComponent.prototype.writeValue = function (value) {
            this.thyNodes = value;
        };
        ThyTreeComponent.prototype.registerOnChange = function (fn) {
            this._onChange = fn;
        };
        ThyTreeComponent.prototype.registerOnTouched = function (fn) {
            this._onTouched = fn;
        };
        // region Public Functions
        ThyTreeComponent.prototype.selectTreeNode = function (node) {
            this._selectionModel.select(node);
        };
        ThyTreeComponent.prototype.getRootNodes = function () {
            return this.treeNodes;
        };
        ThyTreeComponent.prototype.getTreeNode = function (key) {
            return this.thyTreeService.getTreeNode(key);
        };
        ThyTreeComponent.prototype.getSelectedNode = function () {
            return this._selectionModel ? this._selectionModel.selected[0] : null;
        };
        ThyTreeComponent.prototype.getSelectedNodes = function () {
            return this._selectionModel ? this._selectionModel.selected : [];
        };
        ThyTreeComponent.prototype.getExpandedNodes = function () {
            return this.thyTreeService.getExpandedNodes();
        };
        ThyTreeComponent.prototype.getCheckedNodes = function () {
            return this.thyTreeService.getCheckedNodes();
        };
        ThyTreeComponent.prototype.addTreeNode = function (node, parent, index) {
            if (index === void 0) { index = -1; }
            if (parent) {
                parent.addChildren(node, index);
            }
            else {
                if (index > -1) {
                    this.treeNodes.splice(index, 0, new ThyTreeNode(node, null, this.thyTreeService));
                }
                else {
                    this.treeNodes.push(new ThyTreeNode(node, null, this.thyTreeService));
                }
            }
        };
        ThyTreeComponent.prototype.deleteTreeNode = function (node) {
            if (this.isSelected(node)) {
                this._selectionModel.toggle(node);
            }
            this.thyTreeService.deleteTreeNode(node);
        };
        ThyTreeComponent.prototype.expandAllNodes = function () {
            var nodes = this.getRootNodes();
            nodes.forEach(function (n) { return n.setExpanded(true, true); });
        };
        ThyTreeComponent.prototype.collapsedAllNodes = function () {
            var nodes = this.getRootNodes();
            nodes.forEach(function (n) { return n.setExpanded(false, true); });
        };
ThyTreeComponent.ɵfac = function ThyTreeComponent_Factory(t) { return new (t || ThyTreeComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ThyTreeService)); };
ThyTreeComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTreeComponent, selectors: [["thy-tree"]], contentQueries: function ThyTreeComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.emptyChildrenTemplate = _t.first);
    } }, hostVars: 6, hostBindings: function ThyTreeComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-multiple-selection-list", ctx.thyMultiple)("thy-tree", ctx.thyTreeClass)("thy-tree-draggable", ctx.thyDraggable);
    } }, inputs: { thyShowExpand: "thyShowExpand", thyMultiple: "thyMultiple", thyAsync: "thyAsync", thyIcons: "thyIcons", thyTitleTruncate: "thyTitleTruncate", thyNodes: "thyNodes", thyDraggable: "thyDraggable", thyCheckStateResolve: "thyCheckStateResolve", thyType: "thyType", thySelectedKeys: "thySelectedKeys", thyCheckable: "thyCheckable", thySize: "thySize", thyBeforeDragStart: "thyBeforeDragStart", thyBeforeDragDrop: "thyBeforeDragDrop" }, outputs: { thyOnClick: "thyOnClick", thyOnCheckboxChange: "thyOnCheckboxChange", thyOnExpandChange: "thyOnExpandChange", thyOnDragDrop: "thyOnDragDrop" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: forms.NG_VALUE_ACCESSOR,
                useExisting: core.forwardRef(function () { return ThyTreeComponent; }),
                multi: true
            },
            ThyTreeService,
            core$1.UpdateHostClassService
        ]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 7, consts: [[3, "thyDropContainer", "thyDropContainerDisabled", "thyBeforeDragStart", "thyBeforeDragOver", "thyBeforeDragDrop", "thyDragStarted", "thyDragDropped"], [3, "thyDrag", "node", "templateRef", "emptyChildrenTemplateRef", "thyAsync", "thyDraggable", "thyCheckable", "thyMultiple", "thyTitleTruncate", "thyOnClick", "thyOnExpandChange", "thyOnCheckboxChange", 4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "thyDrag", "node", "templateRef", "emptyChildrenTemplateRef", "thyAsync", "thyDraggable", "thyCheckable", "thyMultiple", "thyTitleTruncate", "thyOnClick", "thyOnExpandChange", "thyOnCheckboxChange"]], template: function ThyTreeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("thyDragStarted", function ThyTreeComponent_Template_div_thyDragStarted_0_listener($event) { return ctx.onDragStart($event); })("thyDragDropped", function ThyTreeComponent_Template_div_thyDragDropped_0_listener($event) { return ctx.onDragDrop($event); });
        ɵngcc0.ɵɵtemplate(1, ThyTreeComponent_thy_tree_node_1_Template, 1, 9, "thy-tree-node", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("thyDropContainer", ctx.treeNodes)("thyDropContainerDisabled", !ctx.thyDraggable)("thyBeforeDragStart", ctx.thyBeforeDragStart)("thyBeforeDragOver", ctx.beforeDragOver)("thyBeforeDragDrop", ctx.thyBeforeDragDrop);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.treeNodes)("ngForTrackBy", ctx.trackByFn);
    } }, directives: function () { return [ɵngcc2.ThyDropContainerDirective, ɵngcc3.NgForOf, ThyTreeNodeComponent, ɵngcc2.ThyDragDirective]; }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-tree',
                template: "<div\n  [thyDropContainer]=\"treeNodes\"\n  [thyDropContainerDisabled]=\"!thyDraggable\"\n  [thyBeforeDragStart]=\"thyBeforeDragStart\"\n  [thyBeforeDragOver]=\"beforeDragOver\"\n  [thyBeforeDragDrop]=\"thyBeforeDragDrop\"\n  (thyDragStarted)=\"onDragStart($event)\"\n  (thyDragDropped)=\"onDragDrop($event)\"\n>\n  <thy-tree-node\n    [thyDrag]=\"node\"\n    *ngFor=\"let node of treeNodes; trackBy: trackByFn; let i = index\"\n    [node]=\"node\"\n    [templateRef]=\"templateRef\"\n    [emptyChildrenTemplateRef]=\"emptyChildrenTemplate\"\n    [thyAsync]=\"thyAsync\"\n    [thyDraggable]=\"thyDraggable\"\n    [thyCheckable]=\"thyCheckable\"\n    [thyMultiple]=\"thyMultiple\"\n    [thyTitleTruncate]=\"thyTitleTruncate\"\n    (thyOnClick)=\"thyOnClick.emit($event)\"\n    (thyOnExpandChange)=\"thyOnExpandChange.emit($event)\"\n    (thyOnCheckboxChange)=\"thyOnCheckboxChange.emit($event)\"\n  >\n  </thy-tree-node>\n</div>\n",
                encapsulation: core.ViewEncapsulation.None,
                providers: [
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core.forwardRef(function () { return ThyTreeComponent; }),
                        multi: true
                    },
                    ThyTreeService,
                    core$1.UpdateHostClassService
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.UpdateHostClassService }, { type: ThyTreeService }]; }, { thyShowExpand: [{
            type: core.Input
        }], thyMultiple: [{
            type: core.HostBinding,
            args: ["class.thy-multiple-selection-list"]
        }, {
            type: core.Input
        }], thyAsync: [{
            type: core.Input
        }], thyIcons: [{
            type: core.Input
        }], thyTitleTruncate: [{
            type: core.Input
        }], thyOnClick: [{
            type: core.Output
        }], thyOnCheckboxChange: [{
            type: core.Output
        }], thyOnExpandChange: [{
            type: core.Output
        }], thyOnDragDrop: [{
            type: core.Output
        }], thyTreeClass: [{
            type: core.HostBinding,
            args: ['class.thy-tree']
        }], thyNodes: [{
            type: core.Input
        }], thyDraggable: [{
            type: core.HostBinding,
            args: ['class.thy-tree-draggable']
        }, {
            type: core.Input
        }], thyCheckStateResolve: [{
            type: core.Input
        }], thyType: [{
            type: core.Input
        }], thySelectedKeys: [{
            type: core.Input
        }], templateRef: [{
            type: core.ContentChild,
            args: ['treeNodeTemplate', { static: true }]
        }], thyCheckable: [{
            type: core.Input
        }], thySize: [{
            type: core.Input
        }], thyBeforeDragStart: [{
            type: core.Input
        }], thyBeforeDragDrop: [{
            type: core.Input
        }], emptyChildrenTemplate: [{
            type: core.ContentChild,
            args: ['emptyChildrenTemplate', { static: true }]
        }] }); })();
        return ThyTreeComponent;
    }());
    ThyTreeComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core$1.UpdateHostClassService },
        { type: ThyTreeService }
    ]; };
    ThyTreeComponent.propDecorators = {
        thyNodes: [{ type: core.Input }],
        thyShowExpand: [{ type: core.Input }],
        thyMultiple: [{ type: core.HostBinding, args: ["class.thy-multiple-selection-list",] }, { type: core.Input }],
        thyDraggable: [{ type: core.HostBinding, args: ['class.thy-tree-draggable',] }, { type: core.Input }],
        thyCheckable: [{ type: core.Input }],
        thyCheckStateResolve: [{ type: core.Input }],
        thyAsync: [{ type: core.Input }],
        thyType: [{ type: core.Input }],
        thyIcons: [{ type: core.Input }],
        thySize: [{ type: core.Input }],
        thyTitleTruncate: [{ type: core.Input }],
        thySelectedKeys: [{ type: core.Input }],
        thyBeforeDragStart: [{ type: core.Input }],
        thyBeforeDragDrop: [{ type: core.Input }],
        thyOnClick: [{ type: core.Output }],
        thyOnCheckboxChange: [{ type: core.Output }],
        thyOnExpandChange: [{ type: core.Output }],
        thyOnDragDrop: [{ type: core.Output }],
        templateRef: [{ type: core.ContentChild, args: ['treeNodeTemplate', { static: true },] }],
        emptyChildrenTemplate: [{ type: core.ContentChild, args: ['emptyChildrenTemplate', { static: true },] }],
        thyTreeClass: [{ type: core.HostBinding, args: ['class.thy-tree',] }]
    };

    var ThyTreeNodeComponent = /** @class */ (function () {
        function ThyTreeNodeComponent(root, thyTreeService, ngZone, cdr) {
            var _this = this;
            this.root = root;
            this.thyTreeService = thyTreeService;
            this.ngZone = ngZone;
            this.cdr = cdr;
            this.thyAsync = false;
            this.thyMultiple = false;
            this.thyDraggable = false;
            this.thyCheckable = false;
            this.thyOnClick = new core.EventEmitter();
            this.thyOnExpandChange = new core.EventEmitter();
            this.thyOnCheckboxChange = new core.EventEmitter();
            this.thyTreeNodeClass = true;
            this.destroy$ = new rxjs.Subject();
            this.checkState = exports.ThyTreeNodeCheckState;
            this.thyTreeService
                .statusChanged()
                .pipe(operators.filter(function (data) { return data.node.key === _this.node.key; }), operators.takeUntil(this.destroy$))
                .subscribe(function () {
                _this.markForCheck();
            });
        }
        Object.defineProperty(ThyTreeNodeComponent.prototype, "nodeIcon", {
            get: function () {
                return this.node.origin.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyTreeNodeComponent.prototype, "nodeIconStyle", {
            get: function () {
                return this.node.origin.iconStyle;
            },
            enumerable: false,
            configurable: true
        });
        ThyTreeNodeComponent.prototype.markForCheck = function () {
            this.cdr.markForCheck();
        };
        ThyTreeNodeComponent.prototype.changeDragIconVisibility = function (event, showDragIcon) {
            var nodeElement = event.target;
            var dragIcon = nodeElement.querySelector('.thy-tree-drag-icon');
            if (dragIcon) {
                dragIcon.style.visibility = showDragIcon ? 'visible' : 'hidden';
            }
        };
        ThyTreeNodeComponent.prototype.nodeMouseEnter = function (event) {
            if (!this.root.thyDraggable || this.node.isDisabled) {
                return;
            }
            else if (this.root.thyDraggable && !this.root.thyBeforeDragStart) {
                this.changeDragIconVisibility(event, true);
            }
            else {
                var containerItems = this.node.getParentNode() ? this.node.getParentNode().getChildren() : this.root.treeNodes;
                var dragStartEvent = {
                    event: event,
                    item: this.node,
                    containerItems: containerItems,
                    currentIndex: containerItems.indexOf(this.node)
                };
                this.changeDragIconVisibility(event, this.root.thyBeforeDragStart(dragStartEvent));
            }
        };
        ThyTreeNodeComponent.prototype.nodeMouseLeave = function (event) {
            if (!this.root.thyDraggable || this.node.isDisabled) {
                return;
            }
            else {
                this.changeDragIconVisibility(event, false);
            }
        };
        ThyTreeNodeComponent.prototype.clickNode = function (event) {
            if (!this.root.thyMultiple) {
                this.root.selectTreeNode(this.node);
            }
            else {
                this.root.toggleTreeNode(this.node);
            }
            this.thyOnClick.emit({
                eventName: 'click',
                event: event,
                node: this.node
            });
        };
        ThyTreeNodeComponent.prototype.clickNodeCheck = function (event) {
            event.stopPropagation();
            if (this.node.isChecked === exports.ThyTreeNodeCheckState.unchecked || this.node.isChecked === exports.ThyTreeNodeCheckState.indeterminate) {
                this.node.setChecked(true);
            }
            else {
                this.node.setChecked(false);
            }
            this.thyOnCheckboxChange.emit({
                eventName: 'checkboxChange',
                event: event,
                node: this.node
            });
        };
        ThyTreeNodeComponent.prototype.expandNode = function (event) {
            event.stopPropagation();
            this.node.setExpanded(!this.node.isExpanded);
            if (this.root.thyShowExpand) {
                this.thyOnExpandChange.emit({
                    eventName: 'expand',
                    event: event,
                    node: this.node
                });
                if (this.thyAsync && this.node.children.length === 0) {
                    this.node.setLoading(true);
                }
            }
        };
        ThyTreeNodeComponent.prototype.isShowExpand = function (node) {
            return this.root.isShowExpand(node);
        };
        ThyTreeNodeComponent.prototype.ngOnDestroy = function () {
            this.destroy$.next();
            this.destroy$.complete();
        };
ThyTreeNodeComponent.ɵfac = function ThyTreeNodeComponent_Factory(t) { return new (t || ThyTreeNodeComponent)(ɵngcc0.ɵɵdirectiveInject(ThyTreeComponent), ɵngcc0.ɵɵdirectiveInject(ThyTreeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyTreeNodeComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTreeNodeComponent, selectors: [["thy-tree-node"]], contentQueries: function ThyTreeNodeComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.childrenTreeTemplateRef = _t.first);
    } }, viewQuery: function ThyTreeNodeComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c3, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleInputElementRef = _t.first);
    } }, hostVars: 2, hostBindings: function ThyTreeNodeComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-tree-node", ctx.thyTreeNodeClass);
    } }, inputs: { thyAsync: "thyAsync", thyMultiple: "thyMultiple", thyDraggable: "thyDraggable", thyCheckable: "thyCheckable", node: "node", thyTitleTruncate: "thyTitleTruncate", templateRef: "templateRef", emptyChildrenTemplateRef: "emptyChildrenTemplateRef" }, outputs: { thyOnClick: "thyOnClick", thyOnExpandChange: "thyOnExpandChange", thyOnCheckboxChange: "thyOnCheckboxChange" }, decls: 9, vars: 10, consts: [["thyDragContent", "", 1, "thy-tree-node-wrapper", "thy-sortable-item", 3, "click", "mouseenter", "mouseleave"], ["class", "thy-tree-drag-icon thy-sortable-handle", 4, "ngIf"], [1, "thy-tree-expand", 3, "click"], [4, "ngIf"], [1, "thy-tree-node-content"], ["type", "checkbox", "class", "thy-tree-node-check", 3, "form-check-indeterminate", "checked", "disabled", "click", 4, "ngIf"], ["class", "thy-tree-node-children", 3, "thyDropContainer", "thyDropContainerDisabled", "thyBeforeDragStart", "thyBeforeDragOver", "thyBeforeDragDrop", "thyDragDropped", "thyDragStarted", 4, "ngIf"], [1, "thy-tree-drag-icon", "thy-sortable-handle"], ["thyIconName", "drag"], ["class", "thy-tree-expand-icon", 3, "thyIconName", 4, "ngIf", "ngIfElse"], ["defaultIcon", ""], [1, "thy-tree-expand-icon", 3, "thyIconName"], ["thyIconName", "caret-right-down", 1, "thy-tree-expand-icon"], ["type", "checkbox", 1, "thy-tree-node-check", 3, "checked", "disabled", "click"], [1, "thy-tree-node-title", 3, "title"], [1, "thy-tree-node-icon"], [3, "ngClass", "ngStyle"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "thy-tree-node-children", 3, "thyDropContainer", "thyDropContainerDisabled", "thyBeforeDragStart", "thyBeforeDragOver", "thyBeforeDragDrop", "thyDragDropped", "thyDragStarted"], [3, "node", "thyDrag", "templateRef", "thyAsync", "thyMultiple", "thyDraggable", "thyCheckable", "thyTitleTruncate", "thyOnClick", "thyOnCheckboxChange", "thyOnExpandChange", 4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "node", "thyDrag", "templateRef", "thyAsync", "thyMultiple", "thyDraggable", "thyCheckable", "thyTitleTruncate", "thyOnClick", "thyOnCheckboxChange", "thyOnExpandChange"], [3, "thyDone", 4, "ngIf"], [3, "thyDone"]], template: function ThyTreeNodeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("click", function ThyTreeNodeComponent_Template_div_click_0_listener($event) { return ctx.clickNode($event); })("mouseenter", function ThyTreeNodeComponent_Template_div_mouseenter_0_listener($event) { return ctx.nodeMouseEnter($event); })("mouseleave", function ThyTreeNodeComponent_Template_div_mouseleave_0_listener($event) { return ctx.nodeMouseLeave($event); });
        ɵngcc0.ɵɵtemplate(1, ThyTreeNodeComponent_span_1_Template, 2, 0, "span", 1);
        ɵngcc0.ɵɵelementStart(2, "span", 2);
        ɵngcc0.ɵɵlistener("click", function ThyTreeNodeComponent_Template_span_click_2_listener($event) { return ctx.expandNode($event); });
        ɵngcc0.ɵɵtemplate(3, ThyTreeNodeComponent_ng_container_3_Template, 2, 1, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "span", 4);
        ɵngcc0.ɵɵtemplate(5, ThyTreeNodeComponent_input_5_Template, 1, 4, "input", 5);
        ɵngcc0.ɵɵtemplate(6, ThyTreeNodeComponent_ng_container_6_Template, 4, 5, "ng-container", 3);
        ɵngcc0.ɵɵtemplate(7, ThyTreeNodeComponent_ng_container_7_Template, 2, 6, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, ThyTreeNodeComponent_div_8_Template, 3, 8, "div", 6);
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("active", ctx.root.isSelected(ctx.node))("disabled", ctx.node.isDisabled);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyDraggable);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isShowExpand(ctx.node));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyCheckable);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.templateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.templateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isShowExpand(ctx.node) && ctx.node.isExpanded);
    } }, directives: [ɵngcc2.ThyDragContentDirective, ɵngcc3.NgIf, ɵngcc4.ThyIconComponent, ɵngcc3.NgClass, ɵngcc3.NgStyle, ɵngcc3.NgTemplateOutlet, ɵngcc2.ThyDropContainerDirective, ɵngcc3.NgForOf, ThyTreeNodeComponent, ɵngcc2.ThyDragDirective, ɵngcc5.ThyLoadingComponent], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeNodeComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-tree-node',
                template: "<div\n  thyDragContent\n  class=\"thy-tree-node-wrapper thy-sortable-item\"\n  [class.active]=\"root.isSelected(node)\"\n  [class.disabled]=\"node.isDisabled\"\n  (click)=\"clickNode($event)\"\n  (mouseenter)=\"nodeMouseEnter($event)\"\n  (mouseleave)=\"nodeMouseLeave($event)\"\n>\n  <span *ngIf=\"thyDraggable\" class=\"thy-tree-drag-icon thy-sortable-handle\">\n    <thy-icon thyIconName=\"drag\"></thy-icon>\n  </span>\n  <span class=\"thy-tree-expand\" (click)=\"expandNode($event)\">\n    <ng-container *ngIf=\"isShowExpand(node)\">\n      <ng-container *ngIf=\"(node.children && node.children.length > 0) || thyAsync || emptyChildrenTemplateRef\">\n        <thy-icon\n          *ngIf=\"root.thyIcons?.expand && root.thyIcons?.collapse; else defaultIcon\"\n          class=\"thy-tree-expand-icon\"\n          [thyIconName]=\"node.isExpanded ? root.thyIcons.expand : root.thyIcons.collapse\"\n        ></thy-icon>\n        <ng-template #defaultIcon>\n          <thy-icon class=\"thy-tree-expand-icon\" [class.collapse]=\"!node.isExpanded\" thyIconName=\"caret-right-down\"></thy-icon>\n        </ng-template>\n      </ng-container>\n    </ng-container>\n  </span>\n\n  <span class=\"thy-tree-node-content\">\n    <input\n      *ngIf=\"thyCheckable\"\n      type=\"checkbox\"\n      class=\"thy-tree-node-check\"\n      [class.form-check-indeterminate]=\"node.isChecked === checkState.indeterminate\"\n      [checked]=\"node.isChecked === checkState.checked\"\n      [disabled]=\"node.isDisabled\"\n      (click)=\"clickNodeCheck($event)\"\n    />\n    <ng-container *ngIf=\"!templateRef\">\n      <ng-container *ngIf=\"nodeIcon\">\n        <span class=\"thy-tree-node-icon\"> <i [ngClass]=\"nodeIcon\" [ngStyle]=\"nodeIconStyle\"></i></span>\n      </ng-container>\n      <span [title]=\"node.title\" class=\"thy-tree-node-title\" [class.truncate]=\"thyTitleTruncate\"> {{ node.title }}</span>\n    </ng-container>\n    <ng-container *ngIf=\"templateRef\">\n      <ng-template\n        [ngTemplateOutlet]=\"templateRef\"\n        [ngTemplateOutletContext]=\"{ $implicit: node, node: node, origin: node?.origin }\"\n      ></ng-template>\n    </ng-container>\n  </span>\n</div>\n<div\n  class=\"thy-tree-node-children\"\n  *ngIf=\"isShowExpand(node) && node.isExpanded\"\n  [thyDropContainer]=\"node.children\"\n  [thyDropContainerDisabled]=\"!root.thyDraggable\"\n  [thyBeforeDragStart]=\"root.thyBeforeDragStart\"\n  [thyBeforeDragOver]=\"root.beforeDragOver\"\n  [thyBeforeDragDrop]=\"root.thyBeforeDragDrop\"\n  (thyDragDropped)=\"root.onDragDrop($event)\"\n  (thyDragStarted)=\"root.onDragStart($event)\"\n>\n  <thy-tree-node\n    *ngFor=\"let node of node.children; trackBy: root.trackByFn; let i = index\"\n    [node]=\"node\"\n    [thyDrag]=\"node\"\n    [templateRef]=\"templateRef\"\n    [thyAsync]=\"thyAsync\"\n    [thyMultiple]=\"thyMultiple\"\n    [thyDraggable]=\"thyDraggable\"\n    [thyCheckable]=\"thyCheckable\"\n    [thyTitleTruncate]=\"thyTitleTruncate\"\n    (thyOnClick)=\"thyOnClick.emit($event)\"\n    (thyOnCheckboxChange)=\"thyOnCheckboxChange.emit($event)\"\n    (thyOnExpandChange)=\"thyOnExpandChange.emit($event)\"\n  >\n  </thy-tree-node>\n  <ng-container *ngIf=\"thyAsync\">\n    <thy-loading *ngIf=\"node.isLoading\" [thyDone]=\"!node.isLoading\"></thy-loading>\n  </ng-container>\n</div>\n",
                encapsulation: core.ViewEncapsulation.None
            }]
    }], function () { return [{ type: ThyTreeComponent }, { type: ThyTreeService }, { type: ɵngcc0.NgZone }, { type: ɵngcc0.ChangeDetectorRef }]; }, { thyAsync: [{
            type: core.Input
        }], thyMultiple: [{
            type: core.Input
        }], thyDraggable: [{
            type: core.Input
        }], thyCheckable: [{
            type: core.Input
        }], thyOnClick: [{
            type: core.Output
        }], thyOnExpandChange: [{
            type: core.Output
        }], thyOnCheckboxChange: [{
            type: core.Output
        }], thyTreeNodeClass: [{
            type: core.HostBinding,
            args: ['class.thy-tree-node']
        }], node: [{
            type: core.Input
        }], thyTitleTruncate: [{
            type: core.Input
        }], templateRef: [{
            type: core.Input
        }], emptyChildrenTemplateRef: [{
            type: core.Input
        }], childrenTreeTemplateRef: [{
            type: core.ContentChild,
            args: ['childrenTree']
        }], titleInputElementRef: [{
            type: core.ViewChild,
            args: ['title']
        }] }); })();
        return ThyTreeNodeComponent;
    }());
    ThyTreeNodeComponent.ctorParameters = function () { return [
        { type: ThyTreeComponent },
        { type: ThyTreeService },
        { type: core.NgZone },
        { type: core.ChangeDetectorRef }
    ]; };
    ThyTreeNodeComponent.propDecorators = {
        node: [{ type: core.Input }],
        thyAsync: [{ type: core.Input }],
        thyMultiple: [{ type: core.Input }],
        thyDraggable: [{ type: core.Input }],
        thyCheckable: [{ type: core.Input }],
        thyTitleTruncate: [{ type: core.Input }],
        templateRef: [{ type: core.Input }],
        emptyChildrenTemplateRef: [{ type: core.Input }],
        thyOnClick: [{ type: core.Output }],
        thyOnExpandChange: [{ type: core.Output }],
        thyOnCheckboxChange: [{ type: core.Output }],
        childrenTreeTemplateRef: [{ type: core.ContentChild, args: ['childrenTree',] }],
        titleInputElementRef: [{ type: core.ViewChild, args: ['title',] }],
        thyTreeNodeClass: [{ type: core.HostBinding, args: ['class.thy-tree-node',] }]
    };

    var ThyTreeReplaceRegionComponent = /** @class */ (function () {
        function ThyTreeReplaceRegionComponent(root, viewRef) {
            this.root = root;
            this.viewRef = viewRef;
        }
        ThyTreeReplaceRegionComponent.prototype.ngOnInit = function () { };
ThyTreeReplaceRegionComponent.ɵfac = function ThyTreeReplaceRegionComponent_Factory(t) { return new (t || ThyTreeReplaceRegionComponent)(ɵngcc0.ɵɵdirectiveInject(ThyTreeComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ThyTreeReplaceRegionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTreeReplaceRegionComponent, selectors: [["", "ThyTreeReplaceRegion", ""]], inputs: { thyInstance: "thyInstance" }, attrs: _c5, ngContentSelectors: _c6, decls: 1, vars: 0, template: function ThyTreeReplaceRegionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeReplaceRegionComponent, [{
        type: core.Component,
        args: [{
                selector: '[ThyTreeReplaceRegion]',
                template: '<ng-content></ng-content>',
                encapsulation: core.ViewEncapsulation.None
            }]
    }], function () { return [{ type: ThyTreeComponent }, { type: ɵngcc0.ViewContainerRef }]; }, { thyInstance: [{
            type: core.Input
        }] }); })();
        return ThyTreeReplaceRegionComponent;
    }());
    ThyTreeReplaceRegionComponent.ctorParameters = function () { return [
        { type: ThyTreeComponent },
        { type: core.ViewContainerRef }
    ]; };
    ThyTreeReplaceRegionComponent.propDecorators = {
        thyInstance: [{ type: core.Input }]
    };

    var ThyTreeModule = /** @class */ (function () {
        function ThyTreeModule() {
        }
ThyTreeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTreeModule });
ThyTreeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTreeModule_Factory(t) { return new (t || ThyTreeModule)(); }, providers: [ThyTreeService], imports: [[
            common.CommonModule,
            input.ThyInputModule,
            button.ThyButtonModule,
            shared.ThySharedModule,
            forms.FormsModule,
            list.ThyListModule,
            shared.ThyOptionModule,
            loading.ThyLoadingModule,
            icon.ThyIconModule,
            dragDrop.ThyDragDropModule,
            checkbox.ThyCheckboxModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTreeModule, { declarations: [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent], imports: [ɵngcc3.CommonModule, ɵngcc6.ThyInputModule, ɵngcc7.ThyButtonModule, ɵngcc8.ThySharedModule, ɵngcc9.FormsModule, ɵngcc10.ThyListModule, ɵngcc8.ThyOptionModule, ɵngcc5.ThyLoadingModule, ɵngcc4.ThyIconModule, ɵngcc2.ThyDragDropModule, ɵngcc11.ThyCheckboxModule], exports: [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent],
                imports: [
                    common.CommonModule,
                    input.ThyInputModule,
                    button.ThyButtonModule,
                    shared.ThySharedModule,
                    forms.FormsModule,
                    list.ThyListModule,
                    shared.ThyOptionModule,
                    loading.ThyLoadingModule,
                    icon.ThyIconModule,
                    dragDrop.ThyDragDropModule,
                    checkbox.ThyCheckboxModule
                ],
                entryComponents: [ThyTreeComponent],
                exports: [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent],
                providers: [ThyTreeService]
            }]
    }], function () { return []; }, null); })();
        return ThyTreeModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyTreeComponent = ThyTreeComponent;
    exports.ThyTreeIcons = ThyTreeIcons;
    exports.ThyTreeModule = ThyTreeModule;
    exports.ThyTreeNode = ThyTreeNode;
    exports.ThyTreeNodeComponent = ThyTreeNodeComponent;
    exports.ThyTreeReplaceRegionComponent = ThyTreeReplaceRegionComponent;
    exports.ThyTreeService = ThyTreeService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
