// ngx-tethys/bundles/ngx-tethys.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/layout'),require('ngx-tethys/button'),require('ngx-tethys/back-top'),require('ngx-tethys/icon'),require('ngx-tethys/popover'),require('ngx-tethys/badge'),require('ngx-tethys/table'),require('ngx-tethys/raster'),require('ngx-tethys/avatar'),require('ngx-tethys/label'),require('ngx-tethys/nav'),require('ngx-tethys/menu'),require('ngx-tethys/pagination'),require('ngx-tethys/notify'),require('ngx-tethys/card'),require('ngx-tethys/loading'),require('ngx-tethys/alert'),require('ngx-tethys/action-menu'),require('ngx-tethys/tree'),require('ngx-tethys/empty'),require('ngx-tethys/switch'),require('ngx-tethys/transfer'),require('ngx-tethys/strength'),require('ngx-tethys/form'),require('ngx-tethys/input'),require('ngx-tethys/dropdown'),require('ngx-tethys/copy'),require('ngx-tethys/checkbox'),require('ngx-tethys/select'),require('ngx-tethys/slide'),require('ngx-tethys/radio'),require('ngx-tethys/property-operation'),require('ngx-tethys/uploader'),require('ngx-tethys/date-range'),require('ngx-tethys/shared'),require('ngx-tethys/list'),require('ngx-tethys/tree-select'),require('ngx-tethys/stepper'),require('ngx-tethys/cascader'),require('ngx-tethys/dialog'),require('ngx-tethys/tooltip'),require('ngx-tethys/progress'),require('ngx-tethys/breadcrumb'),require('ngx-tethys/arrow-switcher'),require('ngx-tethys/markdown'),require('ngx-tethys/flexible-text'),require('ngx-tethys/drag-drop'),require('ngx-tethys/skeleton'),require('ngx-tethys/vote'),require('ngx-tethys/result'),require('ngx-tethys/mention'),require('ngx-tethys/date-picker'),require('ngx-tethys/timeline'),require('ngx-tethys/divider'),require('ngx-tethys/time-picker'),require('ngx-tethys/statistic'),require('ngx-tethys/autocomplete'),require('ngx-tethys/anchor'),require('ngx-tethys/affix'),require('ngx-tethys/slider'),require('ngx-tethys/calendar'),require('ngx-tethys/fullscreen'),require('ngx-tethys/guider'),require('ngx-tethys/resizable'),exports, require('ngx-tethys/util'), require('@angular/core'), require('ngx-tethys/action-menu'), require('ngx-tethys/affix'), require('ngx-tethys/alert'), require('ngx-tethys/anchor'), require('ngx-tethys/arrow-switcher'), require('ngx-tethys/autocomplete'), require('ngx-tethys/avatar'), require('ngx-tethys/back-top'), require('ngx-tethys/badge'), require('ngx-tethys/breadcrumb'), require('ngx-tethys/button'), require('ngx-tethys/card'), require('ngx-tethys/cascader'), require('ngx-tethys/checkbox'), require('ngx-tethys/shared'), require('ngx-tethys/date-picker'), require('ngx-tethys/date-range'), require('ngx-tethys/dialog'), require('ngx-tethys/drag-drop'), require('ngx-tethys/dropdown'), require('ngx-tethys/copy'), require('ngx-tethys/empty'), require('ngx-tethys/flexible-text'), require('ngx-tethys/form'), require('ngx-tethys/table'), require('ngx-tethys/icon'), require('ngx-tethys/input'), require('ngx-tethys/label'), require('ngx-tethys/layout'), require('ngx-tethys/list'), require('ngx-tethys/loading'), require('ngx-tethys/markdown'), require('ngx-tethys/mention'), require('ngx-tethys/menu'), require('ngx-tethys/nav'), require('ngx-tethys/notify'), require('ngx-tethys/pagination'), require('ngx-tethys/popover'), require('ngx-tethys/progress'), require('ngx-tethys/property-operation'), require('ngx-tethys/radio'), require('ngx-tethys/raster'), require('ngx-tethys/result'), require('ngx-tethys/select'), require('ngx-tethys/skeleton'), require('ngx-tethys/slide'), require('ngx-tethys/statistic'), require('ngx-tethys/stepper'), require('ngx-tethys/strength'), require('ngx-tethys/switch'), require('ngx-tethys/time-picker'), require('ngx-tethys/tooltip'), require('ngx-tethys/transfer'), require('ngx-tethys/tree-select'), require('ngx-tethys/tree'), require('ngx-tethys/uploader'), require('ngx-tethys/vote'), require('ngx-tethys/slider'), require('ngx-tethys/calendar'), require('ngx-tethys/timeline'), require('ngx-tethys/divider'), require('ngx-tethys/fullscreen'), require('ngx-tethys/guider'), require('ngx-tethys/resizable'), require('ngx-tethys/core'), require('ngx-tethys/positioning'), require('ngx-tethys/store')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys', ['@angular/core','ngx-tethys/layout','ngx-tethys/button','ngx-tethys/back-top','ngx-tethys/icon','ngx-tethys/popover','ngx-tethys/badge','ngx-tethys/table','ngx-tethys/raster','ngx-tethys/avatar','ngx-tethys/label','ngx-tethys/nav','ngx-tethys/menu','ngx-tethys/pagination','ngx-tethys/notify','ngx-tethys/card','ngx-tethys/loading','ngx-tethys/alert','ngx-tethys/action-menu','ngx-tethys/tree','ngx-tethys/empty','ngx-tethys/switch','ngx-tethys/transfer','ngx-tethys/strength','ngx-tethys/form','ngx-tethys/input','ngx-tethys/dropdown','ngx-tethys/copy','ngx-tethys/checkbox','ngx-tethys/select','ngx-tethys/slide','ngx-tethys/radio','ngx-tethys/property-operation','ngx-tethys/uploader','ngx-tethys/date-range','ngx-tethys/shared','ngx-tethys/list','ngx-tethys/tree-select','ngx-tethys/stepper','ngx-tethys/cascader','ngx-tethys/dialog','ngx-tethys/tooltip','ngx-tethys/progress','ngx-tethys/breadcrumb','ngx-tethys/arrow-switcher','ngx-tethys/markdown','ngx-tethys/flexible-text','ngx-tethys/drag-drop','ngx-tethys/skeleton','ngx-tethys/vote','ngx-tethys/result','ngx-tethys/mention','ngx-tethys/date-picker','ngx-tethys/timeline','ngx-tethys/divider','ngx-tethys/time-picker','ngx-tethys/statistic','ngx-tethys/autocomplete','ngx-tethys/anchor','ngx-tethys/affix','ngx-tethys/slider','ngx-tethys/calendar','ngx-tethys/fullscreen','ngx-tethys/guider','ngx-tethys/resizable','exports', 'ngx-tethys/util', '@angular/core', 'ngx-tethys/action-menu', 'ngx-tethys/affix', 'ngx-tethys/alert', 'ngx-tethys/anchor', 'ngx-tethys/arrow-switcher', 'ngx-tethys/autocomplete', 'ngx-tethys/avatar', 'ngx-tethys/back-top', 'ngx-tethys/badge', 'ngx-tethys/breadcrumb', 'ngx-tethys/button', 'ngx-tethys/card', 'ngx-tethys/cascader', 'ngx-tethys/checkbox', 'ngx-tethys/shared', 'ngx-tethys/date-picker', 'ngx-tethys/date-range', 'ngx-tethys/dialog', 'ngx-tethys/drag-drop', 'ngx-tethys/dropdown', 'ngx-tethys/copy', 'ngx-tethys/empty', 'ngx-tethys/flexible-text', 'ngx-tethys/form', 'ngx-tethys/table', 'ngx-tethys/icon', 'ngx-tethys/input', 'ngx-tethys/label', 'ngx-tethys/layout', 'ngx-tethys/list', 'ngx-tethys/loading', 'ngx-tethys/markdown', 'ngx-tethys/mention', 'ngx-tethys/menu', 'ngx-tethys/nav', 'ngx-tethys/notify', 'ngx-tethys/pagination', 'ngx-tethys/popover', 'ngx-tethys/progress', 'ngx-tethys/property-operation', 'ngx-tethys/radio', 'ngx-tethys/raster', 'ngx-tethys/result', 'ngx-tethys/select', 'ngx-tethys/skeleton', 'ngx-tethys/slide', 'ngx-tethys/statistic', 'ngx-tethys/stepper', 'ngx-tethys/strength', 'ngx-tethys/switch', 'ngx-tethys/time-picker', 'ngx-tethys/tooltip', 'ngx-tethys/transfer', 'ngx-tethys/tree-select', 'ngx-tethys/tree', 'ngx-tethys/uploader', 'ngx-tethys/vote', 'ngx-tethys/slider', 'ngx-tethys/calendar', 'ngx-tethys/timeline', 'ngx-tethys/divider', 'ngx-tethys/fullscreen', 'ngx-tethys/guider', 'ngx-tethys/resizable', 'ngx-tethys/core', 'ngx-tethys/positioning', 'ngx-tethys/store'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.layout,global.ngxTethys.button,global.ngxTethys.backTop,global.ngxTethys.icon,global.ngxTethys.popover,global.ngxTethys.badge,global.ngxTethys.table,global.ngxTethys.raster,global.ngxTethys.avatar,global.ngxTethys.label,global.ngxTethys.nav,global.ngxTethys.menu,global.ngxTethys.pagination,global.ngxTethys.notify,global.ngxTethys.card,global.ngxTethys.loading,global.ngxTethys.alert,global.ngxTethys.actionMenu,global.ngxTethys.tree,global.ngxTethys.empty,global.ngxTethys.switch,global.ngxTethys.transfer,global.ngxTethys.strength,global.ngxTethys.form,global.ngxTethys.input,global.ngxTethys.dropdown,global.ngxTethys.copy,global.ngxTethys.checkbox,global.ngxTethys.select,global.ngxTethys.slide,global.ngxTethys.radio,global.ngxTethys.propertyOperation,global.ngxTethys.uploader,global.ngxTethys.dateRange,global.ngxTethys.shared,global.ngxTethys.list,global.ngxTethys.treeSelect,global.ngxTethys.stepper,global.ngxTethys.cascader,global.ngxTethys.dialog,global.ngxTethys.tooltip,global.ngxTethys.progress,global.ngxTethys.breadcrumb,global.ngxTethys.arrowSwitcher,global.ngxTethys.markdown,global.ngxTethys.flexibleText,global.ngxTethys.dragDrop,global.ngxTethys.skeleton,global.ngxTethys.vote,global.ngxTethys.result,global.ngxTethys.mention,global.ngxTethys.datePicker,global.ngxTethys.timeline,global.ngxTethys.divider,global.ngxTethys.timePicker,global.ngxTethys.statistic,global.ngxTethys.autocomplete,global.ngxTethys.anchor,global.ngxTethys.affix,global.ngxTethys.slider,global.ngxTethys.calendar,global.ngxTethys.fullscreen,global.ngxTethys.guider,global.ngxTethys.resizable,global['ngx-tethys'] = {}, global['ngx-tethys'].util, global.ng.core, global['ngx-tethys']['action-menu'], global['ngx-tethys'].affix, global['ngx-tethys'].alert, global['ngx-tethys'].anchor, global['ngx-tethys']['arrow-switcher'], global['ngx-tethys'].autocomplete, global['ngx-tethys'].avatar, global['ngx-tethys']['back-top'], global['ngx-tethys'].badge, global['ngx-tethys'].breadcrumb, global['ngx-tethys'].button, global['ngx-tethys'].card, global['ngx-tethys'].cascader, global['ngx-tethys'].checkbox, global['ngx-tethys'].shared, global['ngx-tethys']['date-picker'], global['ngx-tethys']['date-range'], global['ngx-tethys'].dialog, global['ngx-tethys']['drag-drop'], global['ngx-tethys'].dropdown, global['ngx-tethys'].copy, global['ngx-tethys'].empty, global['ngx-tethys']['flexible-text'], global['ngx-tethys'].form, global['ngx-tethys'].table, global['ngx-tethys'].icon, global['ngx-tethys'].input, global['ngx-tethys'].label, global['ngx-tethys'].layout, global['ngx-tethys'].list, global['ngx-tethys'].loading, global['ngx-tethys'].markdown, global['ngx-tethys'].mention, global['ngx-tethys'].menu, global['ngx-tethys'].nav, global['ngx-tethys'].notify, global['ngx-tethys'].pagination, global['ngx-tethys'].popover, global['ngx-tethys'].progress, global['ngx-tethys']['property-operation'], global['ngx-tethys'].radio, global['ngx-tethys'].raster, global['ngx-tethys'].result, global['ngx-tethys'].select, global['ngx-tethys'].skeleton, global['ngx-tethys'].slide, global['ngx-tethys'].statistic, global['ngx-tethys'].stepper, global['ngx-tethys'].strength, global['ngx-tethys'].switch, global['ngx-tethys']['time-picker'], global['ngx-tethys'].tooltip, global['ngx-tethys'].transfer, global['ngx-tethys']['tree-select'], global['ngx-tethys'].tree, global['ngx-tethys'].uploader, global['ngx-tethys'].vote, global['ngx-tethys'].slider, global['ngx-tethys'].calendar, global['ngx-tethys'].timeline, global['ngx-tethys'].divider, global['ngx-tethys'].fullscreen, global['ngx-tethys'].guider, global['ngx-tethys'].resizable, global['ngx-tethys'].core, global['ngx-tethys'].positioning, global['ngx-tethys'].store));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,ɵngcc10,ɵngcc11,ɵngcc12,ɵngcc13,ɵngcc14,ɵngcc15,ɵngcc16,ɵngcc17,ɵngcc18,ɵngcc19,ɵngcc20,ɵngcc21,ɵngcc22,ɵngcc23,ɵngcc24,ɵngcc25,ɵngcc26,ɵngcc27,ɵngcc28,ɵngcc29,ɵngcc30,ɵngcc31,ɵngcc32,ɵngcc33,ɵngcc34,ɵngcc35,ɵngcc36,ɵngcc37,ɵngcc38,ɵngcc39,ɵngcc40,ɵngcc41,ɵngcc42,ɵngcc43,ɵngcc44,ɵngcc45,ɵngcc46,ɵngcc47,ɵngcc48,ɵngcc49,ɵngcc50,ɵngcc51,ɵngcc52,ɵngcc53,ɵngcc54,ɵngcc55,ɵngcc56,ɵngcc57,ɵngcc58,ɵngcc59,ɵngcc60,ɵngcc61,ɵngcc62,ɵngcc63,ɵngcc64,exports, util, core$1, actionMenu, affix, alert, anchor, arrowSwitcher, autocomplete, avatar, backTop, badge, breadcrumb, button, card, cascader, checkbox, shared, datePicker, dateRange, dialog, dragDrop, dropdown, copy, empty, flexibleText, form, table, icon, input, label, layout, list, loading, markdown, mention, menu, nav, notify, pagination, popover, progress, propertyOperation, radio, raster, result, select, skeleton, slide, statistic, stepper, strength, _switch, timePicker, tooltip, transfer, treeSelect, tree, uploader, vote, slider, calendar, timeline, divider, fullscreen, guider, resizable, core, positioning, store) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var IMPORT_EXPORT_MODULES = [
        layout.ThyLayoutModule,
        button.ThyButtonModule,
        backTop.ThyBackTopModule,
        icon.ThyIconModule,
        popover.ThyPopoverModule,
        badge.ThyBadgeModule,
        table.ThyTableModule,
        raster.ThyRasterModule,
        avatar.ThyAvatarModule,
        label.ThyLabelModule,
        nav.ThyNavModule,
        menu.ThyMenuModule,
        pagination.ThyPaginationModule,
        notify.ThyNotifyModule,
        card.ThyCardModule,
        loading.ThyLoadingModule,
        alert.ThyAlertModule,
        actionMenu.ThyActionMenuModule,
        tree.ThyTreeModule,
        empty.ThyEmptyModule,
        _switch.ThySwitchModule,
        transfer.ThyTransferModule,
        strength.ThyStrengthModule,
        form.ThyFormModule,
        input.ThyInputModule,
        dropdown.ThyDropdownModule,
        copy.ThyCopyModule,
        checkbox.ThyCheckboxModule,
        select.ThySelectModule,
        slide.ThySlideModule,
        radio.ThyRadioModule,
        select.ThySelectModule,
        propertyOperation.ThyPropertyOperationModule,
        uploader.ThyUploaderModule,
        dateRange.ThyDateRangeModule,
        shared.ThySharedModule,
        list.ThyListModule,
        treeSelect.ThyTreeSelectModule,
        stepper.ThyStepperModule,
        cascader.ThyCascaderModule,
        dialog.ThyDialogModule,
        tooltip.ThyTooltipModule,
        progress.ThyProgressModule,
        breadcrumb.ThyBreadcrumbModule,
        arrowSwitcher.ThyArrowSwitcherModule,
        markdown.ThyMarkdownModule,
        flexibleText.ThyFlexibleTextModule,
        dragDrop.ThyDragDropModule,
        shared.ThySelectCommonModule,
        skeleton.ThySkeletonModule,
        vote.ThyVoteModule,
        result.ThyResultModule,
        mention.ThyMentionModule,
        datePicker.ThyDatePickerModule,
        timeline.ThyTimelineModule,
        divider.ThyDividerModule,
        timePicker.ThyTimePickerModule,
        statistic.ThyStatisticModule,
        autocomplete.ThyAutocompleteModule,
        anchor.ThyAnchorModule,
        affix.ThyAffixModule,
        slider.ThySliderModule,
        calendar.ThyCalendarModule,
        fullscreen.ThyFullscreenModule,
        guider.ThyGuiderModule,
        resizable.ThyResizableModule
    ];
    /**
     * @deprecated root module is deprecated, please use component module
     */
    var NgxTethysModule = /** @class */ (function () {
        function NgxTethysModule() {
            util.warnDeprecation('The NgxTethysModule is deprecated, please use second entry point, we will remove main entry point in next major version');
        }
        NgxTethysModule.forRoot = function () {
            return {
                ngModule: NgxTethysModule,
                providers: []
            };
        };
NgxTethysModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxTethysModule });
NgxTethysModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxTethysModule_Factory(t) { return new (t || NgxTethysModule)(); }, providers: [], imports: [__spread(IMPORT_EXPORT_MODULES), ɵngcc1.ThyLayoutModule, ɵngcc2.ThyButtonModule, ɵngcc3.ThyBackTopModule, ɵngcc4.ThyIconModule, ɵngcc5.ThyPopoverModule, ɵngcc6.ThyBadgeModule, ɵngcc7.ThyTableModule, ɵngcc8.ThyRasterModule, ɵngcc9.ThyAvatarModule, ɵngcc10.ThyLabelModule, ɵngcc11.ThyNavModule, ɵngcc12.ThyMenuModule, ɵngcc13.ThyPaginationModule, ɵngcc14.ThyNotifyModule, ɵngcc15.ThyCardModule, ɵngcc16.ThyLoadingModule, ɵngcc17.ThyAlertModule, ɵngcc18.ThyActionMenuModule, ɵngcc19.ThyTreeModule, ɵngcc20.ThyEmptyModule, ɵngcc21.ThySwitchModule, ɵngcc22.ThyTransferModule, ɵngcc23.ThyStrengthModule, ɵngcc24.ThyFormModule, ɵngcc25.ThyInputModule, ɵngcc26.ThyDropdownModule, ɵngcc27.ThyCopyModule, ɵngcc28.ThyCheckboxModule, ɵngcc29.ThySelectModule, ɵngcc30.ThySlideModule, ɵngcc31.ThyRadioModule, ɵngcc29.ThySelectModule, ɵngcc32.ThyPropertyOperationModule, ɵngcc33.ThyUploaderModule, ɵngcc34.ThyDateRangeModule, ɵngcc35.ThySharedModule, ɵngcc36.ThyListModule, ɵngcc37.ThyTreeSelectModule, ɵngcc38.ThyStepperModule, ɵngcc39.ThyCascaderModule, ɵngcc40.ThyDialogModule, ɵngcc41.ThyTooltipModule, ɵngcc42.ThyProgressModule, ɵngcc43.ThyBreadcrumbModule, ɵngcc44.ThyArrowSwitcherModule, ɵngcc45.ThyMarkdownModule, ɵngcc46.ThyFlexibleTextModule, ɵngcc47.ThyDragDropModule, ɵngcc35.ThySelectCommonModule, ɵngcc48.ThySkeletonModule, ɵngcc49.ThyVoteModule, ɵngcc50.ThyResultModule, ɵngcc51.ThyMentionModule, ɵngcc52.ThyDatePickerModule, ɵngcc53.ThyTimelineModule, ɵngcc54.ThyDividerModule, ɵngcc55.ThyTimePickerModule, ɵngcc56.ThyStatisticModule, ɵngcc57.ThyAutocompleteModule, ɵngcc58.ThyAnchorModule, ɵngcc59.ThyAffixModule, ɵngcc60.ThySliderModule, ɵngcc61.ThyCalendarModule, ɵngcc62.ThyFullscreenModule, ɵngcc63.ThyGuiderModule, ɵngcc64.ThyResizableModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxTethysModule, { imports: [ɵngcc1.ThyLayoutModule, ɵngcc2.ThyButtonModule, ɵngcc3.ThyBackTopModule, ɵngcc4.ThyIconModule, ɵngcc5.ThyPopoverModule, ɵngcc6.ThyBadgeModule, ɵngcc7.ThyTableModule, ɵngcc8.ThyRasterModule, ɵngcc9.ThyAvatarModule, ɵngcc10.ThyLabelModule, ɵngcc11.ThyNavModule, ɵngcc12.ThyMenuModule, ɵngcc13.ThyPaginationModule, ɵngcc14.ThyNotifyModule, ɵngcc15.ThyCardModule, ɵngcc16.ThyLoadingModule, ɵngcc17.ThyAlertModule, ɵngcc18.ThyActionMenuModule, ɵngcc19.ThyTreeModule, ɵngcc20.ThyEmptyModule, ɵngcc21.ThySwitchModule, ɵngcc22.ThyTransferModule, ɵngcc23.ThyStrengthModule, ɵngcc24.ThyFormModule, ɵngcc25.ThyInputModule, ɵngcc26.ThyDropdownModule, ɵngcc27.ThyCopyModule, ɵngcc28.ThyCheckboxModule, ɵngcc29.ThySelectModule, ɵngcc30.ThySlideModule, ɵngcc31.ThyRadioModule, ɵngcc29.ThySelectModule, ɵngcc32.ThyPropertyOperationModule, ɵngcc33.ThyUploaderModule, ɵngcc34.ThyDateRangeModule, ɵngcc35.ThySharedModule, ɵngcc36.ThyListModule, ɵngcc37.ThyTreeSelectModule, ɵngcc38.ThyStepperModule, ɵngcc39.ThyCascaderModule, ɵngcc40.ThyDialogModule, ɵngcc41.ThyTooltipModule, ɵngcc42.ThyProgressModule, ɵngcc43.ThyBreadcrumbModule, ɵngcc44.ThyArrowSwitcherModule, ɵngcc45.ThyMarkdownModule, ɵngcc46.ThyFlexibleTextModule, ɵngcc47.ThyDragDropModule, ɵngcc35.ThySelectCommonModule, ɵngcc48.ThySkeletonModule, ɵngcc49.ThyVoteModule, ɵngcc50.ThyResultModule, ɵngcc51.ThyMentionModule, ɵngcc52.ThyDatePickerModule, ɵngcc53.ThyTimelineModule, ɵngcc54.ThyDividerModule, ɵngcc55.ThyTimePickerModule, ɵngcc56.ThyStatisticModule, ɵngcc57.ThyAutocompleteModule, ɵngcc58.ThyAnchorModule, ɵngcc59.ThyAffixModule, ɵngcc60.ThySliderModule, ɵngcc61.ThyCalendarModule, ɵngcc62.ThyFullscreenModule, ɵngcc63.ThyGuiderModule, ɵngcc64.ThyResizableModule], exports: [ɵngcc1.ThyLayoutModule, ɵngcc2.ThyButtonModule, ɵngcc3.ThyBackTopModule, ɵngcc4.ThyIconModule, ɵngcc5.ThyPopoverModule, ɵngcc6.ThyBadgeModule, ɵngcc7.ThyTableModule, ɵngcc8.ThyRasterModule, ɵngcc9.ThyAvatarModule, ɵngcc10.ThyLabelModule, ɵngcc11.ThyNavModule, ɵngcc12.ThyMenuModule, ɵngcc13.ThyPaginationModule, ɵngcc14.ThyNotifyModule, ɵngcc15.ThyCardModule, ɵngcc16.ThyLoadingModule, ɵngcc17.ThyAlertModule, ɵngcc18.ThyActionMenuModule, ɵngcc19.ThyTreeModule, ɵngcc20.ThyEmptyModule, ɵngcc21.ThySwitchModule, ɵngcc22.ThyTransferModule, ɵngcc23.ThyStrengthModule, ɵngcc24.ThyFormModule, ɵngcc25.ThyInputModule, ɵngcc26.ThyDropdownModule, ɵngcc27.ThyCopyModule, ɵngcc28.ThyCheckboxModule, ɵngcc29.ThySelectModule, ɵngcc30.ThySlideModule, ɵngcc31.ThyRadioModule, ɵngcc29.ThySelectModule, ɵngcc32.ThyPropertyOperationModule, ɵngcc33.ThyUploaderModule, ɵngcc34.ThyDateRangeModule, ɵngcc35.ThySharedModule, ɵngcc36.ThyListModule, ɵngcc37.ThyTreeSelectModule, ɵngcc38.ThyStepperModule, ɵngcc39.ThyCascaderModule, ɵngcc40.ThyDialogModule, ɵngcc41.ThyTooltipModule, ɵngcc42.ThyProgressModule, ɵngcc43.ThyBreadcrumbModule, ɵngcc44.ThyArrowSwitcherModule, ɵngcc45.ThyMarkdownModule, ɵngcc46.ThyFlexibleTextModule, ɵngcc47.ThyDragDropModule, ɵngcc35.ThySelectCommonModule, ɵngcc48.ThySkeletonModule, ɵngcc49.ThyVoteModule, ɵngcc50.ThyResultModule, ɵngcc51.ThyMentionModule, ɵngcc52.ThyDatePickerModule, ɵngcc53.ThyTimelineModule, ɵngcc54.ThyDividerModule, ɵngcc55.ThyTimePickerModule, ɵngcc56.ThyStatisticModule, ɵngcc57.ThyAutocompleteModule, ɵngcc58.ThyAnchorModule, ɵngcc59.ThyAffixModule, ɵngcc60.ThySliderModule, ɵngcc61.ThyCalendarModule, ɵngcc62.ThyFullscreenModule, ɵngcc63.ThyGuiderModule, ɵngcc64.ThyResizableModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxTethysModule, [{
        type: core$1.NgModule,
        args: [{
                declarations: [],
                imports: __spread(IMPORT_EXPORT_MODULES),
                exports: IMPORT_EXPORT_MODULES,
                providers: []
            }]
    }], function () { return []; }, null); })();
        return NgxTethysModule;
    }());
    NgxTethysModule.ctorParameters = function () { return []; };

    var VERSION = new core$1.Version('11.0.0');

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgxTethysModule = NgxTethysModule;
    exports.VERSION = VERSION;
    Object.keys(util).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return util[k];
            }
        });
    });
    Object.keys(actionMenu).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return actionMenu[k];
            }
        });
    });
    Object.keys(affix).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return affix[k];
            }
        });
    });
    Object.keys(alert).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return alert[k];
            }
        });
    });
    Object.keys(anchor).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return anchor[k];
            }
        });
    });
    Object.keys(arrowSwitcher).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return arrowSwitcher[k];
            }
        });
    });
    Object.keys(autocomplete).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return autocomplete[k];
            }
        });
    });
    Object.keys(avatar).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return avatar[k];
            }
        });
    });
    Object.keys(backTop).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return backTop[k];
            }
        });
    });
    Object.keys(badge).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return badge[k];
            }
        });
    });
    Object.keys(breadcrumb).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return breadcrumb[k];
            }
        });
    });
    Object.keys(button).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return button[k];
            }
        });
    });
    Object.keys(card).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return card[k];
            }
        });
    });
    Object.keys(cascader).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return cascader[k];
            }
        });
    });
    Object.keys(checkbox).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return checkbox[k];
            }
        });
    });
    Object.keys(shared).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return shared[k];
            }
        });
    });
    Object.keys(datePicker).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return datePicker[k];
            }
        });
    });
    Object.keys(dateRange).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return dateRange[k];
            }
        });
    });
    Object.keys(dialog).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return dialog[k];
            }
        });
    });
    Object.keys(dragDrop).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return dragDrop[k];
            }
        });
    });
    Object.keys(dropdown).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return dropdown[k];
            }
        });
    });
    Object.keys(copy).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return copy[k];
            }
        });
    });
    Object.keys(empty).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return empty[k];
            }
        });
    });
    Object.keys(flexibleText).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return flexibleText[k];
            }
        });
    });
    Object.keys(form).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return form[k];
            }
        });
    });
    Object.keys(table).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return table[k];
            }
        });
    });
    Object.keys(icon).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return icon[k];
            }
        });
    });
    Object.keys(input).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return input[k];
            }
        });
    });
    Object.keys(label).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return label[k];
            }
        });
    });
    Object.keys(layout).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return layout[k];
            }
        });
    });
    Object.keys(list).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return list[k];
            }
        });
    });
    Object.keys(loading).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return loading[k];
            }
        });
    });
    Object.keys(markdown).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return markdown[k];
            }
        });
    });
    Object.keys(mention).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return mention[k];
            }
        });
    });
    Object.keys(menu).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return menu[k];
            }
        });
    });
    Object.keys(nav).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return nav[k];
            }
        });
    });
    Object.keys(notify).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return notify[k];
            }
        });
    });
    Object.keys(pagination).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return pagination[k];
            }
        });
    });
    Object.keys(popover).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return popover[k];
            }
        });
    });
    Object.keys(progress).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return progress[k];
            }
        });
    });
    Object.keys(propertyOperation).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return propertyOperation[k];
            }
        });
    });
    Object.keys(radio).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return radio[k];
            }
        });
    });
    Object.keys(raster).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return raster[k];
            }
        });
    });
    Object.keys(result).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return result[k];
            }
        });
    });
    Object.keys(select).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return select[k];
            }
        });
    });
    Object.keys(skeleton).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return skeleton[k];
            }
        });
    });
    Object.keys(slide).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return slide[k];
            }
        });
    });
    Object.keys(statistic).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return statistic[k];
            }
        });
    });
    Object.keys(stepper).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return stepper[k];
            }
        });
    });
    Object.keys(strength).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return strength[k];
            }
        });
    });
    Object.keys(_switch).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return _switch[k];
            }
        });
    });
    Object.keys(timePicker).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return timePicker[k];
            }
        });
    });
    Object.keys(tooltip).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return tooltip[k];
            }
        });
    });
    Object.keys(transfer).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return transfer[k];
            }
        });
    });
    Object.keys(treeSelect).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return treeSelect[k];
            }
        });
    });
    Object.keys(tree).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return tree[k];
            }
        });
    });
    Object.keys(uploader).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return uploader[k];
            }
        });
    });
    Object.keys(vote).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return vote[k];
            }
        });
    });
    Object.keys(slider).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return slider[k];
            }
        });
    });
    Object.keys(calendar).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return calendar[k];
            }
        });
    });
    Object.keys(timeline).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return timeline[k];
            }
        });
    });
    Object.keys(fullscreen).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return fullscreen[k];
            }
        });
    });
    Object.keys(guider).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return guider[k];
            }
        });
    });
    Object.keys(resizable).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return resizable[k];
            }
        });
    });
    Object.keys(core).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return core[k];
            }
        });
    });
    Object.keys(positioning).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return positioning[k];
            }
        });
    });
    Object.keys(store).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return store[k];
            }
        });
    });

    Object.defineProperty(exports, '__esModule', { value: true });

})));
