// ngx-tethys/bundles/ngx-tethys-raster.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/core'),exports, require('@angular/core'), require('rxjs'), require('ngx-tethys/core'), require('ngx-tethys/util'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/raster', ['@angular/core','ngx-tethys/core','exports', '@angular/core', 'rxjs', 'ngx-tethys/core', 'ngx-tethys/util', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.core,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].raster = {}), global.ng.core, global.rxjs, global['ngx-tethys'].core, global['ngx-tethys'].util, global.rxjs.operators));
}(this, (function (ɵngcc0,ɵngcc1,exports, core$1, rxjs, core, util, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var _MixinBase$1 = core.mixinUnsubscribe(core.MixinBase);
    var ThyRowDirective = /** @class */ (function (_super) {
        __extends(ThyRowDirective, _super);
        function ThyRowDirective(renderer, elementRef) {
            var _this = _super.call(this) || this;
            _this.renderer = renderer;
            _this.elementRef = elementRef;
            _this.thyAlign = null;
            _this.thyJustify = null;
            _this.actualGutter$ = new rxjs.ReplaySubject(1);
            return _this;
        }
        ThyRowDirective.prototype.ngOnInit = function () {
            this._setGutterStyle();
        };
        ThyRowDirective.prototype.ngOnChanges = function () {
            this._setGutterStyle();
        };
        ThyRowDirective.prototype.ngAfterViewInit = function () { };
        ThyRowDirective.prototype._setGutterStyle = function () {
            var _this = this;
            var _a = __read(this._getGutter(), 2), horizontalGutter = _a[0], verticalGutter = _a[1];
            this.actualGutter$.next([horizontalGutter, verticalGutter]);
            var renderGutter = function (name, gutter) {
                var nativeElement = _this.elementRef.nativeElement;
                _this.renderer.setStyle(nativeElement, name, "-" + gutter / 2 + "px");
            };
            if (horizontalGutter > 0) {
                renderGutter('margin-left', horizontalGutter);
                renderGutter('margin-right', horizontalGutter);
            }
            if (verticalGutter > 0) {
                renderGutter('margin-top', verticalGutter);
                renderGutter('margin-bottom', verticalGutter);
            }
        };
        ThyRowDirective.prototype._getGutter = function () {
            if (util.isString(this.thyGutter)) {
                throw Error("thyGutter value can not be string type");
            }
            return [this.thyGutter, 0];
        };
        ThyRowDirective.prototype.ngOnDestroy = function () {
            _super.prototype.ngOnDestroy.call(this);
        };
ThyRowDirective.ɵfac = function ThyRowDirective_Factory(t) { return new (t || ThyRowDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyRowDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyRowDirective, selectors: [["", "thyRow", ""]], hostAttrs: [1, "thy-row"], inputs: { thyGutter: "thyGutter", thyAlign: "thyAlign", thyJustify: "thyJustify" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRowDirective, [{
        type: core$1.Directive,
        args: [{
                selector: '[thyRow]',
                host: {
                    class: 'thy-row'
                }
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { thyGutter: [{
            type: core$1.Input
        }], thyAlign: [{
            type: core$1.Input
        }], thyJustify: [{
            type: core$1.Input
        }] }); })();
        return ThyRowDirective;
    }(_MixinBase$1));
    ThyRowDirective.ctorParameters = function () { return [
        { type: core$1.Renderer2 },
        { type: core$1.ElementRef }
    ]; };
    ThyRowDirective.propDecorators = {
        thyGutter: [{ type: core$1.Input }],
        thyAlign: [{ type: core$1.Input }],
        thyJustify: [{ type: core$1.Input }]
    };

    var _MixinBase = core.mixinUnsubscribe(core.MixinBase);
    var ThyColDirective = /** @class */ (function (_super) {
        __extends(ThyColDirective, _super);
        function ThyColDirective(elementRef, renderer, updateHostClassService, thyRowDirective) {
            var _this = _super.call(this) || this;
            _this.elementRef = elementRef;
            _this.renderer = renderer;
            _this.updateHostClassService = updateHostClassService;
            _this.thyRowDirective = thyRowDirective;
            _this.thyFlex = null;
            _this.thySpan = 24;
            _this.thyOrder = null;
            _this.thyOffset = null;
            _this.thyPush = null;
            _this.thyPull = null;
            _this.thyXs = null;
            _this.thySm = null;
            _this.thyMd = null;
            _this.thyLg = null;
            _this.thyXl = null;
            _this.thyXXl = null;
            _this.updateHostClassService.initializeElement(_this.elementRef);
            return _this;
        }
        ThyColDirective.prototype.ngOnInit = function () {
            this._setHostClassMap();
        };
        ThyColDirective.prototype.ngOnChanges = function (changes) {
            this._setHostClassMap();
        };
        ThyColDirective.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.thyRowDirective) {
                this.thyRowDirective.actualGutter$.pipe(operators.takeUntil(this.ngUnsubscribe$)).subscribe(function (_a) {
                    var _b = __read(_a, 2), horizontalGutter = _b[0], verticalGutter = _b[1];
                    var renderGutter = function (name, gutter) {
                        var nativeElement = _this.elementRef.nativeElement;
                        _this.renderer.setStyle(nativeElement, name, gutter / 2 + "px");
                    };
                    if (horizontalGutter > 0) {
                        renderGutter('padding-left', horizontalGutter);
                        renderGutter('padding-right', horizontalGutter);
                    }
                    if (verticalGutter > 0) {
                        renderGutter('padding-top', verticalGutter);
                        renderGutter('padding-bottom', verticalGutter);
                    }
                });
            }
        };
        ThyColDirective.prototype._setHostClassMap = function () {
            var _a;
            this.updateHostClassService.updateClassByMap((_a = {},
                _a["thy-col-" + this.thySpan] = true,
                _a));
        };
        ThyColDirective.prototype.ngOnDestroy = function () {
            _super.prototype.ngOnDestroy.call(this);
        };
ThyColDirective.ɵfac = function ThyColDirective_Factory(t) { return new (t || ThyColDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ThyRowDirective, 9)); };
ThyColDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyColDirective, selectors: [["", "thyCol", ""]], hostAttrs: [1, "thy-col"], inputs: { thyFlex: "thyFlex", thySpan: "thySpan", thyOrder: "thyOrder", thyOffset: "thyOffset", thyPush: "thyPush", thyPull: "thyPull", thyXs: "thyXs", thySm: "thySm", thyMd: "thyMd", thyLg: "thyLg", thyXl: "thyXl", thyXXl: "thyXXl" }, features: [ɵngcc0.ɵɵProvidersFeature([core.UpdateHostClassService]), ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyColDirective, [{
        type: core$1.Directive,
        args: [{
                selector: '[thyCol]',
                providers: [core.UpdateHostClassService],
                host: {
                    class: 'thy-col'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc1.UpdateHostClassService }, { type: ThyRowDirective, decorators: [{
                type: core$1.Optional
            }, {
                type: core$1.Host
            }] }]; }, { thyFlex: [{
            type: core$1.Input
        }], thySpan: [{
            type: core$1.Input
        }], thyOrder: [{
            type: core$1.Input
        }], thyOffset: [{
            type: core$1.Input
        }], thyPush: [{
            type: core$1.Input
        }], thyPull: [{
            type: core$1.Input
        }], thyXs: [{
            type: core$1.Input
        }], thySm: [{
            type: core$1.Input
        }], thyMd: [{
            type: core$1.Input
        }], thyLg: [{
            type: core$1.Input
        }], thyXl: [{
            type: core$1.Input
        }], thyXXl: [{
            type: core$1.Input
        }] }); })();
        return ThyColDirective;
    }(_MixinBase));
    ThyColDirective.ctorParameters = function () { return [
        { type: core$1.ElementRef },
        { type: core$1.Renderer2 },
        { type: core.UpdateHostClassService },
        { type: ThyRowDirective, decorators: [{ type: core$1.Optional }, { type: core$1.Host }] }
    ]; };
    ThyColDirective.propDecorators = {
        thyFlex: [{ type: core$1.Input }],
        thySpan: [{ type: core$1.Input }],
        thyOrder: [{ type: core$1.Input }],
        thyOffset: [{ type: core$1.Input }],
        thyPush: [{ type: core$1.Input }],
        thyPull: [{ type: core$1.Input }],
        thyXs: [{ type: core$1.Input }],
        thySm: [{ type: core$1.Input }],
        thyMd: [{ type: core$1.Input }],
        thyLg: [{ type: core$1.Input }],
        thyXl: [{ type: core$1.Input }],
        thyXXl: [{ type: core$1.Input }]
    };

    var ThyRasterModule = /** @class */ (function () {
        function ThyRasterModule() {
        }
ThyRasterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyRasterModule });
ThyRasterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyRasterModule_Factory(t) { return new (t || ThyRasterModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyRasterModule, { declarations: [ThyRowDirective, ThyColDirective], exports: [ThyRowDirective, ThyColDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRasterModule, [{
        type: core$1.NgModule,
        args: [{
                declarations: [ThyRowDirective, ThyColDirective],
                exports: [ThyRowDirective, ThyColDirective],
                imports: []
            }]
    }], function () { return []; }, null); })();
        return ThyRasterModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyColDirective = ThyColDirective;
    exports.ThyRasterModule = ThyRasterModule;
    exports.ThyRowDirective = ThyRowDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
