// ngx-tethys/bundles/ngx-tethys-skeleton.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/cdk/portal'),exports, require('@angular/core'), require('@angular/common'), require('@angular/cdk/portal'), require('ngx-tethys/util')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/skeleton', ['@angular/core','@angular/common','@angular/cdk/portal','exports', '@angular/core', '@angular/common', '@angular/cdk/portal', 'ngx-tethys/util'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ng.cdk.portal,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].skeleton = {}), global.ng.core, global.ng.common, global.ng.cdk.portal, global['ngx-tethys'].util));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, portal, util) { 
var _c0 = ["content"];
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1__svg_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var content_r8 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", content_r8);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1_Template, 2, 1, "ng-container", 7);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.contentTemplates);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_5__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_5__svg_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r3.customTemplate);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r4.animationValues[0])("dur", ctx_r4.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r5.animationValues[1])("dur", ctx_r5.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r6.animationValues[2])("dur", ctx_r6.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 2);
    ɵngcc0.ɵɵelement(1, "rect", 3);
    ɵngcc0.ɵɵelementStart(2, "defs");
    ɵngcc0.ɵɵelementStart(3, "clipPath");
    ɵngcc0.ɵɵtemplate(4, ThySkeletonComponent__svg_svg_0__svg_ng_container_4_Template, 2, 1, "ng-container", 1);
    ɵngcc0.ɵɵtemplate(5, ThySkeletonComponent__svg_svg_0__svg_ng_container_5_Template, 2, 1, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "linearGradient");
    ɵngcc0.ɵɵelementStart(7, "stop", 4);
    ɵngcc0.ɵɵtemplate(8, ThySkeletonComponent__svg_svg_0__svg_ng_container_8_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "stop", 5);
    ɵngcc0.ɵɵtemplate(10, ThySkeletonComponent__svg_svg_0__svg_ng_container_10_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "stop", 6);
    ɵngcc0.ɵɵtemplate(12, ThySkeletonComponent__svg_svg_0__svg_ng_container_12_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattributeInterpolate2("viewBox", "0 0 ", ctx_r0.thyViewBoxWidth, " ", ctx_r0.thyViewBoxHeight, "");
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r0.thyStyle);
    ɵngcc0.ɵɵattribute("preserveAspectRatio", ctx_r0.thyPreserveAspectRatio)("height", ctx_r0.thyHeight)("width", ctx_r0.thyWidth);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r0.fillStyle);
    ɵngcc0.ɵɵattribute("clip-path", ctx_r0.clipPath)("width", ctx_r0.thyWidth)("height", ctx_r0.thyHeight);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("id", ctx_r0.idClip);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.contentTemplates);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("id", ctx_r0.idGradient);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thyPrimaryColor)("stop-opacity", ctx_r0.thyPrimaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thySecondaryColor)("stop-opacity", ctx_r0.thySecondaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thyPrimaryColor)("stop-opacity", ctx_r0.thyPrimaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
} }
function ThySkeletonComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementContainerEnd();
} }
var _c1 = ["*"];
function ThySkeletonListComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
    ɵngcc0.ɵɵelement(1, "rect", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelement(3, "rect", 4);
    ɵngcc0.ɵɵelement(4, "rect", 5);
    ɵngcc0.ɵɵelement(5, "rect", 6);
} }
function ThySkeletonBulletListComponent_ng_template_0_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(1, "circle", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("cx", item_r3.circle.cx)("cy", item_r3.circle.cy);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("x", item_r3.rect.x)("y", item_r3.rect.y);
} }
function ThySkeletonBulletListComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThySkeletonBulletListComponent_ng_template_0_ng_container_0_Template, 3, 4, "ng-container", 1);
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.items);
} }
function ThySkeletonTitleComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("width", ctx_r1.width)("height", ctx_r1.height);
} }
function ThySkeletonAvatarComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "circle", 1);
} }
function ThySkeletonParagraphComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
    ɵngcc0.ɵɵelement(1, "rect", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelement(3, "rect", 4);
} }
'use strict';

    var ThySkeletonComponent = /** @class */ (function () {
        function ThySkeletonComponent(platformId) {
            this.platformId = platformId;
            this.addSkeletonClass = true;
            this.thyAnimate = true;
            this.thyBaseUrl = '';
            this.thyWidth = '100%';
            this.thyHeight = '100%';
            this.thyViewBoxWidth = 400;
            this.thyViewBoxHeight = 130;
            this.thySpeed = 2;
            this.thyPreserveAspectRatio = 'none'; // xMidYMid meet
            this.thyPrimaryColor = '#f0f0f0';
            this.thySecondaryColor = '#e0e0e0';
            this.thyPrimaryOpacity = 1;
            this.thySecondaryOpacity = 1;
            this.thyIgnoreBaseUrl = false;
            this.thyLoadingDone = false;
            this.contentTemplates = [];
            this.idClip = util.generateRandomStr();
            this.idGradient = util.generateRandomStr();
            this.defaultAnimation = ['-3; 1', '-2; 2', '-1; 3'];
            this.rtlAnimation = ['1; -3', '2; -2', '3; -1'];
        }
        ThySkeletonComponent.prototype.ngOnInit = function () {
            this.animationValues = this.thyRtl ? this.rtlAnimation : this.defaultAnimation;
            if (this.thyBaseUrl === '' && !this.thyIgnoreBaseUrl && common.isPlatformBrowser(this.platformId)) {
                this.thyBaseUrl = window.location.pathname;
            }
            this.setFillStyle();
            this.setClipPath();
        };
        ThySkeletonComponent.prototype.ngOnChanges = function (changes) {
            if (changes['thyBaseUrl']) {
                if (changes['thyBaseUrl'].previousValue !== changes['thyBaseUrl'].currentValue) {
                    this.setFillStyle();
                    this.setClipPath();
                }
            }
        };
        ThySkeletonComponent.prototype.setFillStyle = function () {
            this.fillStyle = {
                fill: "url(" + this.thyBaseUrl + "#" + this.idGradient + ")"
            };
        };
        ThySkeletonComponent.prototype.setClipPath = function () {
            this.clipPath = "url(" + this.thyBaseUrl + "#" + this.idClip + ")";
        };
        ThySkeletonComponent.prototype.assignInputProperties = function (input) {
            this.thyAnimate = input.thyAnimate;
            this.thyBaseUrl = input.thyBaseUrl;
            this.thyHeight = input.thyHeight;
            this.thyWidth = input.thyWidth;
            this.thyViewBoxHeight = input.thyViewBoxHeight;
            this.thyViewBoxWidth = input.thyViewBoxWidth;
            this.thyPreserveAspectRatio = input.thyPreserveAspectRatio;
            this.thyPrimaryColor = input.thyPrimaryColor;
            this.thyPrimaryOpacity = input.thyPrimaryOpacity;
            this.thySecondaryColor = input.thySecondaryColor;
            this.thySecondaryOpacity = input.thySecondaryOpacity;
            this.thyRtl = this.thyRtl;
            this.thySpeed = this.thySpeed;
            this.thyUniqueKey = this.thyUniqueKey;
        };
        ThySkeletonComponent.prototype.addTemplate = function (template) {
            if (template) {
                this.contentTemplates.push(template);
            }
        };
ThySkeletonComponent.ɵfac = function ThySkeletonComponent_Factory(t) { return new (t || ThySkeletonComponent)(ɵngcc0.ɵɵdirectiveInject(core.PLATFORM_ID)); };
ThySkeletonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonComponent, selectors: [["thy-skeleton"]], contentQueries: function ThySkeletonComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customTemplate = _t.first);
    } }, hostVars: 2, hostBindings: function ThySkeletonComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-skeleton", ctx.addSkeletonClass);
    } }, inputs: { thyAnimate: "thyAnimate", thyBaseUrl: "thyBaseUrl", thyWidth: "thyWidth", thyHeight: "thyHeight", thyViewBoxWidth: "thyViewBoxWidth", thyViewBoxHeight: "thyViewBoxHeight", thySpeed: "thySpeed", thyPreserveAspectRatio: "thyPreserveAspectRatio", thyPrimaryColor: "thyPrimaryColor", thySecondaryColor: "thySecondaryColor", thyPrimaryOpacity: "thyPrimaryOpacity", thySecondaryOpacity: "thySecondaryOpacity", thyIgnoreBaseUrl: "thyIgnoreBaseUrl", thyLoadingDone: "thyLoadingDone", thyRtl: "thyRtl", thyUniqueKey: "thyUniqueKey", thyStyle: "thyStyle" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c1, decls: 2, vars: 2, consts: [[3, "ngStyle", 4, "ngIf"], [4, "ngIf"], [3, "ngStyle"], ["x", "0", "y", "0", 3, "ngStyle"], ["offset", "0%"], ["offset", "50%"], ["offset", "100%"], [4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["attributeName", "offset", "repeatCount", "indefinite"]], template: function ThySkeletonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThySkeletonComponent__svg_svg_0_Template, 13, 23, "svg", 0);
        ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent_ng_container_1_Template, 2, 0, "ng-container", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.thyLoadingDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyLoadingDone);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-skeleton',
                template: "<svg\n  *ngIf=\"!thyLoadingDone\"\n  attr.viewBox=\"0 0 {{ thyViewBoxWidth }} {{ thyViewBoxHeight }}\"\n  [attr.preserveAspectRatio]=\"thyPreserveAspectRatio\"\n  [ngStyle]=\"thyStyle\"\n  [attr.height]=\"thyHeight\"\n  [attr.width]=\"thyWidth\"\n>\n  <rect\n    [attr.clip-path]=\"clipPath\"\n    [ngStyle]=\"fillStyle\"\n    x=\"0\"\n    y=\"0\"\n    [attr.width]=\"thyWidth\"\n    [attr.height]=\"thyHeight\"\n  />\n\n  <defs>\n    <clipPath [attr.id]=\"idClip\">\n      <ng-container *ngIf=\"contentTemplates\">\n        <ng-container *ngFor=\"let content of contentTemplates\">\n          <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n        </ng-container>\n      </ng-container>\n\n      <ng-container *ngIf=\"customTemplate\">\n        <ng-container *ngTemplateOutlet=\"customTemplate\"></ng-container>\n      </ng-container>\n    </clipPath>\n\n    <linearGradient [attr.id]=\"idGradient\">\n      <stop offset=\"0%\" [attr.stop-color]=\"thyPrimaryColor\" [attr.stop-opacity]=\"thyPrimaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[0]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n\n      <stop offset=\"50%\" [attr.stop-color]=\"thySecondaryColor\" [attr.stop-opacity]=\"thySecondaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[1]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n\n      <stop offset=\"100%\" [attr.stop-color]=\"thyPrimaryColor\" [attr.stop-opacity]=\"thyPrimaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[2]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n    </linearGradient>\n  </defs>\n</svg>\n<ng-container *ngIf=\"thyLoadingDone\">\n  <ng-content></ng-content>\n</ng-container>\n"
            }]
    }], function () { return [{ type: String, decorators: [{
                type: core.Inject,
                args: [core.PLATFORM_ID]
            }] }]; }, { addSkeletonClass: [{
            type: core.HostBinding,
            args: ['class.thy-skeleton']
        }], thyAnimate: [{
            type: core.Input
        }], thyBaseUrl: [{
            type: core.Input
        }], thyWidth: [{
            type: core.Input
        }], thyHeight: [{
            type: core.Input
        }], thyViewBoxWidth: [{
            type: core.Input
        }], thyViewBoxHeight: [{
            type: core.Input
        }], thySpeed: [{
            type: core.Input
        }], thyPreserveAspectRatio: [{
            type: core.Input
        }], thyPrimaryColor: [{
            type: core.Input
        }], thySecondaryColor: [{
            type: core.Input
        }], thyPrimaryOpacity: [{
            type: core.Input
        }], thySecondaryOpacity: [{
            type: core.Input
        }], thyIgnoreBaseUrl: [{
            type: core.Input
        }], thyLoadingDone: [{
            type: core.Input
        }], thyRtl: [{
            type: core.Input
        }], thyUniqueKey: [{
            type: core.Input
        }], thyStyle: [{
            type: core.Input
        }], customTemplate: [{
            type: core.ContentChild,
            args: ['content']
        }] }); })();
        return ThySkeletonComponent;
    }());
    ThySkeletonComponent.ctorParameters = function () { return [
        { type: String, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] }
    ]; };
    ThySkeletonComponent.propDecorators = {
        addSkeletonClass: [{ type: core.HostBinding, args: ['class.thy-skeleton',] }],
        thyAnimate: [{ type: core.Input }],
        thyBaseUrl: [{ type: core.Input }],
        thyWidth: [{ type: core.Input }],
        thyHeight: [{ type: core.Input }],
        thyViewBoxWidth: [{ type: core.Input }],
        thyViewBoxHeight: [{ type: core.Input }],
        thySpeed: [{ type: core.Input }],
        thyPreserveAspectRatio: [{ type: core.Input }],
        thyPrimaryColor: [{ type: core.Input }],
        thySecondaryColor: [{ type: core.Input }],
        thyPrimaryOpacity: [{ type: core.Input }],
        thySecondaryOpacity: [{ type: core.Input }],
        thyUniqueKey: [{ type: core.Input }],
        thyRtl: [{ type: core.Input }],
        thyStyle: [{ type: core.Input }],
        thyIgnoreBaseUrl: [{ type: core.Input }],
        thyLoadingDone: [{ type: core.Input }],
        customTemplate: [{ type: core.ContentChild, args: ['content',] }]
    };

    var ThySkeletonListComponent = /** @class */ (function () {
        function ThySkeletonListComponent(skeletonComponent) {
            this.skeletonComponent = skeletonComponent;
        }
        ThySkeletonListComponent.prototype.ngOnInit = function () {
            this.skeletonComponent.addTemplate(this.contentTemplateRef);
        };
ThySkeletonListComponent.ɵfac = function ThySkeletonListComponent_Factory(t) { return new (t || ThySkeletonListComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonListComponent, selectors: [["thy-skeleton-list-template"]], viewQuery: function ThySkeletonListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, decls: 2, vars: 0, consts: [["content", ""], ["x", "0", "y", "0", "rx", "3", "ry", "3", "width", "250", "height", "10"], ["x", "20", "y", "20", "rx", "3", "ry", "3", "width", "220", "height", "10"], ["x", "20", "y", "40", "rx", "3", "ry", "3", "width", "170", "height", "10"], ["x", "0", "y", "60", "rx", "3", "ry", "3", "width", "250", "height", "10"], ["x", "20", "y", "80", "rx", "3", "ry", "3", "width", "200", "height", "10"], ["x", "20", "y", "100", "rx", "3", "ry", "3", "width", "80", "height", "10"]], template: function ThySkeletonListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonListComponent_ng_template_0_Template, 6, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonListComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-skeleton-list-template',
                template: "\n        <ng-template #content>\n            <svg:rect x=\"0\" y=\"0\" rx=\"3\" ry=\"3\" width=\"250\" height=\"10\" />\n            <svg:rect x=\"20\" y=\"20\" rx=\"3\" ry=\"3\" width=\"220\" height=\"10\" />\n            <svg:rect x=\"20\" y=\"40\" rx=\"3\" ry=\"3\" width=\"170\" height=\"10\" />\n            <svg:rect x=\"0\" y=\"60\" rx=\"3\" ry=\"3\" width=\"250\" height=\"10\" />\n            <svg:rect x=\"20\" y=\"80\" rx=\"3\" ry=\"3\" width=\"200\" height=\"10\" />\n            <svg:rect x=\"20\" y=\"100\" rx=\"3\" ry=\"3\" width=\"80\" height=\"10\" />\n        </ng-template>\n    "
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: core.ViewChild,
            args: ['content', { static: true }]
        }] }); })();
        return ThySkeletonListComponent;
    }());
    ThySkeletonListComponent.ctorParameters = function () { return [
        { type: ThySkeletonComponent }
    ]; };
    ThySkeletonListComponent.propDecorators = {
        contentTemplateRef: [{ type: core.ViewChild, args: ['content', { static: true },] }]
    };

    var ThySkeletonBulletListComponent = /** @class */ (function () {
        function ThySkeletonBulletListComponent(skeletonComponent) {
            this.skeletonComponent = skeletonComponent;
            this.thyCount = 5;
            this.items = [];
        }
        ThySkeletonBulletListComponent.prototype.ngOnInit = function () {
            for (var i = 0; i <= this.thyCount; i++) {
                this.items.push({
                    circle: {
                        cx: 10,
                        cy: i * 30 + 20,
                        r: 8
                    },
                    rect: {
                        x: 25,
                        y: i * 2 * 15 + 15,
                        rx: 5,
                        ry: 5,
                        width: '90%',
                        height: 10
                    }
                });
            }
            this.skeletonComponent.addTemplate(this.contentTemplateRef);
        };
ThySkeletonBulletListComponent.ɵfac = function ThySkeletonBulletListComponent_Factory(t) { return new (t || ThySkeletonBulletListComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonBulletListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonBulletListComponent, selectors: [["thy-skeleton-bullet-list-template"]], viewQuery: function ThySkeletonBulletListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, inputs: { thyCount: "thyCount" }, decls: 2, vars: 0, consts: [["content", ""], [4, "ngFor", "ngForOf"], ["r", "8"], ["rx", "5", "ry", "5", "width", "90%", "height", "10"]], template: function ThySkeletonBulletListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonBulletListComponent_ng_template_0_Template, 1, 1, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc1.NgForOf], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonBulletListComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-skeleton-bullet-list-template',
                template: "\n        <ng-template #content>\n            <ng-container *ngFor=\"let item of items\">\n                <svg:circle [attr.cx]=\"item.circle.cx\" [attr.cy]=\"item.circle.cy\" r=\"8\" />\n                <svg:rect [attr.x]=\"item.rect.x\" [attr.y]=\"item.rect.y\" rx=\"5\" ry=\"5\" width=\"90%\" height=\"10\" />\n            </ng-container>\n        </ng-template>\n    "
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { thyCount: [{
            type: core.Input
        }], contentTemplateRef: [{
            type: core.ViewChild,
            args: ['content', { static: true }]
        }] }); })();
        return ThySkeletonBulletListComponent;
    }());
    ThySkeletonBulletListComponent.ctorParameters = function () { return [
        { type: ThySkeletonComponent }
    ]; };
    ThySkeletonBulletListComponent.propDecorators = {
        thyCount: [{ type: core.Input }],
        contentTemplateRef: [{ type: core.ViewChild, args: ['content', { static: true },] }]
    };

    var ThySkeletonTitleComponent = /** @class */ (function () {
        function ThySkeletonTitleComponent(skeletonComponent) {
            this.skeletonComponent = skeletonComponent;
            this.width = 100;
            this.height = 10;
        }
        Object.defineProperty(ThySkeletonTitleComponent.prototype, "thyWidth", {
            set: function (value) {
                if (value) {
                    this.width = value;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThySkeletonTitleComponent.prototype, "thyHeight", {
            set: function (value) {
                if (value) {
                    this.height = value;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThySkeletonTitleComponent.prototype.ngOnInit = function () {
            this.skeletonComponent.addTemplate(this.contentTemplateRef);
        };
ThySkeletonTitleComponent.ɵfac = function ThySkeletonTitleComponent_Factory(t) { return new (t || ThySkeletonTitleComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonTitleComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonTitleComponent, selectors: [["thy-skeleton-title-template"], ["", "thySkeletonTitleTemplate", ""]], viewQuery: function ThySkeletonTitleComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, inputs: { thyWidth: "thyWidth", thyHeight: "thyHeight" }, exportAs: ["thySkeletonTitleTemplate"], decls: 2, vars: 0, consts: [["content", ""], ["x", "15", "y", "13", "rx", "4", "ry", "4"]], template: function ThySkeletonTitleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonTitleComponent_ng_template_0_Template, 1, 2, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonTitleComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-skeleton-title-template,[thySkeletonTitleTemplate]',
                template: "\n        <ng-template #content>\n            <svg:rect x=\"15\" y=\"13\" rx=\"4\" ry=\"4\" [attr.width]=\"width\" [attr.height]=\"height\" />\n        </ng-template>\n    ",
                exportAs: 'thySkeletonTitleTemplate'
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { thyWidth: [{
            type: core.Input
        }], thyHeight: [{
            type: core.Input
        }], contentTemplateRef: [{
            type: core.ViewChild,
            args: ['content', { static: true }]
        }] }); })();
        return ThySkeletonTitleComponent;
    }());
    ThySkeletonTitleComponent.ctorParameters = function () { return [
        { type: ThySkeletonComponent }
    ]; };
    ThySkeletonTitleComponent.propDecorators = {
        contentTemplateRef: [{ type: core.ViewChild, args: ['content', { static: true },] }],
        thyWidth: [{ type: core.Input }],
        thyHeight: [{ type: core.Input }]
    };

    var ThySkeletonAvatarComponent = /** @class */ (function () {
        function ThySkeletonAvatarComponent(skeletonComponent) {
            this.skeletonComponent = skeletonComponent;
        }
        ThySkeletonAvatarComponent.prototype.ngOnInit = function () {
            this.skeletonComponent.addTemplate(this.contentTemplateRef);
        };
ThySkeletonAvatarComponent.ɵfac = function ThySkeletonAvatarComponent_Factory(t) { return new (t || ThySkeletonAvatarComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonAvatarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonAvatarComponent, selectors: [["thy-skeleton-avatar-template"], ["", "thySkeletonAvatarTemplate", ""]], viewQuery: function ThySkeletonAvatarComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, decls: 2, vars: 0, consts: [["content", ""], ["cx", "30", "cy", "30", "r", "30"]], template: function ThySkeletonAvatarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonAvatarComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonAvatarComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-skeleton-avatar-template,[thySkeletonAvatarTemplate]',
                template: "\n        <ng-template #content>\n            <svg:circle cx=\"30\" cy=\"30\" r=\"30\" />\n        </ng-template>\n    "
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: core.ViewChild,
            args: ['content', { static: true }]
        }] }); })();
        return ThySkeletonAvatarComponent;
    }());
    ThySkeletonAvatarComponent.ctorParameters = function () { return [
        { type: ThySkeletonComponent }
    ]; };
    ThySkeletonAvatarComponent.propDecorators = {
        contentTemplateRef: [{ type: core.ViewChild, args: ['content', { static: true },] }]
    };

    var ThySkeletonParagraphComponent = /** @class */ (function () {
        function ThySkeletonParagraphComponent(skeletonComponent) {
            this.skeletonComponent = skeletonComponent;
        }
        ThySkeletonParagraphComponent.prototype.ngOnInit = function () {
            this.skeletonComponent.addTemplate(this.contentTemplateRef);
        };
ThySkeletonParagraphComponent.ɵfac = function ThySkeletonParagraphComponent_Factory(t) { return new (t || ThySkeletonParagraphComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonParagraphComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonParagraphComponent, selectors: [["thy-skeleton-paragraph-template"], ["", "thySkeletonParagraphTemplate", ""]], viewQuery: function ThySkeletonParagraphComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, exportAs: ["thySkeletonParagraphTemplate"], decls: 2, vars: 0, consts: [["content", ""], ["x", "0", "y", "0", "rx", "2", "ry", "2", "width", "38%", "height", "10"], ["x", "20", "y", "20", "rx", "2", "ry", "2", "width", "92%", "height", "10"], ["x", "20", "y", "40", "rx", "2", "ry", "2", "width", "92%", "height", "10"], ["x", "0", "y", "60", "rx", "2", "ry", "2", "width", "61%", "height", "10"]], template: function ThySkeletonParagraphComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonParagraphComponent_ng_template_0_Template, 4, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonParagraphComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-skeleton-paragraph-template,[thySkeletonParagraphTemplate]',
                template: "\n        <ng-template #content>\n            <svg:rect x=\"0\" y=\"0\" rx=\"2\" ry=\"2\" width=\"38%\" height=\"10\" />\n            <svg:rect x=\"20\" y=\"20\" rx=\"2\" ry=\"2\" width=\"92%\" height=\"10\" />\n            <svg:rect x=\"20\" y=\"40\" rx=\"2\" ry=\"2\" width=\"92%\" height=\"10\" />\n            <svg:rect x=\"0\" y=\"60\" rx=\"2\" ry=\"2\" width=\"61%\" height=\"10\" />\n        </ng-template>\n    ",
                exportAs: 'thySkeletonParagraphTemplate'
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: core.ViewChild,
            args: ['content', { static: true }]
        }] }); })();
        return ThySkeletonParagraphComponent;
    }());
    ThySkeletonParagraphComponent.ctorParameters = function () { return [
        { type: ThySkeletonComponent }
    ]; };
    ThySkeletonParagraphComponent.propDecorators = {
        contentTemplateRef: [{ type: core.ViewChild, args: ['content', { static: true },] }]
    };

    var ThySkeletonModule = /** @class */ (function () {
        function ThySkeletonModule() {
        }
ThySkeletonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySkeletonModule });
ThySkeletonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySkeletonModule_Factory(t) { return new (t || ThySkeletonModule)(); }, providers: [], imports: [[common.CommonModule, portal.PortalModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySkeletonModule, { declarations: [ThySkeletonComponent, ThySkeletonListComponent, ThySkeletonBulletListComponent, ThySkeletonAvatarComponent, ThySkeletonTitleComponent, ThySkeletonParagraphComponent], imports: [ɵngcc1.CommonModule, ɵngcc2.PortalModule], exports: [ThySkeletonComponent, ThySkeletonListComponent, ThySkeletonBulletListComponent, ThySkeletonTitleComponent, ThySkeletonAvatarComponent, ThySkeletonParagraphComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    ThySkeletonComponent,
                    ThySkeletonListComponent,
                    ThySkeletonBulletListComponent,
                    ThySkeletonAvatarComponent,
                    ThySkeletonTitleComponent,
                    ThySkeletonParagraphComponent
                ],
                imports: [common.CommonModule, portal.PortalModule],
                entryComponents: [ThySkeletonComponent],
                exports: [
                    ThySkeletonComponent,
                    ThySkeletonListComponent,
                    ThySkeletonBulletListComponent,
                    ThySkeletonTitleComponent,
                    ThySkeletonAvatarComponent,
                    ThySkeletonParagraphComponent
                ],
                providers: []
            }]
    }], function () { return []; }, null); })();
        return ThySkeletonModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThySkeletonAvatarComponent = ThySkeletonAvatarComponent;
    exports.ThySkeletonBulletListComponent = ThySkeletonBulletListComponent;
    exports.ThySkeletonComponent = ThySkeletonComponent;
    exports.ThySkeletonListComponent = ThySkeletonListComponent;
    exports.ThySkeletonModule = ThySkeletonModule;
    exports.ThySkeletonParagraphComponent = ThySkeletonParagraphComponent;
    exports.ThySkeletonTitleComponent = ThySkeletonTitleComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
