// ngx-tethys/bundles/ngx-tethys-arrow-switcher.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-tethys/button'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/button'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/arrow-switcher', ['@angular/core','ngx-tethys/button','@angular/common','exports', '@angular/core', '@angular/common', 'ngx-tethys/button', '@angular/forms'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxTethys.button,global.ng.common,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys']['arrow-switcher'] = {}), global.ng.core, global.ng.common, global['ngx-tethys'].button, global.ng.forms));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, button, forms) { 'use strict';

    var ThyArrowSwitcherComponent = /** @class */ (function () {
        function ThyArrowSwitcherComponent(cd) {
            this.cd = cd;
            this._isArrowSwitcher = true;
            this._isSmallSize = false;
            this.thyPrevious = new core.EventEmitter();
            this.thyNext = new core.EventEmitter();
            this.index = 0;
            this.disabled = false;
            this.previousDisabled = false;
            this.nextDisabled = false;
        }
        Object.defineProperty(ThyArrowSwitcherComponent.prototype, "thyTotal", {
            set: function (value) {
                if (value) {
                    this.total = value;
                    this.getDisabled();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyArrowSwitcherComponent.prototype, "thySize", {
            set: function (size) {
                if (size === 'sm') {
                    this._isSmallSize = true;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyArrowSwitcherComponent.prototype.ngOnInit = function () { };
        ThyArrowSwitcherComponent.prototype.writeValue = function (value) {
            if (value >= 0) {
                this.index = value;
                this.getDisabled();
            }
            this.cd.markForCheck();
        };
        ThyArrowSwitcherComponent.prototype.registerOnChange = function (fn) {
            this.onModelChange = fn;
        };
        ThyArrowSwitcherComponent.prototype.registerOnTouched = function (fn) {
            this.onModelTouched = fn;
        };
        ThyArrowSwitcherComponent.prototype.setDisabledState = function (isDisable) {
            this.disabled = isDisable;
        };
        ThyArrowSwitcherComponent.prototype.getDisabled = function () {
            this.previousDisabled = this.index <= 0 || this.disabled;
            this.nextDisabled = this.index >= this.total - 1 || this.disabled;
        };
        ThyArrowSwitcherComponent.prototype.onPreviousClick = function (event) {
            this.index--;
            this.onModelChange(this.index);
            this.getDisabled();
            this.thyPrevious.emit({ index: this.index, event: event });
        };
        ThyArrowSwitcherComponent.prototype.onNextClick = function (event) {
            this.index++;
            this.onModelChange(this.index);
            this.getDisabled();
            this.thyNext.emit({ index: this.index, event: event });
        };
ThyArrowSwitcherComponent.ɵfac = function ThyArrowSwitcherComponent_Factory(t) { return new (t || ThyArrowSwitcherComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyArrowSwitcherComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyArrowSwitcherComponent, selectors: [["thy-arrow-switcher"]], hostVars: 4, hostBindings: function ThyArrowSwitcherComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-arrow-switcher", ctx._isArrowSwitcher)("thy-arrow-switcher-small", ctx._isSmallSize);
    } }, inputs: { thyTotal: "thyTotal", thySize: "thySize" }, outputs: { thyPrevious: "thyPrevious", thyNext: "thyNext" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: forms.NG_VALUE_ACCESSOR,
                useExisting: core.forwardRef(function () { return ThyArrowSwitcherComponent; }),
                multi: true
            }
        ])], decls: 5, vars: 4, consts: [[1, "switcher-wrapper"], ["thyButtonIcon", "angle-left", 3, "disabled", "click"], ["thyButtonIcon", "angle-right", 3, "disabled", "click"]], template: function ThyArrowSwitcherComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function ThyArrowSwitcherComponent_Template_button_click_1_listener($event) { return ctx.onPreviousClick($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "span");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "button", 2);
        ɵngcc0.ɵɵlistener("click", function ThyArrowSwitcherComponent_Template_button_click_4_listener($event) { return ctx.onNextClick($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", ctx.previousDisabled);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate2("", ctx.index + 1, " / ", ctx.total, "");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", ctx.nextDisabled);
    } }, directives: [ɵngcc1.ThyButtonIconComponent], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyArrowSwitcherComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-arrow-switcher',
                template: "<div class=\"switcher-wrapper\">\n  <button thyButtonIcon=\"angle-left\" [disabled]=\"previousDisabled\" (click)=\"onPreviousClick($event)\"></button>\n  <span>{{ index + 1 }} / {{ total }}</span>\n  <button thyButtonIcon=\"angle-right\" [disabled]=\"nextDisabled\" (click)=\"onNextClick($event)\"></button>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core.forwardRef(function () { return ThyArrowSwitcherComponent; }),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { _isArrowSwitcher: [{
            type: core.HostBinding,
            args: ['class.thy-arrow-switcher']
        }], _isSmallSize: [{
            type: core.HostBinding,
            args: ['class.thy-arrow-switcher-small']
        }], thyPrevious: [{
            type: core.Output
        }], thyNext: [{
            type: core.Output
        }], thyTotal: [{
            type: core.Input
        }], thySize: [{
            type: core.Input
        }] }); })();
        return ThyArrowSwitcherComponent;
    }());
    ThyArrowSwitcherComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef }
    ]; };
    ThyArrowSwitcherComponent.propDecorators = {
        _isArrowSwitcher: [{ type: core.HostBinding, args: ['class.thy-arrow-switcher',] }],
        _isSmallSize: [{ type: core.HostBinding, args: ['class.thy-arrow-switcher-small',] }],
        thyPrevious: [{ type: core.Output }],
        thyNext: [{ type: core.Output }],
        thyTotal: [{ type: core.Input }],
        thySize: [{ type: core.Input }]
    };

    var ThyArrowSwitcherModule = /** @class */ (function () {
        function ThyArrowSwitcherModule() {
        }
ThyArrowSwitcherModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyArrowSwitcherModule });
ThyArrowSwitcherModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyArrowSwitcherModule_Factory(t) { return new (t || ThyArrowSwitcherModule)(); }, imports: [[common.CommonModule, button.ThyButtonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyArrowSwitcherModule, { declarations: [ThyArrowSwitcherComponent], imports: [ɵngcc2.CommonModule, ɵngcc1.ThyButtonModule], exports: [ThyArrowSwitcherComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyArrowSwitcherModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyArrowSwitcherComponent],
                imports: [common.CommonModule, button.ThyButtonModule],
                exports: [ThyArrowSwitcherComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyArrowSwitcherModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyArrowSwitcherComponent = ThyArrowSwitcherComponent;
    exports.ThyArrowSwitcherModule = ThyArrowSwitcherModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
