// ngx-tethys/bundles/ngx-tethys-breadcrumb.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('ngx-tethys/icon'),exports, require('@angular/core'), require('@angular/common'), require('ngx-tethys/icon')) :
    typeof define === 'function' && define.amd ? define('ngx-tethys/breadcrumb', ['@angular/core','@angular/common','ngx-tethys/icon','exports', '@angular/core', '@angular/common', 'ngx-tethys/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ngxTethys.icon,(global['ngx-tethys'] = global['ngx-tethys'] || {}, global['ngx-tethys'].breadcrumb = {}), global.ng.core, global.ng.common, global['ngx-tethys'].icon));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, icon) { 
function ThyBreadcrumbComponent_div_0_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 4);
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r1.svgIconName);
} }
function ThyBreadcrumbComponent_div_0_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 5);
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.iconClasses);
} }
function ThyBreadcrumbComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵtemplate(1, ThyBreadcrumbComponent_div_0_thy_icon_1_Template, 1, 1, "thy-icon", 2);
    ɵngcc0.ɵɵtemplate(2, ThyBreadcrumbComponent_div_0_ng_template_2_Template, 1, 1, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r2 = ɵngcc0.ɵɵreference(3);
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.svgIconName)("ngIfElse", _r2);
} }
var _c0 = ["*"];
'use strict';

    var ThyBreadcrumbComponent = /** @class */ (function () {
        function ThyBreadcrumbComponent() {
            this._isBreadcrumb = true;
            this.isSlash = false;
            this.isBackslash = false;
        }
        Object.defineProperty(ThyBreadcrumbComponent.prototype, "thyIcon", {
            set: function (icon) {
                this.setIcon(icon);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ThyBreadcrumbComponent.prototype, "thySeparator", {
            set: function (type) {
                if (type === 'slash') {
                    this.isSlash = true;
                }
                if (type === 'backslash') {
                    this.isBackslash = true;
                }
            },
            enumerable: false,
            configurable: true
        });
        ThyBreadcrumbComponent.prototype.setIcon = function (icon) {
            if (icon) {
                if (icon.includes('wtf')) {
                    var classes = icon.split(' ');
                    if (classes.length === 1) {
                        classes.unshift('wtf');
                    }
                    this.iconClasses = classes;
                }
                else {
                    this.svgIconName = icon;
                }
            }
            else {
                this.iconClasses = null;
                this.svgIconName = null;
            }
        };
ThyBreadcrumbComponent.ɵfac = function ThyBreadcrumbComponent_Factory(t) { return new (t || ThyBreadcrumbComponent)(); };
ThyBreadcrumbComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyBreadcrumbComponent, selectors: [["thy-breadcrumb"]], hostVars: 6, hostBindings: function ThyBreadcrumbComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-breadcrumb", ctx._isBreadcrumb)("separator-slash", ctx.isSlash)("separator-backslash", ctx.isBackslash);
    } }, inputs: { thyIcon: "thyIcon", thySeparator: "thySeparator" }, exportAs: ["ThyBreadcrumb"], ngContentSelectors: _c0, decls: 2, vars: 1, consts: [["class", "thy-breadcrumb-icon", 4, "ngIf"], [1, "thy-breadcrumb-icon"], [3, "thyIconName", 4, "ngIf", "ngIfElse"], ["iconFont", ""], [3, "thyIconName"], [3, "ngClass"]], template: function ThyBreadcrumbComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThyBreadcrumbComponent_div_0_Template, 4, 2, "div", 0);
        ɵngcc0.ɵɵprojection(1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.svgIconName || ctx.iconClasses);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBreadcrumbComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-breadcrumb',
                template: "\n        <div class=\"thy-breadcrumb-icon\" *ngIf=\"svgIconName || iconClasses\">\n            <thy-icon *ngIf=\"svgIconName; else iconFont\" [thyIconName]=\"svgIconName\"></thy-icon>\n            <ng-template #iconFont>\n                <i [ngClass]=\"iconClasses\"></i>\n            </ng-template>\n        </div>\n        <ng-content></ng-content>\n    ",
                exportAs: 'ThyBreadcrumb',
                changeDetection: core.ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { _isBreadcrumb: [{
            type: core.HostBinding,
            args: ["class.thy-breadcrumb"]
        }], isSlash: [{
            type: core.HostBinding,
            args: ["class.separator-slash"]
        }], isBackslash: [{
            type: core.HostBinding,
            args: ["class.separator-backslash"]
        }], thyIcon: [{
            type: core.Input
        }], thySeparator: [{
            type: core.Input
        }] }); })();
        return ThyBreadcrumbComponent;
    }());
    ThyBreadcrumbComponent.propDecorators = {
        _isBreadcrumb: [{ type: core.HostBinding, args: ["class.thy-breadcrumb",] }],
        isSlash: [{ type: core.HostBinding, args: ["class.separator-slash",] }],
        isBackslash: [{ type: core.HostBinding, args: ["class.separator-backslash",] }],
        thyIcon: [{ type: core.Input }],
        thySeparator: [{ type: core.Input }]
    };

    var ThyBreadcrumbItemComponent = /** @class */ (function () {
        function ThyBreadcrumbItemComponent() {
            this._isBreadcrumbItem = true;
        }
ThyBreadcrumbItemComponent.ɵfac = function ThyBreadcrumbItemComponent_Factory(t) { return new (t || ThyBreadcrumbItemComponent)(); };
ThyBreadcrumbItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyBreadcrumbItemComponent, selectors: [["thy-breadcrumb-item"], ["", "thyBreadcrumbItem", ""]], hostVars: 2, hostBindings: function ThyBreadcrumbItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-breadcrumb-item", ctx._isBreadcrumbItem);
    } }, exportAs: ["ThyBreadcrumbItem"], ngContentSelectors: _c0, decls: 2, vars: 0, consts: [["thyIconName", "angle-right", 1, "default-icon"]], template: function ThyBreadcrumbItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelement(1, "thy-icon", 0);
    } }, directives: [ɵngcc2.ThyIconComponent], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBreadcrumbItemComponent, [{
        type: core.Component,
        args: [{
                selector: 'thy-breadcrumb-item,[thyBreadcrumbItem]',
                template: '<ng-content></ng-content><thy-icon class="default-icon" thyIconName="angle-right"></thy-icon>',
                exportAs: 'ThyBreadcrumbItem',
                changeDetection: core.ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { _isBreadcrumbItem: [{
            type: core.HostBinding,
            args: ["class.thy-breadcrumb-item"]
        }] }); })();
        return ThyBreadcrumbItemComponent;
    }());
    ThyBreadcrumbItemComponent.propDecorators = {
        _isBreadcrumbItem: [{ type: core.HostBinding, args: ["class.thy-breadcrumb-item",] }]
    };

    var ThyBreadcrumbModule = /** @class */ (function () {
        function ThyBreadcrumbModule() {
        }
ThyBreadcrumbModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyBreadcrumbModule });
ThyBreadcrumbModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyBreadcrumbModule_Factory(t) { return new (t || ThyBreadcrumbModule)(); }, imports: [[common.CommonModule, icon.ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyBreadcrumbModule, { declarations: [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent], imports: [ɵngcc1.CommonModule, ɵngcc2.ThyIconModule], exports: [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBreadcrumbModule, [{
        type: core.NgModule,
        args: [{
                declarations: [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent],
                imports: [common.CommonModule, icon.ThyIconModule],
                exports: [ThyBreadcrumbComponent, ThyBreadcrumbItemComponent]
            }]
    }], function () { return []; }, null); })();
        return ThyBreadcrumbModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ThyBreadcrumbComponent = ThyBreadcrumbComponent;
    exports.ThyBreadcrumbItemComponent = ThyBreadcrumbItemComponent;
    exports.ThyBreadcrumbModule = ThyBreadcrumbModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
