// ngx-tethys/esm2015/anchor/anchor.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PlatformModule } from '@angular/cdk/platform';
import { ThyAnchorComponent } from './anchor.component';
import { ThyAnchorLinkComponent } from './anchor-link.component';
import { ThyAffixModule } from 'ngx-tethys/affix';
import * as ɵngcc0 from '@angular/core';
export class ThyAnchorModule {
}
ThyAnchorModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAnchorModule });
ThyAnchorModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAnchorModule_Factory(t) { return new (t || ThyAnchorModule)(); }, imports: [[CommonModule, PlatformModule, ThyAffixModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAnchorModule, { declarations: function () { return [ThyAnchorComponent, ThyAnchorLinkComponent]; }, imports: function () { return [CommonModule, PlatformModule, ThyAffixModule]; }, exports: function () { return [ThyAnchorComponent, ThyAnchorLinkComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAnchorModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyAnchorComponent, ThyAnchorLinkComponent],
                exports: [ThyAnchorComponent, ThyAnchorLinkComponent],
                imports: [CommonModule, PlatformModule, ThyAffixModule]
            }]
    }], null, null); })();
