// ngx-tethys/esm2015/anchor/anchor-link.component.js
import { Platform } from '@angular/cdk/platform';
import { ChangeDetectionStrategy, Component, ContentChild, ElementRef, Input, Renderer2, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { ThyAnchorComponent } from './anchor.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './anchor.component';
import * as ɵngcc2 from '@angular/cdk/platform';
import * as ɵngcc3 from '@angular/common';

const _c0 = ["thyTemplate"];
const _c1 = ["linkTitle"];
function ThyAnchorLinkComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.title);
} }
const _c2 = ["*"];
export class ThyAnchorLinkComponent {
    constructor(elementRef, anchorComponent, platform, renderer) {
        this.elementRef = elementRef;
        this.anchorComponent = anchorComponent;
        this.platform = platform;
        this.renderer = renderer;
        this.thyHref = '#';
        this.title = '';
        this.renderer.addClass(elementRef.nativeElement, 'thy-anchor-link');
    }
    set thyTitle(value) {
        if (value instanceof TemplateRef) {
            this.title = null;
            this.titleTemplate = value;
        }
        else {
            this.title = value;
        }
    }
    ngOnInit() {
        this.anchorComponent.registerLink(this);
    }
    getLinkTitleElement() {
        return this.linkTitle.nativeElement;
    }
    setActive() {
        this.renderer.addClass(this.elementRef.nativeElement, 'thy-anchor-link-active');
    }
    unsetActive() {
        this.renderer.removeClass(this.elementRef.nativeElement, 'thy-anchor-link-active');
    }
    goToClick(e) {
        e.preventDefault();
        e.stopPropagation();
        if (this.platform.isBrowser) {
            this.anchorComponent.handleScrollTo(this);
        }
    }
    ngOnDestroy() {
        this.anchorComponent.unregisterLink(this);
    }
}
ThyAnchorLinkComponent.ɵfac = function ThyAnchorLinkComponent_Factory(t) { return new (t || ThyAnchorLinkComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyAnchorComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyAnchorLinkComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyAnchorLinkComponent, selectors: [["thy-link"]], contentQueries: function ThyAnchorLinkComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.thyTemplate = _t.first);
    } }, viewQuery: function ThyAnchorLinkComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.linkTitle = _t.first);
    } }, inputs: { thyHref: "thyHref", thyTitle: "thyTitle" }, exportAs: ["thyLink"], ngContentSelectors: _c2, decls: 4, vars: 4, consts: [[1, "thy-anchor-link-title", 3, "href", "title", "click"], ["linkTitle", ""], [4, "ngIf", "ngIfElse"]], template: function ThyAnchorLinkComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "a", 0, 1);
        ɵngcc0.ɵɵlistener("click", function ThyAnchorLinkComponent_Template_a_click_0_listener($event) { return ctx.goToClick($event); });
        ɵngcc0.ɵɵtemplate(2, ThyAnchorLinkComponent_span_2_Template, 2, 1, "span", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(3);
    } if (rf & 2) {
        ɵngcc0.ɵɵpropertyInterpolate("href", ctx.thyHref, ɵngcc0.ɵɵsanitizeUrl);
        ɵngcc0.ɵɵpropertyInterpolate("title", ctx.title);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.title)("ngIfElse", ctx.titleTemplate || ctx.thyTemplate);
    } }, directives: [ɵngcc3.NgIf], encapsulation: 2, changeDetection: 0 });
ThyAnchorLinkComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyAnchorComponent },
    { type: Platform },
    { type: Renderer2 }
];
ThyAnchorLinkComponent.propDecorators = {
    thyHref: [{ type: Input }],
    thyTitle: [{ type: Input }],
    thyTemplate: [{ type: ContentChild, args: ['thyTemplate',] }],
    linkTitle: [{ type: ViewChild, args: ['linkTitle', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAnchorLinkComponent, [{
        type: Component,
        args: [{
                selector: 'thy-link',
                exportAs: 'thyLink',
                preserveWhitespaces: false,
                template: `
        <a #linkTitle (click)="goToClick($event)" href="{{ thyHref }}" class="thy-anchor-link-title" title="{{ title }}">
            <span *ngIf="title; else titleTemplate || thyTemplate">{{ title }}</span>
        </a>
        <ng-content></ng-content>
    `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyAnchorComponent }, { type: ɵngcc2.Platform }, { type: ɵngcc0.Renderer2 }]; }, { thyHref: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], thyTemplate: [{
            type: ContentChild,
            args: ['thyTemplate']
        }], linkTitle: [{
            type: ViewChild,
            args: ['linkTitle', { static: true }]
        }] }); })();
