// ngx-tethys/esm2015/dialog/body/dialog-body.component.js
import { Component, Input, HostBinding } from '@angular/core';
import { ThyDialog } from '../dialog.service';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../dialog.service';

const _c0 = ["*"];
export class DialogBodyComponent {
    constructor(dialog) {
        this.dialog = dialog;
        this._isDialogBody = true;
        this.thyClearPaddingClassName = false;
    }
    set thyClearPadding(value) {
        this.thyClearPaddingClassName = coerceBooleanProperty(value);
    }
    ngOnInit() { }
}
DialogBodyComponent.ɵfac = function DialogBodyComponent_Factory(t) { return new (t || DialogBodyComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyDialog)); };
DialogBodyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DialogBodyComponent, selectors: [["thy-dialog-body"]], hostVars: 4, hostBindings: function DialogBodyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dialog-body", ctx._isDialogBody)("dialog-body-clear-padding", ctx.thyClearPaddingClassName);
    } }, inputs: { thyClearPadding: "thyClearPadding" }, exportAs: ["thyDialogBody"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function DialogBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
DialogBodyComponent.ctorParameters = () => [
    { type: ThyDialog }
];
DialogBodyComponent.propDecorators = {
    _isDialogBody: [{ type: HostBinding, args: [`class.dialog-body`,] }],
    thyClearPaddingClassName: [{ type: HostBinding, args: [`class.dialog-body-clear-padding`,] }],
    thyClearPadding: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DialogBodyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-body',
                template: '<ng-content></ng-content>',
                // changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'thyDialogBody'
            }]
    }], function () { return [{ type: ɵngcc1.ThyDialog }]; }, { _isDialogBody: [{
            type: HostBinding,
            args: [`class.dialog-body`]
        }], thyClearPaddingClassName: [{
            type: HostBinding,
            args: [`class.dialog-body-clear-padding`]
        }], thyClearPadding: [{
            type: Input
        }] }); })();
