// ngx-tethys/esm2015/label/label.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyLabelComponent } from './label.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyLabelModule {
}
ThyLabelModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLabelModule });
ThyLabelModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLabelModule_Factory(t) { return new (t || ThyLabelModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLabelModule, { declarations: function () { return [ThyLabelComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyLabelComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLabelModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyLabelComponent],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyLabelComponent]
            }]
    }], null, null); })();
