// ngx-tethys/esm2015/date-picker/date-helper.service.js
import { Injectable } from '@angular/core';
import { formatDate } from '@angular/common';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export function DATE_HELPER_SERVICE_FACTORY() {
    return new DateHelperByDatePipe();
}
export class DateHelperService {
    constructor() {
        this.relyOnDatePipe = this instanceof DateHelperByDatePipe;
    }
}
DateHelperService.ɵfac = function DateHelperService_Factory(t) { return new (t || DateHelperService)(); };
DateHelperService.ɵprov = i0.ɵɵdefineInjectable({ factory: DATE_HELPER_SERVICE_FACTORY, token: DateHelperService, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DateHelperService, [{
        type: Injectable,
        args: [{
                providedIn: 'root',
                useFactory: DATE_HELPER_SERVICE_FACTORY
            }]
    }], function () { return []; }, null); })();
export class DateHelperByDatePipe extends DateHelperService {
    getISOWeek(date) {
        return +this.format(date, 'w');
    }
    getFirstDayOfWeek() {
        return 0;
    }
    format(date, formatStr) {
        return date ? formatDate(date, formatStr, 'zh-Hans') : '';
    }
    transCompatFormat(format) {
        return format && format.replace(/Y/g, 'y').replace(/D/g, 'd');
    }
}
DateHelperByDatePipe.ɵprov = i0.ɵɵdefineInjectable({ factory: DATE_HELPER_SERVICE_FACTORY, token: DateHelperByDatePipe, providedIn: "root" });
