// ngx-tethys/esm2015/date-picker/date-picker.module.js
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule, registerLocaleData } from '@angular/common';
import { NgModule } from '@angular/core';
import localeZhHans from '@angular/common/locales/zh-Hans';
import { LibPackerModule } from './lib/lib-packer.module';
import { BasePickerComponent } from './base-picker.component';
import { HeaderPickerComponent } from './header-picker.component';
import { ThyDatePickerComponent } from './date-picker.component';
import { ThyMonthPickerComponent } from './month-picker.component';
import { ThyRangePickerComponent } from './range-picker.component';
import { ThyWeekPickerComponent } from './week-picker.component';
import { ThyYearPickerComponent } from './year-picker.component';
import { ThyPickerComponent } from './picker.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyInputModule } from 'ngx-tethys/input';
import { DatePickerRequiredValidator, RangePickerRequiredValidator } from './picker.validators';
import { ThyDatePickerFormatPipe, ThyDatePickerFormatStringPipe } from './picker.pipes';
import { ThyDatePickerDirective } from './date-picker.directive';
import { ThyRangePickerDirective } from './range-picker.directive';
import { DatePopupComponent } from './lib/popups/date-popup.component';
import { ThyPopoverModule } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
registerLocaleData(localeZhHans, 'zh-Hans');
export class ThyDatePickerModule {
}
ThyDatePickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDatePickerModule });
ThyDatePickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDatePickerModule_Factory(t) { return new (t || ThyDatePickerModule)(); }, imports: [[CommonModule, OverlayModule, LibPackerModule, ThyIconModule, ThyInputModule, ThyPopoverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDatePickerModule, { declarations: function () { return [HeaderPickerComponent,
        BasePickerComponent,
        ThyPickerComponent,
        ThyDatePickerComponent,
        ThyMonthPickerComponent,
        ThyYearPickerComponent,
        ThyWeekPickerComponent,
        ThyRangePickerComponent,
        DatePickerRequiredValidator,
        RangePickerRequiredValidator,
        ThyDatePickerFormatPipe,
        ThyDatePickerFormatStringPipe,
        ThyDatePickerDirective,
        ThyRangePickerDirective]; }, imports: function () { return [CommonModule, OverlayModule, LibPackerModule, ThyIconModule, ThyInputModule, ThyPopoverModule]; }, exports: function () { return [ThyDatePickerComponent,
        ThyRangePickerComponent,
        ThyMonthPickerComponent,
        ThyYearPickerComponent,
        ThyWeekPickerComponent,
        DatePickerRequiredValidator,
        RangePickerRequiredValidator,
        ThyDatePickerFormatPipe,
        ThyDatePickerFormatStringPipe,
        ThyDatePickerDirective,
        ThyRangePickerDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDatePickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, OverlayModule, LibPackerModule, ThyIconModule, ThyInputModule, ThyPopoverModule],
                exports: [
                    ThyDatePickerComponent,
                    ThyRangePickerComponent,
                    ThyMonthPickerComponent,
                    ThyYearPickerComponent,
                    ThyWeekPickerComponent,
                    DatePickerRequiredValidator,
                    RangePickerRequiredValidator,
                    ThyDatePickerFormatPipe,
                    ThyDatePickerFormatStringPipe,
                    ThyDatePickerDirective,
                    ThyRangePickerDirective
                ],
                declarations: [
                    HeaderPickerComponent,
                    BasePickerComponent,
                    ThyPickerComponent,
                    ThyDatePickerComponent,
                    ThyMonthPickerComponent,
                    ThyYearPickerComponent,
                    ThyWeekPickerComponent,
                    ThyRangePickerComponent,
                    DatePickerRequiredValidator,
                    RangePickerRequiredValidator,
                    ThyDatePickerFormatPipe,
                    ThyDatePickerFormatStringPipe,
                    ThyDatePickerDirective,
                    ThyRangePickerDirective
                ],
                entryComponents: [DatePopupComponent]
            }]
    }], null, null); })();
