// ngx-tethys/esm2015/date-picker/header-picker.component.js
import { ChangeDetectorRef, Component, Input } from '@angular/core';
import { AbstractPickerComponent } from './abstract-picker.component';
import { TinyDate } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
export class HeaderPickerComponent extends AbstractPickerComponent {
    constructor(cdr) {
        super(cdr);
    }
    ngOnInit() {
        super.ngOnInit();
        this.panelMode = this.endPanelMode;
        const allHeaderPanels = ['decade', 'year', 'month'];
        this.supportPanels = allHeaderPanels.slice(0, allHeaderPanels.indexOf(this.endPanelMode) + 1);
    }
    onPanelModeChange(mode) {
        if (this.supportPanels.indexOf(mode) > -1) {
            this.panelMode = mode;
        }
        else {
            // Since the default "click year" logic can be "year panel" -> "date panel", we need force to the end panel otherwise
            this.panelMode = this.endPanelMode;
        }
    }
    onChooseValue(mode, value) {
        if (this.endPanelMode === mode) {
            super.onValueChange(value);
            this.closeOverlay();
        }
    }
    onOpenChange(open) {
        if (!open) {
            this.cleanUp();
        }
        this.thyOpenChange.emit(open);
    }
    // Restore some initial props to let open as new in next time
    cleanUp() {
        this.panelMode = this.endPanelMode;
    }
}
HeaderPickerComponent.ɵfac = function HeaderPickerComponent_Factory(t) { return new (t || HeaderPickerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
HeaderPickerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: HeaderPickerComponent, selectors: [["ng-component"]], inputs: { thyPlaceHolder: "thyPlaceHolder", thyDefaultValue: "thyDefaultValue", thyFormat: "thyFormat" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function HeaderPickerComponent_Template(rf, ctx) { }, encapsulation: 2 });
HeaderPickerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
HeaderPickerComponent.propDecorators = {
    thyPlaceHolder: [{ type: Input }],
    thyDefaultValue: [{ type: Input }],
    thyFormat: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HeaderPickerComponent, [{
        type: Component,
        args: [{
                template: ``
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { thyPlaceHolder: [{
            type: Input
        }], thyDefaultValue: [{
            type: Input
        }], thyFormat: [{
            type: Input
        }] }); })();
