// ngx-tethys/esm2015/date-picker/month-picker.component.js
import { forwardRef, ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, Input, Renderer2 } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { HeaderPickerComponent } from './header-picker.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './picker.component';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './lib/calendar/calendar-header.component';

function ThyMonthPickerComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementStart(2, "div", 2);
    ɵngcc0.ɵɵelementStart(3, "calendar-header", 3);
    ɵngcc0.ɵɵlistener("panelModeChange", function ThyMonthPickerComponent_div_1_Template_calendar_header_panelModeChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.onPanelModeChange($event); })("chooseYear", function ThyMonthPickerComponent_div_1_Template_calendar_header_chooseYear_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onChooseValue("year", $event); })("chooseMonth", function ThyMonthPickerComponent_div_1_Template_calendar_header_chooseMonth_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.onChooseValue("month", $event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMapInterpolate1("thy-calendar-picker-container thy-calendar thy-calendar-month ", ctx_r0.thyPanelClassName, "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("disabledMonth", ctx_r0.thyDisabledDate)("disabledYear", ctx_r0.thyDisabledDate)("panelMode", ctx_r0.panelMode)("value", ctx_r0.thyValue)("enablePrev", true)("enableNext", true);
} }
export class ThyMonthPickerComponent extends HeaderPickerComponent {
    constructor(cdr, renderer, elementRef) {
        super(cdr);
        this.thyFormat = 'yyyy-MM';
        this.endPanelMode = 'month';
        renderer.addClass(elementRef.nativeElement, 'thy-calendar-picker');
    }
}
ThyMonthPickerComponent.ɵfac = function ThyMonthPickerComponent_Factory(t) { return new (t || ThyMonthPickerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyMonthPickerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMonthPickerComponent, selectors: [["thy-month-picker"]], inputs: { thyFormat: "thyFormat" }, exportAs: ["thyMonthPicker"], features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useExisting: forwardRef(() => ThyMonthPickerComponent)
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], decls: 2, vars: 13, consts: [[3, "isRange", "value", "open", "disabled", "format", "allowClear", "className", "autoFocus", "placeholder", "readonly", "size", "suffixIcon", "valueChange", "openChange"], [4, "ngIf"], [1, "thy-calendar-header-wrap"], [3, "disabledMonth", "disabledYear", "panelMode", "value", "enablePrev", "enableNext", "panelModeChange", "chooseYear", "chooseMonth"]], template: function ThyMonthPickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "thy-picker", 0);
        ɵngcc0.ɵɵlistener("valueChange", function ThyMonthPickerComponent_Template_thy_picker_valueChange_0_listener($event) { return ctx.onValueChange($event); })("openChange", function ThyMonthPickerComponent_Template_thy_picker_openChange_0_listener($event) { return ctx.onOpenChange($event); });
        ɵngcc0.ɵɵtemplate(1, ThyMonthPickerComponent_div_1_Template, 4, 9, "div", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("isRange", false)("value", ctx.thyValue)("open", ctx.thyOpen)("disabled", ctx.thyDisabled)("format", ctx.thyFormat)("allowClear", ctx.thyAllowClear)("className", ctx.thyOriginClassName)("autoFocus", ctx.thyAutoFocus)("placeholder", ctx.thyPlaceHolder)("readonly", ctx.thyReadonly)("size", ctx.thySize)("suffixIcon", ctx.thySuffixIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.realOpenState);
    } }, directives: [ɵngcc1.ThyPickerComponent, ɵngcc2.NgIf, ɵngcc3.CalendarHeaderComponent], encapsulation: 2, changeDetection: 0 });
ThyMonthPickerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: Renderer2 },
    { type: ElementRef }
];
ThyMonthPickerComponent.propDecorators = {
    thyFormat: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMonthPickerComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                selector: 'thy-month-picker',
                exportAs: 'thyMonthPicker',
                template: "<thy-picker\n  [isRange]=\"false\"\n  [value]=\"thyValue\"\n  (valueChange)=\"onValueChange($event)\"\n  [open]=\"thyOpen\"\n  [disabled]=\"thyDisabled\"\n  [format]=\"thyFormat\"\n  [allowClear]=\"thyAllowClear\"\n  [className]=\"thyOriginClassName\"\n  [autoFocus]=\"thyAutoFocus\"\n  [placeholder]=\"thyPlaceHolder\"\n  [readonly]=\"thyReadonly\"\n  [size]=\"thySize\"\n  [suffixIcon]=\"thySuffixIcon\"\n  (openChange)=\"onOpenChange($event)\"\n>\n  <div *ngIf=\"realOpenState\">\n    <div class=\"thy-calendar-picker-container thy-calendar thy-calendar-month {{ thyPanelClassName }}\">\n      <div class=\"thy-calendar-header-wrap\">\n        <calendar-header\n          [disabledMonth]=\"thyDisabledDate\"\n          [disabledYear]=\"thyDisabledDate\"\n          [panelMode]=\"panelMode\"\n          (panelModeChange)=\"onPanelModeChange($event)\"\n          [value]=\"thyValue\"\n          (chooseYear)=\"onChooseValue('year', $event)\"\n          (chooseMonth)=\"onChooseValue('month', $event)\"\n          [enablePrev]=\"true\"\n          [enableNext]=\"true\"\n        ></calendar-header>\n      </div>\n    </div>\n  </div>\n</thy-picker>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: forwardRef(() => ThyMonthPickerComponent)
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { thyFormat: [{
            type: Input
        }] }); })();
