// ngx-tethys/esm2015/date-picker/picker.pipes.js
import { Pipe } from '@angular/core';
import { DateHelperService } from './date-helper.service';
import { transformDateValue } from './picker.util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './date-helper.service';
export class ThyDatePickerFormatPipe {
    constructor(dateHelper) {
        this.dateHelper = dateHelper;
    }
    transform(originalValue, formatStr) {
        const { value, withTime } = transformDateValue(originalValue);
        if (!formatStr) {
            formatStr = withTime ? 'yyyy-MM-dd HH:mm' : 'yyyy-MM-dd';
        }
        if (!Array.isArray(value)) {
            return this.dateHelper.format(value, formatStr);
        }
        else {
            return value.map(date => this.dateHelper.format(date, formatStr)).join(' ~ ');
        }
    }
}
ThyDatePickerFormatPipe.ɵfac = function ThyDatePickerFormatPipe_Factory(t) { return new (t || ThyDatePickerFormatPipe)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DateHelperService)); };
ThyDatePickerFormatPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "thyDatePickerFormat", type: ThyDatePickerFormatPipe, pure: true });
ThyDatePickerFormatPipe.ctorParameters = () => [
    { type: DateHelperService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDatePickerFormatPipe, [{
        type: Pipe,
        args: [{ name: 'thyDatePickerFormat' }]
    }], function () { return [{ type: ɵngcc1.DateHelperService }]; }, null); })();
export class ThyDatePickerFormatStringPipe {
    constructor(dateHelper) {
        this.dateHelper = dateHelper;
    }
    transform(originalValue) {
        const { withTime } = transformDateValue(originalValue);
        return withTime ? 'yyyy-MM-dd HH:mm' : 'yyyy-MM-dd';
    }
}
ThyDatePickerFormatStringPipe.ɵfac = function ThyDatePickerFormatStringPipe_Factory(t) { return new (t || ThyDatePickerFormatStringPipe)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DateHelperService)); };
ThyDatePickerFormatStringPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "thyDatePickerFormatString", type: ThyDatePickerFormatStringPipe, pure: true });
ThyDatePickerFormatStringPipe.ctorParameters = () => [
    { type: DateHelperService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDatePickerFormatStringPipe, [{
        type: Pipe,
        args: [{ name: 'thyDatePickerFormatString' }]
    }], function () { return [{ type: ɵngcc1.DateHelperService }]; }, null); })();
