// ngx-tethys/esm2015/date-picker/picker.validators.js
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { instanceOfDateEntry, instanceOfRangeEntry } from './standard-types';
import * as ɵngcc0 from '@angular/core';
const DATE_PICKER_REQUIRED_VALIDATOR = {
    provide: NG_VALIDATORS,
    // tslint:disable-next-line: no-use-before-declare
    useExisting: forwardRef(() => DatePickerRequiredValidator),
    multi: true
};
export class DatePickerRequiredValidator {
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = value != null && value !== false && `${value}` !== 'false';
    }
    validate(control) {
        return this.required ? this.validateRequired(control) : null;
    }
    validateRequired(control) {
        return isEmptyInputDateValue(control.value) ? { required: true } : null;
    }
}
DatePickerRequiredValidator.ɵfac = function DatePickerRequiredValidator_Factory(t) { return new (t || DatePickerRequiredValidator)(); };
DatePickerRequiredValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DatePickerRequiredValidator, selectors: [["", "thyDatePicker", "", "required", "", "formControlName", ""], ["", "thyDatePicker", "", "required", "", "formControl", ""], ["", "thyDatePicker", "", "required", "", "ngModel", ""], ["thy-date-picker", "required", "", "formControlName", ""], ["thy-date-picker", "required", "", "formControl", ""], ["thy-date-picker", "required", "", "ngModel", ""]], inputs: { required: "required" }, features: [ɵngcc0.ɵɵProvidersFeature([DATE_PICKER_REQUIRED_VALIDATOR])] });
DatePickerRequiredValidator.propDecorators = {
    required: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DatePickerRequiredValidator, [{
        type: Directive,
        args: [{
                selector: '[thyDatePicker][required][formControlName],[thyDatePicker][required][formControl],[thyDatePicker][required][ngModel],thy-date-picker[required][formControlName],thy-date-picker[required][formControl],thy-date-picker[required][ngModel]',
                providers: [DATE_PICKER_REQUIRED_VALIDATOR]
            }]
    }], null, { required: [{
            type: Input
        }] }); })();
const RANGE_PICKER_REQUIRED_VALIDATOR = {
    provide: NG_VALIDATORS,
    // tslint:disable-next-line: no-use-before-declare
    useExisting: forwardRef(() => RangePickerRequiredValidator),
    multi: true
};
export class RangePickerRequiredValidator {
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = value != null && value !== false && `${value}` !== 'false';
    }
    validate(control) {
        return this.required ? this.validateRequired(control) : null;
    }
    validateRequired(control) {
        return isEmptyInputRangeValue(control.value) ? { required: true } : null;
    }
}
RangePickerRequiredValidator.ɵfac = function RangePickerRequiredValidator_Factory(t) { return new (t || RangePickerRequiredValidator)(); };
RangePickerRequiredValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: RangePickerRequiredValidator, selectors: [["", "thyRangePicker", "", "required", "", "formControlName", ""], ["", "thyRangePicker", "", "required", "", "formControl", ""], ["", "thyRangePicker", "", "required", "", "ngModel", ""], ["thy-range-picker", "required", "", "formControlName", ""], ["thy-range-picker", "required", "", "formControl", ""], ["thy-range-picker", "required", "", "ngModel", ""]], inputs: { required: "required" }, features: [ɵngcc0.ɵɵProvidersFeature([RANGE_PICKER_REQUIRED_VALIDATOR])] });
RangePickerRequiredValidator.propDecorators = {
    required: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RangePickerRequiredValidator, [{
        type: Directive,
        args: [{
                selector: '[thyRangePicker][required][formControlName],[thyRangePicker][required][formControl],[thyRangePicker][required][ngModel],thy-range-picker[required][formControlName],thy-range-picker[required][formControl],thy-range-picker[required][ngModel]',
                providers: [RANGE_PICKER_REQUIRED_VALIDATOR]
            }]
    }], null, { required: [{
            type: Input
        }] }); })();
function isEmptyInputRangeValue(value) {
    return value == null || (instanceOfRangeEntry(value) && (value.begin == null || value.end == null));
}
function isEmptyInputDateValue(value) {
    return value == null || (instanceOfDateEntry(value) && value.date == null);
}
