// ngx-tethys/esm2015/date-picker/range-picker.directive.js
import { forwardRef, ChangeDetectorRef, ElementRef, Directive } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { PickerDirective } from './abstract-picker.directive';
import { ThyPopover } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
export class ThyRangePickerDirective extends PickerDirective {
    constructor(elementRef, cdr, thyPopover) {
        super(elementRef, cdr, thyPopover);
        this.isRange = true;
    }
}
ThyRangePickerDirective.ɵfac = function ThyRangePickerDirective_Factory(t) { return new (t || ThyRangePickerDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover)); };
ThyRangePickerDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyRangePickerDirective, selectors: [["", "thyRangePicker", ""]], exportAs: ["thyRangePicker"], features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useExisting: forwardRef(() => ThyRangePickerDirective)
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature] });
ThyRangePickerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef },
    { type: ThyPopover }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRangePickerDirective, [{
        type: Directive,
        args: [{
                selector: '[thyRangePicker]',
                exportAs: 'thyRangePicker',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: forwardRef(() => ThyRangePickerDirective)
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.ThyPopover }]; }, null); })();
