// ngx-tethys/esm2015/back-top/back-top.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyBackTopComponent } from './back-top.component';
import * as ɵngcc0 from '@angular/core';
export class ThyBackTopModule {
}
ThyBackTopModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyBackTopModule });
ThyBackTopModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyBackTopModule_Factory(t) { return new (t || ThyBackTopModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyBackTopModule, { declarations: function () { return [ThyBackTopComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyBackTopComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBackTopModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyBackTopComponent],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyBackTopComponent]
            }]
    }], null, null); })();
