// ngx-tethys/esm2015/empty/empty.component.js
import { ThyTranslate } from 'ngx-tethys/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { Component, ContentChild, ElementRef, HostBinding, Input, NgZone, Renderer2, TemplateRef } from '@angular/core';
import { ThyEmptyConfig } from './empty.config';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from './empty.config';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/icon';

const _c0 = ["extra"];
function ThyEmptyComponent_img_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 6);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("src", ctx_r0.thyImageUrl, ɵngcc0.ɵɵsanitizeUrl);
} }
function ThyEmptyComponent_ng_template_2_thy_icon_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 9);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r5.thyIconName);
} }
function ThyEmptyComponent_ng_template_2_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(1, "svg", 10);
    ɵngcc0.ɵɵelementStart(2, "defs");
    ɵngcc0.ɵɵelementStart(3, "filter", 11);
    ɵngcc0.ɵɵelement(4, "feGaussianBlur", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "g", 13);
    ɵngcc0.ɵɵelementStart(6, "g", 14);
    ɵngcc0.ɵɵelement(7, "ellipse", 15);
    ɵngcc0.ɵɵelementStart(8, "g", 16);
    ɵngcc0.ɵɵelement(9, "polygon", 17);
    ɵngcc0.ɵɵelement(10, "path", 18);
    ɵngcc0.ɵɵelement(11, "path", 19);
    ɵngcc0.ɵɵelementStart(12, "text", 20);
    ɵngcc0.ɵɵelementStart(13, "tspan", 21);
    ɵngcc0.ɵɵtext(14, "</null>");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(15, "rect", 22);
    ɵngcc0.ɵɵelement(16, "rect", 23);
    ɵngcc0.ɵɵelement(17, "path", 24);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyEmptyComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyEmptyComponent_ng_template_2_thy_icon_0_Template, 1, 1, "thy-icon", 7);
    ɵngcc0.ɵɵtemplate(1, ThyEmptyComponent_ng_template_2_ng_template_1_Template, 18, 0, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
} if (rf & 2) {
    const _r6 = ɵngcc0.ɵɵreference(2);
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.thyIconName)("ngIfElse", _r6);
} }
function ThyEmptyComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r3.thyDescription, "\n");
} }
function ThyEmptyComponent_div_7_ng_template_1_Template(rf, ctx) { }
function ThyEmptyComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵtemplate(1, ThyEmptyComponent_div_7_ng_template_1_Template, 0, 0, "ng-template", 27);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r4.extraTemplateRef);
} }
const _c1 = function (a0, a1) { return { "thy-empty-text": a0, "thy-empty-text-with-extra": a1 }; };
const sizeClassMap = {
    lg: ['thy-empty-state', 'thy-empty-state--lg'],
    md: ['thy-empty-state'],
    sm: ['thy-empty-state', 'thy-empty-state--sm']
};
const sizeMap = {
    lg: {
        height: 168,
        offsetTop: 30,
        defaultMarginTop: 120 // 不自动计算默认的 top 距离
    },
    md: {
        height: 118,
        offsetTop: 20,
        defaultMarginTop: 10
    },
    sm: {
        height: 78,
        offsetTop: 10,
        defaultMarginTop: 10
    }
};
export class ThyEmptyComponent {
    constructor(thyTranslate, thyEmptyConfig, elementRef, renderer, ngZone) {
        this.thyTranslate = thyTranslate;
        this.thyEmptyConfig = thyEmptyConfig;
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.sizeClass = sizeClassMap['md'].join(' ');
    }
    set thySize(value) {
        const classList = sizeClassMap[value || 'md'];
        if (classList) {
            this.sizeClass = classList.join(' ');
        }
    }
    get displayText() {
        if (this.thyMessage) {
            return this.thyMessage;
        }
        else if (this.thyTranslationKey) {
            return this.thyTranslate.instant(this.thyTranslationKey, this.thyTranslationValues);
        }
        else if (this.thyEntityName) {
            return this.thyTranslate.instant(this.thyEmptyConfig.noResultWithTargetTranslateKey, {
                target: this.thyEntityName
            });
        }
        else if (this.thyEntityNameTranslateKey) {
            return this.thyTranslate.instant(this.thyEmptyConfig.noResultWithTargetTranslateKey, {
                target: this.thyTranslate.instant(this.thyEntityNameTranslateKey)
            });
        }
        else {
            return this.thyTranslate.instant(this.thyEmptyConfig.noResultTranslateKey);
        }
    }
    _calculatePosition() {
        const sizeOptions = sizeMap[this.thySize || 'md'];
        const topAuto = coerceBooleanProperty(this.thyTopAuto);
        let marginTop = null;
        if (topAuto) {
            // 选择参考父容器居中
            const containerElement = this.thyContainer ? this.thyContainer.nativeElement : this.elementRef.nativeElement.parentElement;
            // containerElement.height;
            let emptyStateHeight = this.elementRef.nativeElement.offsetHeight;
            // 高度没有自动计算出来使用默认值
            if (emptyStateHeight <= 10) {
                emptyStateHeight = sizeOptions.height;
            }
            marginTop = (containerElement.offsetHeight - emptyStateHeight) / 2 - sizeOptions.offsetTop;
            // marginTop = (containerElement.offsetHeight - emptyStateHeight) / 2;
            if (marginTop < 0) {
                marginTop = 0; // sizeOptions.defaultMarginTop;
            }
        }
        else {
            if (this.thyMarginTop) {
                marginTop = this.thyMarginTop;
            }
            else {
                marginTop = 0; // sizeOptions.defaultMarginTop;
            }
        }
        if (marginTop) {
            this.renderer.setStyle(this.elementRef.nativeElement, 'marginTop', marginTop + 'px');
        }
    }
    ngOnInit() { }
    ngAfterViewInit() {
        this.ngZone.runOutsideAngular(() => {
            setTimeout(() => {
                this._calculatePosition();
            }, 50);
        });
    }
}
ThyEmptyComponent.ɵfac = function ThyEmptyComponent_Factory(t) { return new (t || ThyEmptyComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyEmptyConfig), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyEmptyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyEmptyComponent, selectors: [["thy-empty"]], contentQueries: function ThyEmptyComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.extraTemplateRef = _t.first);
    } }, hostVars: 2, hostBindings: function ThyEmptyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.sizeClass);
    } }, inputs: { thySize: "thySize", thyMessage: "thyMessage", thyTranslationKey: "thyTranslationKey", thyTranslationValues: "thyTranslationValues", thyEntityName: "thyEntityName", thyEntityNameTranslateKey: "thyEntityNameTranslateKey", thyIconName: "thyIconName", thyMarginTop: "thyMarginTop", thyTopAuto: "thyTopAuto", thyContainer: "thyContainer", thyImageUrl: "thyImageUrl", thyDescription: "thyDescription" }, decls: 8, vars: 9, consts: [[1, "thy-empty-icon"], ["class", "empty-img", 3, "src", 4, "ngIf", "ngIfElse"], ["icon", ""], [3, "ngClass"], ["class", "thy-empty-description", 4, "ngIf"], ["class", "thy-empty-extra", 4, "ngIf"], [1, "empty-img", 3, "src"], ["class", "empty-icon", 3, "thyIconName", 4, "ngIf", "ngIfElse"], ["svg", ""], [1, "empty-icon", 3, "thyIconName"], ["width", "148px", "height", "134px", "viewBox", "0 0 148 134", "version", "1.1", "xmlns", "http://www.w3.org/2000/svg", 0, "xmlns", "xlink", "http://www.w3.org/1999/xlink"], ["x", "0.0%", "y", "0.0%", "width", "100.0%", "height", "100.0%", "filterUnits", "objectBoundingBox", "id", "filter-1"], ["stdDeviation", "0", "in", "SourceGraphic"], ["id", "148x134", "stroke", "none", "stroke-width", "1", "fill", "none", "fill-rule", "evenodd"], ["id", "\u7F16\u7EC4-6", "transform", "translate(1.000000, 1.000000)"], ["id", "\u692D\u5706\u5F62", "fill", "#EDEEF2", "opacity", "0.3", "filter", "url(#filter-1)", "cx", "73.0800017", "cy", "115.920003", "rx", "73.0800017", "ry", "16.8000004"], ["id", "\u7F16\u7EC4-5", "transform", "translate(15.120000, 0.000000)"], ["id", "\u77E9\u5F62", "fill", "#E2E4E9", "points", "19.2789848 49.5600011 99.1200023 48.7200011 117.600003 75.9297673 117.600003 92.313049 0 92.313049 0 75.0356267"], ["d", "M23.5200005,0 L94.0800002,0 C97.7913538,2.06413823e-16 100.8,3.00864655 100.8,6.72000015 L100.8,99.1200023 L100.8,99.1200023 L16.8000004,99.1200023 L16.8000004,6.72000015 C16.8000004,3.00864655 19.8086469,1.56994302e-15 23.5200005,0 Z", "id", "\u77E9\u5F62", "fill", "#F9FAFB"], ["d", "M30.9200007,12.4400003 L86.6800019,12.4400003 C88.5356787,12.4400003 90.040002,13.9443236 90.040002,15.8000004 L90.040002,42.000001 C90.040002,43.8556778 88.5356787,45.360001 86.6800019,45.360001 L30.9200007,45.360001 C29.0643239,45.360001 27.5600006,43.8556778 27.5600006,42.000001 L27.5600006,15.8000004 C27.5600006,13.9443236 29.0643239,12.4400003 30.9200007,12.4400003 Z", "id", "\u77E9\u5F62", "fill", "#E8EAEE"], ["id", "</null>", "font-family", "PingFangSC-Medium, PingFang SC", "font-size", "15.1200003", "font-weight", "400", "fill", "#BCBECD"], ["x", "33.6000008", "y", "32.8000004"], ["id", "\u77E9\u5F62", "fill", "#E8EAEE", "x", "27.5600006", "y", "52.0800012", "width", "61.4800014", "height", "5.04000011", "rx", "2.52000006"], ["id", "\u77E9\u5F62\u5907\u4EFD", "fill", "#E8EAEE", "x", "27.5600006", "y", "63.8400014", "width", "61.4800014", "height", "5.04000011", "rx", "2.52000006"], ["d", "M0,75.6000017 L29.280235,75.6000017 C32.0637502,75.6000017 34.3202352,77.8564866 34.3202352,80.6400018 L34.3202352,86.2591426 C34.3202352,89.0426578 36.5767201,91.2991427 39.3602353,91.2991427 L78.4136737,91.2991427 C81.1971889,91.2991427 83.4536738,89.0426578 83.4536738,86.2591426 L83.4536738,80.6400018 C83.4536738,77.8564866 85.7101587,75.6000017 88.4936739,75.6000017 L117.600003,75.6000017 L117.600003,75.6000017 L117.600003,110.880003 C117.600003,115.519195 113.839194,119.280003 109.200002,119.280003 L8.40000019,119.280003 C3.76080819,119.280003 -6.53729019e-15,115.519195 0,110.880003 L0,75.6000017 L0,75.6000017 Z", "id", "\u77E9\u5F62", "fill", "#EDEFF2"], [1, "thy-empty-description"], [1, "thy-empty-extra"], [3, "ngTemplateOutlet"]], template: function ThyEmptyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyEmptyComponent_img_1_Template, 1, 1, "img", 1);
        ɵngcc0.ɵɵtemplate(2, ThyEmptyComponent_ng_template_2_Template, 3, 2, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, ThyEmptyComponent_div_6_Template, 2, 1, "div", 4);
        ɵngcc0.ɵɵtemplate(7, ThyEmptyComponent_div_7_Template, 2, 1, "div", 5);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(3);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyImageUrl)("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(6, _c1, !ctx.thyDescription, ctx.thyDescription));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.displayText, "\n");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyDescription);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.extraTemplateRef);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgClass, ɵngcc4.ThyIconComponent, ɵngcc3.NgTemplateOutlet], encapsulation: 2 });
ThyEmptyComponent.ctorParameters = () => [
    { type: ThyTranslate },
    { type: ThyEmptyConfig },
    { type: ElementRef },
    { type: Renderer2 },
    { type: NgZone }
];
ThyEmptyComponent.propDecorators = {
    sizeClass: [{ type: HostBinding, args: ['class',] }],
    thyMessage: [{ type: Input }],
    thyTranslationKey: [{ type: Input }],
    thyTranslationValues: [{ type: Input }],
    thyEntityName: [{ type: Input }],
    thyEntityNameTranslateKey: [{ type: Input }],
    thyIconName: [{ type: Input }],
    thySize: [{ type: Input }],
    thyMarginTop: [{ type: Input }],
    thyTopAuto: [{ type: Input }],
    thyContainer: [{ type: Input }],
    thyImageUrl: [{ type: Input }],
    thyDescription: [{ type: Input }],
    extraTemplateRef: [{ type: ContentChild, args: ['extra',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyEmptyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-empty',
                template: "<div class=\"thy-empty-icon\">\n  <img *ngIf=\"thyImageUrl; else icon\" [src]=\"thyImageUrl\" class=\"empty-img\" />\n  <ng-template #icon>\n    <thy-icon class=\"empty-icon\" [thyIconName]=\"thyIconName\" *ngIf=\"thyIconName; else svg\"></thy-icon>\n    <ng-template #svg>\n      <ng-container>\n        <svg\n          width=\"148px\"\n          height=\"134px\"\n          viewBox=\"0 0 148 134\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n        >\n          <defs>\n            <filter x=\"0.0%\" y=\"0.0%\" width=\"100.0%\" height=\"100.0%\" filterUnits=\"objectBoundingBox\" id=\"filter-1\">\n              <feGaussianBlur stdDeviation=\"0\" in=\"SourceGraphic\"></feGaussianBlur>\n            </filter>\n          </defs>\n          <g id=\"148x134\" stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <g id=\"\u7F16\u7EC4-6\" transform=\"translate(1.000000, 1.000000)\">\n              <ellipse\n                id=\"\u692D\u5706\u5F62\"\n                fill=\"#EDEEF2\"\n                opacity=\"0.3\"\n                filter=\"url(#filter-1)\"\n                cx=\"73.0800017\"\n                cy=\"115.920003\"\n                rx=\"73.0800017\"\n                ry=\"16.8000004\"\n              ></ellipse>\n              <g id=\"\u7F16\u7EC4-5\" transform=\"translate(15.120000, 0.000000)\">\n                <polygon\n                  id=\"\u77E9\u5F62\"\n                  fill=\"#E2E4E9\"\n                  points=\"19.2789848 49.5600011 99.1200023 48.7200011 117.600003 75.9297673 117.600003 92.313049 0 92.313049 0 75.0356267\"\n                ></polygon>\n                <path\n                  d=\"M23.5200005,0 L94.0800002,0 C97.7913538,2.06413823e-16 100.8,3.00864655 100.8,6.72000015 L100.8,99.1200023 L100.8,99.1200023 L16.8000004,99.1200023 L16.8000004,6.72000015 C16.8000004,3.00864655 19.8086469,1.56994302e-15 23.5200005,0 Z\"\n                  id=\"\u77E9\u5F62\"\n                  fill=\"#F9FAFB\"\n                ></path>\n                <path\n                  d=\"M30.9200007,12.4400003 L86.6800019,12.4400003 C88.5356787,12.4400003 90.040002,13.9443236 90.040002,15.8000004 L90.040002,42.000001 C90.040002,43.8556778 88.5356787,45.360001 86.6800019,45.360001 L30.9200007,45.360001 C29.0643239,45.360001 27.5600006,43.8556778 27.5600006,42.000001 L27.5600006,15.8000004 C27.5600006,13.9443236 29.0643239,12.4400003 30.9200007,12.4400003 Z\"\n                  id=\"\u77E9\u5F62\"\n                  fill=\"#E8EAEE\"\n                ></path>\n                <text\n                  id=\"&lt;/null&gt;\"\n                  font-family=\"PingFangSC-Medium, PingFang SC\"\n                  font-size=\"15.1200003\"\n                  font-weight=\"400\"\n                  fill=\"#BCBECD\"\n                >\n                  <tspan x=\"33.6000008\" y=\"32.8000004\">&lt;/null&gt;</tspan>\n                </text>\n                <rect id=\"\u77E9\u5F62\" fill=\"#E8EAEE\" x=\"27.5600006\" y=\"52.0800012\" width=\"61.4800014\" height=\"5.04000011\" rx=\"2.52000006\"></rect>\n                <rect\n                  id=\"\u77E9\u5F62\u5907\u4EFD\"\n                  fill=\"#E8EAEE\"\n                  x=\"27.5600006\"\n                  y=\"63.8400014\"\n                  width=\"61.4800014\"\n                  height=\"5.04000011\"\n                  rx=\"2.52000006\"\n                ></rect>\n                <path\n                  d=\"M0,75.6000017 L29.280235,75.6000017 C32.0637502,75.6000017 34.3202352,77.8564866 34.3202352,80.6400018 L34.3202352,86.2591426 C34.3202352,89.0426578 36.5767201,91.2991427 39.3602353,91.2991427 L78.4136737,91.2991427 C81.1971889,91.2991427 83.4536738,89.0426578 83.4536738,86.2591426 L83.4536738,80.6400018 C83.4536738,77.8564866 85.7101587,75.6000017 88.4936739,75.6000017 L117.600003,75.6000017 L117.600003,75.6000017 L117.600003,110.880003 C117.600003,115.519195 113.839194,119.280003 109.200002,119.280003 L8.40000019,119.280003 C3.76080819,119.280003 -6.53729019e-15,115.519195 0,110.880003 L0,75.6000017 L0,75.6000017 Z\"\n                  id=\"\u77E9\u5F62\"\n                  fill=\"#EDEFF2\"\n                ></path>\n              </g>\n            </g>\n          </g>\n        </svg>\n      </ng-container>\n    </ng-template>\n  </ng-template>\n</div>\n\n<div [ngClass]=\"{ 'thy-empty-text': !thyDescription, 'thy-empty-text-with-extra': thyDescription }\">\n  {{ displayText }}\n</div>\n\n<div class=\"thy-empty-description\" *ngIf=\"thyDescription\">\n  {{ thyDescription }}\n</div>\n\n<div class=\"thy-empty-extra\" *ngIf=\"extraTemplateRef\">\n  <ng-template [ngTemplateOutlet]=\"extraTemplateRef\"></ng-template>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ThyTranslate }, { type: ɵngcc2.ThyEmptyConfig }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.NgZone }]; }, { sizeClass: [{
            type: HostBinding,
            args: ['class']
        }], thySize: [{
            type: Input
        }], thyMessage: [{
            type: Input
        }], thyTranslationKey: [{
            type: Input
        }], thyTranslationValues: [{
            type: Input
        }], thyEntityName: [{
            type: Input
        }], thyEntityNameTranslateKey: [{
            type: Input
        }], thyIconName: [{
            type: Input
        }], thyMarginTop: [{
            type: Input
        }], thyTopAuto: [{
            type: Input
        }], thyContainer: [{
            type: Input
        }], thyImageUrl: [{
            type: Input
        }], thyDescription: [{
            type: Input
        }], extraTemplateRef: [{
            type: ContentChild,
            args: ['extra']
        }] }); })();
