// ngx-tethys/esm2015/empty/empty.module.js
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThySharedModule } from 'ngx-tethys/shared';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyEmptyComponent } from './empty.component';
import { ThyEmptyConfig } from './empty.config';
import * as ɵngcc0 from '@angular/core';
export class ThyEmptyModule {
}
ThyEmptyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyEmptyModule });
ThyEmptyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyEmptyModule_Factory(t) { return new (t || ThyEmptyModule)(); }, providers: [ThyEmptyConfig], imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyEmptyModule, { declarations: function () { return [ThyEmptyComponent]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyEmptyComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyEmptyModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyEmptyComponent],
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyEmptyComponent],
                providers: [ThyEmptyConfig]
            }]
    }], null, null); })();
