// ngx-tethys/esm2015/slide/slide-layout/slide-layout.component.js
import { Component, ViewEncapsulation, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThySlideLayoutComponent {
    constructor() {
        this.slideLayout = true;
    }
    ngOnInit() { }
}
ThySlideLayoutComponent.ɵfac = function ThySlideLayoutComponent_Factory(t) { return new (t || ThySlideLayoutComponent)(); };
ThySlideLayoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySlideLayoutComponent, selectors: [["thy-slide-layout"]], hostVars: 2, hostBindings: function ThySlideLayoutComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-slide-layout", ctx.slideLayout);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThySlideLayoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThySlideLayoutComponent.ctorParameters = () => [];
ThySlideLayoutComponent.propDecorators = {
    slideLayout: [{ type: HostBinding, args: ['class.thy-slide-layout',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySlideLayoutComponent, [{
        type: Component,
        args: [{
                selector: 'thy-slide-layout',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return []; }, { slideLayout: [{
            type: HostBinding,
            args: ['class.thy-slide-layout']
        }] }); })();
