// ngx-tethys/esm2015/mention/mention.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThyMentionDirective } from './mention.directive';
import { ThyMentionSuggestionsComponent } from './suggestions/suggestions.component';
import { ThyPopoverModule } from 'ngx-tethys/popover';
import { ThyListModule } from 'ngx-tethys/list';
import { ThyLoadingModule } from 'ngx-tethys/loading';
import * as ɵngcc0 from '@angular/core';
export class ThyMentionModule {
}
ThyMentionModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyMentionModule });
ThyMentionModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyMentionModule_Factory(t) { return new (t || ThyMentionModule)(); }, providers: [], imports: [[CommonModule, FormsModule, ThyPopoverModule, ThyListModule, ThyLoadingModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyMentionModule, { declarations: function () { return [ThyMentionDirective, ThyMentionSuggestionsComponent]; }, imports: function () { return [CommonModule, FormsModule, ThyPopoverModule, ThyListModule, ThyLoadingModule]; }, exports: function () { return [ThyMentionDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMentionModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyMentionDirective, ThyMentionSuggestionsComponent],
                imports: [CommonModule, FormsModule, ThyPopoverModule, ThyListModule, ThyLoadingModule],
                exports: [ThyMentionDirective],
                providers: [],
                entryComponents: [ThyMentionSuggestionsComponent]
            }]
    }], null, null); })();
