// ngx-tethys/schematics/ng-update/update-11/migrations/import-name-change.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportNameChangeMigrationByNg11 = exports.IMPORT_NAME_CHANGE_RELATION = void 0;
const class_1 = require("../../../class");
exports.IMPORT_NAME_CHANGE_RELATION = {
    ThyGridModule: 'ThyTableModule',
    ThyGridColumn: 'ThyTableColumn',
    ThyGridEmptyOptions: 'ThyTableEmptyOptions',
    ThyGridEvent: 'ThyTableEvent',
    ThyGridDraggableEvent: 'ThyTableDraggableEvent',
    ThyGridRowEvent: 'ThyTableRowEvent',
    IThyGridColumnParentComponent: 'IThyTableColumnParentComponent',
    ThyGridTheme: 'ThyTableTheme',
    ThyGridMode: 'ThyTableMode',
    ThyGridSize: 'ThyTableSize',
    ThyGridComponent: 'ThyTableComponent',
    THY_GRID_COLUMN_PARENT_COMPONENT: 'THY_TABLE_COLUMN_PARENT_COMPONENT',
    ThyGridColumnComponent: 'ThyTableColumnComponent'
};
class ImportNameChangeMigrationByNg11 extends class_1.ImportNameChangeBase {
    constructor() {
        super(...arguments);
        this.relation = exports.IMPORT_NAME_CHANGE_RELATION;
    }
}
exports.ImportNameChangeMigrationByNg11 = ImportNameChangeMigrationByNg11;
