// ngx-tethys/esm2015/form/form-group.component.js
import { Component, HostBinding, Optional, Input, ViewEncapsulation, ContentChild, ChangeDetectionStrategy } from '@angular/core';
import { ThyFormDirective } from './form.directive';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { TemplateRef } from '@angular/core';
import { ThyTranslate } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './form.directive';
import * as ɵngcc2 from 'ngx-tethys/core';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/icon';

const _c0 = ["formGroup"];
function ThyFormGroupComponent_ng_container_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r5.feedbackSvgIconName);
} }
function ThyFormGroupComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementContainer(1, 3);
    ɵngcc0.ɵɵtemplate(2, ThyFormGroupComponent_ng_container_0_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    const _r3 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.contentTemplateRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.feedbackSvgIconName)("ngIfElse", _r3);
} }
const _c1 = function (a0, a1, a2) { return { "col-sm-2 col-form-label": a0, "label-required": a1, "pt-0": a2 }; };
function ThyFormGroupComponent_ng_template_1_label_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction3(2, _c1, ctx_r6.isHorizontal, ctx_r6.labelRequired, ctx_r6.labelPaddingTopClear));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r6.labelText);
} }
function ThyFormGroupComponent_ng_template_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r7.feedbackSvgIconName);
} }
function ThyFormGroupComponent_ng_template_1_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r8.tips);
} }
const _c2 = function (a0, a1) { return { "col-sm-10 col-form-control": a0, "position-relative": a1 }; };
function ThyFormGroupComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyFormGroupComponent_ng_template_1_label_0_Template, 2, 6, "label", 5);
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵprojection(2);
    ɵngcc0.ɵɵtemplate(3, ThyFormGroupComponent_ng_template_1_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(4, ThyFormGroupComponent_ng_template_1_span_4_Template, 2, 1, "span", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    const _r3 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵproperty("ngIf", !(!ctx_r2.isHorizontal && !ctx_r2.labelText));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(5, _c2, ctx_r2.isHorizontal, !ctx_r2.isHorizontal));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.feedbackSvgIconName)("ngIfElse", _r3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.tips);
} }
function ThyFormGroupComponent_ng_template_3_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span");
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMapInterpolate1("form-control-feedback ", ctx_r9.feedbackIcon, "");
} }
function ThyFormGroupComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyFormGroupComponent_ng_template_3_span_0_Template, 1, 3, "span", 10);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.feedbackIcon);
} }
const _c3 = ["*"];
const internalIconMap = {
    date: 'wtf wtf-schedule-o'
};
export class ThyFormGroupComponent {
    constructor(thyParentForm, thyTranslate) {
        this.thyParentForm = thyParentForm;
        this.thyTranslate = thyTranslate;
        this.labelRequired = false;
        this.labelPaddingTopClear = false;
        this._rowFill = false;
        this._isFormGroup = true;
        this.isHorizontal = true;
        this.hasFeedback = false;
    }
    set thyLabelText(value) {
        this.labelText = value;
    }
    set thyLabelTextTranslateKey(value) {
        if (value) {
            this.labelText = this.thyTranslate.instant(value);
        }
        else {
            this.labelText = '';
        }
    }
    set thyLabelRequired(value) {
        this.labelRequired = coerceBooleanProperty(value);
    }
    set thyLabelPaddingTopClear(value) {
        this.labelPaddingTopClear = coerceBooleanProperty(value);
    }
    set thyFeedbackIcon(value) {
        this.hasFeedback = true;
        if (internalIconMap[value]) {
            this.feedbackIcon = internalIconMap[value];
            this.feedbackSvgIconName = null;
        }
        else {
            this.feedbackSvgIconName = value;
            this.feedbackIcon = null;
        }
    }
    set thyTips(value) {
        this.tips = value;
    }
    set thyTipsTranslateKey(value) {
        this.tips = this.thyTranslate.instant(value);
    }
    set thyRowFill(value) {
        this._rowFill = coerceBooleanProperty(value);
    }
    ngOnInit() {
        this.isHorizontal = this.thyParentForm ? this.thyParentForm.isHorizontal : true;
    }
}
ThyFormGroupComponent.ɵfac = function ThyFormGroupComponent_Factory(t) { return new (t || ThyFormGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyFormDirective, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyTranslate)); };
ThyFormGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFormGroupComponent, selectors: [["thy-form-group"]], contentQueries: function ThyFormGroupComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, hostVars: 8, hostBindings: function ThyFormGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("row-fill", ctx._rowFill)("form-group", ctx._isFormGroup)("row", ctx.isHorizontal)("has-feedback", ctx.hasFeedback);
    } }, inputs: { thyLabelText: "thyLabelText", thyLabelTextTranslateKey: "thyLabelTextTranslateKey", thyLabelRequired: "thyLabelRequired", thyLabelPaddingTopClear: "thyLabelPaddingTopClear", thyFeedbackIcon: "thyFeedbackIcon", thyTips: "thyTips", thyTipsTranslateKey: "thyTipsTranslateKey", thyRowFill: "thyRowFill" }, ngContentSelectors: _c3, decls: 5, vars: 2, consts: [[4, "ngIf", "ngIfElse"], ["defaultTpl", ""], ["iconFont", ""], [3, "ngTemplateOutlet"], [1, "form-control-feedback", 3, "thyIconName"], ["class", "form-label ", 3, "ngClass", 4, "ngIf"], [3, "ngClass"], ["class", "form-text text-desc", 4, "ngIf"], [1, "form-label", 3, "ngClass"], [1, "form-text", "text-desc"], [3, "class", 4, "ngIf"]], template: function ThyFormGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThyFormGroupComponent_ng_container_0_Template, 3, 3, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, ThyFormGroupComponent_ng_template_1_Template, 5, 8, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(3, ThyFormGroupComponent_ng_template_3_Template, 1, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.contentTemplateRef)("ngIfElse", _r1);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet, ɵngcc4.ThyIconComponent, ɵngcc3.NgClass], encapsulation: 2, changeDetection: 0 });
ThyFormGroupComponent.ctorParameters = () => [
    { type: ThyFormDirective, decorators: [{ type: Optional }] },
    { type: ThyTranslate }
];
ThyFormGroupComponent.propDecorators = {
    _rowFill: [{ type: HostBinding, args: ['class.row-fill',] }],
    _isFormGroup: [{ type: HostBinding, args: ['class.form-group',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.row',] }],
    hasFeedback: [{ type: HostBinding, args: ['class.has-feedback',] }],
    thyLabelText: [{ type: Input }],
    thyLabelTextTranslateKey: [{ type: Input }],
    thyLabelRequired: [{ type: Input }],
    thyLabelPaddingTopClear: [{ type: Input }],
    thyFeedbackIcon: [{ type: Input }],
    thyTips: [{ type: Input }],
    thyTipsTranslateKey: [{ type: Input }],
    thyRowFill: [{ type: Input }],
    contentTemplateRef: [{ type: ContentChild, args: ['formGroup',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-form-group',
                template: "<ng-container *ngIf=\"contentTemplateRef; else defaultTpl\">\n  <ng-container [ngTemplateOutlet]=\"contentTemplateRef\"></ng-container>\n  <ng-container *ngIf=\"feedbackSvgIconName; else iconFont\">\n    <thy-icon class=\"form-control-feedback\" [thyIconName]=\"feedbackSvgIconName\"></thy-icon>\n  </ng-container>\n</ng-container>\n<ng-template #defaultTpl>\n  <label\n    *ngIf=\"!(!isHorizontal && !labelText)\"\n    class=\"form-label \"\n    [ngClass]=\"{\n      'col-sm-2 col-form-label': isHorizontal,\n      'label-required': labelRequired,\n      'pt-0': labelPaddingTopClear\n    }\"\n    >{{ labelText }}</label\n  >\n  <div [ngClass]=\"{ 'col-sm-10 col-form-control': isHorizontal, 'position-relative': !isHorizontal }\">\n    <ng-content></ng-content>\n    <ng-container *ngIf=\"feedbackSvgIconName; else iconFont\">\n      <thy-icon class=\"form-control-feedback\" [thyIconName]=\"feedbackSvgIconName\"></thy-icon>\n    </ng-container>\n    <span *ngIf=\"tips\" class=\"form-text text-desc\">{{ tips }}</span>\n  </div>\n</ng-template>\n\n<ng-template #iconFont>\n  <span *ngIf=\"feedbackIcon\" class=\"form-control-feedback {{ feedbackIcon }}\"></span>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.ThyFormDirective, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.ThyTranslate }]; }, { _rowFill: [{
            type: HostBinding,
            args: ['class.row-fill']
        }], _isFormGroup: [{
            type: HostBinding,
            args: ['class.form-group']
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.row']
        }], hasFeedback: [{
            type: HostBinding,
            args: ['class.has-feedback']
        }], thyLabelText: [{
            type: Input
        }], thyLabelTextTranslateKey: [{
            type: Input
        }], thyLabelRequired: [{
            type: Input
        }], thyLabelPaddingTopClear: [{
            type: Input
        }], thyFeedbackIcon: [{
            type: Input
        }], thyTips: [{
            type: Input
        }], thyTipsTranslateKey: [{
            type: Input
        }], thyRowFill: [{
            type: Input
        }], contentTemplateRef: [{
            type: ContentChild,
            args: ['formGroup']
        }] }); })();
