// ngx-tethys/esm2015/form/form-submit.directive.js
import { Directive, Output, HostListener, EventEmitter } from '@angular/core';
import { NgForm } from '@angular/forms';
import { ThyFormDirective } from './form.directive';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from './form.directive';
export class ThyFormSubmitDirective {
    constructor(ngForm, thyFormDirective) {
        this.ngForm = ngForm;
        this.thyFormDirective = thyFormDirective;
        this.thyFormSubmit = new EventEmitter();
    }
    ngOnInit() {
        this.thyFormDirective.onSubmitSuccess = ($event) => {
            this.thyFormSubmit.emit($event);
        };
    }
    onSubmit($event) {
        this.thyFormDirective.submit($event);
    }
}
ThyFormSubmitDirective.ɵfac = function ThyFormSubmitDirective_Factory(t) { return new (t || ThyFormSubmitDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgForm), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyFormDirective)); };
ThyFormSubmitDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyFormSubmitDirective, selectors: [["", "thyFormSubmit", ""], ["", "thy-form-submit", ""]], hostBindings: function ThyFormSubmitDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyFormSubmitDirective_click_HostBindingHandler($event) { return ctx.onSubmit($event); });
    } }, outputs: { thyFormSubmit: "thyFormSubmit" } });
ThyFormSubmitDirective.ctorParameters = () => [
    { type: NgForm },
    { type: ThyFormDirective }
];
ThyFormSubmitDirective.propDecorators = {
    thyFormSubmit: [{ type: Output }],
    onSubmit: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormSubmitDirective, [{
        type: Directive,
        args: [{
                selector: '[thyFormSubmit],[thy-form-submit]'
            }]
    }], function () { return [{ type: ɵngcc1.NgForm }, { type: ɵngcc2.ThyFormDirective }]; }, { thyFormSubmit: [{
            type: Output
        }], onSubmit: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
