// ngx-tethys/esm2015/form/form-group-label.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { ThyTranslate } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
export class ThyFormGroupLabelDirective {
    constructor(thyTranslate) {
        this.thyTranslate = thyTranslate;
        this.labelRequired = false;
        this._isFormGroupLabel = true;
    }
    set thyLabelText(value) {
        this.labelText = value;
    }
    set thyLabelTranslateKey(translateKey) {
        if (translateKey) {
            this.labelText = this.thyTranslate.instant(translateKey);
        }
    }
    set thyLabelRequired(value) {
        this.labelRequired = coerceBooleanProperty(value);
    }
}
ThyFormGroupLabelDirective.ɵfac = function ThyFormGroupLabelDirective_Factory(t) { return new (t || ThyFormGroupLabelDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate)); };
ThyFormGroupLabelDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyFormGroupLabelDirective, selectors: [["", "thyFormGroupLabel", ""]], hostVars: 4, hostBindings: function ThyFormGroupLabelDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("label-required", ctx.labelRequired)("col-form-label", ctx._isFormGroupLabel);
    } }, inputs: { thyLabelText: "thyLabelText", thyLabelTranslateKey: "thyLabelTranslateKey", thyLabelRequired: "thyLabelRequired" } });
ThyFormGroupLabelDirective.ctorParameters = () => [
    { type: ThyTranslate }
];
ThyFormGroupLabelDirective.propDecorators = {
    labelRequired: [{ type: HostBinding, args: ['class.label-required',] }],
    _isFormGroupLabel: [{ type: HostBinding, args: ['class.col-form-label',] }],
    thyLabelText: [{ type: Input }],
    thyLabelTranslateKey: [{ type: Input }],
    thyLabelRequired: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupLabelDirective, [{
        type: Directive,
        args: [{
                selector: '[thyFormGroupLabel]'
            }]
    }], function () { return [{ type: ɵngcc1.ThyTranslate }]; }, { labelRequired: [{
            type: HostBinding,
            args: ['class.label-required']
        }], _isFormGroupLabel: [{
            type: HostBinding,
            args: ['class.col-form-label']
        }], thyLabelText: [{
            type: Input
        }], thyLabelTranslateKey: [{
            type: Input
        }], thyLabelRequired: [{
            type: Input
        }] }); })();
