// ngx-tethys/esm2015/form/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThyFormDirective } from './form.directive';
import { ThyFormGroupComponent } from './form-group.component';
import { ThyFormGroupLabelDirective } from './form-group-label.directive';
import { ThyFormSubmitDirective } from './form-submit.directive';
import { ThyInputModule } from 'ngx-tethys/input';
import { ThyFormGroupFooterComponent } from './from-group-footer/form-group-footer.component';
import { ThyFormGroupErrorComponent } from './form-group-error/form-group-error.component';
import { ThyFormValidatorLoader } from './form-validator-loader';
import { THY_VALIDATOR_CONFIG, THY_FORM_CONFIG_PROVIDER } from './form.class';
import { ThyUniqueCheckValidator, ThyMaxDirective, ThyMinDirective, ThyConfirmValidatorDirective } from './validator/index';
import { ThyAlertModule } from 'ngx-tethys/alert';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyFormModule {
    static forRoot(config) {
        return {
            ngModule: ThyFormModule,
            providers: [
                {
                    provide: THY_VALIDATOR_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
ThyFormModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyFormModule });
ThyFormModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyFormModule_Factory(t) { return new (t || ThyFormModule)(); }, providers: [ThyFormValidatorLoader, THY_FORM_CONFIG_PROVIDER], imports: [[CommonModule, FormsModule, ThyInputModule, ThyAlertModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyFormModule, { declarations: function () { return [ThyFormDirective,
        ThyFormGroupComponent,
        ThyFormGroupLabelDirective,
        ThyFormSubmitDirective,
        ThyFormGroupFooterComponent,
        ThyUniqueCheckValidator,
        ThyFormGroupErrorComponent,
        ThyMinDirective,
        ThyMaxDirective,
        ThyConfirmValidatorDirective]; }, imports: function () { return [CommonModule, FormsModule, ThyInputModule, ThyAlertModule, ThyIconModule]; }, exports: function () { return [ThyFormDirective,
        ThyFormGroupComponent,
        ThyFormGroupLabelDirective,
        ThyFormSubmitDirective,
        ThyFormGroupFooterComponent,
        ThyUniqueCheckValidator,
        ThyFormGroupErrorComponent,
        ThyMinDirective,
        ThyMaxDirective,
        ThyConfirmValidatorDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ThyInputModule, ThyAlertModule, ThyIconModule],
                declarations: [
                    ThyFormDirective,
                    ThyFormGroupComponent,
                    ThyFormGroupLabelDirective,
                    ThyFormSubmitDirective,
                    ThyFormGroupFooterComponent,
                    ThyUniqueCheckValidator,
                    ThyFormGroupErrorComponent,
                    ThyMinDirective,
                    ThyMaxDirective,
                    ThyConfirmValidatorDirective
                ],
                exports: [
                    ThyFormDirective,
                    ThyFormGroupComponent,
                    ThyFormGroupLabelDirective,
                    ThyFormSubmitDirective,
                    ThyFormGroupFooterComponent,
                    ThyUniqueCheckValidator,
                    ThyFormGroupErrorComponent,
                    ThyMinDirective,
                    ThyMaxDirective,
                    ThyConfirmValidatorDirective
                ],
                providers: [ThyFormValidatorLoader, THY_FORM_CONFIG_PROVIDER]
            }]
    }], null, null); })();
