// ngx-tethys/esm2015/avatar/avatar.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyAvatarComponent } from './avatar.component';
import { AvatarPipes } from './avatar.pipe';
import { ThyAvatarService, ThyDefaultAvatarService } from './avatar.service';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './avatar.pipe';
export class ThyAvatarModule {
}
ThyAvatarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAvatarModule });
ThyAvatarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAvatarModule_Factory(t) { return new (t || ThyAvatarModule)(); }, providers: [
        {
            provide: ThyAvatarService,
            useClass: ThyDefaultAvatarService
        }
    ], imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAvatarModule, { declarations: function () { return [ThyAvatarComponent, ɵngcc1.AvatarShortNamePipe, ɵngcc1.AvatarBgColorPipe, ɵngcc1.AvatarSrcPipe]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyAvatarComponent, ɵngcc1.AvatarShortNamePipe, ɵngcc1.AvatarBgColorPipe, ɵngcc1.AvatarSrcPipe]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAvatarModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyAvatarComponent, AvatarPipes],
                imports: [CommonModule, ThyIconModule],
                providers: [
                    {
                        provide: ThyAvatarService,
                        useClass: ThyDefaultAvatarService
                    }
                ],
                exports: [ThyAvatarComponent, AvatarPipes]
            }]
    }], null, null); })();
