// ngx-tethys/esm2015/avatar/avatar.pipe.js
import { Pipe } from '@angular/core';
import { ThyAvatarService } from './avatar.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './avatar.service';
export class AvatarShortNamePipe {
    transform(name) {
        if (!name) {
            return;
        }
        name = name.trim();
        if (/^[\u4e00-\u9fa5]+$/.test(name)) {
            if (name.length > 2) {
                return name.substr(name.length - 2, 2);
            }
        }
        if (/^[a-zA-Z\/ ]+$/.test(name)) {
            if (name.indexOf(' ') > 0) {
                const ens = name.split(' ');
                return (ens[0].substr(0, 1) + ens[1].substr(0, 1)).toUpperCase();
            }
        }
        if (name.length > 2) {
            return name.substr(0, 2).toUpperCase();
        }
        return name.toUpperCase();
    }
}
AvatarShortNamePipe.ɵfac = function AvatarShortNamePipe_Factory(t) { return new (t || AvatarShortNamePipe)(); };
AvatarShortNamePipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "avatarShortName", type: AvatarShortNamePipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvatarShortNamePipe, [{
        type: Pipe,
        args: [{ name: 'avatarShortName' }]
    }], null, null); })();
export class AvatarBgColorPipe {
    transform(name) {
        if (!name) {
            return;
        }
        const colors = [
            '#56abfb',
            '#5dcfff',
            '#84e17e',
            '#73d897',
            '#ff9f73',
            '#fa8888',
            '#fb7fb7',
            '#9a7ef4',
            '#868af6'
        ];
        const nameArray = name.split('');
        const code = name && name.length > 0
            ? nameArray.reduce(function (result, item) {
                result.value += item.charCodeAt(0);
                return result;
            }, { value: 0 }).value
            : 0;
        return {
            'background-color': colors[code % 9]
        };
    }
}
AvatarBgColorPipe.ɵfac = function AvatarBgColorPipe_Factory(t) { return new (t || AvatarBgColorPipe)(); };
AvatarBgColorPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "avatarBgColor", type: AvatarBgColorPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvatarBgColorPipe, [{
        type: Pipe,
        args: [{ name: 'avatarBgColor' }]
    }], null, null); })();
export class AvatarSrcPipe {
    constructor(thyAvatarService) {
        this.thyAvatarService = thyAvatarService;
    }
    transform(src, size) {
        return this.thyAvatarService.avatarSrcTransform(src, size);
    }
}
AvatarSrcPipe.ɵfac = function AvatarSrcPipe_Factory(t) { return new (t || AvatarSrcPipe)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyAvatarService)); };
AvatarSrcPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "thyAvatarSrc", type: AvatarSrcPipe, pure: true });
AvatarSrcPipe.ctorParameters = () => [
    { type: ThyAvatarService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AvatarSrcPipe, [{
        type: Pipe,
        args: [{ name: 'thyAvatarSrc' }]
    }], function () { return [{ type: ɵngcc1.ThyAvatarService }]; }, null); })();
export const AvatarPipes = [AvatarShortNamePipe, AvatarBgColorPipe, AvatarSrcPipe];
