// ngx-tethys/esm2015/affix/affix.module.js
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyAffixComponent } from './affix.component';
import * as ɵngcc0 from '@angular/core';
export class ThyAffixModule {
}
ThyAffixModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAffixModule });
ThyAffixModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAffixModule_Factory(t) { return new (t || ThyAffixModule)(); }, imports: [[CommonModule, PlatformModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAffixModule, { declarations: function () { return [ThyAffixComponent]; }, imports: function () { return [CommonModule, PlatformModule]; }, exports: function () { return [ThyAffixComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAffixModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyAffixComponent],
                exports: [ThyAffixComponent],
                imports: [CommonModule, PlatformModule]
            }]
    }], null, null); })();
