// ngx-tethys/esm2015/cascader/cascader-li.component.js
import { Component, Input, HostBinding, ChangeDetectionStrategy, ViewEncapsulation } from '@angular/core';
import { UpdateHostClassService } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["thy-cascader-option", ""];
function ThyCascaderOptionComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 1);
} }
export class ThyCascaderOptionComponent {
    constructor() {
        this.item = true;
        this.active = false;
    }
    get disabled() {
        return this.option.disabled;
    }
    get expand() {
        return this.option && !this.option.isLeaf;
    }
    getOptionLabel() {
        return this.option ? this.option[this.thyLabelProperty || 'label'] : '';
    }
    ngOnInit() { }
}
ThyCascaderOptionComponent.ɵfac = function ThyCascaderOptionComponent_Factory(t) { return new (t || ThyCascaderOptionComponent)(); };
ThyCascaderOptionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyCascaderOptionComponent, selectors: [["", "thy-cascader-option", ""]], hostVars: 8, hostBindings: function ThyCascaderOptionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-cascader-menu-item", ctx.item)("thy-cascader-menu-item-active", ctx.active)("thy-cascader-menu-item-disabled", ctx.disabled)("thy-cascader-menu-item-expand", ctx.expand);
    } }, inputs: { active: "active", option: "option", thyLabelProperty: "thyLabelProperty" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], attrs: _c0, decls: 3, vars: 2, consts: [["class", "wtf wtf-angle-right thy-cascader-menu-item-expand-icon", 4, "ngIf"], [1, "wtf", "wtf-angle-right", "thy-cascader-menu-item-expand-icon"]], template: function ThyCascaderOptionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(2, ThyCascaderOptionComponent_span_2_Template, 1, 0, "span", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.getOptionLabel(), "\n");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.option.isLeaf || ctx.option.children && ctx.option.children.length || ctx.option.loading);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2, changeDetection: 0 });
ThyCascaderOptionComponent.ctorParameters = () => [];
ThyCascaderOptionComponent.propDecorators = {
    option: [{ type: Input }],
    item: [{ type: HostBinding, args: ['class.thy-cascader-menu-item',] }],
    active: [{ type: HostBinding, args: ['class.thy-cascader-menu-item-active',] }, { type: Input }],
    disabled: [{ type: HostBinding, args: ['class.thy-cascader-menu-item-disabled',] }],
    expand: [{ type: HostBinding, args: ['class.thy-cascader-menu-item-expand',] }],
    thyLabelProperty: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCascaderOptionComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: '[thy-cascader-option]',
                template: "<ng-container>\n    {{ getOptionLabel() }}\n</ng-container>\n<span *ngIf=\"!option.isLeaf || option.children && option.children.length || option.loading\" class=\"wtf wtf-angle-right thy-cascader-menu-item-expand-icon\">\n</span>",
                providers: [UpdateHostClassService]
            }]
    }], function () { return []; }, { item: [{
            type: HostBinding,
            args: ['class.thy-cascader-menu-item']
        }], active: [{
            type: HostBinding,
            args: ['class.thy-cascader-menu-item-active']
        }, {
            type: Input
        }], disabled: [{
            type: HostBinding,
            args: ['class.thy-cascader-menu-item-disabled']
        }], expand: [{
            type: HostBinding,
            args: ['class.thy-cascader-menu-item-expand']
        }], option: [{
            type: Input
        }], thyLabelProperty: [{
            type: Input
        }] }); })();
