// ngx-tethys/esm2015/cascader/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { OverlayModule } from '@angular/cdk/overlay';
import { ThyInputModule } from 'ngx-tethys/input';
import { ThyCascaderComponent } from './cascader.component';
import { ThyCascaderOptionComponent } from './cascader-li.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyCascaderModule {
}
ThyCascaderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCascaderModule });
ThyCascaderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCascaderModule_Factory(t) { return new (t || ThyCascaderModule)(); }, imports: [[CommonModule, FormsModule, OverlayModule, ThyInputModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCascaderModule, { declarations: function () { return [ThyCascaderComponent, ThyCascaderOptionComponent]; }, imports: function () { return [CommonModule, FormsModule, OverlayModule, ThyInputModule, ThyIconModule]; }, exports: function () { return [ThyCascaderComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCascaderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, OverlayModule, ThyInputModule, ThyIconModule],
                declarations: [ThyCascaderComponent, ThyCascaderOptionComponent],
                exports: [ThyCascaderComponent]
            }]
    }], null, null); })();
