// ngx-tethys/esm2015/menu/menu.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyMenuComponent } from './menu.component';
import { ThyMenuItemComponent } from './item/menu-item.component';
import { ThyMenuGroupComponent } from './group/menu-group.component';
import { ThyMenuItemNameComponent } from './item/name/menu-item-name.component';
import { ThyMenuItemIconComponent } from './item/icon/menu-item-icon.component';
import { ThyMenuItemActionComponent } from './item/action/menu-item-action.component';
import { ThyMenuDividerComponent } from './divider/menu-divider.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyPopoverModule } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
export class ThyMenuModule {
}
ThyMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyMenuModule });
ThyMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyMenuModule_Factory(t) { return new (t || ThyMenuModule)(); }, imports: [[CommonModule, ThyIconModule, ThyPopoverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyMenuModule, { declarations: function () { return [ThyMenuComponent,
        ThyMenuGroupComponent,
        ThyMenuItemComponent,
        ThyMenuItemNameComponent,
        ThyMenuItemIconComponent,
        ThyMenuItemActionComponent,
        ThyMenuDividerComponent]; }, imports: function () { return [CommonModule, ThyIconModule, ThyPopoverModule]; }, exports: function () { return [ThyMenuComponent,
        ThyMenuGroupComponent,
        ThyMenuItemComponent,
        ThyMenuItemNameComponent,
        ThyMenuItemIconComponent,
        ThyMenuItemActionComponent,
        ThyMenuDividerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyMenuComponent,
                    ThyMenuGroupComponent,
                    ThyMenuItemComponent,
                    ThyMenuItemNameComponent,
                    ThyMenuItemIconComponent,
                    ThyMenuItemActionComponent,
                    ThyMenuDividerComponent
                ],
                imports: [CommonModule, ThyIconModule, ThyPopoverModule],
                exports: [
                    ThyMenuComponent,
                    ThyMenuGroupComponent,
                    ThyMenuItemComponent,
                    ThyMenuItemNameComponent,
                    ThyMenuItemIconComponent,
                    ThyMenuItemActionComponent,
                    ThyMenuDividerComponent
                ]
            }]
    }], null, null); })();
