// ngx-tethys/esm2015/menu/menu.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyMenuComponent {
    constructor() {
        this.isThyMenu = true;
    }
    ngOnInit() { }
}
ThyMenuComponent.ɵfac = function ThyMenuComponent_Factory(t) { return new (t || ThyMenuComponent)(); };
ThyMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuComponent, selectors: [["thy-menu"], ["", "thy-menu", ""], ["", "thyMenu", ""]], hostVars: 2, hostBindings: function ThyMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu", ctx.isThyMenu);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuComponent.ctorParameters = () => [];
ThyMenuComponent.propDecorators = {
    isThyMenu: [{ type: HostBinding, args: ['class.thy-menu',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu,[thy-menu],[thyMenu]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenu: [{
            type: HostBinding,
            args: ['class.thy-menu']
        }] }); })();
