// ngx-tethys/esm2015/guider/guider.directive.js
import { ThyGuiderManager } from './guider-manager';
import { Directive, ElementRef, Input, NgZone } from '@angular/core';
import { take } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './guider-manager';
export class ThyGuiderTargetDirective {
    constructor(guiderManager, el, ngZone) {
        this.guiderManager = guiderManager;
        this.el = el;
        this.ngZone = ngZone;
    }
    ngOnInit() {
        this.guiderManager.addStepTarget(this.target, this.el.nativeElement);
        this.ngZone.onStable.pipe(take(1)).subscribe(() => {
            const { key, guiderRef } = this.guiderManager.getActive();
            if (key === this.target) {
                const index = guiderRef.steps.findIndex(step => step.key === this.target);
                this.ngZone.run(() => {
                    guiderRef.active(index);
                });
            }
        });
    }
    ngAfterViewInit() { }
    ngOnDestroy() {
        const { guiderRef } = this.guiderManager.getActive();
        if (guiderRef) {
            guiderRef.close();
        }
        this.guiderManager.removeStepTarget(this.target);
    }
}
ThyGuiderTargetDirective.ɵfac = function ThyGuiderTargetDirective_Factory(t) { return new (t || ThyGuiderTargetDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyGuiderManager), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyGuiderTargetDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyGuiderTargetDirective, selectors: [["", "thyGuiderTarget", ""]], inputs: { target: ["thyGuiderTarget", "target"] }, exportAs: ["thyGuiderTarget"] });
ThyGuiderTargetDirective.ctorParameters = () => [
    { type: ThyGuiderManager },
    { type: ElementRef },
    { type: NgZone }
];
ThyGuiderTargetDirective.propDecorators = {
    target: [{ type: Input, args: ['thyGuiderTarget',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyGuiderTargetDirective, [{
        type: Directive,
        args: [{ selector: '[thyGuiderTarget]', exportAs: 'thyGuiderTarget' }]
    }], function () { return [{ type: ɵngcc1.ThyGuiderManager }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }]; }, { target: [{
            type: Input,
            args: ['thyGuiderTarget']
        }] }); })();
