// ngx-tethys/esm2015/guider/guider-ref.js
import { Router } from '@angular/router';
import { helpers } from 'ngx-tethys/util';
import { DOCUMENT } from '@angular/common';
import { ThyPopover } from 'ngx-tethys/popover';
import { ThyGuiderManager } from './guider-manager';
import { ThyGuiderStepRef } from './guider-step-ref';
import { ReplaySubject, Subject } from 'rxjs';
import { ThyGuiderConfig } from './guider.class';
import { Inject, NgZone, RendererFactory2 } from '@angular/core';
import { Overlay } from '@angular/cdk/overlay';
export class ThyGuiderRef {
    constructor(config, rendererFactory, popover, router, guiderManager, ngZone, overlay, document) {
        this.config = config;
        this.rendererFactory = rendererFactory;
        this.popover = popover;
        this.router = router;
        this.guiderManager = guiderManager;
        this.ngZone = ngZone;
        this.overlay = overlay;
        this.document = document;
        this.stepChange$ = new ReplaySubject();
        this.guiderEnded$ = new Subject();
        this.closed$ = new Subject();
        this.targetClicked$ = new Subject();
        this.stepsRef = config.steps.map((step, index) => {
            return new ThyGuiderStepRef(step, index, this.rendererFactory, this.popover, this.guiderManager, this.overlay, this.document);
        });
        this.steps = config.steps;
    }
    stepChange() {
        return this.stepChange$.asObservable();
    }
    ended() {
        return this.guiderEnded$;
    }
    closed() {
        return this.closed$;
    }
    targetClicked() {
        return this.targetClicked$;
    }
    start(startWith) {
        this.to(startWith);
        return this.stepChange();
    }
    next() {
        if (this.currentStepIndex + 1 > this.steps.length) {
            return;
        }
        this.to(this.currentStepIndex + 1);
    }
    previous() {
        if (this.currentStepIndex - 1 < 0) {
            return;
        }
        this.to(this.currentStepIndex - 1);
    }
    active(indexOrKey) {
        if (helpers.isNumber(indexOrKey)) {
            this.to(indexOrKey);
            return;
        }
        if (helpers.isString(indexOrKey)) {
            const index = this.steps.findIndex(step => step.key === indexOrKey);
            this.to(index);
            return;
        }
    }
    to(index) {
        this.removeExistedStep();
        if (!helpers.isNumber(index) || index >= this.steps.length || index < 0 || Number.isNaN(index)) {
            index = 0;
        }
        this.currentStep = this.steps[index];
        this.currentStepIndex = index;
        // update guiderManager
        this.guiderManager.updateActive(this.currentStep.key, this);
        if (this.currentStep.route && this.currentStep.route !== this.router.url) {
            this.ngZone.run(() => {
                this.router.navigateByUrl(this.currentStep.route);
            });
            return;
        }
        setTimeout(() => {
            this.drawStep();
            this.notifyStepClicked();
        }, 0);
    }
    close() {
        var _a;
        this.removeManagerActiveKey();
        (_a = this.stepsRef[this.currentStepIndex]) === null || _a === void 0 ? void 0 : _a.dispose();
        this.closed$.next(this.currentStep);
    }
    end() {
        this.close();
        this.guiderEnded$.next(this.currentStep);
        this.notifyGuiderIsFinished();
    }
    removeManagerActiveKey() {
        const activeKey = this.guiderManager.getActive().key;
        if (activeKey && this.steps.some(step => step.key === activeKey)) {
            this.guiderManager.updateActive('', undefined);
        }
    }
    notifyStepClicked() {
        this.stepChange$.next(this.currentStep);
    }
    drawStep() {
        this.removeExistedStep();
        this.stepsRef[this.currentStepIndex].show(this);
    }
    notifyGuiderIsFinished() {
        this.stepChange$.complete();
        // this.targetClicked().unsubscribe();
        this.currentStepIndex = 0;
    }
    removeExistedStep() {
        var _a;
        (_a = this.stepsRef[this.currentStepIndex]) === null || _a === void 0 ? void 0 : _a.dispose();
    }
}
ThyGuiderRef.ctorParameters = () => [
    { type: ThyGuiderConfig },
    { type: RendererFactory2 },
    { type: ThyPopover },
    { type: Router },
    { type: ThyGuiderManager },
    { type: NgZone },
    { type: Overlay },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
