// ngx-tethys/esm2015/guider/guider.service.js
import { Router } from '@angular/router';
import { helpers } from 'ngx-tethys/util';
import { DOCUMENT } from '@angular/common';
import { ThyGuiderRef } from './guider-ref';
import { Inject, NgZone } from '@angular/core';
import { ThyPopover } from 'ngx-tethys/popover';
import { ThyGuiderManager } from './guider-manager';
import { Injectable, RendererFactory2 } from '@angular/core';
import { defaultGuiderPositionConfig } from './guider.class';
import { Overlay } from '@angular/cdk/overlay';
import * as i0 from "@angular/core";
import * as i1 from "ngx-tethys/popover";
import * as i2 from "@angular/router";
import * as i3 from "./guider-manager";
import * as i4 from "@angular/cdk/overlay";
import * as i5 from "@angular/common";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
import * as ɵngcc2 from '@angular/router';
import * as ɵngcc3 from './guider-manager';
import * as ɵngcc4 from '@angular/cdk/overlay';
export class ThyGuider {
    constructor(rendererFactory, popover, router, guiderManager, ngZone, overlay, document) {
        this.rendererFactory = rendererFactory;
        this.popover = popover;
        this.router = router;
        this.guiderManager = guiderManager;
        this.ngZone = ngZone;
        this.overlay = overlay;
        this.document = document;
        this.guiderRefs = [];
    }
    create(config) {
        if (!config || !(config === null || config === void 0 ? void 0 : config.steps) || !helpers.isArray(config === null || config === void 0 ? void 0 : config.steps)) {
            throw new Error(`'config.steps' must be an array of length greater than 0`);
        }
        const normalizeConfig = this.normalizeConfig(config);
        this.guiderRef = new ThyGuiderRef(normalizeConfig, this.rendererFactory, this.popover, this.router, this.guiderManager, this.ngZone, this.overlay, this.document);
        this.guiderRef.closed().subscribe(() => {
            const index = this.guiderRefs.findIndex(guiderRef => guiderRef === this.guiderRef);
            this.guiderRefs.splice(index, 1);
        });
        this.guiderRefs.push(this.guiderRef);
        return this.guiderRef;
    }
    normalizeConfig(config) {
        const normalizeConfig = Object.assign({}, defaultGuiderPositionConfig, config);
        normalizeConfig.steps = normalizeConfig.steps.map(step => {
            return this.normalizeStep(step, normalizeConfig);
        });
        return normalizeConfig;
    }
    normalizeStep(step, config) {
        const tempStep = Object.assign({
            hintPlacement: config.hintPlacement,
            hintOffset: config.hintOffset,
            pointOffset: config.pointOffset
        }, step);
        return tempStep;
    }
    close() {
        if (this.guiderRefs.length > 0) {
            const lasGuiderRef = this.guiderRefs[this.guiderRefs.length - 1];
            if (lasGuiderRef) {
                lasGuiderRef.close();
                this.guiderRefs.pop();
            }
        }
    }
}
ThyGuider.ɵfac = function ThyGuider_Factory(t) { return new (t || ThyGuider)(ɵngcc0.ɵɵinject(ɵngcc0.RendererFactory2), ɵngcc0.ɵɵinject(ɵngcc1.ThyPopover), ɵngcc0.ɵɵinject(ɵngcc2.Router), ɵngcc0.ɵɵinject(ɵngcc3.ThyGuiderManager), ɵngcc0.ɵɵinject(ɵngcc0.NgZone), ɵngcc0.ɵɵinject(ɵngcc4.Overlay), ɵngcc0.ɵɵinject(DOCUMENT)); };
ThyGuider.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyGuider_Factory() { return new ThyGuider(i0.ɵɵinject(i0.RendererFactory2), i0.ɵɵinject(i1.ThyPopover), i0.ɵɵinject(i2.Router), i0.ɵɵinject(i3.ThyGuiderManager), i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(i4.Overlay), i0.ɵɵinject(i5.DOCUMENT)); }, token: ThyGuider, providedIn: "root" });
ThyGuider.ctorParameters = () => [
    { type: RendererFactory2 },
    { type: ThyPopover },
    { type: Router },
    { type: ThyGuiderManager },
    { type: NgZone },
    { type: Overlay },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyGuider, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc0.RendererFactory2 }, { type: ɵngcc1.ThyPopover }, { type: ɵngcc2.Router }, { type: ɵngcc3.ThyGuiderManager }, { type: ɵngcc0.NgZone }, { type: ɵngcc4.Overlay }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, null); })();
