// ngx-tethys/esm2015/guider/guider-manager.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export class ThyGuiderManager {
    constructor() {
        this.targetListMap = {};
    }
    updateActive(key, guiderRef) {
        this.activeStepKey = key;
        this.thyGuiderRef = guiderRef;
    }
    addStepTarget(key, el) {
        this.targetListMap[key] = el;
    }
    removeStepTarget(key) {
        delete this.targetListMap[key];
    }
    getActiveTarget(key) {
        return this.targetListMap[key];
    }
    getActive() {
        return {
            key: this.activeStepKey,
            guiderRef: this.thyGuiderRef
        };
    }
}
ThyGuiderManager.ɵfac = function ThyGuiderManager_Factory(t) { return new (t || ThyGuiderManager)(); };
ThyGuiderManager.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyGuiderManager_Factory() { return new ThyGuiderManager(); }, token: ThyGuiderManager, providedIn: "root" });
ThyGuiderManager.ctorParameters = () => [];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyGuiderManager, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
