// ngx-tethys/esm2015/guider/guider.module.js
import { NgModule } from '@angular/core';
import { ThyGuider } from './guider.service';
import { CommonModule } from '@angular/common';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyPopoverModule } from 'ngx-tethys/popover';
import { ThyGuiderTargetDirective } from './guider.directive';
import { ThyGuiderHintComponent } from './guider-hint/guider-hint.component';
import * as ɵngcc0 from '@angular/core';
export class ThyGuiderModule {
}
ThyGuiderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyGuiderModule });
ThyGuiderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyGuiderModule_Factory(t) { return new (t || ThyGuiderModule)(); }, providers: [ThyGuider], imports: [[CommonModule, ThySharedModule, ThyIconModule, ThyButtonModule, ThyPopoverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyGuiderModule, { declarations: function () { return [ThyGuiderHintComponent, ThyGuiderTargetDirective]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule, ThyButtonModule, ThyPopoverModule]; }, exports: function () { return [ThyGuiderHintComponent, ThyGuiderTargetDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyGuiderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ThySharedModule, ThyIconModule, ThyButtonModule, ThyPopoverModule],
                declarations: [ThyGuiderHintComponent, ThyGuiderTargetDirective],
                exports: [ThyGuiderHintComponent, ThyGuiderTargetDirective],
                entryComponents: [ThyGuiderHintComponent],
                providers: [ThyGuider]
            }]
    }], null, null); })();
