// ngx-tethys/esm2015/shared/select/module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThySelectControlComponent } from './select-control/select-control.component';
import { FormsModule } from '@angular/forms';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThySelectCommonModule {
}
ThySelectCommonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySelectCommonModule });
ThySelectCommonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySelectCommonModule_Factory(t) { return new (t || ThySelectCommonModule)(); }, imports: [[CommonModule, FormsModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySelectCommonModule, { declarations: function () { return [ThySelectControlComponent]; }, imports: function () { return [CommonModule, FormsModule, ThyIconModule]; }, exports: function () { return [ThySelectControlComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySelectCommonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ThyIconModule],
                exports: [ThySelectControlComponent],
                declarations: [ThySelectControlComponent]
            }]
    }], null, null); })();
