// ngx-tethys/esm2015/form/from-group-footer/form-group-footer.component.js
import { Component, HostBinding, ViewEncapsulation, Optional, Input, Inject, ContentChild, TemplateRef } from '@angular/core';
import { ThyFormDirective } from '../form.directive';
import { THY_FORM_CONFIG } from '../form.class';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../form.directive';
import * as ɵngcc2 from '@angular/common';

const _c0 = ["description"];
function ThyFormGroupFooterComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementContainer(1, 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.description);
} }
const _c1 = ["*"];
export class ThyFormGroupFooterComponent {
    constructor(thyParentForm, defaultConfig) {
        this.thyParentForm = thyParentForm;
        this.defaultConfig = defaultConfig;
        this._isFormGroup = true;
        this.isHorizontal = true;
        this.footerClassMap = {};
    }
    get align() {
        return !!this.thyAlign ? this.thyAlign : this.defaultConfig.footerAlign;
    }
    ngOnInit() {
        if (this.thyParentForm) {
            this.isHorizontal = this.thyParentForm.isHorizontal;
        }
        this.setFooterClassMap();
    }
    setFooterClassMap() {
        this.footerClassMap = {
            'form-group-footer': true,
            'col-sm-10 offset-sm-2 col-form-control': this.isHorizontal,
            [`form-group-footer-align-${this.align}`]: true
        };
    }
}
ThyFormGroupFooterComponent.ɵfac = function ThyFormGroupFooterComponent_Factory(t) { return new (t || ThyFormGroupFooterComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyFormDirective, 8), ɵngcc0.ɵɵdirectiveInject(THY_FORM_CONFIG)); };
ThyFormGroupFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFormGroupFooterComponent, selectors: [["thy-form-group-footer"]], contentQueries: function ThyFormGroupFooterComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.description = _t.first);
    } }, hostVars: 4, hostBindings: function ThyFormGroupFooterComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("form-group", ctx._isFormGroup)("row", ctx.isHorizontal);
    } }, inputs: { thyAlign: "thyAlign" }, ngContentSelectors: _c1, decls: 4, vars: 2, consts: [[3, "ngClass"], [1, "btn-pair", "form-footer-actions"], ["class", "form-footer-description", 4, "ngIf"], [1, "form-footer-description"], [3, "ngTemplateOutlet"]], template: function ThyFormGroupFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ThyFormGroupFooterComponent_div_3_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ctx.footerClassMap);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.description);
    } }, directives: [ɵngcc2.NgClass, ɵngcc2.NgIf, ɵngcc2.NgTemplateOutlet], encapsulation: 2 });
ThyFormGroupFooterComponent.ctorParameters = () => [
    { type: ThyFormDirective, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Inject, args: [THY_FORM_CONFIG,] }] }
];
ThyFormGroupFooterComponent.propDecorators = {
    _isFormGroup: [{ type: HostBinding, args: ['class.form-group',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.row',] }],
    description: [{ type: ContentChild, args: ['description',] }],
    thyAlign: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupFooterComponent, [{
        type: Component,
        args: [{
                selector: 'thy-form-group-footer',
                template: "<div [ngClass]=\"footerClassMap\">\n  <div class=\"btn-pair form-footer-actions\">\n    <ng-content></ng-content>\n  </div>\n\n  <div class=\"form-footer-description\" *ngIf=\"description\">\n    <ng-container [ngTemplateOutlet]=\"description\"></ng-container>\n  </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc1.ThyFormDirective, decorators: [{
                type: Optional
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_FORM_CONFIG]
            }] }]; }, { _isFormGroup: [{
            type: HostBinding,
            args: ['class.form-group']
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.row']
        }], description: [{
            type: ContentChild,
            args: ['description']
        }], thyAlign: [{
            type: Input
        }] }); })();
