// ngx-tethys/esm2015/mention/suggestions/suggestions.component.js
import { Component, HostBinding, ElementRef, NgZone } from '@angular/core';
import { Subject, Observable, of } from 'rxjs';
import { debounceTime, switchMap, catchError, take } from 'rxjs/operators';
import { ThyPopoverRef } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
import * as ɵngcc2 from 'ngx-tethys/loading';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/list';
import * as ɵngcc5 from 'ngx-tethys/shared';

function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_container_1_1_ng_template_0_Template(rf, ctx) { }
function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_container_1_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_container_1_1_ng_template_0_Template, 0, 0, "ng-template");
} }
const _c0 = function (a0) { return { $implicit: a0 }; };
function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_container_1_1_Template, 1, 0, undefined, 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r6 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r6.mention == null ? null : ctx_r6.mention.displayTemplateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c0, item_r5));
} }
function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r5["name"] || "", " ");
} }
function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "thy-list-option", 6);
    ɵngcc0.ɵɵtemplate(1, ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_container_1_Template, 2, 4, "ng-container", 7);
    ɵngcc0.ɵɵtemplate(2, ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_ng_template_2_Template, 1, 1, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    const _r7 = ɵngcc0.ɵɵreference(3);
    const ctx_r4 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("thyValue", item_r5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.mention == null ? null : ctx_r4.mention.displayTemplateRef)("ngIfElse", _r7);
} }
function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-selection-list", 4);
    ɵngcc0.ɵɵlistener("thySelectionChange", function ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_Template_thy_selection_list_thySelectionChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(2); return ctx_r13.selectionChange($event); });
    ɵngcc0.ɵɵtemplate(1, ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_thy_list_option_1_Template, 4, 3, "thy-list-option", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyScrollContainer", ctx_r3.elementRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.data);
} }
function ThyMentionSuggestionsComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyMentionSuggestionsComponent_ng_container_1_thy_selection_list_1_Template, 2, 2, "thy-selection-list", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    const _r1 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.data == null ? null : ctx_r0.data.length) > 0)("ngIfElse", _r1);
} }
function ThyMentionSuggestionsComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.mention == null ? null : ctx_r2.mention.emptyText, " ");
} }
export class ThyMentionSuggestionsComponent {
    constructor(elementRef, ngZone, popoverRef) {
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.popoverRef = popoverRef;
        this.suggestionSelect$ = new Subject();
        this.debounce = 150;
        this.loadingDone = true;
        this.search$ = new Subject();
        this.suggestionsClass = true;
        this.search$
            .pipe(switchMap(query => {
            const data = this.mention.search(query.term, this.mention.data);
            if (data instanceof Observable) {
                this.loadingDone = false;
                return data.pipe(debounceTime(this.debounce));
            }
            else {
                return of(data);
            }
        }), catchError(() => {
            this.loadingDone = false;
            return [];
        }))
            .subscribe(data => {
            this.loadingDone = true;
            this.data = data || [];
            if (this.popoverRef) {
                if (this.mention.autoClose && this.data.length === 0) {
                    this.popoverRef.close();
                }
                this.ngZone.onStable.pipe(take(1)).subscribe(() => {
                    this.popoverRef.updatePosition();
                });
            }
        });
    }
    ngOnInit() {
        if (this.mention.popoverClass) {
            this.elementRef.nativeElement.classList.add(this.mention.popoverClass);
        }
    }
    search(query) {
        this.search$.next(query);
    }
    select(item, event) {
        this.suggestionSelect$.next({
            event,
            item
        });
    }
    selectionChange(event) {
        this.select(event.value, event.event);
    }
    ngOnDestroy() {
        this.search$.next();
        this.search$.complete();
    }
}
ThyMentionSuggestionsComponent.ɵfac = function ThyMentionSuggestionsComponent_Factory(t) { return new (t || ThyMentionSuggestionsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopoverRef)); };
ThyMentionSuggestionsComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMentionSuggestionsComponent, selectors: [["thy-mention-suggestions"]], hostVars: 2, hostBindings: function ThyMentionSuggestionsComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-mention-suggestions", ctx.suggestionsClass);
    } }, decls: 4, vars: 2, consts: [[3, "thyDone"], [4, "ngIf"], ["empty", ""], ["thyBindKeyEventContainer", "body", "thyMultiple", "false", "thyAutoActiveFirstItem", "true", "thySpaceKeyEnabled", "false", 3, "thyScrollContainer", "thySelectionChange", 4, "ngIf", "ngIfElse"], ["thyBindKeyEventContainer", "body", "thyMultiple", "false", "thyAutoActiveFirstItem", "true", "thySpaceKeyEnabled", "false", 3, "thyScrollContainer", "thySelectionChange"], [3, "thyValue", 4, "ngFor", "ngForOf"], [3, "thyValue"], [4, "ngIf", "ngIfElse"], ["default", ""], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "text-desc", "p-3"]], template: function ThyMentionSuggestionsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "thy-loading", 0);
        ɵngcc0.ɵɵtemplate(1, ThyMentionSuggestionsComponent_ng_container_1_Template, 2, 2, "ng-container", 1);
        ɵngcc0.ɵɵtemplate(2, ThyMentionSuggestionsComponent_ng_template_2_Template, 2, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("thyDone", ctx.loadingDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingDone);
    } }, directives: [ɵngcc2.ThyLoadingComponent, ɵngcc3.NgIf, ɵngcc4.ThySelectionListComponent, ɵngcc3.NgForOf, ɵngcc5.ThyListOptionComponent, ɵngcc3.NgTemplateOutlet], encapsulation: 2 });
ThyMentionSuggestionsComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone },
    { type: ThyPopoverRef }
];
ThyMentionSuggestionsComponent.propDecorators = {
    suggestionsClass: [{ type: HostBinding, args: ['class.thy-mention-suggestions',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMentionSuggestionsComponent, [{
        type: Component,
        args: [{
                selector: 'thy-mention-suggestions',
                template: "<thy-loading [thyDone]=\"loadingDone\"></thy-loading>\n<ng-container *ngIf=\"loadingDone\">\n  <thy-selection-list\n    *ngIf=\"data?.length > 0; else empty\"\n    thyBindKeyEventContainer=\"body\"\n    thyMultiple=\"false\"\n    thyAutoActiveFirstItem=\"true\"\n    thySpaceKeyEnabled=\"false\"\n    [thyScrollContainer]=\"elementRef\"\n    (thySelectionChange)=\"selectionChange($event)\"\n  >\n    <thy-list-option *ngFor=\"let item of data\" [thyValue]=\"item\">\n      <ng-container *ngIf=\"mention?.displayTemplateRef; else default\">\n        <ng-template *ngTemplateOutlet=\"mention?.displayTemplateRef; context: { $implicit: item }\"></ng-template>\n      </ng-container>\n      <ng-template #default>\n        {{ item['name'] || '' }}\n      </ng-template>\n    </thy-list-option>\n  </thy-selection-list>\n</ng-container>\n<ng-template #empty>\n  <div class=\"text-desc p-3\">\n    {{ mention?.emptyText }}\n  </div>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }, { type: ɵngcc1.ThyPopoverRef }]; }, { suggestionsClass: [{
            type: HostBinding,
            args: ['class.thy-mention-suggestions']
        }] }); })();
