// ngx-tethys/esm2015/time-picker/time-picker.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyTimePickerComponent } from './time-picker.component';
import { TimePickerConfig } from './time-picker.config';
import { ThyTimePickerStore } from './time-picker.store';
import * as ɵngcc0 from '@angular/core';
export class ThyTimePickerModule {
}
ThyTimePickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTimePickerModule });
ThyTimePickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTimePickerModule_Factory(t) { return new (t || ThyTimePickerModule)(); }, providers: [TimePickerConfig, ThyTimePickerStore], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTimePickerModule, { declarations: function () { return [ThyTimePickerComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyTimePickerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTimePickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [ThyTimePickerComponent],
                exports: [ThyTimePickerComponent],
                providers: [TimePickerConfig, ThyTimePickerStore]
            }]
    }], null, null); })();
