// ngx-tethys/esm2015/vote/vote.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyVoteComponent } from './vote.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyVoteModule {
}
ThyVoteModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyVoteModule });
ThyVoteModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyVoteModule_Factory(t) { return new (t || ThyVoteModule)(); }, providers: [], imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyVoteModule, { declarations: function () { return [ThyVoteComponent]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyVoteComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyVoteModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyVoteComponent],
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyVoteComponent],
                providers: []
            }]
    }], null, null); })();
