// ngx-tethys/esm2015/slide/slide-body/slide-body-section.component.js
import { Component, Input, HostBinding } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThySlideBodySectionComponent {
    constructor() {
        this.thySlideBodyItem = true;
        this.hasDivider = false;
    }
    set thyDividerBorder(value) {
        this.hasDivider = coerceBooleanProperty(value);
    }
    ngOnInit() { }
}
ThySlideBodySectionComponent.ɵfac = function ThySlideBodySectionComponent_Factory(t) { return new (t || ThySlideBodySectionComponent)(); };
ThySlideBodySectionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySlideBodySectionComponent, selectors: [["thy-slide-body-section"]], hostVars: 4, hostBindings: function ThySlideBodySectionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-slide-body-section", ctx.thySlideBodyItem)("thy-slide-body-section-divider", ctx.hasDivider);
    } }, inputs: { thyDividerBorder: "thyDividerBorder" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThySlideBodySectionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThySlideBodySectionComponent.propDecorators = {
    thySlideBodyItem: [{ type: HostBinding, args: ['class.thy-slide-body-section',] }],
    hasDivider: [{ type: HostBinding, args: ['class.thy-slide-body-section-divider',] }],
    thyDividerBorder: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySlideBodySectionComponent, [{
        type: Component,
        args: [{
                selector: 'thy-slide-body-section',
                template: '<ng-content></ng-content>'
            }]
    }], function () { return []; }, { thySlideBodyItem: [{
            type: HostBinding,
            args: ['class.thy-slide-body-section']
        }], hasDivider: [{
            type: HostBinding,
            args: ['class.thy-slide-body-section-divider']
        }], thyDividerBorder: [{
            type: Input
        }] }); })();
