// ngx-tethys/esm2015/slide/slide-body/slide-body.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThySlideBodyComponent {
    constructor() {
        this.slideLayoutBody = true;
    }
    ngOnInit() { }
}
ThySlideBodyComponent.ɵfac = function ThySlideBodyComponent_Factory(t) { return new (t || ThySlideBodyComponent)(); };
ThySlideBodyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySlideBodyComponent, selectors: [["thy-slide-body"]], hostVars: 2, hostBindings: function ThySlideBodyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-slide-body", ctx.slideLayoutBody);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThySlideBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThySlideBodyComponent.propDecorators = {
    slideLayoutBody: [{ type: HostBinding, args: ['class.thy-slide-body',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySlideBodyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-slide-body',
                template: `
        <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { slideLayoutBody: [{
            type: HostBinding,
            args: ['class.thy-slide-body']
        }] }); })();
