// ngx-tethys/esm2015/stepper/stepper-button.directive.js
import { Directive, HostListener } from '@angular/core';
import { ThyStepperComponent } from './stepper.component';
/** Button that moves to the next step in a stepper workflow. */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './stepper.component';
export class ThyStepperNextDirective {
    constructor(stepper) {
        this.stepper = stepper;
    }
    click($event) {
        this.stepper.next();
    }
}
ThyStepperNextDirective.ɵfac = function ThyStepperNextDirective_Factory(t) { return new (t || ThyStepperNextDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyStepperComponent)); };
ThyStepperNextDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyStepperNextDirective, selectors: [["", "thyStepperNext", ""]], hostBindings: function ThyStepperNextDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyStepperNextDirective_click_HostBindingHandler($event) { return ctx.click($event); });
    } } });
ThyStepperNextDirective.ctorParameters = () => [
    { type: ThyStepperComponent }
];
ThyStepperNextDirective.propDecorators = {
    click: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperNextDirective, [{
        type: Directive,
        args: [{
                selector: '[thyStepperNext]'
            }]
    }], function () { return [{ type: ɵngcc1.ThyStepperComponent }]; }, { click: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
/** Button that moves to the previous step in a stepper workflow. */
export class ThyStepperPreviousDirective {
    constructor(stepper) {
        this.stepper = stepper;
    }
    click($event) {
        this.stepper.previous();
    }
}
ThyStepperPreviousDirective.ɵfac = function ThyStepperPreviousDirective_Factory(t) { return new (t || ThyStepperPreviousDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyStepperComponent)); };
ThyStepperPreviousDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyStepperPreviousDirective, selectors: [["", "thyStepperPrevious", ""]], hostBindings: function ThyStepperPreviousDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyStepperPreviousDirective_click_HostBindingHandler($event) { return ctx.click($event); });
    } } });
ThyStepperPreviousDirective.ctorParameters = () => [
    { type: ThyStepperComponent }
];
ThyStepperPreviousDirective.propDecorators = {
    click: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperPreviousDirective, [{
        type: Directive,
        args: [{
                selector: '[thyStepperPrevious]'
            }]
    }], function () { return [{ type: ɵngcc1.ThyStepperComponent }]; }, { click: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
