// ngx-tethys/esm2015/stepper/step-header.component.js
import { Component, HostBinding, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ThyStepHeaderComponent {
    constructor() {
        this.thyStepHeader = true;
    }
}
ThyStepHeaderComponent.ɵfac = function ThyStepHeaderComponent_Factory(t) { return new (t || ThyStepHeaderComponent)(); };
ThyStepHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStepHeaderComponent, selectors: [["thy-step-header"]], hostVars: 4, hostBindings: function ThyStepHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-stepper-header", ctx.thyStepHeader)("thy-stepper-header-selected", ctx.selected);
    } }, inputs: { label: "label", index: "index", selected: "selected", active: "active" }, decls: 5, vars: 2, consts: [[1, "thy-step-number"], [1, "thy-step-label"]], template: function ThyStepHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "span");
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 1);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.index + 1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.label);
    } }, encapsulation: 2 });
ThyStepHeaderComponent.propDecorators = {
    label: [{ type: Input }],
    index: [{ type: Input }],
    selected: [{ type: HostBinding, args: ['class.thy-stepper-header-selected',] }, { type: Input }],
    active: [{ type: Input }],
    thyStepHeader: [{ type: HostBinding, args: ['class.thy-stepper-header',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-step-header',
                template: "<div class=\"thy-step-number\"> <span>{{index + 1}}</span></div>\n<div class=\"thy-step-label\">{{label}}</div>"
            }]
    }], function () { return []; }, { thyStepHeader: [{
            type: HostBinding,
            args: ['class.thy-stepper-header']
        }], label: [{
            type: Input
        }], index: [{
            type: Input
        }], selected: [{
            type: HostBinding,
            args: ['class.thy-stepper-header-selected']
        }, {
            type: Input
        }], active: [{
            type: Input
        }] }); })();
