// ngx-tethys/esm2015/stepper/stepper.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyStepHeaderComponent } from './step-header.component';
import { ThyStepComponent } from './step.component';
import { ThyStepperComponent } from './stepper.component';
import { ThyStepperNextDirective, ThyStepperPreviousDirective } from './stepper-button.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyStepperModule {
}
ThyStepperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyStepperModule });
ThyStepperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyStepperModule_Factory(t) { return new (t || ThyStepperModule)(); }, providers: [], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyStepperModule, { declarations: function () { return [ThyStepHeaderComponent, ThyStepperComponent, ThyStepComponent, ThyStepperNextDirective, ThyStepperPreviousDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyStepperComponent, ThyStepComponent, ThyStepHeaderComponent, ThyStepperNextDirective, ThyStepperPreviousDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyStepHeaderComponent, ThyStepperComponent, ThyStepComponent, ThyStepperNextDirective, ThyStepperPreviousDirective],
                entryComponents: [],
                imports: [CommonModule],
                exports: [ThyStepperComponent, ThyStepComponent, ThyStepHeaderComponent, ThyStepperNextDirective, ThyStepperPreviousDirective],
                providers: []
            }]
    }], null, null); })();
