// ngx-tethys/esm2015/stepper/step.component.js
import { Component, Input, ViewChild, TemplateRef, Inject, InjectionToken, Optional } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

function ThyStepComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c0 = ["*"];
export const THY_STEPPER_COMPONENT = new InjectionToken('THY_STEPPER_COMPONENT');
export class ThyStepComponent {
    constructor(stepper) {
        this.stepper = stepper;
    }
    select() {
        this.stepper.selected = this;
    }
}
ThyStepComponent.ɵfac = function ThyStepComponent_Factory(t) { return new (t || ThyStepComponent)(ɵngcc0.ɵɵdirectiveInject(THY_STEPPER_COMPONENT, 8)); };
ThyStepComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStepComponent, selectors: [["thy-step"]], viewQuery: function ThyStepComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(TemplateRef, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { label: "label", thyLabel: "thyLabel" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyStepComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThyStepComponent_ng_template_0_Template, 1, 0, "ng-template");
    } }, encapsulation: 2 });
ThyStepComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [THY_STEPPER_COMPONENT,] }] }
];
ThyStepComponent.propDecorators = {
    content: [{ type: ViewChild, args: [TemplateRef, { static: true },] }],
    label: [{ type: Input }],
    thyLabel: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepComponent, [{
        type: Component,
        args: [{
                selector: 'thy-step',
                template: "<ng-template><ng-content></ng-content></ng-template>\n"
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [THY_STEPPER_COMPONENT]
            }] }]; }, { content: [{
            type: ViewChild,
            args: [TemplateRef, { static: true }]
        }], label: [{
            type: Input
        }], thyLabel: [{
            type: Input
        }] }); })();
