// ngx-tethys/esm2015/stepper/stepper.component.js
import { Component, ViewChildren, ContentChildren, QueryList, HostBinding, Input, Output, EventEmitter } from '@angular/core';
import { ThyStepComponent, THY_STEPPER_COMPONENT } from './step.component';
import { ThyStepHeaderComponent } from './step-header.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from './step-header.component';

function ThyStepperComponent_div_0_ng_container_1_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 7);
} if (rf & 2) {
    const i_r3 = ɵngcc0.ɵɵnextContext().index;
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("thy-stepper-line-active", i_r3 < ctx_r5.selectedIndex);
} }
function ThyStepperComponent_div_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-step-header", 5);
    ɵngcc0.ɵɵtemplate(2, ThyStepperComponent_div_0_ng_container_1_div_2_Template, 1, 2, "div", 6);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const step_r2 = ctx.$implicit;
    const i_r3 = ctx.index;
    const isLast_r4 = ctx.last;
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("thy-stepper-header-active", i_r3 <= ctx_r1.selectedIndex);
    ɵngcc0.ɵɵproperty("index", i_r3)("label", step_r2.thyLabel || step_r2.label)("selected", ctx_r1.selectedIndex === i_r3)("active", i_r3 <= ctx_r1.selectedIndex);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !isLast_r4);
} }
function ThyStepperComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵtemplate(1, ThyStepperComponent_div_0_ng_container_1_Template, 3, 7, "ng-container", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.steps);
} }
export class ThyStepperComponent {
    constructor() {
        this.thyShowStepHeader = true;
        this._selectedIndex = 0;
        this.selectionChange = new EventEmitter();
        this.thyStepper = true;
    }
    set thySelectedIndex(value) {
        this.selectedIndex = value;
    }
    set thySelected(value) {
        this.selected = value;
    }
    set selected(step) {
        this.selectedIndex = this.steps ? this.steps.toArray().indexOf(step) : -1;
    }
    get selected() {
        return this.steps ? this.steps.toArray()[this.selectedIndex] : null;
    }
    set selectedIndex(index) {
        if (this.steps) {
            this._updateSelectedItemIndex(index);
        }
        else {
            this._selectedIndex = index;
        }
    }
    get selectedIndex() {
        return this._selectedIndex;
    }
    _updateSelectedItemIndex(newIndex) {
        const stepsArray = this.steps.toArray();
        this.selectionChange.emit({
            selectedIndex: newIndex,
            previouslySelectedIndex: this._selectedIndex,
            selectedStep: stepsArray[newIndex],
            previouslySelectedStep: stepsArray[this._selectedIndex]
        });
        this._selectedIndex = newIndex;
    }
    to(index) {
        this.selectedIndex = Math.min(index, this.steps.length - 1);
    }
    next() {
        this.selectedIndex = Math.min(this._selectedIndex + 1, this.steps.length - 1);
    }
    /** Selects and focuses the previous step in list. */
    previous() {
        this.selectedIndex = Math.max(this._selectedIndex - 1, 0);
    }
}
ThyStepperComponent.ɵfac = function ThyStepperComponent_Factory(t) { return new (t || ThyStepperComponent)(); };
ThyStepperComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStepperComponent, selectors: [["thy-stepper"]], contentQueries: function ThyStepperComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyStepComponent, false);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.steps = _t);
    } }, viewQuery: function ThyStepperComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(ThyStepHeaderComponent, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepHeaders = _t);
    } }, hostVars: 2, hostBindings: function ThyStepperComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-stepper", ctx.thyStepper);
    } }, inputs: { thyShowStepHeader: "thyShowStepHeader", thySelectedIndex: "thySelectedIndex", thySelected: "thySelected" }, outputs: { selectionChange: "selectionChange" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: THY_STEPPER_COMPONENT,
                useExisting: ThyStepperComponent
            }
        ])], decls: 3, vars: 2, consts: [["class", "thy-stepper-header-container", 4, "ngIf"], [1, "thy-stepper-content-container"], [3, "ngTemplateOutlet"], [1, "thy-stepper-header-container"], [4, "ngFor", "ngForOf"], [3, "index", "label", "selected", "active"], ["class", "thy-stepper-line", 3, "thy-stepper-line-active", 4, "ngIf"], [1, "thy-stepper-line"]], template: function ThyStepperComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyStepperComponent_div_0_Template, 2, 1, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementContainer(2, 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowStepHeader);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.selected == null ? null : ctx.selected.content);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgForOf, ɵngcc2.ThyStepHeaderComponent], encapsulation: 2 });
ThyStepperComponent.propDecorators = {
    thySelectedIndex: [{ type: Input }],
    thySelected: [{ type: Input }],
    thyShowStepHeader: [{ type: Input }],
    selectionChange: [{ type: Output }],
    stepHeaders: [{ type: ViewChildren, args: [ThyStepHeaderComponent,] }],
    steps: [{ type: ContentChildren, args: [ThyStepComponent,] }],
    thyStepper: [{ type: HostBinding, args: ['class.thy-stepper',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStepperComponent, [{
        type: Component,
        args: [{
                selector: 'thy-stepper',
                template: "<div *ngIf=\"thyShowStepHeader\" class=\"thy-stepper-header-container\">\n  <ng-container *ngFor=\"let step of steps; let i = index; let isLast = last\">\n    <thy-step-header\n      [index]=\"i\"\n      [label]=\"step.thyLabel || step.label\"\n      [selected]=\"selectedIndex === i\"\n      [class.thy-stepper-header-active]=\"i <= selectedIndex\"\n      [active]=\"i <= selectedIndex\"\n    >\n    </thy-step-header>\n    <div *ngIf=\"!isLast\" class=\"thy-stepper-line\" [class.thy-stepper-line-active]=\"i < selectedIndex\"></div>\n  </ng-container>\n</div>\n\n<div class=\"thy-stepper-content-container\">\n  <ng-container [ngTemplateOutlet]=\"selected?.content\"></ng-container>\n</div>\n",
                providers: [
                    {
                        provide: THY_STEPPER_COMPONENT,
                        useExisting: ThyStepperComponent
                    }
                ]
            }]
    }], function () { return []; }, { thyShowStepHeader: [{
            type: Input
        }], selectionChange: [{
            type: Output
        }], thyStepper: [{
            type: HostBinding,
            args: ['class.thy-stepper']
        }], thySelectedIndex: [{
            type: Input
        }], thySelected: [{
            type: Input
        }], stepHeaders: [{
            type: ViewChildren,
            args: [ThyStepHeaderComponent]
        }], steps: [{
            type: ContentChildren,
            args: [ThyStepComponent]
        }] }); })();
