// ngx-tethys/esm2015/shared/option/group/option-group.component.js
import { Component, HostBinding, Input, ContentChildren, QueryList, NgZone } from '@angular/core';
import { defer, Subject, merge } from 'rxjs';
import { ThyOptionComponent } from '../option.component';
import { take, switchMap, startWith, takeUntil, debounceTime, map } from 'rxjs/operators';
import { THY_OPTION_GROUP_COMPONENT } from '../option.token';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThySelectOptionGroupComponent {
    constructor(_ngZone) {
        this._ngZone = _ngZone;
        this._hidden = false;
        this._isOptionGroup = true;
        this._destroy$ = new Subject();
        this.optionVisibleChanges = defer(() => {
            if (this.options) {
                return merge(...this.options.map(option => option.visibleChange));
            }
            return this._ngZone.onStable.asObservable().pipe(take(1), switchMap(() => this.optionVisibleChanges));
        });
    }
    get hidden() {
        return this._hidden;
    }
    ngAfterContentInit() {
        this.options.changes.pipe(startWith(null), takeUntil(this._destroy$)).subscribe(() => {
            this._resetOptions();
        });
    }
    _resetOptions() {
        const changedOrDestroyed$ = merge(this.options.changes, this._destroy$);
        merge(...this.options.map(option => option.visibleChange))
            .pipe(takeUntil(changedOrDestroyed$), debounceTime(10), map((event) => {
            const hasOption = this.options.find(option => {
                if (!option.hidden) {
                    return true;
                }
            });
            if (hasOption) {
                return false;
            }
            else {
                return true;
            }
        }))
            .subscribe((data) => {
            this._hidden = data;
        });
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
}
ThySelectOptionGroupComponent.ɵfac = function ThySelectOptionGroupComponent_Factory(t) { return new (t || ThySelectOptionGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThySelectOptionGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySelectOptionGroupComponent, selectors: [["thy-option-group"]], contentQueries: function ThySelectOptionGroupComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyOptionComponent, false);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.options = _t);
    } }, hostVars: 6, hostBindings: function ThySelectOptionGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-option-item-group", ctx._isOptionGroup)("thy-select-option-group-hidden", ctx.hidden)("disabled", ctx.thyDisabled);
    } }, inputs: { thyDisabled: "thyDisabled", thyGroupLabel: "thyGroupLabel" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: THY_OPTION_GROUP_COMPONENT,
                useExisting: ThySelectOptionGroupComponent
            }
        ])], ngContentSelectors: _c0, decls: 3, vars: 1, consts: [[1, "group-name", "text-desc"]], template: function ThySelectOptionGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyGroupLabel);
    } }, encapsulation: 2 });
ThySelectOptionGroupComponent.ctorParameters = () => [
    { type: NgZone }
];
ThySelectOptionGroupComponent.propDecorators = {
    thyDisabled: [{ type: Input }, { type: HostBinding, args: [`class.disabled`,] }],
    _isOptionGroup: [{ type: HostBinding, args: ['class.thy-option-item-group',] }],
    hidden: [{ type: HostBinding, args: ['class.thy-select-option-group-hidden',] }],
    thyGroupLabel: [{ type: Input }],
    options: [{ type: ContentChildren, args: [ThyOptionComponent,] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySelectOptionGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-option-group',
                template: "<span class=\"group-name text-desc\">{{ thyGroupLabel }}</span>\n<ng-content></ng-content>\n",
                providers: [
                    {
                        provide: THY_OPTION_GROUP_COMPONENT,
                        useExisting: ThySelectOptionGroupComponent
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.NgZone }]; }, { _isOptionGroup: [{
            type: HostBinding,
            args: ['class.thy-option-item-group']
        }], hidden: [{
            type: HostBinding,
            args: ['class.thy-select-option-group-hidden']
        }], thyDisabled: [{
            type: Input
        }, {
            type: HostBinding,
            args: [`class.disabled`]
        }], thyGroupLabel: [{
            type: Input
        }], options: [{
            type: ContentChildren,
            args: [ThyOptionComponent]
        }] }); })();
