// ngx-tethys/esm2015/store/plugins/redux_devtools.js
export class StorePlugin {
}
export function tinyStateVersion() {
    return 'v0.6.0';
}
export class ReduxDevtoolsPlugin {
    constructor() {
        this._devTools = null;
        this._window = window;
        if (this._window == null) {
            return;
        }
        const globalDevtools = this._window['__REDUX_DEVTOOLS_EXTENSION__'] ||
            this._window['devToolsExtension'];
        if (!globalDevtools) {
            console.log(`未安装Chrome浏览器的拓展插件: Redux DevTools.`);
            console.log(`插件下载地址: https://www.chromefor.com/redux-devtools_v2-17-0/`);
            return;
        }
        this._devTools = globalDevtools.connect({
            name: `NgxStore ${tinyStateVersion()}`
        });
    }
    handleNewState(actionName, state) {
        if (this.isConnectSuccessed()) {
            this._devTools.send(actionName, state);
        }
    }
    isConnectSuccessed() {
        if (this._devTools === null) {
            return false;
        }
        return true;
    }
}
function getReduxDevToolsPlugin() {
    if (!window[`___ReduxDevtoolsPlugin___`]) {
        window[`___ReduxDevtoolsPlugin___`] = new ReduxDevtoolsPlugin();
    }
    return window[`___ReduxDevtoolsPlugin___`];
}
export default getReduxDevToolsPlugin;
