// ngx-tethys/esm2015/dialog/footer/dialog-footer.component.js
import { UpdateHostClassService } from 'ngx-tethys/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { Component, ContentChild, ElementRef, Inject, Input, TemplateRef } from '@angular/core';
import { THY_DIALOG_LAYOUT_CONFIG, ThyDialogLayoutConfig } from '../dialog.config';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '../dialog.config';

const _c0 = ["description"];
function DialogFooterComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementContainer(1, 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.description);
} }
const _c1 = ["*"];
export class DialogFooterComponent {
    constructor(dialogLayoutConfig, updateHostClassService, elementRef) {
        this.dialogLayoutConfig = dialogLayoutConfig;
        this.updateHostClassService = updateHostClassService;
        this.elementRef = elementRef;
        this.updateHostClassService.initializeElement(this.elementRef.nativeElement);
        this.divided = this.dialogLayoutConfig.footerDivided;
    }
    set thyDivided(value) {
        this.divided = coerceBooleanProperty(value);
    }
    get align() {
        return !!this.thyAlign ? this.thyAlign : this.dialogLayoutConfig.footerAlign;
    }
    ngOnInit() {
        this.updateHostClassService.updateClassByMap({
            'dialog-footer': true,
            [`dialog-footer-actions-align-${this.align}`]: true,
            'dialog-footer-border-top': this.divided
        });
    }
}
DialogFooterComponent.ɵfac = function DialogFooterComponent_Factory(t) { return new (t || DialogFooterComponent)(ɵngcc0.ɵɵdirectiveInject(THY_DIALOG_LAYOUT_CONFIG), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
DialogFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DialogFooterComponent, selectors: [["thy-dialog-footer"]], contentQueries: function DialogFooterComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.description = _t.first);
    } }, inputs: { thyDivided: "thyDivided", thyAlign: "thyAlign" }, exportAs: ["thyDialogFooter"], features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], ngContentSelectors: _c1, decls: 3, vars: 1, consts: [[1, "dialog-footer-actions", "btn-pair"], ["class", "dialog-footer-description", 4, "ngIf"], [1, "dialog-footer-description"], [3, "ngTemplateOutlet"]], template: function DialogFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, DialogFooterComponent_div_2_Template, 2, 1, "div", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.description);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgTemplateOutlet], encapsulation: 2 });
DialogFooterComponent.ctorParameters = () => [
    { type: ThyDialogLayoutConfig, decorators: [{ type: Inject, args: [THY_DIALOG_LAYOUT_CONFIG,] }] },
    { type: UpdateHostClassService },
    { type: ElementRef }
];
DialogFooterComponent.propDecorators = {
    description: [{ type: ContentChild, args: ['description',] }],
    thyDivided: [{ type: Input }],
    thyAlign: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DialogFooterComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-footer',
                template: "<div class=\"dialog-footer-actions btn-pair\">\n  <ng-content></ng-content>\n</div>\n<div class=\"dialog-footer-description\" *ngIf=\"description\">\n  <ng-container [ngTemplateOutlet]=\"description\"></ng-container>\n</div>\n",
                // changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'thyDialogFooter',
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: ɵngcc3.ThyDialogLayoutConfig, decorators: [{
                type: Inject,
                args: [THY_DIALOG_LAYOUT_CONFIG]
            }] }, { type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }]; }, { thyDivided: [{
            type: Input
        }], description: [{
            type: ContentChild,
            args: ['description']
        }], thyAlign: [{
            type: Input
        }] }); })();
