// ngx-tethys/esm2015/tree-select/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { OverlayModule } from '@angular/cdk/overlay';
import { ThyLabelModule } from 'ngx-tethys/label';
import { ThyTreeSelectComponent } from './tree-select.component';
import { ThyTreeSelectNodesComponent } from './tree-select-nodes.component';
import { ThyListModule } from 'ngx-tethys/list';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyEmptyModule } from 'ngx-tethys/empty';
import { ThySelectCommonModule } from 'ngx-tethys/shared';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/cdk/overlay';
import * as ɵngcc2 from 'ngx-tethys/shared';
import * as ɵngcc3 from '@angular/common';
export class ThyTreeSelectModule {
}
ThyTreeSelectModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTreeSelectModule });
ThyTreeSelectModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTreeSelectModule_Factory(t) { return new (t || ThyTreeSelectModule)(); }, imports: [[
            CommonModule,
            FormsModule,
            OverlayModule,
            ThyLabelModule,
            ThyListModule,
            ThyIconModule,
            ThyEmptyModule,
            ThySelectCommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTreeSelectModule, { declarations: function () { return [ThyTreeSelectComponent, ThyTreeSelectNodesComponent]; }, imports: function () { return [CommonModule,
        FormsModule,
        OverlayModule,
        ThyLabelModule,
        ThyListModule,
        ThyIconModule,
        ThyEmptyModule,
        ThySelectCommonModule]; }, exports: function () { return [ThyTreeSelectComponent, ThyTreeSelectNodesComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    OverlayModule,
                    ThyLabelModule,
                    ThyListModule,
                    ThyIconModule,
                    ThyEmptyModule,
                    ThySelectCommonModule
                ],
                declarations: [ThyTreeSelectComponent, ThyTreeSelectNodesComponent],
                exports: [ThyTreeSelectComponent, ThyTreeSelectNodesComponent]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(ThyTreeSelectComponent, [ɵngcc1.CdkOverlayOrigin, ɵngcc2.ThySelectControlComponent, ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet, ɵngcc1.CdkConnectedOverlay, ThyTreeSelectNodesComponent], []);
