// ngx-tethys/esm2015/tree-select/tree-select-nodes.component.js
import { Component, HostBinding, Input } from '@angular/core';
import { ThyTreeSelectComponent } from './tree-select.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './tree-select.component';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-tethys/empty';
import * as ɵngcc4 from 'ngx-tethys/icon';

const _c0 = function (a0) { return { $implicit: a0 }; };
function ThyTreeSelectNodesComponent_div_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 5);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    const _r3 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r3)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c0, ctx_r5.treeNodes));
} }
function ThyTreeSelectNodesComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵtemplate(1, ThyTreeSelectNodesComponent_div_0_ng_container_1_Template, 1, 4, "ng-container", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.treeNodes == null ? null : ctx_r0.treeNodes.length) > 0);
} }
function ThyTreeSelectNodesComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-empty", 6);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyMessage", ctx_r2.parent.thyEmptyOptionsText);
} }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_ng_template_1_Template(rf, ctx) { }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_6_ng_template_0_Template(rf, ctx) { }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_6_ng_template_0_Template, 0, 0, "ng-template", 5);
} if (rf & 2) {
    const node_r8 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r13.treeNodeTemplateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c0, node_r8));
} }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const node_r8 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", node_r8[ctx_r15.showKey], " ");
} }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_span_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 17);
    ɵngcc0.ɵɵelement(1, "i", 18);
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = function (a0) { return { active: a0 }; };
const _c2 = function (a0) { return { "padding-left.px": a0 }; };
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵlistener("click", function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const node_r8 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r20 = ɵngcc0.ɵɵnextContext(2); return ctx_r20.selectTreeNode($event, node_r8); });
    ɵngcc0.ɵɵtemplate(1, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_ng_template_1_Template, 0, 0, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(3, "span", 11);
    ɵngcc0.ɵɵlistener("click", function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_Template_span_click_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const node_r8 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r23 = ɵngcc0.ɵɵnextContext(2); return ctx_r23.nodeExpandToggle($event, node_r8); });
    ɵngcc0.ɵɵelement(4, "thy-icon", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "span", 13);
    ɵngcc0.ɵɵtemplate(6, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_6_Template, 1, 4, undefined, 14);
    ɵngcc0.ɵɵtemplate(7, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_ng_template_7_Template, 1, 1, "ng-template", null, 15, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(9, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_span_9_Template, 2, 0, "span", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r14 = ɵngcc0.ɵɵreference(8);
    const node_r8 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r9.treeNodeIsDisable(node_r8));
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(10, _c1, ctx_r9.treeNodeIsSelected(node_r8)))("ngStyle", ɵngcc0.ɵɵpureFunction1(12, _c2, 20 * node_r8.level + ctx_r9.parent.icons.gap));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵclassProp("invisible", !((node_r8.children == null ? null : node_r8.children.length) > 0 || node_r8[ctx_r9.childCountKey] && node_r8[ctx_r9.childCountKey] > 0));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r9.treeNodeIsExpand(node_r8) ? ctx_r9.parent.icons.expand : ctx_r9.parent.icons.collapse);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r9.treeNodeTemplateRef)("ngIfElse", _r14);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r9.isMultiple);
} }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 5);
} if (rf & 2) {
    const node_r8 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵnextContext(2);
    const _r3 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r3)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c0, node_r8.children));
} }
function ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_a_1_Template, 10, 14, "a", 9);
    ɵngcc0.ɵɵtemplate(2, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_ng_container_2_Template, 1, 4, "ng-container", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const node_r8 = ctx.$implicit;
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r7.treeNodeIsHidden(node_r8));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r7.treeNodeIsExpand(node_r8) && (node_r8.children == null ? null : node_r8.children.length) > 0);
} }
function ThyTreeSelectNodesComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵtemplate(1, ThyTreeSelectNodesComponent_ng_template_3_ng_container_1_Template, 3, 2, "ng-container", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const nodes_r6 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", nodes_r6);
} }
export class ThyTreeSelectNodesComponent {
    constructor(parent) {
        this.parent = parent;
        this.primaryKey = this.parent.thyPrimaryKey;
        this.showKey = this.parent.thyShowKey;
        this.isMultiple = this.parent.thyMultiple;
        this.valueIsObject = this.parent.valueIsObject;
        this.selectedValue = this.parent.selectedValue;
        this.childCountKey = this.parent.thyChildCountKey;
        this.treeNodeTemplateRef = this.parent.treeNodeTemplateRef;
    }
    ngOnInit() {
        this.class = this.isMultiple ? 'thy-tree-select-dropdown thy-tree-select-dropdown-multiple' : 'thy-tree-select-dropdown';
    }
    treeNodeIsSelected(node) {
        if (this.parent.thyMultiple) {
            return (this.parent.selectedNodes || []).find(item => {
                return item[this.primaryKey] === node[this.primaryKey];
            });
        }
        else {
            return this.parent.selectedNode && this.parent.selectedNode[this.primaryKey] === node[this.primaryKey];
        }
    }
    treeNodeIsHidden(node) {
        if (this.parent.thyHiddenNodeKey) {
            return node[this.parent.thyHiddenNodeKey];
        }
        if (this.parent.thyHiddenNodeFn) {
            return this.parent.thyHiddenNodeFn(node);
        }
        return false;
    }
    treeNodeIsDisable(node) {
        if (this.parent.thyDisableNodeKey) {
            return node[this.parent.thyDisableNodeKey];
        }
        if (this.parent.thyDisableNodeFn) {
            return this.parent.thyDisableNodeFn(node);
        }
        return false;
    }
    treeNodeIsExpand(node) {
        let isSelectedNodeParent = false;
        if (this.parent.thyMultiple) {
            isSelectedNodeParent = !!(this.parent.selectedNodes || []).find(item => {
                return item.parentValues.indexOf(node[this.primaryKey]) > -1;
            });
        }
        else {
            isSelectedNodeParent = this.parent.selectedNode
                ? this.parent.selectedNode.parentValues.indexOf(node[this.primaryKey]) > -1
                : false;
        }
        const isExpand = node.expand || (Object.keys(node).indexOf('expand') < 0 && isSelectedNodeParent);
        node.expand = isExpand;
        return isExpand;
    }
    getNodeChildren(node) {
        return this.parent.getNodeChildren(node);
    }
    selectTreeNode(event, node) {
        event.stopPropagation();
        if (this.treeNodeIsDisable(node)) {
            return;
        }
        this.parent.selectNode(node);
    }
    nodeExpandToggle(event, node) {
        event.stopPropagation();
        if (Object.keys(node).indexOf('expand') > -1) {
            node.expand = !node.expand;
        }
        else {
            if (this.treeNodeIsExpand(node)) {
                node.expand = false;
            }
            else {
                node.expand = true;
            }
        }
        if (node.expand && this.parent.thyAsyncNode) {
            this.getNodeChildren(node).subscribe(() => {
                this.parent.setPosition();
            });
        }
        this.parent.setPosition();
    }
}
ThyTreeSelectNodesComponent.ɵfac = function ThyTreeSelectNodesComponent_Factory(t) { return new (t || ThyTreeSelectNodesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTreeSelectComponent)); };
ThyTreeSelectNodesComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTreeSelectNodesComponent, selectors: [["thy-tree-select-nodes"]], hostVars: 2, hostBindings: function ThyTreeSelectNodesComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.class);
    } }, inputs: { treeNodes: "treeNodes" }, decls: 5, vars: 2, consts: [["class", "thy-tree-select-options", 4, "ngIf", "ngIfElse"], ["emptyPlaceholder", ""], ["treeSelectNode", ""], [1, "thy-tree-select-options"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["thySize", "sm", 1, "thy-select-empty-content", 3, "thyMessage"], [1, "thy-tree-select-node"], [4, "ngFor", "ngForOf"], ["class", "thy-option-item", 3, "ngClass", "disabled", "ngStyle", "click", 4, "ngIf"], [1, "thy-option-item", 3, "ngClass", "ngStyle", "click"], [1, "thy-tree-select-option-icon", 3, "click"], [1, "node-expand-icon", 3, "thyIconName"], [1, "thy-tree-select-option-text"], [4, "ngIf", "ngIfElse"], ["defaultNodeText", ""], ["class", "checked-icon", 4, "ngIf"], [1, "checked-icon"], [1, "wtf", "wtf-checked"]], template: function ThyTreeSelectNodesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyTreeSelectNodesComponent_div_0_Template, 2, 1, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyTreeSelectNodesComponent_ng_template_1_Template, 1, 1, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(3, ThyTreeSelectNodesComponent_ng_template_3_Template, 2, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(2);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.treeNodes == null ? null : ctx.treeNodes.length) > 0)("ngIfElse", _r1);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgTemplateOutlet, ɵngcc3.ThyEmptyComponent, ɵngcc2.NgForOf, ɵngcc2.NgClass, ɵngcc2.NgStyle, ɵngcc4.ThyIconComponent], encapsulation: 2 });
ThyTreeSelectNodesComponent.ctorParameters = () => [
    { type: ThyTreeSelectComponent }
];
ThyTreeSelectNodesComponent.propDecorators = {
    class: [{ type: HostBinding, args: ['class',] }],
    treeNodes: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeSelectNodesComponent, [{
        type: Component,
        args: [{
                selector: 'thy-tree-select-nodes',
                template: "<div class=\"thy-tree-select-options\" *ngIf=\"treeNodes?.length > 0; else emptyPlaceholder\">\n  <ng-container\n    *ngIf=\"treeNodes?.length > 0\"\n    [ngTemplateOutlet]=\"treeSelectNode\"\n    [ngTemplateOutletContext]=\"{ $implicit: treeNodes }\"\n  ></ng-container>\n</div>\n<ng-template #emptyPlaceholder>\n  <thy-empty class=\"thy-select-empty-content\" thySize=\"sm\" [thyMessage]=\"parent.thyEmptyOptionsText\"></thy-empty>\n</ng-template>\n<ng-template #treeSelectNode let-nodes>\n  <div class=\"thy-tree-select-node\">\n    <ng-container *ngFor=\"let node of nodes\">\n      <a\n        class=\"thy-option-item\"\n        [ngClass]=\"{ active: treeNodeIsSelected(node) }\"\n        [class.disabled]=\"treeNodeIsDisable(node)\"\n        [ngStyle]=\"{ 'padding-left.px': 20 * node.level + parent.icons.gap }\"\n        (click)=\"selectTreeNode($event, node)\"\n        *ngIf=\"!treeNodeIsHidden(node)\"\n      >\n        <ng-template #treeSelectNode> </ng-template>\n        <span\n          class=\"thy-tree-select-option-icon\"\n          [class.invisible]=\"!(node.children?.length > 0 || (node[childCountKey] && node[childCountKey] > 0))\"\n          (click)=\"nodeExpandToggle($event, node)\"\n        >\n          <thy-icon\n            class=\"node-expand-icon\"\n            [thyIconName]=\"treeNodeIsExpand(node) ? parent.icons.expand : parent.icons.collapse\"\n          ></thy-icon>\n        </span>\n        <span class=\"thy-tree-select-option-text\">\n          <ng-template\n            *ngIf=\"treeNodeTemplateRef; else defaultNodeText\"\n            [ngTemplateOutlet]=\"treeNodeTemplateRef\"\n            [ngTemplateOutletContext]=\"{ $implicit: node }\"\n          ></ng-template>\n          <ng-template #defaultNodeText>\n            {{ node[showKey] }}\n          </ng-template>\n        </span>\n        <span class=\"checked-icon\" *ngIf=\"isMultiple\">\n          <i class=\"wtf wtf-checked\"></i>\n        </span>\n      </a>\n      <ng-container\n        *ngIf=\"treeNodeIsExpand(node) && node.children?.length > 0\"\n        [ngTemplateOutlet]=\"treeSelectNode\"\n        [ngTemplateOutletContext]=\"{ $implicit: node.children }\"\n      ></ng-container>\n    </ng-container>\n  </div>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ThyTreeSelectComponent }]; }, { class: [{
            type: HostBinding,
            args: ['class']
        }], treeNodes: [{
            type: Input
        }] }); })();
