// ngx-tethys/esm2015/popover/body/popover-body.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyPopoverBodyComponent {
    constructor() {
        this.isPopoverBody = true;
    }
}
ThyPopoverBodyComponent.ɵfac = function ThyPopoverBodyComponent_Factory(t) { return new (t || ThyPopoverBodyComponent)(); };
ThyPopoverBodyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPopoverBodyComponent, selectors: [["thy-popover-body"]], hostVars: 2, hostBindings: function ThyPopoverBodyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-popover-body", ctx.isPopoverBody);
    } }, exportAs: ["thyPopoverBody"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyPopoverBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyPopoverBodyComponent.propDecorators = {
    isPopoverBody: [{ type: HostBinding, args: ['class.thy-popover-body',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverBodyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-popover-body',
                template: '<ng-content></ng-content>',
                exportAs: 'thyPopoverBody'
            }]
    }], function () { return []; }, { isPopoverBody: [{
            type: HostBinding,
            args: ['class.thy-popover-body']
        }] }); })();
