// ngx-tethys/esm2015/property-operation/module.js
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyFlexibleTextModule } from 'ngx-tethys/flexible-text';
import { ThyIconModule } from 'ngx-tethys/icon';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyPropertyOperationGroupComponent } from './group/property-operation-group.component';
import { ThyPropertyOperationComponent } from './property-operation.component';
import * as ɵngcc0 from '@angular/core';
export class ThyPropertyOperationModule {
}
ThyPropertyOperationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyPropertyOperationModule });
ThyPropertyOperationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyPropertyOperationModule_Factory(t) { return new (t || ThyPropertyOperationModule)(); }, imports: [[CommonModule, ThyButtonModule, ThyIconModule, ThyFlexibleTextModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyPropertyOperationModule, { declarations: function () { return [ThyPropertyOperationComponent, ThyPropertyOperationGroupComponent]; }, imports: function () { return [CommonModule, ThyButtonModule, ThyIconModule, ThyFlexibleTextModule]; }, exports: function () { return [ThyPropertyOperationComponent, ThyPropertyOperationGroupComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPropertyOperationModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyPropertyOperationComponent, ThyPropertyOperationGroupComponent],
                imports: [CommonModule, ThyButtonModule, ThyIconModule, ThyFlexibleTextModule],
                exports: [ThyPropertyOperationComponent, ThyPropertyOperationGroupComponent]
            }]
    }], null, null); })();
