// ngx-tethys/esm2015/dialog/dialog.module.js
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyFormModule } from 'ngx-tethys/form';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThySharedModule } from 'ngx-tethys/shared';
import { DialogBodyComponent } from './body/dialog-body.component';
import { THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER } from './confirm.config';
import { ThyConfirmComponent } from './confirm/confirm.component';
import { ThyDialogContainerComponent } from './dialog-container.component';
import { THY_DIALOG_DEFAULT_OPTIONS_PROVIDER, THY_DIALOG_LAYOUT_CONFIG_PROVIDER } from './dialog.config';
import { ThyDialog } from './dialog.service';
import { DialogFooterComponent } from './footer/dialog-footer.component';
import { DialogHeaderComponent } from './header/dialog-header.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from 'ngx-tethys/form';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/button';
export class ThyDialogModule {
}
ThyDialogModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDialogModule });
ThyDialogModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDialogModule_Factory(t) { return new (t || ThyDialogModule)(); }, providers: [ThyDialog, THY_DIALOG_DEFAULT_OPTIONS_PROVIDER, THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER, THY_DIALOG_LAYOUT_CONFIG_PROVIDER], imports: [[CommonModule, ThySharedModule, PortalModule, OverlayModule, ThyButtonModule, ThyIconModule, ThyFormModule, FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDialogModule, { declarations: function () { return [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent, ThyConfirmComponent]; }, imports: function () { return [CommonModule, ThySharedModule, PortalModule, OverlayModule, ThyButtonModule, ThyIconModule, ThyFormModule, FormsModule]; }, exports: function () { return [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDialogModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent, ThyConfirmComponent],
                imports: [CommonModule, ThySharedModule, PortalModule, OverlayModule, ThyButtonModule, ThyIconModule, ThyFormModule, FormsModule],
                providers: [ThyDialog, THY_DIALOG_DEFAULT_OPTIONS_PROVIDER, THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER, THY_DIALOG_LAYOUT_CONFIG_PROVIDER],
                entryComponents: [ThyDialogContainerComponent, ThyConfirmComponent],
                exports: [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(ThyConfirmComponent, [DialogHeaderComponent, DialogBodyComponent, ɵngcc1.ɵangular_packages_forms_forms_y, ɵngcc1.NgControlStatusGroup, ɵngcc1.NgForm, ɵngcc2.ThyFormDirective, ɵngcc2.ThyFormGroupFooterComponent, ɵngcc3.NgClass, ɵngcc4.ThyButtonComponent], []);
