// ngx-tethys/esm2015/dialog/dialog.config.js
import { InjectionToken } from '@angular/core';
import { ThyAbstractOverlayConfig } from 'ngx-tethys/core';
export var ThyDialogSizes;
(function (ThyDialogSizes) {
    ThyDialogSizes["lg"] = "lg";
    ThyDialogSizes["supperLg"] = "supper-lg";
    ThyDialogSizes["maxLg"] = "max-lg";
    ThyDialogSizes["md"] = "md";
    ThyDialogSizes["sm"] = "sm";
    ThyDialogSizes["full"] = "full";
})(ThyDialogSizes || (ThyDialogSizes = {}));
/**
 * Configuration for opening a modal dialog with the ThyDialog service.
 */
export class ThyDialogConfig extends ThyAbstractOverlayConfig {
    constructor() {
        super(...arguments);
        /** The ARIA role of the dialog element. */
        this.role = 'dialog';
    }
}
/** Injection token that can be used to specify default dialog options. */
export const THY_DIALOG_DEFAULT_OPTIONS = new InjectionToken('thy-dialog-default-options');
const ɵ0 = {
    role: 'dialog',
    hasBackdrop: true,
    backdropClass: '',
    panelClass: '',
    backdropClosable: true,
    closeOnNavigation: true,
    autoFocus: true,
    restoreFocus: true
};
export const THY_DIALOG_DEFAULT_OPTIONS_PROVIDER = {
    provide: THY_DIALOG_DEFAULT_OPTIONS,
    useValue: ɵ0
};
export class ThyDialogLayoutConfig {
}
export const THY_DIALOG_LAYOUT_CONFIG = new InjectionToken('thy-dialog-layout-config');
const ɵ1 = {
    footerAlign: 'left',
    footerDivided: false
};
export const THY_DIALOG_LAYOUT_CONFIG_PROVIDER = {
    provide: THY_DIALOG_LAYOUT_CONFIG,
    useValue: ɵ1
};
export { ɵ0, ɵ1 };
