// ngx-tethys/schematics/ng-update/update-9/migrations/import-entry-point-change.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportEntryPointChangeMigrationByNg9 = void 0;
const class_1 = require("../../../class");
class ImportEntryPointChangeMigrationByNg9 extends class_1.ImportEntryPointChangeMigrationBase {
    constructor() {
        super(...arguments);
        this.changeModulePackageGroup = {
            Id: {
                with: 'ngx-tethys/types',
                replace: 'ngx-tethys/store'
            },
            PaginationInfo: {
                with: 'ngx-tethys/types',
                replace: 'ngx-tethys/store'
            },
            UpdateHostClassService: {
                with: 'ngx-tethys/core',
                replace: 'ngx-tethys/shared'
            }
        };
        this.changePackageGroup = {
            'ngx-tethys/typings': 'ngx-tethys/types',
            'ngx-tethys/directives': 'ngx-tethys/shared'
        };
    }
}
exports.ImportEntryPointChangeMigrationByNg9 = ImportEntryPointChangeMigrationByNg9;
