// ngx-tethys/esm2015/util/date/functions.js
import * as dateFns from 'date-fns';
export const differenceInCalendarDays = dateFns.differenceInCalendarDays;
export const differenceInCalendarMonths = dateFns.differenceInCalendarMonths;
export const differenceInCalendarYears = dateFns.differenceInCalendarYears;
export const differenceInWeeks = dateFns.differenceInWeeks;
export const differenceInHours = dateFns.differenceInHours;
export const differenceInMinutes = dateFns.differenceInMinutes;
export const differenceInSeconds = dateFns.differenceInSeconds;
export const isSameDay = dateFns.isSameDay;
export const isSameHour = dateFns.isSameHour;
export const isSameMinute = dateFns.isSameMinute;
export const isSameMonth = dateFns.isSameMonth;
export const isSameSecond = dateFns.isSameSecond;
export const isSameYear = dateFns.isSameYear;
export const isToday = dateFns.isToday;
export const isTomorrow = dateFns.isTomorrow;
export const isValid = dateFns.isValid;
export const setYear = dateFns.setYear;
export const startOfMonth = dateFns.startOfMonth;
export const startOfWeek = dateFns.startOfWeek;
export const addMonths = dateFns.addMonths;
export const addYears = dateFns.addYears;
export const setDay = dateFns.setDay;
export const setMonth = dateFns.setMonth;
export const getUnixTime = dateFns.getUnixTime;
export const startOfDay = dateFns.startOfDay;
export const endOfDay = dateFns.endOfDay;
export const fromUnixTime = dateFns.fromUnixTime;
export const isWeekend = dateFns.isWeekend;
export const getWeek = dateFns.getWeek;
export const getDaysInMonth = dateFns.getDaysInMonth;
export const addSeconds = dateFns.addSeconds;
export const addMinutes = dateFns.addMinutes;
export const addHours = dateFns.addHours;
export const addWeeks = dateFns.addWeeks;
export const addQuarters = dateFns.addQuarters;
export const startOfQuarter = dateFns.startOfQuarter;
export const startOfYear = dateFns.startOfYear;
export const endOfWeek = dateFns.endOfWeek;
export const endOfMonth = dateFns.endOfMonth;
export const endOfQuarter = dateFns.endOfQuarter;
export const endOfYear = dateFns.endOfYear;
export const format = dateFns.format;
export const getQuarter = dateFns.getQuarter;
export const addDays = dateFns.addDays;
