// ngx-tethys/esm2015/calendar/calendar-cells.js
import { Directive } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ThyDateCellDirective {
}
ThyDateCellDirective.ɵfac = function ThyDateCellDirective_Factory(t) { return new (t || ThyDateCellDirective)(); };
ThyDateCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDateCellDirective, selectors: [["", "thyDateCell", ""]], exportAs: ["thyDateCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDateCellDirective, [{
        type: Directive,
        args: [{
                selector: '[thyDateCell]',
                exportAs: 'thyDateCell'
            }]
    }], null, null); })();
export class ThyMonthCellDirective {
}
ThyMonthCellDirective.ɵfac = function ThyMonthCellDirective_Factory(t) { return new (t || ThyMonthCellDirective)(); };
ThyMonthCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMonthCellDirective, selectors: [["", "thyMonthCell", ""]], exportAs: ["thyMonthCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMonthCellDirective, [{
        type: Directive,
        args: [{
                selector: '[thyMonthCell]',
                exportAs: 'thyMonthCell'
            }]
    }], null, null); })();
export class ThyDateFullCellDirective {
}
ThyDateFullCellDirective.ɵfac = function ThyDateFullCellDirective_Factory(t) { return new (t || ThyDateFullCellDirective)(); };
ThyDateFullCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDateFullCellDirective, selectors: [["", "thyDateFullCell", ""]], exportAs: ["thyDateFullCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDateFullCellDirective, [{
        type: Directive,
        args: [{
                selector: '[thyDateFullCell]',
                exportAs: 'thyDateFullCell'
            }]
    }], null, null); })();
export class ThyMonthFullCellDirective {
}
ThyMonthFullCellDirective.ɵfac = function ThyMonthFullCellDirective_Factory(t) { return new (t || ThyMonthFullCellDirective)(); };
ThyMonthFullCellDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMonthFullCellDirective, selectors: [["", "thyMonthFullCell", ""]], exportAs: ["thyMonthFullCell"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMonthFullCellDirective, [{
        type: Directive,
        args: [{
                selector: '[thyMonthFullCell]',
                exportAs: 'thyMonthFullCell'
            }]
    }], null, null); })();
export class ThyCalendarHeaderOperationDirective {
}
ThyCalendarHeaderOperationDirective.ɵfac = function ThyCalendarHeaderOperationDirective_Factory(t) { return new (t || ThyCalendarHeaderOperationDirective)(); };
ThyCalendarHeaderOperationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyCalendarHeaderOperationDirective, selectors: [["", "thyCalendarHeaderOperation", ""]], exportAs: ["thyCalendarHeaderOperation"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCalendarHeaderOperationDirective, [{
        type: Directive,
        args: [{
                selector: '[thyCalendarHeaderOperation]',
                exportAs: 'thyCalendarHeaderOperation'
            }]
    }], null, null); })();
