// ngx-tethys/esm2015/calendar/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyCalendarHeaderComponent } from './calendar-header.component';
import { ThyCalendarComponent } from './calendar.component';
import { ThyCalendarHeaderOperationDirective, ThyDateCellDirective, ThyDateFullCellDirective, ThyMonthCellDirective, ThyMonthFullCellDirective } from './calendar-cells';
import { LibPackerModule } from 'ngx-tethys/date-picker';
import { ThySelectModule } from 'ngx-tethys/select';
import { ThyRadioModule } from 'ngx-tethys/radio';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyDateRangeModule } from 'ngx-tethys/date-range';
import * as ɵngcc0 from '@angular/core';
export class ThyCalendarModule {
}
ThyCalendarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCalendarModule });
ThyCalendarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCalendarModule_Factory(t) { return new (t || ThyCalendarModule)(); }, imports: [[
            CommonModule,
            FormsModule,
            ThySharedModule,
            ThyIconModule,
            LibPackerModule,
            ThySelectModule,
            ThyRadioModule,
            ThyButtonModule,
            ThyDateRangeModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCalendarModule, { declarations: function () { return [ThyCalendarHeaderComponent,
        ThyCalendarComponent,
        ThyDateCellDirective,
        ThyDateFullCellDirective,
        ThyMonthCellDirective,
        ThyMonthFullCellDirective,
        ThyCalendarHeaderOperationDirective]; }, imports: function () { return [CommonModule,
        FormsModule,
        ThySharedModule,
        ThyIconModule,
        LibPackerModule,
        ThySelectModule,
        ThyRadioModule,
        ThyButtonModule,
        ThyDateRangeModule]; }, exports: function () { return [ThyCalendarHeaderComponent,
        ThyCalendarComponent,
        ThyDateCellDirective,
        ThyDateFullCellDirective,
        ThyMonthCellDirective,
        ThyMonthFullCellDirective,
        ThyCalendarHeaderOperationDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCalendarModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ThySharedModule,
                    ThyIconModule,
                    LibPackerModule,
                    ThySelectModule,
                    ThyRadioModule,
                    ThyButtonModule,
                    ThyDateRangeModule
                ],
                declarations: [
                    ThyCalendarHeaderComponent,
                    ThyCalendarComponent,
                    ThyDateCellDirective,
                    ThyDateFullCellDirective,
                    ThyMonthCellDirective,
                    ThyMonthFullCellDirective,
                    ThyCalendarHeaderOperationDirective
                ],
                exports: [
                    ThyCalendarHeaderComponent,
                    ThyCalendarComponent,
                    ThyDateCellDirective,
                    ThyDateFullCellDirective,
                    ThyMonthCellDirective,
                    ThyMonthFullCellDirective,
                    ThyCalendarHeaderOperationDirective
                ]
            }]
    }], null, null); })();
