// ngx-tethys/esm2015/date-picker/lib/calendar/calendar-footer.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ElementRef, Renderer2, ViewContainerRef } from '@angular/core';
import { TinyDate } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/button';
import * as ɵngcc3 from 'ngx-tethys/icon';
import * as ɵngcc4 from 'ngx-tethys/time-picker';
import * as ɵngcc5 from '@angular/forms';

function CalendarFooterComponent_div_0_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 7);
    ɵngcc0.ɵɵlistener("click", function CalendarFooterComponent_div_0_a_1_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(2); return ctx_r4.changeTimeShowMode("show"); });
    ɵngcc0.ɵɵelement(1, "thy-icon", 8);
    ɵngcc0.ɵɵtext(2, "\u8BBE\u7F6E\u65F6\u95F4 ");
    ɵngcc0.ɵɵelementEnd();
} }
function CalendarFooterComponent_div_0_thy_time_picker_2_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-time-picker", 9);
    ɵngcc0.ɵɵlistener("ngModelChange", function CalendarFooterComponent_div_0_thy_time_picker_2_Template_thy_time_picker_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(2); return ctx_r6.onSelectTime($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("showMeridian", false)("ngModel", ctx_r2.value == null ? null : ctx_r2.value.nativeDate);
} }
function CalendarFooterComponent_div_0_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 10);
    ɵngcc0.ɵɵlistener("click", function CalendarFooterComponent_div_0_button_4_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.onTimeOk(); });
    ɵngcc0.ɵɵtext(1, " \u786E\u5B9A ");
    ɵngcc0.ɵɵelementEnd();
} }
function CalendarFooterComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵlistener("click", function CalendarFooterComponent_div_0_Template_div_click_0_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵtemplate(1, CalendarFooterComponent_div_0_a_1_Template, 3, 0, "a", 2);
    ɵngcc0.ɵɵtemplate(2, CalendarFooterComponent_div_0_thy_time_picker_2_Template, 1, 2, "thy-time-picker", 3);
    ɵngcc0.ɵɵelementStart(3, "div", 4);
    ɵngcc0.ɵɵtemplate(4, CalendarFooterComponent_div_0_button_4_Template, 2, 0, "button", 5);
    ɵngcc0.ɵɵelementStart(5, "button", 6);
    ɵngcc0.ɵɵlistener("click", function CalendarFooterComponent_div_0_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onClear(); });
    ɵngcc0.ɵɵtext(6, " \u6E05\u9664 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isCanTime);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isShowTime);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.isCanTime);
} }
export class CalendarFooterComponent {
    constructor(_elementRef, _renderer, _viewContainerRef) {
        this.showTime = false;
        this.mustShowTime = false;
        this.selectTime = new EventEmitter();
        this.clickOk = new EventEmitter();
        this.clickRemove = new EventEmitter();
        this.showTimePickerChange = new EventEmitter();
        this.isShowTime = false;
        this.isCanTime = false;
    }
    ngOnInit() {
        this._initTimeShowMode();
        if (!this.value) {
            this.value = new TinyDate();
        }
    }
    onSelectTime(date) {
        this.selectTime.emit(new TinyDate(date));
    }
    onTimeOk() {
        this.selectTime.emit(this.value);
        this.clickOk.emit();
    }
    onClear() {
        this.value = null;
        this.clickRemove.emit();
    }
    changeTimeShowMode(type) {
        switch (type) {
            case 'can':
                this.isCanTime = true;
                this.isShowTime = false;
                break;
            case 'show':
                this.isCanTime = false;
                this.isShowTime = true;
                break;
        }
        this.showTimePickerChange.emit(this.isShowTime);
    }
    _initTimeShowMode() {
        if (this.mustShowTime) {
            this.changeTimeShowMode('show');
        }
        else {
            if (this.showTime) {
                this.changeTimeShowMode('can');
            }
        }
    }
}
CalendarFooterComponent.ɵfac = function CalendarFooterComponent_Factory(t) { return new (t || CalendarFooterComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
CalendarFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CalendarFooterComponent, selectors: [["calendar-footer"]], inputs: { showTime: "showTime", mustShowTime: "mustShowTime", value: "value" }, outputs: { selectTime: "selectTime", clickOk: "clickOk", clickRemove: "clickRemove", showTimePickerChange: "showTimePickerChange" }, exportAs: ["calendarFooter"], decls: 1, vars: 1, consts: [["class", "time-picker-section", 3, "click", 4, "ngIf"], [1, "time-picker-section", 3, "click"], ["href", "javascript:;", "class", "link-has-icon time-picker-set-btn", 3, "click", 4, "ngIf"], ["class", "time-picker-wrap", 3, "showMeridian", "ngModel", "ngModelChange", 4, "ngIf"], [1, "time-picker-btn-wrap"], ["class", "time-picker-ok-btn", "thyButton", "primary", "thySize", "sm", 3, "click", 4, "ngIf"], ["thyButton", "link-secondary", "thySize", "sm", 1, "time-picker-clear-btn", 3, "click"], ["href", "javascript:;", 1, "link-has-icon", "time-picker-set-btn", 3, "click"], ["thyIconName", "clock-circle"], [1, "time-picker-wrap", 3, "showMeridian", "ngModel", "ngModelChange"], ["thyButton", "primary", "thySize", "sm", 1, "time-picker-ok-btn", 3, "click"]], template: function CalendarFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, CalendarFooterComponent_div_0_Template, 7, 3, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isShowTime || ctx.isCanTime);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyButtonComponent, ɵngcc3.ThyIconComponent, ɵngcc4.ThyTimePickerComponent, ɵngcc5.NgControlStatus, ɵngcc5.NgModel], encapsulation: 2, changeDetection: 0 });
CalendarFooterComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: ViewContainerRef }
];
CalendarFooterComponent.propDecorators = {
    showTime: [{ type: Input }],
    mustShowTime: [{ type: Input }],
    value: [{ type: Input }],
    selectTime: [{ type: Output }],
    clickOk: [{ type: Output }],
    clickRemove: [{ type: Output }],
    showTimePickerChange: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CalendarFooterComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                selector: 'calendar-footer',
                exportAs: 'calendarFooter',
                template: "<div *ngIf=\"isShowTime || isCanTime\" class=\"time-picker-section\" (click)=\"$event.stopPropagation()\">\n  <a\n    href=\"javascript:;\"\n    *ngIf=\"isCanTime\"\n    class=\"link-has-icon time-picker-set-btn\"\n    (click)=\"changeTimeShowMode('show')\"\n  >\n    <thy-icon thyIconName=\"clock-circle\"></thy-icon>\u8BBE\u7F6E\u65F6\u95F4\n  </a>\n  <thy-time-picker\n    *ngIf=\"isShowTime\"\n    class=\"time-picker-wrap\"\n    [showMeridian]=\"false\"\n    [ngModel]=\"value?.nativeDate\"\n    (ngModelChange)=\"onSelectTime($event)\"\n  ></thy-time-picker>\n  <div class=\"time-picker-btn-wrap\">\n    <button *ngIf=\"!isCanTime\" class=\"time-picker-ok-btn\" thyButton=\"primary\" thySize=\"sm\" (click)=\"onTimeOk()\">\n      \u786E\u5B9A\n    </button>\n    <button class=\"time-picker-clear-btn\" thyButton=\"link-secondary\" thySize=\"sm\" (click)=\"onClear()\">\n      \u6E05\u9664\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ViewContainerRef }]; }, { showTime: [{
            type: Input
        }], mustShowTime: [{
            type: Input
        }], selectTime: [{
            type: Output
        }], clickOk: [{
            type: Output
        }], clickRemove: [{
            type: Output
        }], showTimePickerChange: [{
            type: Output
        }], value: [{
            type: Input
        }] }); })();
