// ngx-tethys/esm2015/autocomplete/overlay/autocomplete-container.component.js
import { Component, ViewChild, Inject, ElementRef, EventEmitter, ChangeDetectorRef, NgZone } from '@angular/core';
import { CdkPortalOutlet } from '@angular/cdk/portal';
import { DOCUMENT } from '@angular/common';
import { ThyAutocompleteConfig } from './autocomplete.config';
import { thyAutocompleteAnimations } from './autocomplete-animations';
import { ThyAbstractOverlayContainer } from 'ngx-tethys/core';
import { autocompleteUpperOverlayOptions } from './autocomplete.options';
import { filter } from 'rxjs/operators';
import { ThyClickDispatcher } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './autocomplete.config';
import * as ɵngcc2 from 'ngx-tethys/core';
import * as ɵngcc3 from '@angular/cdk/portal';

function ThyAutocompleteContainerComponent_ng_template_0_Template(rf, ctx) { }
export class ThyAutocompleteContainerComponent extends ThyAbstractOverlayContainer {
    constructor(elementRef, document, config, changeDetectorRef, thyClickDispatcher, ngZone) {
        super(autocompleteUpperOverlayOptions, changeDetectorRef);
        this.elementRef = elementRef;
        this.document = document;
        this.config = config;
        this.thyClickDispatcher = thyClickDispatcher;
        this.ngZone = ngZone;
        /** State of the autocomplete animation. */
        this.animationState = 'enter';
        /** Emits when an animation state changes. */
        this.animationStateChanged = new EventEmitter();
        this.animationOpeningDone = this.animationStateChanged.pipe(filter((event) => {
            return event.phaseName === 'done' && event.toState === 'enter';
        }));
        this.animationClosingDone = this.animationStateChanged.pipe(filter((event) => {
            return event.phaseName === 'done' && event.toState === 'exit';
        }));
    }
    beforeAttachPortal() { }
    ngAfterViewInit() { }
    /** Callback, invoked whenever an animation on the host completes. */
    onAnimationDone(event) {
        this.animationStateChanged.emit(event);
    }
    /** Callback, invoked when an animation on the host starts. */
    onAnimationStart(event) {
        this.animationStateChanged.emit(event);
    }
    startExitAnimation() {
        this.animationState = 'exit';
        this.changeDetectorRef.markForCheck();
    }
}
ThyAutocompleteContainerComponent.ɵfac = function ThyAutocompleteContainerComponent_Factory(t) { return new (t || ThyAutocompleteContainerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyAutocompleteConfig), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyClickDispatcher), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyAutocompleteContainerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyAutocompleteContainerComponent, selectors: [["thy-autocomplete-container"]], viewQuery: function ThyAutocompleteContainerComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(CdkPortalOutlet, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.portalOutlet = _t.first);
    } }, hostAttrs: ["tabindex", "-1", 1, "thy-autocomplete-container"], hostVars: 2, hostBindings: function ThyAutocompleteContainerComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵsyntheticHostListener("@autocompleteContainer.start", function ThyAutocompleteContainerComponent_animation_autocompleteContainer_start_HostBindingHandler($event) { return ctx.onAnimationStart($event); })("@autocompleteContainer.done", function ThyAutocompleteContainerComponent_animation_autocompleteContainer_done_HostBindingHandler($event) { return ctx.onAnimationDone($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("role", "autocomplete");
        ɵngcc0.ɵɵsyntheticHostProperty("@autocompleteContainer", ctx.animationState);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 0, consts: [["cdkPortalOutlet", ""]], template: function ThyAutocompleteContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyAutocompleteContainerComponent_ng_template_0_Template, 0, 0, "ng-template", 0);
    } }, directives: [ɵngcc3.CdkPortalOutlet], encapsulation: 2, data: { animation: [thyAutocompleteAnimations.autocompleteContainer] } });
ThyAutocompleteContainerComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ThyAutocompleteConfig },
    { type: ChangeDetectorRef },
    { type: ThyClickDispatcher },
    { type: NgZone }
];
ThyAutocompleteContainerComponent.propDecorators = {
    portalOutlet: [{ type: ViewChild, args: [CdkPortalOutlet, { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAutocompleteContainerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-autocomplete-container',
                template: "<ng-template cdkPortalOutlet></ng-template>\n",
                animations: [thyAutocompleteAnimations.autocompleteContainer],
                host: {
                    class: 'thy-autocomplete-container',
                    tabindex: '-1',
                    '[attr.role]': `'autocomplete'`,
                    '[@autocompleteContainer]': 'animationState',
                    '(@autocompleteContainer.start)': 'onAnimationStart($event)',
                    '(@autocompleteContainer.done)': 'onAnimationDone($event)'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc1.ThyAutocompleteConfig }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc2.ThyClickDispatcher }, { type: ɵngcc0.NgZone }]; }, { portalOutlet: [{
            type: ViewChild,
            args: [CdkPortalOutlet, { static: true }]
        }] }); })();
