// ngx-tethys/esm2015/table/table.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { ThyTableComponent } from './table.component';
import { ThyTableColumnComponent } from './table-column.component';
import { ThyPaginationModule } from 'ngx-tethys/pagination';
import { ThySwitchModule } from 'ngx-tethys/switch';
import { ThyLoadingModule } from 'ngx-tethys/loading';
import { ThyEmptyModule } from 'ngx-tethys/empty';
import { TableIsValidModelValuePipe } from './table.pipe';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import { DragDropModule } from '@angular/cdk/drag-drop';
import * as ɵngcc0 from '@angular/core';
export class ThyTableModule {
}
ThyTableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTableModule });
ThyTableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTableModule_Factory(t) { return new (t || ThyTableModule)(); }, imports: [[
            CommonModule,
            FormsModule,
            ThyPaginationModule,
            ThySwitchModule,
            ThyLoadingModule,
            ThyEmptyModule,
            ThySharedModule,
            ThyIconModule,
            DragDropModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTableModule, { declarations: function () { return [ThyTableComponent, ThyTableColumnComponent, TableIsValidModelValuePipe]; }, imports: function () { return [CommonModule,
        FormsModule,
        ThyPaginationModule,
        ThySwitchModule,
        ThyLoadingModule,
        ThyEmptyModule,
        ThySharedModule,
        ThyIconModule,
        DragDropModule]; }, exports: function () { return [ThyTableComponent, ThyTableColumnComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyTableComponent, ThyTableColumnComponent, TableIsValidModelValuePipe],
                imports: [
                    CommonModule,
                    FormsModule,
                    ThyPaginationModule,
                    ThySwitchModule,
                    ThyLoadingModule,
                    ThyEmptyModule,
                    ThySharedModule,
                    ThyIconModule,
                    DragDropModule
                ],
                exports: [ThyTableComponent, ThyTableColumnComponent]
            }]
    }], null, null); })();
