// ngx-tethys/esm2015/notify/notify-queue.store.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Store, Action } from 'ngx-tethys/store';
import * as ɵngcc0 from '@angular/core';
export const notifyQueueInitialState = {
    topLeftQueue: [],
    topRightQueue: [],
    bottomLeftQueue: [],
    bottomRightQueue: []
};
export class NotifyQueueStore extends Store {
    constructor() {
        super(notifyQueueInitialState);
    }
    static topRightSelector(state) {
        return state.topRightQueue;
    }
    static topLeftSelector(state) {
        return state.topLeftQueue;
    }
    static bottomLeftSelector(state) {
        return state.bottomLeftQueue;
    }
    static bottomRightSelector(state) {
        return state.bottomRightQueue;
    }
    convertQueueKey(placement) {
        let key;
        if (placement === 'topLeft') {
            key = 'topLeftQueue';
        }
        else if (placement === 'topRight') {
            key = 'topRightQueue';
        }
        else if (placement === 'bottomLeft') {
            key = 'bottomLeftQueue';
        }
        else if (placement === 'bottomRight') {
            key = 'bottomRightQueue';
        }
        return key;
    }
    addNotify(placement, options) {
        const key = this.convertQueueKey(placement);
        const state = this.snapshot;
        if (state[key].length > options.maxStack) {
            state[key].shift();
        }
        state[key].push(options);
        this.next(state);
    }
    removeNotify(id, placement) {
        const state = this.snapshot;
        if (placement) {
            const queueKey = this.convertQueueKey(placement);
            if (state.hasOwnProperty(queueKey) && state[queueKey].length) {
                state[queueKey] = state[queueKey].filter((item) => {
                    return item.id !== id;
                });
            }
        }
        else {
            Object.keys(state).forEach(queueKey => {
                if (state.hasOwnProperty(queueKey) && state[queueKey].length) {
                    state[queueKey] = state[queueKey].filter((item) => {
                        return item.id !== id;
                    });
                }
            });
        }
        this.next(state);
    }
}
NotifyQueueStore.ɵfac = function NotifyQueueStore_Factory(t) { return new (t || NotifyQueueStore)(); };
NotifyQueueStore.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NotifyQueueStore, factory: NotifyQueueStore.ɵfac });
NotifyQueueStore.ctorParameters = () => [];
__decorate([
    Action(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", void 0)
], NotifyQueueStore.prototype, "addNotify", null);
__decorate([
    Action(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Number, String]),
    __metadata("design:returntype", void 0)
], NotifyQueueStore.prototype, "removeNotify", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotifyQueueStore, [{
        type: Injectable
    }], function () { return []; }, null); })();
