// ngx-tethys/esm2015/notify/notify.component.js
import { Component, Input, HostBinding, HostListener, NgZone } from '@angular/core';
import { trigger, state, style, animate, transition } from '@angular/animations';
import { UpdateHostClassService } from 'ngx-tethys/core';
import { NotifyQueueStore } from './notify-queue.store';
import { helpers } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './notify-queue.store';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-tethys/icon';

function ThyNotifyComponent_ng_container_0_div_8_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 11);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_div_8_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(3); return ctx_r5.triggerDetail(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r4.option == null ? null : ctx_r4.option.detail["link"]);
} }
const _c0 = function (a0) { return { "thy-notify-content--extend": a0 }; };
function ThyNotifyComponent_ng_container_0_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_div_8_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.extendContent(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, ThyNotifyComponent_ng_container_0_div_8_a_2_Template, 2, 1, "a", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx_r2.extendContentClass === true));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.option == null ? null : ctx_r2.option.content, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.option == null ? null : ctx_r2.option.detail);
} }
function ThyNotifyComponent_ng_container_0_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.option == null ? null : ctx_r3.option.detail["content"]);
} }
function ThyNotifyComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.closeNotify(); });
    ɵngcc0.ɵɵelement(2, "thy-icon", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 3);
    ɵngcc0.ɵɵelement(4, "thy-icon", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 5);
    ɵngcc0.ɵɵelementStart(6, "div", 6);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(8, ThyNotifyComponent_ng_container_0_div_8_Template, 3, 5, "div", 7);
    ɵngcc0.ɵɵtemplate(9, ThyNotifyComponent_ng_container_0_div_9_Template, 2, 1, "div", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.notifyIconName);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.option == null ? null : ctx_r0.option.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.option == null ? null : ctx_r0.option.detail) || (ctx_r0.option == null ? null : ctx_r0.option.content));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isShowDetail);
} }
function ThyNotifyComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_1_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.closeNotify(); });
    ɵngcc0.ɵɵelement(2, "thy-icon", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵelement(4, "template", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.option == null ? null : ctx_r1.option.html);
} }
const ANIMATION_IN_DURATION = 100;
const ANIMATION_OUT_DURATION = 150;
const HIDE_STYLE = { transform: 'translateX(0)', opacity: 0, height: 0, paddingTop: 0, paddingBottom: 0, margin: 0 };
export class ThyNotifyComponent {
    constructor(_queueStore, _ngZone) {
        this._queueStore = _queueStore;
        this._ngZone = _ngZone;
        this.className = '';
        this.notifyIconName = '';
        this.extendContentClass = false;
        this.isShowDetail = false;
    }
    set thyOption(value) {
        this.option = value;
        const type = value.type;
        this.placement = value.placement || 'topRight';
        if (this.placement === 'topLeft' || this.placement === 'bottomLeft') {
            this.flyInOut = 'flyInOutLeft';
        }
        else {
            this.flyInOut = 'flyInOutRight';
        }
        this.className = `thy-notify thy-notify-${type}`;
    }
    ngOnInit() {
        const iconName = {
            success: 'check-circle-fill',
            info: 'info-circle-fill',
            warning: 'waring-fill',
            error: 'close-circle-fill'
        };
        this.notifyIconName = iconName[this.option.type];
        this._creatCloseTimer();
    }
    ngOnDestroy() {
        this._clearCloseTimer();
    }
    extendContent() {
        this.extendContentClass = true;
    }
    showDetailToggle() {
        this.isShowDetail = !this.isShowDetail;
    }
    closeNotify() {
        this._ngZone.runOutsideAngular(() => {
            this.flyInOut = 'componentHide';
            setTimeout(() => {
                this._queueStore.removeNotify(this.option.id, this.placement);
            }, ANIMATION_OUT_DURATION);
        });
    }
    triggerDetail() {
        if (helpers.isFunction(this.option.detail.action)) {
            this.option.detail.action();
        }
        if (this.option.detail.content) {
            this.showDetailToggle();
        }
    }
    mouseenter() {
        if (this.option.pauseOnHover) {
            this._clearCloseTimer();
        }
    }
    mouseleave() {
        if (this.option.pauseOnHover) {
            this._creatCloseTimer();
        }
    }
    _creatCloseTimer() {
        if (this.option.duration) {
            this.closeTimer = setInterval(() => {
                clearInterval(this.closeTimer);
                this.closeNotify();
            }, this.option.duration);
        }
    }
    _clearCloseTimer() {
        clearInterval(this.closeTimer);
    }
}
ThyNotifyComponent.ɵfac = function ThyNotifyComponent_Factory(t) { return new (t || ThyNotifyComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NotifyQueueStore), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyNotifyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNotifyComponent, selectors: [["thy-notify"]], hostVars: 3, hostBindings: function ThyNotifyComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mouseenter", function ThyNotifyComponent_mouseenter_HostBindingHandler() { return ctx.mouseenter(); })("mouseleave", function ThyNotifyComponent_mouseleave_HostBindingHandler() { return ctx.mouseleave(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵsyntheticHostProperty("@flyInOut", ctx.flyInOut);
        ɵngcc0.ɵɵclassMap(ctx.className);
    } }, inputs: { thyOption: "thyOption" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], decls: 2, vars: 2, consts: [[4, "ngIf"], ["href", "javascript:;", 1, "thy-notify-close", 3, "click"], ["thyIconName", "close"], [1, "thy-notify-icon-container"], [3, "thyIconName"], [1, "thy-notify-main"], [1, "thy-notify-title"], ["class", "thy-notify-content", 3, "ngClass", "click", 4, "ngIf"], ["class", "thy-notify-detail", 4, "ngIf"], [1, "thy-notify-content", 3, "ngClass", "click"], ["href", "javascript:;", "class", "link-secondary", 3, "click", 4, "ngIf"], ["href", "javascript:;", 1, "link-secondary", 3, "click"], [1, "thy-notify-detail"], [3, "ngTemplateOutlet"]], template: function ThyNotifyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyNotifyComponent_ng_container_0_Template, 10, 4, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, ThyNotifyComponent_ng_container_1_Template, 5, 1, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.option.html);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.option == null ? null : ctx.option.html);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.ThyIconComponent, ɵngcc2.NgClass, ɵngcc2.NgTemplateOutlet], encapsulation: 2, data: { animation: [
            trigger('flyInOut', [
                state('flyInOutRight', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                transition('void => flyInOutRight', [
                    style({ transform: 'translateX(100%)', opacity: 0, height: '*' }),
                    animate(ANIMATION_IN_DURATION)
                ]),
                transition('flyInOutRight => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                transition('flyInOutRight => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                state('flyInOutLeft', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                transition('void => flyInOutLeft', [
                    style({ transform: 'translateX(-100%)', opacity: 0, height: '*' }),
                    animate(ANIMATION_IN_DURATION)
                ]),
                transition('flyInOutLeft => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                transition('flyInOutLeft => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                state('componentHide', style(HIDE_STYLE))
            ])
        ] } });
ThyNotifyComponent.ctorParameters = () => [
    { type: NotifyQueueStore },
    { type: NgZone }
];
ThyNotifyComponent.propDecorators = {
    flyInOut: [{ type: HostBinding, args: ['@flyInOut',] }],
    className: [{ type: HostBinding, args: ['class',] }],
    thyOption: [{ type: Input }],
    mouseenter: [{ type: HostListener, args: ['mouseenter',] }],
    mouseleave: [{ type: HostListener, args: ['mouseleave',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-notify',
                template: "<ng-container *ngIf=\"!option.html\">\n  <a href=\"javascript:;\" class=\"thy-notify-close\" (click)=\"closeNotify()\">\n    <thy-icon thyIconName=\"close\"></thy-icon>\n  </a>\n  <div class=\"thy-notify-icon-container\">\n    <thy-icon [thyIconName]=\"notifyIconName\"></thy-icon>\n  </div>\n  <div class=\"thy-notify-main\">\n    <div class=\"thy-notify-title\">{{ option?.title }}</div>\n    <div\n      *ngIf=\"option?.detail || option?.content\"\n      class=\"thy-notify-content\"\n      [ngClass]=\"{ 'thy-notify-content--extend': extendContentClass === true }\"\n      (click)=\"extendContent()\"\n    >\n      {{ option?.content }}\n      <a href=\"javascript:;\" *ngIf=\"option?.detail\" class=\"link-secondary\" (click)=\"triggerDetail()\">{{ (option?.detail)['link'] }}</a>\n    </div>\n    <div class=\"thy-notify-detail\" *ngIf=\"isShowDetail\">{{ (option?.detail)['content'] }}</div>\n  </div>\n</ng-container>\n<ng-container *ngIf=\"option?.html\">\n  <a href=\"javascript:;\" class=\"thy-notify-close\" (click)=\"closeNotify()\">\n    <thy-icon thyIconName=\"close\"></thy-icon>\n  </a>\n  <div class=\"thy-notify-main\">\n    <template [ngTemplateOutlet]=\"option?.html\"></template>\n  </div>\n</ng-container>\n",
                providers: [UpdateHostClassService],
                animations: [
                    trigger('flyInOut', [
                        state('flyInOutRight', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                        transition('void => flyInOutRight', [
                            style({ transform: 'translateX(100%)', opacity: 0, height: '*' }),
                            animate(ANIMATION_IN_DURATION)
                        ]),
                        transition('flyInOutRight => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                        transition('flyInOutRight => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                        state('flyInOutLeft', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                        transition('void => flyInOutLeft', [
                            style({ transform: 'translateX(-100%)', opacity: 0, height: '*' }),
                            animate(ANIMATION_IN_DURATION)
                        ]),
                        transition('flyInOutLeft => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                        transition('flyInOutLeft => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                        state('componentHide', style(HIDE_STYLE))
                    ])
                ]
            }]
    }], function () { return [{ type: ɵngcc1.NotifyQueueStore }, { type: ɵngcc0.NgZone }]; }, { className: [{
            type: HostBinding,
            args: ['class']
        }], thyOption: [{
            type: Input
        }], flyInOut: [{
            type: HostBinding,
            args: ['@flyInOut']
        }], mouseenter: [{
            type: HostListener,
            args: ['mouseenter']
        }], mouseleave: [{
            type: HostListener,
            args: ['mouseleave']
        }] }); })();
