// ngx-tethys/esm2015/notify/notify.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyNotifyComponent } from './notify.component';
import { ThyNotifyContainerComponent } from './notify.container.component';
import { ThyNotifyService } from './notify.service';
import { ThyIconModule } from 'ngx-tethys/icon';
import { NotifyQueueStore } from './notify-queue.store';
import { THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER } from './notify-option.interface';
import * as ɵngcc0 from '@angular/core';
export class ThyNotifyModule {
}
ThyNotifyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyNotifyModule });
ThyNotifyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyNotifyModule_Factory(t) { return new (t || ThyNotifyModule)(); }, providers: [ThyNotifyService, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER], imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyNotifyModule, { declarations: function () { return [ThyNotifyComponent, ThyNotifyContainerComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyNotifyContainerComponent, ThyNotifyComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyNotifyComponent, ThyNotifyContainerComponent],
                entryComponents: [ThyNotifyContainerComponent],
                providers: [ThyNotifyService, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyNotifyContainerComponent, ThyNotifyComponent]
            }]
    }], null, null); })();
