// ngx-tethys/esm2015/notify/notify.container.component.js
import { Component, HostBinding } from '@angular/core';
import { NotifyQueueStore } from './notify-queue.store';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './notify-queue.store';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './notify.component';

function ThyNotifyContainerComponent_thy_notify_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r4 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r4);
} }
function ThyNotifyContainerComponent_thy_notify_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r5);
} }
function ThyNotifyContainerComponent_thy_notify_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r6 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r6);
} }
function ThyNotifyContainerComponent_thy_notify_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r7 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r7);
} }
export class ThyNotifyContainerComponent {
    constructor(queueStore) {
        this.queueStore = queueStore;
        this.className = true;
    }
    ngOnInit() {
        this.placement = this.initialState.placement;
        let queue$, queueKey;
        if (this.placement === 'bottomRight') {
            this.bottomRight = true;
        }
        else if (this.placement === 'bottomLeft') {
            this.bottomLeft = true;
        }
        else if (this.placement === 'topLeft') {
            this.topLeft = true;
        }
        else {
            this.topRight = true;
        }
        if (this.placement === 'topLeft') {
            queueKey = 'topLeftQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.topLeftSelector);
        }
        else if (this.placement === 'topRight') {
            queueKey = 'topRightQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.topRightSelector);
        }
        else if (this.placement === 'bottomLeft') {
            queueKey = 'bottomLeftQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.bottomLeftSelector);
        }
        else if (this.placement === 'bottomRight') {
            queueKey = 'bottomRightQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.bottomRightSelector);
        }
        queue$.subscribe(data => {
            this[queueKey] = data;
        });
    }
}
ThyNotifyContainerComponent.ɵfac = function ThyNotifyContainerComponent_Factory(t) { return new (t || ThyNotifyContainerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NotifyQueueStore)); };
ThyNotifyContainerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNotifyContainerComponent, selectors: [["thy-notify-container"]], hostVars: 10, hostBindings: function ThyNotifyContainerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-notify-root", ctx.className)("thy-notify-bottomRight", ctx.bottomRight)("thy-notify-bottomLeft", ctx.bottomLeft)("thy-notify-topLeft", ctx.topLeft)("thy-notify-topRight", ctx.topRight);
    } }, decls: 4, vars: 4, consts: [[3, "thyOption", 4, "ngFor", "ngForOf"], [3, "thyOption"]], template: function ThyNotifyContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyNotifyContainerComponent_thy_notify_0_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(1, ThyNotifyContainerComponent_thy_notify_1_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(2, ThyNotifyContainerComponent_thy_notify_2_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(3, ThyNotifyContainerComponent_thy_notify_3_Template, 1, 1, "thy-notify", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.topLeftQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.topRightQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bottomLeftQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bottomRightQueue);
    } }, directives: [ɵngcc2.NgForOf, ɵngcc3.ThyNotifyComponent], encapsulation: 2 });
ThyNotifyContainerComponent.ctorParameters = () => [
    { type: NotifyQueueStore }
];
ThyNotifyContainerComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class.thy-notify-root',] }],
    bottomRight: [{ type: HostBinding, args: ['class.thy-notify-bottomRight',] }],
    bottomLeft: [{ type: HostBinding, args: ['class.thy-notify-bottomLeft',] }],
    topLeft: [{ type: HostBinding, args: ['class.thy-notify-topLeft',] }],
    topRight: [{ type: HostBinding, args: ['class.thy-notify-topRight',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyContainerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-notify-container',
                template: "<thy-notify *ngFor=\"let item of topLeftQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of topRightQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of bottomLeftQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of bottomRightQueue\" [thyOption]=\"item\"></thy-notify>\n"
            }]
    }], function () { return [{ type: ɵngcc1.NotifyQueueStore }]; }, { className: [{
            type: HostBinding,
            args: ['class.thy-notify-root']
        }], bottomRight: [{
            type: HostBinding,
            args: ['class.thy-notify-bottomRight']
        }], bottomLeft: [{
            type: HostBinding,
            args: ['class.thy-notify-bottomLeft']
        }], topLeft: [{
            type: HostBinding,
            args: ['class.thy-notify-topLeft']
        }], topRight: [{
            type: HostBinding,
            args: ['class.thy-notify-topRight']
        }] }); })();
