// ngx-tethys/esm2015/uploader/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { ThyFileSelectComponent } from './file-select.component';
import { ThyFileDropComponent } from './file-drop.component';
import { ThyUploaderService } from './uploader.service';
import { THY_UPLOADER_DEFAULT_OPTIONS_PROVIDER } from './uploader.config';
import { FileSelectBaseComponent } from './file-select-base';
// import { ThyDirectiveModule } from 'ngx-tethys/directive';
import * as ɵngcc0 from '@angular/core';
export class ThyUploaderModule {
}
ThyUploaderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyUploaderModule });
ThyUploaderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyUploaderModule_Factory(t) { return new (t || ThyUploaderModule)(); }, providers: [ThyUploaderService, THY_UPLOADER_DEFAULT_OPTIONS_PROVIDER], imports: [[CommonModule, HttpClientModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyUploaderModule, { declarations: function () { return [ThyFileSelectComponent, ThyFileDropComponent, FileSelectBaseComponent]; }, imports: function () { return [CommonModule, HttpClientModule]; }, exports: function () { return [ThyFileSelectComponent, ThyFileDropComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyUploaderModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyFileSelectComponent, ThyFileDropComponent, FileSelectBaseComponent],
                imports: [CommonModule, HttpClientModule],
                entryComponents: [ThyFileDropComponent],
                providers: [ThyUploaderService, THY_UPLOADER_DEFAULT_OPTIONS_PROVIDER],
                exports: [ThyFileSelectComponent, ThyFileDropComponent]
            }]
    }], null, null); })();
