// ngx-tethys/esm2015/tooltip/tooltip.service.js
import { NgZone, Injectable, Inject } from '@angular/core';
import { Overlay, ScrollDispatcher } from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { FocusMonitor } from '@angular/cdk/a11y';
import { ThyTooltipDirective } from './tooltip.directive';
import { ThyTooltipConfig, THY_TOOLTIP_DEFAULT_CONFIG_TOKEN } from './tooltip.config';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/cdk/overlay';
import * as ɵngcc2 from '@angular/cdk/platform';
import * as ɵngcc3 from '@angular/cdk/a11y';
import * as ɵngcc4 from './tooltip.config';
export class TooltipService {
    constructor(overlay, scrollDispatcher, ngZone, platform, focusMonitor, thyTooltipConfig) {
        this.overlay = overlay;
        this.scrollDispatcher = scrollDispatcher;
        this.ngZone = ngZone;
        this.platform = platform;
        this.focusMonitor = focusMonitor;
        this.thyTooltipConfig = thyTooltipConfig;
    }
    attach(elementRef, viewContainerRef, trigger) {
        this.thyTooltipDirective = new ThyTooltipDirective(this.overlay, elementRef, this.scrollDispatcher, viewContainerRef, this.ngZone, this.platform, this.focusMonitor, this.thyTooltipConfig);
        if (trigger) {
            this.thyTooltipDirective.trigger = trigger;
        }
        this.thyTooltipDirective.ngOnInit();
    }
    detach() {
        this.thyTooltipDirective.ngOnDestroy();
    }
}
TooltipService.ɵfac = function TooltipService_Factory(t) { return new (t || TooltipService)(ɵngcc0.ɵɵinject(ɵngcc1.Overlay), ɵngcc0.ɵɵinject(ɵngcc1.ScrollDispatcher), ɵngcc0.ɵɵinject(ɵngcc0.NgZone), ɵngcc0.ɵɵinject(ɵngcc2.Platform), ɵngcc0.ɵɵinject(ɵngcc3.FocusMonitor), ɵngcc0.ɵɵinject(THY_TOOLTIP_DEFAULT_CONFIG_TOKEN)); };
TooltipService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: TooltipService, factory: TooltipService.ɵfac });
TooltipService.ctorParameters = () => [
    { type: Overlay },
    { type: ScrollDispatcher },
    { type: NgZone },
    { type: Platform },
    { type: FocusMonitor },
    { type: ThyTooltipConfig, decorators: [{ type: Inject, args: [THY_TOOLTIP_DEFAULT_CONFIG_TOKEN,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TooltipService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Overlay }, { type: ɵngcc1.ScrollDispatcher }, { type: ɵngcc0.NgZone }, { type: ɵngcc2.Platform }, { type: ɵngcc3.FocusMonitor }, { type: ɵngcc4.ThyTooltipConfig, decorators: [{
                type: Inject,
                args: [THY_TOOLTIP_DEFAULT_CONFIG_TOKEN]
            }] }]; }, null); })();
