// ngx-tethys/esm2015/tooltip/tooltip.module.js
import { OverlayModule } from '@angular/cdk/overlay';
import { A11yModule } from '@angular/cdk/a11y';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyTooltipDirective } from './tooltip.directive';
import { ThyTooltipComponent } from './tooltip.component';
import { THY_TOOLTIP_DEFAULT_CONFIG_PROVIDER } from './tooltip.config';
import * as ɵngcc0 from '@angular/core';
export class ThyTooltipModule {
}
ThyTooltipModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTooltipModule });
ThyTooltipModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTooltipModule_Factory(t) { return new (t || ThyTooltipModule)(); }, providers: [THY_TOOLTIP_DEFAULT_CONFIG_PROVIDER], imports: [[A11yModule, CommonModule, OverlayModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTooltipModule, { declarations: function () { return [ThyTooltipDirective, ThyTooltipComponent]; }, imports: function () { return [A11yModule, CommonModule, OverlayModule]; }, exports: function () { return [ThyTooltipDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTooltipModule, [{
        type: NgModule,
        args: [{
                imports: [A11yModule, CommonModule, OverlayModule],
                exports: [ThyTooltipDirective],
                declarations: [ThyTooltipDirective, ThyTooltipComponent],
                entryComponents: [ThyTooltipComponent],
                providers: [THY_TOOLTIP_DEFAULT_CONFIG_PROVIDER]
            }]
    }], null, null); })();
