// ngx-tethys/fesm2015/ngx-tethys-pagination.js
import { InjectionToken, EventEmitter, Component, ChangeDetectionStrategy, Optional, Inject, ElementRef, ChangeDetectorRef, Input, Output, HostBinding, Pipe, NgModule } from '@angular/core';
import { UpdateHostClassService } from 'ngx-tethys/core';
import { isTemplateRef } from 'ngx-tethys/util';
import { CommonModule } from '@angular/common';
import { DomSanitizer } from '@angular/platform-browser';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-tethys/icon';
import * as ɵngcc4 from 'ngx-tethys/shared';
import * as ɵngcc5 from '@angular/platform-browser';

function ThyPaginationComponent_ng_container_0_div_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 10);
    ɵngcc0.ɵɵtext(2, " \u7B2C");
    ɵngcc0.ɵɵelementStart(3, "span", 11);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(5, "\u6761 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div");
    ɵngcc0.ɵɵtext(7, " \u5171\u8BA1");
    ɵngcc0.ɵɵelementStart(8, "span", 11);
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(10, "\u6761 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate2(" ", ctx_r9.range.from, "-", ctx_r9.range.to, " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r9.total, " ");
} }
function ThyPaginationComponent_ng_container_0_div_1_ng_template_2_ng_template_0_Template(rf, ctx) { }
const _c0 = function (a0, a1) { return { $implicit: a0, range: a1 }; };
function ThyPaginationComponent_ng_container_0_div_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyPaginationComponent_ng_container_0_div_1_ng_template_2_ng_template_0_Template, 0, 0, "ng-template", 12);
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r11.showTotal)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c0, ctx_r11.total, ctx_r11.range));
} }
function ThyPaginationComponent_ng_container_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵtemplate(1, ThyPaginationComponent_ng_container_0_div_1_ng_container_1_Template, 11, 3, "ng-container", 8);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_div_1_ng_template_2_Template, 1, 5, "ng-template", null, 9, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r10 = ɵngcc0.ɵɵreference(3);
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.isTemplateRef(ctx_r1.showTotal))("ngIfElse", _r10);
} }
function ThyPaginationComponent_ng_container_0_li_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r13.config.firstText);
} }
function ThyPaginationComponent_ng_container_0_li_4_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r14.config.firstIcon);
} }
function ThyPaginationComponent_ng_container_0_li_4_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_4_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.selectPage(ctx_r15.firstIndex); });
    ɵngcc0.ɵɵelementStart(1, "a", 14);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_4_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_4_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r2.disabled || ctx_r2.pageIndex === ctx_r2.firstIndex);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.config.firstText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.config.firstIcon);
} }
function ThyPaginationComponent_ng_container_0_li_5_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r17 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r17.config.previousText);
} }
function ThyPaginationComponent_ng_container_0_li_5_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r18 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r18.config.previousIcon);
} }
function ThyPaginationComponent_ng_container_0_li_5_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_5_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.selectPage(ctx_r19.pageIndex - 1); });
    ɵngcc0.ɵɵelementStart(1, "a", 16);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_5_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_5_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r3.disabled || ctx_r3.pageIndex === ctx_r3.firstIndex);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.config.previousText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.config.previousIcon);
} }
function ThyPaginationComponent_ng_container_0_li_6_Template(rf, ctx) { if (rf & 1) {
    const _r23 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_6_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r23); const page_r21 = ctx.$implicit; const ctx_r22 = ɵngcc0.ɵɵnextContext(2); return ctx_r22.selectPage(page_r21.index); });
    ɵngcc0.ɵɵelementStart(1, "a", 17);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const page_r21 = ctx.$implicit;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("active", page_r21.active)("disabled", ctx_r4.disabled);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(page_r21.text);
} }
function ThyPaginationComponent_ng_container_0_li_7_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r24 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r24.config.nextText);
} }
function ThyPaginationComponent_ng_container_0_li_7_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r25 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r25.config.nextIcon);
} }
function ThyPaginationComponent_ng_container_0_li_7_Template(rf, ctx) { if (rf & 1) {
    const _r27 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 18);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_7_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r27); const ctx_r26 = ɵngcc0.ɵɵnextContext(2); return ctx_r26.selectPage(ctx_r26.pageIndex + 1); });
    ɵngcc0.ɵɵelementStart(1, "a", 19);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_7_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_7_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r5.disabled || ctx_r5.pageIndex === ctx_r5.pageCount);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.config.nextText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.config.nextIcon);
} }
function ThyPaginationComponent_ng_container_0_li_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r28 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r28.config.lastText);
} }
function ThyPaginationComponent_ng_container_0_li_8_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r29 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r29.config.lastIcon);
} }
function ThyPaginationComponent_ng_container_0_li_8_Template(rf, ctx) { if (rf & 1) {
    const _r31 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 13);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_li_8_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r31); const ctx_r30 = ɵngcc0.ɵɵnextContext(2); return ctx_r30.selectPage(ctx_r30.pageCount); });
    ɵngcc0.ɵɵelementStart(1, "a", 20);
    ɵngcc0.ɵɵtemplate(2, ThyPaginationComponent_ng_container_0_li_8_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ThyPaginationComponent_ng_container_0_li_8_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("disabled", ctx_r6.disabled || ctx_r6.pageIndex === ctx_r6.pageCount);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.config.lastText);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.config.lastIcon);
} }
function ThyPaginationComponent_ng_container_0_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "div", 21);
    ɵngcc0.ɵɵpipe(2, "paginationTotalPagesFormat");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind2(2, 1, ctx_r7.pageCount, ctx_r7.config.totalPagesFormat), ɵngcc0.ɵɵsanitizeHtml);
} }
function ThyPaginationComponent_ng_container_0_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    const _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 22);
    ɵngcc0.ɵɵelementStart(2, "div", 23);
    ɵngcc0.ɵɵtext(3, " \u5230 ");
    ɵngcc0.ɵɵelementStart(4, "input", 24, 25);
    ɵngcc0.ɵɵlistener("thyEnter", function ThyPaginationComponent_ng_container_0_ng_container_10_Template_input_thyEnter_4_listener() { ɵngcc0.ɵɵrestoreView(_r34); const _r32 = ɵngcc0.ɵɵreference(5); const ctx_r33 = ɵngcc0.ɵɵnextContext(2); return ctx_r33.jumpPage(_r32); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(6, " \u9875 ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "button", 26);
    ɵngcc0.ɵɵlistener("click", function ThyPaginationComponent_ng_container_0_ng_container_10_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r34); const _r32 = ɵngcc0.ɵɵreference(5); const ctx_r35 = ɵngcc0.ɵɵnextContext(2); return ctx_r35.jumpPage(_r32); });
    ɵngcc0.ɵɵtext(8, "\u786E\u5B9A");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("disabled", ctx_r8.disabled);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("disabled", ctx_r8.disabled);
} }
function ThyPaginationComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyPaginationComponent_ng_container_0_div_1_Template, 4, 2, "div", 1);
    ɵngcc0.ɵɵelementStart(2, "div", 2);
    ɵngcc0.ɵɵelementStart(3, "ul", 3);
    ɵngcc0.ɵɵtemplate(4, ThyPaginationComponent_ng_container_0_li_4_Template, 4, 4, "li", 4);
    ɵngcc0.ɵɵtemplate(5, ThyPaginationComponent_ng_container_0_li_5_Template, 4, 4, "li", 4);
    ɵngcc0.ɵɵtemplate(6, ThyPaginationComponent_ng_container_0_li_6_Template, 3, 5, "li", 5);
    ɵngcc0.ɵɵtemplate(7, ThyPaginationComponent_ng_container_0_li_7_Template, 4, 4, "li", 6);
    ɵngcc0.ɵɵtemplate(8, ThyPaginationComponent_ng_container_0_li_8_Template, 4, 4, "li", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(9, ThyPaginationComponent_ng_container_0_ng_container_9_Template, 3, 4, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(10, ThyPaginationComponent_ng_container_0_ng_container_10_Template, 9, 2, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.showTotal);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.boundaryLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.directionLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.pages);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.directionLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.boundaryLinks);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.totalPagesFormat);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.showQuickJumper);
} }
const DEFAULT_RANGE_COUNT = 5;
const PaginationDefaultConfig = {
    boundaryLinks: false,
    directionLinks: true,
    pageSize: 20,
    maxCount: 7,
    rangeCount: DEFAULT_RANGE_COUNT,
    showQuickJumper: true,
    firstText: '第一页',
    lastText: '最后一页',
    previousText: '',
    nextText: '',
    firstIcon: '',
    lastIcon: '',
    previousIcon: 'angle-left',
    nextIcon: 'angle-right',
    totalPagesFormat: '共{total}页'
};
const THY_PAGINATION_CONFIG = new InjectionToken('THY_PAGINATION_CONFIG', {
    providedIn: 'root',
    factory: () => {
        return {};
    }
});

class ThyPaginationComponent {
    constructor(paginationConfig, updateHostClassService, elementRef, cdr) {
        this.paginationConfig = paginationConfig;
        this.updateHostClassService = updateHostClassService;
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.isTemplateRef = isTemplateRef;
        this.config = Object.assign({}, PaginationDefaultConfig, this.paginationConfig.main);
        this.disabled = false;
        this.pageIndexChange = new EventEmitter();
        this.pageChanged = new EventEmitter();
        this.pages = [];
        this.pageIndex = 1;
        this.range = { from: 0, to: 0 };
        this.firstIndex = 1;
        this.isHideOnSinglePage = false;
        this.initialized = false;
        this.isPaginationClass = true;
        // 是否显示范围和total
        this.showTotal = false;
        this.updateHostClassService.initializeElement(this.elementRef.nativeElement);
    }
    set thyPageIndex(pageIndex) {
        this.pageIndex = pageIndex;
        if (this.initialized) {
            this.setPageIndex(pageIndex);
        }
    }
    set thyPageSize(pageSize) {
        this.pageSize = pageSize;
        if (this.initialized) {
            this.calculatePageCount();
            this.initializePages(this.pageIndex, this.pageCount);
            this.cdr.markForCheck();
        }
    }
    set thyTotal(total) {
        this.total = total;
        if (this.initialized) {
            this.calculatePageCount();
            this.initializePages(this.pageIndex, this.pageCount);
            this.cdr.markForCheck();
        }
    }
    set showQuickJumper(value) {
        this.config.showQuickJumper = value;
    }
    set size(size) {
        this.updateHostClassService.addClass(`thy-pagination-${size}`);
    }
    set maxCount(value) {
        this.config.maxCount = value;
    }
    set thyRangeCount(value) {
        if (Number.isInteger(value)) {
            this.config.rangeCount = value;
            if (this.initialized) {
                this.setMarginalCount(value);
            }
        }
    }
    ngOnInit() {
        this.setMarginalCount(this.config.rangeCount);
        this.calculatePageCount();
        this.setPageIndex(this.pageIndex);
        this.initialized = true;
    }
    setMarginalCount(range) {
        if (!this.marginalCount) {
            this.marginalCount = range <= DEFAULT_RANGE_COUNT ? 1 : 2;
        }
    }
    setPageIndex(pageIndex) {
        this.pageIndex = pageIndex > this.pageCount ? this.pageCount : pageIndex || 1;
        this.range = {
            from: (this.pageIndex - 1) * this.pageSize + 1,
            to: this.pageIndex * this.pageSize
        };
        this.initializePages(this.pageIndex, this.pageCount);
        this.cdr.markForCheck();
    }
    calculatePageCount() {
        const pageCount = this.pageSize < 1 ? 1 : Math.ceil(this.total / this.pageSize);
        this.pageCount = Math.max(pageCount || 0, 1);
    }
    makePage(index, text, active) {
        return { index, text, active };
    }
    initializePages(pageIndex, pageCount) {
        const marginalCount = this.marginalCount;
        const rangeCount = this.config.rangeCount;
        const maxCount = this.config.maxCount;
        let pages = [];
        const isMaxSized = pageCount > maxCount;
        if (isMaxSized) {
            const beforePages = [];
            const afterPages = [];
            // beforePages
            for (let i = 1; i <= marginalCount; i++) {
                beforePages.push(this.makePage(i, i.toString(), i === pageIndex));
            }
            if (pageIndex - Math.ceil(rangeCount / 2) > this.firstIndex) {
                beforePages.push(this.makePage(pageIndex - rangeCount, '...', null));
            }
            // afterPages
            if (pageIndex + Math.ceil(rangeCount / 2) < pageCount) {
                afterPages.push(this.makePage(pageIndex + rangeCount, '...', null));
            }
            for (let i = pageCount - marginalCount + 1; i <= pageCount; i++) {
                afterPages.push(this.makePage(i, i.toString(), i === pageIndex));
            }
            // mainPages
            let start = Math.max(marginalCount + 1, pageIndex - (rangeCount - 1) / 2);
            let end = Math.min(pageIndex + (rangeCount - 1) / 2, pageCount - marginalCount);
            if (pageIndex - 1 <= marginalCount) {
                end = rangeCount;
            }
            if (pageCount - pageIndex <= marginalCount) {
                start = pageCount - rangeCount + 1;
            }
            for (let i = start; i <= end; i++) {
                pages.push({
                    index: i,
                    text: i.toString(),
                    active: i === +pageIndex
                });
            }
            pages = [...beforePages, ...pages, ...afterPages];
        }
        else {
            for (let i = 1; i <= pageCount; i++) {
                pages.push({
                    index: i,
                    text: i.toString(),
                    active: i === +pageIndex
                });
            }
        }
        this.pages = pages;
    }
    pageChange(pageIndex) {
        this.pageIndexChange.emit(pageIndex);
        this.pageChanged.emit({ page: pageIndex });
    }
    selectPage(pageIndex) {
        if (this.disabled || pageIndex === this.firstIndex - 1 || pageIndex === this.pageCount + 1) {
            return;
        }
        this.setPageIndex(pageIndex);
        this.pageChange(this.pageIndex);
    }
    jumpPage(input) {
        const pageIndex = +input.value;
        if (Number.isInteger(pageIndex)) {
            this.selectPage(pageIndex);
        }
        input.value = '';
    }
}
ThyPaginationComponent.ɵfac = function ThyPaginationComponent_Factory(t) { return new (t || ThyPaginationComponent)(ɵngcc0.ɵɵdirectiveInject(THY_PAGINATION_CONFIG, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
ThyPaginationComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyPaginationComponent, selectors: [["thy-pagination"]], hostVars: 4, hostBindings: function ThyPaginationComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-pagination", ctx.isPaginationClass)("thy-pagination-has-total", ctx.showTotal);
    } }, inputs: { disabled: ["thyDisabled", "disabled"], showTotal: ["thyShowTotal", "showTotal"], thyPageIndex: "thyPageIndex", thyPageSize: "thyPageSize", thyTotal: "thyTotal", showQuickJumper: ["thyShowQuickJumper", "showQuickJumper"], size: ["thySize", "size"], maxCount: ["thyMaxCount", "maxCount"], thyRangeCount: "thyRangeCount", marginalCount: ["thyMarginalCount", "marginalCount"], hideOnSinglePage: ["thyHideOnSinglePage", "hideOnSinglePage"] }, outputs: { pageIndexChange: "thyPageIndexChange", pageChanged: "thyPageChanged" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], decls: 1, vars: 1, consts: [[4, "ngIf"], ["class", "thy-pagination-total", 4, "ngIf"], [1, "thy-pagination-content"], [1, "thy-pagination-pages"], ["class", "thy-page-item", 3, "disabled", "click", 4, "ngIf"], ["class", "thy-page-item", 3, "active", "disabled", "click", 4, "ngFor", "ngForOf"], ["class", "thy-page-item page-item-next", 3, "disabled", "click", 4, "ngIf"], [1, "thy-pagination-total"], [4, "ngIf", "ngIfElse"], ["totalTemplate", ""], [1, "mr-3"], [1, "number"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "thy-page-item", 3, "click"], ["href", "javascript:;", 1, "thy-page-link", "thy-page-link-first"], [3, "thyIconName"], ["href", "javascript:;", 1, "thy-page-link", "page-link-pre"], ["href", "javascript:;", 1, "thy-page-link"], [1, "thy-page-item", "page-item-next", 3, "click"], ["href", "javascript:;", 1, "thy-page-link", "thy-page-link-next"], ["href", "javascript:;", 1, "thy-page-link", "thy-page-link-last"], [1, "thy-pagination-count", 3, "innerHTML"], [1, "thy-pagination-jumper"], [1, "thy-pagination-jumper-input"], ["type", "text", 3, "disabled", "thyEnter"], ["jumperInput", ""], ["type", "button", 3, "disabled", "click"]], template: function ThyPaginationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyPaginationComponent_ng_container_0_Template, 11, 8, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.hideOnSinglePage && ctx.total > ctx.pageSize || !ctx.hideOnSinglePage);
    } }, directives: function () { return [ɵngcc2.NgIf, ɵngcc2.NgForOf, ɵngcc2.NgTemplateOutlet, ɵngcc3.ThyIconComponent, ɵngcc4.ThyEnterDirective]; }, pipes: function () { return [PaginationTotalCountFormat]; }, encapsulation: 2, changeDetection: 0 });
ThyPaginationComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [THY_PAGINATION_CONFIG,] }] },
    { type: UpdateHostClassService },
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
ThyPaginationComponent.propDecorators = {
    thyPageIndex: [{ type: Input }],
    thyPageSize: [{ type: Input }],
    thyTotal: [{ type: Input }],
    disabled: [{ type: Input, args: ['thyDisabled',] }],
    showQuickJumper: [{ type: Input, args: ['thyShowQuickJumper',] }],
    size: [{ type: Input, args: ['thySize',] }],
    maxCount: [{ type: Input, args: ['thyMaxCount',] }],
    marginalCount: [{ type: Input, args: ['thyMarginalCount',] }],
    thyRangeCount: [{ type: Input }],
    hideOnSinglePage: [{ type: Input, args: ['thyHideOnSinglePage',] }],
    pageIndexChange: [{ type: Output, args: ['thyPageIndexChange',] }],
    pageChanged: [{ type: Output, args: ['thyPageChanged',] }],
    isPaginationClass: [{ type: HostBinding, args: ['class.thy-pagination',] }],
    showTotal: [{ type: HostBinding, args: ['class.thy-pagination-has-total',] }, { type: Input, args: ['thyShowTotal',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPaginationComponent, [{
        type: Component,
        args: [{
                selector: 'thy-pagination',
                template: "<ng-container *ngIf=\"(hideOnSinglePage && total > pageSize) || !hideOnSinglePage\">\n  <div class=\"thy-pagination-total\" *ngIf=\"showTotal\">\n    <ng-container *ngIf=\"!isTemplateRef(showTotal); else totalTemplate\">\n      <div class=\"mr-3\">\n        \u7B2C<span class=\"number\"> {{ range.from }}-{{ range.to }} </span>\u6761\n      </div>\n      <div>\n        \u5171\u8BA1<span class=\"number\"> {{ total }} </span>\u6761\n      </div>\n    </ng-container>\n    <ng-template #totalTemplate>\n      <ng-template\n        [ngTemplateOutlet]=\"showTotal\"\n        [ngTemplateOutletContext]=\"{ $implicit: total, range: range }\"\n      ></ng-template>\n    </ng-template>\n  </div>\n  <div class=\"thy-pagination-content\">\n    <ul class=\"thy-pagination-pages\">\n      <li\n        class=\"thy-page-item\"\n        [class.disabled]=\"disabled || pageIndex === firstIndex\"\n        *ngIf=\"config.boundaryLinks\"\n        (click)=\"selectPage(firstIndex)\"\n      >\n        <a class=\"thy-page-link thy-page-link-first\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.firstText\">{{ config.firstText }}</ng-container>\n          <ng-container *ngIf=\"config.firstIcon\">\n            <thy-icon [thyIconName]=\"config.firstIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n      <li\n        class=\"thy-page-item\"\n        [class.disabled]=\"disabled || pageIndex === firstIndex\"\n        *ngIf=\"config.directionLinks\"\n        (click)=\"selectPage(pageIndex - 1)\"\n      >\n        <a class=\"thy-page-link  page-link-pre\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.previousText\">{{ config.previousText }}</ng-container>\n          <ng-container *ngIf=\"config.previousIcon\">\n            <thy-icon [thyIconName]=\"config.previousIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n      <li\n        class=\"thy-page-item\"\n        [class.active]=\"page.active\"\n        [class.disabled]=\"disabled\"\n        (click)=\"selectPage(page.index)\"\n        *ngFor=\"let page of pages\"\n      >\n        <a class=\"thy-page-link\" href=\"javascript:;\">{{ page.text }}</a>\n      </li>\n      <li\n        class=\"thy-page-item page-item-next\"\n        [class.disabled]=\"disabled || pageIndex === pageCount\"\n        *ngIf=\"config.directionLinks\"\n        (click)=\"selectPage(pageIndex + 1)\"\n      >\n        <a class=\"thy-page-link thy-page-link-next\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.nextText\">{{ config.nextText }}</ng-container>\n          <ng-container *ngIf=\"config.nextIcon\">\n            <thy-icon [thyIconName]=\"config.nextIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n      <li\n        class=\"thy-page-item\"\n        [class.disabled]=\"disabled || pageIndex === pageCount\"\n        *ngIf=\"config.boundaryLinks\"\n        (click)=\"selectPage(pageCount)\"\n      >\n        <a class=\"thy-page-link thy-page-link-last\" href=\"javascript:;\">\n          <ng-container *ngIf=\"config.lastText\">{{ config.lastText }}</ng-container>\n          <ng-container *ngIf=\"config.lastIcon\">\n            <thy-icon [thyIconName]=\"config.lastIcon\"></thy-icon>\n          </ng-container>\n        </a>\n      </li>\n    </ul>\n    <ng-container *ngIf=\"config.totalPagesFormat\">\n      <div\n        class=\"thy-pagination-count\"\n        [innerHTML]=\"pageCount | paginationTotalPagesFormat: config.totalPagesFormat\"\n      ></div>\n    </ng-container>\n    <ng-container *ngIf=\"config.showQuickJumper\">\n      <div class=\"thy-pagination-jumper\">\n        <div class=\"thy-pagination-jumper-input\">\n          \u5230 <input type=\"text\" [disabled]=\"disabled\" #jumperInput (thyEnter)=\"jumpPage(jumperInput)\" /> \u9875\n        </div>\n        <button type=\"button\" (click)=\"jumpPage(jumperInput)\" [disabled]=\"disabled\">\u786E\u5B9A</button>\n      </div>\n    </ng-container>\n  </div>\n</ng-container>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [THY_PAGINATION_CONFIG]
            }] }, { type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }]; }, { disabled: [{
            type: Input,
            args: ['thyDisabled']
        }], pageIndexChange: [{
            type: Output,
            args: ['thyPageIndexChange']
        }], pageChanged: [{
            type: Output,
            args: ['thyPageChanged']
        }], isPaginationClass: [{
            type: HostBinding,
            args: ['class.thy-pagination']
        }], showTotal: [{
            type: HostBinding,
            args: ['class.thy-pagination-has-total']
        }, {
            type: Input,
            args: ['thyShowTotal']
        }], thyPageIndex: [{
            type: Input
        }], thyPageSize: [{
            type: Input
        }], thyTotal: [{
            type: Input
        }], showQuickJumper: [{
            type: Input,
            args: ['thyShowQuickJumper']
        }], size: [{
            type: Input,
            args: ['thySize']
        }], maxCount: [{
            type: Input,
            args: ['thyMaxCount']
        }], thyRangeCount: [{
            type: Input
        }], marginalCount: [{
            type: Input,
            args: ['thyMarginalCount']
        }], hideOnSinglePage: [{
            type: Input,
            args: ['thyHideOnSinglePage']
        }] }); })();

class PaginationTotalCountFormat {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    transform(count, format) {
        if (count && format) {
            return this.sanitizer.bypassSecurityTrustHtml(format.replace(/\{(.+?)\}/g, `<span>${count}</span>`));
        }
        else {
            return '';
        }
    }
}
PaginationTotalCountFormat.ɵfac = function PaginationTotalCountFormat_Factory(t) { return new (t || PaginationTotalCountFormat)(ɵngcc0.ɵɵdirectiveInject(ɵngcc5.DomSanitizer)); };
PaginationTotalCountFormat.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "paginationTotalPagesFormat", type: PaginationTotalCountFormat, pure: true });
PaginationTotalCountFormat.ctorParameters = () => [
    { type: DomSanitizer }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PaginationTotalCountFormat, [{
        type: Pipe,
        args: [{
                name: 'paginationTotalPagesFormat'
            }]
    }], function () { return [{ type: ɵngcc5.DomSanitizer }]; }, null); })();

class ThyPaginationModule {
}
ThyPaginationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyPaginationModule });
ThyPaginationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyPaginationModule_Factory(t) { return new (t || ThyPaginationModule)(); }, imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyPaginationModule, { declarations: function () { return [ThyPaginationComponent, PaginationTotalCountFormat]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyPaginationComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPaginationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyPaginationComponent],
                declarations: [ThyPaginationComponent, PaginationTotalCountFormat]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { DEFAULT_RANGE_COUNT, PaginationDefaultConfig, PaginationTotalCountFormat, THY_PAGINATION_CONFIG, ThyPaginationComponent, ThyPaginationModule };
