// ngx-tethys/fesm2015/ngx-tethys-menu.js
import { Component, HostBinding, EventEmitter, ViewChild, Input, Output, ElementRef, Renderer2, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { ThyPopover, ThyPopoverModule } from 'ngx-tethys/popover';
import { ThyIconModule } from 'ngx-tethys/icon';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
import * as ɵngcc2 from 'ngx-tethys/icon';
import * as ɵngcc3 from '@angular/common';

const _c0 = ["*"];
const _c1 = ["thyMenuGroup"];
function ThyMenuGroupComponent_thy_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 7);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.iconClass);
} }
function ThyMenuGroupComponent_thy_icon_5_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 8);
    ɵngcc0.ɵɵlistener("click", function ThyMenuGroupComponent_thy_icon_5_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onActionClick($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r1.rightIconClass);
} }
class ThyMenuComponent {
    constructor() {
        this.isThyMenu = true;
    }
    ngOnInit() { }
}
ThyMenuComponent.ɵfac = function ThyMenuComponent_Factory(t) { return new (t || ThyMenuComponent)(); };
ThyMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuComponent, selectors: [["thy-menu"], ["", "thy-menu", ""], ["", "thyMenu", ""]], hostVars: 2, hostBindings: function ThyMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu", ctx.isThyMenu);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuComponent.ctorParameters = () => [];
ThyMenuComponent.propDecorators = {
    isThyMenu: [{ type: HostBinding, args: ['class.thy-menu',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu,[thy-menu],[thyMenu]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenu: [{
            type: HostBinding,
            args: ['class.thy-menu']
        }] }); })();

class ThyMenuItemComponent {
    constructor() {
        this.isThyMenuItem = true;
    }
    ngOnInit() { }
}
ThyMenuItemComponent.ɵfac = function ThyMenuItemComponent_Factory(t) { return new (t || ThyMenuItemComponent)(); };
ThyMenuItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemComponent, selectors: [["thy-menu-item"], ["", "thy-menu-item", ""], ["", "thyMenuItem", ""]], hostVars: 2, hostBindings: function ThyMenuItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item", ctx.isThyMenuItem);
    } }, ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "thy-menu-item-content"]], template: function ThyMenuItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
ThyMenuItemComponent.ctorParameters = () => [];
ThyMenuItemComponent.propDecorators = {
    isThyMenuItem: [{ type: HostBinding, args: ['class.thy-menu-item',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item,[thy-menu-item],[thyMenuItem]',
                template: "<div class=\"thy-menu-item-content\">\n  <ng-content></ng-content>\n</div>\n"
            }]
    }], function () { return []; }, { isThyMenuItem: [{
            type: HostBinding,
            args: ['class.thy-menu-item']
        }] }); })();

class ThyMenuGroupComponent {
    constructor(popover) {
        this.popover = popover;
        this.rightIconClass = 'more';
        this.iconClass = 'folder-bold';
        this.groupHeaderPaddingLeft = 0;
        this.isThyMenuGroup = true;
        this.showIcon = false;
        this.isCollapsed = true;
        this.thyTitle = '';
        this.thyShowAction = false;
        this.thyActionStopPropagation = true;
        this.thyOnActionClick = new EventEmitter();
    }
    set thyExpand(value) {
        this.isCollapsed = !!!value;
    }
    set thyShowIcon(value) {
        this.showIcon = value;
    }
    set thyIcon(value) {
        this.iconClass = value;
    }
    set thyActionIcon(value) {
        this.rightIconClass = value;
    }
    set thyActionMenu(value) {
        this._actionMenu = value;
    }
    ngOnInit() { }
    collapseGroup() {
        this.isCollapsed = !this.isCollapsed;
    }
    onActionClick(event) {
        if (this.thyActionStopPropagation) {
            event.stopPropagation();
        }
        if (this._actionMenu) {
            this.popover.open(this._actionMenu, {
                origin: event.currentTarget,
                insideClosable: true,
                placement: 'bottom'
            });
        }
        else {
            this.thyOnActionClick.emit(event);
        }
    }
}
ThyMenuGroupComponent.ɵfac = function ThyMenuGroupComponent_Factory(t) { return new (t || ThyMenuGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover)); };
ThyMenuGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuGroupComponent, selectors: [["thy-menu-group"], ["", "thy-menu-group", ""], ["", "thyMenuGroup", ""]], viewQuery: function ThyMenuGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._thyMenuGroup = _t.first);
    } }, hostVars: 6, hostBindings: function ThyMenuGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-group", ctx.isThyMenuGroup)("has-icon", ctx.showIcon)("collapsed", ctx.isCollapsed);
    } }, inputs: { thyTitle: "thyTitle", thyShowAction: "thyShowAction", thyActionStopPropagation: "thyActionStopPropagation", thyExpand: "thyExpand", thyShowIcon: "thyShowIcon", thyIcon: "thyIcon", thyActionIcon: "thyActionIcon", thyActionMenu: "thyActionMenu" }, outputs: { thyOnActionClick: "thyOnActionClick" }, ngContentSelectors: _c0, decls: 9, vars: 5, consts: [[1, "thy-menu-group-header", 3, "click"], ["thyIconName", "caret-right-down", 1, "thy-menu-group-arrow"], ["class", "thy-menu-group-folder", 3, "thyIconName", 4, "ngIf"], [1, "thy-menu-group-title", 3, "title"], ["class", "thy-menu-group-more", 3, "thyIconName", "click", 4, "ngIf"], [1, "thy-menu-group-body"], ["thyMenuGroup", ""], [1, "thy-menu-group-folder", 3, "thyIconName"], [1, "thy-menu-group-more", 3, "thyIconName", "click"]], template: function ThyMenuGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("click", function ThyMenuGroupComponent_Template_div_click_0_listener() { return ctx.collapseGroup(); });
        ɵngcc0.ɵɵelement(1, "thy-icon", 1);
        ɵngcc0.ɵɵtemplate(2, ThyMenuGroupComponent_thy_icon_2_Template, 1, 1, "thy-icon", 2);
        ɵngcc0.ɵɵelementStart(3, "span", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ThyMenuGroupComponent_thy_icon_5_Template, 1, 1, "thy-icon", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5, 6);
        ɵngcc0.ɵɵprojection(8);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("title", ctx.thyTitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyTitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowAction);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("@detailsContentAnimation", ctx.isCollapsed);
    } }, directives: [ɵngcc2.ThyIconComponent, ɵngcc3.NgIf], encapsulation: 2, data: { animation: [
            trigger('detailsContentAnimation', [
                state('void', style({
                    height: '*'
                })),
                state('1', style({
                    height: 0,
                    overflow: 'hidden'
                })),
                state('0', style({
                    height: '*'
                })),
                transition('* => *', animate('0ms ease-out'))
            ])
        ] } });
ThyMenuGroupComponent.ctorParameters = () => [
    { type: ThyPopover }
];
ThyMenuGroupComponent.propDecorators = {
    _thyMenuGroup: [{ type: ViewChild, args: ['thyMenuGroup', { static: true },] }],
    isThyMenuGroup: [{ type: HostBinding, args: ['class.thy-menu-group',] }],
    showIcon: [{ type: HostBinding, args: ['class.has-icon',] }],
    isCollapsed: [{ type: HostBinding, args: ['class.collapsed',] }],
    thyTitle: [{ type: Input }],
    thyExpand: [{ type: Input, args: ['thyExpand',] }],
    thyShowIcon: [{ type: Input, args: ['thyShowIcon',] }],
    thyIcon: [{ type: Input, args: ['thyIcon',] }],
    thyActionIcon: [{ type: Input, args: ['thyActionIcon',] }],
    thyShowAction: [{ type: Input }],
    thyActionStopPropagation: [{ type: Input }],
    thyOnActionClick: [{ type: Output }],
    thyActionMenu: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-group, [thy-menu-group],[thyMenuGroup]',
                template: "<div class=\"thy-menu-group-header\" (click)=\"collapseGroup()\">\n  <thy-icon thyIconName=\"caret-right-down\" class=\"thy-menu-group-arrow\"></thy-icon>\n  <thy-icon *ngIf=\"showIcon\" class=\"thy-menu-group-folder\" [thyIconName]=\"iconClass\"></thy-icon>\n  <span class=\"thy-menu-group-title\" [title]=\"thyTitle\">{{ thyTitle }}</span>\n  <thy-icon\n    *ngIf=\"thyShowAction\"\n    class=\"thy-menu-group-more\"\n    [thyIconName]=\"rightIconClass\"\n    (click)=\"onActionClick($event)\"\n  ></thy-icon>\n</div>\n<div class=\"thy-menu-group-body\" #thyMenuGroup [@detailsContentAnimation]=\"isCollapsed\">\n  <ng-content></ng-content>\n</div>\n",
                animations: [
                    trigger('detailsContentAnimation', [
                        state('void', style({
                            height: '*'
                        })),
                        state('1', style({
                            height: 0,
                            overflow: 'hidden'
                        })),
                        state('0', style({
                            height: '*'
                        })),
                        transition('* => *', animate('0ms ease-out'))
                    ])
                ]
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }]; }, { isThyMenuGroup: [{
            type: HostBinding,
            args: ['class.thy-menu-group']
        }], showIcon: [{
            type: HostBinding,
            args: ['class.has-icon']
        }], isCollapsed: [{
            type: HostBinding,
            args: ['class.collapsed']
        }], thyTitle: [{
            type: Input
        }], thyShowAction: [{
            type: Input
        }], thyActionStopPropagation: [{
            type: Input
        }], thyOnActionClick: [{
            type: Output
        }], thyExpand: [{
            type: Input,
            args: ['thyExpand']
        }], thyShowIcon: [{
            type: Input,
            args: ['thyShowIcon']
        }], thyIcon: [{
            type: Input,
            args: ['thyIcon']
        }], thyActionIcon: [{
            type: Input,
            args: ['thyActionIcon']
        }], thyActionMenu: [{
            type: Input
        }], _thyMenuGroup: [{
            type: ViewChild,
            args: ['thyMenuGroup', { static: true }]
        }] }); })();

class ThyMenuItemNameComponent {
    constructor() {
        this.isThyMenuItemName = true;
        this.thyOverflowEllipsis = true;
    }
    ngOnInit() { }
}
ThyMenuItemNameComponent.ɵfac = function ThyMenuItemNameComponent_Factory(t) { return new (t || ThyMenuItemNameComponent)(); };
ThyMenuItemNameComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemNameComponent, selectors: [["thy-menu-item-name"], ["", "thy-menu-item-name", ""], ["", "thyMenuItemName", ""]], hostVars: 4, hostBindings: function ThyMenuItemNameComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-name", ctx.isThyMenuItemName)("thy-menu-item-name-ellipsis", ctx.thyOverflowEllipsis);
    } }, inputs: { thyOverflowEllipsis: "thyOverflowEllipsis" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemNameComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuItemNameComponent.ctorParameters = () => [];
ThyMenuItemNameComponent.propDecorators = {
    isThyMenuItemName: [{ type: HostBinding, args: ['class.thy-menu-item-name',] }],
    thyOverflowEllipsis: [{ type: HostBinding, args: ['class.thy-menu-item-name-ellipsis',] }, { type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemNameComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item-name,[thy-menu-item-name],[thyMenuItemName]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenuItemName: [{
            type: HostBinding,
            args: ['class.thy-menu-item-name']
        }], thyOverflowEllipsis: [{
            type: HostBinding,
            args: ['class.thy-menu-item-name-ellipsis']
        }, {
            type: Input
        }] }); })();

class ThyMenuItemIconComponent {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.isThyMenuItemIcon = true;
    }
    set thyColor(value) {
        if (value) {
            this.renderer.setStyle(this.elementRef.nativeElement, 'color', value);
        }
    }
    ngOnInit() { }
}
ThyMenuItemIconComponent.ɵfac = function ThyMenuItemIconComponent_Factory(t) { return new (t || ThyMenuItemIconComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyMenuItemIconComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemIconComponent, selectors: [["thy-menu-item-icon"], ["", "thy-menu-item-icon", ""], ["", "thyMenuItemIcon", ""]], hostVars: 2, hostBindings: function ThyMenuItemIconComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-icon", ctx.isThyMenuItemIcon);
    } }, inputs: { thyColor: "thyColor" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemIconComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuItemIconComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ThyMenuItemIconComponent.propDecorators = {
    isThyMenuItemIcon: [{ type: HostBinding, args: ['class.thy-menu-item-icon',] }],
    thyColor: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemIconComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item-icon,[thy-menu-item-icon],[thyMenuItemIcon]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { isThyMenuItemIcon: [{
            type: HostBinding,
            args: ['class.thy-menu-item-icon']
        }], thyColor: [{
            type: Input
        }] }); })();

class ThyMenuItemActionComponent {
    constructor(popover, renderer, elementRef) {
        this.popover = popover;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this._boundEvent = false;
        this.isThyMenuItemIconMore = true;
        this.thyStopPropagation = true;
    }
    set thyActionMenu(value) {
        this._actionMenu = value;
        if (this._actionMenu) {
            this.bindClickEvent();
        }
    }
    bindClickEvent() {
        if (this._boundEvent) {
            return;
        }
        this._boundEvent = true;
        this.renderer.listen(this.elementRef.nativeElement, 'click', event => {
            if (this.thyStopPropagation) {
                event.stopPropagation();
            }
            if (this._actionMenu) {
                this.popover.open(this._actionMenu, {
                    origin: event.currentTarget,
                    insideClosable: true,
                    placement: 'bottom'
                });
            }
        });
    }
    ngOnInit() { }
}
ThyMenuItemActionComponent.ɵfac = function ThyMenuItemActionComponent_Factory(t) { return new (t || ThyMenuItemActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyMenuItemActionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemActionComponent, selectors: [["thy-menu-item-action"], ["", "thy-menu-item-action", ""], ["", "thyMenuItemAction", ""]], hostVars: 2, hostBindings: function ThyMenuItemActionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-action", ctx.isThyMenuItemIconMore);
    } }, inputs: { thyStopPropagation: "thyStopPropagation", thyActionMenu: "thyActionMenu" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuItemActionComponent.ctorParameters = () => [
    { type: ThyPopover },
    { type: Renderer2 },
    { type: ElementRef }
];
ThyMenuItemActionComponent.propDecorators = {
    isThyMenuItemIconMore: [{ type: HostBinding, args: ['class.thy-menu-item-action',] }],
    thyActionMenu: [{ type: Input }],
    thyStopPropagation: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemActionComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item-action,[thy-menu-item-action],[thyMenuItemAction]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { isThyMenuItemIconMore: [{
            type: HostBinding,
            args: ['class.thy-menu-item-action']
        }], thyStopPropagation: [{
            type: Input
        }], thyActionMenu: [{
            type: Input
        }] }); })();

class ThyMenuDividerComponent {
    constructor() {
        this.isThyMenuDivider = true;
    }
    ngOnInit() { }
}
ThyMenuDividerComponent.ɵfac = function ThyMenuDividerComponent_Factory(t) { return new (t || ThyMenuDividerComponent)(); };
ThyMenuDividerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuDividerComponent, selectors: [["thy-menu-divider"], ["", "thy-menu-divider", ""], ["", "thyMenuDivider", ""]], hostVars: 2, hostBindings: function ThyMenuDividerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-divider", ctx.isThyMenuDivider);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuDividerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuDividerComponent.ctorParameters = () => [];
ThyMenuDividerComponent.propDecorators = {
    isThyMenuDivider: [{ type: HostBinding, args: ['class.thy-menu-divider',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuDividerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-divider,[thy-menu-divider],[thyMenuDivider]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenuDivider: [{
            type: HostBinding,
            args: ['class.thy-menu-divider']
        }] }); })();

class ThyMenuModule {
}
ThyMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyMenuModule });
ThyMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyMenuModule_Factory(t) { return new (t || ThyMenuModule)(); }, imports: [[CommonModule, ThyIconModule, ThyPopoverModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyMenuModule, { declarations: function () { return [ThyMenuComponent, ThyMenuGroupComponent, ThyMenuItemComponent, ThyMenuItemNameComponent, ThyMenuItemIconComponent, ThyMenuItemActionComponent, ThyMenuDividerComponent]; }, imports: function () { return [CommonModule, ThyIconModule, ThyPopoverModule]; }, exports: function () { return [ThyMenuComponent, ThyMenuGroupComponent, ThyMenuItemComponent, ThyMenuItemNameComponent, ThyMenuItemIconComponent, ThyMenuItemActionComponent, ThyMenuDividerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyMenuComponent,
                    ThyMenuGroupComponent,
                    ThyMenuItemComponent,
                    ThyMenuItemNameComponent,
                    ThyMenuItemIconComponent,
                    ThyMenuItemActionComponent,
                    ThyMenuDividerComponent
                ],
                imports: [CommonModule, ThyIconModule, ThyPopoverModule],
                exports: [
                    ThyMenuComponent,
                    ThyMenuGroupComponent,
                    ThyMenuItemComponent,
                    ThyMenuItemNameComponent,
                    ThyMenuItemIconComponent,
                    ThyMenuItemActionComponent,
                    ThyMenuDividerComponent
                ]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyMenuComponent, ThyMenuDividerComponent, ThyMenuGroupComponent, ThyMenuItemActionComponent, ThyMenuItemComponent, ThyMenuItemIconComponent, ThyMenuItemNameComponent, ThyMenuModule };
