// ngx-tethys/fesm2015/ngx-tethys-dropdown.js
import { Directive, ElementRef, HostBinding, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import * as ɵngcc0 from '@angular/core';
const thyDropdown = 'thyDropdown';
const thyDropdownSplit = 'thyDropdownSplit';
class ThyDropdownDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this._isDropdownClass = true;
        this._isDropdownSplitClass = false;
        const attrs = Array.from(elementRef.nativeElement.attributes);
        if (attrs.find((item) => item.name === thyDropdownSplit.toLocaleLowerCase())) {
            this._isDropdownSplitClass = true;
        }
    }
}
ThyDropdownDirective.ɵfac = function ThyDropdownDirective_Factory(t) { return new (t || ThyDropdownDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyDropdownDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyDropdownDirective, selectors: [["", "thyDropdown", ""], ["", "thyDropdownSplit", ""]], hostVars: 4, hostBindings: function ThyDropdownDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-dropdown-toggle", ctx._isDropdownClass)("thy-dropdown-toggle-split", ctx._isDropdownSplitClass);
    } } });
ThyDropdownDirective.ctorParameters = () => [
    { type: ElementRef }
];
ThyDropdownDirective.propDecorators = {
    _isDropdownClass: [{ type: HostBinding, args: ['class.thy-dropdown-toggle',] }],
    _isDropdownSplitClass: [{ type: HostBinding, args: ['class.thy-dropdown-toggle-split',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDropdownDirective, [{
        type: Directive,
        args: [{
                selector: `[${thyDropdown}],[${thyDropdownSplit}]`
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { _isDropdownClass: [{
            type: HostBinding,
            args: ['class.thy-dropdown-toggle']
        }], _isDropdownSplitClass: [{
            type: HostBinding,
            args: ['class.thy-dropdown-toggle-split']
        }] }); })();

class ThyDropdownModule {
}
ThyDropdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDropdownModule });
ThyDropdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDropdownModule_Factory(t) { return new (t || ThyDropdownModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDropdownModule, { declarations: function () { return [ThyDropdownDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyDropdownDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDropdownModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    ThyDropdownDirective
                ],
                exports: [
                    ThyDropdownDirective
                ]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyDropdownDirective, ThyDropdownModule };
