// ngx-tethys/fesm2015/ngx-tethys-alert.js
import { TemplateRef, Component, HostBinding, Input, ContentChild, Directive, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { isString } from 'ngx-tethys/util';
import { ThyIconModule } from 'ngx-tethys/icon';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/icon';

const _c0 = ["operation"];
function ThyAlertComponent_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 5);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.thyIcon);
} }
function ThyAlertComponent_ng_container_2_1_ng_template_0_Template(rf, ctx) { }
function ThyAlertComponent_ng_container_2_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyAlertComponent_ng_container_2_1_ng_template_0_Template, 0, 0, "ng-template");
} }
function ThyAlertComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyAlertComponent_ng_container_2_1_Template, 1, 0, undefined, 6);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.messageTemplate);
} }
function ThyAlertComponent_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.messageText);
} }
function ThyAlertComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵelementContainer(1, 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r3.alertOperation);
} }
function ThyAlertComponent_a_5_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 9);
    ɵngcc0.ɵɵlistener("click", function ThyAlertComponent_a_5_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.closeAlert(); });
    ɵngcc0.ɵɵelement(1, "thy-icon", 10);
    ɵngcc0.ɵɵelementEnd();
} }
class ThyAlertComponent {
    constructor() {
        // @ViewChild(TemplateRef) content: TemplateRef<any>;
        this._typeIcon = {
            success: 'check-circle-fill',
            warning: 'waring-fill',
            danger: 'close-circle-fill',
            info: 'minus-circle-fill',
            'primary-week': 'question-circle-fill',
            'success-week': 'check-circle-fill',
            'warning-week': 'waring-fill',
            'danger-week': 'close-circle-fill',
            'primary-weak': 'question-circle-fill',
            'success-weak': 'check-circle-fill',
            'warning-weak': 'waring-fill',
            'danger-weak': 'close-circle-fill'
        };
        this._showIcon = true;
        this._type = 'info';
    }
    set thyType(value) {
        if (value && value.indexOf('-week') !== -1) {
            console.warn('The type "xxx-week" will be removed, please use "xxx-weak" instead.');
        }
        this._type = value;
    }
    set thyMessage(value) {
        if (value instanceof TemplateRef) {
            this.messageTemplate = value;
        }
        else {
            this.messageText = value;
        }
    }
    set thyIcon(value) {
        if (value) {
            this._showIcon = true;
            this._icon = isString(value) ? value.toString() : null;
        }
        else {
            this._showIcon = false;
        }
    }
    get thyIcon() {
        if (this._showIcon) {
            return this._icon || this._typeIcon[this._type];
        }
        else {
            return null;
        }
    }
    ngOnInit() {
        this.class = `thy-alert thy-alert-${this._type}`;
    }
    closeAlert() {
        this.class = `${this.class} thy-alert-hidden`;
    }
}
ThyAlertComponent.ɵfac = function ThyAlertComponent_Factory(t) { return new (t || ThyAlertComponent)(); };
ThyAlertComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyAlertComponent, selectors: [["thy-alert"]], contentQueries: function ThyAlertComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.alertOperation = _t.first);
    } }, hostVars: 2, hostBindings: function ThyAlertComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.class);
    } }, inputs: { thyType: "thyType", thyMessage: "thyMessage", thyIcon: "thyIcon", thyCloseable: "thyCloseable" }, decls: 6, vars: 5, consts: [[1, "thy-alert-content"], ["class", "thy-alert-icon", 3, "thyIconName", 4, "ngIf"], [4, "ngIf"], ["class", "thy-alert-operation", 4, "ngIf"], ["href", "javascript:;", "class", "thy-alert-close", 3, "click", 4, "ngIf"], [1, "thy-alert-icon", 3, "thyIconName"], [4, "ngTemplateOutlet"], [1, "thy-alert-operation"], [3, "ngTemplateOutlet"], ["href", "javascript:;", 1, "thy-alert-close", 3, "click"], ["thyIconName", "close"]], template: function ThyAlertComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyAlertComponent_thy_icon_1_Template, 1, 1, "thy-icon", 1);
        ɵngcc0.ɵɵtemplate(2, ThyAlertComponent_ng_container_2_Template, 2, 1, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(3, ThyAlertComponent_span_3_Template, 2, 1, "span", 2);
        ɵngcc0.ɵɵtemplate(4, ThyAlertComponent_div_4_Template, 2, 1, "div", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ThyAlertComponent_a_5_Template, 2, 0, "a", 4);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.messageTemplate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.messageText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.alertOperation);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyCloseable);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
ThyAlertComponent.ctorParameters = () => [];
ThyAlertComponent.propDecorators = {
    class: [{ type: HostBinding, args: ['class',] }],
    thyType: [{ type: Input }],
    thyMessage: [{ type: Input }],
    thyIcon: [{ type: Input }],
    thyCloseable: [{ type: Input }],
    alertOperation: [{ type: ContentChild, args: ['operation',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAlertComponent, [{
        type: Component,
        args: [{
                selector: 'thy-alert',
                template: "<div class=\"thy-alert-content\">\n  <thy-icon *ngIf=\"thyIcon\" class=\"thy-alert-icon\" [thyIconName]=\"thyIcon\"></thy-icon>\n  <ng-container *ngIf=\"messageTemplate\">\n    <ng-template *ngTemplateOutlet=\"messageTemplate\"></ng-template>\n  </ng-container>\n  <span *ngIf=\"messageText\">{{ messageText }}</span>\n  <div class=\"thy-alert-operation\" *ngIf=\"alertOperation\">\n    <ng-container [ngTemplateOutlet]=\"alertOperation\"></ng-container>\n  </div>\n</div>\n<a href=\"javascript:;\" class=\"thy-alert-close\" *ngIf=\"thyCloseable\" (click)=\"closeAlert()\">\n  <thy-icon thyIconName=\"close\"></thy-icon>\n</a>\n"
            }]
    }], function () { return []; }, { thyType: [{
            type: Input
        }], thyMessage: [{
            type: Input
        }], thyIcon: [{
            type: Input
        }], class: [{
            type: HostBinding,
            args: ['class']
        }], thyCloseable: [{
            type: Input
        }], alertOperation: [{
            type: ContentChild,
            args: ['operation']
        }] }); })();

class ThyAlertActionItemDirective {
    constructor() {
        this.class = true;
    }
}
ThyAlertActionItemDirective.ɵfac = function ThyAlertActionItemDirective_Factory(t) { return new (t || ThyAlertActionItemDirective)(); };
ThyAlertActionItemDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyAlertActionItemDirective, selectors: [["", "thyAlertActionItem", ""]], hostVars: 2, hostBindings: function ThyAlertActionItemDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-alert-action", ctx.class);
    } } });
ThyAlertActionItemDirective.ctorParameters = () => [];
ThyAlertActionItemDirective.propDecorators = {
    class: [{ type: HostBinding, args: ['class.thy-alert-action',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAlertActionItemDirective, [{
        type: Directive,
        args: [{ selector: '[thyAlertActionItem]' }]
    }], function () { return []; }, { class: [{
            type: HostBinding,
            args: ['class.thy-alert-action']
        }] }); })();

class ThyAlertModule {
}
ThyAlertModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyAlertModule });
ThyAlertModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyAlertModule_Factory(t) { return new (t || ThyAlertModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyAlertModule, { declarations: function () { return [ThyAlertComponent, ThyAlertActionItemDirective]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyAlertComponent, ThyAlertActionItemDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyAlertModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyAlertComponent, ThyAlertActionItemDirective],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyAlertComponent, ThyAlertActionItemDirective]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyAlertActionItemDirective, ThyAlertComponent, ThyAlertModule };
