// ngx-tethys/fesm2015/ngx-tethys-form.js
import { InjectionToken, Injectable, Optional, Inject, Directive, ElementRef, Renderer2, NgZone, Input, HostBinding, Component, ViewEncapsulation, ChangeDetectionStrategy, ContentChild, EventEmitter, Output, HostListener, forwardRef, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgForm, NG_ASYNC_VALIDATORS, Validators, NG_VALIDATORS, FormsModule } from '@angular/forms';
import { UpdateHostClassService, ThyTranslate, InputBoolean } from 'ngx-tethys/core';
import { helpers, keycodes, coerceBooleanProperty } from 'ngx-tethys/util';
import { ThyInputModule } from 'ngx-tethys/input';
import { __decorate, __metadata } from 'tslib';
import { of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { ThyAlertModule } from 'ngx-tethys/alert';
import { ThyIconModule } from 'ngx-tethys/icon';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from 'ngx-tethys/core';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/icon';
import * as ɵngcc5 from 'ngx-tethys/alert';

const _c0 = ["formGroup"];
function ThyFormGroupComponent_ng_container_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r5.feedbackSvgIconName);
} }
function ThyFormGroupComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementContainer(1, 3);
    ɵngcc0.ɵɵtemplate(2, ThyFormGroupComponent_ng_container_0_ng_container_2_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    const _r3 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.contentTemplateRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.feedbackSvgIconName)("ngIfElse", _r3);
} }
const _c1 = function (a0, a1, a2) { return { "col-sm-2 col-form-label": a0, "label-required": a1, "pt-0": a2 }; };
function ThyFormGroupComponent_ng_template_1_label_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction3(2, _c1, ctx_r6.isHorizontal, ctx_r6.labelRequired, ctx_r6.labelPaddingTopClear));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r6.labelText);
} }
function ThyFormGroupComponent_ng_template_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r7.feedbackSvgIconName);
} }
function ThyFormGroupComponent_ng_template_1_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r8.tips);
} }
const _c2 = function (a0, a1) { return { "col-sm-10 col-form-control": a0, "position-relative": a1 }; };
function ThyFormGroupComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyFormGroupComponent_ng_template_1_label_0_Template, 2, 6, "label", 5);
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵprojection(2);
    ɵngcc0.ɵɵtemplate(3, ThyFormGroupComponent_ng_template_1_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(4, ThyFormGroupComponent_ng_template_1_span_4_Template, 2, 1, "span", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    const _r3 = ɵngcc0.ɵɵreference(4);
    ɵngcc0.ɵɵproperty("ngIf", !(!ctx_r2.isHorizontal && !ctx_r2.labelText));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(5, _c2, ctx_r2.isHorizontal, !ctx_r2.isHorizontal));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.feedbackSvgIconName)("ngIfElse", _r3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.tips);
} }
function ThyFormGroupComponent_ng_template_3_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span");
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMapInterpolate1("form-control-feedback ", ctx_r9.feedbackIcon, "");
} }
function ThyFormGroupComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyFormGroupComponent_ng_template_3_span_0_Template, 1, 3, "span", 10);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.feedbackIcon);
} }
const _c3 = ["*"];
const _c4 = ["description"];
function ThyFormGroupFooterComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementContainer(1, 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.description);
} }
function ThyFormGroupErrorComponent_thy_alert_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-alert", 2);
} if (rf & 2) {
    const error_r1 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyMessage", error_r1);
} }
const _c5 = function (a0) { return { "col-sm-10 offset-sm-2 col-form-control": a0 }; };
const THY_VALIDATOR_CONFIG = new InjectionToken('VALIDATION_CONFIG');
const THY_FORM_CONFIG = new InjectionToken('THY_FORM_CONFIG');
const ɵ0 = { layout: 'horizontal', footerAlign: 'left' };
const THY_FORM_CONFIG_PROVIDER = {
    provide: THY_FORM_CONFIG,
    useValue: ɵ0
};

const ERROR_VALUE_REPLACE_REGEX = /\{(.+?)\}/g;
const INVALID_CLASS = 'is-invalid';
const INVALID_FEEDBACK_CLASS = 'invalid-feedback';
const defaultValidatorConfig = {
    showElementError: true,
    removeElementError: true,
    validationMessages: {}
};
const globalValidationMessages = {
    required: '该选项不能为空',
    maxlength: '该选项输入值长度不能大于{maxlength}',
    minlength: '该选项输入值长度不能小于{minlength}',
    thyUniqueCheck: '输入值已经存在，请重新输入',
    email: '输入邮件的格式不正确',
    confirm: '两次输入不一致',
    pattern: '该选项输入格式不正确',
    number: '必须输入数字',
    url: '输入URL格式不正确',
    max: '该选项输入值不能大于{max}',
    min: '该选项输入值不能小于{min}'
};
class ThyFormValidatorLoader {
    constructor(config) {
        this.config = Object.assign({}, defaultValidatorConfig, config);
    }
    getDefaultValidationMessage(key) {
        if (this.config.globalValidationMessages && this.config.globalValidationMessages[key]) {
            return this.config.globalValidationMessages[key];
        }
        else {
            return globalValidationMessages[key];
        }
    }
    get validationMessages() {
        return this.config.validationMessages;
    }
    getErrorMessage(name, key) {
        if (this.validationMessages[name] && this.validationMessages[name][key]) {
            return this.validationMessages[name][key];
        }
        else {
            return this.getDefaultValidationMessage(key);
        }
    }
    getErrorMessages(name, validationErrors) {
        const messages = [];
        for (const validationError in validationErrors) {
            if (validationErrors.hasOwnProperty(validationError)) {
                messages.push(this.getErrorMessage(name, validationError));
            }
        }
        return messages;
    }
    defaultShowError(element, errorMessages) {
        if (element && element.parentElement) {
            const documentFrag = document.createDocumentFragment();
            const divNode = document.createElement('DIV');
            const textNode = document.createTextNode(errorMessages[0]);
            divNode.appendChild(textNode);
            divNode.setAttribute('class', INVALID_FEEDBACK_CLASS);
            documentFrag.appendChild(divNode);
            element.parentElement.append(documentFrag);
        }
    }
    defaultRemoveError(element) {
        if (element && element.parentElement) {
            const invalidFeedback = element.parentElement.querySelector('.invalid-feedback');
            element.parentElement.removeChild(invalidFeedback);
        }
    }
    removeError(element) {
        element.classList.remove(INVALID_CLASS);
        if (helpers.isFunction(this.config.removeElementError)) {
            this.config.removeElementError(element);
        }
        else if (this.config.showElementError) {
            this.defaultRemoveError(element);
        }
        else {
            // do nothings
        }
    }
    showError(element, errorMessages) {
        element.classList.add(INVALID_CLASS);
        if (helpers.isFunction(this.config.showElementError)) {
            this.config.showElementError(element, errorMessages);
        }
        else if (this.config.showElementError) {
            this.defaultShowError(element, errorMessages);
        }
        else {
            // do nothings
        }
    }
    addValidationMessages(messages) {
        Object.assign(this.config.validationMessages, messages);
    }
    setGlobalValidationMessages(validationMessages) {
        this.config.globalValidationMessages = validationMessages;
    }
}
ThyFormValidatorLoader.ɵfac = function ThyFormValidatorLoader_Factory(t) { return new (t || ThyFormValidatorLoader)(ɵngcc0.ɵɵinject(THY_VALIDATOR_CONFIG, 8)); };
ThyFormValidatorLoader.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyFormValidatorLoader, factory: ThyFormValidatorLoader.ɵfac });
ThyFormValidatorLoader.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [THY_VALIDATOR_CONFIG,] }] }
];

class ThyFormValidatorService {
    constructor(thyFormValidateLoader) {
        this.thyFormValidateLoader = thyFormValidateLoader;
        this.errors = [];
        // 记录所有元素的验证信息
        this.validations = {};
    }
    _getElement(name) {
        const element = this._formElement[name];
        if (element) {
            return element;
        }
        else {
            return this._formElement.querySelector(`[name='${name}']`);
        }
    }
    _clearElementError(name) {
        if (this.validations[name] && this.validations[name].hasError) {
            this.validations[name].hasError = false;
            this.validations[name].errorMessages = [];
            this.thyFormValidateLoader.removeError(this._getElement(name));
        }
    }
    _tryGetValidation(name) {
        if (!this.validations[name]) {
            this._initializeFormControlValidation(name, this._ngForm.controls[name]);
        }
        return this.validations[name];
    }
    _addError(message) {
        this.errors.unshift(message);
    }
    _clearErrors() {
        this.errors = [];
    }
    _initializeFormControlValidation(name, control) {
        this.validations[name] = {
            hasError: false,
            errorMessages: []
        };
        control.valueChanges.subscribe(() => {
            this._clearElementError(name);
            this._clearErrors();
        });
    }
    _restFormControlValidation(name) {
        const validation = this.validations[name];
        if (validation) {
            validation.hasError = false;
            validation.errorMessages = [];
        }
    }
    _formatValidationMessage(name, message) {
        const control = this._ngForm.controls[name];
        if (control) {
            return message.replace(ERROR_VALUE_REPLACE_REGEX, (tag, key) => {
                if (key) {
                    return control.errors[key][key] || control.errors[key].requiredLength;
                }
            });
        }
        else {
            return message;
        }
    }
    _getValidationMessage(name, validationError) {
        let message = null;
        if (this._config &&
            this._config.validationMessages &&
            this._config.validationMessages[name] &&
            this._config.validationMessages[name][validationError]) {
            message = this._config.validationMessages[name][validationError];
        }
        else {
            message = this.thyFormValidateLoader.getErrorMessage(name, validationError);
        }
        return this._formatValidationMessage(name, message);
    }
    _getValidationMessages(name, validationErrors) {
        const messages = [];
        for (const validationError in validationErrors) {
            if (validationErrors.hasOwnProperty(validationError)) {
                messages.push(this._getValidationMessage(name, validationError));
            }
        }
        return messages;
    }
    _setControlValidationError(name, errorMessages) {
        const validation = this._tryGetValidation(name);
        validation.errorMessages = errorMessages;
        validation.hasError = true;
        this.thyFormValidateLoader.showError(this._getElement(name), errorMessages);
    }
    initialize(ngForm, formElement) {
        this._ngForm = ngForm;
        this._formElement = formElement;
    }
    setValidatorConfig(config) {
        this._config = config;
    }
    validateControl(name) {
        this._clearElementError(name);
        const control = this._ngForm.controls[name];
        if (control && control.invalid) {
            const errorMessages = this._getValidationMessages(name, control.errors);
            this._setControlValidationError(name, errorMessages);
        }
    }
    validateControls() {
        // 主要是 无法检测到 ngForm 的 controls 的变化，或者是我没有找到
        // 验证的时候循环 ngForm 的 controls 验证
        // 发现没有 validation 初始化一个，已经存在不会重新初始化，保存缓存数据
        for (const name in this._ngForm.controls) {
            if (this._ngForm.controls.hasOwnProperty(name)) {
                this._tryGetValidation(name);
                this.validateControl(name);
            }
        }
        // 移除已经不存在的 validation
        const names = Object.keys(this.validations);
        names.forEach(name => {
            if (!this._ngForm.controls[name]) {
                delete this.validations[name];
            }
        });
    }
    addError(message) {
        this._addError(message);
    }
    validate($event) {
        this._ngForm.onSubmit($event);
        this.validateControls();
        return this._ngForm.valid;
    }
    reset() {
        this._ngForm.reset();
        for (const name in this.validations) {
            if (this.validations.hasOwnProperty(name)) {
                this._restFormControlValidation(name);
                this._clearElementError(name);
            }
        }
    }
    setElementErrorMessage(name, message) {
        this._clearElementError(name);
        this._setControlValidationError(name, [message]);
    }
}
ThyFormValidatorService.ɵfac = function ThyFormValidatorService_Factory(t) { return new (t || ThyFormValidatorService)(ɵngcc0.ɵɵinject(ThyFormValidatorLoader)); };
ThyFormValidatorService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyFormValidatorService, factory: ThyFormValidatorService.ɵfac });
ThyFormValidatorService.ctorParameters = () => [
    { type: ThyFormValidatorLoader }
];

// 1. submit 按 Enter 键提交, Textare或包含[contenteditable]属性的元素 除外，需要按 Ctrl | Command + Enter 提交
// 2. alwaysSubmit 不管是哪个元素 按 Enter 键都提交
// 3. forbidSubmit Enter 键禁止提交
// 默认 submit
var ThyEnterKeyMode;
(function (ThyEnterKeyMode) {
    ThyEnterKeyMode["submit"] = "submit";
    ThyEnterKeyMode["alwaysSubmit"] = "alwaysSubmit";
    ThyEnterKeyMode["forbidSubmit"] = "forbidSubmit";
})(ThyEnterKeyMode || (ThyEnterKeyMode = {}));
class ThyFormDirective {
    constructor(ngForm, elementRef, renderer, ngZone, updateHostClassService, validator, config) {
        this.ngForm = ngForm;
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.updateHostClassService = updateHostClassService;
        this.validator = validator;
        this.config = config;
        this.initialized = false;
        this.wasValidated = false;
        this.updateHostClassService.initializeElement(this.elementRef.nativeElement);
        this.layout = this.config.layout;
    }
    set thyLayout(value) {
        if (value) {
            this.layout = value;
            if (this.initialized) {
                this.updateClasses();
            }
        }
    }
    get thyLayout() {
        return this.layout;
    }
    get isHorizontal() {
        return this.layout === 'horizontal';
    }
    set thyFormValidatorConfig(config) {
        this.validator.setValidatorConfig(config);
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this._unsubscribe = this.renderer.listen(this.elementRef.nativeElement, 'keydown', this.onKeydown.bind(this));
        });
        this.updateClasses();
        this.initialized = true;
        this.validator.initialize(this.ngForm, this.elementRef.nativeElement);
    }
    submit($event) {
        if (this.validator.validate($event)) {
            this.onSubmitSuccess($event);
        }
        else {
            // this.wasValidated = true;
        }
    }
    updateClasses() {
        this.updateHostClassService.updateClassByMap({
            [`thy-form-${this.thyLayout}`]: true
        });
    }
    submitRunInZone($event) {
        this.ngZone.run(() => {
            this.submit($event);
        });
    }
    onKeydown($event) {
        const currentInput = document.activeElement;
        const key = $event.which || $event.keyCode;
        if (key === keycodes.ENTER && currentInput.tagName) {
            if (!this.thyEnterKeyMode || this.thyEnterKeyMode === ThyEnterKeyMode.submit) {
                // TEXTAREA或包含[contenteditable]属性的元素 Ctrl + Enter 或者 Command + Enter 阻止默认行为并提交
                if (currentInput.tagName === 'TEXTAREA' || coerceBooleanProperty(currentInput.getAttribute('contenteditable'))) {
                    if ($event.ctrlKey || $event.metaKey) {
                        $event.preventDefault();
                        this.submitRunInZone($event);
                    }
                }
                else {
                    // 不是 TEXTAREA Enter 阻止默认行为并提交
                    $event.preventDefault();
                    this.submitRunInZone($event);
                }
            }
            else if (this.thyEnterKeyMode === ThyEnterKeyMode.alwaysSubmit) {
                $event.preventDefault();
                this.submitRunInZone($event);
            }
            else {
                // do nothing
            }
        }
    }
    ngOnDestroy() {
        if (this._unsubscribe) {
            this._unsubscribe();
            this._unsubscribe = null;
        }
    }
}
ThyFormDirective.ɵfac = function ThyFormDirective_Factory(t) { return new (t || ThyFormDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgForm), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ThyFormValidatorService), ɵngcc0.ɵɵdirectiveInject(THY_FORM_CONFIG)); };
ThyFormDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyFormDirective, selectors: [["", "thyForm", ""], ["", "thy-form", ""]], hostAttrs: [1, "thy-form"], hostVars: 2, hostBindings: function ThyFormDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("was-validated", ctx.wasValidated);
    } }, inputs: { thyLayout: "thyLayout", thyFormValidatorConfig: "thyFormValidatorConfig", thyEnterKeyMode: "thyEnterKeyMode" }, exportAs: ["thyForm"], features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService, ThyFormValidatorService])] });
ThyFormDirective.ctorParameters = () => [
    { type: NgForm },
    { type: ElementRef },
    { type: Renderer2 },
    { type: NgZone },
    { type: UpdateHostClassService },
    { type: ThyFormValidatorService },
    { type: undefined, decorators: [{ type: Inject, args: [THY_FORM_CONFIG,] }] }
];
ThyFormDirective.propDecorators = {
    thyLayout: [{ type: Input }],
    thyEnterKeyMode: [{ type: Input }],
    thyFormValidatorConfig: [{ type: Input }],
    wasValidated: [{ type: HostBinding, args: ['class.was-validated',] }]
};

const internalIconMap = {
    date: 'wtf wtf-schedule-o'
};
class ThyFormGroupComponent {
    constructor(thyParentForm, thyTranslate) {
        this.thyParentForm = thyParentForm;
        this.thyTranslate = thyTranslate;
        this.labelRequired = false;
        this.labelPaddingTopClear = false;
        this._rowFill = false;
        this._isFormGroup = true;
        this.isHorizontal = true;
        this.hasFeedback = false;
    }
    set thyLabelText(value) {
        this.labelText = value;
    }
    set thyLabelTextTranslateKey(value) {
        if (value) {
            this.labelText = this.thyTranslate.instant(value);
        }
        else {
            this.labelText = '';
        }
    }
    set thyLabelRequired(value) {
        this.labelRequired = coerceBooleanProperty(value);
    }
    set thyLabelPaddingTopClear(value) {
        this.labelPaddingTopClear = coerceBooleanProperty(value);
    }
    set thyFeedbackIcon(value) {
        this.hasFeedback = true;
        if (internalIconMap[value]) {
            this.feedbackIcon = internalIconMap[value];
            this.feedbackSvgIconName = null;
        }
        else {
            this.feedbackSvgIconName = value;
            this.feedbackIcon = null;
        }
    }
    set thyTips(value) {
        this.tips = value;
    }
    set thyTipsTranslateKey(value) {
        this.tips = this.thyTranslate.instant(value);
    }
    set thyRowFill(value) {
        this._rowFill = coerceBooleanProperty(value);
    }
    ngOnInit() {
        this.isHorizontal = this.thyParentForm ? this.thyParentForm.isHorizontal : true;
    }
}
ThyFormGroupComponent.ɵfac = function ThyFormGroupComponent_Factory(t) { return new (t || ThyFormGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ThyFormDirective, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyTranslate)); };
ThyFormGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFormGroupComponent, selectors: [["thy-form-group"]], contentQueries: function ThyFormGroupComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, hostVars: 8, hostBindings: function ThyFormGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("row-fill", ctx._rowFill)("form-group", ctx._isFormGroup)("row", ctx.isHorizontal)("has-feedback", ctx.hasFeedback);
    } }, inputs: { thyLabelText: "thyLabelText", thyLabelTextTranslateKey: "thyLabelTextTranslateKey", thyLabelRequired: "thyLabelRequired", thyLabelPaddingTopClear: "thyLabelPaddingTopClear", thyFeedbackIcon: "thyFeedbackIcon", thyTips: "thyTips", thyTipsTranslateKey: "thyTipsTranslateKey", thyRowFill: "thyRowFill" }, ngContentSelectors: _c3, decls: 5, vars: 2, consts: [[4, "ngIf", "ngIfElse"], ["defaultTpl", ""], ["iconFont", ""], [3, "ngTemplateOutlet"], [1, "form-control-feedback", 3, "thyIconName"], ["class", "form-label ", 3, "ngClass", 4, "ngIf"], [3, "ngClass"], ["class", "form-text text-desc", 4, "ngIf"], [1, "form-label", 3, "ngClass"], [1, "form-text", "text-desc"], [3, "class", 4, "ngIf"]], template: function ThyFormGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThyFormGroupComponent_ng_container_0_Template, 3, 3, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, ThyFormGroupComponent_ng_template_1_Template, 5, 8, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(3, ThyFormGroupComponent_ng_template_3_Template, 1, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.contentTemplateRef)("ngIfElse", _r1);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet, ɵngcc4.ThyIconComponent, ɵngcc3.NgClass], encapsulation: 2, changeDetection: 0 });
ThyFormGroupComponent.ctorParameters = () => [
    { type: ThyFormDirective, decorators: [{ type: Optional }] },
    { type: ThyTranslate }
];
ThyFormGroupComponent.propDecorators = {
    _rowFill: [{ type: HostBinding, args: ['class.row-fill',] }],
    _isFormGroup: [{ type: HostBinding, args: ['class.form-group',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.row',] }],
    hasFeedback: [{ type: HostBinding, args: ['class.has-feedback',] }],
    thyLabelText: [{ type: Input }],
    thyLabelTextTranslateKey: [{ type: Input }],
    thyLabelRequired: [{ type: Input }],
    thyLabelPaddingTopClear: [{ type: Input }],
    thyFeedbackIcon: [{ type: Input }],
    thyTips: [{ type: Input }],
    thyTipsTranslateKey: [{ type: Input }],
    thyRowFill: [{ type: Input }],
    contentTemplateRef: [{ type: ContentChild, args: ['formGroup',] }]
};

class ThyFormGroupLabelDirective {
    constructor(thyTranslate) {
        this.thyTranslate = thyTranslate;
        this.labelRequired = false;
        this._isFormGroupLabel = true;
    }
    set thyLabelText(value) {
        this.labelText = value;
    }
    set thyLabelTranslateKey(translateKey) {
        if (translateKey) {
            this.labelText = this.thyTranslate.instant(translateKey);
        }
    }
    set thyLabelRequired(value) {
        this.labelRequired = coerceBooleanProperty(value);
    }
}
ThyFormGroupLabelDirective.ɵfac = function ThyFormGroupLabelDirective_Factory(t) { return new (t || ThyFormGroupLabelDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyTranslate)); };
ThyFormGroupLabelDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyFormGroupLabelDirective, selectors: [["", "thyFormGroupLabel", ""]], hostVars: 4, hostBindings: function ThyFormGroupLabelDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("label-required", ctx.labelRequired)("col-form-label", ctx._isFormGroupLabel);
    } }, inputs: { thyLabelText: "thyLabelText", thyLabelTranslateKey: "thyLabelTranslateKey", thyLabelRequired: "thyLabelRequired" } });
ThyFormGroupLabelDirective.ctorParameters = () => [
    { type: ThyTranslate }
];
ThyFormGroupLabelDirective.propDecorators = {
    labelRequired: [{ type: HostBinding, args: ['class.label-required',] }],
    _isFormGroupLabel: [{ type: HostBinding, args: ['class.col-form-label',] }],
    thyLabelText: [{ type: Input }],
    thyLabelTranslateKey: [{ type: Input }],
    thyLabelRequired: [{ type: Input }]
};

class ThyFormSubmitDirective {
    constructor(ngForm, thyFormDirective) {
        this.ngForm = ngForm;
        this.thyFormDirective = thyFormDirective;
        this.thyFormSubmit = new EventEmitter();
    }
    ngOnInit() {
        this.thyFormDirective.onSubmitSuccess = ($event) => {
            this.thyFormSubmit.emit($event);
        };
    }
    onSubmit($event) {
        this.thyFormDirective.submit($event);
    }
}
ThyFormSubmitDirective.ɵfac = function ThyFormSubmitDirective_Factory(t) { return new (t || ThyFormSubmitDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgForm), ɵngcc0.ɵɵdirectiveInject(ThyFormDirective)); };
ThyFormSubmitDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyFormSubmitDirective, selectors: [["", "thyFormSubmit", ""], ["", "thy-form-submit", ""]], hostBindings: function ThyFormSubmitDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyFormSubmitDirective_click_HostBindingHandler($event) { return ctx.onSubmit($event); });
    } }, outputs: { thyFormSubmit: "thyFormSubmit" } });
ThyFormSubmitDirective.ctorParameters = () => [
    { type: NgForm },
    { type: ThyFormDirective }
];
ThyFormSubmitDirective.propDecorators = {
    thyFormSubmit: [{ type: Output }],
    onSubmit: [{ type: HostListener, args: ['click', ['$event'],] }]
};

class ThyFormGroupFooterComponent {
    constructor(thyParentForm, defaultConfig) {
        this.thyParentForm = thyParentForm;
        this.defaultConfig = defaultConfig;
        this._isFormGroup = true;
        this.isHorizontal = true;
        this.footerClassMap = {};
    }
    get align() {
        return !!this.thyAlign ? this.thyAlign : this.defaultConfig.footerAlign;
    }
    ngOnInit() {
        if (this.thyParentForm) {
            this.isHorizontal = this.thyParentForm.isHorizontal;
        }
        this.setFooterClassMap();
    }
    setFooterClassMap() {
        this.footerClassMap = {
            'form-group-footer': true,
            'col-sm-10 offset-sm-2 col-form-control': this.isHorizontal,
            [`form-group-footer-align-${this.align}`]: true
        };
    }
}
ThyFormGroupFooterComponent.ɵfac = function ThyFormGroupFooterComponent_Factory(t) { return new (t || ThyFormGroupFooterComponent)(ɵngcc0.ɵɵdirectiveInject(ThyFormDirective, 8), ɵngcc0.ɵɵdirectiveInject(THY_FORM_CONFIG)); };
ThyFormGroupFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFormGroupFooterComponent, selectors: [["thy-form-group-footer"]], contentQueries: function ThyFormGroupFooterComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c4, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.description = _t.first);
    } }, hostVars: 4, hostBindings: function ThyFormGroupFooterComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("form-group", ctx._isFormGroup)("row", ctx.isHorizontal);
    } }, inputs: { thyAlign: "thyAlign" }, ngContentSelectors: _c3, decls: 4, vars: 2, consts: [[3, "ngClass"], [1, "btn-pair", "form-footer-actions"], ["class", "form-footer-description", 4, "ngIf"], [1, "form-footer-description"], [3, "ngTemplateOutlet"]], template: function ThyFormGroupFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ThyFormGroupFooterComponent_div_3_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ctx.footerClassMap);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.description);
    } }, directives: [ɵngcc3.NgClass, ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet], encapsulation: 2 });
ThyFormGroupFooterComponent.ctorParameters = () => [
    { type: ThyFormDirective, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Inject, args: [THY_FORM_CONFIG,] }] }
];
ThyFormGroupFooterComponent.propDecorators = {
    _isFormGroup: [{ type: HostBinding, args: ['class.form-group',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.row',] }],
    description: [{ type: ContentChild, args: ['description',] }],
    thyAlign: [{ type: Input }]
};

class ThyFormGroupErrorComponent {
    constructor(thyParentForm) {
        this.thyParentForm = thyParentForm;
        this.thyShowFirst = true;
        this.isHorizontal = true;
    }
    set thyErrors(errors) {
        this.errors = errors;
    }
    get thyErrors() {
        const errors = this.errors || this.thyParentForm.validator.errors;
        return errors && errors.length > 0 && this.thyShowFirst ? [errors[0]] : errors;
    }
    get _isFormGroup() {
        return this.thyErrors && this.thyErrors.length > 0;
    }
    ngOnInit() {
        if (this.thyParentForm) {
            this.isHorizontal = this.thyParentForm.isHorizontal;
        }
    }
}
ThyFormGroupErrorComponent.ɵfac = function ThyFormGroupErrorComponent_Factory(t) { return new (t || ThyFormGroupErrorComponent)(ɵngcc0.ɵɵdirectiveInject(ThyFormDirective, 8)); };
ThyFormGroupErrorComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyFormGroupErrorComponent, selectors: [["thy-form-group-error"]], hostVars: 4, hostBindings: function ThyFormGroupErrorComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("row", ctx.isHorizontal)("form-group", ctx._isFormGroup);
    } }, inputs: { thyShowFirst: "thyShowFirst", thyErrors: "thyErrors" }, decls: 2, vars: 4, consts: [[3, "ngClass"], ["thyType", "danger", 3, "thyMessage", 4, "ngFor", "ngForOf"], ["thyType", "danger", 3, "thyMessage"]], template: function ThyFormGroupErrorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyFormGroupErrorComponent_thy_alert_1_Template, 1, 1, "thy-alert", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c5, ctx.isHorizontal));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.thyErrors);
    } }, directives: [ɵngcc3.NgClass, ɵngcc3.NgForOf, ɵngcc5.ThyAlertComponent], encapsulation: 2 });
ThyFormGroupErrorComponent.ctorParameters = () => [
    { type: ThyFormDirective, decorators: [{ type: Optional }] }
];
ThyFormGroupErrorComponent.propDecorators = {
    thyShowFirst: [{ type: Input }],
    thyErrors: [{ type: Input }],
    _isFormGroup: [{ type: HostBinding, args: ['class.form-group',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.row',] }]
};
__decorate([
    InputBoolean(),
    __metadata("design:type", Object)
], ThyFormGroupErrorComponent.prototype, "thyShowFirst", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormValidatorLoader, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [THY_VALIDATOR_CONFIG]
            }] }]; }, null); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormValidatorService, [{
        type: Injectable
    }], function () { return [{ type: ThyFormValidatorLoader }]; }, null); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormDirective, [{
        type: Directive,
        args: [{
                selector: '[thyForm],[thy-form]',
                providers: [UpdateHostClassService, ThyFormValidatorService],
                exportAs: 'thyForm',
                host: {
                    class: 'thy-form'
                }
            }]
    }], function () { return [{ type: ɵngcc1.NgForm }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.NgZone }, { type: ɵngcc2.UpdateHostClassService }, { type: ThyFormValidatorService }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_FORM_CONFIG]
            }] }]; }, { wasValidated: [{
            type: HostBinding,
            args: ['class.was-validated']
        }], thyLayout: [{
            type: Input
        }], thyFormValidatorConfig: [{
            type: Input
        }], thyEnterKeyMode: [{
            type: Input
        }] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-form-group',
                template: "<ng-container *ngIf=\"contentTemplateRef; else defaultTpl\">\n  <ng-container [ngTemplateOutlet]=\"contentTemplateRef\"></ng-container>\n  <ng-container *ngIf=\"feedbackSvgIconName; else iconFont\">\n    <thy-icon class=\"form-control-feedback\" [thyIconName]=\"feedbackSvgIconName\"></thy-icon>\n  </ng-container>\n</ng-container>\n<ng-template #defaultTpl>\n  <label\n    *ngIf=\"!(!isHorizontal && !labelText)\"\n    class=\"form-label \"\n    [ngClass]=\"{\n      'col-sm-2 col-form-label': isHorizontal,\n      'label-required': labelRequired,\n      'pt-0': labelPaddingTopClear\n    }\"\n    >{{ labelText }}</label\n  >\n  <div [ngClass]=\"{ 'col-sm-10 col-form-control': isHorizontal, 'position-relative': !isHorizontal }\">\n    <ng-content></ng-content>\n    <ng-container *ngIf=\"feedbackSvgIconName; else iconFont\">\n      <thy-icon class=\"form-control-feedback\" [thyIconName]=\"feedbackSvgIconName\"></thy-icon>\n    </ng-container>\n    <span *ngIf=\"tips\" class=\"form-text text-desc\">{{ tips }}</span>\n  </div>\n</ng-template>\n\n<ng-template #iconFont>\n  <span *ngIf=\"feedbackIcon\" class=\"form-control-feedback {{ feedbackIcon }}\"></span>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ThyFormDirective, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.ThyTranslate }]; }, { _rowFill: [{
            type: HostBinding,
            args: ['class.row-fill']
        }], _isFormGroup: [{
            type: HostBinding,
            args: ['class.form-group']
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.row']
        }], hasFeedback: [{
            type: HostBinding,
            args: ['class.has-feedback']
        }], thyLabelText: [{
            type: Input
        }], thyLabelTextTranslateKey: [{
            type: Input
        }], thyLabelRequired: [{
            type: Input
        }], thyLabelPaddingTopClear: [{
            type: Input
        }], thyFeedbackIcon: [{
            type: Input
        }], thyTips: [{
            type: Input
        }], thyTipsTranslateKey: [{
            type: Input
        }], thyRowFill: [{
            type: Input
        }], contentTemplateRef: [{
            type: ContentChild,
            args: ['formGroup']
        }] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupLabelDirective, [{
        type: Directive,
        args: [{
                selector: '[thyFormGroupLabel]'
            }]
    }], function () { return [{ type: ɵngcc2.ThyTranslate }]; }, { labelRequired: [{
            type: HostBinding,
            args: ['class.label-required']
        }], _isFormGroupLabel: [{
            type: HostBinding,
            args: ['class.col-form-label']
        }], thyLabelText: [{
            type: Input
        }], thyLabelTranslateKey: [{
            type: Input
        }], thyLabelRequired: [{
            type: Input
        }] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormSubmitDirective, [{
        type: Directive,
        args: [{
                selector: '[thyFormSubmit],[thy-form-submit]'
            }]
    }], function () { return [{ type: ɵngcc1.NgForm }, { type: ThyFormDirective }]; }, { thyFormSubmit: [{
            type: Output
        }], onSubmit: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupFooterComponent, [{
        type: Component,
        args: [{
                selector: 'thy-form-group-footer',
                template: "<div [ngClass]=\"footerClassMap\">\n  <div class=\"btn-pair form-footer-actions\">\n    <ng-content></ng-content>\n  </div>\n\n  <div class=\"form-footer-description\" *ngIf=\"description\">\n    <ng-container [ngTemplateOutlet]=\"description\"></ng-container>\n  </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ThyFormDirective, decorators: [{
                type: Optional
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_FORM_CONFIG]
            }] }]; }, { _isFormGroup: [{
            type: HostBinding,
            args: ['class.form-group']
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.row']
        }], description: [{
            type: ContentChild,
            args: ['description']
        }], thyAlign: [{
            type: Input
        }] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormGroupErrorComponent, [{
        type: Component,
        args: [{
                selector: 'thy-form-group-error',
                template: "<div [ngClass]=\"{ 'col-sm-10 offset-sm-2 col-form-control': isHorizontal }\">\n  <thy-alert *ngFor=\"let error of thyErrors\" thyType=\"danger\" [thyMessage]=\"error\"></thy-alert>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ThyFormDirective, decorators: [{
                type: Optional
            }] }]; }, { thyShowFirst: [{
            type: Input
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.row']
        }], thyErrors: [{
            type: Input
        }], _isFormGroup: [{
            type: HostBinding,
            args: ['class.form-group']
        }] }); })();

class ThyUniqueCheckValidator {
    constructor(elementRef, thyForm) {
        this.elementRef = elementRef;
        this.thyForm = thyForm;
    }
    validate(ctrl) {
        return this.thyUniqueCheck(ctrl.value).pipe(map((failed) => {
            setTimeout(() => {
                if (failed && this.thyForm && this.elementRef.nativeElement.name) {
                    this.thyForm.validator.validateControl(this.elementRef.nativeElement.name);
                }
            });
            return failed ? { thyUniqueCheck: failed } : null;
        }), catchError(() => of(null)));
    }
}
ThyUniqueCheckValidator.ɵfac = function ThyUniqueCheckValidator_Factory(t) { return new (t || ThyUniqueCheckValidator)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ThyFormDirective, 8)); };
ThyUniqueCheckValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyUniqueCheckValidator, selectors: [["", "thyUniqueCheck", ""]], inputs: { thyUniqueCheck: "thyUniqueCheck" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_ASYNC_VALIDATORS,
                useExisting: ThyUniqueCheckValidator,
                multi: true
            }
        ])] });
ThyUniqueCheckValidator.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyFormDirective, decorators: [{ type: Optional }] }
];
ThyUniqueCheckValidator.propDecorators = {
    thyUniqueCheck: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyUniqueCheckValidator, [{
        type: Directive,
        args: [{
                selector: '[thyUniqueCheck]',
                providers: [
                    {
                        provide: NG_ASYNC_VALIDATORS,
                        useExisting: ThyUniqueCheckValidator,
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ThyFormDirective, decorators: [{
                type: Optional
            }] }]; }, { thyUniqueCheck: [{
            type: Input
        }] }); })();

class ThyMinDirective {
    set min(value) {
        this._validator = Validators.min(parseFloat(value));
    }
    validate(control) {
        return this._validator(control);
    }
}
ThyMinDirective.ɵfac = function ThyMinDirective_Factory(t) { return new (t || ThyMinDirective)(); };
ThyMinDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMinDirective, selectors: [["", "min", "", "formControlName", ""], ["", "min", "", "formControl", ""], ["", "min", "", "ngModel", ""]], inputs: { min: "min" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => ThyMinDirective),
                multi: true
            }
        ])] });
ThyMinDirective.propDecorators = {
    min: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMinDirective, [{
        type: Directive,
        args: [{
                selector: '[min][formControlName],[min][formControl],[min][ngModel]',
                providers: [
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => ThyMinDirective),
                        multi: true
                    }
                ]
            }]
    }], null, { min: [{
            type: Input
        }] }); })();

class ThyMaxDirective {
    set max(value) {
        this._validator = Validators.max(parseFloat(value));
    }
    validate(control) {
        return this._validator(control);
    }
}
ThyMaxDirective.ɵfac = function ThyMaxDirective_Factory(t) { return new (t || ThyMaxDirective)(); };
ThyMaxDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyMaxDirective, selectors: [["", "max", "", "formControlName", ""], ["", "max", "", "formControl", ""], ["", "max", "", "ngModel", ""]], inputs: { max: "max" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => ThyMaxDirective),
                multi: true
            }
        ])] });
ThyMaxDirective.propDecorators = {
    max: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMaxDirective, [{
        type: Directive,
        args: [{
                selector: '[max][formControlName],[max][formControl],[max][ngModel]',
                providers: [
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => ThyMaxDirective),
                        multi: true
                    }
                ]
            }]
    }], null, { max: [{
            type: Input
        }] }); })();

function confirmValidator(value) {
    return (control) => {
        const isEqual = value === control.value;
        return !isEqual ? { confirm: { value: value, actual: control.value } } : null;
    };
}
class ThyConfirmValidatorDirective {
    validate(control) {
        return confirmValidator(this.confirm)(control);
    }
}
ThyConfirmValidatorDirective.ɵfac = function ThyConfirmValidatorDirective_Factory(t) { return new (t || ThyConfirmValidatorDirective)(); };
ThyConfirmValidatorDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyConfirmValidatorDirective, selectors: [["", "confirm", ""]], inputs: { confirm: "confirm" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: ThyConfirmValidatorDirective, multi: true }])] });
ThyConfirmValidatorDirective.propDecorators = {
    confirm: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyConfirmValidatorDirective, [{
        type: Directive,
        args: [{
                selector: '[confirm]',
                providers: [{ provide: NG_VALIDATORS, useExisting: ThyConfirmValidatorDirective, multi: true }]
            }]
    }], null, { confirm: [{
            type: Input
        }] }); })();

class ThyFormModule {
    static forRoot(config) {
        return {
            ngModule: ThyFormModule,
            providers: [
                {
                    provide: THY_VALIDATOR_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
ThyFormModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyFormModule });
ThyFormModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyFormModule_Factory(t) { return new (t || ThyFormModule)(); }, providers: [ThyFormValidatorLoader, THY_FORM_CONFIG_PROVIDER], imports: [[CommonModule, FormsModule, ThyInputModule, ThyAlertModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyFormModule, { declarations: function () { return [ThyFormDirective, ThyFormGroupComponent, ThyFormGroupLabelDirective, ThyFormSubmitDirective, ThyFormGroupFooterComponent, ThyUniqueCheckValidator, ThyFormGroupErrorComponent, ThyMinDirective, ThyMaxDirective, ThyConfirmValidatorDirective]; }, imports: function () { return [CommonModule, FormsModule, ThyInputModule, ThyAlertModule, ThyIconModule]; }, exports: function () { return [ThyFormDirective, ThyFormGroupComponent, ThyFormGroupLabelDirective, ThyFormSubmitDirective, ThyFormGroupFooterComponent, ThyUniqueCheckValidator, ThyFormGroupErrorComponent, ThyMinDirective, ThyMaxDirective, ThyConfirmValidatorDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFormModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ThyInputModule, ThyAlertModule, ThyIconModule],
                declarations: [
                    ThyFormDirective,
                    ThyFormGroupComponent,
                    ThyFormGroupLabelDirective,
                    ThyFormSubmitDirective,
                    ThyFormGroupFooterComponent,
                    ThyUniqueCheckValidator,
                    ThyFormGroupErrorComponent,
                    ThyMinDirective,
                    ThyMaxDirective,
                    ThyConfirmValidatorDirective
                ],
                exports: [
                    ThyFormDirective,
                    ThyFormGroupComponent,
                    ThyFormGroupLabelDirective,
                    ThyFormSubmitDirective,
                    ThyFormGroupFooterComponent,
                    ThyUniqueCheckValidator,
                    ThyFormGroupErrorComponent,
                    ThyMinDirective,
                    ThyMaxDirective,
                    ThyConfirmValidatorDirective
                ],
                providers: [ThyFormValidatorLoader, THY_FORM_CONFIG_PROVIDER]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ERROR_VALUE_REPLACE_REGEX, THY_FORM_CONFIG, THY_FORM_CONFIG_PROVIDER, THY_VALIDATOR_CONFIG, ThyConfirmValidatorDirective, ThyEnterKeyMode, ThyFormDirective, ThyFormGroupComponent, ThyFormGroupErrorComponent, ThyFormGroupFooterComponent, ThyFormGroupLabelDirective, ThyFormModule, ThyFormSubmitDirective, ThyFormValidatorLoader, ThyFormValidatorService, ThyMaxDirective, ThyMinDirective, ThyUniqueCheckValidator, confirmValidator, ɵ0 };
