// ngx-tethys/fesm2015/ngx-tethys-skeleton.js
import { Component, Inject, PLATFORM_ID, HostBinding, Input, ContentChild, ViewChild, NgModule } from '@angular/core';
import { isPlatformBrowser, CommonModule } from '@angular/common';
import { PortalModule } from '@angular/cdk/portal';
import { generateRandomStr } from 'ngx-tethys/util';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["content"];
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1__svg_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const content_r8 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", content_r8);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_4__svg_ng_container_1_Template, 2, 1, "ng-container", 7);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.contentTemplates);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_5__svg_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainer(0);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent__svg_svg_0__svg_ng_container_5__svg_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r3.customTemplate);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r4.animationValues[0])("dur", ctx_r4.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r5.animationValues[1])("dur", ctx_r5.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0__svg_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "animate", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("values", ctx_r6.animationValues[2])("dur", ctx_r6.thySpeed);
} }
function ThySkeletonComponent__svg_svg_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 2);
    ɵngcc0.ɵɵelement(1, "rect", 3);
    ɵngcc0.ɵɵelementStart(2, "defs");
    ɵngcc0.ɵɵelementStart(3, "clipPath");
    ɵngcc0.ɵɵtemplate(4, ThySkeletonComponent__svg_svg_0__svg_ng_container_4_Template, 2, 1, "ng-container", 1);
    ɵngcc0.ɵɵtemplate(5, ThySkeletonComponent__svg_svg_0__svg_ng_container_5_Template, 2, 1, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "linearGradient");
    ɵngcc0.ɵɵelementStart(7, "stop", 4);
    ɵngcc0.ɵɵtemplate(8, ThySkeletonComponent__svg_svg_0__svg_ng_container_8_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "stop", 5);
    ɵngcc0.ɵɵtemplate(10, ThySkeletonComponent__svg_svg_0__svg_ng_container_10_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "stop", 6);
    ɵngcc0.ɵɵtemplate(12, ThySkeletonComponent__svg_svg_0__svg_ng_container_12_Template, 2, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattributeInterpolate2("viewBox", "0 0 ", ctx_r0.thyViewBoxWidth, " ", ctx_r0.thyViewBoxHeight, "");
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r0.thyStyle);
    ɵngcc0.ɵɵattribute("preserveAspectRatio", ctx_r0.thyPreserveAspectRatio)("height", ctx_r0.thyHeight)("width", ctx_r0.thyWidth);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ctx_r0.fillStyle);
    ɵngcc0.ɵɵattribute("clip-path", ctx_r0.clipPath)("width", ctx_r0.thyWidth)("height", ctx_r0.thyHeight);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("id", ctx_r0.idClip);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.contentTemplates);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.customTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("id", ctx_r0.idGradient);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thyPrimaryColor)("stop-opacity", ctx_r0.thyPrimaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thySecondaryColor)("stop-opacity", ctx_r0.thySecondaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("stop-color", ctx_r0.thyPrimaryColor)("stop-opacity", ctx_r0.thyPrimaryOpacity);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.thyAnimate);
} }
function ThySkeletonComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementContainerEnd();
} }
const _c1 = ["*"];
function ThySkeletonListComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
    ɵngcc0.ɵɵelement(1, "rect", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelement(3, "rect", 4);
    ɵngcc0.ɵɵelement(4, "rect", 5);
    ɵngcc0.ɵɵelement(5, "rect", 6);
} }
function ThySkeletonBulletListComponent_ng_template_0_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(1, "circle", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("cx", item_r3.circle.cx)("cy", item_r3.circle.cy);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("x", item_r3.rect.x)("y", item_r3.rect.y);
} }
function ThySkeletonBulletListComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThySkeletonBulletListComponent_ng_template_0_ng_container_0_Template, 3, 4, "ng-container", 1);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.items);
} }
function ThySkeletonTitleComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("width", ctx_r1.width)("height", ctx_r1.height);
} }
function ThySkeletonAvatarComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "circle", 1);
} }
function ThySkeletonParagraphComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelement(0, "rect", 1);
    ɵngcc0.ɵɵelement(1, "rect", 2);
    ɵngcc0.ɵɵelement(2, "rect", 3);
    ɵngcc0.ɵɵelement(3, "rect", 4);
} }
class ThySkeletonComponent {
    constructor(platformId) {
        this.platformId = platformId;
        this.addSkeletonClass = true;
        this.thyAnimate = true;
        this.thyBaseUrl = '';
        this.thyWidth = '100%';
        this.thyHeight = '100%';
        this.thyViewBoxWidth = 400;
        this.thyViewBoxHeight = 130;
        this.thySpeed = 2;
        this.thyPreserveAspectRatio = 'none'; // xMidYMid meet
        this.thyPrimaryColor = '#f0f0f0';
        this.thySecondaryColor = '#e0e0e0';
        this.thyPrimaryOpacity = 1;
        this.thySecondaryOpacity = 1;
        this.thyIgnoreBaseUrl = false;
        this.thyLoadingDone = false;
        this.contentTemplates = [];
        this.idClip = generateRandomStr();
        this.idGradient = generateRandomStr();
        this.defaultAnimation = ['-3; 1', '-2; 2', '-1; 3'];
        this.rtlAnimation = ['1; -3', '2; -2', '3; -1'];
    }
    ngOnInit() {
        this.animationValues = this.thyRtl ? this.rtlAnimation : this.defaultAnimation;
        if (this.thyBaseUrl === '' && !this.thyIgnoreBaseUrl && isPlatformBrowser(this.platformId)) {
            this.thyBaseUrl = window.location.pathname;
        }
        this.setFillStyle();
        this.setClipPath();
    }
    ngOnChanges(changes) {
        if (changes['thyBaseUrl']) {
            if (changes['thyBaseUrl'].previousValue !== changes['thyBaseUrl'].currentValue) {
                this.setFillStyle();
                this.setClipPath();
            }
        }
    }
    setFillStyle() {
        this.fillStyle = {
            fill: `url(${this.thyBaseUrl}#${this.idGradient})`
        };
    }
    setClipPath() {
        this.clipPath = `url(${this.thyBaseUrl}#${this.idClip})`;
    }
    assignInputProperties(input) {
        this.thyAnimate = input.thyAnimate;
        this.thyBaseUrl = input.thyBaseUrl;
        this.thyHeight = input.thyHeight;
        this.thyWidth = input.thyWidth;
        this.thyViewBoxHeight = input.thyViewBoxHeight;
        this.thyViewBoxWidth = input.thyViewBoxWidth;
        this.thyPreserveAspectRatio = input.thyPreserveAspectRatio;
        this.thyPrimaryColor = input.thyPrimaryColor;
        this.thyPrimaryOpacity = input.thyPrimaryOpacity;
        this.thySecondaryColor = input.thySecondaryColor;
        this.thySecondaryOpacity = input.thySecondaryOpacity;
        this.thyRtl = this.thyRtl;
        this.thySpeed = this.thySpeed;
        this.thyUniqueKey = this.thyUniqueKey;
    }
    addTemplate(template) {
        if (template) {
            this.contentTemplates.push(template);
        }
    }
}
ThySkeletonComponent.ɵfac = function ThySkeletonComponent_Factory(t) { return new (t || ThySkeletonComponent)(ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID)); };
ThySkeletonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonComponent, selectors: [["thy-skeleton"]], contentQueries: function ThySkeletonComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customTemplate = _t.first);
    } }, hostVars: 2, hostBindings: function ThySkeletonComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-skeleton", ctx.addSkeletonClass);
    } }, inputs: { thyAnimate: "thyAnimate", thyBaseUrl: "thyBaseUrl", thyWidth: "thyWidth", thyHeight: "thyHeight", thyViewBoxWidth: "thyViewBoxWidth", thyViewBoxHeight: "thyViewBoxHeight", thySpeed: "thySpeed", thyPreserveAspectRatio: "thyPreserveAspectRatio", thyPrimaryColor: "thyPrimaryColor", thySecondaryColor: "thySecondaryColor", thyPrimaryOpacity: "thyPrimaryOpacity", thySecondaryOpacity: "thySecondaryOpacity", thyIgnoreBaseUrl: "thyIgnoreBaseUrl", thyLoadingDone: "thyLoadingDone", thyRtl: "thyRtl", thyUniqueKey: "thyUniqueKey", thyStyle: "thyStyle" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c1, decls: 2, vars: 2, consts: [[3, "ngStyle", 4, "ngIf"], [4, "ngIf"], [3, "ngStyle"], ["x", "0", "y", "0", 3, "ngStyle"], ["offset", "0%"], ["offset", "50%"], ["offset", "100%"], [4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["attributeName", "offset", "repeatCount", "indefinite"]], template: function ThySkeletonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ThySkeletonComponent__svg_svg_0_Template, 13, 23, "svg", 0);
        ɵngcc0.ɵɵtemplate(1, ThySkeletonComponent_ng_container_1_Template, 2, 0, "ng-container", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.thyLoadingDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyLoadingDone);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
ThySkeletonComponent.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
ThySkeletonComponent.propDecorators = {
    addSkeletonClass: [{ type: HostBinding, args: ['class.thy-skeleton',] }],
    thyAnimate: [{ type: Input }],
    thyBaseUrl: [{ type: Input }],
    thyWidth: [{ type: Input }],
    thyHeight: [{ type: Input }],
    thyViewBoxWidth: [{ type: Input }],
    thyViewBoxHeight: [{ type: Input }],
    thySpeed: [{ type: Input }],
    thyPreserveAspectRatio: [{ type: Input }],
    thyPrimaryColor: [{ type: Input }],
    thySecondaryColor: [{ type: Input }],
    thyPrimaryOpacity: [{ type: Input }],
    thySecondaryOpacity: [{ type: Input }],
    thyUniqueKey: [{ type: Input }],
    thyRtl: [{ type: Input }],
    thyStyle: [{ type: Input }],
    thyIgnoreBaseUrl: [{ type: Input }],
    thyLoadingDone: [{ type: Input }],
    customTemplate: [{ type: ContentChild, args: ['content',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton',
                template: "<svg\n  *ngIf=\"!thyLoadingDone\"\n  attr.viewBox=\"0 0 {{ thyViewBoxWidth }} {{ thyViewBoxHeight }}\"\n  [attr.preserveAspectRatio]=\"thyPreserveAspectRatio\"\n  [ngStyle]=\"thyStyle\"\n  [attr.height]=\"thyHeight\"\n  [attr.width]=\"thyWidth\"\n>\n  <rect\n    [attr.clip-path]=\"clipPath\"\n    [ngStyle]=\"fillStyle\"\n    x=\"0\"\n    y=\"0\"\n    [attr.width]=\"thyWidth\"\n    [attr.height]=\"thyHeight\"\n  />\n\n  <defs>\n    <clipPath [attr.id]=\"idClip\">\n      <ng-container *ngIf=\"contentTemplates\">\n        <ng-container *ngFor=\"let content of contentTemplates\">\n          <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n        </ng-container>\n      </ng-container>\n\n      <ng-container *ngIf=\"customTemplate\">\n        <ng-container *ngTemplateOutlet=\"customTemplate\"></ng-container>\n      </ng-container>\n    </clipPath>\n\n    <linearGradient [attr.id]=\"idGradient\">\n      <stop offset=\"0%\" [attr.stop-color]=\"thyPrimaryColor\" [attr.stop-opacity]=\"thyPrimaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[0]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n\n      <stop offset=\"50%\" [attr.stop-color]=\"thySecondaryColor\" [attr.stop-opacity]=\"thySecondaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[1]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n\n      <stop offset=\"100%\" [attr.stop-color]=\"thyPrimaryColor\" [attr.stop-opacity]=\"thyPrimaryOpacity\">\n        <ng-container *ngIf=\"thyAnimate\">\n          <animate\n            attributeName=\"offset\"\n            [attr.values]=\"animationValues[2]\"\n            attr.dur=\"{{ thySpeed }}\"\n            repeatCount=\"indefinite\"\n          />\n        </ng-container>\n      </stop>\n    </linearGradient>\n  </defs>\n</svg>\n<ng-container *ngIf=\"thyLoadingDone\">\n  <ng-content></ng-content>\n</ng-container>\n"
            }]
    }], function () { return [{ type: String, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, { addSkeletonClass: [{
            type: HostBinding,
            args: ['class.thy-skeleton']
        }], thyAnimate: [{
            type: Input
        }], thyBaseUrl: [{
            type: Input
        }], thyWidth: [{
            type: Input
        }], thyHeight: [{
            type: Input
        }], thyViewBoxWidth: [{
            type: Input
        }], thyViewBoxHeight: [{
            type: Input
        }], thySpeed: [{
            type: Input
        }], thyPreserveAspectRatio: [{
            type: Input
        }], thyPrimaryColor: [{
            type: Input
        }], thySecondaryColor: [{
            type: Input
        }], thyPrimaryOpacity: [{
            type: Input
        }], thySecondaryOpacity: [{
            type: Input
        }], thyIgnoreBaseUrl: [{
            type: Input
        }], thyLoadingDone: [{
            type: Input
        }], thyRtl: [{
            type: Input
        }], thyUniqueKey: [{
            type: Input
        }], thyStyle: [{
            type: Input
        }], customTemplate: [{
            type: ContentChild,
            args: ['content']
        }] }); })();

class ThySkeletonListComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonListComponent.ɵfac = function ThySkeletonListComponent_Factory(t) { return new (t || ThySkeletonListComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonListComponent, selectors: [["thy-skeleton-list-template"]], viewQuery: function ThySkeletonListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, decls: 2, vars: 0, consts: [["content", ""], ["x", "0", "y", "0", "rx", "3", "ry", "3", "width", "250", "height", "10"], ["x", "20", "y", "20", "rx", "3", "ry", "3", "width", "220", "height", "10"], ["x", "20", "y", "40", "rx", "3", "ry", "3", "width", "170", "height", "10"], ["x", "0", "y", "60", "rx", "3", "ry", "3", "width", "250", "height", "10"], ["x", "20", "y", "80", "rx", "3", "ry", "3", "width", "200", "height", "10"], ["x", "20", "y", "100", "rx", "3", "ry", "3", "width", "80", "height", "10"]], template: function ThySkeletonListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonListComponent_ng_template_0_Template, 6, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonListComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonListComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonListComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-list-template',
                template: `
        <ng-template #content>
            <svg:rect x="0" y="0" rx="3" ry="3" width="250" height="10" />
            <svg:rect x="20" y="20" rx="3" ry="3" width="220" height="10" />
            <svg:rect x="20" y="40" rx="3" ry="3" width="170" height="10" />
            <svg:rect x="0" y="60" rx="3" ry="3" width="250" height="10" />
            <svg:rect x="20" y="80" rx="3" ry="3" width="200" height="10" />
            <svg:rect x="20" y="100" rx="3" ry="3" width="80" height="10" />
        </ng-template>
    `
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();

class ThySkeletonBulletListComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
        this.thyCount = 5;
        this.items = [];
    }
    ngOnInit() {
        for (let i = 0; i <= this.thyCount; i++) {
            this.items.push({
                circle: {
                    cx: 10,
                    cy: i * 30 + 20,
                    r: 8
                },
                rect: {
                    x: 25,
                    y: i * 2 * 15 + 15,
                    rx: 5,
                    ry: 5,
                    width: '90%',
                    height: 10
                }
            });
        }
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonBulletListComponent.ɵfac = function ThySkeletonBulletListComponent_Factory(t) { return new (t || ThySkeletonBulletListComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonBulletListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonBulletListComponent, selectors: [["thy-skeleton-bullet-list-template"]], viewQuery: function ThySkeletonBulletListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, inputs: { thyCount: "thyCount" }, decls: 2, vars: 0, consts: [["content", ""], [4, "ngFor", "ngForOf"], ["r", "8"], ["rx", "5", "ry", "5", "width", "90%", "height", "10"]], template: function ThySkeletonBulletListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonBulletListComponent_ng_template_0_Template, 1, 1, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc1.NgForOf], encapsulation: 2 });
ThySkeletonBulletListComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonBulletListComponent.propDecorators = {
    thyCount: [{ type: Input }],
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonBulletListComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-bullet-list-template',
                template: `
        <ng-template #content>
            <ng-container *ngFor="let item of items">
                <svg:circle [attr.cx]="item.circle.cx" [attr.cy]="item.circle.cy" r="8" />
                <svg:rect [attr.x]="item.rect.x" [attr.y]="item.rect.y" rx="5" ry="5" width="90%" height="10" />
            </ng-container>
        </ng-template>
    `
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { thyCount: [{
            type: Input
        }], contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();

class ThySkeletonTitleComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
        this.width = 100;
        this.height = 10;
    }
    set thyWidth(value) {
        if (value) {
            this.width = value;
        }
    }
    set thyHeight(value) {
        if (value) {
            this.height = value;
        }
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonTitleComponent.ɵfac = function ThySkeletonTitleComponent_Factory(t) { return new (t || ThySkeletonTitleComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonTitleComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonTitleComponent, selectors: [["thy-skeleton-title-template"], ["", "thySkeletonTitleTemplate", ""]], viewQuery: function ThySkeletonTitleComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, inputs: { thyWidth: "thyWidth", thyHeight: "thyHeight" }, exportAs: ["thySkeletonTitleTemplate"], decls: 2, vars: 0, consts: [["content", ""], ["x", "15", "y", "13", "rx", "4", "ry", "4"]], template: function ThySkeletonTitleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonTitleComponent_ng_template_0_Template, 1, 2, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonTitleComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonTitleComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }],
    thyWidth: [{ type: Input }],
    thyHeight: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonTitleComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-title-template,[thySkeletonTitleTemplate]',
                template: `
        <ng-template #content>
            <svg:rect x="15" y="13" rx="4" ry="4" [attr.width]="width" [attr.height]="height" />
        </ng-template>
    `,
                exportAs: 'thySkeletonTitleTemplate'
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { thyWidth: [{
            type: Input
        }], thyHeight: [{
            type: Input
        }], contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();

class ThySkeletonAvatarComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonAvatarComponent.ɵfac = function ThySkeletonAvatarComponent_Factory(t) { return new (t || ThySkeletonAvatarComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonAvatarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonAvatarComponent, selectors: [["thy-skeleton-avatar-template"], ["", "thySkeletonAvatarTemplate", ""]], viewQuery: function ThySkeletonAvatarComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, decls: 2, vars: 0, consts: [["content", ""], ["cx", "30", "cy", "30", "r", "30"]], template: function ThySkeletonAvatarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonAvatarComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonAvatarComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonAvatarComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonAvatarComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-avatar-template,[thySkeletonAvatarTemplate]',
                template: `
        <ng-template #content>
            <svg:circle cx="30" cy="30" r="30" />
        </ng-template>
    `
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();

class ThySkeletonParagraphComponent {
    constructor(skeletonComponent) {
        this.skeletonComponent = skeletonComponent;
    }
    ngOnInit() {
        this.skeletonComponent.addTemplate(this.contentTemplateRef);
    }
}
ThySkeletonParagraphComponent.ɵfac = function ThySkeletonParagraphComponent_Factory(t) { return new (t || ThySkeletonParagraphComponent)(ɵngcc0.ɵɵdirectiveInject(ThySkeletonComponent)); };
ThySkeletonParagraphComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySkeletonParagraphComponent, selectors: [["thy-skeleton-paragraph-template"], ["", "thySkeletonParagraphTemplate", ""]], viewQuery: function ThySkeletonParagraphComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
    } }, exportAs: ["thySkeletonParagraphTemplate"], decls: 2, vars: 0, consts: [["content", ""], ["x", "0", "y", "0", "rx", "2", "ry", "2", "width", "38%", "height", "10"], ["x", "20", "y", "20", "rx", "2", "ry", "2", "width", "92%", "height", "10"], ["x", "20", "y", "40", "rx", "2", "ry", "2", "width", "92%", "height", "10"], ["x", "0", "y", "60", "rx", "2", "ry", "2", "width", "61%", "height", "10"]], template: function ThySkeletonParagraphComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThySkeletonParagraphComponent_ng_template_0_Template, 4, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ThySkeletonParagraphComponent.ctorParameters = () => [
    { type: ThySkeletonComponent }
];
ThySkeletonParagraphComponent.propDecorators = {
    contentTemplateRef: [{ type: ViewChild, args: ['content', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonParagraphComponent, [{
        type: Component,
        args: [{
                selector: 'thy-skeleton-paragraph-template,[thySkeletonParagraphTemplate]',
                template: `
        <ng-template #content>
            <svg:rect x="0" y="0" rx="2" ry="2" width="38%" height="10" />
            <svg:rect x="20" y="20" rx="2" ry="2" width="92%" height="10" />
            <svg:rect x="20" y="40" rx="2" ry="2" width="92%" height="10" />
            <svg:rect x="0" y="60" rx="2" ry="2" width="61%" height="10" />
        </ng-template>
    `,
                exportAs: 'thySkeletonParagraphTemplate'
            }]
    }], function () { return [{ type: ThySkeletonComponent }]; }, { contentTemplateRef: [{
            type: ViewChild,
            args: ['content', { static: true }]
        }] }); })();

class ThySkeletonModule {
}
ThySkeletonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySkeletonModule });
ThySkeletonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySkeletonModule_Factory(t) { return new (t || ThySkeletonModule)(); }, providers: [], imports: [[CommonModule, PortalModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySkeletonModule, { declarations: function () { return [ThySkeletonComponent, ThySkeletonListComponent, ThySkeletonBulletListComponent, ThySkeletonAvatarComponent, ThySkeletonTitleComponent, ThySkeletonParagraphComponent]; }, imports: function () { return [CommonModule, PortalModule]; }, exports: function () { return [ThySkeletonComponent, ThySkeletonListComponent, ThySkeletonBulletListComponent, ThySkeletonTitleComponent, ThySkeletonAvatarComponent, ThySkeletonParagraphComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySkeletonModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThySkeletonComponent,
                    ThySkeletonListComponent,
                    ThySkeletonBulletListComponent,
                    ThySkeletonAvatarComponent,
                    ThySkeletonTitleComponent,
                    ThySkeletonParagraphComponent
                ],
                imports: [CommonModule, PortalModule],
                entryComponents: [ThySkeletonComponent],
                exports: [
                    ThySkeletonComponent,
                    ThySkeletonListComponent,
                    ThySkeletonBulletListComponent,
                    ThySkeletonTitleComponent,
                    ThySkeletonAvatarComponent,
                    ThySkeletonParagraphComponent
                ],
                providers: []
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThySkeletonAvatarComponent, ThySkeletonBulletListComponent, ThySkeletonComponent, ThySkeletonListComponent, ThySkeletonModule, ThySkeletonParagraphComponent, ThySkeletonTitleComponent };
