// ngx-tethys/fesm2015/ngx-tethys-dialog.js
import * as i1 from '@angular/cdk/overlay';
import { Overlay, OverlayModule } from '@angular/cdk/overlay';
import { CdkPortalOutlet, ComponentPortal, PortalModule } from '@angular/cdk/portal';
import { DOCUMENT, CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { InjectionToken, EventEmitter, Component, ChangeDetectionStrategy, ElementRef, Inject, ChangeDetectorRef, ViewChild, HostBinding, Input, Injector, Injectable, Optional, ContentChild, Output, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThyFormModule } from 'ngx-tethys/form';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThySharedModule } from 'ngx-tethys/shared';
import { of } from 'rxjs';
import * as i3 from 'ngx-tethys/core';
import { ThyAbstractOverlayConfig, ThyAbstractOverlayContainer, ThyClickPositioner, ThyAbstractOverlayRef, ThyAbstractInternalOverlayRef, ThyAbstractOverlayService, UpdateHostClassService, ThyTranslate } from 'ngx-tethys/core';
import { state, style, transition, animate, trigger } from '@angular/animations';
import { FocusTrapFactory } from '@angular/cdk/a11y';
import { filter, finalize } from 'rxjs/operators';
import { Directionality } from '@angular/cdk/bidi';
import { coerceBooleanProperty } from 'ngx-tethys/util';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '@angular/cdk/a11y';
import * as ɵngcc3 from '@angular/cdk/portal';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from 'ngx-tethys/form';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from 'ngx-tethys/button';
import * as ɵngcc8 from '@angular/cdk/overlay';
import * as ɵngcc9 from 'ngx-tethys/icon';

function ThyDialogContainerComponent_ng_template_0_Template(rf, ctx) { }
const _c0 = ["*"];
const _c1 = ["description"];
function DialogFooterComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementContainer(1, 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.description);
} }
const _c2 = ["dialogHeader"];
function DialogHeaderComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "template", 2);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.headerTemplate);
} }
function DialogHeaderComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "h3", 3);
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "button", 5);
    ɵngcc0.ɵɵlistener("click", function DialogHeaderComponent_ng_template_1_Template_button_click_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.close($event); });
    ɵngcc0.ɵɵelement(4, "thy-icon", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r2.thyIcon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.thyTitle);
} }
var ThyDialogSizes;
(function (ThyDialogSizes) {
    ThyDialogSizes["lg"] = "lg";
    ThyDialogSizes["supperLg"] = "supper-lg";
    ThyDialogSizes["maxLg"] = "max-lg";
    ThyDialogSizes["md"] = "md";
    ThyDialogSizes["sm"] = "sm";
    ThyDialogSizes["full"] = "full";
})(ThyDialogSizes || (ThyDialogSizes = {}));
/**
 * Configuration for opening a modal dialog with the ThyDialog service.
 */
class ThyDialogConfig extends ThyAbstractOverlayConfig {
    constructor() {
        super(...arguments);
        /** The ARIA role of the dialog element. */
        this.role = 'dialog';
    }
}
/** Injection token that can be used to specify default dialog options. */
const THY_DIALOG_DEFAULT_OPTIONS = new InjectionToken('thy-dialog-default-options');
const ɵ0 = {
    role: 'dialog',
    hasBackdrop: true,
    backdropClass: '',
    panelClass: '',
    backdropClosable: true,
    closeOnNavigation: true,
    autoFocus: true,
    restoreFocus: true
};
const THY_DIALOG_DEFAULT_OPTIONS_PROVIDER = {
    provide: THY_DIALOG_DEFAULT_OPTIONS,
    useValue: ɵ0
};
class ThyDialogLayoutConfig {
}
const THY_DIALOG_LAYOUT_CONFIG = new InjectionToken('thy-dialog-layout-config');
const ɵ1 = {
    footerAlign: 'left',
    footerDivided: false
};
const THY_DIALOG_LAYOUT_CONFIG_PROVIDER = {
    provide: THY_DIALOG_LAYOUT_CONFIG,
    useValue: ɵ1
};

const animationBody = [
    // Note: The `enter` animation transitions to `transform: none`, because for some reason
    // specifying the transform explicitly, causes IE both to blur the dialog content and
    // decimate the animation performance. Leaving it as `none` solves both issues. translate(-300px,-100px)
    state('void, exit', style({ opacity: 0, transform: 'scale(0.1)' })),
    state('enter', style({ transform: 'none' })),
    transition('* => enter', animate('200ms cubic-bezier(0, 0, 0.57, 1)', style({ transform: 'none', opacity: 1 }))),
    transition('* => exit', animate('150ms cubic-bezier(0.24, 0.18, 0.96, 0.52)', style({ transform: 'scale(0.3)', opacity: 0 })))
];
/**
 * Animations used by MatDialog.
 * @docs-private
 */
const thyDialogAnimations = {
    dialogContainer: trigger('dialogContainer', animationBody)
};
// export const setAnimationStart

const dialogUpperOverlayOptions = {
    name: 'dialog',
    animationEnabled: true,
    disposeWhenClose: true
};

class ThyDialogContainerComponent extends ThyAbstractOverlayContainer {
    constructor(elementRef, document, config, changeDetectorRef, clickPositioner, focusTrapFactory) {
        super(dialogUpperOverlayOptions, changeDetectorRef);
        this.elementRef = elementRef;
        this.document = document;
        this.config = config;
        this.clickPositioner = clickPositioner;
        this.focusTrapFactory = focusTrapFactory;
        /** State of the dialog animation. */
        this.animationState = 'void';
        /** Emits when an animation state changes. */
        this.animationStateChanged = new EventEmitter();
        /** ID of the element that should be considered as the dialog's label. */
        this.ariaLabelledBy = null;
        /** Element that was focused before the dialog was opened. Save this to restore upon close. */
        this.elementFocusedBeforeDialogWasOpened = null;
        this.animationOpeningDone = this.animationStateChanged.pipe(filter((event) => {
            return event.phaseName === 'done' && event.toState === 'void';
        }));
        this.animationClosingDone = this.animationStateChanged.pipe(filter((event) => {
            return event.phaseName === 'done' && event.toState === 'exit';
        }));
    }
    savePreviouslyFocusedElement() {
        if (this.document) {
            this.elementFocusedBeforeDialogWasOpened = this.document.activeElement;
            // Note that there is no focus method when rendering on the server.
            if (this.elementRef.nativeElement.focus) {
                // Move focus onto the dialog immediately in order to prevent the user from accidentally
                // opening multiple dialogs at the same time. Needs to be async, because the element
                // may not be focusable immediately.
                Promise.resolve().then(() => this.elementRef.nativeElement.focus());
            }
        }
    }
    /** Moves the focus inside the focus trap. */
    trapFocus() {
        const element = this.elementRef.nativeElement;
        if (!this.focusTrap) {
            this.focusTrap = this.focusTrapFactory.create(element);
        }
        // If we were to attempt to focus immediately, then the content of the dialog would not yet be
        // ready in instances where change detection has to run first. To deal with this, we simply
        // wait for the microtask queue to be empty.
        if (this.config.autoFocus) {
            this.focusTrap.focusInitialElementWhenReady();
        }
        else {
            const activeElement = this.document.activeElement;
            // Otherwise ensure that focus is on the dialog container. It's possible that a different
            // component tried to move focus while the open animation was running. See:
            // https://github.com/angular/components/issues/16215. Note that we only want to do this
            // if the focus isn't inside the dialog already, because it's possible that the consumer
            // turned off `autoFocus` in order to move focus themselves.
            if (activeElement !== element && !element.contains(activeElement)) {
                element.focus();
            }
        }
    }
    restoreFocus() {
        const toFocus = this.elementFocusedBeforeDialogWasOpened;
        // We need the extra check, because IE can set the `activeElement` to null in some cases.
        if (this.config.restoreFocus && toFocus && typeof toFocus.focus === 'function') {
            toFocus.focus();
        }
        if (this.focusTrap) {
            this.focusTrap.destroy();
        }
    }
    setTransformOrigin() {
        this.clickPositioner.runTaskUseLastPosition(lastPosition => {
            if (lastPosition) {
                const containerElement = this.elementRef.nativeElement;
                const transformOrigin = `${lastPosition.x - containerElement.offsetLeft}px ${lastPosition.y -
                    containerElement.offsetTop}px 0px`;
                containerElement.style['transform-origin'] = transformOrigin;
                // 手动修改动画状态为从 void 到 enter, 开启动画
            }
            this.animationState = 'enter';
        });
    }
    beforeAttachPortal() {
        this.setTransformOrigin();
        this.savePreviouslyFocusedElement();
    }
    /** Callback, invoked whenever an animation on the host completes. */
    onAnimationDone(event) {
        if (event.toState === 'void') {
            this.trapFocus();
        }
        else if (event.toState === 'exit') {
            this.restoreFocus();
        }
        this.animationStateChanged.emit(event);
    }
    /** Callback, invoked when an animation on the host starts. */
    onAnimationStart(event) {
        this.animationStateChanged.emit(event);
    }
    ngOnDestroy() {
        super.destroy();
    }
}
ThyDialogContainerComponent.ɵfac = function ThyDialogContainerComponent_Factory(t) { return new (t || ThyDialogContainerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ThyDialogConfig), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyClickPositioner), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FocusTrapFactory)); };
ThyDialogContainerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyDialogContainerComponent, selectors: [["thy-dialog-container"]], viewQuery: function ThyDialogContainerComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(CdkPortalOutlet, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.portalOutlet = _t.first);
    } }, hostAttrs: ["tabindex", "-1", "aria-modal", "true", 1, "thy-dialog-container"], hostVars: 6, hostBindings: function ThyDialogContainerComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵsyntheticHostListener("@dialogContainer.start", function ThyDialogContainerComponent_animation_dialogContainer_start_HostBindingHandler($event) { return ctx.onAnimationStart($event); })("@dialogContainer.done", function ThyDialogContainerComponent_animation_dialogContainer_done_HostBindingHandler($event) { return ctx.onAnimationDone($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("id", ctx.id)("role", ctx.config.role)("aria-labelledby", ctx.config.ariaLabel ? null : ctx.ariaLabelledBy)("aria-label", ctx.config.ariaLabel)("aria-describedby", ctx.config.ariaDescribedBy || null);
        ɵngcc0.ɵɵsyntheticHostProperty("@dialogContainer", ctx.animationState);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 0, consts: [["cdkPortalOutlet", ""]], template: function ThyDialogContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyDialogContainerComponent_ng_template_0_Template, 0, 0, "ng-template", 0);
    } }, directives: [ɵngcc3.CdkPortalOutlet], encapsulation: 2, data: { animation: [thyDialogAnimations.dialogContainer] } });
ThyDialogContainerComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ThyDialogConfig },
    { type: ChangeDetectorRef },
    { type: ThyClickPositioner },
    { type: FocusTrapFactory }
];
ThyDialogContainerComponent.propDecorators = {
    portalOutlet: [{ type: ViewChild, args: [CdkPortalOutlet, { static: true },] }],
    id: [{ type: HostBinding, args: [`attr.id`,] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDialogContainerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-container',
                template: `
        <ng-template cdkPortalOutlet></ng-template>
    `,
                // Using OnPush for dialogs caused some sync issues, e.g. custom ngModel can't to detect changes
                // Disabled until we can track them down.
                changeDetection: ChangeDetectionStrategy.Default,
                animations: [thyDialogAnimations.dialogContainer],
                host: {
                    class: 'thy-dialog-container',
                    tabindex: '-1',
                    'aria-modal': 'true',
                    '[attr.id]': 'id',
                    '[attr.role]': 'config.role',
                    '[attr.aria-labelledby]': 'config.ariaLabel ? null : ariaLabelledBy',
                    '[attr.aria-label]': 'config.ariaLabel',
                    '[attr.aria-describedby]': 'config.ariaDescribedBy || null',
                    '[@dialogContainer]': 'animationState',
                    '(@dialogContainer.start)': 'onAnimationStart($event)',
                    '(@dialogContainer.done)': 'onAnimationDone($event)'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ThyDialogConfig }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.ThyClickPositioner }, { type: ɵngcc2.FocusTrapFactory }]; }, { portalOutlet: [{
            type: ViewChild,
            args: [CdkPortalOutlet, { static: true }]
        }], id: [{
            type: HostBinding,
            args: [`attr.id`]
        }] }); })();

class ThyDialogRef extends ThyAbstractOverlayRef {
}
class ThyInternalDialogRef extends ThyAbstractInternalOverlayRef {
    constructor(overlayRef, containerInstance, config) {
        super(dialogUpperOverlayOptions, overlayRef, containerInstance, config);
    }
    /**
     * Updates the dialog's position.
     * @param position New dialog position.
     */
    updatePosition(position) {
        this.updateGlobalPosition(position);
        return this;
    }
    /**
     * Updates the dialog's width and height.
     * @param width New width of the dialog.
     * @param height New height of the dialog.
     */
    updateSizeAndPosition(width = '', height = '', position) {
        this.getPositionStrategy().width(width).height(height);
        this.updatePosition(position);
        return this;
    }
}

const THY_CONFIRM_DEFAULT_OPTIONS = new InjectionToken('thy-confirm-default-options');
const THY_CONFIRM_DEFAULT_OPTIONS_VALUE = {
    title: '确认删除',
    okText: '确认',
    okType: 'danger',
    cancelText: '取消',
    footerAlign: 'left'
};
const THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER = {
    provide: THY_CONFIRM_DEFAULT_OPTIONS,
    useValue: THY_CONFIRM_DEFAULT_OPTIONS_VALUE
};

class ThyConfirmComponent {
    constructor(dialogRef, changeDetectorRef, defaultConfig) {
        this.dialogRef = dialogRef;
        this.changeDetectorRef = changeDetectorRef;
        this.defaultConfig = defaultConfig;
        this.defaultConfig = Object.assign(Object.assign({}, THY_CONFIRM_DEFAULT_OPTIONS_VALUE), this.defaultConfig);
    }
    ngOnInit() {
        this.defaultConfig = Object.assign(Object.assign({}, this.defaultConfig), this.options);
        this.title = this.defaultConfig.title;
        this.content = this.defaultConfig.content;
        this.okText = this.defaultConfig.okText;
        this.okType = this.defaultConfig.okType;
        this.cancelText = this.defaultConfig.cancelText;
        this.okLoadingText = this.defaultConfig.okLoadingText || this.okText;
        this.footerAlign = this.defaultConfig.footerAlign;
    }
    confirm() {
        this.loading = true;
        const result = this.options.onOk();
        if (result && result.subscribe) {
            result
                .pipe(finalize(() => {
                this.loading = false;
                this.changeDetectorRef.markForCheck();
            }))
                .subscribe(success => {
                if (success) {
                    this.close();
                }
            });
        }
        else {
            this.close();
        }
    }
    close() {
        this.dialogRef.close();
    }
    ngOnDestroy() { }
}
ThyConfirmComponent.ɵfac = function ThyConfirmComponent_Factory(t) { return new (t || ThyConfirmComponent)(ɵngcc0.ɵɵdirectiveInject(ThyDialogRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(THY_CONFIRM_DEFAULT_OPTIONS)); };
ThyConfirmComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyConfirmComponent, selectors: [["thy-confirm-default"]], inputs: { options: "options" }, decls: 11, vars: 9, consts: [[3, "thyTitle", "thyOnClose"], [1, "thy-form"], ["thyForm", "", "name", "thyConfirm", "thyLayout", "vertical"], [1, "form-group"], [3, "innerHTML"], [3, "thyAlign"], [1, "thy-confirm-footer", 3, "ngClass"], [3, "thyButton", "thyLoading", "thyLoadingText", "click"], ["thyButton", "link-secondary", 1, "thy-confirm-cancel", 3, "click"]], template: function ThyConfirmComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "thy-dialog-header", 0);
        ɵngcc0.ɵɵlistener("thyOnClose", function ThyConfirmComponent_Template_thy_dialog_header_thyOnClose_0_listener() { return ctx.close(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(1, "thy-dialog-body", 1);
        ɵngcc0.ɵɵelementStart(2, "form", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelement(4, "p", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "thy-form-group-footer", 5);
        ɵngcc0.ɵɵelementStart(6, "div", 6);
        ɵngcc0.ɵɵelementStart(7, "button", 7);
        ɵngcc0.ɵɵlistener("click", function ThyConfirmComponent_Template_button_click_7_listener() { return ctx.confirm(); });
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "button", 8);
        ɵngcc0.ɵɵlistener("click", function ThyConfirmComponent_Template_button_click_9_listener() { return ctx.close(); });
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("thyTitle", ctx.title || "Install Angular");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("innerHTML", ctx.content, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thyAlign", ctx.footerAlign);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", "thy-confirm-footer-" + ctx.footerAlign);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thyButton", ctx.okType)("thyLoading", ctx.loading)("thyLoadingText", ctx.okLoadingText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.okText);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.cancelText);
    } }, directives: function () { return [DialogHeaderComponent, DialogBodyComponent, ɵngcc4.ɵangular_packages_forms_forms_y, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc5.ThyFormDirective, ɵngcc5.ThyFormGroupFooterComponent, ɵngcc6.NgClass, ɵngcc7.ThyButtonComponent]; }, encapsulation: 2, changeDetection: 0 });
ThyConfirmComponent.ctorParameters = () => [
    { type: ThyDialogRef },
    { type: ChangeDetectorRef },
    { type: undefined, decorators: [{ type: Inject, args: [THY_CONFIRM_DEFAULT_OPTIONS,] }] }
];
ThyConfirmComponent.propDecorators = {
    options: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyConfirmComponent, [{
        type: Component,
        args: [{
                selector: 'thy-confirm-default',
                template: "<thy-dialog-header [thyTitle]=\"title || 'Install Angular'\" (thyOnClose)=\"close()\"> </thy-dialog-header>\n<thy-dialog-body class=\"thy-form\">\n  <form thyForm name=\"thyConfirm\" thyLayout=\"vertical\">\n    <div class=\"form-group\">\n      <p [innerHTML]=\"content\"></p>\n    </div>\n    <thy-form-group-footer [thyAlign]=\"footerAlign\">\n      <div class=\"thy-confirm-footer\" [ngClass]=\"'thy-confirm-footer-' + footerAlign\">\n        <button [thyButton]=\"okType\" (click)=\"confirm()\" [thyLoading]=\"loading\" [thyLoadingText]=\"okLoadingText\">{{ okText }}</button>\n        <button thyButton=\"link-secondary\" (click)=\"close()\" class=\"thy-confirm-cancel\">{{ cancelText }}</button>\n      </div>\n    </thy-form-group-footer>\n  </form>\n</thy-dialog-body>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ThyDialogRef }, { type: ɵngcc0.ChangeDetectorRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_CONFIRM_DEFAULT_OPTIONS]
            }] }]; }, { options: [{
            type: Input
        }] }); })();

class ThyDialog extends ThyAbstractOverlayService {
    constructor(overlay, injector, defaultConfig, clickPositioner) {
        super(dialogUpperOverlayOptions, overlay, injector, defaultConfig);
        clickPositioner.initialize();
    }
    buildOverlayConfig(config) {
        const size = config.size || ThyDialogSizes.md;
        const overlayConfig = this.buildBaseOverlayConfig(config, [`dialog-${size}`]);
        overlayConfig.positionStrategy = this.overlay.position().global();
        overlayConfig.scrollStrategy = config.scrollStrategy || this.overlay.scrollStrategies.block();
        return overlayConfig;
    }
    attachUpperOverlayContainer(overlay, config) {
        const userInjector = config && config.viewContainerRef && config.viewContainerRef.injector;
        const injector = Injector.create({
            parent: userInjector || this.injector,
            providers: [{ provide: ThyDialogConfig, useValue: config }]
        });
        const containerPortal = new ComponentPortal(ThyDialogContainerComponent, config.viewContainerRef, injector);
        const containerRef = overlay.attach(containerPortal);
        return containerRef.instance;
    }
    createUpperOverlayRef(overlayRef, containerInstance, config) {
        return new ThyInternalDialogRef(overlayRef, containerInstance, config);
    }
    createInjector(config, dialogRef, dialogContainer) {
        const userInjector = config && config.viewContainerRef && config.viewContainerRef.injector;
        const injectionTokens = [
            { provide: ThyDialogContainerComponent, useValue: dialogContainer },
            {
                provide: ThyDialogRef,
                useValue: dialogRef
            }
        ];
        if (config.direction && (!userInjector || !userInjector.get(Directionality, null))) {
            injectionTokens.push({
                provide: Directionality,
                useValue: {
                    value: config.direction,
                    change: of()
                }
            });
        }
        return Injector.create({ parent: userInjector || this.injector, providers: injectionTokens });
    }
    open(componentOrTemplateRef, config) {
        const dialogRef = this.openUpperOverlay(componentOrTemplateRef, config);
        const dialogRefInternal = dialogRef;
        dialogRefInternal.updateSizeAndPosition(dialogRef.containerInstance.config.width, dialogRef.containerInstance.config.height, dialogRef.containerInstance.config.position);
        return dialogRef;
    }
    confirm(options) {
        return this.open(ThyConfirmComponent, {
            initialState: {
                options: options
            }
        });
    }
    getDialogById(id) {
        return this.getUpperOverlayById(id);
    }
    /**
     * Finds the closest ThyDialogRef to an element by looking at the DOM.
     */
    getClosestDialog(element) {
        let parent = element.parentElement;
        while (parent && !parent.classList.contains('thy-dialog-container')) {
            parent = parent.parentElement;
        }
        if (parent && parent.id) {
            return this.getDialogById(parent.id);
        }
        return null;
    }
    ngOnDestroy() {
        this.dispose();
    }
}
ThyDialog.ɵfac = function ThyDialog_Factory(t) { return new (t || ThyDialog)(ɵngcc0.ɵɵinject(ɵngcc8.Overlay), ɵngcc0.ɵɵinject(ɵngcc0.Injector), ɵngcc0.ɵɵinject(THY_DIALOG_DEFAULT_OPTIONS, 8), ɵngcc0.ɵɵinject(ɵngcc1.ThyClickPositioner)); };
ThyDialog.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThyDialog_Factory() { return new ThyDialog(i0.ɵɵinject(i1.Overlay), i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(THY_DIALOG_DEFAULT_OPTIONS, 8), i0.ɵɵinject(i3.ThyClickPositioner)); }, token: ThyDialog, providedIn: "root" });
ThyDialog.ctorParameters = () => [
    { type: Overlay },
    { type: Injector },
    { type: ThyDialogConfig, decorators: [{ type: Optional }, { type: Inject, args: [THY_DIALOG_DEFAULT_OPTIONS,] }] },
    { type: ThyClickPositioner }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDialog, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc8.Overlay }, { type: ɵngcc0.Injector }, { type: ThyDialogConfig, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [THY_DIALOG_DEFAULT_OPTIONS]
            }] }, { type: ɵngcc1.ThyClickPositioner }]; }, null); })();

class DialogBodyComponent {
    constructor(dialog) {
        this.dialog = dialog;
        this._isDialogBody = true;
        this.thyClearPaddingClassName = false;
    }
    set thyClearPadding(value) {
        this.thyClearPaddingClassName = coerceBooleanProperty(value);
    }
    ngOnInit() { }
}
DialogBodyComponent.ɵfac = function DialogBodyComponent_Factory(t) { return new (t || DialogBodyComponent)(ɵngcc0.ɵɵdirectiveInject(ThyDialog)); };
DialogBodyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DialogBodyComponent, selectors: [["thy-dialog-body"]], hostVars: 4, hostBindings: function DialogBodyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dialog-body", ctx._isDialogBody)("dialog-body-clear-padding", ctx.thyClearPaddingClassName);
    } }, inputs: { thyClearPadding: "thyClearPadding" }, exportAs: ["thyDialogBody"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function DialogBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
DialogBodyComponent.ctorParameters = () => [
    { type: ThyDialog }
];
DialogBodyComponent.propDecorators = {
    _isDialogBody: [{ type: HostBinding, args: [`class.dialog-body`,] }],
    thyClearPaddingClassName: [{ type: HostBinding, args: [`class.dialog-body-clear-padding`,] }],
    thyClearPadding: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DialogBodyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-body',
                template: '<ng-content></ng-content>',
                // changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'thyDialogBody'
            }]
    }], function () { return [{ type: ThyDialog }]; }, { _isDialogBody: [{
            type: HostBinding,
            args: [`class.dialog-body`]
        }], thyClearPaddingClassName: [{
            type: HostBinding,
            args: [`class.dialog-body-clear-padding`]
        }], thyClearPadding: [{
            type: Input
        }] }); })();

class DialogFooterComponent {
    constructor(dialogLayoutConfig, updateHostClassService, elementRef) {
        this.dialogLayoutConfig = dialogLayoutConfig;
        this.updateHostClassService = updateHostClassService;
        this.elementRef = elementRef;
        this.updateHostClassService.initializeElement(this.elementRef.nativeElement);
        this.divided = this.dialogLayoutConfig.footerDivided;
    }
    set thyDivided(value) {
        this.divided = coerceBooleanProperty(value);
    }
    get align() {
        return !!this.thyAlign ? this.thyAlign : this.dialogLayoutConfig.footerAlign;
    }
    ngOnInit() {
        this.updateHostClassService.updateClassByMap({
            'dialog-footer': true,
            [`dialog-footer-actions-align-${this.align}`]: true,
            'dialog-footer-border-top': this.divided
        });
    }
}
DialogFooterComponent.ɵfac = function DialogFooterComponent_Factory(t) { return new (t || DialogFooterComponent)(ɵngcc0.ɵɵdirectiveInject(THY_DIALOG_LAYOUT_CONFIG), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
DialogFooterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DialogFooterComponent, selectors: [["thy-dialog-footer"]], contentQueries: function DialogFooterComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.description = _t.first);
    } }, inputs: { thyDivided: "thyDivided", thyAlign: "thyAlign" }, exportAs: ["thyDialogFooter"], features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], ngContentSelectors: _c0, decls: 3, vars: 1, consts: [[1, "dialog-footer-actions", "btn-pair"], ["class", "dialog-footer-description", 4, "ngIf"], [1, "dialog-footer-description"], [3, "ngTemplateOutlet"]], template: function DialogFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, DialogFooterComponent_div_2_Template, 2, 1, "div", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.description);
    } }, directives: [ɵngcc6.NgIf, ɵngcc6.NgTemplateOutlet], encapsulation: 2 });
DialogFooterComponent.ctorParameters = () => [
    { type: ThyDialogLayoutConfig, decorators: [{ type: Inject, args: [THY_DIALOG_LAYOUT_CONFIG,] }] },
    { type: UpdateHostClassService },
    { type: ElementRef }
];
DialogFooterComponent.propDecorators = {
    description: [{ type: ContentChild, args: ['description',] }],
    thyDivided: [{ type: Input }],
    thyAlign: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DialogFooterComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-footer',
                template: "<div class=\"dialog-footer-actions btn-pair\">\n  <ng-content></ng-content>\n</div>\n<div class=\"dialog-footer-description\" *ngIf=\"description\">\n  <ng-container [ngTemplateOutlet]=\"description\"></ng-container>\n</div>\n",
                // changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'thyDialogFooter',
                providers: [UpdateHostClassService]
            }]
    }], function () { return [{ type: ThyDialogLayoutConfig, decorators: [{
                type: Inject,
                args: [THY_DIALOG_LAYOUT_CONFIG]
            }] }, { type: ɵngcc1.UpdateHostClassService }, { type: ɵngcc0.ElementRef }]; }, { thyDivided: [{
            type: Input
        }], description: [{
            type: ContentChild,
            args: ['description']
        }], thyAlign: [{
            type: Input
        }] }); })();

class DialogHeaderComponent {
    constructor(elementRef, dialog, translate, dialogContainer) {
        this.elementRef = elementRef;
        this.dialog = dialog;
        this.translate = translate;
        this.dialogContainer = dialogContainer;
        this.isDialogHeader = true;
        this.isDialogHeaderLg = false;
        this.thyOnClose = new EventEmitter();
    }
    set thySize(value) {
        this.isDialogHeaderLg = value === 'lg';
    }
    set thyTitleTranslationKey(key) {
        if (key && !this.thyTitle) {
            this.thyTitle = this.translate.instant(key);
        }
    }
    ngOnInit() {
        if (!this.dialogContainer) {
            // When this directive is included in a dialog via TemplateRef (rather than being
            // in a Component), the ThyDialogContainerComponent isn't available via injection because embedded
            // views cannot be given a custom injector. Instead, we look up the ThyDialogContainerComponent by
            // ID. This must occur in `onInit`, as the ID binding for the dialog container won't
            // be resolved at constructor time.
            const dialogRef = this.dialog.getClosestDialog(this.elementRef.nativeElement);
            this.dialogContainer = dialogRef ? dialogRef.containerInstance : null;
        }
        // change in next microtask avoid throw ExpressionChangedAfterItHasBeenCheckedError
        // because sub component change parent's HostBinding property (ariaLabelledBy)
        Promise.resolve().then(() => {
            if (this.dialogContainer) {
                this.dialogContainer.ariaLabelledBy = this.thyTitle;
            }
        });
    }
    close(event) {
        if (this.thyOnClose.observers.length > 0) {
            this.thyOnClose.emit(event);
        }
        else {
            this.dialog.close();
        }
    }
}
DialogHeaderComponent.ɵfac = function DialogHeaderComponent_Factory(t) { return new (t || DialogHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ThyDialog), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate), ɵngcc0.ɵɵdirectiveInject(ThyDialogContainerComponent, 8)); };
DialogHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DialogHeaderComponent, selectors: [["thy-dialog-header"]], contentQueries: function DialogHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerTemplate = _t.first);
    } }, hostVars: 4, hostBindings: function DialogHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dialog-header", ctx.isDialogHeader)("dialog-header-lg", ctx.isDialogHeaderLg);
    } }, inputs: { thySize: "thySize", thyTitleTranslationKey: "thyTitleTranslationKey", thyTitle: "thyTitle", thyIcon: "thyIcon" }, outputs: { thyOnClose: "thyOnClose" }, exportAs: ["thyDialogHeader"], decls: 3, vars: 2, consts: [[4, "ngIf", "ngIfElse"], ["defaultHeader", ""], [3, "ngTemplateOutlet"], [1, "dialog-title"], [3, "thyIconName"], ["type", "button", 1, "close", 3, "click"], ["thyIconName", "close-bold"]], template: function DialogHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, DialogHeaderComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, DialogHeaderComponent_ng_template_1_Template, 5, 2, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.headerTemplate)("ngIfElse", _r1);
    } }, directives: [ɵngcc6.NgIf, ɵngcc6.NgTemplateOutlet, ɵngcc9.ThyIconComponent], encapsulation: 2 });
DialogHeaderComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyDialog },
    { type: ThyTranslate },
    { type: ThyDialogContainerComponent, decorators: [{ type: Optional }] }
];
DialogHeaderComponent.propDecorators = {
    isDialogHeader: [{ type: HostBinding, args: [`class.dialog-header`,] }],
    isDialogHeaderLg: [{ type: HostBinding, args: [`class.dialog-header-lg`,] }],
    headerTemplate: [{ type: ContentChild, args: ['dialogHeader',] }],
    thyTitle: [{ type: Input }],
    thySize: [{ type: Input }],
    thyTitleTranslationKey: [{ type: Input }],
    thyIcon: [{ type: Input }],
    thyOnClose: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DialogHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-header',
                template: "<ng-container *ngIf=\"headerTemplate; else defaultHeader\">\n  <template [ngTemplateOutlet]=\"headerTemplate\"></template>\n</ng-container>\n<ng-template #defaultHeader>\n  <h3 class=\"dialog-title\"><thy-icon [thyIconName]=\"thyIcon\"></thy-icon>{{ thyTitle }}</h3>\n  <button type=\"button\" class=\"close\" (click)=\"close($event)\">\n    <thy-icon thyIconName=\"close-bold\"></thy-icon>\n  </button>\n</ng-template>\n",
                // changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'thyDialogHeader'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ThyDialog }, { type: ɵngcc1.ThyTranslate }, { type: ThyDialogContainerComponent, decorators: [{
                type: Optional
            }] }]; }, { isDialogHeader: [{
            type: HostBinding,
            args: [`class.dialog-header`]
        }], isDialogHeaderLg: [{
            type: HostBinding,
            args: [`class.dialog-header-lg`]
        }], thyOnClose: [{
            type: Output
        }], thySize: [{
            type: Input
        }], thyTitleTranslationKey: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], headerTemplate: [{
            type: ContentChild,
            args: ['dialogHeader']
        }], thyIcon: [{
            type: Input
        }] }); })();

class ThyDialogModule {
}
ThyDialogModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyDialogModule });
ThyDialogModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyDialogModule_Factory(t) { return new (t || ThyDialogModule)(); }, providers: [ThyDialog, THY_DIALOG_DEFAULT_OPTIONS_PROVIDER, THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER, THY_DIALOG_LAYOUT_CONFIG_PROVIDER], imports: [[CommonModule, ThySharedModule, PortalModule, OverlayModule, ThyButtonModule, ThyIconModule, ThyFormModule, FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyDialogModule, { declarations: function () { return [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent, ThyConfirmComponent]; }, imports: function () { return [CommonModule, ThySharedModule, PortalModule, OverlayModule, ThyButtonModule, ThyIconModule, ThyFormModule, FormsModule]; }, exports: function () { return [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyDialogModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent, ThyConfirmComponent],
                imports: [CommonModule, ThySharedModule, PortalModule, OverlayModule, ThyButtonModule, ThyIconModule, ThyFormModule, FormsModule],
                providers: [ThyDialog, THY_DIALOG_DEFAULT_OPTIONS_PROVIDER, THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER, THY_DIALOG_LAYOUT_CONFIG_PROVIDER],
                entryComponents: [ThyDialogContainerComponent, ThyConfirmComponent],
                exports: [ThyDialogContainerComponent, DialogHeaderComponent, DialogBodyComponent, DialogFooterComponent]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { DialogBodyComponent, DialogFooterComponent, DialogHeaderComponent, THY_CONFIRM_DEFAULT_OPTIONS, THY_CONFIRM_DEFAULT_OPTIONS_PROVIDER, THY_CONFIRM_DEFAULT_OPTIONS_VALUE, THY_DIALOG_DEFAULT_OPTIONS, THY_DIALOG_DEFAULT_OPTIONS_PROVIDER, THY_DIALOG_LAYOUT_CONFIG, THY_DIALOG_LAYOUT_CONFIG_PROVIDER, ThyConfirmComponent as ThyConfirmComponent$1, ThyDialog, ThyDialogConfig, ThyDialogContainerComponent, ThyDialogLayoutConfig, ThyDialogModule, ThyDialogRef, ThyDialogSizes, thyDialogAnimations, ɵ0, ɵ1 };
