// ngx-tethys/fesm2015/ngx-tethys-strength.js
import { Component, forwardRef, HostBinding, Input, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ThyTranslate } from 'ngx-tethys/core';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
var ThyStrengthEnum;
(function (ThyStrengthEnum) {
    ThyStrengthEnum[ThyStrengthEnum["highest"] = 4] = "highest";
    ThyStrengthEnum[ThyStrengthEnum["high"] = 3] = "high";
    ThyStrengthEnum[ThyStrengthEnum["average"] = 2] = "average";
    ThyStrengthEnum[ThyStrengthEnum["low"] = 1] = "low";
})(ThyStrengthEnum || (ThyStrengthEnum = {}));
const strengthMap = {
    [ThyStrengthEnum.highest]: {
        level: 'highest',
        text: '最高'
    },
    [ThyStrengthEnum.high]: {
        level: 'high',
        text: '高'
    },
    [ThyStrengthEnum.average]: {
        level: 'average',
        text: '中'
    },
    [ThyStrengthEnum.low]: {
        level: 'low',
        text: '低'
    }
};
class ThyStrengthComponent {
    constructor(translate) {
        this.translate = translate;
        this.styleClass = true;
        this.strengthMap = JSON.parse(JSON.stringify(strengthMap));
        this._onChange = Function.prototype;
        this._onTouched = Function.prototype;
    }
    set titleKey(value) {
        this.strengthTitle = this.translate.instant(value);
    }
    set highestKey(value) {
        this.strengthMap[ThyStrengthEnum.highest].text = this.translate.instant(value);
    }
    set highKey(value) {
        this.strengthMap[ThyStrengthEnum.high].text = this.translate.instant(value);
    }
    set averageKey(value) {
        this.strengthMap[ThyStrengthEnum.average].text = this.translate.instant(value);
    }
    set lowKey(value) {
        this.strengthMap[ThyStrengthEnum.low].text = this.translate.instant(value);
    }
    ngOnInit() { }
    writeValue(value) {
        this.strength = value;
    }
    registerOnChange(fn) {
        this._onChange = fn;
    }
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
}
ThyStrengthComponent.ɵfac = function ThyStrengthComponent_Factory(t) { return new (t || ThyStrengthComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTranslate)); };
ThyStrengthComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyStrengthComponent, selectors: [["thy-strength"]], hostVars: 2, hostBindings: function ThyStrengthComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("password-strength-container", ctx.styleClass);
    } }, inputs: { titleKey: "titleKey", highestKey: "highestKey", highKey: "highKey", averageKey: "averageKey", lowKey: "lowKey" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => ThyStrengthComponent),
                multi: true
            }
        ])], decls: 6, vars: 8, template: function ThyStrengthComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementStart(2, "span");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div");
        ɵngcc0.ɵɵelement(5, "span");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1("", ctx.strengthTitle, " ");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMapInterpolate1("strength-text strength-text-", ctx.strengthMap[ctx.strength] == null ? null : ctx.strengthMap[ctx.strength].level, "");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.strengthMap[ctx.strength] == null ? null : ctx.strengthMap[ctx.strength].text);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMapInterpolate1("strength-level strength-level-", ctx.strengthMap[ctx.strength] == null ? null : ctx.strengthMap[ctx.strength].level, "");
    } }, encapsulation: 2 });
ThyStrengthComponent.ctorParameters = () => [
    { type: ThyTranslate }
];
ThyStrengthComponent.propDecorators = {
    styleClass: [{ type: HostBinding, args: ['class.password-strength-container',] }],
    titleKey: [{ type: Input }],
    highestKey: [{ type: Input }],
    highKey: [{ type: Input }],
    averageKey: [{ type: Input }],
    lowKey: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStrengthComponent, [{
        type: Component,
        args: [{
                selector: 'thy-strength',
                template: "<div>{{strengthTitle}} <span class=\"strength-text strength-text-{{strengthMap[strength]?.level}}\">{{strengthMap[strength]?.text}}</span></div>\n<div class=\"strength-level strength-level-{{strengthMap[strength]?.level}}\">\n    <span></span>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => ThyStrengthComponent),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.ThyTranslate }]; }, { styleClass: [{
            type: HostBinding,
            args: ['class.password-strength-container']
        }], titleKey: [{
            type: Input
        }], highestKey: [{
            type: Input
        }], highKey: [{
            type: Input
        }], averageKey: [{
            type: Input
        }], lowKey: [{
            type: Input
        }] }); })();

class ThyStrengthModule {
}
ThyStrengthModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyStrengthModule });
ThyStrengthModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyStrengthModule_Factory(t) { return new (t || ThyStrengthModule)(); }, providers: [], imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyStrengthModule, { declarations: [ThyStrengthComponent], exports: [ThyStrengthComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStrengthModule, [{
        type: NgModule,
        args: [{
                imports: [],
                exports: [ThyStrengthComponent],
                declarations: [ThyStrengthComponent],
                providers: []
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyStrengthComponent, ThyStrengthModule };
