// ngx-tethys/fesm2015/ngx-tethys-notify.js
import { Injectable, Component, NgZone, HostBinding, Input, HostListener, InjectionToken, Injector, ComponentFactoryResolver, ApplicationRef, Inject, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { UpdateHostClassService } from 'ngx-tethys/core';
import { __decorate, __metadata } from 'tslib';
import { Store, Action } from 'ngx-tethys/store';
import { helpers } from 'ngx-tethys/util';
import { Subject } from 'rxjs';
import { DomPortalOutlet, ComponentPortal } from '@angular/cdk/portal';
import { ThyIconModule } from 'ngx-tethys/icon';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/icon';

function ThyNotifyComponent_ng_container_0_div_8_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 11);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_div_8_a_2_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(3); return ctx_r5.triggerDetail(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r4.option == null ? null : ctx_r4.option.detail["link"]);
} }
const _c0 = function (a0) { return { "thy-notify-content--extend": a0 }; };
function ThyNotifyComponent_ng_container_0_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_div_8_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.extendContent(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, ThyNotifyComponent_ng_container_0_div_8_a_2_Template, 2, 1, "a", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx_r2.extendContentClass === true));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.option == null ? null : ctx_r2.option.content, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.option == null ? null : ctx_r2.option.detail);
} }
function ThyNotifyComponent_ng_container_0_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.option == null ? null : ctx_r3.option.detail["content"]);
} }
function ThyNotifyComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_0_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.closeNotify(); });
    ɵngcc0.ɵɵelement(2, "thy-icon", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 3);
    ɵngcc0.ɵɵelement(4, "thy-icon", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 5);
    ɵngcc0.ɵɵelementStart(6, "div", 6);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(8, ThyNotifyComponent_ng_container_0_div_8_Template, 3, 5, "div", 7);
    ɵngcc0.ɵɵtemplate(9, ThyNotifyComponent_ng_container_0_div_9_Template, 2, 1, "div", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.notifyIconName);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.option == null ? null : ctx_r0.option.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.option == null ? null : ctx_r0.option.detail) || (ctx_r0.option == null ? null : ctx_r0.option.content));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isShowDetail);
} }
function ThyNotifyComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵlistener("click", function ThyNotifyComponent_ng_container_1_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.closeNotify(); });
    ɵngcc0.ɵɵelement(2, "thy-icon", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵelement(4, "template", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.option == null ? null : ctx_r1.option.html);
} }
function ThyNotifyContainerComponent_thy_notify_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r4 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r4);
} }
function ThyNotifyContainerComponent_thy_notify_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r5);
} }
function ThyNotifyContainerComponent_thy_notify_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r6 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r6);
} }
function ThyNotifyContainerComponent_thy_notify_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-notify", 1);
} if (rf & 2) {
    const item_r7 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("thyOption", item_r7);
} }
const notifyQueueInitialState = {
    topLeftQueue: [],
    topRightQueue: [],
    bottomLeftQueue: [],
    bottomRightQueue: []
};
class NotifyQueueStore extends Store {
    constructor() {
        super(notifyQueueInitialState);
    }
    static topRightSelector(state) {
        return state.topRightQueue;
    }
    static topLeftSelector(state) {
        return state.topLeftQueue;
    }
    static bottomLeftSelector(state) {
        return state.bottomLeftQueue;
    }
    static bottomRightSelector(state) {
        return state.bottomRightQueue;
    }
    convertQueueKey(placement) {
        let key;
        if (placement === 'topLeft') {
            key = 'topLeftQueue';
        }
        else if (placement === 'topRight') {
            key = 'topRightQueue';
        }
        else if (placement === 'bottomLeft') {
            key = 'bottomLeftQueue';
        }
        else if (placement === 'bottomRight') {
            key = 'bottomRightQueue';
        }
        return key;
    }
    addNotify(placement, options) {
        const key = this.convertQueueKey(placement);
        const state = this.snapshot;
        if (state[key].length > options.maxStack) {
            state[key].shift();
        }
        state[key].push(options);
        this.next(state);
    }
    removeNotify(id, placement) {
        const state = this.snapshot;
        if (placement) {
            const queueKey = this.convertQueueKey(placement);
            if (state.hasOwnProperty(queueKey) && state[queueKey].length) {
                state[queueKey] = state[queueKey].filter((item) => {
                    return item.id !== id;
                });
            }
        }
        else {
            Object.keys(state).forEach(queueKey => {
                if (state.hasOwnProperty(queueKey) && state[queueKey].length) {
                    state[queueKey] = state[queueKey].filter((item) => {
                        return item.id !== id;
                    });
                }
            });
        }
        this.next(state);
    }
}
NotifyQueueStore.ɵfac = function NotifyQueueStore_Factory(t) { return new (t || NotifyQueueStore)(); };
NotifyQueueStore.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NotifyQueueStore, factory: NotifyQueueStore.ɵfac });
NotifyQueueStore.ctorParameters = () => [];
__decorate([
    Action(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", void 0)
], NotifyQueueStore.prototype, "addNotify", null);
__decorate([
    Action(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Number, String]),
    __metadata("design:returntype", void 0)
], NotifyQueueStore.prototype, "removeNotify", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotifyQueueStore, [{
        type: Injectable
    }], function () { return []; }, null); })();

const ANIMATION_IN_DURATION = 100;
const ANIMATION_OUT_DURATION = 150;
const HIDE_STYLE = { transform: 'translateX(0)', opacity: 0, height: 0, paddingTop: 0, paddingBottom: 0, margin: 0 };
class ThyNotifyComponent {
    constructor(_queueStore, _ngZone) {
        this._queueStore = _queueStore;
        this._ngZone = _ngZone;
        this.className = '';
        this.notifyIconName = '';
        this.extendContentClass = false;
        this.isShowDetail = false;
    }
    set thyOption(value) {
        this.option = value;
        const type = value.type;
        this.placement = value.placement || 'topRight';
        if (this.placement === 'topLeft' || this.placement === 'bottomLeft') {
            this.flyInOut = 'flyInOutLeft';
        }
        else {
            this.flyInOut = 'flyInOutRight';
        }
        this.className = `thy-notify thy-notify-${type}`;
    }
    ngOnInit() {
        const iconName = {
            success: 'check-circle-fill',
            info: 'info-circle-fill',
            warning: 'waring-fill',
            error: 'close-circle-fill'
        };
        this.notifyIconName = iconName[this.option.type];
        this._creatCloseTimer();
    }
    ngOnDestroy() {
        this._clearCloseTimer();
    }
    extendContent() {
        this.extendContentClass = true;
    }
    showDetailToggle() {
        this.isShowDetail = !this.isShowDetail;
    }
    closeNotify() {
        this._ngZone.runOutsideAngular(() => {
            this.flyInOut = 'componentHide';
            setTimeout(() => {
                this._queueStore.removeNotify(this.option.id, this.placement);
            }, ANIMATION_OUT_DURATION);
        });
    }
    triggerDetail() {
        if (helpers.isFunction(this.option.detail.action)) {
            this.option.detail.action();
        }
        if (this.option.detail.content) {
            this.showDetailToggle();
        }
    }
    mouseenter() {
        if (this.option.pauseOnHover) {
            this._clearCloseTimer();
        }
    }
    mouseleave() {
        if (this.option.pauseOnHover) {
            this._creatCloseTimer();
        }
    }
    _creatCloseTimer() {
        if (this.option.duration) {
            this.closeTimer = setInterval(() => {
                clearInterval(this.closeTimer);
                this.closeNotify();
            }, this.option.duration);
        }
    }
    _clearCloseTimer() {
        clearInterval(this.closeTimer);
    }
}
ThyNotifyComponent.ɵfac = function ThyNotifyComponent_Factory(t) { return new (t || ThyNotifyComponent)(ɵngcc0.ɵɵdirectiveInject(NotifyQueueStore), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyNotifyComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNotifyComponent, selectors: [["thy-notify"]], hostVars: 3, hostBindings: function ThyNotifyComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mouseenter", function ThyNotifyComponent_mouseenter_HostBindingHandler() { return ctx.mouseenter(); })("mouseleave", function ThyNotifyComponent_mouseleave_HostBindingHandler() { return ctx.mouseleave(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵsyntheticHostProperty("@flyInOut", ctx.flyInOut);
        ɵngcc0.ɵɵclassMap(ctx.className);
    } }, inputs: { thyOption: "thyOption" }, features: [ɵngcc0.ɵɵProvidersFeature([UpdateHostClassService])], decls: 2, vars: 2, consts: [[4, "ngIf"], ["href", "javascript:;", 1, "thy-notify-close", 3, "click"], ["thyIconName", "close"], [1, "thy-notify-icon-container"], [3, "thyIconName"], [1, "thy-notify-main"], [1, "thy-notify-title"], ["class", "thy-notify-content", 3, "ngClass", "click", 4, "ngIf"], ["class", "thy-notify-detail", 4, "ngIf"], [1, "thy-notify-content", 3, "ngClass", "click"], ["href", "javascript:;", "class", "link-secondary", 3, "click", 4, "ngIf"], ["href", "javascript:;", 1, "link-secondary", 3, "click"], [1, "thy-notify-detail"], [3, "ngTemplateOutlet"]], template: function ThyNotifyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyNotifyComponent_ng_container_0_Template, 10, 4, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, ThyNotifyComponent_ng_container_1_Template, 5, 1, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.option.html);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.option == null ? null : ctx.option.html);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass, ɵngcc1.NgTemplateOutlet], encapsulation: 2, data: { animation: [
            trigger('flyInOut', [
                state('flyInOutRight', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                transition('void => flyInOutRight', [
                    style({ transform: 'translateX(100%)', opacity: 0, height: '*' }),
                    animate(ANIMATION_IN_DURATION)
                ]),
                transition('flyInOutRight => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                transition('flyInOutRight => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                state('flyInOutLeft', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                transition('void => flyInOutLeft', [
                    style({ transform: 'translateX(-100%)', opacity: 0, height: '*' }),
                    animate(ANIMATION_IN_DURATION)
                ]),
                transition('flyInOutLeft => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                transition('flyInOutLeft => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                state('componentHide', style(HIDE_STYLE))
            ])
        ] } });
ThyNotifyComponent.ctorParameters = () => [
    { type: NotifyQueueStore },
    { type: NgZone }
];
ThyNotifyComponent.propDecorators = {
    flyInOut: [{ type: HostBinding, args: ['@flyInOut',] }],
    className: [{ type: HostBinding, args: ['class',] }],
    thyOption: [{ type: Input }],
    mouseenter: [{ type: HostListener, args: ['mouseenter',] }],
    mouseleave: [{ type: HostListener, args: ['mouseleave',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyComponent, [{
        type: Component,
        args: [{
                selector: 'thy-notify',
                template: "<ng-container *ngIf=\"!option.html\">\n  <a href=\"javascript:;\" class=\"thy-notify-close\" (click)=\"closeNotify()\">\n    <thy-icon thyIconName=\"close\"></thy-icon>\n  </a>\n  <div class=\"thy-notify-icon-container\">\n    <thy-icon [thyIconName]=\"notifyIconName\"></thy-icon>\n  </div>\n  <div class=\"thy-notify-main\">\n    <div class=\"thy-notify-title\">{{ option?.title }}</div>\n    <div\n      *ngIf=\"option?.detail || option?.content\"\n      class=\"thy-notify-content\"\n      [ngClass]=\"{ 'thy-notify-content--extend': extendContentClass === true }\"\n      (click)=\"extendContent()\"\n    >\n      {{ option?.content }}\n      <a href=\"javascript:;\" *ngIf=\"option?.detail\" class=\"link-secondary\" (click)=\"triggerDetail()\">{{ (option?.detail)['link'] }}</a>\n    </div>\n    <div class=\"thy-notify-detail\" *ngIf=\"isShowDetail\">{{ (option?.detail)['content'] }}</div>\n  </div>\n</ng-container>\n<ng-container *ngIf=\"option?.html\">\n  <a href=\"javascript:;\" class=\"thy-notify-close\" (click)=\"closeNotify()\">\n    <thy-icon thyIconName=\"close\"></thy-icon>\n  </a>\n  <div class=\"thy-notify-main\">\n    <template [ngTemplateOutlet]=\"option?.html\"></template>\n  </div>\n</ng-container>\n",
                providers: [UpdateHostClassService],
                animations: [
                    trigger('flyInOut', [
                        state('flyInOutRight', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                        transition('void => flyInOutRight', [
                            style({ transform: 'translateX(100%)', opacity: 0, height: '*' }),
                            animate(ANIMATION_IN_DURATION)
                        ]),
                        transition('flyInOutRight => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                        transition('flyInOutRight => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                        state('flyInOutLeft', style({ transform: 'translateX(0)', opacity: 1, height: '*' })),
                        transition('void => flyInOutLeft', [
                            style({ transform: 'translateX(-100%)', opacity: 0, height: '*' }),
                            animate(ANIMATION_IN_DURATION)
                        ]),
                        transition('flyInOutLeft => componentHide', [animate(ANIMATION_OUT_DURATION, style(HIDE_STYLE))]),
                        transition('flyInOutLeft => void', [animate(ANIMATION_IN_DURATION, style(HIDE_STYLE))]),
                        state('componentHide', style(HIDE_STYLE))
                    ])
                ]
            }]
    }], function () { return [{ type: NotifyQueueStore }, { type: ɵngcc0.NgZone }]; }, { className: [{
            type: HostBinding,
            args: ['class']
        }], thyOption: [{
            type: Input
        }], flyInOut: [{
            type: HostBinding,
            args: ['@flyInOut']
        }], mouseenter: [{
            type: HostListener,
            args: ['mouseenter']
        }], mouseleave: [{
            type: HostListener,
            args: ['mouseleave']
        }] }); })();

class ThyNotifyContainerComponent {
    constructor(queueStore) {
        this.queueStore = queueStore;
        this.className = true;
    }
    ngOnInit() {
        this.placement = this.initialState.placement;
        let queue$, queueKey;
        if (this.placement === 'bottomRight') {
            this.bottomRight = true;
        }
        else if (this.placement === 'bottomLeft') {
            this.bottomLeft = true;
        }
        else if (this.placement === 'topLeft') {
            this.topLeft = true;
        }
        else {
            this.topRight = true;
        }
        if (this.placement === 'topLeft') {
            queueKey = 'topLeftQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.topLeftSelector);
        }
        else if (this.placement === 'topRight') {
            queueKey = 'topRightQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.topRightSelector);
        }
        else if (this.placement === 'bottomLeft') {
            queueKey = 'bottomLeftQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.bottomLeftSelector);
        }
        else if (this.placement === 'bottomRight') {
            queueKey = 'bottomRightQueue';
            queue$ = this.queueStore.select(NotifyQueueStore.bottomRightSelector);
        }
        queue$.subscribe(data => {
            this[queueKey] = data;
        });
    }
}
ThyNotifyContainerComponent.ɵfac = function ThyNotifyContainerComponent_Factory(t) { return new (t || ThyNotifyContainerComponent)(ɵngcc0.ɵɵdirectiveInject(NotifyQueueStore)); };
ThyNotifyContainerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyNotifyContainerComponent, selectors: [["thy-notify-container"]], hostVars: 10, hostBindings: function ThyNotifyContainerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-notify-root", ctx.className)("thy-notify-bottomRight", ctx.bottomRight)("thy-notify-bottomLeft", ctx.bottomLeft)("thy-notify-topLeft", ctx.topLeft)("thy-notify-topRight", ctx.topRight);
    } }, decls: 4, vars: 4, consts: [[3, "thyOption", 4, "ngFor", "ngForOf"], [3, "thyOption"]], template: function ThyNotifyContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyNotifyContainerComponent_thy_notify_0_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(1, ThyNotifyContainerComponent_thy_notify_1_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(2, ThyNotifyContainerComponent_thy_notify_2_Template, 1, 1, "thy-notify", 0);
        ɵngcc0.ɵɵtemplate(3, ThyNotifyContainerComponent_thy_notify_3_Template, 1, 1, "thy-notify", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.topLeftQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.topRightQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bottomLeftQueue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bottomRightQueue);
    } }, directives: [ɵngcc1.NgForOf, ThyNotifyComponent], encapsulation: 2 });
ThyNotifyContainerComponent.ctorParameters = () => [
    { type: NotifyQueueStore }
];
ThyNotifyContainerComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class.thy-notify-root',] }],
    bottomRight: [{ type: HostBinding, args: ['class.thy-notify-bottomRight',] }],
    bottomLeft: [{ type: HostBinding, args: ['class.thy-notify-bottomLeft',] }],
    topLeft: [{ type: HostBinding, args: ['class.thy-notify-topLeft',] }],
    topRight: [{ type: HostBinding, args: ['class.thy-notify-topRight',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyContainerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-notify-container',
                template: "<thy-notify *ngFor=\"let item of topLeftQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of topRightQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of bottomLeftQueue\" [thyOption]=\"item\"></thy-notify>\n<thy-notify *ngFor=\"let item of bottomRightQueue\" [thyOption]=\"item\"></thy-notify>\n"
            }]
    }], function () { return [{ type: NotifyQueueStore }]; }, { className: [{
            type: HostBinding,
            args: ['class.thy-notify-root']
        }], bottomRight: [{
            type: HostBinding,
            args: ['class.thy-notify-bottomRight']
        }], bottomLeft: [{
            type: HostBinding,
            args: ['class.thy-notify-bottomLeft']
        }], topLeft: [{
            type: HostBinding,
            args: ['class.thy-notify-topLeft']
        }], topRight: [{
            type: HostBinding,
            args: ['class.thy-notify-topRight']
        }] }); })();

const CONTAINER_PLACEMENT = new InjectionToken('CONTAINER_PLACEMENT');
const THY_NOTIFY_DEFAULT_OPTIONS = new InjectionToken('thy-notify-default-options');
const ɵ0 = {};
const THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER = {
    provide: THY_NOTIFY_DEFAULT_OPTIONS,
    useValue: ɵ0
};

const NOTIFY_OPTION_DEFAULT = {
    duration: 4500,
    pauseOnHover: true,
    maxStack: 8,
    placement: 'topRight'
};
class ThyNotifyService {
    constructor(injector, componentFactoryResolver, appRef, queueStore, defaultConfig) {
        this.injector = injector;
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.queueStore = queueStore;
        this.defaultConfig = defaultConfig;
        this.notifyQueue$ = new Subject();
        this._lastNotifyId = 0;
    }
    show(options) {
        const notifyConfig = this.formatOptions(options);
        const { placement } = notifyConfig;
        this.queueStore.addNotify(placement, notifyConfig);
        this._initContainer(placement);
    }
    success(title, content, options) {
        this.show(Object.assign(Object.assign({}, (options || {})), { type: 'success', title: title || (options === null || options === void 0 ? void 0 : options.title) || '成功', content: content || (options === null || options === void 0 ? void 0 : options.content) }));
    }
    info(title, content, options) {
        this.show(Object.assign(Object.assign({}, (options || {})), { type: 'info', title: title || (options === null || options === void 0 ? void 0 : options.title) || '提示', content: content || (options === null || options === void 0 ? void 0 : options.content) }));
    }
    warning(title, content, options) {
        this.show(Object.assign(Object.assign({}, (options || {})), { type: 'warning', title: title || (options === null || options === void 0 ? void 0 : options.title) || '警告', content: content || (options === null || options === void 0 ? void 0 : options.content) }));
    }
    error(title, content, options) {
        var _a, _b;
        const config = helpers.isString(options)
            ? { type: 'error', title: title || '错误', content: content, detail: options }
            : Object.assign(Object.assign({}, (options || {})), { type: 'error', title: title || ((_a = options) === null || _a === void 0 ? void 0 : _a.title) || '错误', content: content || ((_b = options) === null || _b === void 0 ? void 0 : _b.content) });
        this.show(config);
    }
    removeNotifyById(id) {
        this.queueStore.removeNotify(id);
    }
    _initContainer(placement) {
        if (placement === 'topRight') {
            this.containerRefTopRight = this._loadNotifyContainerComponent(this.containerRefTopRight, placement);
        }
        else if (placement === 'bottomRight') {
            this.containerRefBottomRight = this._loadNotifyContainerComponent(this.containerRefBottomRight, placement);
        }
        else if (placement === 'bottomLeft') {
            this.containerRefBottomLeft = this._loadNotifyContainerComponent(this.containerRefBottomLeft, placement);
        }
        else if (placement === 'topLeft') {
            this.containerRefTopLeft = this._loadNotifyContainerComponent(this.containerRefTopLeft, placement);
        }
    }
    _loadNotifyContainerComponent(containerRef, placement) {
        if (!containerRef) {
            const portalOutlet = new DomPortalOutlet(document.body, this.componentFactoryResolver, this.appRef, this.injector);
            const componentPortal = new ComponentPortal(ThyNotifyContainerComponent, null);
            containerRef = portalOutlet.attachComponentPortal(componentPortal);
            Object.assign(containerRef.instance, {
                initialState: {
                    placement
                }
            });
            containerRef.changeDetectorRef.detectChanges();
        }
        return containerRef;
    }
    formatOptions(options) {
        if (helpers.isString(options.detail)) {
            options = Object.assign(Object.assign({}, options), { detail: { link: '[详情]', content: options.detail } });
        }
        return Object.assign({}, NOTIFY_OPTION_DEFAULT, { id: this._lastNotifyId++ }, this.defaultConfig, options);
    }
}
ThyNotifyService.ɵfac = function ThyNotifyService_Factory(t) { return new (t || ThyNotifyService)(ɵngcc0.ɵɵinject(ɵngcc0.Injector), ɵngcc0.ɵɵinject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵinject(ɵngcc0.ApplicationRef), ɵngcc0.ɵɵinject(NotifyQueueStore), ɵngcc0.ɵɵinject(THY_NOTIFY_DEFAULT_OPTIONS)); };
ThyNotifyService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ThyNotifyService, factory: ThyNotifyService.ɵfac });
ThyNotifyService.ctorParameters = () => [
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: ApplicationRef },
    { type: NotifyQueueStore },
    { type: undefined, decorators: [{ type: Inject, args: [THY_NOTIFY_DEFAULT_OPTIONS,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.Injector }, { type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc0.ApplicationRef }, { type: NotifyQueueStore }, { type: undefined, decorators: [{
                type: Inject,
                args: [THY_NOTIFY_DEFAULT_OPTIONS]
            }] }]; }, null); })();

class ThyNotifyModule {
}
ThyNotifyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyNotifyModule });
ThyNotifyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyNotifyModule_Factory(t) { return new (t || ThyNotifyModule)(); }, providers: [ThyNotifyService, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER], imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyNotifyModule, { declarations: function () { return [ThyNotifyComponent, ThyNotifyContainerComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyNotifyContainerComponent, ThyNotifyComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNotifyModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyNotifyComponent, ThyNotifyContainerComponent],
                entryComponents: [ThyNotifyContainerComponent],
                providers: [ThyNotifyService, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyNotifyContainerComponent, ThyNotifyComponent]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { CONTAINER_PLACEMENT, NotifyQueueStore, THY_NOTIFY_DEFAULT_OPTIONS, THY_NOTIFY_DEFAULT_OPTIONS_PROVIDER, ThyNotifyComponent, ThyNotifyContainerComponent, ThyNotifyModule, ThyNotifyService, notifyQueueInitialState, ɵ0 };
