// ngx-tethys/fesm2015/ngx-tethys-back-top.js
import { EventEmitter, Component, ChangeDetectionStrategy, ViewEncapsulation, Inject, ChangeDetectorRef, NgZone, HostBinding, Input, Output, NgModule } from '@angular/core';
import { DOCUMENT, CommonModule } from '@angular/common';
import { ThyIconModule } from 'ngx-tethys/icon';
import { Subject, fromEvent } from 'rxjs';
import { Platform } from '@angular/cdk/platform';
import { throttleTime, takeUntil } from 'rxjs/operators';
import { fadeMotion, ThyScrollService } from 'ngx-tethys/core';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/core';
import * as ɵngcc2 from '@angular/cdk/platform';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/icon';

function ThyBackTopComponent_div_0_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelement(1, "thy-icon", 5);
    ɵngcc0.ɵɵelementEnd();
} }
function ThyBackTopComponent_div_0_ng_template_3_Template(rf, ctx) { }
function ThyBackTopComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵlistener("click", function ThyBackTopComponent_div_0_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.clickBackTop(); });
    ɵngcc0.ɵɵtemplate(1, ThyBackTopComponent_div_0_ng_template_1_Template, 2, 0, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(3, ThyBackTopComponent_div_0_ng_template_3_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r1 = ɵngcc0.ɵɵreference(2);
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@fadeMotion", undefined);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.thyTemplate || _r1);
} }
class ThyBackTopComponent {
    constructor(doc, thyScrollService, platform, cdr, zone) {
        this.doc = doc;
        this.thyScrollService = thyScrollService;
        this.platform = platform;
        this.cdr = cdr;
        this.zone = zone;
        this.classNames = true;
        this.thyVisibilityHeight = 400;
        this.thyClick = new EventEmitter();
        this.visibleChange = new EventEmitter();
        this.visible = false;
        this.scrollListenerDestroy$ = new Subject();
        this.target = null;
    }
    ngOnInit() {
        this.registerScrollEvent();
    }
    clickBackTop() {
        this.thyScrollService.scrollTo(this.getTarget(), 0);
        this.thyClick.emit(true);
    }
    getTarget() {
        return this.target || window;
    }
    handleScroll() {
        if (this.visible === this.thyScrollService.getScroll(this.getTarget()) > this.thyVisibilityHeight) {
            return;
        }
        this.visible = !this.visible;
        this.cdr.detectChanges();
        this.zone.run(() => {
            this.visibleChange.emit(this.visible);
        });
    }
    registerScrollEvent() {
        if (!this.platform.isBrowser) {
            return;
        }
        this.scrollListenerDestroy$.next();
        this.handleScroll();
        this.zone.runOutsideAngular(() => {
            fromEvent(this.getTarget(), 'scroll')
                .pipe(throttleTime(50), takeUntil(this.scrollListenerDestroy$))
                .subscribe(() => this.handleScroll());
        });
    }
    ngOnDestroy() {
        this.scrollListenerDestroy$.next();
        this.scrollListenerDestroy$.complete();
    }
    ngOnChanges(changes) {
        const { thyContainer } = changes;
        if (thyContainer) {
            this.target = typeof this.thyContainer === 'string' ? this.doc.querySelector(this.thyContainer) : this.thyContainer;
            this.registerScrollEvent();
        }
    }
}
ThyBackTopComponent.ɵfac = function ThyBackTopComponent_Factory(t) { return new (t || ThyBackTopComponent)(ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyScrollService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThyBackTopComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyBackTopComponent, selectors: [["thy-back-top"], ["", "thyBackTop", ""]], hostVars: 2, hostBindings: function ThyBackTopComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-back-top-container", ctx.classNames);
    } }, inputs: { thyVisibilityHeight: "thyVisibilityHeight", thyTemplate: "thyTemplate", thyContainer: "thyContainer" }, outputs: { thyClick: "thyClick", visibleChange: "visibleChange" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [["class", "thy-back-top", 3, "click", 4, "ngIf"], [1, "thy-back-top", 3, "click"], ["defaultContent", ""], [3, "ngTemplateOutlet"], [1, "thy-back-top-content"], ["thyIconName", "arrow-up", 1, "back-top-icon", "text-muted", "font-size-lg"]], template: function ThyBackTopComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ThyBackTopComponent_div_0_Template, 4, 2, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.visible);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgTemplateOutlet, ɵngcc4.ThyIconComponent], encapsulation: 2, data: { animation: [fadeMotion] }, changeDetection: 0 });
ThyBackTopComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ThyScrollService },
    { type: Platform },
    { type: ChangeDetectorRef },
    { type: NgZone }
];
ThyBackTopComponent.propDecorators = {
    classNames: [{ type: HostBinding, args: ['class.thy-back-top-container',] }],
    thyTemplate: [{ type: Input }],
    thyVisibilityHeight: [{ type: Input }],
    thyContainer: [{ type: Input }],
    thyClick: [{ type: Output }],
    visibleChange: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBackTopComponent, [{
        type: Component,
        args: [{
                selector: 'thy-back-top,[thyBackTop]',
                template: "<div class=\"thy-back-top\" (click)=\"clickBackTop()\" @fadeMotion *ngIf=\"visible\">\n  <ng-template #defaultContent>\n    <div class=\"thy-back-top-content\">\n      <thy-icon class=\"back-top-icon text-muted font-size-lg\" thyIconName=\"arrow-up\"></thy-icon>\n    </div>\n  </ng-template>\n  <ng-template [ngTemplateOutlet]=\"thyTemplate || defaultContent\"></ng-template>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                animations: [fadeMotion]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc1.ThyScrollService }, { type: ɵngcc2.Platform }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.NgZone }]; }, { classNames: [{
            type: HostBinding,
            args: ['class.thy-back-top-container']
        }], thyVisibilityHeight: [{
            type: Input
        }], thyClick: [{
            type: Output
        }], visibleChange: [{
            type: Output
        }], thyTemplate: [{
            type: Input
        }], thyContainer: [{
            type: Input
        }] }); })();

class ThyBackTopModule {
}
ThyBackTopModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyBackTopModule });
ThyBackTopModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyBackTopModule_Factory(t) { return new (t || ThyBackTopModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyBackTopModule, { declarations: function () { return [ThyBackTopComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyBackTopComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBackTopModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyBackTopComponent],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyBackTopComponent]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyBackTopComponent, ThyBackTopModule };
