// ngx-tethys/fesm2015/ngx-tethys-table.js
import { mixinUnsubscribe, MixinBase, UpdateHostClassService } from 'ngx-tethys/core';
import { coerceBooleanProperty, get, isString, set, keyBy } from 'ngx-tethys/util';
import { merge, of } from 'rxjs';
import { delay, takeUntil } from 'rxjs/operators';
import { moveItemInArray, DragDropModule } from '@angular/cdk/drag-drop';
import { ViewportRuler } from '@angular/cdk/overlay';
import { DOCUMENT, CommonModule } from '@angular/common';
import { InjectionToken, Component, ViewEncapsulation, ElementRef, Optional, Inject, Input, ContentChild, TemplateRef, EventEmitter, IterableDiffers, ViewChild, HostBinding, Output, ContentChildren, Pipe, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ThyPaginationModule } from 'ngx-tethys/pagination';
import { ThySwitchModule } from 'ngx-tethys/switch';
import { ThyLoadingModule } from 'ngx-tethys/loading';
import { ThyEmptyModule } from 'ngx-tethys/empty';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';

/**
 * Injection token used to provide the parent component to options.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/cdk/overlay';
import * as ɵngcc2 from 'ngx-tethys/core';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-tethys/loading';
import * as ɵngcc5 from 'ngx-tethys/shared';
import * as ɵngcc6 from '@angular/cdk/drag-drop';
import * as ɵngcc7 from 'ngx-tethys/icon';
import * as ɵngcc8 from '@angular/forms';
import * as ɵngcc9 from 'ngx-tethys/switch';
import * as ɵngcc10 from 'ngx-tethys/empty';
import * as ɵngcc11 from 'ngx-tethys/pagination';

const _c0 = ["cell"];
const _c1 = ["header"];
const _c2 = ["*"];
const _c3 = ["empty"];
const _c4 = ["group"];
const _c5 = ["table"];
function ThyTableComponent_col_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "col", 8);
} if (rf & 2) {
    const column_r7 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("width", column_r7.width);
} }
function ThyTableComponent_thead_4_th_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", column_r9.title, " ");
} }
function ThyTableComponent_thead_4_th_2_ng_container_2_ng_template_1_Template(rf, ctx) { }
const _c6 = function (a0) { return { $implicit: a0 }; };
function ThyTableComponent_thead_4_th_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_thead_4_th_2_ng_container_2_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", column_r9.headerTemplateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c6, column_r9));
} }
function ThyTableComponent_thead_4_th_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 10);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_thead_4_th_2_ng_container_1_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_thead_4_th_2_ng_container_2_Template, 2, 4, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r9 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngClass", column_r9.headerClassName);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !column_r9.headerTemplateRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r9.headerTemplateRef);
} }
function ThyTableComponent_thead_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "thead");
    ɵngcc0.ɵɵelementStart(1, "tr");
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_thead_4_th_2_Template, 3, 3, "th", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.columns);
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_ng_template_1_Template(rf, ctx) { }
const _c7 = function (a0, a1) { return { $implicit: a0, group: a1 }; };
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const group_r23 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r24 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r24.groupTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c7, group_r23.origin, group_r23.origin));
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_ng_template_1_Template(rf, ctx) { }
const _c8 = function (a0, a1) { return { row: a0, index: a1, level: 0 }; };
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const row_r29 = ctx.$implicit;
    const i_r30 = ctx.index;
    ɵngcc0.ɵɵnextContext(4);
    const _r18 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r18)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c8, row_r29, i_r30));
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_ng_container_1_Template, 2, 5, "ng-container", 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const group_r23 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r25 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", group_r23.children)("ngForTrackBy", ctx_r25.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "tr", 16);
    ɵngcc0.ɵɵlistener("click", function ThyTableComponent_tbody_5_ng_container_1_ng_container_1_Template_tr_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r34); const group_r23 = ctx.$implicit; const ctx_r33 = ɵngcc0.ɵɵnextContext(3); return ctx_r33.expandGroup(group_r23); });
    ɵngcc0.ɵɵelementStart(2, "td");
    ɵngcc0.ɵɵelementStart(3, "div", 17);
    ɵngcc0.ɵɵelement(4, "thy-icon", 18);
    ɵngcc0.ɵɵtemplate(5, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_5_Template, 2, 5, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_ng_container_6_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const group_r23 = ctx.$implicit;
    const ctx_r22 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("colspan", ctx_r22.columns.length);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("thyIconName", group_r23.expand ? "angle-down" : "angle-right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r22.groupTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (group_r23.children == null ? null : group_r23.children.length) && group_r23.expand);
} }
function ThyTableComponent_tbody_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_ng_container_1_Template, 7, 4, "ng-container", 15);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r15.groups);
} }
function ThyTableComponent_tbody_5_ng_container_2_tr_1_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template(rf, ctx) { if (rf & 1) {
    const _r40 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 21);
    ɵngcc0.ɵɵlistener("cdkDragStarted", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_cdkDragStarted_0_listener() { ɵngcc0.ɵɵrestoreView(_r40); const ctx_r39 = ɵngcc0.ɵɵnextContext(3); return ctx_r39.onDragStarted(); })("mousedown", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_mousedown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); const ctx_r41 = ɵngcc0.ɵɵnextContext(3); return ctx_r41.draggableStopPropagation($event); })("touchstart", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_touchstart_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); const ctx_r42 = ɵngcc0.ɵɵnextContext(3); return ctx_r42.draggableStopPropagation($event); })("pointerdown", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_pointerdown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); const ctx_r43 = ɵngcc0.ɵɵnextContext(3); return ctx_r43.draggableStopPropagation($event); })("click", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); const row_r36 = ctx.$implicit; const ctx_r44 = ɵngcc0.ɵɵnextContext(3); return ctx_r44.onRowClick($event, row_r36); })("thyContextMenu", function ThyTableComponent_tbody_5_ng_container_2_tr_1_Template_tr_thyContextMenu_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r40); const row_r36 = ctx.$implicit; const ctx_r45 = ɵngcc0.ɵɵnextContext(3); return ctx_r45.onRowContextMenu($event, row_r36); });
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_2_tr_1_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const row_r36 = ctx.$implicit;
    const i_r37 = ctx.index;
    ɵngcc0.ɵɵnextContext(2);
    const _r20 = ɵngcc0.ɵɵreference(7);
    const ctx_r35 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDragData", row_r36)("ngClass", ctx_r35.renderRowClassName(row_r36, i_r37));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r20)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(4, _c8, row_r36, i_r37));
} }
function ThyTableComponent_tbody_5_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_2_tr_1_Template, 2, 7, "tr", 20);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r16.model)("ngForTrackBy", ctx_r16.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_container_3_ng_container_1_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_3_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const row_r47 = ctx.$implicit;
    const i_r48 = ctx.index;
    ɵngcc0.ɵɵnextContext(2);
    const _r18 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r18)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c8, row_r47, i_r48));
} }
function ThyTableComponent_tbody_5_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_3_ng_container_1_Template, 2, 5, "ng-container", 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r17 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r17.model)("ngForTrackBy", ctx_r17.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_template_4_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_ng_template_0_Template(rf, ctx) { }
const _c9 = function (a0, a1, a2) { return { row: a0, index: a1, level: a2 }; };
function ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_ng_template_0_Template, 0, 0, "ng-template", 11);
} if (rf & 2) {
    const child_r56 = ctx.$implicit;
    const j_r57 = ctx.index;
    const level_r52 = ɵngcc0.ɵɵnextContext(2).level;
    ɵngcc0.ɵɵnextContext();
    const _r18 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r18)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(2, _c9, child_r56, j_r57, level_r52 + 1));
} }
function ThyTableComponent_tbody_5_ng_template_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_4_ng_container_2_1_Template, 1, 6, undefined, 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const row_r50 = ɵngcc0.ɵɵnextContext().row;
    const ctx_r54 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", row_r50[ctx_r54.thyChildrenKey])("ngForTrackBy", ctx_r54.trackByFn);
} }
function ThyTableComponent_tbody_5_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    const _r62 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 21);
    ɵngcc0.ɵɵlistener("cdkDragStarted", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_cdkDragStarted_0_listener() { ɵngcc0.ɵɵrestoreView(_r62); const ctx_r61 = ɵngcc0.ɵɵnextContext(2); return ctx_r61.onDragStarted(); })("mousedown", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_mousedown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); const ctx_r63 = ɵngcc0.ɵɵnextContext(2); return ctx_r63.draggableStopPropagation($event); })("touchstart", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_touchstart_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); const ctx_r64 = ɵngcc0.ɵɵnextContext(2); return ctx_r64.draggableStopPropagation($event); })("pointerdown", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_pointerdown_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); const ctx_r65 = ɵngcc0.ɵɵnextContext(2); return ctx_r65.draggableStopPropagation($event); })("click", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); const row_r50 = ctx.row; const ctx_r66 = ɵngcc0.ɵɵnextContext(2); return ctx_r66.onRowClick($event, row_r50); })("thyContextMenu", function ThyTableComponent_tbody_5_ng_template_4_Template_tr_thyContextMenu_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r62); const row_r50 = ctx.row; const ctx_r67 = ɵngcc0.ɵɵnextContext(2); return ctx_r67.onRowContextMenu($event, row_r50); });
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_4_ng_template_1_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_4_ng_container_2_Template, 2, 2, "ng-container", 3);
} if (rf & 2) {
    const row_r50 = ctx.row;
    const i_r51 = ctx.index;
    const level_r52 = ctx.level;
    ɵngcc0.ɵɵnextContext();
    const _r20 = ɵngcc0.ɵɵreference(7);
    const ctx_r19 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDragData", row_r50)("ngClass", ctx_r19.renderRowClassName(row_r50, i_r51));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r20)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(5, _c9, row_r50, i_r51, level_r52 + 1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r19.mode === "tree" && ctx_r19.isExpanded(row_r50));
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-icon", 24);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    const _r81 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 26);
    ɵngcc0.ɵɵlistener("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_thy_icon_1_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r81); const row_r68 = ɵngcc0.ɵɵnextContext(3).row; const ctx_r79 = ɵngcc0.ɵɵnextContext(2); return ctx_r79.expandChildren($event, row_r68); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r82 = ɵngcc0.ɵɵnextContext(3);
    const row_r68 = ctx_r82.row;
    const level_r70 = ctx_r82.level;
    const ctx_r77 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵstyleProp("visibility", ctx_r77.showExpand(row_r68) ? "visible" : "hidden")("left", ctx_r77.iconIndentComputed(level_r70), "px");
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r77.isExpanded(row_r68) ? "angle-down" : "angle-right");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_ng_template_2_Template(rf, ctx) { }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_thy_icon_1_Template, 1, 5, "thy-icon", 25);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_ng_template_2_Template, 0, 0, "ng-template", 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext().$implicit;
    const row_r68 = ɵngcc0.ɵɵnextContext().row;
    const ctx_r75 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r75.mode === "tree" && column_r72.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", column_r72.templateRef)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(3, _c6, row_r68));
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    const _r96 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 26);
    ɵngcc0.ɵɵlistener("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_thy_icon_1_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r96); const row_r68 = ɵngcc0.ɵɵnextContext(4).row; const ctx_r94 = ɵngcc0.ɵɵnextContext(2); return ctx_r94.expandChildren($event, row_r68); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r97 = ɵngcc0.ɵɵnextContext(4);
    const row_r68 = ctx_r97.row;
    const level_r70 = ctx_r97.level;
    const ctx_r90 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵstyleProp("visibility", ctx_r90.showExpand(row_r68) ? "visible" : "hidden")("margin-left", ctx_r90.iconIndentComputed(level_r70), "px");
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r90.isExpanded(row_r68) ? "angle-down" : "angle-right");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const row_r68 = ɵngcc0.ɵɵnextContext().row;
    const ctx_r91 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r91.getModelValue(row_r68, column_r72.model), " ");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 30);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", column_r72.defaultText, " ");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_thy_icon_1_Template, 1, 5, "thy-icon", 27);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_container_2_Template, 2, 1, "ng-container", 28);
    ɵngcc0.ɵɵpipe(3, "isValidModelValue");
    ɵngcc0.ɵɵtemplate(4, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_ng_template_4_Template, 2, 1, "ng-template", null, 29, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r92 = ɵngcc0.ɵɵreference(5);
    const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const row_r68 = ɵngcc0.ɵɵnextContext().row;
    const ctx_r85 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r85.mode === "tree" && column_r72.expand);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(3, 3, ctx_r85.getModelValue(row_r68, column_r72.model)))("ngIfElse", _r92);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const i_r69 = ɵngcc0.ɵɵnextContext(3).index;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", i_r69 + 1, " ");
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    const _r106 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "input", 31);
    ɵngcc0.ɵɵlistener("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r106); const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r68 = ɵngcc0.ɵɵnextContext().row; return (row_r68[column_r72.key] = $event); })("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template_input_ngModelChange_1_listener() { ɵngcc0.ɵɵrestoreView(_r106); const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r68 = ɵngcc0.ɵɵnextContext().row; const ctx_r108 = ɵngcc0.ɵɵnextContext(2); return ctx_r108.onCheckboxChange(row_r68, column_r72); })("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template_input_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r106); const ctx_r111 = ɵngcc0.ɵɵnextContext(5); return ctx_r111.onStopPropagation($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const row_r68 = ɵngcc0.ɵɵnextContext().row;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", row_r68[column_r72.key])("disabled", column_r72.disabled);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    const _r115 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "input", 32);
    ɵngcc0.ɵɵlistener("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r115); const ctx_r114 = ɵngcc0.ɵɵnextContext(5); return ctx_r114.selectedRadioRow = $event; })("click", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template_input_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r115); const ctx_r116 = ɵngcc0.ɵɵnextContext(5); return ctx_r116.onStopPropagation($event); })("change", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template_input_change_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r115); const row_r68 = ɵngcc0.ɵɵnextContext(3).row; const ctx_r117 = ɵngcc0.ɵɵnextContext(2); return ctx_r117.onRadioSelectChange($event, row_r68); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const row_r68 = ɵngcc0.ɵɵnextContext().row;
    const ctx_r88 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r88.selectedRadioRow)("value", row_r68)("disabled", column_r72.disabled);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    const _r123 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "thy-switch", 33);
    ɵngcc0.ɵɵlistener("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template_thy_switch_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r123); const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r68 = ɵngcc0.ɵɵnextContext().row; return (row_r68[column_r72.key] = $event); })("ngModelChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template_thy_switch_ngModelChange_1_listener() { ɵngcc0.ɵɵrestoreView(_r123); const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r68 = ɵngcc0.ɵɵnextContext().row; const ctx_r125 = ɵngcc0.ɵɵnextContext(2); return ctx_r125.onModelChange(row_r68, column_r72); })("thyChange", function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template_thy_switch_thyChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r123); const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r68 = ɵngcc0.ɵɵnextContext().row; const ctx_r128 = ɵngcc0.ɵɵnextContext(2); return ctx_r128.onSwitchChange($event, row_r68, column_r72); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const row_r68 = ɵngcc0.ɵɵnextContext().row;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", row_r68[column_r72.key])("disabled", column_r72.disabled);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_1_Template, 6, 5, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_2_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_3_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(4, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_4_Template, 2, 3, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(5, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_ng_container_5_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r72 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r76 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.model && !column_r72.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.index);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.checkbox);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.radio);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r72.type === ctx_r76.customType.switch);
} }
function ThyTableComponent_tbody_5_ng_template_6_td_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 23);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_1_Template, 2, 0, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_2_Template, 3, 5, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTableComponent_tbody_5_ng_template_6_td_0_ng_container_3_Template, 6, 5, "ng-container", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r72 = ctx.$implicit;
    const j_r73 = ctx.index;
    const level_r70 = ɵngcc0.ɵɵnextContext().level;
    const ctx_r71 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("thy-table-checkbox-column", column_r72.type === ctx_r71.customType.checkbox);
    ɵngcc0.ɵɵproperty("ngClass", column_r72.className)("ngStyle", ctx_r71.mode === "tree" && column_r72.expand ? ctx_r71.tdIndentComputed(level_r70) : null);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", j_r73 === 0 && ctx_r71.draggable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r71.isTemplateRef(column_r72.templateRef));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r71.isTemplateRef(column_r72.templateRef));
} }
function ThyTableComponent_tbody_5_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyTableComponent_tbody_5_ng_template_6_td_0_Template, 4, 7, "td", 22);
} if (rf & 2) {
    const ctx_r21 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r21.columns);
} }
function ThyTableComponent_tbody_5_Template(rf, ctx) { if (rf & 1) {
    const _r136 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tbody", 12);
    ɵngcc0.ɵɵlistener("cdkDropListDropped", function ThyTableComponent_tbody_5_Template_tbody_cdkDropListDropped_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r136); const ctx_r135 = ɵngcc0.ɵɵnextContext(); return ctx_r135.onDragDropped($event); });
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_tbody_5_ng_container_1_Template, 2, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_tbody_5_ng_container_2_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, ThyTableComponent_tbody_5_ng_container_3_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(4, ThyTableComponent_tbody_5_ng_template_4_Template, 3, 9, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(6, ThyTableComponent_tbody_5_ng_template_6_Template, 1, 1, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDropListDisabled", !ctx_r3.draggable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "group");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "list");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "tree");
} }
function ThyTableComponent_div_6_ng_container_1_ng_template_1_Template(rf, ctx) { }
function ThyTableComponent_div_6_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_div_6_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 36);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r137 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r137.emptyTemplate);
} }
function ThyTableComponent_div_6_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-empty", 37);
} if (rf & 2) {
    const ctx_r139 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyMessage", ctx_r139.emptyOptions.message)("thyTranslationKey", ctx_r139.emptyOptions.translationKey)("thyTranslationValues", ctx_r139.emptyOptions.translationValues)("thyEntityName", ctx_r139.emptyOptions.entityName)("thyEntityNameTranslateKey", ctx_r139.emptyOptions.entityNameTranslateKey)("thyIconName", ctx_r139.emptyOptions.iconName)("thySize", ctx_r139.emptyOptions.size)("thyMarginTop", ctx_r139.emptyOptions.marginTop)("thyTopAuto", ctx_r139.emptyOptions.topAuto)("thyContainer", ctx_r139.emptyOptions.container);
} }
function ThyTableComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 34);
    ɵngcc0.ɵɵtemplate(1, ThyTableComponent_div_6_ng_container_1_Template, 2, 1, "ng-container", 28);
    ɵngcc0.ɵɵtemplate(2, ThyTableComponent_div_6_ng_template_2_Template, 1, 10, "ng-template", null, 35, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r138 = ɵngcc0.ɵɵreference(3);
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.emptyTemplate)("ngIfElse", _r138);
} }
function ThyTableComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r142 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵelementStart(1, "thy-pagination", 39);
    ɵngcc0.ɵɵlistener("thyPageChanged", function ThyTableComponent_div_8_Template_thy_pagination_thyPageChanged_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r142); const ctx_r141 = ɵngcc0.ɵɵnextContext(); return ctx_r141.onPageChange($event); })("thyPageIndexChange", function ThyTableComponent_div_8_Template_thy_pagination_thyPageIndexChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r142); const ctx_r143 = ɵngcc0.ɵɵnextContext(); return ctx_r143.onPageIndexChange($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("thy-table-footer-has-padding", ctx_r5.theme === "default");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyPageIndex", ctx_r5.pagination.index)("thyTotal", ctx_r5.pagination.total)("thyPageSize", ctx_r5.pagination.size)("thyShowTotal", ctx_r5.showTotal);
} }
function ThyTableComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c10 = function (a0) { return [a0]; };
const THY_TABLE_COLUMN_PARENT_COMPONENT = new InjectionToken('THY_TABLE_COLUMN_PARENT_COMPONENT');
/**
 * @deprecated THY_GRID_COLUMN_PARENT_COMPONENT is deprecated, please use THY_TABLE_COLUMN_PARENT_COMPONENT
 */
const THY_GRID_COLUMN_PARENT_COMPONENT = THY_TABLE_COLUMN_PARENT_COMPONENT;
class ThyTableColumnComponent {
    constructor(el, parent) {
        this.el = el;
        this.parent = parent;
        this.thyModelKey = '';
        this.thyTitle = '';
        this.thyType = '';
        this.thyWidth = '';
        this.thyClassName = '';
        this.thyHeaderClassName = '';
        this.thyDisabled = false;
        this.thyDefaultText = '';
        this.thyExpand = false;
        this.selections = [];
        this._firstChange = true;
        this.expand = false;
    }
    set thySelections(value) {
        if (value) {
            if (Array.isArray(value)) {
                this.selections = value;
            }
            else {
                this.selections = [value];
            }
            if (!this._firstChange) {
                this.parent.updateColumnSelections(this.key, this.selections);
            }
        }
    }
    set templateRef(value) {
        if (value) {
            if (!this.headerTemplateRef && !this.cellTemplateRef) {
                this.cellTemplateRef = value;
            }
        }
    }
    ngOnInit() {
        this.key = this._generateKey();
        this.model = this.thyModelKey;
        this.title = this.thyTitle;
        this.type = this.thyType;
        this.width = this.thyWidth;
        this.className = this.thyClassName;
        this.headerClassName = this.thyHeaderClassName;
        this.disabled = this.thyDisabled;
        this.defaultText = this.thyDefaultText;
        this._firstChange = false;
        this.expand = this.thyExpand;
    }
    _generateKey() {
        return ('[$$column]' +
            Math.random()
                .toString(16)
                .substr(2, 8));
    }
}
ThyTableColumnComponent.ɵfac = function ThyTableColumnComponent_Factory(t) { return new (t || ThyTableColumnComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(THY_GRID_COLUMN_PARENT_COMPONENT, 8)); };
ThyTableColumnComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTableColumnComponent, selectors: [["thy-grid-column"], ["thy-table-column"]], contentQueries: function ThyTableColumnComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, TemplateRef, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.cellTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerTemplateRef = _t.first);
    } }, inputs: { thyModelKey: "thyModelKey", thyTitle: "thyTitle", thyType: "thyType", thyWidth: "thyWidth", thyClassName: "thyClassName", thyHeaderClassName: "thyHeaderClassName", thyDisabled: "thyDisabled", thyDefaultText: "thyDefaultText", thyExpand: "thyExpand", thySelections: "thySelections" }, ngContentSelectors: _c2, decls: 1, vars: 0, template: function ThyTableColumnComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyTableColumnComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [THY_GRID_COLUMN_PARENT_COMPONENT,] }] }
];
ThyTableColumnComponent.propDecorators = {
    thyModelKey: [{ type: Input }],
    thyTitle: [{ type: Input }],
    thyType: [{ type: Input }],
    thyWidth: [{ type: Input }],
    thyClassName: [{ type: Input }],
    thyHeaderClassName: [{ type: Input }],
    thyDisabled: [{ type: Input }],
    thySelections: [{ type: Input }],
    thyDefaultText: [{ type: Input }],
    thyExpand: [{ type: Input }],
    headerTemplateRef: [{ type: ContentChild, args: ['header', { static: true },] }],
    cellTemplateRef: [{ type: ContentChild, args: ['cell', { static: true },] }],
    templateRef: [{ type: ContentChild, args: [TemplateRef, { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableColumnComponent, [{
        type: Component,
        args: [{
                selector: 'thy-grid-column,thy-table-column',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [THY_GRID_COLUMN_PARENT_COMPONENT]
            }] }]; }, { thyModelKey: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], thyType: [{
            type: Input
        }], thyWidth: [{
            type: Input
        }], thyClassName: [{
            type: Input
        }], thyHeaderClassName: [{
            type: Input
        }], thyDisabled: [{
            type: Input
        }], thyDefaultText: [{
            type: Input
        }], thyExpand: [{
            type: Input
        }], thySelections: [{
            type: Input
        }], templateRef: [{
            type: ContentChild,
            args: [TemplateRef, { static: true }]
        }], cellTemplateRef: [{
            type: ContentChild,
            args: ['cell', { static: true }]
        }], headerTemplateRef: [{
            type: ContentChild,
            args: ['header', { static: true }]
        }] }); })();
/**
 * @deprecated ThyTableColumnComponent is deprecated, please use ThyTableColumnComponent
 */
const ThyGridColumnComponent = ThyTableColumnComponent;

const tableThemeMap = {
    default: 'table-default',
    bordered: 'table-bordered'
};
const tableSizeMap = {
    sm: 'table-sm'
};
const customType = {
    index: 'index',
    checkbox: 'checkbox',
    radio: 'radio',
    switch: 'switch'
};
const _MixinBase = mixinUnsubscribe(MixinBase);
class ThyTableComponent extends _MixinBase {
    constructor(elementRef, _differs, viewportRuler, updateHostClassService, document) {
        super();
        this.elementRef = elementRef;
        this._differs = _differs;
        this.viewportRuler = viewportRuler;
        this.updateHostClassService = updateHostClassService;
        this.document = document;
        this.customType = customType;
        this.model = [];
        this.groups = [];
        this.rowKey = '_id';
        this.mode = 'list';
        this.columns = [];
        this.theme = 'default';
        this.className = '';
        this.size = 'default';
        this.loadingDone = true;
        this.emptyOptions = {};
        this.draggable = false;
        this.selectedRadioRow = null;
        this.pagination = { index: 1, size: 20, total: 0 };
        this.wholeRowSelect = false;
        this._filter = null;
        this.initialized = false;
        this._oldThyClassName = '';
        this.thyShowHeader = true;
        this.showTotal = false;
        this.thyIndent = 20;
        this.thyChildrenKey = 'children';
        this.thyOnSwitchChange = new EventEmitter();
        this.thyOnPageChange = new EventEmitter();
        this.thyOnPageIndexChange = new EventEmitter();
        this.thyOnMultiSelectChange = new EventEmitter();
        this.thyOnRadioSelectChange = new EventEmitter();
        this.thyOnDraggableChange = new EventEmitter();
        this.thyOnRowClick = new EventEmitter();
        this.thyOnRowContextMenu = new EventEmitter();
        this.isTableClass = true;
        // 数据的折叠展开状态
        this.expandStatusMap = {};
        this._bindTrackFn();
    }
    set thyMode(value) {
        this.mode = value || this.mode;
    }
    set thyGroupBy(value) {
        this.groupBy = value;
    }
    set thyRowKey(value) {
        this.rowKey = value || this.rowKey;
    }
    set thyGroups(value) {
        if (this.mode === 'group') {
            this.buildGroups(value);
        }
    }
    set thyModel(value) {
        this.model = value || [];
        this._diff = this._differs.find(this.model).create();
        this._initializeDataModel();
        if (this.mode === 'group') {
            this.buildModel();
        }
    }
    set thyTheme(value) {
        this.theme = value || this.theme;
        this._setClass();
    }
    set thySize(value) {
        this.size = value || this.size;
        this._setClass();
    }
    set thyClassName(value) {
        const list = this.className.split(' ').filter(a => a.trim());
        const index = list.findIndex(item => item === this._oldThyClassName);
        if (index !== -1) {
            list.splice(index, 1, value);
        }
        else {
            list.push(value);
        }
        this._oldThyClassName = value;
        this.className = list.join(' ');
    }
    set thyRowClassName(value) {
        this.rowClassName = value;
    }
    set thyLoadingDone(value) {
        this.loadingDone = value;
    }
    set thyLoadingText(value) {
        this.loadingText = value;
    }
    set thyEmptyOptions(value) {
        this.emptyOptions = value;
    }
    set thyDraggable(value) {
        this.draggable = coerceBooleanProperty(value);
        if (this.mode !== 'list' && this.draggable) {
            throw new Error('Only list mode sorting is supported');
        }
    }
    set thyFilter(value) {
        this._filter = value;
    }
    set thyPageIndex(value) {
        this.pagination.index = value;
    }
    set thyPageSize(value) {
        this.pagination.size = value;
    }
    set thyPageTotal(value) {
        this.pagination.total = value;
    }
    set thyWholeRowSelect(value) {
        if (value) {
            this.className += ' table-hover';
        }
        this.wholeRowSelect = value;
    }
    set listOfColumnComponents(components) {
        if (components) {
            this._listOfColumnComponents = components;
            this._initializeColumns();
            this._initializeDataModel();
        }
    }
    _getSelectionKeys(selections) {
        return selections.map((item) => {
            if (typeof item === 'number' || typeof item === 'string') {
                return item;
            }
            else {
                return item[this.rowKey];
            }
        });
    }
    _initializeColumns() {
        const components = this._listOfColumnComponents ? this._listOfColumnComponents.toArray() : [];
        const hasExpand = components.some(item => item.expand === true);
        this.columns = components.map((component, i) => {
            const selections = this._getSelectionKeys(component.selections);
            return {
                key: component.key,
                model: component.model,
                title: component.title,
                type: component.type,
                selections: selections,
                width: component.width,
                className: component.className,
                headerClassName: component.headerClassName,
                disabled: component.disabled,
                defaultText: component.defaultText,
                expand: hasExpand ? component.expand : i === 0,
                templateRef: component.cellTemplateRef,
                headerTemplateRef: component.headerTemplateRef
            };
        });
    }
    _initializeDataModel() {
        this.model.forEach(row => {
            this.columns.forEach(column => {
                this._initialSelections(row, column);
                this._initialCustomModelValue(row, column);
            });
        });
    }
    _initialSelections(row, column) {
        if (column.selections) {
            if (column.type === 'checkbox') {
                row[column.key] = column.selections.includes(row[this.rowKey]);
                this.onModelChange(row, column);
            }
            if (column.type === 'radio') {
                if (column.selections.includes(row[this.rowKey])) {
                    this.selectedRadioRow = row;
                }
            }
        }
    }
    _initialCustomModelValue(row, column) {
        if (column.type === customType.switch) {
            row[column.key] = get(row, column.model);
        }
    }
    _refreshCustomModelValue(row) {
        this.columns.forEach(column => {
            this._initialCustomModelValue(row, column);
        });
    }
    _applyDiffChanges(changes) {
        if (changes) {
            changes.forEachAddedItem((record) => {
                this._refreshCustomModelValue(record.item);
            });
        }
    }
    _bindTrackFn() {
        this.trackByFn = function (index, row) {
            return row && this.rowKey ? row[this.rowKey] : index;
        }.bind(this);
    }
    _destroyInvalidAttribute() {
        this.model.forEach(row => {
            for (const key in row) {
                if (key.includes('[$$column]')) {
                    delete row[key];
                }
            }
        });
    }
    _setClass(first = false) {
        if (!first && !this.initialized) {
            return;
        }
        const classNames = [];
        if (tableSizeMap[this.size]) {
            classNames.push(tableSizeMap[this.size]);
        }
        if (tableThemeMap[this.theme]) {
            classNames.push(tableThemeMap[this.theme]);
        }
        if (tableSizeMap[this.size] === tableSizeMap['sm'] && tableThemeMap[this.theme] === tableThemeMap['default']) {
            classNames.push('table-default-sm-bottom-padding');
        }
        this.updateHostClassService.updateClass(classNames);
    }
    _filterModel() {
        if (this.model && this.model.length > 0) {
            if (this._filter) {
            }
        }
    }
    updateColumnSelections(key, selections) {
        const column = this.columns.find(item => item.key === key);
        column.selections = this._getSelectionKeys(selections);
        this.model.forEach(row => {
            this._initialSelections(row, column);
        });
    }
    isTemplateRef(ref) {
        return ref instanceof TemplateRef;
    }
    getModelValue(row, path) {
        return get(row, path);
    }
    renderRowClassName(row, index) {
        if (!this.rowClassName) {
            return null;
        }
        if (isString(this.rowClassName)) {
            return this.rowClassName;
        }
        else {
            return this.rowClassName(row, index);
        }
    }
    onModelChange(row, column) {
        if (column.model) {
            set(row, column.model, row[column.key]);
        }
    }
    onStopPropagation(event) {
        if (this.wholeRowSelect) {
            event.stopPropagation();
        }
    }
    onPageChange(event) {
        this.thyOnPageChange.emit(event);
    }
    onPageIndexChange(event) {
        this.thyOnPageIndexChange.emit(event);
    }
    onCheckboxChange(row, column) {
        this.onModelChange(row, column);
        this.onMultiSelectChange(null, row, column);
    }
    onMultiSelectChange(event, row, column) {
        const rows = this.model.filter(item => {
            return item[column.key];
        });
        const multiSelectEvent = {
            event: event,
            row: row,
            rows: rows
        };
        this.thyOnMultiSelectChange.emit(multiSelectEvent);
    }
    onRadioSelectChange(event, row) {
        const radioSelectEvent = {
            event: event,
            row: row
        };
        this.thyOnRadioSelectChange.emit(radioSelectEvent);
    }
    onSwitchChange(event, row, column) {
        const switchEvent = {
            event: event,
            row: row,
            refresh: (value) => {
                value = value || row;
                setTimeout(() => {
                    value[column.key] = get(value, column.model);
                });
            }
        };
        this.thyOnSwitchChange.emit(switchEvent);
    }
    showExpand(row) {
        return row[this.thyChildrenKey] && row[this.thyChildrenKey].length > 0;
    }
    isExpanded(row) {
        return this.expandStatusMap[row[this.rowKey]];
    }
    iconIndentComputed(level) {
        if (this.mode === 'tree') {
            return level * this.thyIndent - 10;
        }
    }
    tdIndentComputed(level) {
        return {
            position: 'relative',
            paddingLeft: `${(level + 1) * this.thyIndent - 10}px`
        };
    }
    expandChildren(event, row) {
        event.stopPropagation();
        if (this.isExpanded(row)) {
            this.expandStatusMap[row[this.rowKey]] = false;
        }
        else {
            this.expandStatusMap[row[this.rowKey]] = true;
        }
    }
    onDragStarted() {
        setTimeout(() => {
            const preview = this.document.getElementsByClassName('cdk-drag-preview')[0];
            if (preview) {
                preview.classList.add('thy-table-drag-preview');
            }
        });
    }
    onDragDropped(event) {
        const dragEvent = {
            model: event.item,
            models: this.model,
            oldIndex: event.previousIndex,
            newIndex: event.currentIndex
        };
        moveItemInArray(this.model, event.previousIndex, event.currentIndex);
        this.thyOnDraggableChange.emit(dragEvent);
    }
    onRowClick(event, row) {
        if (this.wholeRowSelect) {
            const column = this.columns.find(item => {
                return item.type === customType.checkbox || item.type === customType.radio;
            });
            if (!column.disabled) {
                if (column.type === customType.checkbox) {
                    row[column.key] = !row[column.key];
                    this.onModelChange(row, column);
                    this.onMultiSelectChange(event, row, column);
                }
                if (column.type === customType.radio) {
                    this.selectedRadioRow = row;
                    this.onRadioSelectChange(event, row);
                }
            }
        }
        const rowEvent = {
            event: event,
            row: row
        };
        this.thyOnRowClick.emit(rowEvent);
    }
    onRowContextMenu(event, row) {
        const contextMenuEvent = {
            event: event,
            row: row
        };
        this.thyOnRowContextMenu.emit(contextMenuEvent);
    }
    draggableStopPropagation(event) {
        if (!this.draggable) {
            event.stopPropagation();
        }
    }
    _refreshColumns() {
        const components = this._listOfColumnComponents ? this._listOfColumnComponents.toArray() : [];
        const _columns = components.map(component => {
            return {
                width: component.width,
                className: component.className
            };
        });
        this.columns.forEach((n, i) => {
            Object.assign(n, _columns[i]);
        });
    }
    buildGroups(originGroups) {
        const collapsedIds = this.groups.filter(group => !group.expand).map(group => group.id);
        this.groups = [];
        originGroups.forEach((origin) => {
            const group = { id: origin[this.rowKey], children: [], origin };
            group.expand = !collapsedIds.includes(group.id);
            this.groups.push(group);
        });
    }
    buildModel() {
        const groupsMap = keyBy(this.groups, 'id');
        this.model.forEach(row => {
            const group = groupsMap[row[this.groupBy]];
            if (group) {
                group.children.push(row);
            }
        });
    }
    expandGroup(group) {
        group.expand = !group.expand;
    }
    ngOnInit() {
        this.updateHostClassService.initializeElement(this.tableElementRef.nativeElement);
        this._setClass(true);
        this.initialized = true;
        merge(this.viewportRuler.change(200), of(null).pipe(delay(200)))
            .pipe(takeUntil(this.ngUnsubscribe$))
            .subscribe(() => {
            this._refreshColumns();
        });
    }
    ngOnChanges(simpleChangs) {
        const modeChange = simpleChangs.thyMode;
        const thyGroupsChange = simpleChangs.thyGroups;
        const isGroupMode = modeChange && modeChange.currentValue === 'group';
        if (isGroupMode && thyGroupsChange && thyGroupsChange.firstChange) {
            this.buildGroups(thyGroupsChange.currentValue);
            this.buildModel();
        }
    }
    ngDoCheck() {
        if (this._diff) {
            const changes = this._diff.diff(this.model);
            this._applyDiffChanges(changes);
        }
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        this._destroyInvalidAttribute();
    }
}
ThyTableComponent.ɵfac = function ThyTableComponent_Factory(t) { return new (t || ThyTableComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.IterableDiffers), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ViewportRuler), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UpdateHostClassService), ɵngcc0.ɵɵdirectiveInject(DOCUMENT)); };
ThyTableComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTableComponent, selectors: [["thy-grid"], ["thy-table"]], contentQueries: function ThyTableComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c3, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, _c4, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, ThyTableColumnComponent, false);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.emptyTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.groupTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.listOfColumnComponents = _t);
    } }, viewQuery: function ThyTableComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c5, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tableElementRef = _t.first);
    } }, hostVars: 4, hostBindings: function ThyTableComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-table", ctx.isTableClass)("thy-table-hover-display-operation", ctx.thyHoverDisplayOperation);
    } }, inputs: { thyShowHeader: "thyShowHeader", showTotal: ["thyShowTotal", "showTotal"], thyIndent: "thyIndent", thyChildrenKey: "thyChildrenKey", thyMode: "thyMode", thyGroupBy: "thyGroupBy", thyRowKey: "thyRowKey", thyGroups: "thyGroups", thyModel: "thyModel", thyTheme: "thyTheme", thySize: "thySize", thyClassName: "thyClassName", thyRowClassName: "thyRowClassName", thyLoadingDone: "thyLoadingDone", thyLoadingText: "thyLoadingText", thyEmptyOptions: "thyEmptyOptions", thyDraggable: "thyDraggable", thyFilter: "thyFilter", thyPageIndex: "thyPageIndex", thyPageSize: "thyPageSize", thyPageTotal: "thyPageTotal", thyWholeRowSelect: "thyWholeRowSelect", thyHoverDisplayOperation: "thyHoverDisplayOperation" }, outputs: { thyOnSwitchChange: "thyOnSwitchChange", thyOnPageChange: "thyOnPageChange", thyOnPageIndexChange: "thyOnPageIndexChange", thyOnMultiSelectChange: "thyOnMultiSelectChange", thyOnRadioSelectChange: "thyOnRadioSelectChange", thyOnDraggableChange: "thyOnDraggableChange", thyOnRowClick: "thyOnRowClick", thyOnRowContextMenu: "thyOnRowContextMenu" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: THY_TABLE_COLUMN_PARENT_COMPONENT,
                useExisting: ThyTableComponent
            },
            UpdateHostClassService
        ]), ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c2, decls: 10, vars: 14, consts: [[1, "table", 3, "ngClass"], ["table", ""], [3, "width", 4, "ngFor", "ngForOf"], [4, "ngIf"], ["thyDragDrop", "", "cdkDropList", "", 3, "cdkDropListDisabled", "cdkDropListDropped", 4, "ngIf"], ["class", "thy-table-empty clear-hover", 4, "ngIf"], [3, "thyDone", "thyTip"], ["class", "thy-table-footer", 3, "thy-table-footer-has-padding", 4, "ngIf"], [3, "width"], [3, "ngClass", 4, "ngFor", "ngForOf"], [3, "ngClass"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["thyDragDrop", "", "cdkDropList", "", 3, "cdkDropListDisabled", "cdkDropListDropped"], ["trTemplate", ""], ["tdsTemplate", ""], [4, "ngFor", "ngForOf"], [1, "thy-table-group", 3, "click"], [1, "thy-table-group-container"], [1, "expand-icon", 3, "thyIconName"], [4, "ngFor", "ngForOf", "ngForTrackBy"], ["class", "thy-sortable-item", "cdkDrag", "", 3, "cdkDragData", "ngClass", "cdkDragStarted", "mousedown", "touchstart", "pointerdown", "click", "thyContextMenu", 4, "ngFor", "ngForOf", "ngForTrackBy"], ["cdkDrag", "", 1, "thy-sortable-item", 3, "cdkDragData", "ngClass", "cdkDragStarted", "mousedown", "touchstart", "pointerdown", "click", "thyContextMenu"], [3, "ngClass", "thy-table-checkbox-column", "ngStyle", 4, "ngFor", "ngForOf"], [3, "ngClass", "ngStyle"], ["thyIconName", "drag", 1, "table-draggable-icon"], ["class", "tree-expand-icon cursor-pointer", 3, "thyIconName", "visibility", "left", "click", 4, "ngIf"], [1, "tree-expand-icon", "cursor-pointer", 3, "thyIconName", "click"], ["class", "tree-expand-icon cursor-pointer", 3, "thyIconName", "visibility", "marginLeft", "click", 4, "ngIf"], [4, "ngIf", "ngIfElse"], ["default", ""], [1, "text-desc"], ["type", "checkbox", 3, "ngModel", "disabled", "ngModelChange", "click"], ["type", "radio", 3, "ngModel", "value", "disabled", "ngModelChange", "click", "change"], [3, "ngModel", "disabled", "ngModelChange", "thyChange"], [1, "thy-table-empty", "clear-hover"], ["defaultEmpty", ""], [3, "ngTemplateOutlet"], [3, "thyMessage", "thyTranslationKey", "thyTranslationValues", "thyEntityName", "thyEntityNameTranslateKey", "thyIconName", "thySize", "thyMarginTop", "thyTopAuto", "thyContainer"], [1, "thy-table-footer"], [3, "thyPageIndex", "thyTotal", "thyPageSize", "thyShowTotal", "thyPageChanged", "thyPageIndexChange"]], template: function ThyTableComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "table", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "colgroup");
        ɵngcc0.ɵɵtemplate(3, ThyTableComponent_col_3_Template, 1, 1, "col", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, ThyTableComponent_thead_4_Template, 3, 1, "thead", 3);
        ɵngcc0.ɵɵtemplate(5, ThyTableComponent_tbody_5_Template, 8, 4, "tbody", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, ThyTableComponent_div_6_Template, 4, 2, "div", 5);
        ɵngcc0.ɵɵelement(7, "thy-loading", 6);
        ɵngcc0.ɵɵtemplate(8, ThyTableComponent_div_8_Template, 2, 6, "div", 7);
        ɵngcc0.ɵɵtemplate(9, ThyTableComponent_ng_template_9_Template, 1, 0, "ng-template");
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("table-draggable", ctx.draggable)("table-group", ctx.mode === "group");
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(12, _c10, ctx.className));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.columns);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowHeader);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingDone);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingDone && ctx.model.length === 0 && ctx.groups.length === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("thyDone", ctx.loadingDone)("thyTip", ctx.loadingText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.pagination.total > ctx.pagination.size);
    } }, directives: function () { return [ɵngcc3.NgClass, ɵngcc3.NgForOf, ɵngcc3.NgIf, ɵngcc4.ThyLoadingComponent, ɵngcc3.NgTemplateOutlet, ɵngcc5.ThyDragDropDirective, ɵngcc6.CdkDropList, ɵngcc7.ThyIconComponent, ɵngcc6.CdkDrag, ɵngcc5.ThyContextMenuDirective, ɵngcc3.NgStyle, ɵngcc8.CheckboxControlValueAccessor, ɵngcc8.NgControlStatus, ɵngcc8.NgModel, ɵngcc8.RadioControlValueAccessor, ɵngcc8.DefaultValueAccessor, ɵngcc9.ThySwitchComponent, ɵngcc10.ThyEmptyComponent, ɵngcc11.ThyPaginationComponent]; }, pipes: function () { return [TableIsValidModelValuePipe]; }, encapsulation: 2 });
ThyTableComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: IterableDiffers },
    { type: ViewportRuler },
    { type: UpdateHostClassService },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
ThyTableComponent.propDecorators = {
    emptyTemplate: [{ type: ContentChild, args: ['empty',] }],
    tableElementRef: [{ type: ViewChild, args: ['table', { static: true },] }],
    thyMode: [{ type: Input }],
    thyGroupBy: [{ type: Input }],
    thyRowKey: [{ type: Input }],
    thyGroups: [{ type: Input }],
    thyModel: [{ type: Input }],
    thyTheme: [{ type: Input }],
    thySize: [{ type: Input }],
    thyClassName: [{ type: Input }],
    thyRowClassName: [{ type: Input }],
    thyLoadingDone: [{ type: Input }],
    thyLoadingText: [{ type: Input }],
    thyEmptyOptions: [{ type: Input }],
    thyDraggable: [{ type: Input }],
    thyFilter: [{ type: Input }],
    thyPageIndex: [{ type: Input }],
    thyPageSize: [{ type: Input }],
    thyPageTotal: [{ type: Input }],
    thyWholeRowSelect: [{ type: Input }],
    thyShowHeader: [{ type: Input }],
    showTotal: [{ type: Input, args: ['thyShowTotal',] }],
    thyIndent: [{ type: Input }],
    thyChildrenKey: [{ type: Input }],
    thyHoverDisplayOperation: [{ type: HostBinding, args: ['class.thy-table-hover-display-operation',] }, { type: Input }],
    thyOnSwitchChange: [{ type: Output }],
    thyOnPageChange: [{ type: Output }],
    thyOnPageIndexChange: [{ type: Output }],
    thyOnMultiSelectChange: [{ type: Output }],
    thyOnRadioSelectChange: [{ type: Output }],
    thyOnDraggableChange: [{ type: Output }],
    thyOnRowClick: [{ type: Output }],
    thyOnRowContextMenu: [{ type: Output }],
    groupTemplate: [{ type: ContentChild, args: ['group', { static: true },] }],
    listOfColumnComponents: [{ type: ContentChildren, args: [ThyTableColumnComponent,] }],
    isTableClass: [{ type: HostBinding, args: ['class.thy-table',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableComponent, [{
        type: Component,
        args: [{
                selector: 'thy-grid,thy-table',
                template: "<table #table class=\"table\" [ngClass]=\"[className]\" [class.table-draggable]=\"draggable\" [class.table-group]=\"mode === 'group'\">\n  <colgroup>\n    <col *ngFor=\"let column of columns\" [width]=\"column.width\" />\n  </colgroup>\n\n  <thead *ngIf=\"thyShowHeader\">\n    <tr>\n      <th *ngFor=\"let column of columns\" [ngClass]=\"column.headerClassName\">\n        <ng-container *ngIf=\"!column.headerTemplateRef\">\n          {{ column.title }}\n        </ng-container>\n        <ng-container *ngIf=\"column.headerTemplateRef\">\n          <ng-template [ngTemplateOutlet]=\"column.headerTemplateRef\" [ngTemplateOutletContext]=\"{ $implicit: column }\"></ng-template>\n        </ng-container>\n      </th>\n    </tr>\n  </thead>\n\n  <tbody *ngIf=\"loadingDone\" thyDragDrop cdkDropList [cdkDropListDisabled]=\"!draggable\" (cdkDropListDropped)=\"onDragDropped($event)\">\n    <!-- group -->\n    <ng-container *ngIf=\"mode === 'group'\">\n      <ng-container *ngFor=\"let group of groups\">\n        <tr class=\"thy-table-group\" (click)=\"expandGroup(group)\">\n          <td [attr.colspan]=\"columns.length\">\n            <div class=\"thy-table-group-container\">\n              <thy-icon class=\"expand-icon\" [thyIconName]=\"group.expand ? 'angle-down' : 'angle-right'\"></thy-icon>\n              <ng-container *ngIf=\"groupTemplate\">\n                <ng-template\n                  [ngTemplateOutlet]=\"groupTemplate\"\n                  [ngTemplateOutletContext]=\"{ $implicit: group.origin, group: group.origin }\"\n                ></ng-template>\n              </ng-container>\n            </div>\n          </td>\n        </tr>\n        <ng-container *ngIf=\"group.children?.length && group.expand\">\n          <ng-container *ngFor=\"let row of group.children; trackBy: trackByFn; let i = index\">\n            <ng-template [ngTemplateOutlet]=\"trTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: 0 }\"></ng-template>\n          </ng-container>\n        </ng-container>\n      </ng-container>\n    </ng-container>\n\n    <!-- list\u6A21\u5F0F\u4E0B\u6CA1\u6709\u76F4\u63A5\u4F7F\u7528item\u6A21\u677F\uFF0C\u662F\u56E0\u4E3A\u4F7F\u7528\u4E86\u6A21\u677F\u5F71\u54CD\u4E86Angular\u7684\u4F9D\u8D56\u6811\uFF0C\u5BFC\u81F4CdkDragDrop\u6392\u5E8F\u5931\u6548\uFF0C\u76EE\u524D\u8FD9\u6837\u4FEE\u6539\u53EF\u4EE5\u6682\u65F6\u89E3\u51B3\u7EBF\u4E0A\u4E0D\u80FD\u6392\u5E8F\u7684\u95EE\u9898\uFF0CTree\u6A21\u5F0F\u7684\u6392\u5E8F\u6682\u65F6\u4E0D\u652F\u6301 -->\n    <ng-container *ngIf=\"mode === 'list'\">\n      <tr\n        *ngFor=\"let row of model; trackBy: trackByFn; let i = index\"\n        class=\"thy-sortable-item\"\n        cdkDrag\n        [cdkDragData]=\"row\"\n        (cdkDragStarted)=\"onDragStarted()\"\n        [ngClass]=\"renderRowClassName(row, i)\"\n        (mousedown)=\"draggableStopPropagation($event)\"\n        (touchstart)=\"draggableStopPropagation($event)\"\n        (pointerdown)=\"draggableStopPropagation($event)\"\n        (click)=\"onRowClick($event, row)\"\n        (thyContextMenu)=\"onRowContextMenu($event, row)\"\n      >\n        <ng-template [ngTemplateOutlet]=\"tdsTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: 0 }\"></ng-template>\n      </tr>\n    </ng-container>\n\n    <ng-container *ngIf=\"mode === 'tree'\">\n      <ng-container *ngFor=\"let row of model; trackBy: trackByFn; let i = index\">\n        <ng-template [ngTemplateOutlet]=\"trTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: 0 }\"></ng-template>\n      </ng-container>\n    </ng-container>\n\n    <ng-template #trTemplate let-row=\"row\" let-i=\"index\" let-level=\"level\">\n      <tr\n        class=\"thy-sortable-item\"\n        cdkDrag\n        [cdkDragData]=\"row\"\n        (cdkDragStarted)=\"onDragStarted()\"\n        [ngClass]=\"renderRowClassName(row, i)\"\n        (mousedown)=\"draggableStopPropagation($event)\"\n        (touchstart)=\"draggableStopPropagation($event)\"\n        (pointerdown)=\"draggableStopPropagation($event)\"\n        (click)=\"onRowClick($event, row)\"\n        (thyContextMenu)=\"onRowContextMenu($event, row)\"\n      >\n        <ng-template [ngTemplateOutlet]=\"tdsTemplate\" [ngTemplateOutletContext]=\"{ row: row, index: i, level: level + 1 }\"></ng-template>\n      </tr>\n\n      <ng-container *ngIf=\"mode === 'tree' && isExpanded(row)\">\n        <ng-template\n          *ngFor=\"let child of row[thyChildrenKey]; trackBy: trackByFn; let j = index\"\n          [ngTemplateOutlet]=\"trTemplate\"\n          [ngTemplateOutletContext]=\"{ row: child, index: j, level: level + 1 }\"\n        ></ng-template>\n      </ng-container>\n    </ng-template>\n\n    <ng-template #tdsTemplate let-row=\"row\" let-i=\"index\" let-level=\"level\">\n      <td\n        *ngFor=\"let column of columns; let j = index\"\n        [ngClass]=\"column.className\"\n        [class.thy-table-checkbox-column]=\"column.type === customType.checkbox\"\n        [ngStyle]=\"mode === 'tree' && column.expand ? tdIndentComputed(level) : null\"\n      >\n        <ng-container *ngIf=\"j === 0 && draggable\">\n          <thy-icon class=\"table-draggable-icon\" thyIconName=\"drag\"></thy-icon>\n        </ng-container>\n\n        <!--template-->\n        <ng-container *ngIf=\"isTemplateRef(column.templateRef)\">\n          <thy-icon\n            class=\"tree-expand-icon cursor-pointer\"\n            *ngIf=\"mode === 'tree' && column.expand\"\n            [thyIconName]=\"isExpanded(row) ? 'angle-down' : 'angle-right'\"\n            [style.visibility]=\"showExpand(row) ? 'visible' : 'hidden'\"\n            [style.left.px]=\"iconIndentComputed(level)\"\n            (click)=\"expandChildren($event, row)\"\n          ></thy-icon>\n          <ng-template [ngTemplateOutlet]=\"column.templateRef\" [ngTemplateOutletContext]=\"{ $implicit: row }\"></ng-template>\n        </ng-container>\n\n        <!--not template-->\n        <ng-container *ngIf=\"!isTemplateRef(column.templateRef)\">\n          <!-- default -->\n          <ng-container *ngIf=\"column.model && !column.type\">\n            <thy-icon\n              class=\"tree-expand-icon cursor-pointer\"\n              *ngIf=\"mode === 'tree' && column.expand\"\n              [thyIconName]=\"isExpanded(row) ? 'angle-down' : 'angle-right'\"\n              [style.visibility]=\"showExpand(row) ? 'visible' : 'hidden'\"\n              [style.marginLeft.px]=\"iconIndentComputed(level)\"\n              (click)=\"expandChildren($event, row)\"\n            ></thy-icon>\n            <ng-container *ngIf=\"getModelValue(row, column.model) | isValidModelValue; else default\">\n              {{ getModelValue(row, column.model) }}\n            </ng-container>\n            <ng-template #default>\n              <div class=\"text-desc\">\n                {{ column.defaultText }}\n              </div>\n            </ng-template>\n          </ng-container>\n\n          <!-- index -->\n          <ng-container *ngIf=\"column.type === customType.index\">\n            {{ i + 1 }}\n          </ng-container>\n\n          <!-- checkbox -->\n          <ng-container *ngIf=\"column.type === customType.checkbox\">\n            <input\n              type=\"checkbox\"\n              [(ngModel)]=\"row[column.key]\"\n              (ngModelChange)=\"onCheckboxChange(row, column)\"\n              (click)=\"onStopPropagation($event)\"\n              [disabled]=\"column.disabled\"\n            />\n          </ng-container>\n\n          <!-- radio -->\n          <ng-container *ngIf=\"column.type === customType.radio\">\n            <input\n              type=\"radio\"\n              [(ngModel)]=\"selectedRadioRow\"\n              [value]=\"row\"\n              [disabled]=\"column.disabled\"\n              (click)=\"onStopPropagation($event)\"\n              (change)=\"onRadioSelectChange($event, row)\"\n            />\n          </ng-container>\n\n          <!-- switch -->\n          <ng-container *ngIf=\"column.type === customType.switch\">\n            <thy-switch\n              [(ngModel)]=\"row[column.key]\"\n              (ngModelChange)=\"onModelChange(row, column)\"\n              [disabled]=\"column.disabled\"\n              (thyChange)=\"onSwitchChange($event, row, column)\"\n            ></thy-switch>\n          </ng-container>\n        </ng-container>\n      </td>\n    </ng-template>\n  </tbody>\n</table>\n\n<div *ngIf=\"loadingDone && model.length === 0 && groups.length === 0\" class=\"thy-table-empty clear-hover\">\n  <ng-container *ngIf=\"emptyTemplate; else defaultEmpty\">\n    <ng-template [ngTemplateOutlet]=\"emptyTemplate\"></ng-template>\n  </ng-container>\n  <ng-template #defaultEmpty>\n    <thy-empty\n      [thyMessage]=\"emptyOptions.message\"\n      [thyTranslationKey]=\"emptyOptions.translationKey\"\n      [thyTranslationValues]=\"emptyOptions.translationValues\"\n      [thyEntityName]=\"emptyOptions.entityName\"\n      [thyEntityNameTranslateKey]=\"emptyOptions.entityNameTranslateKey\"\n      [thyIconName]=\"emptyOptions.iconName\"\n      [thySize]=\"emptyOptions.size\"\n      [thyMarginTop]=\"emptyOptions.marginTop\"\n      [thyTopAuto]=\"emptyOptions.topAuto\"\n      [thyContainer]=\"emptyOptions.container\"\n    ></thy-empty>\n  </ng-template>\n</div>\n\n<thy-loading [thyDone]=\"loadingDone\" [thyTip]=\"loadingText\"></thy-loading>\n\n<div class=\"thy-table-footer\" *ngIf=\"pagination.total > pagination.size\" [class.thy-table-footer-has-padding]=\"theme === 'default'\">\n  <thy-pagination\n    [thyPageIndex]=\"pagination.index\"\n    [thyTotal]=\"pagination.total\"\n    [thyPageSize]=\"pagination.size\"\n    [thyShowTotal]=\"showTotal\"\n    (thyPageChanged)=\"onPageChange($event)\"\n    (thyPageIndexChange)=\"onPageIndexChange($event)\"\n  ></thy-pagination>\n</div>\n\n<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                providers: [
                    {
                        provide: THY_TABLE_COLUMN_PARENT_COMPONENT,
                        useExisting: ThyTableComponent
                    },
                    UpdateHostClassService
                ],
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.IterableDiffers }, { type: ɵngcc1.ViewportRuler }, { type: ɵngcc2.UpdateHostClassService }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, { thyShowHeader: [{
            type: Input
        }], showTotal: [{
            type: Input,
            args: ['thyShowTotal']
        }], thyIndent: [{
            type: Input
        }], thyChildrenKey: [{
            type: Input
        }], thyOnSwitchChange: [{
            type: Output
        }], thyOnPageChange: [{
            type: Output
        }], thyOnPageIndexChange: [{
            type: Output
        }], thyOnMultiSelectChange: [{
            type: Output
        }], thyOnRadioSelectChange: [{
            type: Output
        }], thyOnDraggableChange: [{
            type: Output
        }], thyOnRowClick: [{
            type: Output
        }], thyOnRowContextMenu: [{
            type: Output
        }], isTableClass: [{
            type: HostBinding,
            args: ['class.thy-table']
        }], thyMode: [{
            type: Input
        }], thyGroupBy: [{
            type: Input
        }], thyRowKey: [{
            type: Input
        }], thyGroups: [{
            type: Input
        }], thyModel: [{
            type: Input
        }], thyTheme: [{
            type: Input
        }], thySize: [{
            type: Input
        }], thyClassName: [{
            type: Input
        }], thyRowClassName: [{
            type: Input
        }], thyLoadingDone: [{
            type: Input
        }], thyLoadingText: [{
            type: Input
        }], thyEmptyOptions: [{
            type: Input
        }], thyDraggable: [{
            type: Input
        }], thyFilter: [{
            type: Input
        }], thyPageIndex: [{
            type: Input
        }], thyPageSize: [{
            type: Input
        }], thyPageTotal: [{
            type: Input
        }], thyWholeRowSelect: [{
            type: Input
        }], listOfColumnComponents: [{
            type: ContentChildren,
            args: [ThyTableColumnComponent]
        }], emptyTemplate: [{
            type: ContentChild,
            args: ['empty']
        }], tableElementRef: [{
            type: ViewChild,
            args: ['table', { static: true }]
        }], thyHoverDisplayOperation: [{
            type: HostBinding,
            args: ['class.thy-table-hover-display-operation']
        }, {
            type: Input
        }], groupTemplate: [{
            type: ContentChild,
            args: ['group', { static: true }]
        }] }); })();
/**
 * @deprecated ThyGridComponent is deprecated, please use ThyTableComponent
 */
const ThyGridComponent = ThyTableComponent;

class TableIsValidModelValuePipe {
    transform(value) {
        return value !== '' && value !== undefined && value !== null;
    }
}
TableIsValidModelValuePipe.ɵfac = function TableIsValidModelValuePipe_Factory(t) { return new (t || TableIsValidModelValuePipe)(); };
TableIsValidModelValuePipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "isValidModelValue", type: TableIsValidModelValuePipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TableIsValidModelValuePipe, [{
        type: Pipe,
        args: [{ name: 'isValidModelValue' }]
    }], null, null); })();

class ThyTableModule {
}
ThyTableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTableModule });
ThyTableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTableModule_Factory(t) { return new (t || ThyTableModule)(); }, imports: [[
            CommonModule,
            FormsModule,
            ThyPaginationModule,
            ThySwitchModule,
            ThyLoadingModule,
            ThyEmptyModule,
            ThySharedModule,
            ThyIconModule,
            DragDropModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTableModule, { declarations: function () { return [ThyTableComponent, ThyTableColumnComponent, TableIsValidModelValuePipe]; }, imports: function () { return [CommonModule,
        FormsModule,
        ThyPaginationModule,
        ThySwitchModule,
        ThyLoadingModule,
        ThyEmptyModule,
        ThySharedModule,
        ThyIconModule,
        DragDropModule]; }, exports: function () { return [ThyTableComponent, ThyTableColumnComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTableModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyTableComponent, ThyTableColumnComponent, TableIsValidModelValuePipe],
                imports: [
                    CommonModule,
                    FormsModule,
                    ThyPaginationModule,
                    ThySwitchModule,
                    ThyLoadingModule,
                    ThyEmptyModule,
                    ThySharedModule,
                    ThyIconModule,
                    DragDropModule
                ],
                exports: [ThyTableComponent, ThyTableColumnComponent]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { THY_GRID_COLUMN_PARENT_COMPONENT, THY_TABLE_COLUMN_PARENT_COMPONENT, TableIsValidModelValuePipe, ThyGridColumnComponent, ThyGridComponent, ThyTableColumnComponent, ThyTableComponent, ThyTableModule };
