// ngx-tethys/fesm2015/ngx-tethys-layout.js
import { Component, HostBinding, Input, ContentChild, Optional, Host, Renderer2, ElementRef, NgZone, ViewChild, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { ThyIconModule } from 'ngx-tethys/icon';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/icon';
import * as ɵngcc3 from '@angular/cdk/drag-drop';

const _c0 = ["*"];
const _c1 = ["headerTitle"];
const _c2 = ["headerContent"];
const _c3 = ["headerOperation"];
function ThyHeaderComponent_ng_template_1_Template(rf, ctx) { }
function ThyHeaderComponent_ng_container_2_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 8);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r4.svgIconName);
} }
function ThyHeaderComponent_ng_container_2_ng_template_2_i_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 10);
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r7.iconClass);
} }
function ThyHeaderComponent_ng_container_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyHeaderComponent_ng_container_2_ng_template_2_i_0_Template, 1, 1, "i", 9);
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.iconClass);
} }
function ThyHeaderComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyHeaderComponent_ng_container_2_thy_icon_1_Template, 1, 1, "thy-icon", 5);
    ɵngcc0.ɵɵtemplate(2, ThyHeaderComponent_ng_container_2_ng_template_2_Template, 1, 1, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(4, "span", 7);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r5 = ɵngcc0.ɵɵreference(3);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.svgIconName)("ngIfElse", _r5);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.thyTitle);
} }
function ThyHeaderComponent_ng_template_4_Template(rf, ctx) { }
function ThyHeaderComponent_ng_template_6_Template(rf, ctx) { }
const _c4 = ["dragRef"];
class ThyLayoutComponent {
    constructor() {
        this.thyLayoutClass = true;
        this.hasSidebar = false;
    }
}
ThyLayoutComponent.ɵfac = function ThyLayoutComponent_Factory(t) { return new (t || ThyLayoutComponent)(); };
ThyLayoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyLayoutComponent, selectors: [["thy-layout"]], hostVars: 4, hostBindings: function ThyLayoutComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout", ctx.thyLayoutClass)("thy-layout--has-sidebar", ctx.hasSidebar);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyLayoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyLayoutComponent.propDecorators = {
    thyLayoutClass: [{ type: HostBinding, args: ['class.thy-layout',] }],
    hasSidebar: [{ type: HostBinding, args: ['class.thy-layout--has-sidebar',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLayoutComponent, [{
        type: Component,
        args: [{
                selector: 'thy-layout',
                template: `
    <ng-content></ng-content>
  `
            }]
    }], function () { return []; }, { thyLayoutClass: [{
            type: HostBinding,
            args: ['class.thy-layout']
        }], hasSidebar: [{
            type: HostBinding,
            args: ['class.thy-layout--has-sidebar']
        }] }); })();

class ThyHeaderComponent {
    constructor() {
        this.thyLayoutHeaderClass = true;
        this._thyHasBorder = false;
        this._thySizeSm = false;
        this.thyIconPrefix = 'wtf';
    }
    set thyHasBorder(value) {
        this._thyHasBorder = coerceBooleanProperty(value);
    }
    set thySize(value) {
        this._thySizeSm = value === 'sm';
    }
    set thyIcon(value) {
        if (value) {
            if (value.includes('wtf')) {
                this.iconClass = `${this.thyIconPrefix} ${value}`;
            }
            else {
                this.svgIconName = value;
            }
        }
        else {
            this.iconClass = null;
            this.svgIconName = null;
        }
    }
    ngOnInit() { }
}
ThyHeaderComponent.ɵfac = function ThyHeaderComponent_Factory(t) { return new (t || ThyHeaderComponent)(); };
ThyHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyHeaderComponent, selectors: [["thy-header"]], contentQueries: function ThyHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c3, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationTemplateRef = _t.first);
    } }, hostVars: 6, hostBindings: function ThyHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-header", ctx.thyLayoutHeaderClass)("header-has-border", ctx._thyHasBorder)("thy-layout-header-sm", ctx._thySizeSm);
    } }, inputs: { thyIconPrefix: "thyIconPrefix", thyHasBorder: "thyHasBorder", thySize: "thySize", thyIcon: "thyIcon", thyTitle: "thyTitle" }, decls: 7, vars: 4, consts: [[1, "layout-header-title"], [3, "ngTemplateOutlet"], [4, "ngIf"], [1, "layout-header-content"], [1, "layout-header-operation"], ["class", "prefix-icon", 3, "thyIconName", 4, "ngIf", "ngIfElse"], ["iconFont", ""], [1, "title-name"], [1, "prefix-icon", 3, "thyIconName"], ["class", "prefix-icon", 3, "ngClass", 4, "ngIf"], [1, "prefix-icon", 3, "ngClass"]], template: function ThyHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyHeaderComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(2, ThyHeaderComponent_ng_container_2_Template, 6, 3, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyHeaderComponent_ng_template_4_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵtemplate(6, ThyHeaderComponent_ng_template_6_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.contentTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.operationTemplateRef);
    } }, directives: [ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass], encapsulation: 2 });
ThyHeaderComponent.ctorParameters = () => [];
ThyHeaderComponent.propDecorators = {
    thyLayoutHeaderClass: [{ type: HostBinding, args: ['class.thy-layout-header',] }],
    _thyHasBorder: [{ type: HostBinding, args: ['class.header-has-border',] }],
    _thySizeSm: [{ type: HostBinding, args: ['class.thy-layout-header-sm',] }],
    thyHasBorder: [{ type: Input, args: ['thyHasBorder',] }],
    thySize: [{ type: Input, args: ['thySize',] }],
    thyTitle: [{ type: Input }],
    thyIconPrefix: [{ type: Input }],
    thyIcon: [{ type: Input, args: ['thyIcon',] }],
    titleTemplateRef: [{ type: ContentChild, args: ['headerTitle',] }],
    contentTemplateRef: [{ type: ContentChild, args: ['headerContent',] }],
    operationTemplateRef: [{ type: ContentChild, args: ['headerOperation',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-header',
                preserveWhitespaces: false,
                template: "<div class=\"layout-header-title\">\n  <ng-template [ngTemplateOutlet]=\"titleTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!titleTemplateRef\">\n    <thy-icon *ngIf=\"svgIconName; else iconFont\" class=\"prefix-icon\" [thyIconName]=\"svgIconName\"></thy-icon>\n    <ng-template #iconFont>\n      <i *ngIf=\"iconClass\" class=\"prefix-icon\" [ngClass]=\"iconClass\"></i>\n    </ng-template>\n    <span class=\"title-name\">{{ thyTitle }}</span></ng-container\n  >\n</div>\n<div class=\"layout-header-content\">\n  <ng-template [ngTemplateOutlet]=\"contentTemplateRef\"></ng-template>\n</div>\n<div class=\"layout-header-operation\">\n  <ng-template [ngTemplateOutlet]=\"operationTemplateRef\"></ng-template>\n</div>\n"
            }]
    }], function () { return []; }, { thyLayoutHeaderClass: [{
            type: HostBinding,
            args: ['class.thy-layout-header']
        }], _thyHasBorder: [{
            type: HostBinding,
            args: ['class.header-has-border']
        }], _thySizeSm: [{
            type: HostBinding,
            args: ['class.thy-layout-header-sm']
        }], thyIconPrefix: [{
            type: Input
        }], thyHasBorder: [{
            type: Input,
            args: ['thyHasBorder']
        }], thySize: [{
            type: Input,
            args: ['thySize']
        }], thyIcon: [{
            type: Input,
            args: ['thyIcon']
        }], thyTitle: [{
            type: Input
        }], titleTemplateRef: [{
            type: ContentChild,
            args: ['headerTitle']
        }], contentTemplateRef: [{
            type: ContentChild,
            args: ['headerContent']
        }], operationTemplateRef: [{
            type: ContentChild,
            args: ['headerOperation']
        }] }); })();

class ThyContentComponent {
    constructor() {
        this._isLayoutContent = true;
    }
}
ThyContentComponent.ɵfac = function ThyContentComponent_Factory(t) { return new (t || ThyContentComponent)(); };
ThyContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentComponent, selectors: [["thy-content"]], hostVars: 2, hostBindings: function ThyContentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content", ctx._isLayoutContent);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyContentComponent.propDecorators = {
    _isLayoutContent: [{ type: HostBinding, args: ['class.thy-layout-content',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentComponent, [{
        type: Component,
        args: [{
                selector: 'thy-content',
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { _isLayoutContent: [{
            type: HostBinding,
            args: ['class.thy-layout-content']
        }] }); })();

const LG_WIDTH = 300;
class ThySidebarComponent {
    constructor(thyLayoutComponent, renderer, elementRef, ngZone) {
        this.thyLayoutComponent = thyLayoutComponent;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.thyLayoutSidebarClass = true;
        this.thyLayoutSidebarClearBorderRightClass = false;
        this.sidebarIsolated = false;
    }
    set thyWidth(value) {
        if (value === 'lg') {
            value = LG_WIDTH;
        }
        this.thyLayoutSidebarWidth = value;
        this.widthPassive = value;
    }
    set thyHasBorderRight(value) {
        this.thyLayoutSidebarClearBorderRightClass = !coerceBooleanProperty(value);
    }
    set thyIsolated(value) {
        this.sidebarIsolated = coerceBooleanProperty(value);
    }
    set thyIsDraggableWidth(value) {
        if (coerceBooleanProperty(value)) {
            this.thyLayoutSidebarPointerEvents = 'all';
        }
    }
    ngOnInit() {
        if (this.thyLayoutComponent) {
            this.thyLayoutComponent.hasSidebar = true;
        }
        this.ngZone.runOutsideAngular(() => {
            setTimeout(() => {
                this.widthPassive = this.elementRef.nativeElement.clientWidth;
                this.renderer.setStyle(this.dragRef.nativeElement, 'left', this.numberConvertToFloor(this.widthPassive) + 'px');
            }, 0);
        });
    }
    dragStartedHandler() {
        this.dragStartedX = this.dragRef.nativeElement.getBoundingClientRect().x;
    }
    dragEndedHandler() {
        const x = this.dragRef.nativeElement.getBoundingClientRect().x;
        this.widthPassive = this.numberConvertToFloor(this.widthPassive * 1 + (x - this.dragStartedX));
        this.thyLayoutSidebarWidth = this.widthPassive;
    }
    numberConvertToFloor(value) {
        let result = Math.floor(value * 1);
        if (result < 1) {
            result = 1;
        }
        return result;
    }
}
ThySidebarComponent.ɵfac = function ThySidebarComponent_Factory(t) { return new (t || ThySidebarComponent)(ɵngcc0.ɵɵdirectiveInject(ThyLayoutComponent, 9), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThySidebarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySidebarComponent, selectors: [["thy-sidebar"]], viewQuery: function ThySidebarComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c4, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dragRef = _t.first);
    } }, hostVars: 10, hostBindings: function ThySidebarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", ctx.thyLayoutSidebarWidth, "px")("pointer-events", ctx.thyLayoutSidebarPointerEvents);
        ɵngcc0.ɵɵclassProp("thy-layout-sidebar", ctx.thyLayoutSidebarClass)("thy-layout-sidebar--clear-border-right", ctx.thyLayoutSidebarClearBorderRightClass)("thy-layout-sidebar-isolated", ctx.sidebarIsolated);
    } }, inputs: { thyWidth: "thyWidth", thyHasBorderRight: "thyHasBorderRight", thyIsolated: "thyIsolated", thyIsDraggableWidth: "thyIsDraggableWidth" }, ngContentSelectors: _c0, decls: 3, vars: 0, consts: [["cdkDrag", "", "cdkDragLockAxis", "x", "cdkDragBoundary", ".thy-layout", 1, "sidebar-drag", 3, "cdkDragStarted", "cdkDragEnded"], ["dragRef", ""]], template: function ThySidebarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0, 1);
        ɵngcc0.ɵɵlistener("cdkDragStarted", function ThySidebarComponent_Template_div_cdkDragStarted_1_listener() { return ctx.dragStartedHandler(); })("cdkDragEnded", function ThySidebarComponent_Template_div_cdkDragEnded_1_listener() { return ctx.dragEndedHandler(); });
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc3.CdkDrag], encapsulation: 2 });
ThySidebarComponent.ctorParameters = () => [
    { type: ThyLayoutComponent, decorators: [{ type: Optional }, { type: Host }] },
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone }
];
ThySidebarComponent.propDecorators = {
    thyLayoutSidebarClass: [{ type: HostBinding, args: ['class.thy-layout-sidebar',] }],
    thyLayoutSidebarClearBorderRightClass: [{ type: HostBinding, args: ['class.thy-layout-sidebar--clear-border-right',] }],
    thyLayoutSidebarWidth: [{ type: HostBinding, args: ['style.width.px',] }],
    thyLayoutSidebarPointerEvents: [{ type: HostBinding, args: ['style.pointer-events',] }],
    sidebarIsolated: [{ type: HostBinding, args: ['class.thy-layout-sidebar-isolated',] }],
    thyWidth: [{ type: Input, args: ['thyWidth',] }],
    thyHasBorderRight: [{ type: Input, args: ['thyHasBorderRight',] }],
    thyIsolated: [{ type: Input, args: ['thyIsolated',] }],
    thyIsDraggableWidth: [{ type: Input, args: ['thyIsDraggableWidth',] }],
    dragRef: [{ type: ViewChild, args: ['dragRef', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySidebarComponent, [{
        type: Component,
        args: [{
                selector: 'thy-sidebar',
                preserveWhitespaces: false,
                template: `
        <ng-content></ng-content>
        <div
            class="sidebar-drag"
            #dragRef
            cdkDrag
            cdkDragLockAxis="x"
            cdkDragBoundary=".thy-layout"
            (cdkDragStarted)="dragStartedHandler()"
            (cdkDragEnded)="dragEndedHandler()"
        ></div>
    `
            }]
    }], function () { return [{ type: ThyLayoutComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }] }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }]; }, { thyLayoutSidebarClass: [{
            type: HostBinding,
            args: ['class.thy-layout-sidebar']
        }], thyLayoutSidebarClearBorderRightClass: [{
            type: HostBinding,
            args: ['class.thy-layout-sidebar--clear-border-right']
        }], sidebarIsolated: [{
            type: HostBinding,
            args: ['class.thy-layout-sidebar-isolated']
        }], thyWidth: [{
            type: Input,
            args: ['thyWidth']
        }], thyLayoutSidebarWidth: [{
            type: HostBinding,
            args: ['style.width.px']
        }], thyHasBorderRight: [{
            type: Input,
            args: ['thyHasBorderRight']
        }], thyIsolated: [{
            type: Input,
            args: ['thyIsolated']
        }], thyIsDraggableWidth: [{
            type: Input,
            args: ['thyIsDraggableWidth']
        }], thyLayoutSidebarPointerEvents: [{
            type: HostBinding,
            args: ['style.pointer-events']
        }], dragRef: [{
            type: ViewChild,
            args: ['dragRef', { static: true }]
        }] }); })();

class ThyContentSectionComponent {
    constructor() {
        this.thyLayoutContentSectionClass = true;
    }
}
ThyContentSectionComponent.ɵfac = function ThyContentSectionComponent_Factory(t) { return new (t || ThyContentSectionComponent)(); };
ThyContentSectionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentSectionComponent, selectors: [["thy-content-section"]], hostVars: 2, hostBindings: function ThyContentSectionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content-section", ctx.thyLayoutContentSectionClass);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentSectionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyContentSectionComponent.propDecorators = {
    thyLayoutContentSectionClass: [{ type: HostBinding, args: ['class.thy-layout-content-section',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentSectionComponent, [{
        type: Component,
        args: [{
                selector: 'thy-content-section',
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { thyLayoutContentSectionClass: [{
            type: HostBinding,
            args: ['class.thy-layout-content-section']
        }] }); })();

class ThyContentMainComponent {
    constructor() {
        this.thyLayoutContentClass = true;
    }
}
ThyContentMainComponent.ɵfac = function ThyContentMainComponent_Factory(t) { return new (t || ThyContentMainComponent)(); };
ThyContentMainComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentMainComponent, selectors: [["thy-content-main"]], hostVars: 2, hostBindings: function ThyContentMainComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content-main", ctx.thyLayoutContentClass);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentMainComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyContentMainComponent.propDecorators = {
    thyLayoutContentClass: [{ type: HostBinding, args: ['class.thy-layout-content-main',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentMainComponent, [{
        type: Component,
        args: [{
                selector: 'thy-content-main',
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { thyLayoutContentClass: [{
            type: HostBinding,
            args: ['class.thy-layout-content-main']
        }] }); })();

class ThyLayoutModule {
}
ThyLayoutModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLayoutModule });
ThyLayoutModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLayoutModule_Factory(t) { return new (t || ThyLayoutModule)(); }, imports: [[CommonModule, DragDropModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLayoutModule, { declarations: function () { return [ThyLayoutComponent, ThyHeaderComponent, ThyContentComponent, ThySidebarComponent, ThyContentSectionComponent, ThyContentMainComponent]; }, imports: function () { return [CommonModule, DragDropModule, ThyIconModule]; }, exports: function () { return [ThyLayoutComponent, ThyHeaderComponent, ThyContentComponent, ThySidebarComponent, ThyContentSectionComponent, ThyContentMainComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLayoutModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyLayoutComponent,
                    ThyHeaderComponent,
                    ThyContentComponent,
                    ThySidebarComponent,
                    ThyContentSectionComponent,
                    ThyContentMainComponent
                ],
                imports: [CommonModule, DragDropModule, ThyIconModule],
                exports: [
                    ThyLayoutComponent,
                    ThyHeaderComponent,
                    ThyContentComponent,
                    ThySidebarComponent,
                    ThyContentSectionComponent,
                    ThyContentMainComponent
                ]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyContentComponent, ThyContentMainComponent, ThyContentSectionComponent, ThyHeaderComponent, ThyLayoutComponent, ThyLayoutModule, ThySidebarComponent };
