// ngx-tethys/fesm2015/ngx-tethys-copy.js
import { EventEmitter, Directive, Inject, ElementRef, ViewContainerRef, Output, Input, HostListener, NgModule } from '@angular/core';
import { DOCUMENT, CommonModule } from '@angular/common';
import { coerceElement } from '@angular/cdk/coercion';
import { TooltipService } from 'ngx-tethys/tooltip';
import { ThyNotifyService } from 'ngx-tethys/notify';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/tooltip';
import * as ɵngcc2 from 'ngx-tethys/notify';
class ThyCopyDirective {
    constructor(document, tooltipService, elementRef, viewContainerRef, notifyService) {
        this.document = document;
        this.tooltipService = tooltipService;
        this.elementRef = elementRef;
        this.viewContainerRef = viewContainerRef;
        this.notifyService = notifyService;
        // 默认为点击标签，可传复制目标标签
        this.thyCopy = new EventEmitter();
        this.thyCopySuccessText = '复制成功';
        this.thyCopyTips = '点击复制';
    }
    ngOnInit() {
        this.tooltipService.attach(this.elementRef, this.viewContainerRef, 'hover');
        this.tooltipService.thyTooltipDirective.content = this.thyCopyTips ? this.thyCopyTips : '点击复制';
    }
    getContent(event) {
        if (typeof this.thyCopyContent === 'string') {
            return this.thyCopyContent;
        }
        else {
            const target = this.thyCopyContent ? coerceElement(this.thyCopyContent) : event.target;
            return target.value || target.textContent;
        }
    }
    onClick(event) {
        const textarea = this.document.createElement('textarea');
        this.document.body.appendChild(textarea);
        textarea.value = this.getContent(event);
        textarea.select();
        try {
            document.execCommand('copy', false, null);
            this.thyCopy.emit({ isSuccess: true, event });
            this.notifyService.success(this.thyCopySuccessText);
        }
        catch (err) {
            this.thyCopy.emit({ isSuccess: false, event });
            this.notifyService.error('复制失败');
        }
        finally {
            textarea.remove();
        }
    }
    ngOnDestroy() {
        this.tooltipService.detach();
    }
}
ThyCopyDirective.ɵfac = function ThyCopyDirective_Factory(t) { return new (t || ThyCopyDirective)(ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TooltipService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyNotifyService)); };
ThyCopyDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyCopyDirective, selectors: [["", "thyCopy", ""]], hostBindings: function ThyCopyDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ThyCopyDirective_click_HostBindingHandler($event) { return ctx.onClick($event); });
    } }, inputs: { thyCopySuccessText: "thyCopySuccessText", thyCopyTips: "thyCopyTips", thyCopyContent: "thyCopyContent" }, outputs: { thyCopy: "thyCopy" }, features: [ɵngcc0.ɵɵProvidersFeature([TooltipService, ThyNotifyService])] });
ThyCopyDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: TooltipService },
    { type: ElementRef },
    { type: ViewContainerRef },
    { type: ThyNotifyService }
];
ThyCopyDirective.propDecorators = {
    thyCopy: [{ type: Output }],
    thyCopySuccessText: [{ type: Input }],
    thyCopyTips: [{ type: Input }],
    thyCopyContent: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCopyDirective, [{
        type: Directive,
        args: [{
                selector: '[thyCopy]',
                providers: [TooltipService, ThyNotifyService]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc1.TooltipService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc2.ThyNotifyService }]; }, { thyCopy: [{
            type: Output
        }], thyCopySuccessText: [{
            type: Input
        }], thyCopyTips: [{
            type: Input
        }], onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }], thyCopyContent: [{
            type: Input
        }] }); })();

class ThyCopyModule {
}
ThyCopyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyCopyModule });
ThyCopyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyCopyModule_Factory(t) { return new (t || ThyCopyModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyCopyModule, { declarations: function () { return [ThyCopyDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyCopyDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyCopyModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyCopyDirective],
                imports: [CommonModule],
                exports: [ThyCopyDirective]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ThyCopyDirective, ThyCopyModule };
