// ngx-tethys/esm2015/shared/directives/thy-contextmenu.directive.js
import { Directive, Output, ElementRef, EventEmitter, NgZone, Renderer2 } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ThyContextMenuDirective {
    constructor(ngZone, elementRef, renderer) {
        this.ngZone = ngZone;
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.thyContextMenu = new EventEmitter();
    }
    rightClick(event) {
        event.preventDefault();
        this.ngZone.run(() => {
            this.thyContextMenu.emit(event);
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this.renderer.listen(this.elementRef.nativeElement, 'contextmenu', this.rightClick.bind(this));
        });
    }
}
ThyContextMenuDirective.ɵfac = function ThyContextMenuDirective_Factory(t) { return new (t || ThyContextMenuDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyContextMenuDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyContextMenuDirective, selectors: [["", "thyContextMenu", ""]], outputs: { thyContextMenu: "thyContextMenu" } });
ThyContextMenuDirective.ctorParameters = () => [
    { type: NgZone },
    { type: ElementRef },
    { type: Renderer2 }
];
ThyContextMenuDirective.propDecorators = {
    thyContextMenu: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContextMenuDirective, [{
        type: Directive,
        args: [{
                selector: '[thyContextMenu]'
            }]
    }], function () { return [{ type: ɵngcc0.NgZone }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { thyContextMenu: [{
            type: Output
        }] }); })();
