// ngx-tethys/esm2015/shared/directives/thy-stop-propagation.directive.js
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
/**
 * 将来会移动到 thy 组件库中
 */
import * as ɵngcc0 from '@angular/core';
export class ThyStopPropagationDirective {
    constructor(_elementRef, _renderer) {
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this._eventName = 'click';
        this.isStopPropagation = true;
    }
    set thyStopPropagation(value) {
        if (value === false || value === 'false') {
            this.isStopPropagation = false;
        }
        else {
            this.isStopPropagation = true;
            this._eventName = value || 'click';
        }
    }
    ngOnInit() {
        if (this.isStopPropagation) {
            this._listener = this._renderer.listen(this._elementRef.nativeElement, this._eventName, ($event) => {
                $event.stopPropagation();
            });
        }
    }
    ngOnDestroy() {
        if (this._listener) {
            this._listener();
        }
    }
}
ThyStopPropagationDirective.ɵfac = function ThyStopPropagationDirective_Factory(t) { return new (t || ThyStopPropagationDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ThyStopPropagationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyStopPropagationDirective, selectors: [["", "thyStopPropagation", ""]], inputs: { thyStopPropagation: "thyStopPropagation" } });
ThyStopPropagationDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ThyStopPropagationDirective.propDecorators = {
    thyStopPropagation: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyStopPropagationDirective, [{
        type: Directive,
        args: [{
                selector: '[thyStopPropagation]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { thyStopPropagation: [{
            type: Input
        }] }); })();
