// ngx-tethys/esm2015/util/immutable.js
import { coerceArray, isFunction, isUndefinedOrNull } from './helpers';
export class Producer {
    constructor(entities, options) {
        this.idKey = '_id';
        this.entities = entities || [];
        if (options && options.idKey) {
            this.idKey = options.idKey;
        }
    }
    /**
     * Add an entity or entities.
     *
     * @example
     * produce([users]).add(Entity);
     * produce([users]).add([Entity, Entity]);
     * produce([users]).add(Entity, { prepend: true });
     * produce([users]).add(Entity, { afterId: '' });
     */
    add(entity, addOptions) {
        const addEntities = coerceArray(entity);
        if (addEntities.length === 0) {
            return this.entities;
        }
        if (addOptions && (addOptions.afterId || addOptions.prepend)) {
            if (addOptions.afterId) {
                const entities = [...this.entities];
                const index = this.entities.findIndex(item => {
                    return item[this.idKey] === addOptions.afterId;
                }) + 1;
                entities.splice(index, 0, ...addEntities);
                this.entities = [...entities];
            }
            else if (addOptions.prepend) {
                this.entities = [...addEntities, ...this.entities];
            }
        }
        else {
            this.entities = [...this.entities, ...addEntities];
        }
        return this.entities;
    }
    update(idsOrFn, newStateOrFn) {
        const ids = coerceArray(idsOrFn);
        for (let i = 0; i < this.entities.length; i++) {
            const oldEntity = this.entities[i];
            if (ids.indexOf(oldEntity[this.idKey]) >= 0) {
                const newState = isFunction(newStateOrFn) ? newStateOrFn(oldEntity) : newStateOrFn;
                this.entities[i] = Object.assign(Object.assign({}, oldEntity), newState);
            }
        }
        return [...this.entities];
    }
    remove(idsOrFn) {
        if (isFunction(idsOrFn)) {
            this.entities = this.entities.filter(entity => {
                return !idsOrFn(entity);
            });
        }
        else {
            const ids = coerceArray(idsOrFn);
            this.entities = this.entities.filter(entity => {
                return ids.indexOf(entity[this.idKey]) === -1;
            });
        }
        return this.entities;
    }
    /**
     *
     * Move one or more entities:
     *
     * @example
     * produce([users]).move(5, {afterId: 2});
     * produce([users]).move(5, {toIndex: 0});
     */
    move(id, moveOptions) {
        const fromIndex = this.entities.findIndex(item => item[this.idKey] === id);
        let toIndex = 0;
        const newEntities = [...this.entities];
        if (!id || fromIndex < 0) {
            return [...this.entities];
        }
        if (moveOptions) {
            if (!isUndefinedOrNull(moveOptions.afterId)) {
                toIndex = this.entities.findIndex(item => item[this.idKey] === moveOptions.afterId);
            }
            else if (moveOptions.toIndex) {
                toIndex = moveOptions.toIndex;
            }
        }
        toIndex = Math.max(0, Math.min(this.entities.length - 1, toIndex));
        if (toIndex === fromIndex) {
            return [...this.entities];
        }
        else {
            const target = this.entities[fromIndex];
            const delta = toIndex < fromIndex ? -1 : 1;
            for (let i = fromIndex; i !== toIndex; i += delta) {
                newEntities[i] = newEntities[i + delta];
            }
            newEntities[toIndex] = target;
            return [...newEntities];
        }
    }
}
export function produce(entities, options) {
    return new Producer(entities, options);
}
