// ngx-tethys/esm2015/menu/divider/menu-divider.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyMenuDividerComponent {
    constructor() {
        this.isThyMenuDivider = true;
    }
    ngOnInit() { }
}
ThyMenuDividerComponent.ɵfac = function ThyMenuDividerComponent_Factory(t) { return new (t || ThyMenuDividerComponent)(); };
ThyMenuDividerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuDividerComponent, selectors: [["thy-menu-divider"], ["", "thy-menu-divider", ""], ["", "thyMenuDivider", ""]], hostVars: 2, hostBindings: function ThyMenuDividerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-divider", ctx.isThyMenuDivider);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuDividerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuDividerComponent.ctorParameters = () => [];
ThyMenuDividerComponent.propDecorators = {
    isThyMenuDivider: [{ type: HostBinding, args: ['class.thy-menu-divider',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuDividerComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-divider,[thy-menu-divider],[thyMenuDivider]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenuDivider: [{
            type: HostBinding,
            args: ['class.thy-menu-divider']
        }] }); })();
