// ngx-tethys/esm2015/slide/slide.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThySlideService } from './slide.service';
import { ThySlideContainerComponent } from './slide-container.component';
import { ThyInternalSlideRef } from './slide-ref.service';
import { ThySlideLayoutComponent } from './slide-layout/slide-layout.component';
import { ThySlideHeaderComponent } from './slide-header/slide-header.component';
import { ThySlideBodyComponent } from './slide-body/slide-body.component';
import { ThySlideBodySectionComponent } from './slide-body/slide-body-section.component';
import { ThySlideFooterComponent } from './slide-footer/slide-footer.component';
import { ThyIconModule } from 'ngx-tethys/icon';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { THY_SLIDE_DEFAULT_CONFIG_PROVIDER } from './slide.config';
import { ThyDrawerContainerDirective } from './drawer/drawer-container.directive';
import * as ɵngcc0 from '@angular/core';
export class ThySlideModule {
}
ThySlideModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySlideModule });
ThySlideModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySlideModule_Factory(t) { return new (t || ThySlideModule)(); }, providers: [ThyInternalSlideRef, ThySlideService, THY_SLIDE_DEFAULT_CONFIG_PROVIDER], imports: [[CommonModule, ThySharedModule, ThyIconModule, OverlayModule, PortalModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySlideModule, { declarations: function () { return [ThySlideContainerComponent,
        ThySlideLayoutComponent,
        ThySlideHeaderComponent,
        ThySlideBodyComponent,
        ThySlideBodySectionComponent,
        ThySlideFooterComponent,
        ThyDrawerContainerDirective]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule, OverlayModule, PortalModule]; }, exports: function () { return [ThySlideLayoutComponent,
        ThySlideHeaderComponent,
        ThySlideBodyComponent,
        ThySlideBodySectionComponent,
        ThySlideFooterComponent,
        ThyDrawerContainerDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySlideModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThySlideContainerComponent,
                    ThySlideLayoutComponent,
                    ThySlideHeaderComponent,
                    ThySlideBodyComponent,
                    ThySlideBodySectionComponent,
                    ThySlideFooterComponent,
                    ThyDrawerContainerDirective
                ],
                entryComponents: [ThySlideContainerComponent],
                imports: [CommonModule, ThySharedModule, ThyIconModule, OverlayModule, PortalModule],
                exports: [
                    ThySlideLayoutComponent,
                    ThySlideHeaderComponent,
                    ThySlideBodyComponent,
                    ThySlideBodySectionComponent,
                    ThySlideFooterComponent,
                    ThyDrawerContainerDirective
                ],
                providers: [ThyInternalSlideRef, ThySlideService, THY_SLIDE_DEFAULT_CONFIG_PROVIDER]
            }]
    }], null, null); })();
