// ngx-tethys/esm2015/nav/nav-link.directive.js
import { Directive, Input, HostBinding } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
export class ThyNavLinkDirective {
    constructor() {
        this.navLinkActive = false;
        this.navLinkClass = true;
        // @HostBinding('attr.href') navLinkHref = 'javascript:;';
    }
    set thyNavLinkActive(active) {
        this.navLinkActive = coerceBooleanProperty(active);
    }
}
ThyNavLinkDirective.ɵfac = function ThyNavLinkDirective_Factory(t) { return new (t || ThyNavLinkDirective)(); };
ThyNavLinkDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThyNavLinkDirective, selectors: [["", "thyNavLink", ""]], hostVars: 4, hostBindings: function ThyNavLinkDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("active", ctx.navLinkActive)("nav-link", ctx.navLinkClass);
    } }, inputs: { thyNavLinkActive: "thyNavLinkActive" } });
ThyNavLinkDirective.propDecorators = {
    thyNavLinkActive: [{ type: Input }],
    navLinkActive: [{ type: HostBinding, args: ['class.active',] }],
    navLinkClass: [{ type: HostBinding, args: ['class.nav-link',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNavLinkDirective, [{
        type: Directive,
        args: [{
                selector: '[thyNavLink]'
            }]
    }], function () { return []; }, { navLinkActive: [{
            type: HostBinding,
            args: ['class.active']
        }], navLinkClass: [{
            type: HostBinding,
            args: ['class.nav-link']
        }], thyNavLinkActive: [{
            type: Input
        }] }); })();
