// ngx-tethys/esm2015/nav/nav.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyNavComponent } from './nav.component';
import { ThyNavLinkDirective } from './nav-link.directive';
import { ThyIconNavComponent } from './icon-nav/icon-nav.component';
import { ThyIconNavLinkComponent } from './icon-nav/icon-nav-link.directive';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyNavModule {
}
ThyNavModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyNavModule });
ThyNavModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyNavModule_Factory(t) { return new (t || ThyNavModule)(); }, imports: [[CommonModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyNavModule, { declarations: function () { return [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent]; }, imports: function () { return [CommonModule, ThyIconModule]; }, exports: function () { return [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyNavModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent],
                imports: [CommonModule, ThyIconModule],
                exports: [ThyNavComponent, ThyNavLinkDirective, ThyIconNavComponent, ThyIconNavLinkComponent]
            }]
    }], null, null); })();
