// ngx-tethys/esm2015/popover/module.js
import { ThyIconModule } from 'ngx-tethys/icon';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThyPopoverBodyComponent } from './body/popover-body.component';
import { ThyPopoverHeaderComponent } from './header/popover-header.component';
import { ThyPopoverContainerComponent } from './popover-container.component';
import { THY_POPOVER_DEFAULT_CONFIG_PROVIDER, THY_POPOVER_SCROLL_STRATEGY_PROVIDER } from './popover.config';
import { ThyPopoverDirective } from './popover.directive';
import * as ɵngcc0 from '@angular/core';
export class ThyPopoverModule {
}
ThyPopoverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyPopoverModule });
ThyPopoverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyPopoverModule_Factory(t) { return new (t || ThyPopoverModule)(); }, providers: [THY_POPOVER_DEFAULT_CONFIG_PROVIDER, THY_POPOVER_SCROLL_STRATEGY_PROVIDER], imports: [[CommonModule, OverlayModule, PortalModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyPopoverModule, { declarations: function () { return [ThyPopoverContainerComponent, ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent]; }, imports: function () { return [CommonModule, OverlayModule, PortalModule, ThyIconModule]; }, exports: function () { return [ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPopoverModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyPopoverContainerComponent, ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent],
                entryComponents: [ThyPopoverContainerComponent],
                imports: [CommonModule, OverlayModule, PortalModule, ThyIconModule],
                exports: [ThyPopoverDirective, ThyPopoverHeaderComponent, ThyPopoverBodyComponent],
                providers: [THY_POPOVER_DEFAULT_CONFIG_PROVIDER, THY_POPOVER_SCROLL_STRATEGY_PROVIDER]
            }]
    }], null, null); })();
