// ngx-tethys/esm2015/popover/popover.config.js
import { ThyAbstractOverlayConfig } from 'ngx-tethys/core';
import { Overlay } from '@angular/cdk/overlay';
import { InjectionToken } from '@angular/core';
export class ThyPopoverConfig extends ThyAbstractOverlayConfig {
}
export const THY_POPOVER_DEFAULT_CONFIG = new InjectionToken('thy-popover-default-config');
export const THY_POPOVER_DEFAULT_CONFIG_VALUE = {
    hasBackdrop: true,
    backdropClass: 'thy-popover-backdrop',
    panelClass: '',
    offset: 4,
    backdropClosable: true,
    closeOnNavigation: true,
    placement: 'bottom',
    insideClosable: false,
    manualClosure: false,
    originActiveClass: 'thy-popover-origin-active'
};
export const THY_POPOVER_DEFAULT_CONFIG_PROVIDER = {
    provide: THY_POPOVER_DEFAULT_CONFIG,
    useValue: THY_POPOVER_DEFAULT_CONFIG_VALUE
};
export const THY_POPOVER_SCROLL_STRATEGY = new InjectionToken('thy-popover-scroll-strategy');
export const THY_POPOVER_SCROLL_STRATEGY_FACTORY = (overlay) => {
    return () => overlay.scrollStrategies.block();
};
export const THY_POPOVER_SCROLL_STRATEGY_PROVIDER = {
    provide: THY_POPOVER_SCROLL_STRATEGY,
    deps: [Overlay],
    useFactory: THY_POPOVER_SCROLL_STRATEGY_FACTORY
};
