// ngx-tethys/esm2015/guider/guider-hint/guider-hint.component.js
import { HostBinding, Component } from '@angular/core';
import { helpers } from 'ngx-tethys/util';
import { defaultGuiderPositionConfig } from '../guider.class';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function ThyGuiderHintComponent_ng_container_8_ng_template_1_Template(rf, ctx) { }
function ThyGuiderHintComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyGuiderHintComponent_ng_container_8_ng_template_1_Template, 0, 0, "ng-template", 14);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.descriptionTemplateRef);
} }
function ThyGuiderHintComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.descriptionString);
} }
const _c0 = function (a0) { return { active: a0 }; };
function ThyGuiderHintComponent_span_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 15);
} if (rf & 2) {
    const i_r10 = ctx.index;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(1, _c0, i_r10 === ctx_r3.stepRef.stepIndex));
} }
function ThyGuiderHintComponent_ng_container_15_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 16);
    ɵngcc0.ɵɵlistener("click", function ThyGuiderHintComponent_ng_container_15_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.prev(); });
    ɵngcc0.ɵɵtext(2, "\u4E0A\u4E00\u6B65");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyGuiderHintComponent_ng_container_16_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 17);
    ɵngcc0.ɵɵlistener("click", function ThyGuiderHintComponent_ng_container_16_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.next(); });
    ɵngcc0.ɵɵtext(2, "\u4E0B\u4E00\u6B65");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ThyGuiderHintComponent_ng_template_17_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 17);
    ɵngcc0.ɵɵlistener("click", function ThyGuiderHintComponent_ng_template_17_Template_a_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.end(); });
    ɵngcc0.ɵɵtext(1, "\u5B8C\u6210");
    ɵngcc0.ɵɵelementEnd();
} }
export class ThyGuiderHintComponent {
    constructor() {
        this.guiderHint = true;
    }
    ngOnInit() {
        this.setDescription(this.stepRef.step.data.description);
    }
    setDescription(value) {
        if (helpers.isString(value)) {
            this.descriptionString = value;
        }
        else {
            this.descriptionTemplateRef = value;
        }
    }
    jump() {
        this.guiderRef.end();
    }
    prev() {
        this.guiderRef.previous();
    }
    next() {
        this.guiderRef.next();
    }
    end() {
        this.guiderRef.end();
    }
    trackByFn(index, step) {
        return step.key || index;
    }
}
ThyGuiderHintComponent.ɵfac = function ThyGuiderHintComponent_Factory(t) { return new (t || ThyGuiderHintComponent)(); };
ThyGuiderHintComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyGuiderHintComponent, selectors: [["thy-guider-hint"]], hostVars: 2, hostBindings: function ThyGuiderHintComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-guider-tip-container", ctx.guiderHint);
    } }, decls: 19, vars: 10, consts: [[1, "thy-guider-tip-top"], [3, "src", "alt"], ["href", "javascript:;", 1, "thy-guider-tip-top-btn", 3, "click"], [1, "thy-guider-tip-content", "bg-white"], [1, "thy-guider-tip-content-title"], [1, "thy-guider-tip-content-description"], [4, "ngIf", "ngIfElse"], ["stringTemplateRef", ""], [1, "thy-guider-tip-step"], [1, "thy-guider-tip-step-pagination"], ["class", "pagination-item", 3, "ngClass", 4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "thy-guider-tip-step-btn"], [4, "ngIf"], ["finish", ""], [3, "ngTemplateOutlet"], [1, "pagination-item", 3, "ngClass"], ["href", "javascript:;", 1, "thy-guider-tip-step-pre", 3, "click"], ["href", "javascript:;", 1, "thy-guider-tip-step-next", 3, "click"]], template: function ThyGuiderHintComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelement(1, "img", 1);
        ɵngcc0.ɵɵelementStart(2, "a", 2);
        ɵngcc0.ɵɵlistener("click", function ThyGuiderHintComponent_Template_a_click_2_listener() { return ctx.jump(); });
        ɵngcc0.ɵɵtext(3, "\u8DF3\u8FC7");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 5);
        ɵngcc0.ɵɵtemplate(8, ThyGuiderHintComponent_ng_container_8_Template, 2, 1, "ng-container", 6);
        ɵngcc0.ɵɵtemplate(9, ThyGuiderHintComponent_ng_template_9_Template, 1, 1, "ng-template", null, 7, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 8);
        ɵngcc0.ɵɵelementStart(12, "div", 9);
        ɵngcc0.ɵɵtemplate(13, ThyGuiderHintComponent_span_13_Template, 1, 3, "span", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 11);
        ɵngcc0.ɵɵtemplate(15, ThyGuiderHintComponent_ng_container_15_Template, 3, 0, "ng-container", 12);
        ɵngcc0.ɵɵtemplate(16, ThyGuiderHintComponent_ng_container_16_Template, 3, 0, "ng-container", 6);
        ɵngcc0.ɵɵtemplate(17, ThyGuiderHintComponent_ng_template_17_Template, 2, 0, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(10);
        const _r6 = ɵngcc0.ɵɵreference(18);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("src", ctx.stepRef.step.data.image, ɵngcc0.ɵɵsanitizeUrl)("alt", ctx.stepRef.step.data.title);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ctx.stepRef.step.data.title);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.descriptionTemplateRef)("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.guiderRef.steps)("ngForTrackBy", ctx.trackByFn);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepRef.stepIndex !== 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepRef.stepIndex !== ctx.guiderRef.steps.length - 1)("ngIfElse", _r6);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgClass], encapsulation: 2 });
ThyGuiderHintComponent.ctorParameters = () => [];
ThyGuiderHintComponent.propDecorators = {
    guiderHint: [{ type: HostBinding, args: ['class.thy-guider-tip-container',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyGuiderHintComponent, [{
        type: Component,
        args: [{
                selector: 'thy-guider-hint',
                template: "<div class=\"thy-guider-tip-top\">\n  <img [src]=\"stepRef.step.data.image\" [alt]=\"stepRef.step.data.title\" />\n  <a href=\"javascript:;\" class=\"thy-guider-tip-top-btn\" (click)=\"jump()\">\u8DF3\u8FC7</a>\n</div>\n<div class=\"thy-guider-tip-content bg-white\">\n  <div class=\"thy-guider-tip-content-title\">{{ stepRef.step.data.title }}</div>\n  <div class=\"thy-guider-tip-content-description\">\n    <ng-container *ngIf=\"descriptionTemplateRef; else stringTemplateRef\">\n      <ng-template [ngTemplateOutlet]=\"descriptionTemplateRef\"></ng-template>\n    </ng-container>\n    <ng-template #stringTemplateRef>{{ descriptionString }}</ng-template>\n  </div>\n  <div class=\"thy-guider-tip-step\">\n    <div class=\"thy-guider-tip-step-pagination\">\n      <span\n        class=\"pagination-item\"\n        *ngFor=\"let step of guiderRef.steps; let i = index; trackBy: trackByFn\"\n        [ngClass]=\"{ active: i === stepRef.stepIndex }\"\n      ></span>\n    </div>\n    <div class=\"thy-guider-tip-step-btn\">\n      <ng-container *ngIf=\"stepRef.stepIndex !== 0\">\n        <a href=\"javascript:;\" (click)=\"prev()\" class=\"thy-guider-tip-step-pre\">\u4E0A\u4E00\u6B65</a>\n      </ng-container>\n      <ng-container *ngIf=\"stepRef.stepIndex !== guiderRef.steps.length - 1; else finish\">\n        <a href=\"javascript:;\" (click)=\"next()\" class=\"thy-guider-tip-step-next\">\u4E0B\u4E00\u6B65</a>\n      </ng-container>\n      <ng-template #finish>\n        <a href=\"javascript:;\" (click)=\"end()\" class=\"thy-guider-tip-step-next\">\u5B8C\u6210</a>\n      </ng-template>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { guiderHint: [{
            type: HostBinding,
            args: ['class.thy-guider-tip-container']
        }] }); })();
defaultGuiderPositionConfig.hintComponent = ThyGuiderHintComponent;
