// ngx-tethys/esm2015/radio/module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ThyRadioComponent } from './radio.component';
import { ThyRadioButtonComponent } from './button/radio-button.component';
import { ThyRadioGroupComponent } from './group/radio-group.component';
import * as ɵngcc0 from '@angular/core';
export class ThyRadioModule {
}
ThyRadioModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyRadioModule });
ThyRadioModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyRadioModule_Factory(t) { return new (t || ThyRadioModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyRadioModule, { declarations: function () { return [ThyRadioComponent,
        ThyRadioGroupComponent,
        ThyRadioButtonComponent]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [ThyRadioComponent,
        ThyRadioGroupComponent,
        ThyRadioButtonComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyRadioModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                declarations: [
                    ThyRadioComponent,
                    ThyRadioGroupComponent,
                    ThyRadioButtonComponent
                ],
                exports: [
                    ThyRadioComponent,
                    ThyRadioGroupComponent,
                    ThyRadioButtonComponent
                ]
            }]
    }], null, null); })();
