// ngx-tethys/esm2015/dialog/header/dialog-header.component.js
import { Component, Input, Output, EventEmitter, ContentChild, TemplateRef, HostBinding, Optional, ElementRef } from '@angular/core';
import { ThyDialog } from '../dialog.service';
import { ThyDialogContainerComponent } from '../dialog-container.component';
import { ThyTranslate } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../dialog.service';
import * as ɵngcc2 from 'ngx-tethys/core';
import * as ɵngcc3 from '../dialog-container.component';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from 'ngx-tethys/icon';

const _c0 = ["dialogHeader"];
function DialogHeaderComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "template", 2);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.headerTemplate);
} }
function DialogHeaderComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "h3", 3);
    ɵngcc0.ɵɵelement(1, "thy-icon", 4);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "button", 5);
    ɵngcc0.ɵɵlistener("click", function DialogHeaderComponent_ng_template_1_Template_button_click_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.close($event); });
    ɵngcc0.ɵɵelement(4, "thy-icon", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r2.thyIcon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.thyTitle);
} }
export class DialogHeaderComponent {
    constructor(elementRef, dialog, translate, dialogContainer) {
        this.elementRef = elementRef;
        this.dialog = dialog;
        this.translate = translate;
        this.dialogContainer = dialogContainer;
        this.isDialogHeader = true;
        this.isDialogHeaderLg = false;
        this.thyOnClose = new EventEmitter();
    }
    set thySize(value) {
        this.isDialogHeaderLg = value === 'lg';
    }
    set thyTitleTranslationKey(key) {
        if (key && !this.thyTitle) {
            this.thyTitle = this.translate.instant(key);
        }
    }
    ngOnInit() {
        if (!this.dialogContainer) {
            // When this directive is included in a dialog via TemplateRef (rather than being
            // in a Component), the ThyDialogContainerComponent isn't available via injection because embedded
            // views cannot be given a custom injector. Instead, we look up the ThyDialogContainerComponent by
            // ID. This must occur in `onInit`, as the ID binding for the dialog container won't
            // be resolved at constructor time.
            const dialogRef = this.dialog.getClosestDialog(this.elementRef.nativeElement);
            this.dialogContainer = dialogRef ? dialogRef.containerInstance : null;
        }
        // change in next microtask avoid throw ExpressionChangedAfterItHasBeenCheckedError
        // because sub component change parent's HostBinding property (ariaLabelledBy)
        Promise.resolve().then(() => {
            if (this.dialogContainer) {
                this.dialogContainer.ariaLabelledBy = this.thyTitle;
            }
        });
    }
    close(event) {
        if (this.thyOnClose.observers.length > 0) {
            this.thyOnClose.emit(event);
        }
        else {
            this.dialog.close();
        }
    }
}
DialogHeaderComponent.ɵfac = function DialogHeaderComponent_Factory(t) { return new (t || DialogHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyDialog), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ThyTranslate), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ThyDialogContainerComponent, 8)); };
DialogHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DialogHeaderComponent, selectors: [["thy-dialog-header"]], contentQueries: function DialogHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerTemplate = _t.first);
    } }, hostVars: 4, hostBindings: function DialogHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dialog-header", ctx.isDialogHeader)("dialog-header-lg", ctx.isDialogHeaderLg);
    } }, inputs: { thySize: "thySize", thyTitleTranslationKey: "thyTitleTranslationKey", thyTitle: "thyTitle", thyIcon: "thyIcon" }, outputs: { thyOnClose: "thyOnClose" }, exportAs: ["thyDialogHeader"], decls: 3, vars: 2, consts: [[4, "ngIf", "ngIfElse"], ["defaultHeader", ""], [3, "ngTemplateOutlet"], [1, "dialog-title"], [3, "thyIconName"], ["type", "button", 1, "close", 3, "click"], ["thyIconName", "close-bold"]], template: function DialogHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, DialogHeaderComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, DialogHeaderComponent_ng_template_1_Template, 5, 2, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.headerTemplate)("ngIfElse", _r1);
    } }, directives: [ɵngcc4.NgIf, ɵngcc4.NgTemplateOutlet, ɵngcc5.ThyIconComponent], encapsulation: 2 });
DialogHeaderComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ThyDialog },
    { type: ThyTranslate },
    { type: ThyDialogContainerComponent, decorators: [{ type: Optional }] }
];
DialogHeaderComponent.propDecorators = {
    isDialogHeader: [{ type: HostBinding, args: [`class.dialog-header`,] }],
    isDialogHeaderLg: [{ type: HostBinding, args: [`class.dialog-header-lg`,] }],
    headerTemplate: [{ type: ContentChild, args: ['dialogHeader',] }],
    thyTitle: [{ type: Input }],
    thySize: [{ type: Input }],
    thyTitleTranslationKey: [{ type: Input }],
    thyIcon: [{ type: Input }],
    thyOnClose: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DialogHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-dialog-header',
                template: "<ng-container *ngIf=\"headerTemplate; else defaultHeader\">\n  <template [ngTemplateOutlet]=\"headerTemplate\"></template>\n</ng-container>\n<ng-template #defaultHeader>\n  <h3 class=\"dialog-title\"><thy-icon [thyIconName]=\"thyIcon\"></thy-icon>{{ thyTitle }}</h3>\n  <button type=\"button\" class=\"close\" (click)=\"close($event)\">\n    <thy-icon thyIconName=\"close-bold\"></thy-icon>\n  </button>\n</ng-template>\n",
                // changeDetection: ChangeDetectionStrategy.OnPush,
                exportAs: 'thyDialogHeader'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.ThyDialog }, { type: ɵngcc2.ThyTranslate }, { type: ɵngcc3.ThyDialogContainerComponent, decorators: [{
                type: Optional
            }] }]; }, { isDialogHeader: [{
            type: HostBinding,
            args: [`class.dialog-header`]
        }], isDialogHeaderLg: [{
            type: HostBinding,
            args: [`class.dialog-header-lg`]
        }], thyOnClose: [{
            type: Output
        }], thySize: [{
            type: Input
        }], thyTitleTranslationKey: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], headerTemplate: [{
            type: ContentChild,
            args: ['dialogHeader']
        }], thyIcon: [{
            type: Input
        }] }); })();
