// ngx-tethys/esm2015/menu/item/name/menu-item-name.component.js
import { Component, HostBinding, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyMenuItemNameComponent {
    constructor() {
        this.isThyMenuItemName = true;
        this.thyOverflowEllipsis = true;
    }
    ngOnInit() { }
}
ThyMenuItemNameComponent.ɵfac = function ThyMenuItemNameComponent_Factory(t) { return new (t || ThyMenuItemNameComponent)(); };
ThyMenuItemNameComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemNameComponent, selectors: [["thy-menu-item-name"], ["", "thy-menu-item-name", ""], ["", "thyMenuItemName", ""]], hostVars: 4, hostBindings: function ThyMenuItemNameComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-name", ctx.isThyMenuItemName)("thy-menu-item-name-ellipsis", ctx.thyOverflowEllipsis);
    } }, inputs: { thyOverflowEllipsis: "thyOverflowEllipsis" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemNameComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuItemNameComponent.ctorParameters = () => [];
ThyMenuItemNameComponent.propDecorators = {
    isThyMenuItemName: [{ type: HostBinding, args: ['class.thy-menu-item-name',] }],
    thyOverflowEllipsis: [{ type: HostBinding, args: ['class.thy-menu-item-name-ellipsis',] }, { type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemNameComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item-name,[thy-menu-item-name],[thyMenuItemName]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return []; }, { isThyMenuItemName: [{
            type: HostBinding,
            args: ['class.thy-menu-item-name']
        }], thyOverflowEllipsis: [{
            type: HostBinding,
            args: ['class.thy-menu-item-name-ellipsis']
        }, {
            type: Input
        }] }); })();
