// ngx-tethys/esm2015/menu/item/action/menu-item-action.component.js
import { Component, HostBinding, ElementRef, Input, Renderer2 } from '@angular/core';
import { ThyPopover } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';

const _c0 = ["*"];
export class ThyMenuItemActionComponent {
    constructor(popover, renderer, elementRef) {
        this.popover = popover;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this._boundEvent = false;
        this.isThyMenuItemIconMore = true;
        this.thyStopPropagation = true;
    }
    set thyActionMenu(value) {
        this._actionMenu = value;
        if (this._actionMenu) {
            this.bindClickEvent();
        }
    }
    bindClickEvent() {
        if (this._boundEvent) {
            return;
        }
        this._boundEvent = true;
        this.renderer.listen(this.elementRef.nativeElement, 'click', event => {
            if (this.thyStopPropagation) {
                event.stopPropagation();
            }
            if (this._actionMenu) {
                this.popover.open(this._actionMenu, {
                    origin: event.currentTarget,
                    insideClosable: true,
                    placement: 'bottom'
                });
            }
        });
    }
    ngOnInit() { }
}
ThyMenuItemActionComponent.ɵfac = function ThyMenuItemActionComponent_Factory(t) { return new (t || ThyMenuItemActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ThyMenuItemActionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuItemActionComponent, selectors: [["thy-menu-item-action"], ["", "thy-menu-item-action", ""], ["", "thyMenuItemAction", ""]], hostVars: 2, hostBindings: function ThyMenuItemActionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-item-action", ctx.isThyMenuItemIconMore);
    } }, inputs: { thyStopPropagation: "thyStopPropagation", thyActionMenu: "thyActionMenu" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyMenuItemActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyMenuItemActionComponent.ctorParameters = () => [
    { type: ThyPopover },
    { type: Renderer2 },
    { type: ElementRef }
];
ThyMenuItemActionComponent.propDecorators = {
    isThyMenuItemIconMore: [{ type: HostBinding, args: ['class.thy-menu-item-action',] }],
    thyActionMenu: [{ type: Input }],
    thyStopPropagation: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuItemActionComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-item-action,[thy-menu-item-action],[thyMenuItemAction]',
                template: "<ng-content></ng-content>"
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { isThyMenuItemIconMore: [{
            type: HostBinding,
            args: ['class.thy-menu-item-action']
        }], thyStopPropagation: [{
            type: Input
        }], thyActionMenu: [{
            type: Input
        }] }); })();
