// ngx-tethys/esm2015/layout/content.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyContentComponent {
    constructor() {
        this._isLayoutContent = true;
    }
}
ThyContentComponent.ɵfac = function ThyContentComponent_Factory(t) { return new (t || ThyContentComponent)(); };
ThyContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentComponent, selectors: [["thy-content"]], hostVars: 2, hostBindings: function ThyContentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content", ctx._isLayoutContent);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyContentComponent.propDecorators = {
    _isLayoutContent: [{ type: HostBinding, args: ['class.thy-layout-content',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentComponent, [{
        type: Component,
        args: [{
                selector: 'thy-content',
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { _isLayoutContent: [{
            type: HostBinding,
            args: ['class.thy-layout-content']
        }] }); })();
