// ngx-tethys/esm2015/layout/content-main.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyContentMainComponent {
    constructor() {
        this.thyLayoutContentClass = true;
    }
}
ThyContentMainComponent.ɵfac = function ThyContentMainComponent_Factory(t) { return new (t || ThyContentMainComponent)(); };
ThyContentMainComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentMainComponent, selectors: [["thy-content-main"]], hostVars: 2, hostBindings: function ThyContentMainComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content-main", ctx.thyLayoutContentClass);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentMainComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyContentMainComponent.propDecorators = {
    thyLayoutContentClass: [{ type: HostBinding, args: ['class.thy-layout-content-main',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentMainComponent, [{
        type: Component,
        args: [{
                selector: 'thy-content-main',
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { thyLayoutContentClass: [{
            type: HostBinding,
            args: ['class.thy-layout-content-main']
        }] }); })();
