// ngx-tethys/esm2015/layout/content-section.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyContentSectionComponent {
    constructor() {
        this.thyLayoutContentSectionClass = true;
    }
}
ThyContentSectionComponent.ɵfac = function ThyContentSectionComponent_Factory(t) { return new (t || ThyContentSectionComponent)(); };
ThyContentSectionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyContentSectionComponent, selectors: [["thy-content-section"]], hostVars: 2, hostBindings: function ThyContentSectionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-content-section", ctx.thyLayoutContentSectionClass);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyContentSectionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyContentSectionComponent.propDecorators = {
    thyLayoutContentSectionClass: [{ type: HostBinding, args: ['class.thy-layout-content-section',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyContentSectionComponent, [{
        type: Component,
        args: [{
                selector: 'thy-content-section',
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
    `
            }]
    }], function () { return []; }, { thyLayoutContentSectionClass: [{
            type: HostBinding,
            args: ['class.thy-layout-content-section']
        }] }); })();
