// ngx-tethys/esm2015/layout/header.component.js
import { Component, HostBinding, Input, TemplateRef, ContentChild } from '@angular/core';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/icon';

const _c0 = ["headerTitle"];
const _c1 = ["headerContent"];
const _c2 = ["headerOperation"];
function ThyHeaderComponent_ng_template_1_Template(rf, ctx) { }
function ThyHeaderComponent_ng_container_2_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 8);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r4.svgIconName);
} }
function ThyHeaderComponent_ng_container_2_ng_template_2_i_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 10);
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r7.iconClass);
} }
function ThyHeaderComponent_ng_container_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThyHeaderComponent_ng_container_2_ng_template_2_i_0_Template, 1, 1, "i", 9);
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.iconClass);
} }
function ThyHeaderComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ThyHeaderComponent_ng_container_2_thy_icon_1_Template, 1, 1, "thy-icon", 5);
    ɵngcc0.ɵɵtemplate(2, ThyHeaderComponent_ng_container_2_ng_template_2_Template, 1, 1, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(4, "span", 7);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r5 = ɵngcc0.ɵɵreference(3);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.svgIconName)("ngIfElse", _r5);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.thyTitle);
} }
function ThyHeaderComponent_ng_template_4_Template(rf, ctx) { }
function ThyHeaderComponent_ng_template_6_Template(rf, ctx) { }
export class ThyHeaderComponent {
    constructor() {
        this.thyLayoutHeaderClass = true;
        this._thyHasBorder = false;
        this._thySizeSm = false;
        this.thyIconPrefix = 'wtf';
    }
    set thyHasBorder(value) {
        this._thyHasBorder = coerceBooleanProperty(value);
    }
    set thySize(value) {
        this._thySizeSm = value === 'sm';
    }
    set thyIcon(value) {
        if (value) {
            if (value.includes('wtf')) {
                this.iconClass = `${this.thyIconPrefix} ${value}`;
            }
            else {
                this.svgIconName = value;
            }
        }
        else {
            this.iconClass = null;
            this.svgIconName = null;
        }
    }
    ngOnInit() { }
}
ThyHeaderComponent.ɵfac = function ThyHeaderComponent_Factory(t) { return new (t || ThyHeaderComponent)(); };
ThyHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyHeaderComponent, selectors: [["thy-header"]], contentQueries: function ThyHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationTemplateRef = _t.first);
    } }, hostVars: 6, hostBindings: function ThyHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout-header", ctx.thyLayoutHeaderClass)("header-has-border", ctx._thyHasBorder)("thy-layout-header-sm", ctx._thySizeSm);
    } }, inputs: { thyIconPrefix: "thyIconPrefix", thyHasBorder: "thyHasBorder", thySize: "thySize", thyIcon: "thyIcon", thyTitle: "thyTitle" }, decls: 7, vars: 4, consts: [[1, "layout-header-title"], [3, "ngTemplateOutlet"], [4, "ngIf"], [1, "layout-header-content"], [1, "layout-header-operation"], ["class", "prefix-icon", 3, "thyIconName", 4, "ngIf", "ngIfElse"], ["iconFont", ""], [1, "title-name"], [1, "prefix-icon", 3, "thyIconName"], ["class", "prefix-icon", 3, "ngClass", 4, "ngIf"], [1, "prefix-icon", 3, "ngClass"]], template: function ThyHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyHeaderComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵtemplate(2, ThyHeaderComponent_ng_container_2_Template, 6, 3, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyHeaderComponent_ng_template_4_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵtemplate(6, ThyHeaderComponent_ng_template_6_Template, 0, 0, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.contentTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.operationTemplateRef);
    } }, directives: [ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf, ɵngcc2.ThyIconComponent, ɵngcc1.NgClass], encapsulation: 2 });
ThyHeaderComponent.ctorParameters = () => [];
ThyHeaderComponent.propDecorators = {
    thyLayoutHeaderClass: [{ type: HostBinding, args: ['class.thy-layout-header',] }],
    _thyHasBorder: [{ type: HostBinding, args: ['class.header-has-border',] }],
    _thySizeSm: [{ type: HostBinding, args: ['class.thy-layout-header-sm',] }],
    thyHasBorder: [{ type: Input, args: ['thyHasBorder',] }],
    thySize: [{ type: Input, args: ['thySize',] }],
    thyTitle: [{ type: Input }],
    thyIconPrefix: [{ type: Input }],
    thyIcon: [{ type: Input, args: ['thyIcon',] }],
    titleTemplateRef: [{ type: ContentChild, args: ['headerTitle',] }],
    contentTemplateRef: [{ type: ContentChild, args: ['headerContent',] }],
    operationTemplateRef: [{ type: ContentChild, args: ['headerOperation',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-header',
                preserveWhitespaces: false,
                template: "<div class=\"layout-header-title\">\n  <ng-template [ngTemplateOutlet]=\"titleTemplateRef\"></ng-template>\n  <ng-container *ngIf=\"!titleTemplateRef\">\n    <thy-icon *ngIf=\"svgIconName; else iconFont\" class=\"prefix-icon\" [thyIconName]=\"svgIconName\"></thy-icon>\n    <ng-template #iconFont>\n      <i *ngIf=\"iconClass\" class=\"prefix-icon\" [ngClass]=\"iconClass\"></i>\n    </ng-template>\n    <span class=\"title-name\">{{ thyTitle }}</span></ng-container\n  >\n</div>\n<div class=\"layout-header-content\">\n  <ng-template [ngTemplateOutlet]=\"contentTemplateRef\"></ng-template>\n</div>\n<div class=\"layout-header-operation\">\n  <ng-template [ngTemplateOutlet]=\"operationTemplateRef\"></ng-template>\n</div>\n"
            }]
    }], function () { return []; }, { thyLayoutHeaderClass: [{
            type: HostBinding,
            args: ['class.thy-layout-header']
        }], _thyHasBorder: [{
            type: HostBinding,
            args: ['class.header-has-border']
        }], _thySizeSm: [{
            type: HostBinding,
            args: ['class.thy-layout-header-sm']
        }], thyIconPrefix: [{
            type: Input
        }], thyHasBorder: [{
            type: Input,
            args: ['thyHasBorder']
        }], thySize: [{
            type: Input,
            args: ['thySize']
        }], thyIcon: [{
            type: Input,
            args: ['thyIcon']
        }], thyTitle: [{
            type: Input
        }], titleTemplateRef: [{
            type: ContentChild,
            args: ['headerTitle']
        }], contentTemplateRef: [{
            type: ContentChild,
            args: ['headerContent']
        }], operationTemplateRef: [{
            type: ContentChild,
            args: ['headerOperation']
        }] }); })();
