// ngx-tethys/esm2015/layout/layout.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyLayoutComponent } from './layout.component';
import { ThyHeaderComponent } from './header.component';
import { ThyContentComponent } from './content.component';
import { ThySidebarComponent } from './sidebar.component';
import { ThyContentSectionComponent } from './content-section.component';
import { ThyContentMainComponent } from './content-main.component';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyLayoutModule {
}
ThyLayoutModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyLayoutModule });
ThyLayoutModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyLayoutModule_Factory(t) { return new (t || ThyLayoutModule)(); }, imports: [[CommonModule, DragDropModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyLayoutModule, { declarations: function () { return [ThyLayoutComponent,
        ThyHeaderComponent,
        ThyContentComponent,
        ThySidebarComponent,
        ThyContentSectionComponent,
        ThyContentMainComponent]; }, imports: function () { return [CommonModule, DragDropModule, ThyIconModule]; }, exports: function () { return [ThyLayoutComponent,
        ThyHeaderComponent,
        ThyContentComponent,
        ThySidebarComponent,
        ThyContentSectionComponent,
        ThyContentMainComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLayoutModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyLayoutComponent,
                    ThyHeaderComponent,
                    ThyContentComponent,
                    ThySidebarComponent,
                    ThyContentSectionComponent,
                    ThyContentMainComponent
                ],
                imports: [CommonModule, DragDropModule, ThyIconModule],
                exports: [
                    ThyLayoutComponent,
                    ThyHeaderComponent,
                    ThyContentComponent,
                    ThySidebarComponent,
                    ThyContentSectionComponent,
                    ThyContentMainComponent
                ]
            }]
    }], null, null); })();
