// ngx-tethys/esm2015/layout/layout.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyLayoutComponent {
    constructor() {
        this.thyLayoutClass = true;
        this.hasSidebar = false;
    }
}
ThyLayoutComponent.ɵfac = function ThyLayoutComponent_Factory(t) { return new (t || ThyLayoutComponent)(); };
ThyLayoutComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyLayoutComponent, selectors: [["thy-layout"]], hostVars: 4, hostBindings: function ThyLayoutComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-layout", ctx.thyLayoutClass)("thy-layout--has-sidebar", ctx.hasSidebar);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyLayoutComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyLayoutComponent.propDecorators = {
    thyLayoutClass: [{ type: HostBinding, args: ['class.thy-layout',] }],
    hasSidebar: [{ type: HostBinding, args: ['class.thy-layout--has-sidebar',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyLayoutComponent, [{
        type: Component,
        args: [{
                selector: 'thy-layout',
                template: `
    <ng-content></ng-content>
  `
            }]
    }], function () { return []; }, { thyLayoutClass: [{
            type: HostBinding,
            args: ['class.thy-layout']
        }], hasSidebar: [{
            type: HostBinding,
            args: ['class.thy-layout--has-sidebar']
        }] }); })();
