// ngx-tethys/esm2015/layout/sidebar.component.js
import { Component, HostBinding, Host, Optional, Input, ViewChild, Renderer2, ElementRef, NgZone } from '@angular/core';
import { ThyLayoutComponent } from './layout.component';
import { coerceBooleanProperty } from 'ngx-tethys/util';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './layout.component';
import * as ɵngcc2 from '@angular/cdk/drag-drop';

const _c0 = ["dragRef"];
const _c1 = ["*"];
const LG_WIDTH = 300;
export class ThySidebarComponent {
    constructor(thyLayoutComponent, renderer, elementRef, ngZone) {
        this.thyLayoutComponent = thyLayoutComponent;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.thyLayoutSidebarClass = true;
        this.thyLayoutSidebarClearBorderRightClass = false;
        this.sidebarIsolated = false;
    }
    set thyWidth(value) {
        if (value === 'lg') {
            value = LG_WIDTH;
        }
        this.thyLayoutSidebarWidth = value;
        this.widthPassive = value;
    }
    set thyHasBorderRight(value) {
        this.thyLayoutSidebarClearBorderRightClass = !coerceBooleanProperty(value);
    }
    set thyIsolated(value) {
        this.sidebarIsolated = coerceBooleanProperty(value);
    }
    set thyIsDraggableWidth(value) {
        if (coerceBooleanProperty(value)) {
            this.thyLayoutSidebarPointerEvents = 'all';
        }
    }
    ngOnInit() {
        if (this.thyLayoutComponent) {
            this.thyLayoutComponent.hasSidebar = true;
        }
        this.ngZone.runOutsideAngular(() => {
            setTimeout(() => {
                this.widthPassive = this.elementRef.nativeElement.clientWidth;
                this.renderer.setStyle(this.dragRef.nativeElement, 'left', this.numberConvertToFloor(this.widthPassive) + 'px');
            }, 0);
        });
    }
    dragStartedHandler() {
        this.dragStartedX = this.dragRef.nativeElement.getBoundingClientRect().x;
    }
    dragEndedHandler() {
        const x = this.dragRef.nativeElement.getBoundingClientRect().x;
        this.widthPassive = this.numberConvertToFloor(this.widthPassive * 1 + (x - this.dragStartedX));
        this.thyLayoutSidebarWidth = this.widthPassive;
    }
    numberConvertToFloor(value) {
        let result = Math.floor(value * 1);
        if (result < 1) {
            result = 1;
        }
        return result;
    }
}
ThySidebarComponent.ɵfac = function ThySidebarComponent_Factory(t) { return new (t || ThySidebarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyLayoutComponent, 9), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ThySidebarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySidebarComponent, selectors: [["thy-sidebar"]], viewQuery: function ThySidebarComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dragRef = _t.first);
    } }, hostVars: 10, hostBindings: function ThySidebarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", ctx.thyLayoutSidebarWidth, "px")("pointer-events", ctx.thyLayoutSidebarPointerEvents);
        ɵngcc0.ɵɵclassProp("thy-layout-sidebar", ctx.thyLayoutSidebarClass)("thy-layout-sidebar--clear-border-right", ctx.thyLayoutSidebarClearBorderRightClass)("thy-layout-sidebar-isolated", ctx.sidebarIsolated);
    } }, inputs: { thyWidth: "thyWidth", thyHasBorderRight: "thyHasBorderRight", thyIsolated: "thyIsolated", thyIsDraggableWidth: "thyIsDraggableWidth" }, ngContentSelectors: _c1, decls: 3, vars: 0, consts: [["cdkDrag", "", "cdkDragLockAxis", "x", "cdkDragBoundary", ".thy-layout", 1, "sidebar-drag", 3, "cdkDragStarted", "cdkDragEnded"], ["dragRef", ""]], template: function ThySidebarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0, 1);
        ɵngcc0.ɵɵlistener("cdkDragStarted", function ThySidebarComponent_Template_div_cdkDragStarted_1_listener() { return ctx.dragStartedHandler(); })("cdkDragEnded", function ThySidebarComponent_Template_div_cdkDragEnded_1_listener() { return ctx.dragEndedHandler(); });
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc2.CdkDrag], encapsulation: 2 });
ThySidebarComponent.ctorParameters = () => [
    { type: ThyLayoutComponent, decorators: [{ type: Optional }, { type: Host }] },
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone }
];
ThySidebarComponent.propDecorators = {
    thyLayoutSidebarClass: [{ type: HostBinding, args: ['class.thy-layout-sidebar',] }],
    thyLayoutSidebarClearBorderRightClass: [{ type: HostBinding, args: ['class.thy-layout-sidebar--clear-border-right',] }],
    thyLayoutSidebarWidth: [{ type: HostBinding, args: ['style.width.px',] }],
    thyLayoutSidebarPointerEvents: [{ type: HostBinding, args: ['style.pointer-events',] }],
    sidebarIsolated: [{ type: HostBinding, args: ['class.thy-layout-sidebar-isolated',] }],
    thyWidth: [{ type: Input, args: ['thyWidth',] }],
    thyHasBorderRight: [{ type: Input, args: ['thyHasBorderRight',] }],
    thyIsolated: [{ type: Input, args: ['thyIsolated',] }],
    thyIsDraggableWidth: [{ type: Input, args: ['thyIsDraggableWidth',] }],
    dragRef: [{ type: ViewChild, args: ['dragRef', { static: true },] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySidebarComponent, [{
        type: Component,
        args: [{
                selector: 'thy-sidebar',
                preserveWhitespaces: false,
                template: `
        <ng-content></ng-content>
        <div
            class="sidebar-drag"
            #dragRef
            cdkDrag
            cdkDragLockAxis="x"
            cdkDragBoundary=".thy-layout"
            (cdkDragStarted)="dragStartedHandler()"
            (cdkDragEnded)="dragEndedHandler()"
        ></div>
    `
            }]
    }], function () { return [{ type: ɵngcc1.ThyLayoutComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }] }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }]; }, { thyLayoutSidebarClass: [{
            type: HostBinding,
            args: ['class.thy-layout-sidebar']
        }], thyLayoutSidebarClearBorderRightClass: [{
            type: HostBinding,
            args: ['class.thy-layout-sidebar--clear-border-right']
        }], sidebarIsolated: [{
            type: HostBinding,
            args: ['class.thy-layout-sidebar-isolated']
        }], thyWidth: [{
            type: Input,
            args: ['thyWidth']
        }], thyLayoutSidebarWidth: [{
            type: HostBinding,
            args: ['style.width.px']
        }], thyHasBorderRight: [{
            type: Input,
            args: ['thyHasBorderRight']
        }], thyIsolated: [{
            type: Input,
            args: ['thyIsolated']
        }], thyIsDraggableWidth: [{
            type: Input,
            args: ['thyIsDraggableWidth']
        }], thyLayoutSidebarPointerEvents: [{
            type: HostBinding,
            args: ['style.pointer-events']
        }], dragRef: [{
            type: ViewChild,
            args: ['dragRef', { static: true }]
        }] }); })();
