// ngx-tethys/esm2015/slider/slider.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { ThySliderComponent } from './slider.component';
import * as ɵngcc0 from '@angular/core';
export class ThySliderModule {
}
ThySliderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThySliderModule });
ThySliderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThySliderModule_Factory(t) { return new (t || ThySliderModule)(); }, providers: [], imports: [[CommonModule, FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThySliderModule, { declarations: function () { return [ThySliderComponent]; }, imports: function () { return [CommonModule, FormsModule]; }, exports: function () { return [ThySliderComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySliderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule],
                exports: [ThySliderComponent],
                declarations: [ThySliderComponent],
                providers: []
            }]
    }], null, null); })();
