// ngx-tethys/esm2015/list/list.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, HostBinding } from '@angular/core';
import { InputBoolean } from 'ngx-tethys/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyListComponent {
    constructor() {
        this._isDivided = false;
        this._isList = true;
    }
    set thyDivided(value) {
        this._isDivided = value;
    }
}
ThyListComponent.ɵfac = function ThyListComponent_Factory(t) { return new (t || ThyListComponent)(); };
ThyListComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyListComponent, selectors: [["thy-list"]], hostVars: 4, hostBindings: function ThyListComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-list-divided", ctx._isDivided)("thy-list", ctx._isList);
    } }, inputs: { thyDivided: "thyDivided" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyListComponent.ctorParameters = () => [];
ThyListComponent.propDecorators = {
    thyDivided: [{ type: Input }],
    _isDivided: [{ type: HostBinding, args: [`class.thy-list-divided`,] }],
    _isList: [{ type: HostBinding, args: [`class.thy-list`,] }]
};
__decorate([
    InputBoolean(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], ThyListComponent.prototype, "thyDivided", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyListComponent, [{
        type: Component,
        args: [{
                selector: 'thy-list',
                template: '<ng-content></ng-content>'
            }]
    }], function () { return []; }, { _isDivided: [{
            type: HostBinding,
            args: [`class.thy-list-divided`]
        }], _isList: [{
            type: HostBinding,
            args: [`class.thy-list`]
        }], thyDivided: [{
            type: Input
        }] }); })();
