// ngx-tethys/esm2015/list/list-item-meta.component.js
import { Component, Input, HostBinding, ChangeDetectionStrategy, TemplateRef, ContentChild } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-tethys/avatar';

const _c0 = ["metaAvatar"];
const _c1 = ["metaTitle"];
const _c2 = ["metaDescription"];
function ThyListItemMetaComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "thy-avatar", 5);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("thySrc", ctx_r0.thyAvatar);
} }
function ThyListItemMetaComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.thyTitle);
} }
function ThyListItemMetaComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.thyDescription);
} }
export class ThyListItemMetaComponent {
    constructor() {
        this.className = `thy-list-item-meta`;
    }
}
ThyListItemMetaComponent.ɵfac = function ThyListItemMetaComponent_Factory(t) { return new (t || ThyListItemMetaComponent)(); };
ThyListItemMetaComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyListItemMetaComponent, selectors: [["thy-list-item-meta"], ["", "thy-list-item-meta", ""]], contentQueries: function ThyListItemMetaComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c2, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.avatarTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.titleTemplateRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.descriptionTemplateRef = _t.first);
    } }, hostVars: 2, hostBindings: function ThyListItemMetaComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.className);
    } }, inputs: { thyAvatar: "thyAvatar", thyTitle: "thyTitle", thyDescription: "thyDescription" }, decls: 7, vars: 6, consts: [[1, "thy-list-item-meta-avatar"], [4, "ngIf", "ngIfElse"], [1, "thy-list-item-meta-content"], [1, "thy-list-item-meta-title"], [1, "thy-list-item-meta-description"], [3, "thySrc"]], template: function ThyListItemMetaComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ThyListItemMetaComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ThyListItemMetaComponent_ng_container_4_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵtemplate(6, ThyListItemMetaComponent_ng_container_6_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyAvatar)("ngIfElse", ctx.avatarTemplateRef);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyTitle)("ngIfElse", ctx.titleTemplateRef);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyDescription)("ngIfElse", ctx.descriptionTemplateRef);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ThyAvatarComponent], encapsulation: 2, changeDetection: 0 });
ThyListItemMetaComponent.ctorParameters = () => [];
ThyListItemMetaComponent.propDecorators = {
    thyAvatar: [{ type: Input }],
    thyTitle: [{ type: Input }],
    thyDescription: [{ type: Input }],
    avatarTemplateRef: [{ type: ContentChild, args: ['metaAvatar',] }],
    titleTemplateRef: [{ type: ContentChild, args: ['metaTitle',] }],
    descriptionTemplateRef: [{ type: ContentChild, args: ['metaDescription',] }],
    className: [{ type: HostBinding, args: ['class',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyListItemMetaComponent, [{
        type: Component,
        args: [{
                selector: 'thy-list-item-meta,[thy-list-item-meta]',
                template: "<div class=\"thy-list-item-meta-avatar\">\n  <ng-container *ngIf=\"thyAvatar; else avatarTemplateRef\">\n    <thy-avatar [thySrc]=\"thyAvatar\"></thy-avatar>\n  </ng-container>\n</div>\n<div class=\"thy-list-item-meta-content\">\n  <div class=\"thy-list-item-meta-title\">\n    <ng-container *ngIf=\"thyTitle; else titleTemplateRef\">{{ thyTitle }}</ng-container>\n  </div>\n  <div class=\"thy-list-item-meta-description\">\n    <ng-container *ngIf=\"thyDescription; else descriptionTemplateRef\">{{ thyDescription }}</ng-container>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, { className: [{
            type: HostBinding,
            args: ['class']
        }], thyAvatar: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], thyDescription: [{
            type: Input
        }], avatarTemplateRef: [{
            type: ContentChild,
            args: ['metaAvatar']
        }], titleTemplateRef: [{
            type: ContentChild,
            args: ['metaTitle']
        }], descriptionTemplateRef: [{
            type: ContentChild,
            args: ['metaDescription']
        }] }); })();
