// ngx-tethys/esm2015/list/list.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyListComponent } from './list.component';
import { ThyListItemComponent } from './list-item.component';
import { ThySelectionListComponent } from './selection/selection-list';
// import { ThyListOptionComponent } from './option/option';
import { ThyOptionModule } from 'ngx-tethys/shared';
import { ThyListItemMetaComponent } from './list-item-meta.component';
import { ThyAvatarModule } from 'ngx-tethys/avatar';
import * as ɵngcc0 from '@angular/core';
export class ThyListModule {
}
ThyListModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyListModule });
ThyListModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyListModule_Factory(t) { return new (t || ThyListModule)(); }, imports: [[CommonModule, ThyOptionModule, ThyAvatarModule], ThyOptionModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyListModule, { declarations: function () { return [ThyListComponent, ThyListItemComponent, ThySelectionListComponent, ThyListItemMetaComponent]; }, imports: function () { return [CommonModule, ThyOptionModule, ThyAvatarModule]; }, exports: function () { return [ThyListComponent, ThyListItemComponent, ThySelectionListComponent, ThyListItemMetaComponent, ThyOptionModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyListModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyListComponent, ThyListItemComponent, ThySelectionListComponent, ThyListItemMetaComponent],
                imports: [CommonModule, ThyOptionModule, ThyAvatarModule],
                exports: [ThyListComponent, ThyListItemComponent, ThySelectionListComponent, ThyListItemMetaComponent, ThyOptionModule]
            }]
    }], null, null); })();
