// ngx-tethys/esm2015/list/list-item.component.js
import { Component, HostBinding } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ThyListItemComponent {
    constructor() {
        this._isListItem = true;
    }
}
ThyListItemComponent.ɵfac = function ThyListItemComponent_Factory(t) { return new (t || ThyListItemComponent)(); };
ThyListItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyListItemComponent, selectors: [["thy-list-item"], ["", "thy-list-item", ""]], hostVars: 2, hostBindings: function ThyListItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-list-item", ctx._isListItem);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ThyListItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyListItemComponent.ctorParameters = () => [];
ThyListItemComponent.propDecorators = {
    _isListItem: [{ type: HostBinding, args: [`class.thy-list-item`,] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyListItemComponent, [{
        type: Component,
        args: [{
                selector: 'thy-list-item,[thy-list-item]',
                template: '<ng-content></ng-content>'
            }]
    }], function () { return []; }, { _isListItem: [{
            type: HostBinding,
            args: [`class.thy-list-item`]
        }] }); })();
