// ngx-tethys/esm2015/pagination/pagination.pipe.js
import { Pipe } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/platform-browser';
export class PaginationTotalCountFormat {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    transform(count, format) {
        if (count && format) {
            return this.sanitizer.bypassSecurityTrustHtml(format.replace(/\{(.+?)\}/g, `<span>${count}</span>`));
        }
        else {
            return '';
        }
    }
}
PaginationTotalCountFormat.ɵfac = function PaginationTotalCountFormat_Factory(t) { return new (t || PaginationTotalCountFormat)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DomSanitizer)); };
PaginationTotalCountFormat.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "paginationTotalPagesFormat", type: PaginationTotalCountFormat, pure: true });
PaginationTotalCountFormat.ctorParameters = () => [
    { type: DomSanitizer }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PaginationTotalCountFormat, [{
        type: Pipe,
        args: [{
                name: 'paginationTotalPagesFormat'
            }]
    }], function () { return [{ type: ɵngcc1.DomSanitizer }]; }, null); })();
