// ngx-tethys/esm2015/pagination/pagination.module.js
import { NgModule } from '@angular/core';
import { ThyPaginationComponent } from './pagination.component';
import { CommonModule } from '@angular/common';
import { PaginationTotalCountFormat } from './pagination.pipe';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyIconModule } from 'ngx-tethys/icon';
import * as ɵngcc0 from '@angular/core';
export class ThyPaginationModule {
}
ThyPaginationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyPaginationModule });
ThyPaginationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyPaginationModule_Factory(t) { return new (t || ThyPaginationModule)(); }, imports: [[CommonModule, ThySharedModule, ThyIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyPaginationModule, { declarations: function () { return [ThyPaginationComponent, PaginationTotalCountFormat]; }, imports: function () { return [CommonModule, ThySharedModule, ThyIconModule]; }, exports: function () { return [ThyPaginationComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyPaginationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ThySharedModule, ThyIconModule],
                exports: [ThyPaginationComponent],
                declarations: [ThyPaginationComponent, PaginationTotalCountFormat]
            }]
    }], null, null); })();
