// ngx-tethys/esm2015/badge/badge.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyBadgeComponent } from './badge.component';
import * as ɵngcc0 from '@angular/core';
export class ThyBadgeModule {
    constructor() { }
}
ThyBadgeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyBadgeModule });
ThyBadgeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyBadgeModule_Factory(t) { return new (t || ThyBadgeModule)(); }, imports: [[
            CommonModule
        ]] });
ThyBadgeModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyBadgeModule, { declarations: function () { return [ThyBadgeComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ThyBadgeComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyBadgeModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ThyBadgeComponent
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ThyBadgeComponent
                ]
            }]
    }], function () { return []; }, null); })();
