// ngx-tethys/esm2015/slide/slide-header/slide-header.component.js
import { Component, ContentChild, TemplateRef, Input, HostBinding } from '@angular/core';
import { ThySlideService } from '../slide.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../slide.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-tethys/icon';

const _c0 = ["thyHeader"];
const _c1 = ["thyHeaderOperate"];
function ThySlideHeaderComponent_1_ng_template_0_Template(rf, ctx) { }
function ThySlideHeaderComponent_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThySlideHeaderComponent_1_ng_template_0_Template, 0, 0, "ng-template", 2);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r0.headerTemplate);
} }
function ThySlideHeaderComponent_ng_template_2_thy_icon_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 10);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r4.thyIcon);
} }
function ThySlideHeaderComponent_ng_template_2_ng_template_2_i_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i");
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassMapInterpolate1("wft ", ctx_r8.thyIcon, "");
} }
function ThySlideHeaderComponent_ng_template_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThySlideHeaderComponent_ng_template_2_ng_template_2_i_0_Template, 1, 3, "i", 11);
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.thyIcon);
} }
function ThySlideHeaderComponent_ng_template_2_7_ng_template_0_Template(rf, ctx) { }
function ThySlideHeaderComponent_ng_template_2_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ThySlideHeaderComponent_ng_template_2_7_ng_template_0_Template, 0, 0, "ng-template", 2);
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r7.headerOperateTemplate);
} }
function ThySlideHeaderComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "h3", 3);
    ɵngcc0.ɵɵtemplate(1, ThySlideHeaderComponent_ng_template_2_thy_icon_1_Template, 1, 1, "thy-icon", 4);
    ɵngcc0.ɵɵtemplate(2, ThySlideHeaderComponent_ng_template_2_ng_template_2_Template, 1, 1, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 6);
    ɵngcc0.ɵɵelementContainerStart(6);
    ɵngcc0.ɵɵtemplate(7, ThySlideHeaderComponent_ng_template_2_7_Template, 1, 1, undefined, 7);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "button", 8);
    ɵngcc0.ɵɵlistener("click", function ThySlideHeaderComponent_ng_template_2_Template_button_click_8_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.closeModal($event); });
    ɵngcc0.ɵɵelement(9, "thy-icon", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r5 = ɵngcc0.ɵɵreference(3);
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.isIconFont && ctx_r2.thyIcon)("ngIfElse", _r5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.thyTitle, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.headerOperateTemplate);
} }
export class ThySlideHeaderComponent {
    constructor(thySlideService) {
        this.thySlideService = thySlideService;
        this.isIconFont = false;
        this._iconName = '';
        this.slideLayoutHeader = true;
    }
    set thyIcon(value) {
        this._iconName = value;
        if (value.includes('wtf')) {
            this.isIconFont = true;
        }
        else {
            this.isIconFont = false;
        }
    }
    get thyIcon() {
        return this._iconName;
    }
    ngOnInit() { }
    closeModal(event) {
        event.stopPropagation();
        this.thySlideService.close();
    }
}
ThySlideHeaderComponent.ɵfac = function ThySlideHeaderComponent_Factory(t) { return new (t || ThySlideHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThySlideService)); };
ThySlideHeaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThySlideHeaderComponent, selectors: [["thy-slide-header"]], contentQueries: function ThySlideHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c1, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerOperateTemplate = _t.first);
    } }, hostVars: 2, hostBindings: function ThySlideHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-slide-header", ctx.slideLayoutHeader);
    } }, inputs: { thyIcon: "thyIcon", thyTitle: "thyTitle" }, decls: 4, vars: 2, consts: [[4, "ngIf", "ngIfElse"], ["default", ""], [3, "ngTemplateOutlet"], [1, "thy-slide-header-title"], [3, "thyIconName", 4, "ngIf", "ngIfElse"], ["iconFont", ""], [1, "thy-slide-header-main"], [4, "ngIf"], ["type", "button", 1, "close", 3, "click"], ["thyIconName", "close"], [3, "thyIconName"], [3, "class", 4, "ngIf"]], template: function ThySlideHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0);
        ɵngcc0.ɵɵtemplate(1, ThySlideHeaderComponent_1_Template, 1, 1, undefined, 0);
        ɵngcc0.ɵɵtemplate(2, ThySlideHeaderComponent_ng_template_2_Template, 10, 4, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementContainerEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(3);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.headerTemplate)("ngIfElse", _r1);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgTemplateOutlet, ɵngcc3.ThyIconComponent], encapsulation: 2 });
ThySlideHeaderComponent.ctorParameters = () => [
    { type: ThySlideService }
];
ThySlideHeaderComponent.propDecorators = {
    slideLayoutHeader: [{ type: HostBinding, args: ['class.thy-slide-header',] }],
    thyTitle: [{ type: Input }],
    thyIcon: [{ type: Input }],
    headerTemplate: [{ type: ContentChild, args: ['thyHeader',] }],
    headerOperateTemplate: [{ type: ContentChild, args: ['thyHeaderOperate',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThySlideHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'thy-slide-header',
                template: "<ng-container>\n  <ng-template *ngIf=\"headerTemplate; else default\" [ngTemplateOutlet]=\"headerTemplate\"></ng-template>\n  <ng-template #default>\n    <h3 class=\"thy-slide-header-title\">\n      <thy-icon *ngIf=\"!isIconFont && thyIcon; else iconFont\" [thyIconName]=\"thyIcon\"></thy-icon>\n      <ng-template #iconFont>\n        <i *ngIf=\"thyIcon\" class=\"wft {{ thyIcon }}\"></i>\n      </ng-template>\n      {{ thyTitle }}\n    </h3>\n\n    <div class=\"thy-slide-header-main\">\n      <ng-container>\n        <ng-template *ngIf=\"headerOperateTemplate\" [ngTemplateOutlet]=\"headerOperateTemplate\"> </ng-template>\n      </ng-container>\n    </div>\n    <button type=\"button\" class=\"close\" (click)=\"closeModal($event)\">\n      <thy-icon thyIconName=\"close\"></thy-icon>\n    </button>\n  </ng-template>\n</ng-container>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ThySlideService }]; }, { slideLayoutHeader: [{
            type: HostBinding,
            args: ['class.thy-slide-header']
        }], thyIcon: [{
            type: Input
        }], thyTitle: [{
            type: Input
        }], headerTemplate: [{
            type: ContentChild,
            args: ['thyHeader']
        }], headerOperateTemplate: [{
            type: ContentChild,
            args: ['thyHeaderOperate']
        }] }); })();
