// ngx-tethys/esm2015/core/overlay/abstract-overlay-container.js
import { EventEmitter } from '@angular/core';
import { Subject } from 'rxjs';
export function throwPopoverContentAlreadyAttachedError(name) {
    throw Error(`Attempting to attach ${name} content after content is already attached`);
}
export class ThyAbstractOverlayContainer {
    constructor(options, changeDetectorRef) {
        this.options = options;
        this.changeDetectorRef = changeDetectorRef;
        this.animationStateChanged = new EventEmitter();
        this.containerDestroy = new Subject();
    }
    /** Before detach content*/
    beforeDetachPortal() { }
    /**
     * Attach a TemplatePortal as content to this overlay container.
     * @param portal Portal to be attached as the overlay content.
     */
    attachTemplatePortal(portal) {
        if (this.portalOutlet.hasAttached()) {
            throwPopoverContentAlreadyAttachedError(this.options.name);
        }
        this.beforeAttachPortal();
        return this.portalOutlet.attachTemplatePortal(portal);
    }
    /**
     * Attach a ComponentPortal as content to this overlay container.
     * @param portal Portal to be attached as the overlay content.
     */
    attachComponentPortal(portal) {
        if (this.portalOutlet.hasAttached()) {
            throwPopoverContentAlreadyAttachedError(this.options.name);
        }
        this.beforeAttachPortal();
        return this.portalOutlet.attachComponentPortal(portal);
    }
    startExitAnimation() {
        if (this.options.animationEnabled) {
            this.animationState = 'exit';
        }
        else {
            // this.animationClosingDone.
        }
        this.beforeDetachPortal();
        // Mark the container for check so it can react if the
        // view container is using OnPush change detection.
        this.changeDetectorRef.markForCheck();
    }
    destroy() {
        this.containerDestroy.next();
    }
}
