// ngx-tethys/esm2015/tree/tree-replace-region.component.js
import { Component, ViewEncapsulation, ViewContainerRef, Input } from '@angular/core';
import { ThyTreeComponent } from './tree.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './tree.component';

const _c0 = ["ThyTreeReplaceRegion", ""];
const _c1 = ["*"];
export class ThyTreeReplaceRegionComponent {
    constructor(root, viewRef) {
        this.root = root;
        this.viewRef = viewRef;
    }
    ngOnInit() { }
}
ThyTreeReplaceRegionComponent.ɵfac = function ThyTreeReplaceRegionComponent_Factory(t) { return new (t || ThyTreeReplaceRegionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyTreeComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ThyTreeReplaceRegionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyTreeReplaceRegionComponent, selectors: [["", "ThyTreeReplaceRegion", ""]], inputs: { thyInstance: "thyInstance" }, attrs: _c0, ngContentSelectors: _c1, decls: 1, vars: 0, template: function ThyTreeReplaceRegionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ThyTreeReplaceRegionComponent.ctorParameters = () => [
    { type: ThyTreeComponent },
    { type: ViewContainerRef }
];
ThyTreeReplaceRegionComponent.propDecorators = {
    thyInstance: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeReplaceRegionComponent, [{
        type: Component,
        args: [{
                selector: '[ThyTreeReplaceRegion]',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc1.ThyTreeComponent }, { type: ɵngcc0.ViewContainerRef }]; }, { thyInstance: [{
            type: Input
        }] }); })();
