// ngx-tethys/esm2015/tree/tree.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyTreeComponent } from './tree.component';
import { ThyTreeNodeComponent } from './tree-node.component';
import { ThyTreeReplaceRegionComponent } from './tree-replace-region.component';
import { ThyInputModule } from 'ngx-tethys/input';
import { ThyButtonModule } from 'ngx-tethys/button';
import { ThySharedModule } from 'ngx-tethys/shared';
import { ThyTreeService } from './tree.service';
import { ThyListModule } from 'ngx-tethys/list';
import { ThyOptionModule } from 'ngx-tethys/shared';
import { FormsModule } from '@angular/forms';
import { ThyLoadingModule } from 'ngx-tethys/loading';
import { ThyIconModule } from 'ngx-tethys/icon';
import { ThyDragDropModule } from 'ngx-tethys/drag-drop';
import { ThyCheckboxModule } from 'ngx-tethys/checkbox';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/drag-drop';
import * as ɵngcc2 from '@angular/common';
export class ThyTreeModule {
}
ThyTreeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyTreeModule });
ThyTreeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyTreeModule_Factory(t) { return new (t || ThyTreeModule)(); }, providers: [ThyTreeService], imports: [[
            CommonModule,
            ThyInputModule,
            ThyButtonModule,
            ThySharedModule,
            FormsModule,
            ThyListModule,
            ThyOptionModule,
            ThyLoadingModule,
            ThyIconModule,
            ThyDragDropModule,
            ThyCheckboxModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyTreeModule, { declarations: function () { return [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent]; }, imports: function () { return [CommonModule,
        ThyInputModule,
        ThyButtonModule,
        ThySharedModule,
        FormsModule,
        ThyListModule,
        ThyOptionModule,
        ThyLoadingModule,
        ThyIconModule,
        ThyDragDropModule,
        ThyCheckboxModule]; }, exports: function () { return [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyTreeModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent],
                imports: [
                    CommonModule,
                    ThyInputModule,
                    ThyButtonModule,
                    ThySharedModule,
                    FormsModule,
                    ThyListModule,
                    ThyOptionModule,
                    ThyLoadingModule,
                    ThyIconModule,
                    ThyDragDropModule,
                    ThyCheckboxModule
                ],
                entryComponents: [ThyTreeComponent],
                exports: [ThyTreeComponent, ThyTreeNodeComponent, ThyTreeReplaceRegionComponent],
                providers: [ThyTreeService]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(ThyTreeComponent, [ɵngcc1.ThyDropContainerDirective, ɵngcc2.NgForOf, ThyTreeNodeComponent, ɵngcc1.ThyDragDirective], []);
