// ngx-tethys/esm2015/menu/group/menu-group.component.js
import { Component, HostBinding, Input, Output, EventEmitter, ElementRef, ViewChild } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { ThyPopover } from 'ngx-tethys/popover';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-tethys/popover';
import * as ɵngcc2 from 'ngx-tethys/icon';
import * as ɵngcc3 from '@angular/common';

const _c0 = ["thyMenuGroup"];
function ThyMenuGroupComponent_thy_icon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "thy-icon", 7);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r0.iconClass);
} }
function ThyMenuGroupComponent_thy_icon_5_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "thy-icon", 8);
    ɵngcc0.ɵɵlistener("click", function ThyMenuGroupComponent_thy_icon_5_Template_thy_icon_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onActionClick($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("thyIconName", ctx_r1.rightIconClass);
} }
const _c1 = ["*"];
export class ThyMenuGroupComponent {
    constructor(popover) {
        this.popover = popover;
        this.rightIconClass = 'more';
        this.iconClass = 'folder-bold';
        this.groupHeaderPaddingLeft = 0;
        this.isThyMenuGroup = true;
        this.showIcon = false;
        this.isCollapsed = true;
        this.thyTitle = '';
        this.thyShowAction = false;
        this.thyActionStopPropagation = true;
        this.thyOnActionClick = new EventEmitter();
    }
    set thyExpand(value) {
        this.isCollapsed = !!!value;
    }
    set thyShowIcon(value) {
        this.showIcon = value;
    }
    set thyIcon(value) {
        this.iconClass = value;
    }
    set thyActionIcon(value) {
        this.rightIconClass = value;
    }
    set thyActionMenu(value) {
        this._actionMenu = value;
    }
    ngOnInit() { }
    collapseGroup() {
        this.isCollapsed = !this.isCollapsed;
    }
    onActionClick(event) {
        if (this.thyActionStopPropagation) {
            event.stopPropagation();
        }
        if (this._actionMenu) {
            this.popover.open(this._actionMenu, {
                origin: event.currentTarget,
                insideClosable: true,
                placement: 'bottom'
            });
        }
        else {
            this.thyOnActionClick.emit(event);
        }
    }
}
ThyMenuGroupComponent.ɵfac = function ThyMenuGroupComponent_Factory(t) { return new (t || ThyMenuGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ThyPopover)); };
ThyMenuGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ThyMenuGroupComponent, selectors: [["thy-menu-group"], ["", "thy-menu-group", ""], ["", "thyMenuGroup", ""]], viewQuery: function ThyMenuGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._thyMenuGroup = _t.first);
    } }, hostVars: 6, hostBindings: function ThyMenuGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("thy-menu-group", ctx.isThyMenuGroup)("has-icon", ctx.showIcon)("collapsed", ctx.isCollapsed);
    } }, inputs: { thyTitle: "thyTitle", thyShowAction: "thyShowAction", thyActionStopPropagation: "thyActionStopPropagation", thyExpand: "thyExpand", thyShowIcon: "thyShowIcon", thyIcon: "thyIcon", thyActionIcon: "thyActionIcon", thyActionMenu: "thyActionMenu" }, outputs: { thyOnActionClick: "thyOnActionClick" }, ngContentSelectors: _c1, decls: 9, vars: 5, consts: [[1, "thy-menu-group-header", 3, "click"], ["thyIconName", "caret-right-down", 1, "thy-menu-group-arrow"], ["class", "thy-menu-group-folder", 3, "thyIconName", 4, "ngIf"], [1, "thy-menu-group-title", 3, "title"], ["class", "thy-menu-group-more", 3, "thyIconName", "click", 4, "ngIf"], [1, "thy-menu-group-body"], ["thyMenuGroup", ""], [1, "thy-menu-group-folder", 3, "thyIconName"], [1, "thy-menu-group-more", 3, "thyIconName", "click"]], template: function ThyMenuGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("click", function ThyMenuGroupComponent_Template_div_click_0_listener() { return ctx.collapseGroup(); });
        ɵngcc0.ɵɵelement(1, "thy-icon", 1);
        ɵngcc0.ɵɵtemplate(2, ThyMenuGroupComponent_thy_icon_2_Template, 1, 1, "thy-icon", 2);
        ɵngcc0.ɵɵelementStart(3, "span", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ThyMenuGroupComponent_thy_icon_5_Template, 1, 1, "thy-icon", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5, 6);
        ɵngcc0.ɵɵprojection(8);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("title", ctx.thyTitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.thyTitle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.thyShowAction);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("@detailsContentAnimation", ctx.isCollapsed);
    } }, directives: [ɵngcc2.ThyIconComponent, ɵngcc3.NgIf], encapsulation: 2, data: { animation: [
            trigger('detailsContentAnimation', [
                state('void', style({
                    height: '*'
                })),
                state('1', style({
                    height: 0,
                    overflow: 'hidden'
                })),
                state('0', style({
                    height: '*'
                })),
                transition('* => *', animate('0ms ease-out'))
            ])
        ] } });
ThyMenuGroupComponent.ctorParameters = () => [
    { type: ThyPopover }
];
ThyMenuGroupComponent.propDecorators = {
    _thyMenuGroup: [{ type: ViewChild, args: ['thyMenuGroup', { static: true },] }],
    isThyMenuGroup: [{ type: HostBinding, args: ['class.thy-menu-group',] }],
    showIcon: [{ type: HostBinding, args: ['class.has-icon',] }],
    isCollapsed: [{ type: HostBinding, args: ['class.collapsed',] }],
    thyTitle: [{ type: Input }],
    thyExpand: [{ type: Input, args: ['thyExpand',] }],
    thyShowIcon: [{ type: Input, args: ['thyShowIcon',] }],
    thyIcon: [{ type: Input, args: ['thyIcon',] }],
    thyActionIcon: [{ type: Input, args: ['thyActionIcon',] }],
    thyShowAction: [{ type: Input }],
    thyActionStopPropagation: [{ type: Input }],
    thyOnActionClick: [{ type: Output }],
    thyActionMenu: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyMenuGroupComponent, [{
        type: Component,
        args: [{
                selector: 'thy-menu-group, [thy-menu-group],[thyMenuGroup]',
                template: "<div class=\"thy-menu-group-header\" (click)=\"collapseGroup()\">\n  <thy-icon thyIconName=\"caret-right-down\" class=\"thy-menu-group-arrow\"></thy-icon>\n  <thy-icon *ngIf=\"showIcon\" class=\"thy-menu-group-folder\" [thyIconName]=\"iconClass\"></thy-icon>\n  <span class=\"thy-menu-group-title\" [title]=\"thyTitle\">{{ thyTitle }}</span>\n  <thy-icon\n    *ngIf=\"thyShowAction\"\n    class=\"thy-menu-group-more\"\n    [thyIconName]=\"rightIconClass\"\n    (click)=\"onActionClick($event)\"\n  ></thy-icon>\n</div>\n<div class=\"thy-menu-group-body\" #thyMenuGroup [@detailsContentAnimation]=\"isCollapsed\">\n  <ng-content></ng-content>\n</div>\n",
                animations: [
                    trigger('detailsContentAnimation', [
                        state('void', style({
                            height: '*'
                        })),
                        state('1', style({
                            height: 0,
                            overflow: 'hidden'
                        })),
                        state('0', style({
                            height: '*'
                        })),
                        transition('* => *', animate('0ms ease-out'))
                    ])
                ]
            }]
    }], function () { return [{ type: ɵngcc1.ThyPopover }]; }, { isThyMenuGroup: [{
            type: HostBinding,
            args: ['class.thy-menu-group']
        }], showIcon: [{
            type: HostBinding,
            args: ['class.has-icon']
        }], isCollapsed: [{
            type: HostBinding,
            args: ['class.collapsed']
        }], thyTitle: [{
            type: Input
        }], thyShowAction: [{
            type: Input
        }], thyActionStopPropagation: [{
            type: Input
        }], thyOnActionClick: [{
            type: Output
        }], thyExpand: [{
            type: Input,
            args: ['thyExpand']
        }], thyShowIcon: [{
            type: Input,
            args: ['thyShowIcon']
        }], thyIcon: [{
            type: Input,
            args: ['thyIcon']
        }], thyActionIcon: [{
            type: Input,
            args: ['thyActionIcon']
        }], thyActionMenu: [{
            type: Input
        }], _thyMenuGroup: [{
            type: ViewChild,
            args: ['thyMenuGroup', { static: true }]
        }] }); })();
