// ngx-tethys/esm2015/flexible-text/flexible-text.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThyFlexibleTextComponent } from './flexible-text.component';
import { ThyTooltipModule } from 'ngx-tethys/tooltip';
import { ObserversModule } from '@angular/cdk/observers';
import * as ɵngcc0 from '@angular/core';
export class ThyFlexibleTextModule {
}
ThyFlexibleTextModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThyFlexibleTextModule });
ThyFlexibleTextModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThyFlexibleTextModule_Factory(t) { return new (t || ThyFlexibleTextModule)(); }, imports: [[CommonModule, ThyTooltipModule, ObserversModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThyFlexibleTextModule, { declarations: function () { return [ThyFlexibleTextComponent]; }, imports: function () { return [CommonModule, ThyTooltipModule, ObserversModule]; }, exports: function () { return [ThyFlexibleTextComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThyFlexibleTextModule, [{
        type: NgModule,
        args: [{
                declarations: [ThyFlexibleTextComponent],
                imports: [CommonModule, ThyTooltipModule, ObserversModule],
                exports: [ThyFlexibleTextComponent]
            }]
    }], null, null); })();
